{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "265e8129",
   "metadata": {},
   "source": [
    "This notebook uses iFeatureOmega, a feature generation software, to add to our feature space for a RandomForestClassifier that predicts protein pair functionality."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4afbc58f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats\n",
    "import sklearn.preprocessing\n",
    "import sklearn.model_selection\n",
    "import sklearn.ensemble\n",
    "import sklearn.feature_selection\n",
    "import unittest\n",
    "import iFeatureOmegaCLI\n",
    "import Bio.SeqIO\n",
    "import Bio.SeqRecord\n",
    "import io\n",
    "from io import StringIO\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ba87d47f",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/loganroberts/Learn2Therm/ValidProt/notebooks\n"
     ]
    }
   ],
   "source": [
    "cd /Users/loganroberts/Learn2Therm/ValidProt/notebooks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f90c1cdd",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Unnamed: 0', 'local_gap_compressed_percent_id',\n",
       "       'scaled_local_query_percent_id', 'scaled_local_symmetric_percent_id',\n",
       "       'query_align_len', 'query_align_cov', 'subject_align_len',\n",
       "       'subject_align_cov', 'bit_score', 'thermo_index', 'meso_index',\n",
       "       'prot_pair_index', 'meso_protein_int_index', 'thermo_protein_int_index',\n",
       "       'taxa_pair_index', 'local_gap_compressed_percent_id_16s',\n",
       "       'scaled_local_query_percent_id_16s',\n",
       "       'scaled_local_symmetric_percent_id_16s', 'query_align_cov_16s',\n",
       "       'subject_align_cov_16s', 'bit_score_16s', 'm_ogt', 't_ogt',\n",
       "       'ogt_difference', 'm_protein_seq', 't_protein_seq', 'm_protein_desc',\n",
       "       't_protein_desc', 'm_protein_len', 't_protein_len'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#convert to pandas df\n",
    "df = pd.read_csv('learn2therm_sample_50k_exploration.csv')\n",
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "99990cd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_fasta_from_dataframe(dataframe, output_file_a, output_file_b):\n",
    "    #adjust this to write function with BioPython\n",
    "    #separate functions for each of the input sequences\n",
    "    #in training, seq_a = meso and seq_b = thermo\n",
    "    \n",
    "    \n",
    "    #meso sequence to fasta\n",
    "    with open(output_file_a, 'w') as f:\n",
    "        for _, row in df.iterrows():\n",
    "            f.write('>{}\\n{}\\n'.format((row['prot_pair_index']), row['m_protein_seq']))\n",
    "    \n",
    "    #thermo sequence to fasta\n",
    "    with open(output_file_b, 'w') as f:\n",
    "        for _, row in df.iterrows():\n",
    "            f.write('>{}\\n{}\\n'.format((row['prot_pair_index']), (row['t_protein_seq'])))\n",
    "   \n",
    "    #return output files\n",
    "    return [output_file_a, output_file_b]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "653eb96d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_protein_descriptors(fasta_file, descriptors=[]):\n",
    "    \n",
    "    \"\"\"\n",
    "    Generates features from a protein sequence\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    Fasta file with amino acid sequences.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    Vector of descriptors\n",
    "    \"\"\"\n",
    "    \n",
    "    #create iProtein object\n",
    "    protein = iFeatureOmegaCLI.iProtein(fasta_file)\n",
    "    \n",
    "    #not sure why we need this yet. Right now it is stored in local directory.\n",
    "    params = protein.import_parameters('protein_parameters.json')\n",
    "    \n",
    "    protein_descriptors = {}\n",
    "    \n",
    "    for descriptor in descriptors:\n",
    "        protein.get_descriptor(descriptor)\n",
    "        protein_descriptors.update({f'{descriptor}':protein.encodings})\n",
    "        \n",
    "    return protein_descriptors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fe48f193",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_new_dataframe(dataframe, output_files, descriptors=[]):\n",
    "    \"\"\"\n",
    "    Creates new dataframe with descriptors added.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    Pandas dataframe, list of descriptors as strings, output file name.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    Dataframe including vector(s) of descriptors\n",
    "    \"\"\"\n",
    "\n",
    "    fasta_files = get_fasta_from_dataframe(dataframe, output_files[0], output_files[1])\n",
    "    \n",
    "    \n",
    "    feature_dict_a = get_protein_descriptors(fasta_files[0], descriptors)\n",
    "    \n",
    "    feature_dict_b = get_protein_descriptors(fasta_files[1], descriptors)\n",
    "\n",
    "\n",
    "    df = dataframe.reset_index()\n",
    "\n",
    "    for desc in descriptors:\n",
    "\n",
    "        feature_dict_a[desc].index = feature_dict_a[desc].index.astype(int)\n",
    "        features_a = feature_dict_a[desc].reset_index()\n",
    "        \n",
    "        feature_dict_b[desc].index = feature_dict_b[desc].index.astype(int)\n",
    "        features_b = feature_dict_b[desc].reset_index()\n",
    "    \n",
    "    feature_df = pd.merge(\n",
    "            df,\n",
    "            features_a,\n",
    "            how='outer',\n",
    "            left_index=True,\n",
    "            right_index=True)\n",
    "    \n",
    "    new_df = pd.merge(\n",
    "            feature_df,\n",
    "            features_b,\n",
    "            how='outer',\n",
    "            left_index=True,\n",
    "            right_index=True)\n",
    "\n",
    "    return new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d7f0a563",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/loganroberts/Learn2Therm/ValidProt/notebooks\n"
     ]
    }
   ],
   "source": [
    "cd /Users/loganroberts/Learn2Therm/ValidProt/notebooks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "043bf4dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>prot_pair_index</th>\n",
       "      <th>protein_match</th>\n",
       "      <th>Jaccard_Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>48641291</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>92992745</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>157628663</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>136708305</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>133672542</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48845</th>\n",
       "      <td>4875</td>\n",
       "      <td>78849058</td>\n",
       "      <td>0</td>\n",
       "      <td>0.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48846</th>\n",
       "      <td>4876</td>\n",
       "      <td>108797464</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48847</th>\n",
       "      <td>4877</td>\n",
       "      <td>161110219</td>\n",
       "      <td>0</td>\n",
       "      <td>0.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48848</th>\n",
       "      <td>4878</td>\n",
       "      <td>74177185</td>\n",
       "      <td>1</td>\n",
       "      <td>0.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48849</th>\n",
       "      <td>4879</td>\n",
       "      <td>143737378</td>\n",
       "      <td>0</td>\n",
       "      <td>0.25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>48850 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0  prot_pair_index  protein_match  Jaccard_Score\n",
       "0               0         48641291              1           1.00\n",
       "1               1         92992745              1           1.00\n",
       "2               2        157628663              1           1.00\n",
       "3               3        136708305              1           1.00\n",
       "4               4        133672542              1           1.00\n",
       "...           ...              ...            ...            ...\n",
       "48845        4875         78849058              0           0.25\n",
       "48846        4876        108797464              1           1.00\n",
       "48847        4877        161110219              0           0.25\n",
       "48848        4878         74177185              1           0.50\n",
       "48849        4879        143737378              0           0.25\n",
       "\n",
       "[48850 rows x 4 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target = pd.read_csv('protein_match_50k')\n",
    "target['protein_match'] = target['protein_match'].map({'Yes': 1, 'No': 0})\n",
    "target"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94da64e9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b2d3db4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.utils import resample\n",
    "\n",
    "# Assuming your data is in a pandas DataFrame called 'data'\n",
    "# Separate the majority and minority classes\n",
    "majority_class = target[target['protein_match'] == 1]\n",
    "minority_class = target[target['protein_match'] == 0]\n",
    "\n",
    "# Undersample the majority class to match the number of minority class samples\n",
    "n_samples = len(minority_class)\n",
    "undersampled_majority = resample(majority_class, n_samples=n_samples, replace=False)\n",
    "\n",
    "# Combine the undersampled majority class with the minority class\n",
    "balanced_data = pd.concat([undersampled_majority, minority_class])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "60de9006",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>prot_pair_index</th>\n",
       "      <th>protein_match</th>\n",
       "      <th>Jaccard_Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>31723</th>\n",
       "      <td>2431</td>\n",
       "      <td>60539488</td>\n",
       "      <td>1</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26282</th>\n",
       "      <td>1874</td>\n",
       "      <td>87049990</td>\n",
       "      <td>1</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15250</th>\n",
       "      <td>596</td>\n",
       "      <td>148005611</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29439</th>\n",
       "      <td>147</td>\n",
       "      <td>163910669</td>\n",
       "      <td>1</td>\n",
       "      <td>0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24293</th>\n",
       "      <td>4756</td>\n",
       "      <td>140356755</td>\n",
       "      <td>1</td>\n",
       "      <td>0.454545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48833</th>\n",
       "      <td>4863</td>\n",
       "      <td>70996712</td>\n",
       "      <td>0</td>\n",
       "      <td>0.285714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48839</th>\n",
       "      <td>4869</td>\n",
       "      <td>14456723</td>\n",
       "      <td>0</td>\n",
       "      <td>0.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48845</th>\n",
       "      <td>4875</td>\n",
       "      <td>78849058</td>\n",
       "      <td>0</td>\n",
       "      <td>0.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48847</th>\n",
       "      <td>4877</td>\n",
       "      <td>161110219</td>\n",
       "      <td>0</td>\n",
       "      <td>0.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48849</th>\n",
       "      <td>4879</td>\n",
       "      <td>143737378</td>\n",
       "      <td>0</td>\n",
       "      <td>0.250000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17108 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0  prot_pair_index  protein_match  Jaccard_Score\n",
       "31723        2431         60539488              1       0.333333\n",
       "26282        1874         87049990              1       0.500000\n",
       "15250         596        148005611              1       1.000000\n",
       "29439         147        163910669              1       0.666667\n",
       "24293        4756        140356755              1       0.454545\n",
       "...           ...              ...            ...            ...\n",
       "48833        4863         70996712              0       0.285714\n",
       "48839        4869         14456723              0       0.250000\n",
       "48845        4875         78849058              0       0.250000\n",
       "48847        4877        161110219              0       0.250000\n",
       "48849        4879        143737378              0       0.250000\n",
       "\n",
       "[17108 rows x 4 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "balanced_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "792508bb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(17108, 33)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.merge(df, balanced_data, on=['prot_pair_index'])\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "62efbea0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0_x</th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>thermo_index</th>\n",
       "      <th>...</th>\n",
       "      <th>ogt_difference</th>\n",
       "      <th>m_protein_seq</th>\n",
       "      <th>t_protein_seq</th>\n",
       "      <th>m_protein_desc</th>\n",
       "      <th>t_protein_desc</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "      <th>Unnamed: 0_y</th>\n",
       "      <th>protein_match</th>\n",
       "      <th>Jaccard_Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>0.275281</td>\n",
       "      <td>0.182156</td>\n",
       "      <td>0.152174</td>\n",
       "      <td>185</td>\n",
       "      <td>0.687732</td>\n",
       "      <td>197</td>\n",
       "      <td>0.525333</td>\n",
       "      <td>48</td>\n",
       "      <td>14963</td>\n",
       "      <td>...</td>\n",
       "      <td>31.5</td>\n",
       "      <td>MRAIGELWPTFDDVHEIAVLRGGGLGDLMFALPAIDALAAAYPEAR...</td>\n",
       "      <td>MATTTAEIGVIGGSGFYSFLDDPHEVTVQTPYGPPSDPIAVGTVAG...</td>\n",
       "      <td>glycosyltransferase family 9 protein</td>\n",
       "      <td>S-methyl-5'-thioadenosine phosphorylase</td>\n",
       "      <td>375</td>\n",
       "      <td>269</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12</td>\n",
       "      <td>0.354067</td>\n",
       "      <td>0.245847</td>\n",
       "      <td>0.225954</td>\n",
       "      <td>208</td>\n",
       "      <td>0.691030</td>\n",
       "      <td>216</td>\n",
       "      <td>0.610170</td>\n",
       "      <td>264</td>\n",
       "      <td>7134</td>\n",
       "      <td>...</td>\n",
       "      <td>24.5</td>\n",
       "      <td>MDSLITVRNLHKLFRVDQREDSGLWPAIKSLIRRERAEVVAVDGVS...</td>\n",
       "      <td>MSEHILETDRLTRRYGDRLAVDAVSMAVRPREVYGFLGPNGAGKTT...</td>\n",
       "      <td>ATP-binding cassette domain-containing protein</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>354</td>\n",
       "      <td>301</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>0.400000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>32</td>\n",
       "      <td>0.314554</td>\n",
       "      <td>0.241007</td>\n",
       "      <td>0.228669</td>\n",
       "      <td>218</td>\n",
       "      <td>0.784173</td>\n",
       "      <td>212</td>\n",
       "      <td>0.688312</td>\n",
       "      <td>228</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MRKQYGDFHAVKGIDLDVRPGECFGLLGPNGAGKSTTMRMLAGTSQ...</td>\n",
       "      <td>MTNGHTDLSGSGLTLGYDRRIVSRDLDVSIPPGSFTVIIGPNACGK...</td>\n",
       "      <td>ATP-binding cassette domain-containing protein</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>308</td>\n",
       "      <td>278</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>0.222222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>33</td>\n",
       "      <td>0.487685</td>\n",
       "      <td>0.452055</td>\n",
       "      <td>0.458333</td>\n",
       "      <td>202</td>\n",
       "      <td>0.922374</td>\n",
       "      <td>203</td>\n",
       "      <td>0.953052</td>\n",
       "      <td>480</td>\n",
       "      <td>4508</td>\n",
       "      <td>...</td>\n",
       "      <td>22.5</td>\n",
       "      <td>MAWTREQMAARAAKELQDGFYVNLGIGIPTLVANYIPDGVHVTLQS...</td>\n",
       "      <td>MDMLVEAAPLDRHGMARRLAADIPEGWVVNLGIGIPTLVSDHVPPE...</td>\n",
       "      <td>3-oxoacid CoA-transferase subunit B</td>\n",
       "      <td>3-oxoacid CoA-transferase subunit B</td>\n",
       "      <td>213</td>\n",
       "      <td>219</td>\n",
       "      <td>31</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>36</td>\n",
       "      <td>0.298701</td>\n",
       "      <td>0.205357</td>\n",
       "      <td>0.206897</td>\n",
       "      <td>254</td>\n",
       "      <td>0.755952</td>\n",
       "      <td>238</td>\n",
       "      <td>0.719033</td>\n",
       "      <td>128</td>\n",
       "      <td>11324</td>\n",
       "      <td>...</td>\n",
       "      <td>21.5</td>\n",
       "      <td>MSATDTPSRPLASLDIKKEVERYALVLGWIFLIAIFGLLLPDTFLT...</td>\n",
       "      <td>MDLLNAIVALLNFVIIPATAYGAQLALGALGVTMIYGILRFSNFAH...</td>\n",
       "      <td>ABC transporter permease</td>\n",
       "      <td>branched-chain amino acid ABC transporter perm...</td>\n",
       "      <td>331</td>\n",
       "      <td>336</td>\n",
       "      <td>34</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17103</th>\n",
       "      <td>49989</td>\n",
       "      <td>0.284091</td>\n",
       "      <td>0.223214</td>\n",
       "      <td>0.245700</td>\n",
       "      <td>190</td>\n",
       "      <td>0.848214</td>\n",
       "      <td>175</td>\n",
       "      <td>0.956284</td>\n",
       "      <td>112</td>\n",
       "      <td>7134</td>\n",
       "      <td>...</td>\n",
       "      <td>22.5</td>\n",
       "      <td>MRRQVGSVPGVTRTAGAAGGGAAARTAPGGPPDVVLMSCSHPRAAE...</td>\n",
       "      <td>MAESHSTTRSGQIRVFVLDDHEVVRRGVRDLLDAEPDITVVGEAET...</td>\n",
       "      <td>response regulator transcription factor</td>\n",
       "      <td>response regulator transcription factor</td>\n",
       "      <td>183</td>\n",
       "      <td>224</td>\n",
       "      <td>4869</td>\n",
       "      <td>0</td>\n",
       "      <td>0.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17104</th>\n",
       "      <td>49994</td>\n",
       "      <td>0.322368</td>\n",
       "      <td>0.303406</td>\n",
       "      <td>0.294737</td>\n",
       "      <td>319</td>\n",
       "      <td>0.987616</td>\n",
       "      <td>335</td>\n",
       "      <td>0.979532</td>\n",
       "      <td>122</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MLTVSYVGAHKIEIREQDPEPVHAGQVQVEVAYAGICGTDLHILHG...</td>\n",
       "      <td>MRAVTISEPGGPEKLQWTEVPDPRPGAGEVLLEVVASAVNRADVLQ...</td>\n",
       "      <td>alcohol dehydrogenase catalytic domain-contain...</td>\n",
       "      <td>NAD(P)H-quinone oxidoreductase</td>\n",
       "      <td>342</td>\n",
       "      <td>323</td>\n",
       "      <td>4874</td>\n",
       "      <td>1</td>\n",
       "      <td>0.600000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17105</th>\n",
       "      <td>49995</td>\n",
       "      <td>0.344828</td>\n",
       "      <td>0.169972</td>\n",
       "      <td>0.192308</td>\n",
       "      <td>180</td>\n",
       "      <td>0.509915</td>\n",
       "      <td>200</td>\n",
       "      <td>0.738007</td>\n",
       "      <td>53</td>\n",
       "      <td>14963</td>\n",
       "      <td>...</td>\n",
       "      <td>24.5</td>\n",
       "      <td>MDNATFRLGDDLSVRLPGHSRWIGQVEREQRWLPWLAPRLPLTVST...</td>\n",
       "      <td>MPPQPRPLRPNDPREIGGFALLGRLGEGGQGTVYLGGAPDGRRVAV...</td>\n",
       "      <td>aminoglycoside phosphotransferase family protein</td>\n",
       "      <td>serine/threonine protein kinase</td>\n",
       "      <td>271</td>\n",
       "      <td>353</td>\n",
       "      <td>4875</td>\n",
       "      <td>0</td>\n",
       "      <td>0.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17106</th>\n",
       "      <td>49997</td>\n",
       "      <td>0.352174</td>\n",
       "      <td>0.239645</td>\n",
       "      <td>0.242152</td>\n",
       "      <td>230</td>\n",
       "      <td>0.680473</td>\n",
       "      <td>227</td>\n",
       "      <td>0.685801</td>\n",
       "      <td>279</td>\n",
       "      <td>7134</td>\n",
       "      <td>...</td>\n",
       "      <td>24.5</td>\n",
       "      <td>MIRLAELTKTYPGQQHPAVDGISMEVAEGEIVVLVGPSGCGKTTTL...</td>\n",
       "      <td>MTEQPILSARGLTVDFRLRGGRRARAVDGVDLDLAPGEVLALAGES...</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>331</td>\n",
       "      <td>338</td>\n",
       "      <td>4877</td>\n",
       "      <td>0</td>\n",
       "      <td>0.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17107</th>\n",
       "      <td>49999</td>\n",
       "      <td>0.336207</td>\n",
       "      <td>0.184834</td>\n",
       "      <td>0.174497</td>\n",
       "      <td>113</td>\n",
       "      <td>0.535545</td>\n",
       "      <td>117</td>\n",
       "      <td>0.495763</td>\n",
       "      <td>165</td>\n",
       "      <td>7134</td>\n",
       "      <td>...</td>\n",
       "      <td>26.5</td>\n",
       "      <td>MTGQGASGSETGVSDDSPITVVVVDDQELLRTGLRDLAEHDGDIAV...</td>\n",
       "      <td>MTTEAQPRRVVLAEDEALIRLDLAEMLAEEGYEVVGQAADGETALA...</td>\n",
       "      <td>response regulator transcription factor</td>\n",
       "      <td>response regulator</td>\n",
       "      <td>236</td>\n",
       "      <td>211</td>\n",
       "      <td>4879</td>\n",
       "      <td>0</td>\n",
       "      <td>0.250000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17108 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0_x  local_gap_compressed_percent_id  \\\n",
       "0                11                         0.275281   \n",
       "1                12                         0.354067   \n",
       "2                32                         0.314554   \n",
       "3                33                         0.487685   \n",
       "4                36                         0.298701   \n",
       "...             ...                              ...   \n",
       "17103         49989                         0.284091   \n",
       "17104         49994                         0.322368   \n",
       "17105         49995                         0.344828   \n",
       "17106         49997                         0.352174   \n",
       "17107         49999                         0.336207   \n",
       "\n",
       "       scaled_local_query_percent_id  scaled_local_symmetric_percent_id  \\\n",
       "0                           0.182156                           0.152174   \n",
       "1                           0.245847                           0.225954   \n",
       "2                           0.241007                           0.228669   \n",
       "3                           0.452055                           0.458333   \n",
       "4                           0.205357                           0.206897   \n",
       "...                              ...                                ...   \n",
       "17103                       0.223214                           0.245700   \n",
       "17104                       0.303406                           0.294737   \n",
       "17105                       0.169972                           0.192308   \n",
       "17106                       0.239645                           0.242152   \n",
       "17107                       0.184834                           0.174497   \n",
       "\n",
       "       query_align_len  query_align_cov  subject_align_len  subject_align_cov  \\\n",
       "0                  185         0.687732                197           0.525333   \n",
       "1                  208         0.691030                216           0.610170   \n",
       "2                  218         0.784173                212           0.688312   \n",
       "3                  202         0.922374                203           0.953052   \n",
       "4                  254         0.755952                238           0.719033   \n",
       "...                ...              ...                ...                ...   \n",
       "17103              190         0.848214                175           0.956284   \n",
       "17104              319         0.987616                335           0.979532   \n",
       "17105              180         0.509915                200           0.738007   \n",
       "17106              230         0.680473                227           0.685801   \n",
       "17107              113         0.535545                117           0.495763   \n",
       "\n",
       "       bit_score  thermo_index  ...  ogt_difference  \\\n",
       "0             48         14963  ...            31.5   \n",
       "1            264          7134  ...            24.5   \n",
       "2            228           875  ...            22.0   \n",
       "3            480          4508  ...            22.5   \n",
       "4            128         11324  ...            21.5   \n",
       "...          ...           ...  ...             ...   \n",
       "17103        112          7134  ...            22.5   \n",
       "17104        122           875  ...            22.0   \n",
       "17105         53         14963  ...            24.5   \n",
       "17106        279          7134  ...            24.5   \n",
       "17107        165          7134  ...            26.5   \n",
       "\n",
       "                                           m_protein_seq  \\\n",
       "0      MRAIGELWPTFDDVHEIAVLRGGGLGDLMFALPAIDALAAAYPEAR...   \n",
       "1      MDSLITVRNLHKLFRVDQREDSGLWPAIKSLIRRERAEVVAVDGVS...   \n",
       "2      MRKQYGDFHAVKGIDLDVRPGECFGLLGPNGAGKSTTMRMLAGTSQ...   \n",
       "3      MAWTREQMAARAAKELQDGFYVNLGIGIPTLVANYIPDGVHVTLQS...   \n",
       "4      MSATDTPSRPLASLDIKKEVERYALVLGWIFLIAIFGLLLPDTFLT...   \n",
       "...                                                  ...   \n",
       "17103  MRRQVGSVPGVTRTAGAAGGGAAARTAPGGPPDVVLMSCSHPRAAE...   \n",
       "17104  MLTVSYVGAHKIEIREQDPEPVHAGQVQVEVAYAGICGTDLHILHG...   \n",
       "17105  MDNATFRLGDDLSVRLPGHSRWIGQVEREQRWLPWLAPRLPLTVST...   \n",
       "17106  MIRLAELTKTYPGQQHPAVDGISMEVAEGEIVVLVGPSGCGKTTTL...   \n",
       "17107  MTGQGASGSETGVSDDSPITVVVVDDQELLRTGLRDLAEHDGDIAV...   \n",
       "\n",
       "                                           t_protein_seq  \\\n",
       "0      MATTTAEIGVIGGSGFYSFLDDPHEVTVQTPYGPPSDPIAVGTVAG...   \n",
       "1      MSEHILETDRLTRRYGDRLAVDAVSMAVRPREVYGFLGPNGAGKTT...   \n",
       "2      MTNGHTDLSGSGLTLGYDRRIVSRDLDVSIPPGSFTVIIGPNACGK...   \n",
       "3      MDMLVEAAPLDRHGMARRLAADIPEGWVVNLGIGIPTLVSDHVPPE...   \n",
       "4      MDLLNAIVALLNFVIIPATAYGAQLALGALGVTMIYGILRFSNFAH...   \n",
       "...                                                  ...   \n",
       "17103  MAESHSTTRSGQIRVFVLDDHEVVRRGVRDLLDAEPDITVVGEAET...   \n",
       "17104  MRAVTISEPGGPEKLQWTEVPDPRPGAGEVLLEVVASAVNRADVLQ...   \n",
       "17105  MPPQPRPLRPNDPREIGGFALLGRLGEGGQGTVYLGGAPDGRRVAV...   \n",
       "17106  MTEQPILSARGLTVDFRLRGGRRARAVDGVDLDLAPGEVLALAGES...   \n",
       "17107  MTTEAQPRRVVLAEDEALIRLDLAEMLAEEGYEVVGQAADGETALA...   \n",
       "\n",
       "                                          m_protein_desc  \\\n",
       "0                   glycosyltransferase family 9 protein   \n",
       "1         ATP-binding cassette domain-containing protein   \n",
       "2         ATP-binding cassette domain-containing protein   \n",
       "3                    3-oxoacid CoA-transferase subunit B   \n",
       "4                               ABC transporter permease   \n",
       "...                                                  ...   \n",
       "17103            response regulator transcription factor   \n",
       "17104  alcohol dehydrogenase catalytic domain-contain...   \n",
       "17105   aminoglycoside phosphotransferase family protein   \n",
       "17106                ABC transporter ATP-binding protein   \n",
       "17107            response regulator transcription factor   \n",
       "\n",
       "                                          t_protein_desc  m_protein_len  \\\n",
       "0                S-methyl-5'-thioadenosine phosphorylase            375   \n",
       "1                    ABC transporter ATP-binding protein            354   \n",
       "2                    ABC transporter ATP-binding protein            308   \n",
       "3                    3-oxoacid CoA-transferase subunit B            213   \n",
       "4      branched-chain amino acid ABC transporter perm...            331   \n",
       "...                                                  ...            ...   \n",
       "17103            response regulator transcription factor            183   \n",
       "17104                     NAD(P)H-quinone oxidoreductase            342   \n",
       "17105                    serine/threonine protein kinase            271   \n",
       "17106                ABC transporter ATP-binding protein            331   \n",
       "17107                                 response regulator            236   \n",
       "\n",
       "       t_protein_len  Unnamed: 0_y  protein_match  Jaccard_Score  \n",
       "0                269            11              0       0.000000  \n",
       "1                301            12              1       0.400000  \n",
       "2                278            30              0       0.222222  \n",
       "3                219            31              1       1.000000  \n",
       "4                336            34              1       1.000000  \n",
       "...              ...           ...            ...            ...  \n",
       "17103            224          4869              0       0.250000  \n",
       "17104            323          4874              1       0.600000  \n",
       "17105            353          4875              0       0.250000  \n",
       "17106            338          4877              0       0.250000  \n",
       "17107            211          4879              0       0.250000  \n",
       "\n",
       "[17108 rows x 33 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6519dcaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#drop columns that don't exihibit signficant pearson correlation with bit_score\n",
    "\n",
    "df = df.drop(columns = ['meso_index', 'meso_protein_int_index', 'local_gap_compressed_percent_id_16s', \n",
    "                        'scaled_local_query_percent_id_16s', 'scaled_local_symmetric_percent_id_16s',\n",
    "                       'bit_score_16s', 'm_ogt', 't_ogt', 'taxa_pair_index', 'thermo_protein_int_index'\n",
    "                       ,'ogt_difference', 'Jaccard_Score',\n",
    "                       'query_align_cov_16s', 'subject_align_cov_16s',\n",
    "                       'Unnamed: 0_x', 'Unnamed: 0_y', 'thermo_index', \n",
    "                       'm_protein_desc', 't_protein_desc'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "16f0544e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['local_gap_compressed_percent_id', 'scaled_local_query_percent_id',\n",
       "       'scaled_local_symmetric_percent_id', 'query_align_len',\n",
       "       'query_align_cov', 'subject_align_len', 'subject_align_cov',\n",
       "       'bit_score', 'prot_pair_index', 'm_protein_seq', 't_protein_seq',\n",
       "       'm_protein_len', 't_protein_len', 'protein_match'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6894895e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.replace([np.inf, -np.inf], np.nan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8004811b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "local_gap_compressed_percent_id      0\n",
      "scaled_local_query_percent_id        0\n",
      "scaled_local_symmetric_percent_id    0\n",
      "query_align_len                      0\n",
      "query_align_cov                      0\n",
      "subject_align_len                    0\n",
      "subject_align_cov                    0\n",
      "bit_score                            0\n",
      "prot_pair_index                      0\n",
      "m_protein_seq                        0\n",
      "t_protein_seq                        0\n",
      "m_protein_len                        0\n",
      "t_protein_len                        0\n",
      "protein_match                        0\n",
      "dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([0])"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nan_counts = df.isna().sum()\n",
    "print(nan_counts)\n",
    "nan_counts.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "76ebfd92",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.dropna(axis=1, how='any')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7c070789",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.drop(columns=['m_protein_seq', 't_protein_seq'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "dd6c520d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['local_gap_compressed_percent_id', 'scaled_local_query_percent_id',\n",
       "       'scaled_local_symmetric_percent_id', 'query_align_len',\n",
       "       'query_align_cov', 'subject_align_len', 'subject_align_cov',\n",
       "       'bit_score', 'prot_pair_index', 'm_protein_len', 't_protein_len',\n",
       "       'protein_match'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e13ec5f3",
   "metadata": {},
   "source": [
    "Use MRMR to select for the best features. Going to start by grouping into different categories of features generated from iFeature Omega."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "f158c7bb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:00<00:00, 669.80it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['t_protein_len']"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#use MRMR to select for the best features from PseKRAAC\n",
    "df_subset = df.loc[:, df.columns != 'protein_match']\n",
    "print(type(df_subset))\n",
    "\n",
    "# select top 10 features using mRMR\n",
    "from mrmr import mrmr_classif\n",
    "selected_features = mrmr_classif(X=df_subset.iloc[:,10:], y=df['protein_match'], K=20)\n",
    "\n",
    "selected_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "5bbe1195",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>prot_pair_index</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.338710</td>\n",
       "      <td>0.318182</td>\n",
       "      <td>0.287671</td>\n",
       "      <td>60</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>71</td>\n",
       "      <td>0.887500</td>\n",
       "      <td>61</td>\n",
       "      <td>133672542</td>\n",
       "      <td>80</td>\n",
       "      <td>66</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.295082</td>\n",
       "      <td>0.291498</td>\n",
       "      <td>218</td>\n",
       "      <td>0.893443</td>\n",
       "      <td>221</td>\n",
       "      <td>0.884000</td>\n",
       "      <td>158</td>\n",
       "      <td>85102366</td>\n",
       "      <td>250</td>\n",
       "      <td>244</td>\n",
       "      <td>244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.275281</td>\n",
       "      <td>0.182156</td>\n",
       "      <td>0.152174</td>\n",
       "      <td>185</td>\n",
       "      <td>0.687732</td>\n",
       "      <td>197</td>\n",
       "      <td>0.525333</td>\n",
       "      <td>48</td>\n",
       "      <td>37128170</td>\n",
       "      <td>375</td>\n",
       "      <td>269</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.270000</td>\n",
       "      <td>0.220408</td>\n",
       "      <td>0.222680</td>\n",
       "      <td>195</td>\n",
       "      <td>0.795918</td>\n",
       "      <td>210</td>\n",
       "      <td>0.875000</td>\n",
       "      <td>173</td>\n",
       "      <td>114387951</td>\n",
       "      <td>240</td>\n",
       "      <td>245</td>\n",
       "      <td>245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.394495</td>\n",
       "      <td>0.346774</td>\n",
       "      <td>0.369099</td>\n",
       "      <td>108</td>\n",
       "      <td>0.870968</td>\n",
       "      <td>106</td>\n",
       "      <td>0.972477</td>\n",
       "      <td>130</td>\n",
       "      <td>166997534</td>\n",
       "      <td>109</td>\n",
       "      <td>124</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17103</th>\n",
       "      <td>0.251397</td>\n",
       "      <td>0.140187</td>\n",
       "      <td>0.134529</td>\n",
       "      <td>186</td>\n",
       "      <td>0.579439</td>\n",
       "      <td>181</td>\n",
       "      <td>0.520115</td>\n",
       "      <td>117</td>\n",
       "      <td>135090204</td>\n",
       "      <td>348</td>\n",
       "      <td>321</td>\n",
       "      <td>321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17104</th>\n",
       "      <td>0.304878</td>\n",
       "      <td>0.188679</td>\n",
       "      <td>0.163666</td>\n",
       "      <td>166</td>\n",
       "      <td>0.626415</td>\n",
       "      <td>170</td>\n",
       "      <td>0.491329</td>\n",
       "      <td>125</td>\n",
       "      <td>133762466</td>\n",
       "      <td>346</td>\n",
       "      <td>265</td>\n",
       "      <td>265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17105</th>\n",
       "      <td>0.344828</td>\n",
       "      <td>0.169972</td>\n",
       "      <td>0.192308</td>\n",
       "      <td>180</td>\n",
       "      <td>0.509915</td>\n",
       "      <td>200</td>\n",
       "      <td>0.738007</td>\n",
       "      <td>53</td>\n",
       "      <td>78849058</td>\n",
       "      <td>271</td>\n",
       "      <td>353</td>\n",
       "      <td>353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17106</th>\n",
       "      <td>0.352174</td>\n",
       "      <td>0.239645</td>\n",
       "      <td>0.242152</td>\n",
       "      <td>230</td>\n",
       "      <td>0.680473</td>\n",
       "      <td>227</td>\n",
       "      <td>0.685801</td>\n",
       "      <td>279</td>\n",
       "      <td>161110219</td>\n",
       "      <td>331</td>\n",
       "      <td>338</td>\n",
       "      <td>338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17107</th>\n",
       "      <td>0.336207</td>\n",
       "      <td>0.184834</td>\n",
       "      <td>0.174497</td>\n",
       "      <td>113</td>\n",
       "      <td>0.535545</td>\n",
       "      <td>117</td>\n",
       "      <td>0.495763</td>\n",
       "      <td>165</td>\n",
       "      <td>143737378</td>\n",
       "      <td>236</td>\n",
       "      <td>211</td>\n",
       "      <td>211</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17108 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       local_gap_compressed_percent_id  scaled_local_query_percent_id  \\\n",
       "0                             0.338710                       0.318182   \n",
       "1                             0.333333                       0.295082   \n",
       "2                             0.275281                       0.182156   \n",
       "3                             0.270000                       0.220408   \n",
       "4                             0.394495                       0.346774   \n",
       "...                                ...                            ...   \n",
       "17103                         0.251397                       0.140187   \n",
       "17104                         0.304878                       0.188679   \n",
       "17105                         0.344828                       0.169972   \n",
       "17106                         0.352174                       0.239645   \n",
       "17107                         0.336207                       0.184834   \n",
       "\n",
       "       scaled_local_symmetric_percent_id  query_align_len  query_align_cov  \\\n",
       "0                               0.287671               60         0.909091   \n",
       "1                               0.291498              218         0.893443   \n",
       "2                               0.152174              185         0.687732   \n",
       "3                               0.222680              195         0.795918   \n",
       "4                               0.369099              108         0.870968   \n",
       "...                                  ...              ...              ...   \n",
       "17103                           0.134529              186         0.579439   \n",
       "17104                           0.163666              166         0.626415   \n",
       "17105                           0.192308              180         0.509915   \n",
       "17106                           0.242152              230         0.680473   \n",
       "17107                           0.174497              113         0.535545   \n",
       "\n",
       "       subject_align_len  subject_align_cov  bit_score  prot_pair_index  \\\n",
       "0                     71           0.887500         61        133672542   \n",
       "1                    221           0.884000        158         85102366   \n",
       "2                    197           0.525333         48         37128170   \n",
       "3                    210           0.875000        173        114387951   \n",
       "4                    106           0.972477        130        166997534   \n",
       "...                  ...                ...        ...              ...   \n",
       "17103                181           0.520115        117        135090204   \n",
       "17104                170           0.491329        125        133762466   \n",
       "17105                200           0.738007         53         78849058   \n",
       "17106                227           0.685801        279        161110219   \n",
       "17107                117           0.495763        165        143737378   \n",
       "\n",
       "       m_protein_len  t_protein_len  t_protein_len  \n",
       "0                 80             66             66  \n",
       "1                250            244            244  \n",
       "2                375            269            269  \n",
       "3                240            245            245  \n",
       "4                109            124            124  \n",
       "...              ...            ...            ...  \n",
       "17103            348            321            321  \n",
       "17104            346            265            265  \n",
       "17105            271            353            353  \n",
       "17106            331            338            338  \n",
       "17107            236            211            211  \n",
       "\n",
       "[17108 rows x 12 columns]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_features_df = df[[feature for feature in selected_features]]\n",
    "\n",
    "#concatenates original feature vector back into the dataframe\n",
    "df = pd.concat([df.iloc[:, :11], best_features_df], axis=1)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "af9f23cf",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(14541, 12)\n",
      "(2567, 12)\n",
      "(12359, 12)\n",
      "(2182, 12)\n"
     ]
    }
   ],
   "source": [
    "#choosing 80/20 split instead of 85/15 because of volume of data\n",
    "\n",
    "dev, test = sklearn.model_selection.train_test_split(df, test_size=0.15, random_state=1)\n",
    "\n",
    "train, val = sklearn.model_selection.train_test_split(dev, test_size=0.15, random_state=1)\n",
    "\n",
    "print(dev.shape)\n",
    "print(test.shape)\n",
    "print(train.shape)\n",
    "print(val.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "2663ce04",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>prot_pair_index</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "      <th>protein_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1968</th>\n",
       "      <td>0.314516</td>\n",
       "      <td>0.157258</td>\n",
       "      <td>0.174107</td>\n",
       "      <td>128</td>\n",
       "      <td>0.516129</td>\n",
       "      <td>122</td>\n",
       "      <td>0.610000</td>\n",
       "      <td>127</td>\n",
       "      <td>157036297</td>\n",
       "      <td>200</td>\n",
       "      <td>248</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6217</th>\n",
       "      <td>0.371429</td>\n",
       "      <td>0.237443</td>\n",
       "      <td>0.232662</td>\n",
       "      <td>157</td>\n",
       "      <td>0.716895</td>\n",
       "      <td>137</td>\n",
       "      <td>0.600877</td>\n",
       "      <td>119</td>\n",
       "      <td>144857841</td>\n",
       "      <td>228</td>\n",
       "      <td>219</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4435</th>\n",
       "      <td>0.331897</td>\n",
       "      <td>0.292776</td>\n",
       "      <td>0.308000</td>\n",
       "      <td>239</td>\n",
       "      <td>0.908745</td>\n",
       "      <td>229</td>\n",
       "      <td>0.966245</td>\n",
       "      <td>319</td>\n",
       "      <td>103555510</td>\n",
       "      <td>237</td>\n",
       "      <td>263</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10376</th>\n",
       "      <td>0.597156</td>\n",
       "      <td>0.591549</td>\n",
       "      <td>0.565022</td>\n",
       "      <td>210</td>\n",
       "      <td>0.985915</td>\n",
       "      <td>210</td>\n",
       "      <td>0.901288</td>\n",
       "      <td>571</td>\n",
       "      <td>72720308</td>\n",
       "      <td>233</td>\n",
       "      <td>213</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7375</th>\n",
       "      <td>0.309179</td>\n",
       "      <td>0.252964</td>\n",
       "      <td>0.259109</td>\n",
       "      <td>213</td>\n",
       "      <td>0.841897</td>\n",
       "      <td>205</td>\n",
       "      <td>0.850622</td>\n",
       "      <td>267</td>\n",
       "      <td>124384823</td>\n",
       "      <td>241</td>\n",
       "      <td>253</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13199</th>\n",
       "      <td>0.325444</td>\n",
       "      <td>0.255814</td>\n",
       "      <td>0.248307</td>\n",
       "      <td>169</td>\n",
       "      <td>0.786046</td>\n",
       "      <td>179</td>\n",
       "      <td>0.785088</td>\n",
       "      <td>175</td>\n",
       "      <td>42719131</td>\n",
       "      <td>228</td>\n",
       "      <td>215</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16196</th>\n",
       "      <td>0.324742</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>0.233333</td>\n",
       "      <td>195</td>\n",
       "      <td>0.844156</td>\n",
       "      <td>206</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>206</td>\n",
       "      <td>79907519</td>\n",
       "      <td>309</td>\n",
       "      <td>231</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10844</th>\n",
       "      <td>0.309417</td>\n",
       "      <td>0.274900</td>\n",
       "      <td>0.281059</td>\n",
       "      <td>228</td>\n",
       "      <td>0.908367</td>\n",
       "      <td>218</td>\n",
       "      <td>0.908333</td>\n",
       "      <td>180</td>\n",
       "      <td>181100188</td>\n",
       "      <td>240</td>\n",
       "      <td>251</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6910</th>\n",
       "      <td>0.241525</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>0.212687</td>\n",
       "      <td>246</td>\n",
       "      <td>0.863158</td>\n",
       "      <td>233</td>\n",
       "      <td>0.928287</td>\n",
       "      <td>173</td>\n",
       "      <td>51999008</td>\n",
       "      <td>251</td>\n",
       "      <td>285</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4562</th>\n",
       "      <td>0.300493</td>\n",
       "      <td>0.240157</td>\n",
       "      <td>0.239216</td>\n",
       "      <td>212</td>\n",
       "      <td>0.834646</td>\n",
       "      <td>213</td>\n",
       "      <td>0.832031</td>\n",
       "      <td>97</td>\n",
       "      <td>45170187</td>\n",
       "      <td>256</td>\n",
       "      <td>254</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2182 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       local_gap_compressed_percent_id  scaled_local_query_percent_id  \\\n",
       "1968                          0.314516                       0.157258   \n",
       "6217                          0.371429                       0.237443   \n",
       "4435                          0.331897                       0.292776   \n",
       "10376                         0.597156                       0.591549   \n",
       "7375                          0.309179                       0.252964   \n",
       "...                                ...                            ...   \n",
       "13199                         0.325444                       0.255814   \n",
       "16196                         0.324742                       0.272727   \n",
       "10844                         0.309417                       0.274900   \n",
       "6910                          0.241525                       0.200000   \n",
       "4562                          0.300493                       0.240157   \n",
       "\n",
       "       scaled_local_symmetric_percent_id  query_align_len  query_align_cov  \\\n",
       "1968                            0.174107              128         0.516129   \n",
       "6217                            0.232662              157         0.716895   \n",
       "4435                            0.308000              239         0.908745   \n",
       "10376                           0.565022              210         0.985915   \n",
       "7375                            0.259109              213         0.841897   \n",
       "...                                  ...              ...              ...   \n",
       "13199                           0.248307              169         0.786046   \n",
       "16196                           0.233333              195         0.844156   \n",
       "10844                           0.281059              228         0.908367   \n",
       "6910                            0.212687              246         0.863158   \n",
       "4562                            0.239216              212         0.834646   \n",
       "\n",
       "       subject_align_len  subject_align_cov  bit_score  prot_pair_index  \\\n",
       "1968                 122           0.610000        127        157036297   \n",
       "6217                 137           0.600877        119        144857841   \n",
       "4435                 229           0.966245        319        103555510   \n",
       "10376                210           0.901288        571         72720308   \n",
       "7375                 205           0.850622        267        124384823   \n",
       "...                  ...                ...        ...              ...   \n",
       "13199                179           0.785088        175         42719131   \n",
       "16196                206           0.666667        206         79907519   \n",
       "10844                218           0.908333        180        181100188   \n",
       "6910                 233           0.928287        173         51999008   \n",
       "4562                 213           0.832031         97         45170187   \n",
       "\n",
       "       m_protein_len  t_protein_len  protein_match  \n",
       "1968             200            248              0  \n",
       "6217             228            219              0  \n",
       "4435             237            263              1  \n",
       "10376            233            213              1  \n",
       "7375             241            253              1  \n",
       "...              ...            ...            ...  \n",
       "13199            228            215              1  \n",
       "16196            309            231              0  \n",
       "10844            240            251              0  \n",
       "6910             251            285              0  \n",
       "4562             256            254              0  \n",
       "\n",
       "[2182 rows x 12 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "val"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8470550d",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#ID target and features, separate into separate arrays\n",
    "\n",
    "target = 'protein_match'\n",
    "input_features = [columns for columns in df]\n",
    "input_features.remove(target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7d45661d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['local_gap_compressed_percent_id',\n",
       " 'scaled_local_query_percent_id',\n",
       " 'scaled_local_symmetric_percent_id',\n",
       " 'query_align_len',\n",
       " 'query_align_cov',\n",
       " 'subject_align_len',\n",
       " 'subject_align_cov',\n",
       " 'bit_score',\n",
       " 'prot_pair_index',\n",
       " 'm_protein_seq',\n",
       " 't_protein_seq',\n",
       " 'm_protein_len',\n",
       " 't_protein_len']"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "1ab62bff",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(14541, 11) (2567, 11) (14541, 1) (2567, 1)\n"
     ]
    }
   ],
   "source": [
    "#split X and y\n",
    "\n",
    "dev_X = dev[input_features].values\n",
    "test_X = test[input_features].values\n",
    "\n",
    "dev_y = dev[target].values.reshape(-1,1)\n",
    "test_y = test[target].values.reshape(-1,1)  \n",
    "\n",
    "print(dev_X.shape, test_X.shape, dev_y.shape, test_y.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "9ac52ac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#same thing for training and validation data\n",
    "\n",
    "train_X = train[input_features].values\n",
    "val_X = val[input_features].values\n",
    "\n",
    "train_y = train[target].values.reshape(-1,1)\n",
    "val_y = val[target].values.reshape(-1,1) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c374f2d7",
   "metadata": {},
   "source": [
    "Scale the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "2b48b2e5",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "scaler = sklearn.preprocessing.StandardScaler()\n",
    "dev_X = scaler.fit_transform(dev_X)\n",
    "test_X = scaler.fit_transform(test_X)\n",
    "train_X = scaler.fit_transform(train_X)\n",
    "val_X = scaler.fit_transform(val_X)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3022371",
   "metadata": {},
   "source": [
    "Train the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f4411c95",
   "metadata": {},
   "outputs": [],
   "source": [
    "def k_fold_cross_val(dataframe, n_splits=10):  \n",
    "    \"\"\"\n",
    "    Runs k-fold cross validation on dataset.\n",
    "    Default = 10-fold.\n",
    "\n",
    "    Params\n",
    "    ----------\n",
    "    -dataframe: Pandas dataframe\n",
    "    -n_splits: Number of cross validations (int)\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    -vector of predictions\n",
    "    \"\"\"\n",
    "    \n",
    "    dev, test = sklearn.model_selection.train_test_split(dataframe, test_size=0.15, random_state=1)\n",
    "\n",
    "    train, val = sklearn.model_selection.train_test_split(dev, test_size=0.15, random_state=1)\n",
    "    \n",
    "    target = 'protein_match'\n",
    "    input_features = [columns for columns in df]\n",
    "    input_features.remove(target)\n",
    "    \n",
    "    dev_X = dev[input_features].values\n",
    "    test_X = test[input_features].values\n",
    "\n",
    "    dev_y = dev[target].values.reshape(-1,1)\n",
    "    test_y = test[target].values.reshape(-1,1) \n",
    "\n",
    "    from sklearn.model_selection import StratifiedKFold\n",
    "\n",
    "    cv = StratifiedKFold(n_splits, shuffle=True)\n",
    "\n",
    "    for (train_index, test_index) in cv.split(dev_X, dev_y):\n",
    "\n",
    "        train_X = dev_X[train_index]\n",
    "        val_X = dev_X[test_index]\n",
    "\n",
    "        train_y = dev_y[train_index]\n",
    "        val_y = dev_y[test_index]\n",
    "\n",
    "        model.fit(train_X, train_y)\n",
    "\n",
    "        preds = model.predict(val_X)\n",
    "\n",
    "        return preds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "6c8325f2",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'preds' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-28-c66c8509e20f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mpreds\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'preds' is not defined"
     ]
    }
   ],
   "source": [
    "preds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "be4e09b0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-1 {color: black;background-color: white;}#sk-container-id-1 pre{padding: 0;}#sk-container-id-1 div.sk-toggleable {background-color: white;}#sk-container-id-1 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-1 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-1 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-1 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-1 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-1 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-1 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-1 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-1 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-1 div.sk-item {position: relative;z-index: 1;}#sk-container-id-1 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-1 div.sk-item::before, #sk-container-id-1 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-1 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-1 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-1 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-1 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-1 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-1 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-1 div.sk-label-container {text-align: center;}#sk-container-id-1 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-1 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>RandomForestClassifier()</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" checked><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">RandomForestClassifier</label><div class=\"sk-toggleable__content\"><pre>RandomForestClassifier()</pre></div></div></div></div></div>"
      ],
      "text/plain": [
       "RandomForestClassifier()"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Random Forest\n",
    "\n",
    "#hyperparameters determiend with optuna\n",
    "model = sklearn.ensemble.RandomForestClassifier()\n",
    "\n",
    "model.fit(train_X, train_y.ravel())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69ebf269",
   "metadata": {},
   "source": [
    "Test the model, report relevant statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "c73327e9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'val_X' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-57-bf2b9d67c999>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mscore\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mscore\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mval_X\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mval_y\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Model score is: {}'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mscore\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mpreds\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpredict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtest_X\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpreds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'val_X' is not defined"
     ]
    }
   ],
   "source": [
    "score = model.score(val_X, val_y)\n",
    "print('Model score is: {}'.format(score))\n",
    "\n",
    "preds = model.predict(test_X)\n",
    "print(preds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "23254534",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'val_X' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-58-5354276309fc>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mproba_y\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpredict_proba\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mval_X\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'val_X' is not defined"
     ]
    }
   ],
   "source": [
    "proba_y = model.predict_proba(val_X)[:,1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "efcc8c3b",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'proba_y' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-59-069e8d20f9f7>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mproba_y\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'proba_y' is not defined"
     ]
    }
   ],
   "source": [
    "proba_y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e9dce62",
   "metadata": {},
   "outputs": [],
   "source": [
    "#confusion matrix\n",
    "\n",
    "confusion_matrix = sklearn.metrics.confusion_matrix(preds, test_y)\n",
    "sklearn.metrics.ConfusionMatrixDisplay(confusion_matrix).plot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f838ddbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.naive_bayes import GaussianNB\n",
    "from sklearn.ensemble import BaggingClassifier, \\\n",
    "    RandomForestClassifier, \\\n",
    "    AdaBoostClassifier, \\\n",
    "    GradientBoostingClassifier, \\\n",
    "    ExtraTreesClassifier\n",
    "\n",
    "from sklearn.discriminant_analysis import LinearDiscriminantAnalysis\n",
    "\n",
    "names = ['LR', 'KNN', 'DT', 'NB', 'RF', 'Bagging', 'AB', 'GB', 'SVM']\n",
    "\n",
    "#list of classifiers (hyperparameters optimized)\n",
    "classifiers = [\n",
    "    #Regression\n",
    "    LogisticRegression(),\n",
    "    #KNN (neighbors optimized iteratively)\n",
    "    KNeighborsClassifier(n_neighbors=20),\n",
    "    #Decision Tree\n",
    "    DecisionTreeClassifier(max_features=None),\n",
    "    #Gaussian\n",
    "    GaussianNB(),\n",
    "    #RF Classifier (with optuna)\n",
    "    RandomForestClassifier(\n",
    "        n_estimators=200,\n",
    "        max_depth=None,\n",
    "        max_samples=0.3,\n",
    "        max_features=0.5,\n",
    "        min_weight_fraction_leaf=0,\n",
    "        min_samples_split=17),\n",
    "    #RF Classifier with bagging (with optuna)\n",
    "    BaggingClassifier(sklearn.ensemble.RandomForestClassifier\n",
    "    (n_estimators=200, max_depth=None, \n",
    "     min_weight_fraction_leaf=0.000215),max_samples=0.5, \n",
    "    max_features=0.5),\n",
    "    #AdaBoost (with optuna)\n",
    "    AdaBoostClassifier(n_estimators=53, learning_rate=0.156),\n",
    "    #Gradient Boosting (with optuna)\n",
    "    GradientBoostingClassifier(n_estimators=100, learning_rate=1.0, \n",
    "                                max_depth=1),   \n",
    "    #C-support vector classification (9)\n",
    "#     SVC(),\n",
    "]\n",
    "\n",
    "F = open('evaluationResults.txt', 'w')\n",
    "\n",
    "F.write('Evaluation Scale:'+'\\n')\n",
    "F.write('0.0% <=Accuracy<= 100.0%'+'\\n')\n",
    "F.write('0.0 <=AUC<= 1.0'+'\\n') #area under curve\n",
    "F.write('0.0 <=auPR<= 1.0'+'\\n')  # average_Precision\n",
    "F.write('0.0 <=F1_Score<= 1.0'+'\\n')\n",
    "F.write('-1.0 <=MCC<= 1.0'+'\\n')\n",
    "F.write('_______________________________________'+'\\n')\n",
    "\n",
    "def runClassifiers(filename:str, dataframe, columns=[], target=[], model=RandomForestClassifier()):\n",
    "\n",
    "    \"\"\"\n",
    "    Takes dataframe and splits it into a training and testing set. \n",
    "    Trains a RF Classifier with data.\n",
    "\n",
    "    Params\n",
    "    ----------\n",
    "    dataframe: Pandas dataframe\n",
    "    columns: list of strings, representing input features\n",
    "    target: list of strings, representing target feature(s)\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    -Accuracy score\n",
    "    -area under ROC curve\n",
    "    -train data (target)\n",
    "    -validation data (features)\n",
    "    -validation data (target)\n",
    "    \"\"\"\n",
    "    \n",
    "    dev, test = sklearn.model_selection.train_test_split(dataframe, test_size=0.15, random_state=1)\n",
    "\n",
    "    train, val = sklearn.model_selection.train_test_split(dev, test_size=0.15, random_state=1)\n",
    "    \n",
    "    dev_X = dev[columns].values\n",
    "    test_X = test[columns].values\n",
    "\n",
    "    dev_y = dev[target].values.reshape(-1,1)\n",
    "    test_y = test[target].values.reshape(-1,1) \n",
    "\n",
    "    # test input arguments\n",
    "    assert \"pandas.core.frame.DataFrame\" in str(type(train))\n",
    "    assert \"pandas.core.frame.DataFrame\" in str(type(val))\n",
    "    assert \"str\" in str(type(columns[0]))\n",
    "    assert \"str\" in str(type(target[0]))\n",
    "\n",
    "    # split into input and output feature(s)\n",
    "    train_X = train[columns].values\n",
    "    val_X = val[columns].values\n",
    "\n",
    "    train_y = train[target].values.reshape(-1, 1)\n",
    "    val_y = val[target].values.reshape(-1, 1)\n",
    "\n",
    "    # scale data\n",
    "    scaler = sklearn.preprocessing.StandardScaler()\n",
    "    train_X = scaler.fit_transform(train_X)\n",
    "    val_X = scaler.fit_transform(val_X)\n",
    "\n",
    "    Results = {}  # compare algorithms\n",
    "\n",
    "    from sklearn.metrics import accuracy_score, \\\n",
    "        confusion_matrix, \\\n",
    "        roc_auc_score,\\\n",
    "        average_precision_score,\\\n",
    "        auc,\\\n",
    "        roc_curve, f1_score, recall_score, matthews_corrcoef, auc\n",
    "\n",
    "    for classifier, name in zip(classifiers, names):\n",
    "        accuracy = []\n",
    "        avg_precision = []\n",
    "        F1_Score = []\n",
    "        AUC = []\n",
    "        MCC = []\n",
    "        Recall = []\n",
    "        \n",
    "        mean_TPR = 0.0\n",
    "        mean_FPR = np.linspace(0, 1, 100)\n",
    "\n",
    "        print('{} is done.'.format(classifier.__class__.__name__))\n",
    "\n",
    "        model = classifier\n",
    "\n",
    "        # model\n",
    "        model.fit(train_X, train_y)\n",
    "\n",
    "        preds = model.predict(val_X)\n",
    "\n",
    "        # Calculate ROC Curve and Area the Curve\n",
    "        proba_y = model.predict_proba(val_X)[:,1]\n",
    "        FPR, TPR, _ = roc_curve(val_y, proba_y, pos_label=1)\n",
    "        roc_auc = auc(FPR, TPR)\n",
    "        \n",
    "        #calculate scoring metrics\n",
    "        #include option to return these scores\n",
    "        accuracy.append(accuracy_score(y_pred=preds, y_true=val_y))\n",
    "        avg_precision.append(average_precision_score(y_true=val_y, y_score=proba_y, pos_label=1))\n",
    "        F1_Score.append(f1_score(y_true=val_y, y_pred=preds, pos_label=1))\n",
    "        MCC.append(matthews_corrcoef(y_true=val_y, y_pred=preds))\n",
    "        Recall.append(recall_score(y_true=val_y, y_pred=preds, pos_label=1))\n",
    "        AUC.append(roc_auc)\n",
    "\n",
    "        confusion_matrix = sklearn.metrics.confusion_matrix(y_pred=preds, y_true=val_y)\n",
    "        sklearn.metrics.ConfusionMatrixDisplay(confusion_matrix).plot()\n",
    "\n",
    "        accuracy = [_*100.0 for _ in accuracy]\n",
    "        Results[name + ' Accuracy, F1 Score'] = [accuracy, F1_Score]\n",
    "        \n",
    "        F.write('Classifier: {}\\n'.format(name))\n",
    "        F.write('Accuracy: {0:.4f}%\\n'.format(np.mean(accuracy)))\n",
    "        F.write('AUC: {0:.4f}\\n'.format( np.mean(AUC)))\n",
    "        F.write('auPR: {0:.4f}\\n'.format(np.mean(avg_precision))) # average_Precision\n",
    "        F.write('F1_Score: {0:.4f}\\n'.format(np.mean(F1_Score)))\n",
    "        F.write('MCC: {0:.4f}\\n'.format(np.mean(MCC)))\n",
    "\n",
    "#         TN, FP, FN, TP = CM.ravel()\n",
    "        F.write('Recall: {0:.4f}\\n'.format( np.mean(Recall)) )\n",
    "        F.write('_______________________________________'+'\\n')\n",
    "    \n",
    "    F.close()\n",
    "    \n",
    "    return Results, model\n",
    "\n",
    "\n",
    "# if __name__ == '__main__':\n",
    "#     # print('Please, enter number of cross validation:')\n",
    "#     import argparse\n",
    "#     p = argparse.ArgumentParser(description='Run Machine Learning Classifiers.')\n",
    "\n",
    "#     p.add_argument('-cv', '--nFCV', type=int, help='Number of crossValidation', default=10)\n",
    "#     p.add_argument('-data', '--dataset', type=str, help='~/dataset.csv', default='optimumDataset.csv')\n",
    "#     p.add_argument('-roc', '--auROC', type=int, help='Print ROC Curve', default=1, choices=[0, 1])\n",
    "#     p.add_argument('-box', '--boxPlot', type=int, help='Print Accuracy Box Plaot', default=1, choices=[0, 1])\n",
    "\n",
    "#     args = p.parse_args()\n",
    "\n",
    "#     runClassifiers(args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "07228656",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "input_features.remove('m_protein_seq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c279e219",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_features.remove('t_protein_seq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ee6b2977",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_features.remove('prot_pair_index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f31bd3b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['local_gap_compressed_percent_id',\n",
       " 'scaled_local_query_percent_id',\n",
       " 'scaled_local_symmetric_percent_id',\n",
       " 'query_align_len',\n",
       " 'query_align_cov',\n",
       " 'subject_align_len',\n",
       " 'subject_align_cov',\n",
       " 'bit_score',\n",
       " 'm_protein_len',\n",
       " 't_protein_len']"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "fa9a7b06",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df = df.drop(columns=['m_protein_seq', 't_protein_seq'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "5fadd7f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>prot_pair_index</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "      <th>protein_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.275281</td>\n",
       "      <td>0.182156</td>\n",
       "      <td>0.152174</td>\n",
       "      <td>185</td>\n",
       "      <td>0.687732</td>\n",
       "      <td>197</td>\n",
       "      <td>0.525333</td>\n",
       "      <td>48</td>\n",
       "      <td>37128170</td>\n",
       "      <td>375</td>\n",
       "      <td>269</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.354067</td>\n",
       "      <td>0.245847</td>\n",
       "      <td>0.225954</td>\n",
       "      <td>208</td>\n",
       "      <td>0.691030</td>\n",
       "      <td>216</td>\n",
       "      <td>0.610170</td>\n",
       "      <td>264</td>\n",
       "      <td>55279368</td>\n",
       "      <td>354</td>\n",
       "      <td>301</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.314554</td>\n",
       "      <td>0.241007</td>\n",
       "      <td>0.228669</td>\n",
       "      <td>218</td>\n",
       "      <td>0.784173</td>\n",
       "      <td>212</td>\n",
       "      <td>0.688312</td>\n",
       "      <td>228</td>\n",
       "      <td>107256987</td>\n",
       "      <td>308</td>\n",
       "      <td>278</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.487685</td>\n",
       "      <td>0.452055</td>\n",
       "      <td>0.458333</td>\n",
       "      <td>202</td>\n",
       "      <td>0.922374</td>\n",
       "      <td>203</td>\n",
       "      <td>0.953052</td>\n",
       "      <td>480</td>\n",
       "      <td>121070359</td>\n",
       "      <td>213</td>\n",
       "      <td>219</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.298701</td>\n",
       "      <td>0.205357</td>\n",
       "      <td>0.206897</td>\n",
       "      <td>254</td>\n",
       "      <td>0.755952</td>\n",
       "      <td>238</td>\n",
       "      <td>0.719033</td>\n",
       "      <td>128</td>\n",
       "      <td>131454663</td>\n",
       "      <td>331</td>\n",
       "      <td>336</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17103</th>\n",
       "      <td>0.284091</td>\n",
       "      <td>0.223214</td>\n",
       "      <td>0.245700</td>\n",
       "      <td>190</td>\n",
       "      <td>0.848214</td>\n",
       "      <td>175</td>\n",
       "      <td>0.956284</td>\n",
       "      <td>112</td>\n",
       "      <td>14456723</td>\n",
       "      <td>183</td>\n",
       "      <td>224</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17104</th>\n",
       "      <td>0.322368</td>\n",
       "      <td>0.303406</td>\n",
       "      <td>0.294737</td>\n",
       "      <td>319</td>\n",
       "      <td>0.987616</td>\n",
       "      <td>335</td>\n",
       "      <td>0.979532</td>\n",
       "      <td>122</td>\n",
       "      <td>46032974</td>\n",
       "      <td>342</td>\n",
       "      <td>323</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17105</th>\n",
       "      <td>0.344828</td>\n",
       "      <td>0.169972</td>\n",
       "      <td>0.192308</td>\n",
       "      <td>180</td>\n",
       "      <td>0.509915</td>\n",
       "      <td>200</td>\n",
       "      <td>0.738007</td>\n",
       "      <td>53</td>\n",
       "      <td>78849058</td>\n",
       "      <td>271</td>\n",
       "      <td>353</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17106</th>\n",
       "      <td>0.352174</td>\n",
       "      <td>0.239645</td>\n",
       "      <td>0.242152</td>\n",
       "      <td>230</td>\n",
       "      <td>0.680473</td>\n",
       "      <td>227</td>\n",
       "      <td>0.685801</td>\n",
       "      <td>279</td>\n",
       "      <td>161110219</td>\n",
       "      <td>331</td>\n",
       "      <td>338</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17107</th>\n",
       "      <td>0.336207</td>\n",
       "      <td>0.184834</td>\n",
       "      <td>0.174497</td>\n",
       "      <td>113</td>\n",
       "      <td>0.535545</td>\n",
       "      <td>117</td>\n",
       "      <td>0.495763</td>\n",
       "      <td>165</td>\n",
       "      <td>143737378</td>\n",
       "      <td>236</td>\n",
       "      <td>211</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17108 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       local_gap_compressed_percent_id  scaled_local_query_percent_id  \\\n",
       "0                             0.275281                       0.182156   \n",
       "1                             0.354067                       0.245847   \n",
       "2                             0.314554                       0.241007   \n",
       "3                             0.487685                       0.452055   \n",
       "4                             0.298701                       0.205357   \n",
       "...                                ...                            ...   \n",
       "17103                         0.284091                       0.223214   \n",
       "17104                         0.322368                       0.303406   \n",
       "17105                         0.344828                       0.169972   \n",
       "17106                         0.352174                       0.239645   \n",
       "17107                         0.336207                       0.184834   \n",
       "\n",
       "       scaled_local_symmetric_percent_id  query_align_len  query_align_cov  \\\n",
       "0                               0.152174              185         0.687732   \n",
       "1                               0.225954              208         0.691030   \n",
       "2                               0.228669              218         0.784173   \n",
       "3                               0.458333              202         0.922374   \n",
       "4                               0.206897              254         0.755952   \n",
       "...                                  ...              ...              ...   \n",
       "17103                           0.245700              190         0.848214   \n",
       "17104                           0.294737              319         0.987616   \n",
       "17105                           0.192308              180         0.509915   \n",
       "17106                           0.242152              230         0.680473   \n",
       "17107                           0.174497              113         0.535545   \n",
       "\n",
       "       subject_align_len  subject_align_cov  bit_score  prot_pair_index  \\\n",
       "0                    197           0.525333         48         37128170   \n",
       "1                    216           0.610170        264         55279368   \n",
       "2                    212           0.688312        228        107256987   \n",
       "3                    203           0.953052        480        121070359   \n",
       "4                    238           0.719033        128        131454663   \n",
       "...                  ...                ...        ...              ...   \n",
       "17103                175           0.956284        112         14456723   \n",
       "17104                335           0.979532        122         46032974   \n",
       "17105                200           0.738007         53         78849058   \n",
       "17106                227           0.685801        279        161110219   \n",
       "17107                117           0.495763        165        143737378   \n",
       "\n",
       "       m_protein_len  t_protein_len  protein_match  \n",
       "0                375            269              0  \n",
       "1                354            301              1  \n",
       "2                308            278              0  \n",
       "3                213            219              1  \n",
       "4                331            336              1  \n",
       "...              ...            ...            ...  \n",
       "17103            183            224              0  \n",
       "17104            342            323              1  \n",
       "17105            271            353              0  \n",
       "17106            331            338              0  \n",
       "17107            236            211              0  \n",
       "\n",
       "[17108 rows x 12 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0a3fc7ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "target = 'protein_match'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "849a7b36",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LogisticRegression is done.\n",
      "KNeighborsClassifier is done.\n",
      "DecisionTreeClassifier is done.\n",
      "GaussianNB is done.\n",
      "RandomForestClassifier is done.\n",
      "BaggingClassifier is done.\n",
      "AdaBoostClassifier is done.\n",
      "GradientBoostingClassifier is done.\n"
     ]
    }
   ],
   "source": [
    "result = runClassifiers('sample.csv', df, columns=input_features, target=target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "1a8b753a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sklearn.ensemble._gb.GradientBoostingClassifier"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = result[1]\n",
    "type(model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "8c209ce3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>prot_pair_index</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "      <th>protein_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.275281</td>\n",
       "      <td>0.182156</td>\n",
       "      <td>0.152174</td>\n",
       "      <td>185</td>\n",
       "      <td>0.687732</td>\n",
       "      <td>197</td>\n",
       "      <td>0.525333</td>\n",
       "      <td>48</td>\n",
       "      <td>37128170</td>\n",
       "      <td>375</td>\n",
       "      <td>269</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.322430</td>\n",
       "      <td>0.264368</td>\n",
       "      <td>0.245552</td>\n",
       "      <td>232</td>\n",
       "      <td>0.888889</td>\n",
       "      <td>208</td>\n",
       "      <td>0.691030</td>\n",
       "      <td>227</td>\n",
       "      <td>172639070</td>\n",
       "      <td>301</td>\n",
       "      <td>261</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.270000</td>\n",
       "      <td>0.220408</td>\n",
       "      <td>0.222680</td>\n",
       "      <td>195</td>\n",
       "      <td>0.795918</td>\n",
       "      <td>210</td>\n",
       "      <td>0.875000</td>\n",
       "      <td>173</td>\n",
       "      <td>114387951</td>\n",
       "      <td>240</td>\n",
       "      <td>245</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.268908</td>\n",
       "      <td>0.232727</td>\n",
       "      <td>0.247582</td>\n",
       "      <td>239</td>\n",
       "      <td>0.869091</td>\n",
       "      <td>236</td>\n",
       "      <td>0.975207</td>\n",
       "      <td>191</td>\n",
       "      <td>157332201</td>\n",
       "      <td>242</td>\n",
       "      <td>275</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.497143</td>\n",
       "      <td>0.426471</td>\n",
       "      <td>0.429630</td>\n",
       "      <td>175</td>\n",
       "      <td>0.857843</td>\n",
       "      <td>174</td>\n",
       "      <td>0.865672</td>\n",
       "      <td>422</td>\n",
       "      <td>101047443</td>\n",
       "      <td>201</td>\n",
       "      <td>204</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17103</th>\n",
       "      <td>0.328000</td>\n",
       "      <td>0.278912</td>\n",
       "      <td>0.299817</td>\n",
       "      <td>278</td>\n",
       "      <td>0.945578</td>\n",
       "      <td>244</td>\n",
       "      <td>0.964427</td>\n",
       "      <td>166</td>\n",
       "      <td>136836774</td>\n",
       "      <td>253</td>\n",
       "      <td>294</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17104</th>\n",
       "      <td>0.284091</td>\n",
       "      <td>0.223214</td>\n",
       "      <td>0.245700</td>\n",
       "      <td>190</td>\n",
       "      <td>0.848214</td>\n",
       "      <td>175</td>\n",
       "      <td>0.956284</td>\n",
       "      <td>112</td>\n",
       "      <td>14456723</td>\n",
       "      <td>183</td>\n",
       "      <td>224</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17105</th>\n",
       "      <td>0.344828</td>\n",
       "      <td>0.169972</td>\n",
       "      <td>0.192308</td>\n",
       "      <td>180</td>\n",
       "      <td>0.509915</td>\n",
       "      <td>200</td>\n",
       "      <td>0.738007</td>\n",
       "      <td>53</td>\n",
       "      <td>78849058</td>\n",
       "      <td>271</td>\n",
       "      <td>353</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17106</th>\n",
       "      <td>0.352174</td>\n",
       "      <td>0.239645</td>\n",
       "      <td>0.242152</td>\n",
       "      <td>230</td>\n",
       "      <td>0.680473</td>\n",
       "      <td>227</td>\n",
       "      <td>0.685801</td>\n",
       "      <td>279</td>\n",
       "      <td>161110219</td>\n",
       "      <td>331</td>\n",
       "      <td>338</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17107</th>\n",
       "      <td>0.336207</td>\n",
       "      <td>0.184834</td>\n",
       "      <td>0.174497</td>\n",
       "      <td>113</td>\n",
       "      <td>0.535545</td>\n",
       "      <td>117</td>\n",
       "      <td>0.495763</td>\n",
       "      <td>165</td>\n",
       "      <td>143737378</td>\n",
       "      <td>236</td>\n",
       "      <td>211</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17108 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       local_gap_compressed_percent_id  scaled_local_query_percent_id  \\\n",
       "0                             0.275281                       0.182156   \n",
       "1                             0.322430                       0.264368   \n",
       "2                             0.270000                       0.220408   \n",
       "3                             0.268908                       0.232727   \n",
       "4                             0.497143                       0.426471   \n",
       "...                                ...                            ...   \n",
       "17103                         0.328000                       0.278912   \n",
       "17104                         0.284091                       0.223214   \n",
       "17105                         0.344828                       0.169972   \n",
       "17106                         0.352174                       0.239645   \n",
       "17107                         0.336207                       0.184834   \n",
       "\n",
       "       scaled_local_symmetric_percent_id  query_align_len  query_align_cov  \\\n",
       "0                               0.152174              185         0.687732   \n",
       "1                               0.245552              232         0.888889   \n",
       "2                               0.222680              195         0.795918   \n",
       "3                               0.247582              239         0.869091   \n",
       "4                               0.429630              175         0.857843   \n",
       "...                                  ...              ...              ...   \n",
       "17103                           0.299817              278         0.945578   \n",
       "17104                           0.245700              190         0.848214   \n",
       "17105                           0.192308              180         0.509915   \n",
       "17106                           0.242152              230         0.680473   \n",
       "17107                           0.174497              113         0.535545   \n",
       "\n",
       "       subject_align_len  subject_align_cov  bit_score  prot_pair_index  \\\n",
       "0                    197           0.525333         48         37128170   \n",
       "1                    208           0.691030        227        172639070   \n",
       "2                    210           0.875000        173        114387951   \n",
       "3                    236           0.975207        191        157332201   \n",
       "4                    174           0.865672        422        101047443   \n",
       "...                  ...                ...        ...              ...   \n",
       "17103                244           0.964427        166        136836774   \n",
       "17104                175           0.956284        112         14456723   \n",
       "17105                200           0.738007         53         78849058   \n",
       "17106                227           0.685801        279        161110219   \n",
       "17107                117           0.495763        165        143737378   \n",
       "\n",
       "       m_protein_len  t_protein_len  protein_match  \n",
       "0                375            269              0  \n",
       "1                301            261              1  \n",
       "2                240            245              1  \n",
       "3                242            275              1  \n",
       "4                201            204              1  \n",
       "...              ...            ...            ...  \n",
       "17103            253            294              1  \n",
       "17104            183            224              0  \n",
       "17105            271            353              0  \n",
       "17106            331            338              0  \n",
       "17107            236            211              0  \n",
       "\n",
       "[17108 rows x 12 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "03972032",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import accuracy_score, \\\n",
    "    confusion_matrix, \\\n",
    "    roc_auc_score,\\\n",
    "    average_precision_score,\\\n",
    "    auc,\\\n",
    "    roc_curve, f1_score, recall_score, matthews_corrcoef, auc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "7414184d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def evaluate_model(output_path, model, dataframe):\n",
    "    '''\n",
    "    Takes a trained model and test data and tests the model.\n",
    "\n",
    "    Args:\n",
    "        output path: File path: str\n",
    "        model: sklearn.neighbors.KNeighborsClassifier\n",
    "        test_X: numpy array\n",
    "        test_y: numpy array\n",
    "\n",
    "    Returns:\n",
    "        Vector of predictions (numpy arrray)\n",
    "    '''\n",
    "    # initialize empty eval results file\n",
    "    F = open('evaluationResults.txt', 'w')\n",
    "\n",
    "    F.write('Evaluation Scale:' + '\\n')\n",
    "    F.write('0.0% <=Accuracy<= 100.0%' + '\\n')\n",
    "    F.write('0.0 <=AUC<= 1.0' + '\\n')  # area under curve\n",
    "    F.write('0.0 <=auPR<= 1.0' + '\\n')  # average_Precision\n",
    "    F.write('0.0 <=F1_Score<= 1.0' + '\\n')\n",
    "    F.write('-1.0 <=MCC<= 1.0' + '\\n')\n",
    "    F.write('_______________________________________' + '\\n')\n",
    "\n",
    "    model = model\n",
    "\n",
    "    df_seqs = dataframe[['m_protein_seq', 't_protein_seq', 'prot_pair_index']]\n",
    "\n",
    "    dataframe = dataframe.drop(columns=['m_protein_seq', 't_protein_seq', 'prot_pair_index'])\n",
    "\n",
    "    target = 'protein_match'\n",
    "    features = [columns for columns in dataframe]\n",
    "    features.remove(target)\n",
    "    print(features)\n",
    "\n",
    "    # split into input and output feature(s)\n",
    "    test_X = dataframe[features].values\n",
    "    test_y = dataframe[target].values.reshape(-1, 1)\n",
    "\n",
    "    # scale data\n",
    "    scaler = sklearn.preprocessing.StandardScaler()\n",
    "    test_X = scaler.fit_transform(test_X)\n",
    "\n",
    "    accuracy = []\n",
    "    avg_precision = []\n",
    "    F1_Score = []\n",
    "    AUC = []\n",
    "    MCC = []\n",
    "    Recall = []\n",
    "\n",
    "    # test input arguments\n",
    "    assert \"sklearn\" in str(type(model))\n",
    "    assert \"numpy.ndarray\" in str(type(test_X))\n",
    "    assert \"numpy.ndarray\" in str(type(test_y))\n",
    "\n",
    "    # vector of predictions\n",
    "    preds = model.predict(test_X)\n",
    "\n",
    "    # calculate precision score\n",
    "    precision_score = sklearn.metrics.precision_score(test_y, preds)\n",
    "\n",
    "    # Calculate ROC Curve and Area the Curve\n",
    "    proba_y = model.predict_proba(test_X)[:, 1]\n",
    "    FPR, TPR, _ = roc_curve(test_y, proba_y, pos_label=1)\n",
    "    roc_auc = auc(FPR, TPR)\n",
    "\n",
    "    # calculate scoring metrics\n",
    "    # include option to return these scores\n",
    "    accuracy.append(accuracy_score(y_pred=preds, y_true=test_y))\n",
    "    avg_precision.append(\n",
    "        average_precision_score(\n",
    "            y_true=test_y,\n",
    "            y_score=proba_y,\n",
    "            pos_label=1))\n",
    "    F1_Score.append(f1_score(y_true=test_y, y_pred=preds, pos_label=1))\n",
    "    MCC.append(matthews_corrcoef(y_true=test_y, y_pred=preds))\n",
    "    Recall.append(recall_score(y_true=test_y, y_pred=preds, pos_label=1))\n",
    "    AUC.append(roc_auc)\n",
    "\n",
    "    confusion_matrix = sklearn.metrics.confusion_matrix(\n",
    "        y_pred=preds, y_true=test_y)\n",
    "    sklearn.metrics.ConfusionMatrixDisplay(confusion_matrix).plot()\n",
    "\n",
    "    accuracy = [_*100.0 for _ in accuracy]\n",
    "\n",
    "    F.write('Accuracy: {0:.4f}%\\n'.format(np.mean(accuracy)))\n",
    "    F.write('AUC: {0:.4f}\\n'.format(np.mean(AUC)))\n",
    "    F.write(\n",
    "        'auPR: {0:.4f}\\n'.format(\n",
    "            np.mean(avg_precision)))  # average_Precision\n",
    "    F.write('F1_Score: {0:.4f}\\n'.format(np.mean(F1_Score)))\n",
    "    F.write('MCC: {0:.4f}\\n'.format(np.mean(MCC)))\n",
    "\n",
    "#         TN, FP, FN, TP = CM.ravel()\n",
    "    F.write('Recall: {0:.4f}\\n'.format(np.mean(Recall)))\n",
    "    F.write('_______________________________________' + '\\n')\n",
    "\n",
    "    # merge dataframes together to report results\n",
    "    df_seqs['prediction'] = preds\n",
    "\n",
    "    # save to csv\n",
    "    df_seqs.to_csv('predictions.csv')\n",
    "\n",
    "    return preds, precision_score, df_seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "30ff6ff1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['local_gap_compressed_percent_id', 'scaled_local_query_percent_id',\n",
       "       'scaled_local_symmetric_percent_id', 'query_align_len',\n",
       "       'query_align_cov', 'subject_align_len', 'subject_align_cov',\n",
       "       'bit_score', 'prot_pair_index', 'm_protein_seq', 't_protein_seq',\n",
       "       'm_protein_len', 't_protein_len', 'protein_match'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a30efe71",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['local_gap_compressed_percent_id', 'scaled_local_query_percent_id', 'scaled_local_symmetric_percent_id', 'query_align_len', 'query_align_cov', 'subject_align_len', 'subject_align_cov', 'bit_score', 'm_protein_len', 't_protein_len']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-34-db1ae9a750aa>:99: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_seqs['prediction'] = preds\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(array([0, 0, 0, ..., 0, 0, 0]),\n",
       " 0.7725343320848939,\n",
       "                                            m_protein_seq  \\\n",
       " 0      MRAIGELWPTFDDVHEIAVLRGGGLGDLMFALPAIDALAAAYPEAR...   \n",
       " 1      MDSLITVRNLHKLFRVDQREDSGLWPAIKSLIRRERAEVVAVDGVS...   \n",
       " 2      MRKQYGDFHAVKGIDLDVRPGECFGLLGPNGAGKSTTMRMLAGTSQ...   \n",
       " 3      MAWTREQMAARAAKELQDGFYVNLGIGIPTLVANYIPDGVHVTLQS...   \n",
       " 4      MSATDTPSRPLASLDIKKEVERYALVLGWIFLIAIFGLLLPDTFLT...   \n",
       " ...                                                  ...   \n",
       " 17103  MRRQVGSVPGVTRTAGAAGGGAAARTAPGGPPDVVLMSCSHPRAAE...   \n",
       " 17104  MLTVSYVGAHKIEIREQDPEPVHAGQVQVEVAYAGICGTDLHILHG...   \n",
       " 17105  MDNATFRLGDDLSVRLPGHSRWIGQVEREQRWLPWLAPRLPLTVST...   \n",
       " 17106  MIRLAELTKTYPGQQHPAVDGISMEVAEGEIVVLVGPSGCGKTTTL...   \n",
       " 17107  MTGQGASGSETGVSDDSPITVVVVDDQELLRTGLRDLAEHDGDIAV...   \n",
       " \n",
       "                                            t_protein_seq  prot_pair_index  \\\n",
       " 0      MATTTAEIGVIGGSGFYSFLDDPHEVTVQTPYGPPSDPIAVGTVAG...         37128170   \n",
       " 1      MSEHILETDRLTRRYGDRLAVDAVSMAVRPREVYGFLGPNGAGKTT...         55279368   \n",
       " 2      MTNGHTDLSGSGLTLGYDRRIVSRDLDVSIPPGSFTVIIGPNACGK...        107256987   \n",
       " 3      MDMLVEAAPLDRHGMARRLAADIPEGWVVNLGIGIPTLVSDHVPPE...        121070359   \n",
       " 4      MDLLNAIVALLNFVIIPATAYGAQLALGALGVTMIYGILRFSNFAH...        131454663   \n",
       " ...                                                  ...              ...   \n",
       " 17103  MAESHSTTRSGQIRVFVLDDHEVVRRGVRDLLDAEPDITVVGEAET...         14456723   \n",
       " 17104  MRAVTISEPGGPEKLQWTEVPDPRPGAGEVLLEVVASAVNRADVLQ...         46032974   \n",
       " 17105  MPPQPRPLRPNDPREIGGFALLGRLGEGGQGTVYLGGAPDGRRVAV...         78849058   \n",
       " 17106  MTEQPILSARGLTVDFRLRGGRRARAVDGVDLDLAPGEVLALAGES...        161110219   \n",
       " 17107  MTTEAQPRRVVLAEDEALIRLDLAEMLAEEGYEVVGQAADGETALA...        143737378   \n",
       " \n",
       "        prediction  \n",
       " 0               0  \n",
       " 1               0  \n",
       " 2               0  \n",
       " 3               1  \n",
       " 4               0  \n",
       " ...           ...  \n",
       " 17103           1  \n",
       " 17104           1  \n",
       " 17105           0  \n",
       " 17106           0  \n",
       " 17107           0  \n",
       " \n",
       " [17108 rows x 4 columns])"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "evaluate_model('notebooks', model, df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "fd0dc8aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>prot_pair_index</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "      <th>protein_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.275281</td>\n",
       "      <td>0.182156</td>\n",
       "      <td>0.152174</td>\n",
       "      <td>185</td>\n",
       "      <td>0.687732</td>\n",
       "      <td>197</td>\n",
       "      <td>0.525333</td>\n",
       "      <td>48</td>\n",
       "      <td>37128170</td>\n",
       "      <td>375</td>\n",
       "      <td>269</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.322430</td>\n",
       "      <td>0.264368</td>\n",
       "      <td>0.245552</td>\n",
       "      <td>232</td>\n",
       "      <td>0.888889</td>\n",
       "      <td>208</td>\n",
       "      <td>0.691030</td>\n",
       "      <td>227</td>\n",
       "      <td>172639070</td>\n",
       "      <td>301</td>\n",
       "      <td>261</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.270000</td>\n",
       "      <td>0.220408</td>\n",
       "      <td>0.222680</td>\n",
       "      <td>195</td>\n",
       "      <td>0.795918</td>\n",
       "      <td>210</td>\n",
       "      <td>0.875000</td>\n",
       "      <td>173</td>\n",
       "      <td>114387951</td>\n",
       "      <td>240</td>\n",
       "      <td>245</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.268908</td>\n",
       "      <td>0.232727</td>\n",
       "      <td>0.247582</td>\n",
       "      <td>239</td>\n",
       "      <td>0.869091</td>\n",
       "      <td>236</td>\n",
       "      <td>0.975207</td>\n",
       "      <td>191</td>\n",
       "      <td>157332201</td>\n",
       "      <td>242</td>\n",
       "      <td>275</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.497143</td>\n",
       "      <td>0.426471</td>\n",
       "      <td>0.429630</td>\n",
       "      <td>175</td>\n",
       "      <td>0.857843</td>\n",
       "      <td>174</td>\n",
       "      <td>0.865672</td>\n",
       "      <td>422</td>\n",
       "      <td>101047443</td>\n",
       "      <td>201</td>\n",
       "      <td>204</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17103</th>\n",
       "      <td>0.328000</td>\n",
       "      <td>0.278912</td>\n",
       "      <td>0.299817</td>\n",
       "      <td>278</td>\n",
       "      <td>0.945578</td>\n",
       "      <td>244</td>\n",
       "      <td>0.964427</td>\n",
       "      <td>166</td>\n",
       "      <td>136836774</td>\n",
       "      <td>253</td>\n",
       "      <td>294</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17104</th>\n",
       "      <td>0.284091</td>\n",
       "      <td>0.223214</td>\n",
       "      <td>0.245700</td>\n",
       "      <td>190</td>\n",
       "      <td>0.848214</td>\n",
       "      <td>175</td>\n",
       "      <td>0.956284</td>\n",
       "      <td>112</td>\n",
       "      <td>14456723</td>\n",
       "      <td>183</td>\n",
       "      <td>224</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17105</th>\n",
       "      <td>0.344828</td>\n",
       "      <td>0.169972</td>\n",
       "      <td>0.192308</td>\n",
       "      <td>180</td>\n",
       "      <td>0.509915</td>\n",
       "      <td>200</td>\n",
       "      <td>0.738007</td>\n",
       "      <td>53</td>\n",
       "      <td>78849058</td>\n",
       "      <td>271</td>\n",
       "      <td>353</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17106</th>\n",
       "      <td>0.352174</td>\n",
       "      <td>0.239645</td>\n",
       "      <td>0.242152</td>\n",
       "      <td>230</td>\n",
       "      <td>0.680473</td>\n",
       "      <td>227</td>\n",
       "      <td>0.685801</td>\n",
       "      <td>279</td>\n",
       "      <td>161110219</td>\n",
       "      <td>331</td>\n",
       "      <td>338</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17107</th>\n",
       "      <td>0.336207</td>\n",
       "      <td>0.184834</td>\n",
       "      <td>0.174497</td>\n",
       "      <td>113</td>\n",
       "      <td>0.535545</td>\n",
       "      <td>117</td>\n",
       "      <td>0.495763</td>\n",
       "      <td>165</td>\n",
       "      <td>143737378</td>\n",
       "      <td>236</td>\n",
       "      <td>211</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17108 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       local_gap_compressed_percent_id  scaled_local_query_percent_id  \\\n",
       "0                             0.275281                       0.182156   \n",
       "1                             0.322430                       0.264368   \n",
       "2                             0.270000                       0.220408   \n",
       "3                             0.268908                       0.232727   \n",
       "4                             0.497143                       0.426471   \n",
       "...                                ...                            ...   \n",
       "17103                         0.328000                       0.278912   \n",
       "17104                         0.284091                       0.223214   \n",
       "17105                         0.344828                       0.169972   \n",
       "17106                         0.352174                       0.239645   \n",
       "17107                         0.336207                       0.184834   \n",
       "\n",
       "       scaled_local_symmetric_percent_id  query_align_len  query_align_cov  \\\n",
       "0                               0.152174              185         0.687732   \n",
       "1                               0.245552              232         0.888889   \n",
       "2                               0.222680              195         0.795918   \n",
       "3                               0.247582              239         0.869091   \n",
       "4                               0.429630              175         0.857843   \n",
       "...                                  ...              ...              ...   \n",
       "17103                           0.299817              278         0.945578   \n",
       "17104                           0.245700              190         0.848214   \n",
       "17105                           0.192308              180         0.509915   \n",
       "17106                           0.242152              230         0.680473   \n",
       "17107                           0.174497              113         0.535545   \n",
       "\n",
       "       subject_align_len  subject_align_cov  bit_score  prot_pair_index  \\\n",
       "0                    197           0.525333         48         37128170   \n",
       "1                    208           0.691030        227        172639070   \n",
       "2                    210           0.875000        173        114387951   \n",
       "3                    236           0.975207        191        157332201   \n",
       "4                    174           0.865672        422        101047443   \n",
       "...                  ...                ...        ...              ...   \n",
       "17103                244           0.964427        166        136836774   \n",
       "17104                175           0.956284        112         14456723   \n",
       "17105                200           0.738007         53         78849058   \n",
       "17106                227           0.685801        279        161110219   \n",
       "17107                117           0.495763        165        143737378   \n",
       "\n",
       "       m_protein_len  t_protein_len  protein_match  \n",
       "0                375            269              0  \n",
       "1                301            261              1  \n",
       "2                240            245              1  \n",
       "3                242            275              1  \n",
       "4                201            204              1  \n",
       "...              ...            ...            ...  \n",
       "17103            253            294              1  \n",
       "17104            183            224              0  \n",
       "17105            271            353              0  \n",
       "17106            331            338              0  \n",
       "17107            236            211              0  \n",
       "\n",
       "[17108 rows x 12 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5add2339",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
