{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "265e8129",
   "metadata": {},
   "source": [
    "This notebook uses iFeatureOmega, a feature generation software, to add to our feature space for a RandomForestClassifier that predicts protein pair functionality."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4afbc58f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats\n",
    "import sklearn.preprocessing\n",
    "import sklearn.model_selection\n",
    "import sklearn.ensemble\n",
    "import sklearn.feature_selection\n",
    "import unittest\n",
    "import iFeatureOmegaCLI\n",
    "import Bio.SeqIO\n",
    "import Bio.SeqRecord\n",
    "import io\n",
    "from io import StringIO\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ba87d47f",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/loganroberts/Learn2Therm/ValidProt/FAFSA\n"
     ]
    }
   ],
   "source": [
    "cd /Users/loganroberts/Learn2Therm/ValidProt/FAFSA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f90c1cdd",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Unnamed: 0', 'local_gap_compressed_percent_id',\n",
       "       'scaled_local_query_percent_id', 'scaled_local_symmetric_percent_id',\n",
       "       'query_align_len', 'query_align_cov', 'subject_align_len',\n",
       "       'subject_align_cov', 'bit_score', 'thermo_index', 'meso_index',\n",
       "       'prot_pair_index', 'meso_protein_int_index', 'thermo_protein_int_index',\n",
       "       'taxa_pair_index', 'local_gap_compressed_percent_id_16s',\n",
       "       'scaled_local_query_percent_id_16s',\n",
       "       'scaled_local_symmetric_percent_id_16s', 'query_align_cov_16s',\n",
       "       'subject_align_cov_16s', 'bit_score_16s', 'm_ogt', 't_ogt',\n",
       "       'ogt_difference', 'm_protein_seq', 't_protein_seq', 'm_protein_desc',\n",
       "       't_protein_desc', 'm_protein_len', 't_protein_len'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#convert to pandas df\n",
    "df = pd.read_csv('learn2therm_sample_50k.csv')\n",
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "99990cd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_fasta_from_dataframe(dataframe, output_file_a, output_file_b):\n",
    "    #adjust this to write function with BioPython\n",
    "    #separate functions for each of the input sequences\n",
    "    #in training, seq_a = meso and seq_b = thermo\n",
    "    \n",
    "    \n",
    "    #meso sequence to fasta\n",
    "    with open(output_file_a, 'w') as f:\n",
    "        for _, row in df.iterrows():\n",
    "            f.write('>{}\\n{}\\n'.format((row['prot_pair_index']), row['m_protein_seq']))\n",
    "    \n",
    "    #thermo sequence to fasta\n",
    "    with open(output_file_b, 'w') as f:\n",
    "        for _, row in df.iterrows():\n",
    "            f.write('>{}\\n{}\\n'.format((row['prot_pair_index']), (row['t_protein_seq'])))\n",
    "   \n",
    "    #return output files\n",
    "    return [output_file_a, output_file_b]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "653eb96d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_protein_descriptors(fasta_file, descriptors=[]):\n",
    "    \n",
    "    \"\"\"\n",
    "    Generates features from a protein sequence\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    Fasta file with amino acid sequences.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    Vector of descriptors\n",
    "    \"\"\"\n",
    "    \n",
    "    #create iProtein object\n",
    "    protein = iFeatureOmegaCLI.iProtein(fasta_file)\n",
    "    \n",
    "    #not sure why we need this yet. Right now it is stored in local directory.\n",
    "    params = protein.import_parameters('protein_parameters.json')\n",
    "    \n",
    "    protein_descriptors = {}\n",
    "    \n",
    "    for descriptor in descriptors:\n",
    "        protein.get_descriptor(descriptor)\n",
    "        protein_descriptors.update({f'{descriptor}':protein.encodings})\n",
    "        \n",
    "    return protein_descriptors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fe48f193",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_new_dataframe(dataframe, output_files, descriptors=[]):\n",
    "    \"\"\"\n",
    "    Creates new dataframe with descriptors added.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    Pandas dataframe, list of descriptors as strings, output file name.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    Dataframe including vector(s) of descriptors\n",
    "    \"\"\"\n",
    "\n",
    "    fasta_files = get_fasta_from_dataframe(dataframe, output_files[0], output_files[1])\n",
    "    \n",
    "    \n",
    "    feature_dict_a = get_protein_descriptors(fasta_files[0], descriptors)\n",
    "    \n",
    "    feature_dict_b = get_protein_descriptors(fasta_files[1], descriptors)\n",
    "\n",
    "\n",
    "    df = dataframe.reset_index()\n",
    "\n",
    "    for desc in descriptors:\n",
    "\n",
    "        feature_dict_a[desc].index = feature_dict_a[desc].index.astype(int)\n",
    "        features_a = feature_dict_a[desc].reset_index()\n",
    "        \n",
    "        feature_dict_b[desc].index = feature_dict_b[desc].index.astype(int)\n",
    "        features_b = feature_dict_b[desc].reset_index()\n",
    "    \n",
    "    feature_df = pd.merge(\n",
    "            df,\n",
    "            features_a,\n",
    "            how='outer',\n",
    "            left_index=True,\n",
    "            right_index=True)\n",
    "    \n",
    "    new_df = pd.merge(\n",
    "            feature_df,\n",
    "            features_b,\n",
    "            how='outer',\n",
    "            left_index=True,\n",
    "            right_index=True)\n",
    "\n",
    "    return new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d7f0a563",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/loganroberts/Learn2Therm/ValidProt/notebooks\n"
     ]
    }
   ],
   "source": [
    "cd /Users/loganroberts/Learn2Therm/ValidProt/notebooks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "043bf4dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>prot_pair_index</th>\n",
       "      <th>protein_match</th>\n",
       "      <th>Jaccard_Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>48641291</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>92992745</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>157628663</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>136708305</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>133672542</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48845</th>\n",
       "      <td>4875</td>\n",
       "      <td>78849058</td>\n",
       "      <td>0</td>\n",
       "      <td>0.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48846</th>\n",
       "      <td>4876</td>\n",
       "      <td>108797464</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48847</th>\n",
       "      <td>4877</td>\n",
       "      <td>161110219</td>\n",
       "      <td>0</td>\n",
       "      <td>0.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48848</th>\n",
       "      <td>4878</td>\n",
       "      <td>74177185</td>\n",
       "      <td>1</td>\n",
       "      <td>0.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48849</th>\n",
       "      <td>4879</td>\n",
       "      <td>143737378</td>\n",
       "      <td>0</td>\n",
       "      <td>0.25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>48850 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0  prot_pair_index  protein_match  Jaccard_Score\n",
       "0               0         48641291              1           1.00\n",
       "1               1         92992745              1           1.00\n",
       "2               2        157628663              1           1.00\n",
       "3               3        136708305              1           1.00\n",
       "4               4        133672542              1           1.00\n",
       "...           ...              ...            ...            ...\n",
       "48845        4875         78849058              0           0.25\n",
       "48846        4876        108797464              1           1.00\n",
       "48847        4877        161110219              0           0.25\n",
       "48848        4878         74177185              1           0.50\n",
       "48849        4879        143737378              0           0.25\n",
       "\n",
       "[48850 rows x 4 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target = pd.read_csv('protein_match_50k')\n",
    "target['protein_match'] = target['protein_match'].map({'Yes': 1, 'No': 0})\n",
    "target"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94da64e9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b2d3db4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.utils import resample\n",
    "\n",
    "# Assuming your data is in a pandas DataFrame called 'data'\n",
    "# Separate the majority and minority classes\n",
    "majority_class = target[target['protein_match'] == 1]\n",
    "minority_class = target[target['protein_match'] == 0]\n",
    "\n",
    "# Undersample the majority class to match the number of minority class samples\n",
    "n_samples = len(minority_class)\n",
    "undersampled_majority = resample(majority_class, n_samples=n_samples, replace=False)\n",
    "\n",
    "# Combine the undersampled majority class with the minority class\n",
    "balanced_data = pd.concat([undersampled_majority, minority_class])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "60de9006",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>prot_pair_index</th>\n",
       "      <th>protein_match</th>\n",
       "      <th>Jaccard_Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15519</th>\n",
       "      <td>865</td>\n",
       "      <td>45982583</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25031</th>\n",
       "      <td>623</td>\n",
       "      <td>65376717</td>\n",
       "      <td>1</td>\n",
       "      <td>0.600000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20705</th>\n",
       "      <td>1168</td>\n",
       "      <td>96763833</td>\n",
       "      <td>1</td>\n",
       "      <td>0.400000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48450</th>\n",
       "      <td>4480</td>\n",
       "      <td>16236144</td>\n",
       "      <td>1</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47129</th>\n",
       "      <td>3159</td>\n",
       "      <td>151972561</td>\n",
       "      <td>1</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48833</th>\n",
       "      <td>4863</td>\n",
       "      <td>70996712</td>\n",
       "      <td>0</td>\n",
       "      <td>0.285714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48839</th>\n",
       "      <td>4869</td>\n",
       "      <td>14456723</td>\n",
       "      <td>0</td>\n",
       "      <td>0.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48845</th>\n",
       "      <td>4875</td>\n",
       "      <td>78849058</td>\n",
       "      <td>0</td>\n",
       "      <td>0.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48847</th>\n",
       "      <td>4877</td>\n",
       "      <td>161110219</td>\n",
       "      <td>0</td>\n",
       "      <td>0.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48849</th>\n",
       "      <td>4879</td>\n",
       "      <td>143737378</td>\n",
       "      <td>0</td>\n",
       "      <td>0.250000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17108 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0  prot_pair_index  protein_match  Jaccard_Score\n",
       "15519         865         45982583              1       1.000000\n",
       "25031         623         65376717              1       0.600000\n",
       "20705        1168         96763833              1       0.400000\n",
       "48450        4480         16236144              1       0.500000\n",
       "47129        3159        151972561              1       0.500000\n",
       "...           ...              ...            ...            ...\n",
       "48833        4863         70996712              0       0.285714\n",
       "48839        4869         14456723              0       0.250000\n",
       "48845        4875         78849058              0       0.250000\n",
       "48847        4877        161110219              0       0.250000\n",
       "48849        4879        143737378              0       0.250000\n",
       "\n",
       "[17108 rows x 4 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "balanced_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "792508bb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(17108, 33)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.merge(df, balanced_data, on=['prot_pair_index'])\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "62efbea0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0_x</th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>thermo_index</th>\n",
       "      <th>...</th>\n",
       "      <th>ogt_difference</th>\n",
       "      <th>m_protein_seq</th>\n",
       "      <th>t_protein_seq</th>\n",
       "      <th>m_protein_desc</th>\n",
       "      <th>t_protein_desc</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "      <th>Unnamed: 0_y</th>\n",
       "      <th>protein_match</th>\n",
       "      <th>Jaccard_Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>0.327273</td>\n",
       "      <td>0.200743</td>\n",
       "      <td>0.214712</td>\n",
       "      <td>166</td>\n",
       "      <td>0.617100</td>\n",
       "      <td>163</td>\n",
       "      <td>0.696581</td>\n",
       "      <td>175</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MTSGLWERVLDGVWVTIQLLVLSALLATAVSFVVGIARTHRLWIVR...</td>\n",
       "      <td>MAMSRRKRGQLARGIQYAILVIVVVVLALLADWGKIGKAFFDWEAA...</td>\n",
       "      <td>ectoine/hydroxyectoine ABC transporter permeas...</td>\n",
       "      <td>amino acid ABC transporter permease</td>\n",
       "      <td>234</td>\n",
       "      <td>269</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6</td>\n",
       "      <td>0.283582</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.240506</td>\n",
       "      <td>133</td>\n",
       "      <td>0.810976</td>\n",
       "      <td>134</td>\n",
       "      <td>0.881579</td>\n",
       "      <td>99</td>\n",
       "      <td>7134</td>\n",
       "      <td>...</td>\n",
       "      <td>24.5</td>\n",
       "      <td>MGSTDRPDLAAMLAPLTRTLIAMERPVLETYGLTMWAYSVLVALSR...</td>\n",
       "      <td>MGHKDELIARIQRAEREFRRSVVSQAATDFFSVDLTMPQLRVVFFL...</td>\n",
       "      <td>MarR family transcriptional regulator</td>\n",
       "      <td>MarR family transcriptional regulator</td>\n",
       "      <td>152</td>\n",
       "      <td>164</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.295082</td>\n",
       "      <td>0.291498</td>\n",
       "      <td>218</td>\n",
       "      <td>0.893443</td>\n",
       "      <td>221</td>\n",
       "      <td>0.884000</td>\n",
       "      <td>158</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>24.0</td>\n",
       "      <td>MQIKDSVAVVTGGASGLGLATTKRLLDAGGSVVVIDLKGEDVVAEL...</td>\n",
       "      <td>MTGTVVITGGSRGIGAATAVLAAERGWQVAVSFRERRDAAEQVVRR...</td>\n",
       "      <td>3-hydroxyacyl-CoA dehydrogenase</td>\n",
       "      <td>SDR family oxidoreductase</td>\n",
       "      <td>250</td>\n",
       "      <td>244</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>0.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11</td>\n",
       "      <td>0.275281</td>\n",
       "      <td>0.182156</td>\n",
       "      <td>0.152174</td>\n",
       "      <td>185</td>\n",
       "      <td>0.687732</td>\n",
       "      <td>197</td>\n",
       "      <td>0.525333</td>\n",
       "      <td>48</td>\n",
       "      <td>14963</td>\n",
       "      <td>...</td>\n",
       "      <td>31.5</td>\n",
       "      <td>MRAIGELWPTFDDVHEIAVLRGGGLGDLMFALPAIDALAAAYPEAR...</td>\n",
       "      <td>MATTTAEIGVIGGSGFYSFLDDPHEVTVQTPYGPPSDPIAVGTVAG...</td>\n",
       "      <td>glycosyltransferase family 9 protein</td>\n",
       "      <td>S-methyl-5'-thioadenosine phosphorylase</td>\n",
       "      <td>375</td>\n",
       "      <td>269</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>18</td>\n",
       "      <td>0.394495</td>\n",
       "      <td>0.346774</td>\n",
       "      <td>0.369099</td>\n",
       "      <td>108</td>\n",
       "      <td>0.870968</td>\n",
       "      <td>106</td>\n",
       "      <td>0.972477</td>\n",
       "      <td>130</td>\n",
       "      <td>14963</td>\n",
       "      <td>...</td>\n",
       "      <td>24.5</td>\n",
       "      <td>MHELSIAQSVVEAVCERADGRRVRSVRMVVGALCAVVPDSMLFCFE...</td>\n",
       "      <td>MGLCEAIVDAVLQRAEGRRVRSVRVRVAGHPVVREVVDQGFAMAAA...</td>\n",
       "      <td>hydrogenase maturation nickel metallochaperone...</td>\n",
       "      <td>hydrogenase maturation nickel metallochaperone...</td>\n",
       "      <td>109</td>\n",
       "      <td>124</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17103</th>\n",
       "      <td>49992</td>\n",
       "      <td>0.331707</td>\n",
       "      <td>0.294372</td>\n",
       "      <td>0.282158</td>\n",
       "      <td>202</td>\n",
       "      <td>0.874459</td>\n",
       "      <td>218</td>\n",
       "      <td>0.868526</td>\n",
       "      <td>232</td>\n",
       "      <td>11324</td>\n",
       "      <td>...</td>\n",
       "      <td>24.0</td>\n",
       "      <td>MSTLVADTVSKRFGGVTALSNVSLELREGEIHGLIGPNGSGKTTLL...</td>\n",
       "      <td>MLEVKNIDLFYGASRALRSVSLTAQAGQVTAVMGRNGVGKTSLIRA...</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>urea ABC transporter ATP-binding subunit UrtE</td>\n",
       "      <td>251</td>\n",
       "      <td>231</td>\n",
       "      <td>4872</td>\n",
       "      <td>1</td>\n",
       "      <td>0.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17104</th>\n",
       "      <td>49993</td>\n",
       "      <td>0.304878</td>\n",
       "      <td>0.188679</td>\n",
       "      <td>0.163666</td>\n",
       "      <td>166</td>\n",
       "      <td>0.626415</td>\n",
       "      <td>170</td>\n",
       "      <td>0.491329</td>\n",
       "      <td>125</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MQVGDATVVITGASSGIGRATALAFADAGANLVLTARAKNPLRAAE...</td>\n",
       "      <td>MGRLSGKRAVVTGGATLIGQAVVRTLHREGARVAIADIDEAGAKAV...</td>\n",
       "      <td>SDR family oxidoreductase</td>\n",
       "      <td>SDR family oxidoreductase</td>\n",
       "      <td>346</td>\n",
       "      <td>265</td>\n",
       "      <td>4873</td>\n",
       "      <td>1</td>\n",
       "      <td>0.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17105</th>\n",
       "      <td>49995</td>\n",
       "      <td>0.344828</td>\n",
       "      <td>0.169972</td>\n",
       "      <td>0.192308</td>\n",
       "      <td>180</td>\n",
       "      <td>0.509915</td>\n",
       "      <td>200</td>\n",
       "      <td>0.738007</td>\n",
       "      <td>53</td>\n",
       "      <td>14963</td>\n",
       "      <td>...</td>\n",
       "      <td>24.5</td>\n",
       "      <td>MDNATFRLGDDLSVRLPGHSRWIGQVEREQRWLPWLAPRLPLTVST...</td>\n",
       "      <td>MPPQPRPLRPNDPREIGGFALLGRLGEGGQGTVYLGGAPDGRRVAV...</td>\n",
       "      <td>aminoglycoside phosphotransferase family protein</td>\n",
       "      <td>serine/threonine protein kinase</td>\n",
       "      <td>271</td>\n",
       "      <td>353</td>\n",
       "      <td>4875</td>\n",
       "      <td>0</td>\n",
       "      <td>0.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17106</th>\n",
       "      <td>49997</td>\n",
       "      <td>0.352174</td>\n",
       "      <td>0.239645</td>\n",
       "      <td>0.242152</td>\n",
       "      <td>230</td>\n",
       "      <td>0.680473</td>\n",
       "      <td>227</td>\n",
       "      <td>0.685801</td>\n",
       "      <td>279</td>\n",
       "      <td>7134</td>\n",
       "      <td>...</td>\n",
       "      <td>24.5</td>\n",
       "      <td>MIRLAELTKTYPGQQHPAVDGISMEVAEGEIVVLVGPSGCGKTTTL...</td>\n",
       "      <td>MTEQPILSARGLTVDFRLRGGRRARAVDGVDLDLAPGEVLALAGES...</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>331</td>\n",
       "      <td>338</td>\n",
       "      <td>4877</td>\n",
       "      <td>0</td>\n",
       "      <td>0.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17107</th>\n",
       "      <td>49999</td>\n",
       "      <td>0.336207</td>\n",
       "      <td>0.184834</td>\n",
       "      <td>0.174497</td>\n",
       "      <td>113</td>\n",
       "      <td>0.535545</td>\n",
       "      <td>117</td>\n",
       "      <td>0.495763</td>\n",
       "      <td>165</td>\n",
       "      <td>7134</td>\n",
       "      <td>...</td>\n",
       "      <td>26.5</td>\n",
       "      <td>MTGQGASGSETGVSDDSPITVVVVDDQELLRTGLRDLAEHDGDIAV...</td>\n",
       "      <td>MTTEAQPRRVVLAEDEALIRLDLAEMLAEEGYEVVGQAADGETALA...</td>\n",
       "      <td>response regulator transcription factor</td>\n",
       "      <td>response regulator</td>\n",
       "      <td>236</td>\n",
       "      <td>211</td>\n",
       "      <td>4879</td>\n",
       "      <td>0</td>\n",
       "      <td>0.25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17108 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0_x  local_gap_compressed_percent_id  \\\n",
       "0                 3                         0.327273   \n",
       "1                 6                         0.283582   \n",
       "2                 9                         0.333333   \n",
       "3                11                         0.275281   \n",
       "4                18                         0.394495   \n",
       "...             ...                              ...   \n",
       "17103         49992                         0.331707   \n",
       "17104         49993                         0.304878   \n",
       "17105         49995                         0.344828   \n",
       "17106         49997                         0.352174   \n",
       "17107         49999                         0.336207   \n",
       "\n",
       "       scaled_local_query_percent_id  scaled_local_symmetric_percent_id  \\\n",
       "0                           0.200743                           0.214712   \n",
       "1                           0.231707                           0.240506   \n",
       "2                           0.295082                           0.291498   \n",
       "3                           0.182156                           0.152174   \n",
       "4                           0.346774                           0.369099   \n",
       "...                              ...                                ...   \n",
       "17103                       0.294372                           0.282158   \n",
       "17104                       0.188679                           0.163666   \n",
       "17105                       0.169972                           0.192308   \n",
       "17106                       0.239645                           0.242152   \n",
       "17107                       0.184834                           0.174497   \n",
       "\n",
       "       query_align_len  query_align_cov  subject_align_len  subject_align_cov  \\\n",
       "0                  166         0.617100                163           0.696581   \n",
       "1                  133         0.810976                134           0.881579   \n",
       "2                  218         0.893443                221           0.884000   \n",
       "3                  185         0.687732                197           0.525333   \n",
       "4                  108         0.870968                106           0.972477   \n",
       "...                ...              ...                ...                ...   \n",
       "17103              202         0.874459                218           0.868526   \n",
       "17104              166         0.626415                170           0.491329   \n",
       "17105              180         0.509915                200           0.738007   \n",
       "17106              230         0.680473                227           0.685801   \n",
       "17107              113         0.535545                117           0.495763   \n",
       "\n",
       "       bit_score  thermo_index  ...  ogt_difference  \\\n",
       "0            175           875  ...            22.0   \n",
       "1             99          7134  ...            24.5   \n",
       "2            158           875  ...            24.0   \n",
       "3             48         14963  ...            31.5   \n",
       "4            130         14963  ...            24.5   \n",
       "...          ...           ...  ...             ...   \n",
       "17103        232         11324  ...            24.0   \n",
       "17104        125           875  ...            22.0   \n",
       "17105         53         14963  ...            24.5   \n",
       "17106        279          7134  ...            24.5   \n",
       "17107        165          7134  ...            26.5   \n",
       "\n",
       "                                           m_protein_seq  \\\n",
       "0      MTSGLWERVLDGVWVTIQLLVLSALLATAVSFVVGIARTHRLWIVR...   \n",
       "1      MGSTDRPDLAAMLAPLTRTLIAMERPVLETYGLTMWAYSVLVALSR...   \n",
       "2      MQIKDSVAVVTGGASGLGLATTKRLLDAGGSVVVIDLKGEDVVAEL...   \n",
       "3      MRAIGELWPTFDDVHEIAVLRGGGLGDLMFALPAIDALAAAYPEAR...   \n",
       "4      MHELSIAQSVVEAVCERADGRRVRSVRMVVGALCAVVPDSMLFCFE...   \n",
       "...                                                  ...   \n",
       "17103  MSTLVADTVSKRFGGVTALSNVSLELREGEIHGLIGPNGSGKTTLL...   \n",
       "17104  MQVGDATVVITGASSGIGRATALAFADAGANLVLTARAKNPLRAAE...   \n",
       "17105  MDNATFRLGDDLSVRLPGHSRWIGQVEREQRWLPWLAPRLPLTVST...   \n",
       "17106  MIRLAELTKTYPGQQHPAVDGISMEVAEGEIVVLVGPSGCGKTTTL...   \n",
       "17107  MTGQGASGSETGVSDDSPITVVVVDDQELLRTGLRDLAEHDGDIAV...   \n",
       "\n",
       "                                           t_protein_seq  \\\n",
       "0      MAMSRRKRGQLARGIQYAILVIVVVVLALLADWGKIGKAFFDWEAA...   \n",
       "1      MGHKDELIARIQRAEREFRRSVVSQAATDFFSVDLTMPQLRVVFFL...   \n",
       "2      MTGTVVITGGSRGIGAATAVLAAERGWQVAVSFRERRDAAEQVVRR...   \n",
       "3      MATTTAEIGVIGGSGFYSFLDDPHEVTVQTPYGPPSDPIAVGTVAG...   \n",
       "4      MGLCEAIVDAVLQRAEGRRVRSVRVRVAGHPVVREVVDQGFAMAAA...   \n",
       "...                                                  ...   \n",
       "17103  MLEVKNIDLFYGASRALRSVSLTAQAGQVTAVMGRNGVGKTSLIRA...   \n",
       "17104  MGRLSGKRAVVTGGATLIGQAVVRTLHREGARVAIADIDEAGAKAV...   \n",
       "17105  MPPQPRPLRPNDPREIGGFALLGRLGEGGQGTVYLGGAPDGRRVAV...   \n",
       "17106  MTEQPILSARGLTVDFRLRGGRRARAVDGVDLDLAPGEVLALAGES...   \n",
       "17107  MTTEAQPRRVVLAEDEALIRLDLAEMLAEEGYEVVGQAADGETALA...   \n",
       "\n",
       "                                          m_protein_desc  \\\n",
       "0      ectoine/hydroxyectoine ABC transporter permeas...   \n",
       "1                  MarR family transcriptional regulator   \n",
       "2                        3-hydroxyacyl-CoA dehydrogenase   \n",
       "3                   glycosyltransferase family 9 protein   \n",
       "4      hydrogenase maturation nickel metallochaperone...   \n",
       "...                                                  ...   \n",
       "17103                ABC transporter ATP-binding protein   \n",
       "17104                          SDR family oxidoreductase   \n",
       "17105   aminoglycoside phosphotransferase family protein   \n",
       "17106                ABC transporter ATP-binding protein   \n",
       "17107            response regulator transcription factor   \n",
       "\n",
       "                                          t_protein_desc  m_protein_len  \\\n",
       "0                    amino acid ABC transporter permease            234   \n",
       "1                  MarR family transcriptional regulator            152   \n",
       "2                              SDR family oxidoreductase            250   \n",
       "3                S-methyl-5'-thioadenosine phosphorylase            375   \n",
       "4      hydrogenase maturation nickel metallochaperone...            109   \n",
       "...                                                  ...            ...   \n",
       "17103      urea ABC transporter ATP-binding subunit UrtE            251   \n",
       "17104                          SDR family oxidoreductase            346   \n",
       "17105                    serine/threonine protein kinase            271   \n",
       "17106                ABC transporter ATP-binding protein            331   \n",
       "17107                                 response regulator            236   \n",
       "\n",
       "       t_protein_len  Unnamed: 0_y  protein_match  Jaccard_Score  \n",
       "0                269             3              1           1.00  \n",
       "1                164             6              1           1.00  \n",
       "2                244             9              1           0.75  \n",
       "3                269            11              0           0.00  \n",
       "4                124            18              1           1.00  \n",
       "...              ...           ...            ...            ...  \n",
       "17103            231          4872              1           0.50  \n",
       "17104            265          4873              1           0.60  \n",
       "17105            353          4875              0           0.25  \n",
       "17106            338          4877              0           0.25  \n",
       "17107            211          4879              0           0.25  \n",
       "\n",
       "[17108 rows x 33 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6519dcaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#drop columns that don't exihibit signficant pearson correlation with bit_score\n",
    "\n",
    "df = df.drop(columns = ['meso_index', 'meso_protein_int_index', 'local_gap_compressed_percent_id_16s', \n",
    "                        'scaled_local_query_percent_id_16s', 'scaled_local_symmetric_percent_id_16s',\n",
    "                       'bit_score_16s', 'm_ogt', 't_ogt', 'taxa_pair_index', 'thermo_protein_int_index'\n",
    "                       , 'prot_pair_index', 'ogt_difference', 'Jaccard_Score',\n",
    "                       'query_align_cov_16s', 'subject_align_cov_16s',\n",
    "                       'Unnamed: 0_x', 'Unnamed: 0_y', 'thermo_index','m_protein_seq', 't_protein_seq', \n",
    "                       'm_protein_desc', 't_protein_desc'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "16f0544e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['local_gap_compressed_percent_id', 'scaled_local_query_percent_id',\n",
       "       'scaled_local_symmetric_percent_id', 'query_align_len',\n",
       "       'query_align_cov', 'subject_align_len', 'subject_align_cov',\n",
       "       'bit_score', 'm_protein_len', 't_protein_len', 'protein_match'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6894895e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.replace([np.inf, -np.inf], np.nan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8004811b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "local_gap_compressed_percent_id      0\n",
      "scaled_local_query_percent_id        0\n",
      "scaled_local_symmetric_percent_id    0\n",
      "query_align_len                      0\n",
      "query_align_cov                      0\n",
      "subject_align_len                    0\n",
      "subject_align_cov                    0\n",
      "bit_score                            0\n",
      "m_protein_len                        0\n",
      "t_protein_len                        0\n",
      "protein_match                        0\n",
      "dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([0])"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nan_counts = df.isna().sum()\n",
    "print(nan_counts)\n",
    "nan_counts.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "76ebfd92",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.dropna(axis=1, how='any')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e13ec5f3",
   "metadata": {},
   "source": [
    "Use MRMR to select for the best features. Going to start by grouping into different categories of features generated from iFeature Omega."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f158c7bb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "number sections must be larger than 0.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-19-c89dd90952f0>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;31m# select top 10 features using mRMR\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mmrmr\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mmrmr_classif\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m \u001b[0mselected_features\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmrmr_classif\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdf_subset\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'protein_match'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mK\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m20\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      8\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0mselected_features\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/mrmr/pandas.py\u001b[0m in \u001b[0;36mmrmr_classif\u001b[0;34m(X, y, K, relevance, redundancy, denominator, cat_features, cat_encoding, only_same_domain, return_scores, n_jobs, show_progress)\u001b[0m\n\u001b[1;32m    169\u001b[0m     \u001b[0mredundancy_args\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m'X'\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mX\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    170\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 171\u001b[0;31m     return mrmr_base(K=K, relevance_func=relevance_func, redundancy_func=redundancy_func,\n\u001b[0m\u001b[1;32m    172\u001b[0m                      \u001b[0mrelevance_args\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrelevance_args\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mredundancy_args\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mredundancy_args\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    173\u001b[0m                      \u001b[0mdenominator_func\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdenominator_func\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0monly_same_domain\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0monly_same_domain\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/mrmr/main.py\u001b[0m in \u001b[0;36mmrmr_base\u001b[0;34m(K, relevance_func, redundancy_func, relevance_args, redundancy_args, denominator_func, only_same_domain, return_scores, show_progress)\u001b[0m\n\u001b[1;32m     96\u001b[0m     \"\"\"\n\u001b[1;32m     97\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 98\u001b[0;31m     \u001b[0mrelevance\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrelevance_func\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mrelevance_args\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     99\u001b[0m     \u001b[0mfeatures\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrelevance\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mrelevance\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfillna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_list\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    100\u001b[0m     \u001b[0mrelevance\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrelevance\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mfeatures\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/mrmr/pandas.py\u001b[0m in \u001b[0;36mf_classif\u001b[0;34m(X, y, n_jobs)\u001b[0m\n\u001b[1;32m     43\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     44\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mf_classif\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn_jobs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 45\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mparallel_df\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_f_classif\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mX\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn_jobs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mn_jobs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     46\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     47\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/mrmr/pandas.py\u001b[0m in \u001b[0;36mparallel_df\u001b[0;34m(func, df, series, n_jobs)\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mparallel_df\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mseries\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn_jobs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m     \u001b[0mn_jobs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcpu_count\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mn_jobs\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0mmin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcpu_count\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn_jobs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 16\u001b[0;31m     \u001b[0mcol_chunks\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray_split\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn_jobs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     17\u001b[0m     lst = Parallel(n_jobs=n_jobs)(\n\u001b[1;32m     18\u001b[0m         \u001b[0mdelayed\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcol_chunk\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mseries\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/numpy/core/overrides.py\u001b[0m in \u001b[0;36marray_split\u001b[0;34m(*args, **kwargs)\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/numpy/lib/shape_base.py\u001b[0m in \u001b[0;36marray_split\u001b[0;34m(ary, indices_or_sections, axis)\u001b[0m\n\u001b[1;32m    776\u001b[0m         \u001b[0mNsections\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindices_or_sections\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    777\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mNsections\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 778\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'number sections must be larger than 0.'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    779\u001b[0m         \u001b[0mNeach_section\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mextras\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdivmod\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mNtotal\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mNsections\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    780\u001b[0m         section_sizes = ([0] +\n",
      "\u001b[0;31mValueError\u001b[0m: number sections must be larger than 0."
     ]
    }
   ],
   "source": [
    "#use MRMR to select for the best features from PseKRAAC\n",
    "df_subset = df.loc[:, df.columns != 'protein_match']\n",
    "print(type(df_subset))\n",
    "\n",
    "# select top 10 features using mRMR\n",
    "from mrmr import mrmr_classif\n",
    "selected_features = mrmr_classif(X=df_subset.iloc[:,10:], y=df['protein_match'], K=20)\n",
    "\n",
    "selected_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "5bbe1195",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "      <th>...</th>\n",
       "      <th>AAC_K</th>\n",
       "      <th>AAC_Q</th>\n",
       "      <th>AAC_I</th>\n",
       "      <th>QSOrder_Grantham.Xd.1</th>\n",
       "      <th>AAC_S</th>\n",
       "      <th>AAC_N</th>\n",
       "      <th>AAC_F</th>\n",
       "      <th>AAC_D</th>\n",
       "      <th>GAAC_uncharge</th>\n",
       "      <th>QSOrder_Schneider.Xd.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.287582</td>\n",
       "      <td>0.217822</td>\n",
       "      <td>0.215686</td>\n",
       "      <td>160</td>\n",
       "      <td>0.792079</td>\n",
       "      <td>152</td>\n",
       "      <td>0.737864</td>\n",
       "      <td>131</td>\n",
       "      <td>206</td>\n",
       "      <td>202</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.029991</td>\n",
       "      <td>-0.020571</td>\n",
       "      <td>-0.029991</td>\n",
       "      <td>-0.064588</td>\n",
       "      <td>-0.015524</td>\n",
       "      <td>-0.019898</td>\n",
       "      <td>-0.015140</td>\n",
       "      <td>0.013410</td>\n",
       "      <td>-0.033452</td>\n",
       "      <td>0.022842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.319635</td>\n",
       "      <td>0.295359</td>\n",
       "      <td>0.297872</td>\n",
       "      <td>218</td>\n",
       "      <td>0.919831</td>\n",
       "      <td>226</td>\n",
       "      <td>0.969957</td>\n",
       "      <td>282</td>\n",
       "      <td>233</td>\n",
       "      <td>237</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.007497</td>\n",
       "      <td>0.021821</td>\n",
       "      <td>-0.016008</td>\n",
       "      <td>-0.093397</td>\n",
       "      <td>0.004944</td>\n",
       "      <td>-0.016443</td>\n",
       "      <td>0.008946</td>\n",
       "      <td>0.035711</td>\n",
       "      <td>-0.017982</td>\n",
       "      <td>0.055143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.279621</td>\n",
       "      <td>0.234127</td>\n",
       "      <td>0.218924</td>\n",
       "      <td>211</td>\n",
       "      <td>0.837302</td>\n",
       "      <td>210</td>\n",
       "      <td>0.731707</td>\n",
       "      <td>96</td>\n",
       "      <td>287</td>\n",
       "      <td>252</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.017325</td>\n",
       "      <td>0.002033</td>\n",
       "      <td>-0.035134</td>\n",
       "      <td>0.034032</td>\n",
       "      <td>-0.067847</td>\n",
       "      <td>-0.018777</td>\n",
       "      <td>-0.004355</td>\n",
       "      <td>0.027003</td>\n",
       "      <td>-0.114015</td>\n",
       "      <td>0.005861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.327273</td>\n",
       "      <td>0.200743</td>\n",
       "      <td>0.214712</td>\n",
       "      <td>166</td>\n",
       "      <td>0.617100</td>\n",
       "      <td>163</td>\n",
       "      <td>0.696581</td>\n",
       "      <td>175</td>\n",
       "      <td>234</td>\n",
       "      <td>269</td>\n",
       "      <td>...</td>\n",
       "      <td>0.008722</td>\n",
       "      <td>-0.000381</td>\n",
       "      <td>-0.018238</td>\n",
       "      <td>-0.082094</td>\n",
       "      <td>0.003511</td>\n",
       "      <td>-0.006323</td>\n",
       "      <td>0.010946</td>\n",
       "      <td>-0.016919</td>\n",
       "      <td>-0.020144</td>\n",
       "      <td>0.055484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.338710</td>\n",
       "      <td>0.318182</td>\n",
       "      <td>0.287671</td>\n",
       "      <td>60</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>71</td>\n",
       "      <td>0.887500</td>\n",
       "      <td>61</td>\n",
       "      <td>80</td>\n",
       "      <td>66</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.032955</td>\n",
       "      <td>0.022348</td>\n",
       "      <td>0.001894</td>\n",
       "      <td>-0.024133</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.007197</td>\n",
       "      <td>-0.066667</td>\n",
       "      <td>-0.005303</td>\n",
       "      <td>0.051136</td>\n",
       "      <td>0.049946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6343</th>\n",
       "      <td>0.387574</td>\n",
       "      <td>0.367978</td>\n",
       "      <td>0.360882</td>\n",
       "      <td>336</td>\n",
       "      <td>0.943820</td>\n",
       "      <td>337</td>\n",
       "      <td>0.910811</td>\n",
       "      <td>588</td>\n",
       "      <td>370</td>\n",
       "      <td>356</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002855</td>\n",
       "      <td>0.030626</td>\n",
       "      <td>-0.015958</td>\n",
       "      <td>0.004543</td>\n",
       "      <td>-0.001488</td>\n",
       "      <td>-0.007319</td>\n",
       "      <td>-0.015639</td>\n",
       "      <td>0.028135</td>\n",
       "      <td>0.023504</td>\n",
       "      <td>-0.002964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6344</th>\n",
       "      <td>0.488462</td>\n",
       "      <td>0.488462</td>\n",
       "      <td>0.455197</td>\n",
       "      <td>258</td>\n",
       "      <td>0.992308</td>\n",
       "      <td>263</td>\n",
       "      <td>0.882550</td>\n",
       "      <td>619</td>\n",
       "      <td>298</td>\n",
       "      <td>260</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.001962</td>\n",
       "      <td>0.004750</td>\n",
       "      <td>-0.005885</td>\n",
       "      <td>0.052205</td>\n",
       "      <td>0.010403</td>\n",
       "      <td>-0.000077</td>\n",
       "      <td>0.011461</td>\n",
       "      <td>0.001652</td>\n",
       "      <td>0.007073</td>\n",
       "      <td>-0.015008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6345</th>\n",
       "      <td>0.383838</td>\n",
       "      <td>0.376238</td>\n",
       "      <td>0.374384</td>\n",
       "      <td>296</td>\n",
       "      <td>0.976898</td>\n",
       "      <td>299</td>\n",
       "      <td>0.977124</td>\n",
       "      <td>436</td>\n",
       "      <td>306</td>\n",
       "      <td>303</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009707</td>\n",
       "      <td>-0.003494</td>\n",
       "      <td>0.035462</td>\n",
       "      <td>-0.063804</td>\n",
       "      <td>0.005986</td>\n",
       "      <td>0.022811</td>\n",
       "      <td>-0.000356</td>\n",
       "      <td>-0.000453</td>\n",
       "      <td>0.040866</td>\n",
       "      <td>-0.007821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6346</th>\n",
       "      <td>0.769231</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>78</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>78</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>258</td>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>...</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.023367</td>\n",
       "      <td>0.051282</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.012821</td>\n",
       "      <td>-0.012821</td>\n",
       "      <td>-0.038462</td>\n",
       "      <td>0.012262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6347</th>\n",
       "      <td>0.343220</td>\n",
       "      <td>0.223757</td>\n",
       "      <td>0.264274</td>\n",
       "      <td>234</td>\n",
       "      <td>0.646409</td>\n",
       "      <td>238</td>\n",
       "      <td>0.948207</td>\n",
       "      <td>347</td>\n",
       "      <td>251</td>\n",
       "      <td>362</td>\n",
       "      <td>...</td>\n",
       "      <td>0.050835</td>\n",
       "      <td>0.010676</td>\n",
       "      <td>0.003610</td>\n",
       "      <td>0.024203</td>\n",
       "      <td>0.013438</td>\n",
       "      <td>0.012535</td>\n",
       "      <td>0.008232</td>\n",
       "      <td>-0.001013</td>\n",
       "      <td>0.003654</td>\n",
       "      <td>0.048942</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6348 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      local_gap_compressed_percent_id  scaled_local_query_percent_id  \\\n",
       "0                            0.287582                       0.217822   \n",
       "1                            0.319635                       0.295359   \n",
       "2                            0.279621                       0.234127   \n",
       "3                            0.327273                       0.200743   \n",
       "4                            0.338710                       0.318182   \n",
       "...                               ...                            ...   \n",
       "6343                         0.387574                       0.367978   \n",
       "6344                         0.488462                       0.488462   \n",
       "6345                         0.383838                       0.376238   \n",
       "6346                         0.769231                       0.769231   \n",
       "6347                         0.343220                       0.223757   \n",
       "\n",
       "      scaled_local_symmetric_percent_id  query_align_len  query_align_cov  \\\n",
       "0                              0.215686              160         0.792079   \n",
       "1                              0.297872              218         0.919831   \n",
       "2                              0.218924              211         0.837302   \n",
       "3                              0.214712              166         0.617100   \n",
       "4                              0.287671               60         0.909091   \n",
       "...                                 ...              ...              ...   \n",
       "6343                           0.360882              336         0.943820   \n",
       "6344                           0.455197              258         0.992308   \n",
       "6345                           0.374384              296         0.976898   \n",
       "6346                           0.769231               78         1.000000   \n",
       "6347                           0.264274              234         0.646409   \n",
       "\n",
       "      subject_align_len  subject_align_cov  bit_score  m_protein_len  \\\n",
       "0                   152           0.737864        131            206   \n",
       "1                   226           0.969957        282            233   \n",
       "2                   210           0.731707         96            287   \n",
       "3                   163           0.696581        175            234   \n",
       "4                    71           0.887500         61             80   \n",
       "...                 ...                ...        ...            ...   \n",
       "6343                337           0.910811        588            370   \n",
       "6344                263           0.882550        619            298   \n",
       "6345                299           0.977124        436            306   \n",
       "6346                 78           1.000000        258             78   \n",
       "6347                238           0.948207        347            251   \n",
       "\n",
       "      t_protein_len  ...     AAC_K     AAC_Q     AAC_I  QSOrder_Grantham.Xd.1  \\\n",
       "0               202  ... -0.029991 -0.020571 -0.029991              -0.064588   \n",
       "1               237  ... -0.007497  0.021821 -0.016008              -0.093397   \n",
       "2               252  ... -0.017325  0.002033 -0.035134               0.034032   \n",
       "3               269  ...  0.008722 -0.000381 -0.018238              -0.082094   \n",
       "4                66  ... -0.032955  0.022348  0.001894              -0.024133   \n",
       "...             ...  ...       ...       ...       ...                    ...   \n",
       "6343            356  ...  0.002855  0.030626 -0.015958               0.004543   \n",
       "6344            260  ... -0.001962  0.004750 -0.005885               0.052205   \n",
       "6345            303  ...  0.009707 -0.003494  0.035462              -0.063804   \n",
       "6346             78  ...  0.025641  0.000000  0.000000               0.023367   \n",
       "6347            362  ...  0.050835  0.010676  0.003610               0.024203   \n",
       "\n",
       "         AAC_S     AAC_N     AAC_F     AAC_D  GAAC_uncharge  \\\n",
       "0    -0.015524 -0.019898 -0.015140  0.013410      -0.033452   \n",
       "1     0.004944 -0.016443  0.008946  0.035711      -0.017982   \n",
       "2    -0.067847 -0.018777 -0.004355  0.027003      -0.114015   \n",
       "3     0.003511 -0.006323  0.010946 -0.016919      -0.020144   \n",
       "4     0.037500  0.007197 -0.066667 -0.005303       0.051136   \n",
       "...        ...       ...       ...       ...            ...   \n",
       "6343 -0.001488 -0.007319 -0.015639  0.028135       0.023504   \n",
       "6344  0.010403 -0.000077  0.011461  0.001652       0.007073   \n",
       "6345  0.005986  0.022811 -0.000356 -0.000453       0.040866   \n",
       "6346  0.051282  0.000000 -0.012821 -0.012821      -0.038462   \n",
       "6347  0.013438  0.012535  0.008232 -0.001013       0.003654   \n",
       "\n",
       "      QSOrder_Schneider.Xd.2  \n",
       "0                   0.022842  \n",
       "1                   0.055143  \n",
       "2                   0.005861  \n",
       "3                   0.055484  \n",
       "4                   0.049946  \n",
       "...                      ...  \n",
       "6343               -0.002964  \n",
       "6344               -0.015008  \n",
       "6345               -0.007821  \n",
       "6346                0.012262  \n",
       "6347                0.048942  \n",
       "\n",
       "[6348 rows x 31 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_features_df = df[[feature for feature in selected_features]]\n",
    "\n",
    "#concatenates original feature vector back into the dataframe\n",
    "df = pd.concat([df.iloc[:, :11], best_features_df], axis=1)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "af9f23cf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(14541, 11)\n",
      "(2567, 11)\n",
      "(12359, 11)\n",
      "(2182, 11)\n"
     ]
    }
   ],
   "source": [
    "#choosing 80/20 split instead of 85/15 because of volume of data\n",
    "\n",
    "dev, test = sklearn.model_selection.train_test_split(df, test_size=0.15, random_state=1)\n",
    "\n",
    "train, val = sklearn.model_selection.train_test_split(dev, test_size=0.15, random_state=1)\n",
    "\n",
    "print(dev.shape)\n",
    "print(test.shape)\n",
    "print(train.shape)\n",
    "print(val.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "8470550d",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#ID target and features, separate into separate arrays\n",
    "\n",
    "target = 'protein_match'\n",
    "input_features = [columns for columns in df]\n",
    "input_features.remove(target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1ab62bff",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(14541, 10) (2567, 10) (14541, 1) (2567, 1)\n"
     ]
    }
   ],
   "source": [
    "#split X and y\n",
    "\n",
    "dev_X = dev[input_features].values\n",
    "test_X = test[input_features].values\n",
    "\n",
    "dev_y = dev[target].values.reshape(-1,1)\n",
    "test_y = test[target].values.reshape(-1,1)  \n",
    "\n",
    "print(dev_X.shape, test_X.shape, dev_y.shape, test_y.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "9ac52ac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#same thing for training and validation data\n",
    "\n",
    "train_X = train[input_features].values\n",
    "val_X = val[input_features].values\n",
    "\n",
    "train_y = train[target].values.reshape(-1,1)\n",
    "val_y = val[target].values.reshape(-1,1) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c374f2d7",
   "metadata": {},
   "source": [
    "Scale the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "2b48b2e5",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "scaler = sklearn.preprocessing.StandardScaler()\n",
    "dev_X = scaler.fit_transform(dev_X)\n",
    "test_X = scaler.fit_transform(test_X)\n",
    "train_X = scaler.fit_transform(train_X)\n",
    "val_X = scaler.fit_transform(val_X)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3022371",
   "metadata": {},
   "source": [
    "Train the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "f4411c95",
   "metadata": {},
   "outputs": [],
   "source": [
    "def k_fold_cross_val(dataframe, n_splits=10):  \n",
    "    \"\"\"\n",
    "    Runs k-fold cross validation on dataset.\n",
    "    Default = 10-fold.\n",
    "\n",
    "    Params\n",
    "    ----------\n",
    "    -dataframe: Pandas dataframe\n",
    "    -n_splits: Number of cross validations (int)\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    -vector of predictions\n",
    "    \"\"\"\n",
    "    \n",
    "    dev, test = sklearn.model_selection.train_test_split(dataframe, test_size=0.15, random_state=1)\n",
    "\n",
    "    train, val = sklearn.model_selection.train_test_split(dev, test_size=0.15, random_state=1)\n",
    "    \n",
    "    target = 'protein_match'\n",
    "    input_features = [columns for columns in df]\n",
    "    input_features.remove(target)\n",
    "    \n",
    "    dev_X = dev[input_features].values\n",
    "    test_X = test[input_features].values\n",
    "\n",
    "    dev_y = dev[target].values.reshape(-1,1)\n",
    "    test_y = test[target].values.reshape(-1,1) \n",
    "\n",
    "    from sklearn.model_selection import StratifiedKFold\n",
    "\n",
    "    cv = StratifiedKFold(n_splits, shuffle=True)\n",
    "\n",
    "    for (train_index, test_index) in cv.split(dev_X, dev_y):\n",
    "\n",
    "        train_X = dev_X[train_index]\n",
    "        val_X = dev_X[test_index]\n",
    "\n",
    "        train_y = dev_y[train_index]\n",
    "        val_y = dev_y[test_index]\n",
    "\n",
    "        model.fit(train_X, train_y)\n",
    "\n",
    "        preds = model.predict(val_X)\n",
    "\n",
    "        return preds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "6c8325f2",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'preds' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-25-c66c8509e20f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mpreds\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'preds' is not defined"
     ]
    }
   ],
   "source": [
    "preds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "be4e09b0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-2 {color: black;background-color: white;}#sk-container-id-2 pre{padding: 0;}#sk-container-id-2 div.sk-toggleable {background-color: white;}#sk-container-id-2 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-2 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-2 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-2 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-2 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-2 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-2 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-2 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-container-id-2 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-2 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-2 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-2 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-2 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-2 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-2 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-2 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-2 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-2 div.sk-item {position: relative;z-index: 1;}#sk-container-id-2 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-2 div.sk-item::before, #sk-container-id-2 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-2 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-2 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-2 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-2 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-2 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-2 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-2 div.sk-label-container {text-align: center;}#sk-container-id-2 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-2 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-2\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>RandomForestClassifier()</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-2\" type=\"checkbox\" checked><label for=\"sk-estimator-id-2\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">RandomForestClassifier</label><div class=\"sk-toggleable__content\"><pre>RandomForestClassifier()</pre></div></div></div></div></div>"
      ],
      "text/plain": [
       "RandomForestClassifier()"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Random Forest\n",
    "\n",
    "#hyperparameters determiend with optuna\n",
    "model = sklearn.ensemble.RandomForestClassifier()\n",
    "\n",
    "model.fit(train_X, train_y.ravel())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69ebf269",
   "metadata": {},
   "source": [
    "Test the model, report relevant statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "c73327e9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model score is: 0.7653528872593951\n",
      "[0 0 1 ... 0 1 0]\n"
     ]
    }
   ],
   "source": [
    "score = model.score(val_X, val_y)\n",
    "print('Model score is: {}'.format(score))\n",
    "\n",
    "preds = model.predict(test_X)\n",
    "print(preds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "23254534",
   "metadata": {},
   "outputs": [],
   "source": [
    "proba_y = model.predict_proba(val_X)[:,1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "efcc8c3b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0.36, 0.92, 0.4 , ..., 1.  , 0.28, 0.37])"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proba_y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "4e9dce62",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<sklearn.metrics._plot.confusion_matrix.ConfusionMatrixDisplay at 0x7fab12e5e790>"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#confusion matrix\n",
    "\n",
    "confusion_matrix = sklearn.metrics.confusion_matrix(preds, test_y)\n",
    "sklearn.metrics.ConfusionMatrixDisplay(confusion_matrix).plot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "f838ddbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.naive_bayes import GaussianNB\n",
    "from sklearn.ensemble import BaggingClassifier, \\\n",
    "    RandomForestClassifier, \\\n",
    "    AdaBoostClassifier, \\\n",
    "    GradientBoostingClassifier, \\\n",
    "    ExtraTreesClassifier\n",
    "\n",
    "from sklearn.discriminant_analysis import LinearDiscriminantAnalysis\n",
    "\n",
    "names = ['LR', 'KNN', 'DT', 'NB', 'RF', 'Bagging', 'AB', 'GB', 'SVM']\n",
    "\n",
    "#list of classifiers (hyperparameters optimized)\n",
    "classifiers = [\n",
    "    #Regression\n",
    "    LogisticRegression(),\n",
    "    #KNN (neighbors optimized iteratively)\n",
    "    KNeighborsClassifier(n_neighbors=20),\n",
    "    #Decision Tree\n",
    "    DecisionTreeClassifier(max_features=None),\n",
    "    #Gaussian\n",
    "    GaussianNB(),\n",
    "    #RF Classifier (with optuna)\n",
    "    RandomForestClassifier(\n",
    "        n_estimators=200,\n",
    "        max_depth=None,\n",
    "        max_samples=0.3,\n",
    "        max_features=0.5,\n",
    "        min_weight_fraction_leaf=0,\n",
    "        min_samples_split=17),\n",
    "    #RF Classifier with bagging (with optuna)\n",
    "    BaggingClassifier(sklearn.ensemble.RandomForestClassifier\n",
    "    (n_estimators=200, max_depth=None, \n",
    "     min_weight_fraction_leaf=0.000215),max_samples=0.5, \n",
    "    max_features=0.5),\n",
    "    #AdaBoost (with optuna)\n",
    "    AdaBoostClassifier(n_estimators=53, learning_rate=0.156),\n",
    "    #Gradient Boosting (with optuna)\n",
    "    GradientBoostingClassifier(n_estimators=100, learning_rate=1.0, \n",
    "                                max_depth=1),   \n",
    "    #C-support vector classification (9)\n",
    "#     SVC(),\n",
    "]\n",
    "\n",
    "F = open('evaluationResults.txt', 'w')\n",
    "\n",
    "F.write('Evaluation Scale:'+'\\n')\n",
    "F.write('0.0% <=Accuracy<= 100.0%'+'\\n')\n",
    "F.write('0.0 <=AUC<= 1.0'+'\\n') #area under curve\n",
    "F.write('0.0 <=auPR<= 1.0'+'\\n')  # average_Precision\n",
    "F.write('0.0 <=F1_Score<= 1.0'+'\\n')\n",
    "F.write('-1.0 <=MCC<= 1.0'+'\\n')\n",
    "F.write('_______________________________________'+'\\n')\n",
    "\n",
    "def runClassifiers(filename:str, dataframe, columns=[], target=[], model=RandomForestClassifier()):\n",
    "\n",
    "    \"\"\"\n",
    "    Takes dataframe and splits it into a training and testing set. \n",
    "    Trains a RF Classifier with data.\n",
    "\n",
    "    Params\n",
    "    ----------\n",
    "    dataframe: Pandas dataframe\n",
    "    columns: list of strings, representing input features\n",
    "    target: list of strings, representing target feature(s)\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    -Accuracy score\n",
    "    -area under ROC curve\n",
    "    -train data (target)\n",
    "    -validation data (features)\n",
    "    -validation data (target)\n",
    "    \"\"\"\n",
    "    \n",
    "    dev, test = sklearn.model_selection.train_test_split(dataframe, test_size=0.15, random_state=1)\n",
    "\n",
    "    train, val = sklearn.model_selection.train_test_split(dev, test_size=0.15, random_state=1)\n",
    "    \n",
    "    dev_X = dev[columns].values\n",
    "    test_X = test[columns].values\n",
    "\n",
    "    dev_y = dev[target].values.reshape(-1,1)\n",
    "    test_y = test[target].values.reshape(-1,1) \n",
    "\n",
    "    # test input arguments\n",
    "    assert \"pandas.core.frame.DataFrame\" in str(type(train))\n",
    "    assert \"pandas.core.frame.DataFrame\" in str(type(val))\n",
    "    assert \"str\" in str(type(columns[0]))\n",
    "    assert \"str\" in str(type(target[0]))\n",
    "\n",
    "    # split into input and output feature(s)\n",
    "    train_X = train[columns].values\n",
    "    val_X = val[columns].values\n",
    "\n",
    "    train_y = train[target].values.reshape(-1, 1)\n",
    "    val_y = val[target].values.reshape(-1, 1)\n",
    "\n",
    "    # scale data\n",
    "    scaler = sklearn.preprocessing.StandardScaler()\n",
    "    train_X = scaler.fit_transform(train_X)\n",
    "    val_X = scaler.fit_transform(val_X)\n",
    "\n",
    "    Results = {}  # compare algorithms\n",
    "\n",
    "    from sklearn.metrics import accuracy_score, \\\n",
    "        confusion_matrix, \\\n",
    "        roc_auc_score,\\\n",
    "        average_precision_score,\\\n",
    "        auc,\\\n",
    "        roc_curve, f1_score, recall_score, matthews_corrcoef, auc\n",
    "\n",
    "    for classifier, name in zip(classifiers, names):\n",
    "        accuracy = []\n",
    "        avg_precision = []\n",
    "        F1_Score = []\n",
    "        AUC = []\n",
    "        MCC = []\n",
    "        Recall = []\n",
    "        \n",
    "        mean_TPR = 0.0\n",
    "        mean_FPR = np.linspace(0, 1, 100)\n",
    "\n",
    "        print('{} is done.'.format(classifier.__class__.__name__))\n",
    "\n",
    "        model = classifier\n",
    "\n",
    "        # model\n",
    "        model.fit(train_X, train_y)\n",
    "\n",
    "        preds = model.predict(val_X)\n",
    "\n",
    "        # Calculate ROC Curve and Area the Curve\n",
    "        proba_y = model.predict_proba(val_X)[:,1]\n",
    "        FPR, TPR, _ = roc_curve(val_y, proba_y, pos_label=1)\n",
    "        roc_auc = auc(FPR, TPR)\n",
    "        \n",
    "        #calculate scoring metrics\n",
    "        #include option to return these scores\n",
    "        accuracy.append(accuracy_score(y_pred=preds, y_true=val_y))\n",
    "        avg_precision.append(average_precision_score(y_true=val_y, y_score=proba_y, pos_label=1))\n",
    "        F1_Score.append(f1_score(y_true=val_y, y_pred=preds, pos_label=1))\n",
    "        MCC.append(matthews_corrcoef(y_true=val_y, y_pred=preds))\n",
    "        Recall.append(recall_score(y_true=val_y, y_pred=preds, pos_label=1))\n",
    "        AUC.append(roc_auc)\n",
    "\n",
    "        confusion_matrix = sklearn.metrics.confusion_matrix(y_pred=preds, y_true=val_y)\n",
    "        sklearn.metrics.ConfusionMatrixDisplay(confusion_matrix).plot()\n",
    "\n",
    "        accuracy = [_*100.0 for _ in accuracy]\n",
    "        Results[name + ' Accuracy, F1 Score'] = [accuracy, F1_Score]\n",
    "        \n",
    "        F.write('Classifier: {}\\n'.format(name))\n",
    "        F.write('Accuracy: {0:.4f}%\\n'.format(np.mean(accuracy)))\n",
    "        F.write('AUC: {0:.4f}\\n'.format( np.mean(AUC)))\n",
    "        F.write('auPR: {0:.4f}\\n'.format(np.mean(avg_precision))) # average_Precision\n",
    "        F.write('F1_Score: {0:.4f}\\n'.format(np.mean(F1_Score)))\n",
    "        F.write('MCC: {0:.4f}\\n'.format(np.mean(MCC)))\n",
    "\n",
    "#         TN, FP, FN, TP = CM.ravel()\n",
    "        F.write('Recall: {0:.4f}\\n'.format( np.mean(Recall)) )\n",
    "        F.write('_______________________________________'+'\\n')\n",
    "    \n",
    "    F.close()\n",
    "    \n",
    "    return Results\n",
    "\n",
    "\n",
    "# if __name__ == '__main__':\n",
    "#     # print('Please, enter number of cross validation:')\n",
    "#     import argparse\n",
    "#     p = argparse.ArgumentParser(description='Run Machine Learning Classifiers.')\n",
    "\n",
    "#     p.add_argument('-cv', '--nFCV', type=int, help='Number of crossValidation', default=10)\n",
    "#     p.add_argument('-data', '--dataset', type=str, help='~/dataset.csv', default='optimumDataset.csv')\n",
    "#     p.add_argument('-roc', '--auROC', type=int, help='Print ROC Curve', default=1, choices=[0, 1])\n",
    "#     p.add_argument('-box', '--boxPlot', type=int, help='Print Accuracy Box Plaot', default=1, choices=[0, 1])\n",
    "\n",
    "#     args = p.parse_args()\n",
    "\n",
    "#     runClassifiers(args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "849a7b36",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LogisticRegression is done.\n",
      "KNeighborsClassifier is done.\n",
      "DecisionTreeClassifier is done.\n",
      "GaussianNB is done.\n",
      "RandomForestClassifier is done.\n",
      "BaggingClassifier is done.\n"
     ]
    }
   ],
   "source": [
    "result = runClassifiers('sample.csv', df, columns=input_features, target=target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a8b753a",
   "metadata": {},
   "outputs": [],
   "source": [
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c674541",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
