{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Dev work- attempting to work with interpro/pfam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# system dependecies\n",
    "import subprocess\n",
    "from pathlib import Path\n",
    "\n",
    "\n",
    "# library dependencies\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from collections import defaultdict\n",
    "\n",
    "## biopython\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SearchIO\n",
    "\n",
    "\n",
    "# local dependencies/utils\n",
    "\n",
    "## Paths\n",
    "PFAM_PATH = Path(\"/Users/humoodalanzi/pfam/Pfam-A.hmm\")\n",
    "ID_DB_PATH = Path(\"/Users/humoodalanzi/pfam/proteins_id.zip\")\n",
    "#probably need path of unit tests\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Will read API and work on this. HMMER website if interpro is slow.\n",
    "# I need to install GNU make. All of this is done in my test_env which has HMMER and biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_database = pd.read_csv(ID_DB_PATH, index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_seq</th>\n",
       "      <th>protein_int_index</th>\n",
       "      <th>identity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MLATITFVFLALLFIVLKLVLVVQMKEVCVIERLGKFRAVLAPGLH...</td>\n",
       "      <td>60589906</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MIAVDGNTLDLAVLAVWAFIFLFLAYKFFQAICLVPTKSAYIVERL...</td>\n",
       "      <td>60589907</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MYKCRAPKNKAELAKYYQLRWQALRKPWQQPLGSEKDELESQSYHR...</td>\n",
       "      <td>60589911</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MIALLQRVTEASVSVNNNIVGEIEQGLLVLLAIEPEDTEVKAKRLA...</td>\n",
       "      <td>60589913</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MIHIARFYIFYMLFTQAFSAFAGSAKVYVWRNEQGVLVFSDSPKPG...</td>\n",
       "      <td>60589920</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712212</th>\n",
       "      <td>MPKINVLDMNGSPVGEMELSESVFGIEPNEAVLHDAVVMQQASQRR...</td>\n",
       "      <td>43784908</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712213</th>\n",
       "      <td>MTQVFTEDGTLIPVTVISAGPCTILQKKEKASDGYESVQLGFADKK...</td>\n",
       "      <td>43784909</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712214</th>\n",
       "      <td>MAKQKIRIRLKAYDHRALDQSAEKIVDTANRTGANVSGPIPLPTER...</td>\n",
       "      <td>43784910</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712216</th>\n",
       "      <td>MELKKVLDLPTYYGKNLDVLWDCLTGWVSLPITLEWRNFNQCKKNI...</td>\n",
       "      <td>43784912</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712217</th>\n",
       "      <td>MTLFEECLDALGEHCEVLSEEKSNRVVEAMCRLFPVAINGRIDWDK...</td>\n",
       "      <td>43784913</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4262435 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               protein_seq  protein_int_index  \\\n",
       "0        MLATITFVFLALLFIVLKLVLVVQMKEVCVIERLGKFRAVLAPGLH...           60589906   \n",
       "1        MIAVDGNTLDLAVLAVWAFIFLFLAYKFFQAICLVPTKSAYIVERL...           60589907   \n",
       "2        MYKCRAPKNKAELAKYYQLRWQALRKPWQQPLGSEKDELESQSYHR...           60589911   \n",
       "3        MIALLQRVTEASVSVNNNIVGEIEQGLLVLLAIEPEDTEVKAKRLA...           60589913   \n",
       "4        MIHIARFYIFYMLFTQAFSAFAGSAKVYVWRNEQGVLVFSDSPKPG...           60589920   \n",
       "...                                                    ...                ...   \n",
       "6712212  MPKINVLDMNGSPVGEMELSESVFGIEPNEAVLHDAVVMQQASQRR...           43784908   \n",
       "6712213  MTQVFTEDGTLIPVTVISAGPCTILQKKEKASDGYESVQLGFADKK...           43784909   \n",
       "6712214  MAKQKIRIRLKAYDHRALDQSAEKIVDTANRTGANVSGPIPLPTER...           43784910   \n",
       "6712216  MELKKVLDLPTYYGKNLDVLWDCLTGWVSLPITLEWRNFNQCKKNI...           43784912   \n",
       "6712217  MTLFEECLDALGEHCEVLSEEKSNRVVEAMCRLFPVAINGRIDWDK...           43784913   \n",
       "\n",
       "        identity  \n",
       "0              m  \n",
       "1              m  \n",
       "2              m  \n",
       "3              m  \n",
       "4              m  \n",
       "...          ...  \n",
       "6712212        t  \n",
       "6712213        t  \n",
       "6712214        t  \n",
       "6712216        t  \n",
       "6712217        t  \n",
       "\n",
       "[4262435 rows x 3 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0          m\n",
       "1          m\n",
       "2          m\n",
       "3          m\n",
       "4          m\n",
       "          ..\n",
       "6712212    t\n",
       "6712213    t\n",
       "6712214    t\n",
       "6712216    t\n",
       "6712217    t\n",
       "Name: identity, Length: 4262435, dtype: object"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_database[\"identity\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Quick calculations, aside {outdated}\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.36497369423937065"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "2449783/6712218 # the proteins with NaN are 30%!!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4262435"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "6712218 - 2449783"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### continued"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_db = protein_database.dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4262435, 3)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_db.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_seq</th>\n",
       "      <th>protein_int_index</th>\n",
       "      <th>identity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MLATITFVFLALLFIVLKLVLVVQMKEVCVIERLGKFRAVLAPGLH...</td>\n",
       "      <td>60589906</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MIAVDGNTLDLAVLAVWAFIFLFLAYKFFQAICLVPTKSAYIVERL...</td>\n",
       "      <td>60589907</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MYKCRAPKNKAELAKYYQLRWQALRKPWQQPLGSEKDELESQSYHR...</td>\n",
       "      <td>60589911</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MIALLQRVTEASVSVNNNIVGEIEQGLLVLLAIEPEDTEVKAKRLA...</td>\n",
       "      <td>60589913</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MIHIARFYIFYMLFTQAFSAFAGSAKVYVWRNEQGVLVFSDSPKPG...</td>\n",
       "      <td>60589920</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712212</th>\n",
       "      <td>MPKINVLDMNGSPVGEMELSESVFGIEPNEAVLHDAVVMQQASQRR...</td>\n",
       "      <td>43784908</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712213</th>\n",
       "      <td>MTQVFTEDGTLIPVTVISAGPCTILQKKEKASDGYESVQLGFADKK...</td>\n",
       "      <td>43784909</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712214</th>\n",
       "      <td>MAKQKIRIRLKAYDHRALDQSAEKIVDTANRTGANVSGPIPLPTER...</td>\n",
       "      <td>43784910</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712216</th>\n",
       "      <td>MELKKVLDLPTYYGKNLDVLWDCLTGWVSLPITLEWRNFNQCKKNI...</td>\n",
       "      <td>43784912</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712217</th>\n",
       "      <td>MTLFEECLDALGEHCEVLSEEKSNRVVEAMCRLFPVAINGRIDWDK...</td>\n",
       "      <td>43784913</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4262435 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               protein_seq  protein_int_index  \\\n",
       "0        MLATITFVFLALLFIVLKLVLVVQMKEVCVIERLGKFRAVLAPGLH...           60589906   \n",
       "1        MIAVDGNTLDLAVLAVWAFIFLFLAYKFFQAICLVPTKSAYIVERL...           60589907   \n",
       "2        MYKCRAPKNKAELAKYYQLRWQALRKPWQQPLGSEKDELESQSYHR...           60589911   \n",
       "3        MIALLQRVTEASVSVNNNIVGEIEQGLLVLLAIEPEDTEVKAKRLA...           60589913   \n",
       "4        MIHIARFYIFYMLFTQAFSAFAGSAKVYVWRNEQGVLVFSDSPKPG...           60589920   \n",
       "...                                                    ...                ...   \n",
       "6712212  MPKINVLDMNGSPVGEMELSESVFGIEPNEAVLHDAVVMQQASQRR...           43784908   \n",
       "6712213  MTQVFTEDGTLIPVTVISAGPCTILQKKEKASDGYESVQLGFADKK...           43784909   \n",
       "6712214  MAKQKIRIRLKAYDHRALDQSAEKIVDTANRTGANVSGPIPLPTER...           43784910   \n",
       "6712216  MELKKVLDLPTYYGKNLDVLWDCLTGWVSLPITLEWRNFNQCKKNI...           43784912   \n",
       "6712217  MTLFEECLDALGEHCEVLSEEKSNRVVEAMCRLFPVAINGRIDWDK...           43784913   \n",
       "\n",
       "        identity  \n",
       "0              m  \n",
       "1              m  \n",
       "2              m  \n",
       "3              m  \n",
       "4              m  \n",
       "...          ...  \n",
       "6712212        t  \n",
       "6712213        t  \n",
       "6712214        t  \n",
       "6712216        t  \n",
       "6712217        t  \n",
       "\n",
       "[4262435 rows x 3 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "meso_seq_db = protein_db.loc[protein_db[\"identity\"] == \"m\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "thermo_seq_db = protein_db.loc[protein_db[\"identity\"] == \"t\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_seq</th>\n",
       "      <th>protein_int_index</th>\n",
       "      <th>identity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MLATITFVFLALLFIVLKLVLVVQMKEVCVIERLGKFRAVLAPGLH...</td>\n",
       "      <td>60589906</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MIAVDGNTLDLAVLAVWAFIFLFLAYKFFQAICLVPTKSAYIVERL...</td>\n",
       "      <td>60589907</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MYKCRAPKNKAELAKYYQLRWQALRKPWQQPLGSEKDELESQSYHR...</td>\n",
       "      <td>60589911</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MIALLQRVTEASVSVNNNIVGEIEQGLLVLLAIEPEDTEVKAKRLA...</td>\n",
       "      <td>60589913</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MIHIARFYIFYMLFTQAFSAFAGSAKVYVWRNEQGVLVFSDSPKPG...</td>\n",
       "      <td>60589920</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712105</th>\n",
       "      <td>MTDKTIEMIIKEYALSDVPSQYVSSDEEPMTQRECAAAFGDQAVER...</td>\n",
       "      <td>42584603</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712107</th>\n",
       "      <td>EPMAFTASIQIKGWQVRKRPIGMRPRIEAVPSVATAPNERWSTDLC...</td>\n",
       "      <td>42584609</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712108</th>\n",
       "      <td>MTTPINMLAIDLAKGSFQVCAIGQDGAVLYNRVLSRTRLASLLAEQ...</td>\n",
       "      <td>42584610</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712109</th>\n",
       "      <td>MAFMEPGLGARTLELCRYGTSRTLFRGPERSLDAPYWAVLGGTACH...</td>\n",
       "      <td>42584611</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712110</th>\n",
       "      <td>MAETPKLKPKDMPELGVFDWSDAFRLEDQLTEEERMLRDAARAYAQ...</td>\n",
       "      <td>42584612</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4108740 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               protein_seq  protein_int_index  \\\n",
       "0        MLATITFVFLALLFIVLKLVLVVQMKEVCVIERLGKFRAVLAPGLH...           60589906   \n",
       "1        MIAVDGNTLDLAVLAVWAFIFLFLAYKFFQAICLVPTKSAYIVERL...           60589907   \n",
       "2        MYKCRAPKNKAELAKYYQLRWQALRKPWQQPLGSEKDELESQSYHR...           60589911   \n",
       "3        MIALLQRVTEASVSVNNNIVGEIEQGLLVLLAIEPEDTEVKAKRLA...           60589913   \n",
       "4        MIHIARFYIFYMLFTQAFSAFAGSAKVYVWRNEQGVLVFSDSPKPG...           60589920   \n",
       "...                                                    ...                ...   \n",
       "6712105  MTDKTIEMIIKEYALSDVPSQYVSSDEEPMTQRECAAAFGDQAVER...           42584603   \n",
       "6712107  EPMAFTASIQIKGWQVRKRPIGMRPRIEAVPSVATAPNERWSTDLC...           42584609   \n",
       "6712108  MTTPINMLAIDLAKGSFQVCAIGQDGAVLYNRVLSRTRLASLLAEQ...           42584610   \n",
       "6712109  MAFMEPGLGARTLELCRYGTSRTLFRGPERSLDAPYWAVLGGTACH...           42584611   \n",
       "6712110  MAETPKLKPKDMPELGVFDWSDAFRLEDQLTEEERMLRDAARAYAQ...           42584612   \n",
       "\n",
       "        identity  \n",
       "0              m  \n",
       "1              m  \n",
       "2              m  \n",
       "3              m  \n",
       "4              m  \n",
       "...          ...  \n",
       "6712105        m  \n",
       "6712107        m  \n",
       "6712108        m  \n",
       "6712109        m  \n",
       "6712110        m  \n",
       "\n",
       "[4108740 rows x 3 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meso_seq_db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_seq</th>\n",
       "      <th>protein_int_index</th>\n",
       "      <th>identity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>120140</th>\n",
       "      <td>ETSYSPEGLRKVAKEIRALVQGDAMRPFNNLKRVFPECGDFPLANW...</td>\n",
       "      <td>16287043</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120141</th>\n",
       "      <td>MTDWKAAEPGGVQDPSKALDEALASFQQEAKKLGELTKLWQEGETT...</td>\n",
       "      <td>16287045</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120142</th>\n",
       "      <td>MTTSPADETRKTSGKATKKTSSRAAARRTRQSSETQADAPTKKAST...</td>\n",
       "      <td>16287050</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120143</th>\n",
       "      <td>MIRVMFADDEEIVRSSLRTMLSGAEDMEVVGEAADGRSAVEVARRH...</td>\n",
       "      <td>16287051</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120144</th>\n",
       "      <td>MMLLALELIGVVAFAASGALAAVHARLDLFGVAVLATTTGLGGGIM...</td>\n",
       "      <td>16287052</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712212</th>\n",
       "      <td>MPKINVLDMNGSPVGEMELSESVFGIEPNEAVLHDAVVMQQASQRR...</td>\n",
       "      <td>43784908</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712213</th>\n",
       "      <td>MTQVFTEDGTLIPVTVISAGPCTILQKKEKASDGYESVQLGFADKK...</td>\n",
       "      <td>43784909</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712214</th>\n",
       "      <td>MAKQKIRIRLKAYDHRALDQSAEKIVDTANRTGANVSGPIPLPTER...</td>\n",
       "      <td>43784910</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712216</th>\n",
       "      <td>MELKKVLDLPTYYGKNLDVLWDCLTGWVSLPITLEWRNFNQCKKNI...</td>\n",
       "      <td>43784912</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712217</th>\n",
       "      <td>MTLFEECLDALGEHCEVLSEEKSNRVVEAMCRLFPVAINGRIDWDK...</td>\n",
       "      <td>43784913</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>153695 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               protein_seq  protein_int_index  \\\n",
       "120140   ETSYSPEGLRKVAKEIRALVQGDAMRPFNNLKRVFPECGDFPLANW...           16287043   \n",
       "120141   MTDWKAAEPGGVQDPSKALDEALASFQQEAKKLGELTKLWQEGETT...           16287045   \n",
       "120142   MTTSPADETRKTSGKATKKTSSRAAARRTRQSSETQADAPTKKAST...           16287050   \n",
       "120143   MIRVMFADDEEIVRSSLRTMLSGAEDMEVVGEAADGRSAVEVARRH...           16287051   \n",
       "120144   MMLLALELIGVVAFAASGALAAVHARLDLFGVAVLATTTGLGGGIM...           16287052   \n",
       "...                                                    ...                ...   \n",
       "6712212  MPKINVLDMNGSPVGEMELSESVFGIEPNEAVLHDAVVMQQASQRR...           43784908   \n",
       "6712213  MTQVFTEDGTLIPVTVISAGPCTILQKKEKASDGYESVQLGFADKK...           43784909   \n",
       "6712214  MAKQKIRIRLKAYDHRALDQSAEKIVDTANRTGANVSGPIPLPTER...           43784910   \n",
       "6712216  MELKKVLDLPTYYGKNLDVLWDCLTGWVSLPITLEWRNFNQCKKNI...           43784912   \n",
       "6712217  MTLFEECLDALGEHCEVLSEEKSNRVVEAMCRLFPVAINGRIDWDK...           43784913   \n",
       "\n",
       "        identity  \n",
       "120140         t  \n",
       "120141         t  \n",
       "120142         t  \n",
       "120143         t  \n",
       "120144         t  \n",
       "...          ...  \n",
       "6712212        t  \n",
       "6712213        t  \n",
       "6712214        t  \n",
       "6712216        t  \n",
       "6712217        t  \n",
       "\n",
       "[153695 rows x 3 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "thermo_seq_db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "26.733075246429618"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "4108740/153695 # ~27-fold more meso than thermo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = pd.read_csv(\"../data/Sample.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Old code archive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Define the amino acid sequence\n",
    "# meso_seq_list = [df['meso_seq'].iloc[i] for i in range(1001)] # loop-over 1000 meso sequences\n",
    "\n",
    "# # Create a list of SeqRecord objects\n",
    "# records = []\n",
    "# for i, seq in enumerate(meso_seq_list):\n",
    "#     record = SeqRecord(Seq(seq), id=f\"meso_seq{i+1}\")\n",
    "#     records.append(record)\n",
    "\n",
    "# # Write the list of SeqRecord objects to a FASTA file\n",
    "# SeqIO.write(records, \"sequences.fasta\", \"fasta\")\n",
    "\n",
    "# # Write sequence to file in FASTA format\n",
    "# with open(\"input.fasta\", \"w\") as input_file:\n",
    "#     SeqIO.write(records, input_file, \"fasta\")\n",
    "\n",
    "# # Run HMMER search against the Pfam database using the hmmscan command\n",
    "# subprocess.run([\"hmmscan\", \"--cpu\", \"4\", \"--domtblout\", \"output.domtblout\", \"/Users/humoodalanzi/pfam/Pfam-A.hmm\", \"input.fasta\"])\n",
    "\n",
    "\n",
    "# # Parse the HMMER results file\n",
    "# best_hit = None\n",
    "# best_evalue = float(\"inf\")\n",
    "# best_bitscore = float(\"-inf\")\n",
    "# with open(\"output.domtblout\", \"r\") as results_file:\n",
    "#     for line in results_file:\n",
    "#         if not line.startswith(\"#\"):\n",
    "#             fields = line.strip().split()\n",
    "#             family_id = fields[0]\n",
    "#             evalue = float(fields[6])\n",
    "#             bitscore = float(fields[7])\n",
    "#             devalue = float(fields[11])\n",
    "#             start = int(fields[19])\n",
    "#             end = int(fields[20])\n",
    "#             if best_hit is None or evalue < best_hit[\"evalue\"]:\n",
    "#                 best_hit = {\"family_id\": family_id, \"evalue\": evalue,\"devalue\":devalue, \"bitscore\": bitscore, \"start\": start, \"end\": end}\n",
    "\n",
    "\n",
    "# # Print the best hit and its attributes\n",
    "# if best_hit is not None:\n",
    "#     print(\"Best Pfam family match: \" + best_hit[\"family_id\"])\n",
    "#     print(\"Total E-value: \" + str(best_hit[\"evalue\"]))\n",
    "#     print(\"Domain E-vlaue: \" + str(best_hit[\"devalue\"]))\n",
    "#     print(\"Bit score: \" + str(best_hit[\"bitscore\"]))\n",
    "#     print(\"Start position: \" + str(best_hit[\"start\"]))\n",
    "#     print(\"End position: \" + str(best_hit[\"end\"]))\n",
    "# else:\n",
    "#     print(\"No Pfam family match found.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below is old code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Get the best hit (hsps: high-scoring segment pair)\n",
    "# best_hit = None\n",
    "# for qresult in qresults:\n",
    "#     for hit in qresult.hits:\n",
    "#         for hsp in hit.hsps:\n",
    "#             family_id = hit.id\n",
    "#             evalue = hsp.evalue\n",
    "#             seqlen = hit.seq_len\n",
    "#             bitscore = hsp.bitscore\n",
    "#             devalue = hsp.bitscore / hit.seq_len\n",
    "#             bias = hsp.bias\n",
    "#             start = hsp.query_start\n",
    "#             end = hsp.query_end\n",
    "#             if best_hit is None or evalue < best_hit[\"evalue\"]:\n",
    "#                 best_hit = {\"family_id\": family_id, \"evalue\": evalue,\"seqlen\": seqlen ,\"bias\": bias, \"devalue\": devalue, \"bitscore\": bitscore, \"start\": start, \"end\": end}\n",
    "\n",
    "\n",
    "# # Print the best hit and its attributes\n",
    "# if best_hit is not None:\n",
    "#     print(\"Best Pfam family match: \" + best_hit[\"family_id\"])\n",
    "#     print(\"Total E-value: \" + str(best_hit[\"evalue\"]))\n",
    "#     print(\"length of hit sequence or HMM: \" + str(best_hit[\"seqlen\"]))\n",
    "#     print(\"High-scoring pair bias: \" + str(best_hit[\"bias\"]))\n",
    "#     print(\"Domain E-vlaue: \" + str(best_hit[\"devalue\"]))\n",
    "#     print(\"Bit score: \" + str(best_hit[\"bitscore\"]))\n",
    "#     print(\"Start position: \" + str(best_hit[\"start\"]))\n",
    "#     print(\"End position: \" + str(best_hit[\"end\"]))\n",
    "# else:\n",
    "#     print(\"No Pfam family match found.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# meso_seq_list = df['meso_seq'].iloc[:1000]\n",
    "# meso_seq_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from collections import Counter, defaultdict\n",
    "# import subprocess\n",
    "# from typing import Dict, List, Optional\n",
    "# from Bio import SeqIO\n",
    "# from Bio.SeqRecord import SeqRecord\n",
    "# from Bio.Seq import Seq\n",
    "# from Bio.Alphabet import generic_protein\n",
    "# from Bio import SearchIO\n",
    "\n",
    "# def read_sequences(filename: str, n: Optional[int] = None) -> List[SeqRecord]:\n",
    "#     \"\"\"Reads n protein sequences from a file and returns them as a list of SeqRecord objects.\"\"\"\n",
    "#     records = []\n",
    "#     with open(filename, \"r\") as f:\n",
    "#         for i, record in enumerate(SeqIO.parse(f, \"fasta\")):\n",
    "#             if n is not None and i >= n:\n",
    "#                 break\n",
    "#             records.append(record)\n",
    "#     return records\n",
    "\n",
    "# def write_sequences(records: List[SeqRecord], filename: str) -> None:\n",
    "#     \"\"\"Writes a list of SeqRecord objects to a FASTA file.\"\"\"\n",
    "#     with open(filename, \"w\", encoding=\"utf-8\") as f:\n",
    "#         SeqIO.write(records, f, \"fasta\")\n",
    "\n",
    "# def run_hmmer(pfam_path: str, input_filename: str, output_filename: str, num_cpu: int = 4) -> None:\n",
    "#     \"\"\"Runs HMMER search against the Pfam database using the hmmscan command.\"\"\"\n",
    "#     with open(output_filename, \"w\", encoding=\"utf-8\") as f:\n",
    "#         subprocess.run([\"hmmscan\", \"--cpu\", str(num_cpu), \"--domtblout\", output_filename, pfam_path, input_filename], stdout=f)\n",
    "\n",
    "# def find_best_hit(query_id: str, hits: List) -> Optional[Dict]:\n",
    "#     \"\"\"\n",
    "#     Given a list of HMMER hits, returns the best hit based on e-value.\n",
    "\n",
    "#     Parameters:\n",
    "#     ------------\n",
    "#     query_id : str\n",
    "#         The ID of the query sequence.\n",
    "#     hits : List\n",
    "#         List of HMMER hits to search through.\n",
    "\n",
    "#     Returns:\n",
    "#     ------------\n",
    "#     dict: A dictionary containing information about the best hit, including family ID, e-value, \n",
    "#           sequence length, bit score, high-scoring pair bias, domain e-value, start position, and end position.\n",
    "#           If no hit is found, returns None.\n",
    "#     \"\"\"\n",
    "#     best_hit = None\n",
    "#     for hit in hits:\n",
    "#         for hsp in hit.hsps:\n",
    "#             family_id = hit.id\n",
    "#             evalue = hit.evalue\n",
    "#             seqlen = hit.seq_len\n",
    "#             bitscore = hsp.bitscore\n",
    "#             devalue = hsp.evalue\n",
    "#             bias = hsp.bias\n",
    "#             start = hsp.query_start\n",
    "#             end = hsp.query_end\n",
    "#             if best_hit is None or evalue < best_hit[\"evalue\"]:\n",
    "#                 best_hit = {\"query_id\": query_id, \"family_id\": family_id, \"evalue\": evalue,\"seqlen\": seqlen ,\"bias\": bias, \"devalue\": devalue, \"bitscore\": bitscore, \"start\": start, \"end\": end}\n",
    "#     return best_hit\n",
    "\n",
    "# if __name__ == \"__main__\":\n",
    "#     # Define input and output filenames\n",
    "#     input_filename = \"input.fasta\"\n",
    "#     output_filename = \"output.domtblout\"\n",
    "\n",
    "#     # Define the Pfam database path\n",
    "#     pfam_path = \"/path/to/Pfam-A.hmm\"\n",
    "\n",
    "#     # Read in protein sequences from a FASTA file\n",
    "#     sequences = read_sequences(input_filename)\n",
    "\n",
    "#     # Write protein sequences to a FASTA file\n",
    "#     write_sequences(sequences, input_filename)\n",
    "\n",
    "#     # Run HMMER search against the Pfam database\n",
    "#     run_hmmer(pfam_path, input_filename, output_filename)\n",
    "\n",
    "#     # Parse HMMER results\n",
    "#     hits_dict = defaultdict(list)\n",
    "#     for query in SearchIO.parse(output_filename, 'hmmer3-domtbl'):\n",
    "#         query_id = query.id\n",
    "#         hits = query.hsps\n",
    "#         best_hit = find_best_hit(query_id, hits)\n",
    "#         if best_hit:\n",
    "#             hits_dict[best_hit['family_id']].append(best_hit)\n",
    "\n",
    "#     # Print summary of hits by family\n",
    "#     for family_id, hits in hits_dict.items():\n",
    "#         print(f\"Family ID: {family_id}\")\n",
    "#         print(f\"Number of hits: {len(hits)}\")\n",
    "#         evalues = [hit['evalue'] for hit in hits]\n",
    "#         avg_evalue = sum(evalues) / len(evalues)\n",
    "#         print(f\"Average e-value: {avg_evalue:.2g}\")\n",
    "#         print(\"\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from collections import Counter\n",
    "# import subprocess\n",
    "# from typing import Dict, List, Optional\n",
    "# from Bio import SeqIO\n",
    "# from Bio.SeqRecord import SeqRecord\n",
    "# from Bio.Seq import Seq\n",
    "# from Bio.Alphabet import generic_protein\n",
    "# from Bio import SearchIO\n",
    "\n",
    "\n",
    "# def read_sequences(filename: str, n: Optional[int] = None) -> List[SeqRecord]:\n",
    "#     \"\"\"Reads n protein sequences from a file and returns them as a list of SeqRecord objects.\"\"\"\n",
    "#     records = []\n",
    "#     with open(filename, \"r\") as f:\n",
    "#         for i, record in enumerate(SeqIO.parse(f, \"fasta\")):\n",
    "#             if n is not None and i >= n:\n",
    "#                 break\n",
    "#             records.append(record)\n",
    "#     return records\n",
    "\n",
    "\n",
    "# def write_sequences(records: List[SeqRecord], filename: str) -> None:\n",
    "#     \"\"\"Writes a list of SeqRecord objects to a FASTA file.\"\"\"\n",
    "#     with open(filename, \"w\", encoding=\"utf-8\") as f:\n",
    "#         SeqIO.write(records, f, \"fasta\")\n",
    "\n",
    "\n",
    "# def run_hmmer(pfam_path: str, input_filename: str, output_filename: str, num_cpu: int = 4) -> None:\n",
    "#     \"\"\"Runs HMMER search against the Pfam database using the hmmscan command.\"\"\"\n",
    "#     with open(output_filename, \"w\", encoding=\"utf-8\") as f:\n",
    "#         subprocess.run(\n",
    "#             [\n",
    "#                 \"hmmscan\",\n",
    "#                 \"--cpu\",\n",
    "#                 str(num_cpu),\n",
    "#                 \"--domtblout\",\n",
    "#                 output_filename,\n",
    "#                 pfam_path,\n",
    "#                 input_filename,\n",
    "#             ],\n",
    "#             stdout=f,\n",
    "#         )\n",
    "\n",
    "\n",
    "# def find_best_hit(query_id: str, hits: List) -> Optional[Dict]:\n",
    "#     \"\"\"\n",
    "#     Given a list of HMMER hits, returns the best hit based on e-value.\n",
    "\n",
    "#     Parameters:\n",
    "#     ------------\n",
    "#     query_id : str\n",
    "#         The ID of the query sequence.\n",
    "#     hits : List[SearchIO.Hit]\n",
    "#         List of HMMER hits.\n",
    "#     \"\"\"\n",
    "#     hits_by_evalue = sorted(hits, key=lambda hit: hit.evalue)\n",
    "#     for hit in hits_by_evalue:\n",
    "#         if hit.id == query_id:\n",
    "#             continue\n",
    "#         return {\n",
    "#             \"query_id\": query_id,\n",
    "#             \"hit_id\": hit.id,\n",
    "#             \"evalue\": hit.evalue,\n",
    "#             \"bitscore\": hit.bitscore,\n",
    "#         }\n",
    "#     return None\n",
    "\n",
    "\n",
    "# def count_domains(hits: List[SearchIO.Hit]) -> Dict[str, int]:\n",
    "#     \"\"\"Counts the frequency of each Pfam domain in a list of HMMER hits.\"\"\"\n",
    "#     domains = Counter(hit.accession.split(\".\")[0] for hit in hits)\n",
    "#     return domains\n",
    "\n",
    "\n",
    "# def main():\n",
    "#     # Paths to input/output files and the Pfam database\n",
    "#     input_filename = \"input.fasta\"\n",
    "#     output_filename = \"output.tbl\"\n",
    "#     pfam_path = \"/path/to/Pfam-A.hmm\"\n",
    "\n",
    "#     # Read input sequences from file\n",
    "#     input_seqs = read_sequences(input_filename)\n",
    "\n",
    "#     # Write input sequences to file\n",
    "#     input_seqs_file = \"input_seqs.fasta\"\n",
    "#     write_sequences(input_seqs, input_seqs_file)\n",
    "\n",
    "#     # Run HMMER search against the Pfam database\n",
    "#     run_hmmer(pfam_path, input_seqs_file, output_filename)\n",
    "\n",
    "#     # Parse the HMMER output\n",
    "#     hits = []\n",
    "#     with open(output_filename, \"r\") as f:\n",
    "#         for qresult in SearchIO.parse(f, \"hmmer3-tab\"):\n",
    "#             for hit in qresult.hits:\n",
    "#                 hits.append(hit)\n",
    "\n",
    "#     # Count the frequency of each Pfam domain in the hits\n",
    "#     domain_counts = count_domains(hits)\n",
    "\n",
    "#     # Find the best hit for each input sequence\n",
    "#     best_hits\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Optimizing the code"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Things I want to do:\n",
    "- Neeed to make the list of AA sequence a function, i.e. more modualr \n",
    "- "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run_hmmer(PFAM_PATH, \"test.fasta\", \"output_test.domtblout\", 4) # works!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parse_hmmer(\"output_test.domtblout\") # works!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Scrap\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# records = []\n",
    "# for i, seq in enumerate(meso_seq_list):\n",
    "#     record = SeqRecord(Seq(seq), id=str(i))\n",
    "#     records.append(record)\n",
    "\n",
    "# # Write the list of SeqRecord objects to a FASTA file\n",
    "# with open(\"input.fasta\", \"w\") as f:\n",
    "#     SeqIO.write(records, f, \"fasta\")\n",
    "\n",
    "\n",
    "# # Run HMMER search against the Pfam database using the hmmscan command\n",
    "# with open(\"output.domtblout\", \"w\") as f:\n",
    "#     subprocess.run([\"hmmscan\", \"--cpu\", \"4\", \"--domtblout\", \"output.domtblout\", PFAM_PATH, \"input.fasta\"], stdout=f)\n",
    "\n",
    "# # Define a function to find the best hit (hsps: high-scoring segment pair)\n",
    "# def find_best_hit(query_id, hits):\n",
    "#     \"\"\"\"\n",
    "#     Given a list of HMMER hits, returns the best hit based on e-value.\n",
    "    \n",
    "#     Parameters:\n",
    "#     ------------\n",
    "#     query_id : str \n",
    "#         The ID of the query sequence.\n",
    "#     hits : list \n",
    "#         List of HMMER hits to search through.\n",
    "    \n",
    "#     Returns:\n",
    "#     ------------\n",
    "#     dict :\n",
    "#         A dictionary containing information about the best hit, including family ID, e-value, \n",
    "#         sequence length, bit score, high-scoring pair bias, domain e-value, start position, and end position.\n",
    "#         If no hit is found, returns None.\n",
    "    \n",
    "#     Raises\n",
    "#     -------\n",
    "    \n",
    "#     \"\"\"\n",
    "#     best_hit = None\n",
    "#     for hit in hits:\n",
    "#         for hsp in hit.hsps:\n",
    "#             family_id = hit.id\n",
    "#             evalue = hit.evalue\n",
    "#             seqlen = hit.seq_len\n",
    "#             bitscore = hsp.bitscore\n",
    "#             devalue = hsp.evalue\n",
    "#             bias = hsp.bias\n",
    "#             start = hsp.query_start\n",
    "#             end = hsp.query_end\n",
    "#             if best_hit is None or evalue < best_hit[\"evalue\"]:\n",
    "#                 best_hit = {\"query_id\": query_id, \"family_id\": family_id, \"evalue\": evalue,\"seqlen\": seqlen ,\"bias\": bias, \"devalue\": devalue, \"bitscore\": bitscore, \"start\": start, \"end\": end}\n",
    "#     return best_hit\n",
    "\n",
    "# # Parse the HMMER results file using HmmerIO\n",
    "# with open(\"output.domtblout\", \"r\") as results_file:\n",
    "#     qresults = list(SearchIO.parse(results_file, 'hmmscan3-domtab'))\n",
    "\n",
    "# # Store the counts of hits for each family ID\n",
    "# hit_counts = defaultdict(int)\n",
    "# for qresult in qresults:\n",
    "#     query_id = qresult.id\n",
    "#     best_hit = find_best_hit(query_id, qresult.hits)\n",
    "#     if best_hit is not None:\n",
    "#         family_id = best_hit[\"family_id\"]\n",
    "#         hit_counts[family_id] += 1\n",
    "#         print(f\"Query sequence ID: {query_id}\")\n",
    "#         print(f\"Best Pfam family match: {family_id}\")\n",
    "#         print(f\"Total E-value: {best_hit['evalue']}\")\n",
    "#         print(f\"Length of hit sequence or HMM: {best_hit['seqlen']}\")\n",
    "#         print(f\"High-scoring pair bias: {best_hit['bias']}\")\n",
    "#         print(f\"Domain E-vlaue: {best_hit['devalue']}\")\n",
    "#         print(f\"Bit score: {best_hit['bitscore']}\")\n",
    "#         print(f\"Start position: {best_hit['start']}\")\n",
    "#         print(f\"End position: {best_hit['end']}\")\n",
    "#         print(\"----\")\n",
    "#     else:\n",
    "#         print(\"No Pfam family match found for query sequence ID:\", query_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ###################\n",
    "#     # reading the data\n",
    "#     protein_database = pd.read_csv(ID_DB_PATH, index_col=0)\n",
    "\n",
    "#     # separating the meso and thermo\n",
    "#     meso_seq_db = protein_database.loc[protein_database[\"identity\"] == \"m\"]\n",
    "#     thermo_seq_db = protein_database.loc[protein_database[\"identity\"] == \"t\"]\n",
    "\n",
    "#     ## processing meso to be suitable for HMMER\n",
    "#     meso_seq_list = meso_seq_db[[\"protein_seq\", \"protein_int_index\"]].set_index(\"protein_int_index\").iloc[:300]\n",
    "#     meso_seq_list.index.name = None\n",
    "\n",
    "#     def hmmer_wrapper_parallel(chunk_index, seq, which):\n",
    "#         hmmer_wrapper(seq=seq, input_filename=f\"{which}_input_{chunk_index}.fasta\", output_filename_with_ext=f\"{which}_output_{chunk_index}.domtblout\", pfam_path=PFAM_PATH, cpu=1)\n",
    "\n",
    "#     meso_chunks = [meso_seq_list[i:i+100] for i in range(0,len(meso_seq_list),100)]\n",
    "\n",
    "#     Parallel(n_jobs=3)(delayed(hmmer_wrapper_parallel)(i, chunk, \"meso\") for i, chunk in enumerate(meso_chunks))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Final version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a list of SeqRecord objects\n",
    "# Write the list of SeqRecord objects to a FASTA file\n",
    "def read_seq(lists: pd.core.frame.DataFrame, inputname: str = \"input\"):\n",
    "    \"\"\"\n",
    "    Returns a list of SeqRecord objects and creates a corresponding input Fasta of them\n",
    "\n",
    "    Parameters:\n",
    "    ------------\n",
    "    list : pandas.core.frame.DataFrame\n",
    "        a dataframe with string amino acid sequences in a 'seq' column\n",
    "    input name : str, default = 'input'\n",
    "        a name for the input fasta file\n",
    "\n",
    "    \n",
    "    Returns:\n",
    "    ------------\n",
    "    file : TextIOWrapper\n",
    "        the input fasta file created from the list of SeqRecord objects\n",
    "\n",
    "    Raises\n",
    "    -------\n",
    "    ValueError : \n",
    "        if the input dataframe is empty\n",
    "    AttributeError :\n",
    "        if any of the sequences are invalid\n",
    "    \"\"\"\n",
    "    # check if input is empty\n",
    "    if lists.empty:\n",
    "        raise ValueError(\"Input dataframe is empty\")\n",
    "    \n",
    "    # check if sequences are valid\n",
    "    for seq in lists['protein_seq']:\n",
    "        try:\n",
    "            Seq(seq)\n",
    "        except:\n",
    "            raise AttributeError(\"Invalid sequence\")\n",
    "\n",
    "    # function    \n",
    "    records = []\n",
    "    for index, seq in lists.itertuples():\n",
    "        try:\n",
    "            record = SeqRecord(Seq(seq), id=str(index))\n",
    "            records.append(record)\n",
    "        except AttributeError:\n",
    "            raise AttributeError(f\"Invalid sequence: {seq}\")\n",
    "    \n",
    "    # raise error if seq not valid\n",
    "    if not records:\n",
    "        raise AttributeError(\"No valid sequences found in input\")\n",
    "    \n",
    "    with open(f\"{inputname}.fasta\", \"w\") as file:\n",
    "            SeqIO.write(records, file, \"fasta\")\n",
    "    return file\n",
    "\n",
    "def run_hmmer(DB_PATH: str, input_filename: str, output_filename: str, num_cpu: int = 4) -> None:\n",
    "    \"\"\"\n",
    "    Runs HMMER search against the Pfam database using the hmmscan command\n",
    "    \n",
    "    Parameters:\n",
    "    ------------\n",
    "    DB_PATH : str\n",
    "        the path of the HMMER/pfam database\n",
    "    input_filename : str\n",
    "        a name for the input fasta file (don't forget extension!)\n",
    "    output_filename : str\n",
    "        a name for the domtblout output file (don't forget extension!)\n",
    "    num_cpu : int\n",
    "\n",
    "    \n",
    "    Returns:\n",
    "    ------------\n",
    "    domtblout : \n",
    "        an output domtblout file of the HMMER/pfam results\n",
    "\n",
    "    Raises\n",
    "    -------\n",
    "    TypeError\n",
    "    \"\"\"\n",
    "    with open(output_filename, \"w\", encoding=\"utf-8\") as file:\n",
    "        subprocess.run([\"hmmscan\", \"--cpu\", str(num_cpu), \"--domtblout\", output_filename, DB_PATH, input_filename], stdout=file)\n",
    "    return file\n",
    "\n",
    "# # Define a function to find the best hit (hsps: high-scoring segment pair)\n",
    "# def find_best_hit(query_id, hits):\n",
    "#     \"\"\"\"\n",
    "#     Given a list of HMMER hits, returns the best hit based on e-value.\n",
    "    \n",
    "#     Parameters:\n",
    "#     ------------\n",
    "#     query_id : str \n",
    "#         The ID of the query sequence.\n",
    "#     hits : list \n",
    "#         List of HMMER hits to search through.\n",
    "    \n",
    "#     Returns:\n",
    "#     ------------\n",
    "#     dict :\n",
    "#         A dictionary containing information about the best hit, including family ID, e-value, \n",
    "#         sequence length, bit score, high-scoring pair bias, domain e-value, start position, and end position.\n",
    "#         If no hit is found, returns None.\n",
    "    \n",
    "#     Raises\n",
    "#     -------\n",
    "    \n",
    "#     \"\"\"\n",
    "#     best_hit = None\n",
    "#     for hit in hits:\n",
    "#         for hsp in hit.hsps:\n",
    "#             family_id = hit.id\n",
    "#             evalue = hit.evalue\n",
    "#             seqlen = hit.seq_len\n",
    "#             bitscore = hsp.bitscore\n",
    "#             devalue = hsp.evalue\n",
    "#             bias = hsp.bias\n",
    "#             start = hsp.query_start\n",
    "#             end = hsp.query_end\n",
    "#             if best_hit is None or evalue < best_hit[\"evalue\"]:\n",
    "#                 best_hit = {\"query_id\": query_id, \"family_id\": family_id, \"evalue\": evalue,\"seqlen\": seqlen ,\"bias\": bias, \"devalue\": devalue, \"bitscore\": bitscore, \"start\": start, \"end\": end}\n",
    "#     return best_hit\n",
    "\n",
    "# # Parse the HMMER results file using HmmerIO\n",
    "# # Store the counts of hits for each family ID\n",
    "# def parse_hmmer(output_name: str):\n",
    "#     \"\"\"\n",
    "#     parses HMMER output result\n",
    "\n",
    "#     Parameters:\n",
    "#     ------------\n",
    "#     output_name : str\n",
    "#         a name for the domtblout output file (don't forget extension!)\n",
    "\n",
    "    \n",
    "#     Returns:\n",
    "#     ------------\n",
    "#     file : \n",
    "#         an input fasta file from a list of SeqRecord objects of originally string the amino acid sequences\n",
    "\n",
    "#     Raises\n",
    "#     -------\n",
    "#     TypeError\n",
    "#     \"\"\"\n",
    "#     # Parse the HMMER results file using HmmerIO\n",
    "#     with open(output_name, \"r\") as results_file:\n",
    "#         qresults = list(SearchIO.parse(results_file, 'hmmscan3-domtab'))\n",
    "\n",
    "#     # Store the counts of hits for each family ID\n",
    "#     hit_counts = defaultdict(int)\n",
    "#     for qresult in qresults:\n",
    "#         query_id = qresult.id\n",
    "#         best_hit = find_best_hit(query_id, qresult.hits)\n",
    "#         if best_hit is not None:\n",
    "#             family_id = best_hit[\"family_id\"]\n",
    "#             hit_counts[family_id] += 1\n",
    "#             print(f\"Query sequence ID: {query_id}\")\n",
    "#             print(f\"Best Pfam family match: {family_id}\")\n",
    "#             print(f\"Total E-value: {best_hit['evalue']}\")\n",
    "#             print(f\"Length of hit sequence or HMM: {best_hit['seqlen']}\")\n",
    "#             print(f\"High-scoring pair bias: {best_hit['bias']}\")\n",
    "#             print(f\"Domain E-vlaue: {best_hit['devalue']}\")\n",
    "#             print(f\"Bit score: {best_hit['bitscore']}\")\n",
    "#             print(f\"Start position: {best_hit['start']}\")\n",
    "#             print(f\"End position: {best_hit['end']}\")\n",
    "#             print(\"----\")\n",
    "#         else:\n",
    "#             print(\"No Pfam family match found for query sequence ID:\", query_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<_io.TextIOWrapper name='thermo_output.domtblout' mode='w' encoding='utf-8'>"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###### Execution (it works)\n",
    "\n",
    "# Define the amino acid sequence\n",
    "\n",
    "## Meso (+a bit of processing)\n",
    "meso_seq_pre = meso_seq_db[[\"protein_seq\", \"protein_int_index\"]]\n",
    "meso_seq_list = meso_seq_pre.set_index(\"protein_int_index\").iloc[:20]\n",
    "meso_seq_list.index.name = None\n",
    "\n",
    "\n",
    "## Thermo (+ a bit of processing)\n",
    "thermo_seq_pre = thermo_seq_db[[\"protein_seq\", \"protein_int_index\"]]\n",
    "thermo_seq_list = thermo_seq_pre.set_index(\"protein_int_index\").iloc[:20]\n",
    "thermo_seq_list.index.name = None\n",
    "\n",
    "\n",
    "# generate meso and thermo files\n",
    "read_seq(meso_seq_list, \"meso_input\")\n",
    "read_seq(thermo_seq_list, \"thermo_input\")\n",
    "\n",
    "# place files into HMMER/pfam\n",
    "run_hmmer(PFAM_PATH, \"meso_input.fasta\", \"meso_output.domtblout\", 4)\n",
    "run_hmmer(PFAM_PATH, \"thermo_input.fasta\", \"thermo_output.domtblout\", 4)\n",
    "\n",
    "# # parse outputs of HMMER/pfam\n",
    "# parse_hmmer(\"meso_output.domtblout\")\n",
    "# parse_hmmer(\"thermo_output.domtblout\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Attempting wrapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hmmer_wrapper(seq: pd.core.series.Series, input_filename: str, pfam_path: str, input_filename_with_ext: str, output_filename_with_ext: str, cpu: int = 4):\n",
    "    \"\"\"\n",
    "    Executes the HMMER against pfam and parsing the results\n",
    "\n",
    "    Parameters:\n",
    "    ------------\n",
    "    seq : str\n",
    "        a pandas series of of amino acids (has to be processed in a certain way)\n",
    "    input_filename : str\n",
    "        A file name for the input of the transformed seq to FASTA\n",
    "    pfam_path : str\n",
    "        path of the HMMER/pfam db\n",
    "    input_filename_with_ext : str\n",
    "        A file name for the input FASTA has to include the ext. FASTA\n",
    "    output_filename_with_ext : str\n",
    "        output file name perferred extension is domtblout\n",
    "    cpu : 4\n",
    "        number of cpus for i/o\n",
    "    \n",
    "    Returns:\n",
    "    ------------\n",
    "    Input fasta file\n",
    "    Output domtblout file\n",
    "        \n",
    "    Raises\n",
    "    -------\n",
    "    TypeError\n",
    "    \"\"\"\n",
    "    \n",
    "    # Create a list of SeqRecord objects\n",
    "    # Write the list of SeqRecord objects to a FASTA file\n",
    "    def read_seq(lists: pd.core.frame.DataFrame, inputname: str = \"input\"):\n",
    "        \"\"\"\n",
    "        Returns a list of SeqRecord objects and creates a corresponding input Fasta of them\n",
    "\n",
    "        Parameters:\n",
    "        ------------\n",
    "        list : pandas.core.frame.DataFrame\n",
    "            a dataframe with string amino acid sequences in a 'seq' column\n",
    "        input name : str, default = 'input'\n",
    "            a name for the input fasta file\n",
    "\n",
    "    \n",
    "        Returns:\n",
    "        ------------\n",
    "        file : TextIOWrapper\n",
    "            the input fasta file created from the list of SeqRecord objects\n",
    "\n",
    "        Raises\n",
    "        -------\n",
    "        ValueError : \n",
    "            if the input dataframe is empty\n",
    "        AttributeError :\n",
    "            if any of the sequences are invalid\n",
    "        \"\"\"\n",
    "        # check if input is empty\n",
    "        if lists.empty:\n",
    "            raise ValueError(\"Input dataframe is empty\")\n",
    "    \n",
    "        # check if sequences are valid\n",
    "        for seq in lists['seq']:\n",
    "            try:\n",
    "                Seq(seq)\n",
    "            except:\n",
    "                raise AttributeError(\"Invalid sequence\")\n",
    "\n",
    "        # function    \n",
    "        records = []\n",
    "        for index, seq in lists.itertuples():\n",
    "            try:\n",
    "                record = SeqRecord(Seq(seq), id=str(index))\n",
    "                records.append(record)\n",
    "            except AttributeError:\n",
    "                raise AttributeError(f\"Invalid sequence: {seq}\")\n",
    "    \n",
    "        # raise error if seq not valid\n",
    "        if not records:\n",
    "            raise AttributeError(\"No valid sequences found in input\")\n",
    "    \n",
    "        with open(f\"{inputname}.fasta\", \"w\") as file:\n",
    "            SeqIO.write(records, file, \"fasta\")\n",
    "        return file\n",
    "\n",
    "    def run_hmmer(DB_PATH: str, input_filename: str, output_filename: str, num_cpu: int = 4) -> None:\n",
    "        \"\"\"\n",
    "        Runs HMMER search against the Pfam database using the hmmscan command\n",
    "    \n",
    "        Parameters:\n",
    "        ------------\n",
    "        DB_PATH : str\n",
    "            the path of the HMMER/pfam database\n",
    "        input_filename : str\n",
    "            a name for the input fasta file (don't forget extension!)\n",
    "        output_filename : str\n",
    "            a name for the domtblout output file (don't forget extension!)\n",
    "        num_cpu : int\n",
    "\n",
    "    \n",
    "        Returns:\n",
    "        ------------\n",
    "        domtblout : \n",
    "            an output domtblout file of the HMMER/pfam results\n",
    "\n",
    "        Raises\n",
    "        -------\n",
    "        AttributeError\n",
    "        \"\"\"\n",
    "        with open(output_filename, \"w\", encoding=\"utf-8\") as file:\n",
    "            subprocess.run([\"hmmscan\", \"--cpu\", str(num_cpu), \"--domtblout\", output_filename, DB_PATH, input_filename], stdout=file)\n",
    "        return file\n",
    "\n",
    "    # Define a function to find the best hit (hsps: high-scoring segment pair)\n",
    "    def find_best_hit(query_id, hits):\n",
    "        \"\"\"\"\n",
    "        Given a list of HMMER hits, returns the best hit based on e-value.\n",
    "    \n",
    "        Parameters:\n",
    "        ------------\n",
    "        query_id : str \n",
    "            The ID of the query sequence.\n",
    "        hits : list \n",
    "            List of HMMER hits to search through.\n",
    "    \n",
    "        Returns:\n",
    "        ------------\n",
    "        dict :\n",
    "            A dictionary containing information about the best hit, including family ID, e-value, \n",
    "            sequence length, bit score, high-scoring pair bias, domain e-value, start position, and end position.\n",
    "            If no hit is found, returns None.\n",
    "    \n",
    "        Raises\n",
    "        -------\n",
    "    \n",
    "        \"\"\"\n",
    "        best_hit = None\n",
    "        for hit in hits:\n",
    "            for hsp in hit.hsps:\n",
    "                family_id = hit.id\n",
    "                evalue = hit.evalue\n",
    "                seqlen = hit.seq_len\n",
    "                bitscore = hsp.bitscore\n",
    "                devalue = hsp.evalue\n",
    "                bias = hsp.bias\n",
    "                start = hsp.query_start\n",
    "                end = hsp.query_end\n",
    "                if best_hit is None or evalue < best_hit[\"evalue\"]:\n",
    "                    best_hit = {\"query_id\": query_id, \"family_id\": family_id, \"evalue\": evalue,\"seqlen\": seqlen ,\"bias\": bias, \"devalue\": devalue, \"bitscore\": bitscore, \"start\": start, \"end\": end}\n",
    "        return best_hit\n",
    "\n",
    "    # Parse the HMMER results file using HmmerIO\n",
    "    # Store the counts of hits for each family ID\n",
    "    def parse_hmmer(output_name: str):\n",
    "        \"\"\"\n",
    "        parses HMMER output result\n",
    "\n",
    "        Parameters:\n",
    "        ------------\n",
    "        output_name : str\n",
    "            a name for the domtblout output file (don't forget extension!)\n",
    "\n",
    "    \n",
    "        Returns:\n",
    "        ------------\n",
    "        file : \n",
    "            an input fasta file from a list of SeqRecord objects of originally string the amino acid sequences\n",
    "\n",
    "        Raises\n",
    "        -------\n",
    "        TypeError\n",
    "        \"\"\"\n",
    "        # Parse the HMMER results file using HmmerIO\n",
    "        with open(output_name, \"r\") as results_file:\n",
    "            qresults = list(SearchIO.parse(results_file, 'hmmscan3-domtab'))\n",
    "\n",
    "        # Store the counts of hits for each family ID\n",
    "        hit_counts = defaultdict(int)\n",
    "        for qresult in qresults:\n",
    "            query_id = qresult.id\n",
    "            best_hit = find_best_hit(query_id, qresult.hits)\n",
    "            if best_hit is not None:\n",
    "                family_id = best_hit[\"family_id\"]\n",
    "                hit_counts[family_id] += 1\n",
    "                print(f\"Query sequence ID: {query_id}\")\n",
    "                print(f\"Best Pfam family match: {family_id}\")\n",
    "                print(f\"Total E-value: {best_hit['evalue']}\")\n",
    "                print(f\"Length of hit sequence or HMM: {best_hit['seqlen']}\")\n",
    "                print(f\"High-scoring pair bias: {best_hit['bias']}\")\n",
    "                print(f\"Domain E-vlaue: {best_hit['devalue']}\")\n",
    "                print(f\"Bit score: {best_hit['bitscore']}\")\n",
    "                print(f\"Start position: {best_hit['start']}\")\n",
    "                print(f\"End position: {best_hit['end']}\")\n",
    "                print(\"----\")\n",
    "            else:\n",
    "                print(\"No Pfam family match found for query sequence ID:\", query_id)\n",
    "    \n",
    "    ### wrapper execution\n",
    "    # User should define an amino acid sequence as Pandas series\n",
    "    \n",
    "    # generate meso and thermo files\n",
    "    read_seq(seq, input_filename)\n",
    "\n",
    "    # place files into HMMER/pfam\n",
    "    run_hmmer(pfam_path, input_filename_with_ext, output_filename_with_ext, cpu)\n",
    "\n",
    "    # parse outputs of HMMER/pfam\n",
    "    parse_hmmer(output_filename_with_ext)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query sequence ID: 60589906\n",
      "Best Pfam family match: Band_7\n",
      "Total E-value: 2.2e-30\n",
      "Length of hit sequence or HMM: 180\n",
      "High-scoring pair bias: 0.5\n",
      "Domain E-vlaue: 3.2e-30\n",
      "Bit score: 105.6\n",
      "Start position: 21\n",
      "End position: 193\n",
      "----\n",
      "Query sequence ID: 60589907\n",
      "Best Pfam family match: Band_7\n",
      "Total E-value: 7.9e-23\n",
      "Length of hit sequence or HMM: 180\n",
      "High-scoring pair bias: 0.3\n",
      "Domain E-vlaue: 1.4e-22\n",
      "Bit score: 80.7\n",
      "Start position: 34\n",
      "End position: 200\n",
      "----\n",
      "Query sequence ID: 60589911\n",
      "Best Pfam family match: YiiD_C\n",
      "Total E-value: 1.3e-41\n",
      "Length of hit sequence or HMM: 144\n",
      "High-scoring pair bias: 0.1\n",
      "Domain E-vlaue: 1.9e-41\n",
      "Bit score: 141.5\n",
      "Start position: 155\n",
      "End position: 294\n",
      "----\n",
      "Query sequence ID: 60589913\n",
      "Best Pfam family match: Tyr_Deacylase\n",
      "Total E-value: 1.6e-52\n",
      "Length of hit sequence or HMM: 144\n",
      "High-scoring pair bias: 0.0\n",
      "Domain E-vlaue: 1.8e-52\n",
      "Bit score: 177.8\n",
      "Start position: 2\n",
      "End position: 143\n",
      "----\n",
      "Query sequence ID: 60589920\n",
      "Best Pfam family match: DUF4124\n",
      "Total E-value: 2.1e-07\n",
      "Length of hit sequence or HMM: 56\n",
      "High-scoring pair bias: 0.0\n",
      "Domain E-vlaue: 3.9e-07\n",
      "Bit score: 30.4\n",
      "Start position: 16\n",
      "End position: 52\n",
      "----\n",
      "Query sequence ID: 60589921\n",
      "Best Pfam family match: HATPase_c\n",
      "Total E-value: 4.1e-19\n",
      "Length of hit sequence or HMM: 112\n",
      "High-scoring pair bias: 0.0\n",
      "Domain E-vlaue: 2600.0\n",
      "Bit score: -1.0\n",
      "Start position: 89\n",
      "End position: 110\n",
      "----\n",
      "Query sequence ID: 60589923\n",
      "Best Pfam family match: tRNA-synt_2\n",
      "Total E-value: 6.7e-43\n",
      "Length of hit sequence or HMM: 314\n",
      "High-scoring pair bias: 0.1\n",
      "Domain E-vlaue: 2e-42\n",
      "Bit score: 145.5\n",
      "Start position: 10\n",
      "End position: 322\n",
      "----\n",
      "Query sequence ID: 60589924\n",
      "Best Pfam family match: THF_DHG_CYH_C\n",
      "Total E-value: 1.7e-66\n",
      "Length of hit sequence or HMM: 161\n",
      "High-scoring pair bias: 3.7\n",
      "Domain E-vlaue: 2.4e-66\n",
      "Bit score: 222.1\n",
      "Start position: 120\n",
      "End position: 284\n",
      "----\n",
      "Query sequence ID: 60589935\n",
      "Best Pfam family match: DTW\n",
      "Total E-value: 1.9e-44\n",
      "Length of hit sequence or HMM: 202\n",
      "High-scoring pair bias: 4.3\n",
      "Domain E-vlaue: 2.1e-44\n",
      "Bit score: 151.6\n",
      "Start position: 2\n",
      "End position: 205\n",
      "----\n",
      "Query sequence ID: 60589938\n",
      "Best Pfam family match: PALP\n",
      "Total E-value: 5.3e-63\n",
      "Length of hit sequence or HMM: 294\n",
      "High-scoring pair bias: 0.5\n",
      "Domain E-vlaue: 6.6e-63\n",
      "Bit score: 213.0\n",
      "Start position: 10\n",
      "End position: 303\n",
      "----\n",
      "Query sequence ID: 60589942\n",
      "Best Pfam family match: DUF599\n",
      "Total E-value: 2.9e-58\n",
      "Length of hit sequence or HMM: 210\n",
      "High-scoring pair bias: 7.9\n",
      "Domain E-vlaue: 3.4e-58\n",
      "Bit score: 196.6\n",
      "Start position: 8\n",
      "End position: 218\n",
      "----\n",
      "Query sequence ID: 60589944\n",
      "Best Pfam family match: RNase_T\n",
      "Total E-value: 1.1e-28\n",
      "Length of hit sequence or HMM: 165\n",
      "High-scoring pair bias: 0.1\n",
      "Domain E-vlaue: 1.4e-28\n",
      "Bit score: 100.6\n",
      "Start position: 11\n",
      "End position: 175\n",
      "----\n",
      "Query sequence ID: 60589946\n",
      "Best Pfam family match: Y1_Tnp\n",
      "Total E-value: 4e-06\n",
      "Length of hit sequence or HMM: 122\n",
      "High-scoring pair bias: 0.0\n",
      "Domain E-vlaue: 1.4e-05\n",
      "Bit score: 25.3\n",
      "Start position: 14\n",
      "End position: 185\n",
      "----\n",
      "Query sequence ID: 60589951\n",
      "Best Pfam family match: tRNA_edit\n",
      "Total E-value: 9.4e-22\n",
      "Length of hit sequence or HMM: 123\n",
      "High-scoring pair bias: 0.5\n",
      "Domain E-vlaue: 1.3e-21\n",
      "Bit score: 77.0\n",
      "Start position: 31\n",
      "End position: 141\n",
      "----\n",
      "Query sequence ID: 60589954\n",
      "Best Pfam family match: PqiA\n",
      "Total E-value: 3.5e-88\n",
      "Length of hit sequence or HMM: 156\n",
      "High-scoring pair bias: 10.8\n",
      "Domain E-vlaue: 9.3e-34\n",
      "Bit score: 116.6\n",
      "Start position: 52\n",
      "End position: 205\n",
      "----\n"
     ]
    }
   ],
   "source": [
    "##### Testing wrapper\n",
    "## Meso (+a bit of processing)\n",
    "meso_seq_pre = meso_seq_db[[\"protein_seq\", \"protein_int_index\"]]\n",
    "meso_seq_list = meso_seq_pre.set_index(\"protein_int_index\").iloc[:15]\n",
    "meso_seq_list.index.name = None\n",
    "\n",
    "\n",
    "## Thermo (+ a bit of processing)\n",
    "thermo_seq_pre = thermo_seq_db[[\"protein_seq\", \"protein_int_index\"]]\n",
    "thermo_seq_list = thermo_seq_pre.set_index(\"protein_int_index\").iloc[:15]\n",
    "thermo_seq_list.index.name = None\n",
    "\n",
    "hmmer_wrapper(meso_seq_list, \"meso_input\", PFAM_PATH, \"meso_input.fasta\", \"meso_output.domtblout\", 4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Put best pfam metrics as a column in the original DB make sure they correspond to the original IDs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### ideas for unit tests:\n",
    "\n",
    "- empty pfam input\n",
    "- collections key error\n",
    "```py \n",
    "from collections import defaultdict\n",
    "# create a defaultdict that returns 0 as default value\n",
    "my_dict = defaultdict(int)\n",
    "\n",
    "# add some values\n",
    "my_dict['a'] = 1\n",
    "my_dict['b'] = 2\n",
    "\n",
    "# try to access a nonexistent key\n",
    "print(my_dict['c'])  # prints 0, instead of raising KeyError\n",
    "```\n",
    "- pfam output error handling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Unit test developmemt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Functions to tests:\n",
    "- read_seq\n",
    "    - {done}\n",
    "- run_hmmer\n",
    "    - {very hard}\n",
    "- find_best_hits\n",
    "    - {placeholder}\n",
    "- parse_hmmer\n",
    "    - {placeholder}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from io import StringIO\n",
    "import tempfile\n",
    "from tempfile import TemporaryDirectory\n",
    "import os\n",
    "import unittest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "......\n",
      "----------------------------------------------------------------------\n",
      "Ran 6 tests in 0.017s\n",
      "\n",
      "OK\n"
     ]
    }
   ],
   "source": [
    "class TestReadSeq(unittest.TestCase):\n",
    "    \"\"\"\n",
    "    Tests for read_seq function\n",
    "    \"\"\"\n",
    "\n",
    "    # inputs in setup function\n",
    "    def setUp(self):\n",
    "        self.sequences = pd.DataFrame({'protein_seq': ['MTITLVEGSEQLRKQVAYVEELRS',\n",
    "                                           'MFRDYVYDYLMRLWRVEYRH',\n",
    "                                           'MVTVLHESDQSLQGQAYLAEELRS',\n",
    "                                           'MTRTLVEGSEQLRKQVAYVEELRS']})\n",
    "        self.inputname = 'test_input'\n",
    "        self.empty_seq = pd.DataFrame({'protein_seq': []})\n",
    "        self.invalid_seq = pd.DataFrame({'protein_seq': ['AAGCTYY']})\n",
    "        read_seq(self.sequences, self.inputname)\n",
    "\n",
    "        # create a new empty dataframe instance\n",
    "        self.empty_df = pd.DataFrame()\n",
    "        \n",
    "    ## Methods/tests\n",
    "        \n",
    "    def test_file_created(self):\n",
    "        \"\"\" Tests if the file input is created\"\"\"\n",
    "        self.assertTrue(os.path.isfile(f\"{self.inputname}.fasta\"))\n",
    "        \n",
    "    def test_file_contains_correct_sequences(self):\n",
    "        \"\"\" Tests if the input is formatted correctly\"\"\"\n",
    "        read_seq(self.sequences, self.inputname)\n",
    "        records = list(SeqIO.parse(f\"{self.inputname}.fasta\", \"fasta\"))\n",
    "        self.assertEqual(len(records), len(self.sequences))\n",
    "        for i, record in enumerate(records):\n",
    "            self.assertEqual(str(record.seq), self.sequences.iloc[i]['protein_seq'])\n",
    "    \n",
    "    def test_invalid_input(self):\n",
    "        \"\"\"Tests if the function raises an error with invalid input\"\"\"\n",
    "        with self.assertRaises(ValueError):\n",
    "            read_seq(self.empty_df, self.inputname)\n",
    "    \n",
    "    def test_invalid_type(self):\n",
    "        \"\"\"Tests if the function raises an error when given an invalid type\"\"\"\n",
    "        with self.assertRaises(AttributeError):\n",
    "            read_seq(123, self.inputname)\n",
    "        \n",
    "    def test_empty_sequences(self):\n",
    "        \"\"\"Tests if the function raises an error when given an empty DataFrame\"\"\"\n",
    "        with self.assertRaises(ValueError) as cm:\n",
    "            read_seq(self.empty_df, self.inputname)\n",
    "        self.assertEqual(str(cm.exception), \"Input dataframe is empty\")\n",
    "        \n",
    "    def test_invalid_sequence(self):\n",
    "        \"\"\"Tests if the function raises an error when given an invalid sequence\"\"\"\n",
    "        with self.assertRaises(AttributeError) as cm:\n",
    "            read_seq(self.invalid_seq, self.inputname)\n",
    "            print(cm.exception)\n",
    "\n",
    "\n",
    "    # tear down of set up\n",
    "    def tearDown(self):\n",
    "        try:\n",
    "            os.remove(f\"{self.inputname}.fasta\")\n",
    "        except FileNotFoundError:\n",
    "            pass\n",
    "    \n",
    "suite = unittest.TestLoader().loadTestsFromTestCase(TestReadSeq)\n",
    "_ = unittest.TextTestRunner().run(suite) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Error: Failed to open binary auxfiles for /Users/humoodalanzi/pfam/hmmer_test/Pfam-A.hmm: use hmmpress first\n",
      "\n",
      "F\n",
      "Error: File format problem, trying to open HMM file /Users/humoodalanzi/Desktop.\n",
      "File exists, but appears to be empty?\n",
      "\n",
      "F\n",
      "Error: Failed to open binary auxfiles for /Users/humoodalanzi/pfam/hmmer_test/Pfam-A.hmm: use hmmpress first\n",
      "\n",
      "F\n",
      "Error: Failed to open binary auxfiles for /Users/humoodalanzi/pfam/hmmer_test/Pfam-A.hmm: use hmmpress first\n",
      "\n",
      ".\n",
      "======================================================================\n",
      "FAIL: test_empty_file (__main__.TestRunHmmer)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/var/folders/0x/cl_8377x4478rhw2_0zwyxdw0000gn/T/ipykernel_27581/2388465039.py\", line 51, in test_empty_file\n",
      "    run_hmmer(self.db_path, self.fasta_file, self.output_path, num_cpu=1)\n",
      "AssertionError: AssertionError not raised\n",
      "\n",
      "======================================================================\n",
      "FAIL: test_invalid_database (__main__.TestRunHmmer)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/var/folders/0x/cl_8377x4478rhw2_0zwyxdw0000gn/T/ipykernel_27581/2388465039.py\", line 66, in test_invalid_database\n",
      "    run_hmmer(\"/Users/humoodalanzi/Desktop\", self.fasta_file, self.output_path, num_cpu=1)\n",
      "AssertionError: FileNotFoundError not raised\n",
      "\n",
      "======================================================================\n",
      "FAIL: test_invalid_fasta_file (__main__.TestRunHmmer)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/var/folders/0x/cl_8377x4478rhw2_0zwyxdw0000gn/T/ipykernel_27581/2388465039.py\", line 61, in test_invalid_fasta_file\n",
      "    run_hmmer(self.db_path, self.fasta_file, self.output_path, num_cpu=1)\n",
      "AssertionError: FileNotFoundError not raised\n",
      "\n",
      "----------------------------------------------------------------------\n",
      "Ran 4 tests in 0.069s\n",
      "\n",
      "FAILED (failures=3)\n"
     ]
    }
   ],
   "source": [
    "class TestRunHmmer(unittest.TestCase):\n",
    "    \"\"\"Tests for run_hmmer function {Hell}\"\"\"\n",
    "    \n",
    "    \n",
    "    def setUp(self):\n",
    "        # Set up a temporary directory to hold output files\n",
    "        self.temp_dir = tempfile.TemporaryDirectory()\n",
    "\n",
    "        # Create synthetic input FASTA file\n",
    "        self.input_filename = \"test_input.fasta\"\n",
    "        input_filepath = os.path.join(self.temp_dir.name, self.input_filename)\n",
    "        with open(input_filepath, \"w\") as f:\n",
    "            f.write(\">seq1\\nATCG\\n>seq2\\nGCTA\\n\")\n",
    "\n",
    "        # Create a synthetic output file\n",
    "        output_filename = \"synthetic_output.txt\"\n",
    "        output_path = os.path.join(self.temp_dir.name, output_filename)\n",
    "        with open(output_path, \"w\") as f:\n",
    "            f.write(\"# HMMER3/f [3.3]\\n\")\n",
    "            f.write(\"# hmmer :: {{build}} :: {{3.3}}\\n\")\n",
    "            f.write(\"# target sequence database: /Users/humoodalanzi/pfam/hmmer_test/Pfam-A.hmm\\n\")\n",
    "            f.write(\"# target sequence input file: synthetic_input.fasta\\n\")\n",
    "            f.write(\"# output directed to: stdout\\n\")\n",
    "            f.write(\"#\\n\")\n",
    "            f.write(\"# Query sequence name                    Accession   E-value  Score (bits)   Strand  Start    End   \")\n",
    "            f.write(\"#------------------------------------------------------------------------------\")\n",
    "            f.write(\"     ----- no hits -----\\n\")\n",
    "\n",
    "        # Set up the paths to input and output files\n",
    "        self.db_path = \"/Users/humoodalanzi/pfam/hmmer_test/Pfam-A.hmm\"\n",
    "        self.fasta_file = \"synthetic_input.fasta\"\n",
    "        self.output_filename = \"synthetic_output.domtblout\"\n",
    "        self.output_path = os.path.join(self.temp_dir.name, self.output_filename)\n",
    "\n",
    "\n",
    "    def test_valid_input(self):\n",
    "        # Run the HMMER search with valid input\n",
    "        run_hmmer(self.db_path, self.fasta_file, self.output_path, num_cpu=1)\n",
    "\n",
    "        # Check that the output file was created\n",
    "        self.assertTrue(os.path.exists(self.output_path))\n",
    "\n",
    "    def test_empty_file(self):\n",
    "    # Create an empty input file\n",
    "        with open(self.fasta_file, \"w\") as f:\n",
    "            f.write(\"\")\n",
    "            f.flush()  # make sure contents are written to disk\n",
    "\n",
    "        # Run the HMMER search with the empty input file\n",
    "        with self.assertRaises(AssertionError):\n",
    "            run_hmmer(self.db_path, self.fasta_file, self.output_path, num_cpu=1)\n",
    "\n",
    "    def test_invalid_fasta_file(self):\n",
    "        # Create an invalid input file\n",
    "        with open(self.fasta_file, \"w\") as f:\n",
    "            f.write(\">header\\n\")\n",
    "            f.write(\"ACDEFGHIKLMNPQRSTVWY\\n\")\n",
    "\n",
    "        # Run the HMMER search with the invalid input file\n",
    "        with self.assertRaises(FileNotFoundError):\n",
    "            run_hmmer(self.db_path, self.fasta_file, self.output_path, num_cpu=1)\n",
    "\n",
    "    def test_invalid_database(self):\n",
    "    # Run the HMMER search with an invalid database path\n",
    "        with self.assertRaises(FileNotFoundError):\n",
    "            run_hmmer(\"/Users/humoodalanzi/Desktop\", self.fasta_file, self.output_path, num_cpu=1)\n",
    "\n",
    "    def tearDown(self):\n",
    "        # Clean up the temporary directory\n",
    "        self.temp_dir.cleanup()\n",
    "\n",
    "suite = unittest.TestLoader().loadTestsFromTestCase(TestRunHmmer)\n",
    "_ = unittest.TextTestRunner().run(suite)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class TestParseHmmer(unittest.TestCase):\n",
    "    \"\"\"Tests for parse_hmmer function\"\"\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "suite = unittest.TestLoader().loadTestsFromTestCase(TestParseHmmer)\n",
    "_ = unittest.TextTestRunner().run(suite)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class TestBestHits(unittest.TestCase):\n",
    "    \"\"\"Tests for best_hits function {not done}\"\"\"\n",
    "    \n",
    "    def setUp(self):\n",
    "        # Define test hits\n",
    "        self.hits = [\n",
    "            HmmerHit(\"PF00001\", 1, 100, [HmmerHsp(1, 100, 50, 0.01)]),\n",
    "            HmmerHit(\"PF00002\", 1, 200, [HmmerHsp(1, 200, 100, 0.001)]),\n",
    "            HmmerHit(\"PF00003\", 1, 300, [HmmerHsp(1, 300, 150, 0.0001)]),\n",
    "        ]\n",
    "\n",
    "    def test_find_best_hit(self):\n",
    "        # Test finding best hit\n",
    "        query_id = \"query1\"\n",
    "        best_hit = find_best_hit(query_id, self.hits)\n",
    "        expected_hit = {\n",
    "            \"query_id\": query_id,\n",
    "            \"family_id\": \"PF00003\",\n",
    "            \"evalue\": 0.0001,\n",
    "            \"seqlen\": 300,\n",
    "            \"bias\": 150,\n",
    "            \"devalue\": 0.0001,\n",
    "            \"bitscore\": 150,\n",
    "            \"start\": 1,\n",
    "            \"end\": 300,\n",
    "        }\n",
    "        self.assertEqual(best_hit, expected_hit)\n",
    "\n",
    "    def test_no_hits(self):\n",
    "        # Test when no hits are found\n",
    "        query_id = \"query2\"\n",
    "        best_hit = find_best_hit(query_id, [])\n",
    "        self.assertIsNone(best_hit)\n",
    "\n",
    "suite = unittest.TestLoader().loadTestsFromTestCase(TestBestHits)\n",
    "_ = unittest.TextTestRunner().run(suite)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class TestHmmerWrapper(unittest.TestCase):\n",
    "    \"\"\"Tests for hmmer_wrapper function\"\"\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "suite = unittest.TestLoader().loadTestsFromTestCase(TestHmmerWrapper)\n",
    "_ = unittest.TextTestRunner().run(suite) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### functional pair finding development"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### old stuff"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```py\n",
    "# Define a function to find the best hit (hsps: high-scoring segment pair)\n",
    "def find_best_hit(query_id, hits):\n",
    "    \"\"\"\"\n",
    "    Given a list of HMMER hits, returns the best hit based on e-value.\n",
    "    \n",
    "    Parameters:\n",
    "    ------------\n",
    "    query_id : str \n",
    "        The ID of the query sequence.\n",
    "    hits : list \n",
    "        List of HMMER hits to search through.\n",
    "    \n",
    "    Returns:\n",
    "    ------------\n",
    "    dict :\n",
    "        A dictionary containing information about the best hit, including family ID, e-value, \n",
    "        sequence length, bit score, high-scoring pair bias, domain e-value, start position, and end position.\n",
    "        If no hit is found, returns None.\n",
    "    \n",
    "    Raises\n",
    "    -------\n",
    "    \n",
    "    \"\"\"\n",
    "    best_hit = None\n",
    "    for hit in hits:\n",
    "        for hsp in hit.hsps:\n",
    "            family_id = hit.id\n",
    "            evalue = hit.evalue\n",
    "            seqlen = hit.seq_len\n",
    "            bitscore = hsp.bitscore\n",
    "            devalue = hsp.evalue\n",
    "            bias = hsp.bias\n",
    "            start = hsp.query_start\n",
    "            end = hsp.query_end\n",
    "            if best_hit is None or evalue < best_hit[\"evalue\"]:\n",
    "                best_hit = {\"query_id\": query_id, \"family_id\": family_id, \"evalue\": evalue,\"seqlen\": seqlen ,\"bias\": bias, \"devalue\": devalue, \"bitscore\": bitscore, \"start\": start, \"end\": end}\n",
    "    return best_hit\n",
    "\n",
    "# Parse the HMMER results file using HmmerIO\n",
    "# Store the counts of hits for each family ID\n",
    "def parse_hmmer(output_name: str):\n",
    "    \"\"\"\n",
    "    parses HMMER output result\n",
    "\n",
    "    Parameters:\n",
    "    ------------\n",
    "    output_name : str\n",
    "        a name for the domtblout output file (don't forget extension!)\n",
    "\n",
    "    \n",
    "    Returns:\n",
    "    ------------\n",
    "    file : \n",
    "        an input fasta file from a list of SeqRecord objects of originally string the amino acid sequences\n",
    "\n",
    "    Raises\n",
    "    -------\n",
    "    TypeError\n",
    "    \"\"\"\n",
    "    # Parse the HMMER results file using HmmerIO\n",
    "    with open(output_name, \"r\") as results_file:\n",
    "        qresults = list(SearchIO.parse(results_file, 'hmmscan3-domtab'))\n",
    "\n",
    "    # Store the counts of hits for each family ID\n",
    "    hit_counts = defaultdict(int)\n",
    "    for qresult in qresults:\n",
    "        query_id = qresult.id\n",
    "        best_hit = find_best_hit(query_id, qresult.hits)\n",
    "        if best_hit is not None:\n",
    "            family_id = best_hit[\"family_id\"]\n",
    "            hit_counts[family_id] += 1\n",
    "            print(f\"Query sequence ID: {query_id}\")\n",
    "            print(f\"Best Pfam family match: {family_id}\")\n",
    "            print(f\"Total E-value: {best_hit['evalue']}\")\n",
    "            print(f\"Length of hit sequence or HMM: {best_hit['seqlen']}\")\n",
    "            print(f\"High-scoring pair bias: {best_hit['bias']}\")\n",
    "            print(f\"Domain E-vlaue: {best_hit['devalue']}\")\n",
    "            print(f\"Bit score: {best_hit['bitscore']}\")\n",
    "            print(f\"Start position: {best_hit['start']}\")\n",
    "            print(f\"End position: {best_hit['end']}\")\n",
    "            print(\"----\")\n",
    "        else:\n",
    "            print(\"No Pfam family match found for query sequence ID:\", query_id)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_matching_pairs(hitsA, hitsB, how: str, **kwargs):\n",
    "    \"\"\"\n",
    "    Parses the HMMER output and compute the protein pair functional matching\n",
    "    \n",
    "    Parameters:\n",
    "    ------------\n",
    "    hitsA : \n",
    "        TODO\n",
    "    hitsB :\n",
    "        TODO\n",
    "    how : str\n",
    "        method to test hitsA and hitsB:\n",
    "            - \"best-hit-match\"\n",
    "            - a more sophisticated method (TODO)\n",
    "\n",
    "\n",
    "    Returns:\n",
    "    ------------\n",
    "    TODO : \n",
    "        TODO\n",
    "\n",
    "    Raises\n",
    "    -------\n",
    "    AttributeError :\n",
    "        TODO\n",
    "    \"\"\"\n",
    "    # \n",
    "    if len(hitsA)==0 or len(hitsB)==0\n",
    "        status = (len(hitsA)<0, len(hitsB)<0)\n",
    "        if all(status):\n",
    "            None\n",
    "        elif any(status):\n",
    "            False\n",
    "        return\n",
    "    \n",
    "    #\n",
    "    if how == \"get-best-hit\":\n",
    "        hitA = get_best_hit(hitsA)\n",
    "        hitB = get_best_hit(hitsB)\n",
    "    \n",
    "    else:\n",
    "        raise NotImplementedError(f\"Type {how} not supported yet\")\n",
    "    \n",
    "    ### how methods\n",
    "    # Define a function to find the best hit (hsps: high-scoring segment pair)\n",
    "def find_best_hit(query_id, hits):\n",
    "    \"\"\"\"\n",
    "    Given a list of HMMER hits, returns the best hit based on e-value.\n",
    "\n",
    "    Parameters:\n",
    "    ------------\n",
    "    query_id : str \n",
    "        The ID of the query sequence.\n",
    "    hits : list \n",
    "        List of HMMER hits to search through.\n",
    "\n",
    "    Returns:\n",
    "    ------------\n",
    "    dict :\n",
    "        A dictionary containing information about the best hit, including family ID, e-value, \n",
    "        sequence length, bit score, high-scoring pair bias, domain e-value, start position, and end position.\n",
    "        If no hit is found, returns None.\n",
    "\n",
    "    Raises\n",
    "    -------\n",
    "\n",
    "    \"\"\"\n",
    "    best_hit = None\n",
    "    for hit in hits:\n",
    "        for hsp in hit.hsps:\n",
    "            family_id = hit.id\n",
    "            evalue = hit.evalue\n",
    "            seqlen = hit.seq_len\n",
    "            bitscore = hsp.bitscore\n",
    "            devalue = hsp.evalue\n",
    "            bias = hsp.bias\n",
    "            start = hsp.query_start\n",
    "            end = hsp.query_end\n",
    "            if best_hit is None or evalue < best_hit[\"evalue\"]:\n",
    "                best_hit = {\"query_id\": query_id, \"family_id\": family_id, \"evalue\": evalue,\"seqlen\": seqlen ,\"bias\": bias, \"devalue\": devalue, \"bitscore\": bitscore, \"start\": start, \"end\": end}\n",
    "    return best_hit"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "vscode": {
   "interpreter": {
    "hash": "4a9b5cea0784dd84a54eb32a9034ea0f7d299d9a522482e4c348f8e08377d626"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
