{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d12015c8",
   "metadata": {},
   "source": [
    "Progress:\n",
    "\n",
    "DONE:\n",
    "-get package to work from command line\n",
    "-generate output from fasta files\n",
    "-programatically get package to work on one sequence\n",
    "-figure out indexing/confirm that package runs through dataframe in order\n",
    "\n",
    "NOT DONE:\n",
    "-send outfiles files into dataframe\n",
    "-get command line argument incorporated into script\n",
    "-scale beyond 5 sequences\n",
    "-train model with embeddings and compare results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "61c53eee",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats\n",
    "import sklearn.preprocessing\n",
    "import sklearn.model_selection\n",
    "import sklearn.ensemble\n",
    "import sklearn.feature_selection\n",
    "import unittest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "aae5ffbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "import esm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ba87d47f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/loganroberts/Learn2Therm/ValidProt/notebooks\n"
     ]
    }
   ],
   "source": [
    "cd /Users/loganroberts/Learn2Therm/ValidProt/notebooks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f90c1cdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#convert to pandas df\n",
    "df = pd.read_csv('learn2therm_sample_50k_exploration.csv').sample(200,random_state=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "51d90c54",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>thermo_index</th>\n",
       "      <th>...</th>\n",
       "      <th>bit_score_16s</th>\n",
       "      <th>m_ogt</th>\n",
       "      <th>t_ogt</th>\n",
       "      <th>ogt_difference</th>\n",
       "      <th>m_protein_seq</th>\n",
       "      <th>t_protein_seq</th>\n",
       "      <th>m_protein_desc</th>\n",
       "      <th>t_protein_desc</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>33553</th>\n",
       "      <td>33553</td>\n",
       "      <td>0.303922</td>\n",
       "      <td>0.276786</td>\n",
       "      <td>0.216783</td>\n",
       "      <td>199</td>\n",
       "      <td>0.888393</td>\n",
       "      <td>211</td>\n",
       "      <td>0.606322</td>\n",
       "      <td>221</td>\n",
       "      <td>9827</td>\n",
       "      <td>...</td>\n",
       "      <td>1004.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>MTNCKVDNKGWINVSDGSQLLDIQHLHTGFRIKDEYYDAVDDVSFE...</td>\n",
       "      <td>MDIKLHHVNKSYETGVETINVLKDIDLTIESGSWLTITGPSGSGKT...</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>348</td>\n",
       "      <td>224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9427</th>\n",
       "      <td>9427</td>\n",
       "      <td>0.303371</td>\n",
       "      <td>0.173633</td>\n",
       "      <td>0.190476</td>\n",
       "      <td>178</td>\n",
       "      <td>0.572347</td>\n",
       "      <td>196</td>\n",
       "      <td>0.765625</td>\n",
       "      <td>70</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>1035.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>MRNVAALSMPPRVPVEEGRTFTVDEGKRLLDAVADHRMGVLVLLAL...</td>\n",
       "      <td>MSGTRPDRVIRAFVDHLVVERGMARNTVDSYARDLRRYHAHLQARG...</td>\n",
       "      <td>site-specific integrase</td>\n",
       "      <td>site-specific tyrosine recombinase XerD</td>\n",
       "      <td>256</td>\n",
       "      <td>311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>199</td>\n",
       "      <td>0.292453</td>\n",
       "      <td>0.185629</td>\n",
       "      <td>0.184799</td>\n",
       "      <td>213</td>\n",
       "      <td>0.637725</td>\n",
       "      <td>215</td>\n",
       "      <td>0.637982</td>\n",
       "      <td>140</td>\n",
       "      <td>11324</td>\n",
       "      <td>...</td>\n",
       "      <td>989.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MFTLKHATAASAVVLGLVFATETFAADVTWNVSLWGKRRAFTEHVE...</td>\n",
       "      <td>MKIAREFAGALAGTLAGFAMAAVTALPVFAEELKFANFTPPFHTIN...</td>\n",
       "      <td>C4-dicarboxylate TRAP transporter substrate-bi...</td>\n",
       "      <td>TRAP transporter substrate-binding protein</td>\n",
       "      <td>337</td>\n",
       "      <td>334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12447</th>\n",
       "      <td>12447</td>\n",
       "      <td>0.406250</td>\n",
       "      <td>0.369919</td>\n",
       "      <td>0.366936</td>\n",
       "      <td>221</td>\n",
       "      <td>0.898374</td>\n",
       "      <td>230</td>\n",
       "      <td>0.920000</td>\n",
       "      <td>315</td>\n",
       "      <td>14963</td>\n",
       "      <td>...</td>\n",
       "      <td>1081.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>52.5</td>\n",
       "      <td>24.5</td>\n",
       "      <td>MVKADRVHKNFGALKVLKGISLEVGRGEVLCLVGPSGSGKSTFLRC...</td>\n",
       "      <td>MPPVLEARDIAKSFGPTPALRGLSLTIDAGEVVAITGPSGSGKSTL...</td>\n",
       "      <td>amino acid ABC transporter ATP-binding protein</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>250</td>\n",
       "      <td>246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39489</th>\n",
       "      <td>39489</td>\n",
       "      <td>0.412371</td>\n",
       "      <td>0.207254</td>\n",
       "      <td>0.172414</td>\n",
       "      <td>111</td>\n",
       "      <td>0.575129</td>\n",
       "      <td>102</td>\n",
       "      <td>0.376384</td>\n",
       "      <td>47</td>\n",
       "      <td>12872</td>\n",
       "      <td>...</td>\n",
       "      <td>1044.0</td>\n",
       "      <td>21.5</td>\n",
       "      <td>45.0</td>\n",
       "      <td>23.5</td>\n",
       "      <td>MIISLPAALGAAVLFGVASALQAYAARTEPPGSGLRGLLRVVLRLP...</td>\n",
       "      <td>MISLTVTAVVFGVVFLAELPDKTALAGLVLGTRYRASYVFAGVAAA...</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>TMEM165/GDT1 family protein</td>\n",
       "      <td>271</td>\n",
       "      <td>193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37891</th>\n",
       "      <td>37891</td>\n",
       "      <td>0.279070</td>\n",
       "      <td>0.255319</td>\n",
       "      <td>0.257143</td>\n",
       "      <td>258</td>\n",
       "      <td>0.914894</td>\n",
       "      <td>256</td>\n",
       "      <td>0.920863</td>\n",
       "      <td>251</td>\n",
       "      <td>14963</td>\n",
       "      <td>...</td>\n",
       "      <td>1179.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>52.5</td>\n",
       "      <td>21.5</td>\n",
       "      <td>MVKRLNFMGVIARVFMWSFLLGLAVIVLYPLLWMALNGFKTNAELF...</td>\n",
       "      <td>MATPQAGARRYGPGRYAVLTAYLVLAVMMLVPFAIVLLNAIKTPRE...</td>\n",
       "      <td>carbohydrate ABC transporter permease</td>\n",
       "      <td>carbohydrate ABC transporter permease</td>\n",
       "      <td>278</td>\n",
       "      <td>282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26178</th>\n",
       "      <td>26178</td>\n",
       "      <td>0.280000</td>\n",
       "      <td>0.230047</td>\n",
       "      <td>0.208511</td>\n",
       "      <td>171</td>\n",
       "      <td>0.802817</td>\n",
       "      <td>189</td>\n",
       "      <td>0.735409</td>\n",
       "      <td>115</td>\n",
       "      <td>6260</td>\n",
       "      <td>...</td>\n",
       "      <td>994.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>MSDNIRLSVKGVNKRFGGLQALSDVGLEIKSGQIYGLIGPNGAGKT...</td>\n",
       "      <td>MLEASALTCIKGDRILFRDLTHVFRPGVLTRIAGPNGVGKTSLLRL...</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>cytochrome c biogenesis heme-transporting ATPa...</td>\n",
       "      <td>257</td>\n",
       "      <td>213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45892</th>\n",
       "      <td>45892</td>\n",
       "      <td>0.368000</td>\n",
       "      <td>0.212963</td>\n",
       "      <td>0.216981</td>\n",
       "      <td>142</td>\n",
       "      <td>0.657407</td>\n",
       "      <td>131</td>\n",
       "      <td>0.629808</td>\n",
       "      <td>61</td>\n",
       "      <td>14963</td>\n",
       "      <td>...</td>\n",
       "      <td>1267.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>52.5</td>\n",
       "      <td>22.5</td>\n",
       "      <td>MSRADPPAPPALKPGRGRRPAEDVRRAVLAAAAELLLAEGMPGMTF...</td>\n",
       "      <td>MTSRAPSREPQQDRSRATRRRLLEAAIDCLASVGWAGTTVAVVAER...</td>\n",
       "      <td>TetR/AcrR family transcriptional regulator C-t...</td>\n",
       "      <td>TetR/AcrR family transcriptional regulator</td>\n",
       "      <td>208</td>\n",
       "      <td>216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20933</th>\n",
       "      <td>20933</td>\n",
       "      <td>0.268775</td>\n",
       "      <td>0.231293</td>\n",
       "      <td>0.235702</td>\n",
       "      <td>255</td>\n",
       "      <td>0.867347</td>\n",
       "      <td>251</td>\n",
       "      <td>0.886926</td>\n",
       "      <td>184</td>\n",
       "      <td>6048</td>\n",
       "      <td>...</td>\n",
       "      <td>1096.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>MVYSEKAAKWVLSFPLLLFWVVMVVFPAGYAIYLSLHDVSMGQGMS...</td>\n",
       "      <td>MTRIVRQAQYLVFLVPILVVYSLFTIYPLLKAFFLSFTNFDGYSKV...</td>\n",
       "      <td>sugar ABC transporter permease</td>\n",
       "      <td>sugar ABC transporter permease</td>\n",
       "      <td>283</td>\n",
       "      <td>294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40971</th>\n",
       "      <td>40971</td>\n",
       "      <td>0.271777</td>\n",
       "      <td>0.242991</td>\n",
       "      <td>0.240370</td>\n",
       "      <td>289</td>\n",
       "      <td>0.900312</td>\n",
       "      <td>296</td>\n",
       "      <td>0.902439</td>\n",
       "      <td>131</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>1108.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>MTKTVRFAVEGGPEVLELVDVELGDPGAGELRLRVEAIGLNRAEAM...</td>\n",
       "      <td>MFAVYATEPHPDDPLAGLAVGERPEPEAPEGWVTVDVKAASLNMHD...</td>\n",
       "      <td>zinc-dependent alcohol dehydrogenase family pr...</td>\n",
       "      <td>quinone oxidoreductase</td>\n",
       "      <td>328</td>\n",
       "      <td>321</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>200 rows Ã— 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0  local_gap_compressed_percent_id  \\\n",
       "33553       33553                         0.303922   \n",
       "9427         9427                         0.303371   \n",
       "199           199                         0.292453   \n",
       "12447       12447                         0.406250   \n",
       "39489       39489                         0.412371   \n",
       "...           ...                              ...   \n",
       "37891       37891                         0.279070   \n",
       "26178       26178                         0.280000   \n",
       "45892       45892                         0.368000   \n",
       "20933       20933                         0.268775   \n",
       "40971       40971                         0.271777   \n",
       "\n",
       "       scaled_local_query_percent_id  scaled_local_symmetric_percent_id  \\\n",
       "33553                       0.276786                           0.216783   \n",
       "9427                        0.173633                           0.190476   \n",
       "199                         0.185629                           0.184799   \n",
       "12447                       0.369919                           0.366936   \n",
       "39489                       0.207254                           0.172414   \n",
       "...                              ...                                ...   \n",
       "37891                       0.255319                           0.257143   \n",
       "26178                       0.230047                           0.208511   \n",
       "45892                       0.212963                           0.216981   \n",
       "20933                       0.231293                           0.235702   \n",
       "40971                       0.242991                           0.240370   \n",
       "\n",
       "       query_align_len  query_align_cov  subject_align_len  subject_align_cov  \\\n",
       "33553              199         0.888393                211           0.606322   \n",
       "9427               178         0.572347                196           0.765625   \n",
       "199                213         0.637725                215           0.637982   \n",
       "12447              221         0.898374                230           0.920000   \n",
       "39489              111         0.575129                102           0.376384   \n",
       "...                ...              ...                ...                ...   \n",
       "37891              258         0.914894                256           0.920863   \n",
       "26178              171         0.802817                189           0.735409   \n",
       "45892              142         0.657407                131           0.629808   \n",
       "20933              255         0.867347                251           0.886926   \n",
       "40971              289         0.900312                296           0.902439   \n",
       "\n",
       "       bit_score  thermo_index  ...  bit_score_16s  m_ogt  t_ogt  \\\n",
       "33553        221          9827  ...         1004.0   26.0   50.0   \n",
       "9427          70           875  ...         1035.0   30.0   50.0   \n",
       "199          140         11324  ...          989.0   32.0   54.0   \n",
       "12447        315         14963  ...         1081.0   28.0   52.5   \n",
       "39489         47         12872  ...         1044.0   21.5   45.0   \n",
       "...          ...           ...  ...            ...    ...    ...   \n",
       "37891        251         14963  ...         1179.0   31.0   52.5   \n",
       "26178        115          6260  ...          994.0   25.0   55.0   \n",
       "45892         61         14963  ...         1267.0   30.0   52.5   \n",
       "20933        184          6048  ...         1096.0   30.0   50.0   \n",
       "40971        131           875  ...         1108.0   25.0   50.0   \n",
       "\n",
       "       ogt_difference                                      m_protein_seq  \\\n",
       "33553            24.0  MTNCKVDNKGWINVSDGSQLLDIQHLHTGFRIKDEYYDAVDDVSFE...   \n",
       "9427             20.0  MRNVAALSMPPRVPVEEGRTFTVDEGKRLLDAVADHRMGVLVLLAL...   \n",
       "199              22.0  MFTLKHATAASAVVLGLVFATETFAADVTWNVSLWGKRRAFTEHVE...   \n",
       "12447            24.5  MVKADRVHKNFGALKVLKGISLEVGRGEVLCLVGPSGSGKSTFLRC...   \n",
       "39489            23.5  MIISLPAALGAAVLFGVASALQAYAARTEPPGSGLRGLLRVVLRLP...   \n",
       "...               ...                                                ...   \n",
       "37891            21.5  MVKRLNFMGVIARVFMWSFLLGLAVIVLYPLLWMALNGFKTNAELF...   \n",
       "26178            30.0  MSDNIRLSVKGVNKRFGGLQALSDVGLEIKSGQIYGLIGPNGAGKT...   \n",
       "45892            22.5  MSRADPPAPPALKPGRGRRPAEDVRRAVLAAAAELLLAEGMPGMTF...   \n",
       "20933            20.0  MVYSEKAAKWVLSFPLLLFWVVMVVFPAGYAIYLSLHDVSMGQGMS...   \n",
       "40971            25.0  MTKTVRFAVEGGPEVLELVDVELGDPGAGELRLRVEAIGLNRAEAM...   \n",
       "\n",
       "                                           t_protein_seq  \\\n",
       "33553  MDIKLHHVNKSYETGVETINVLKDIDLTIESGSWLTITGPSGSGKT...   \n",
       "9427   MSGTRPDRVIRAFVDHLVVERGMARNTVDSYARDLRRYHAHLQARG...   \n",
       "199    MKIAREFAGALAGTLAGFAMAAVTALPVFAEELKFANFTPPFHTIN...   \n",
       "12447  MPPVLEARDIAKSFGPTPALRGLSLTIDAGEVVAITGPSGSGKSTL...   \n",
       "39489  MISLTVTAVVFGVVFLAELPDKTALAGLVLGTRYRASYVFAGVAAA...   \n",
       "...                                                  ...   \n",
       "37891  MATPQAGARRYGPGRYAVLTAYLVLAVMMLVPFAIVLLNAIKTPRE...   \n",
       "26178  MLEASALTCIKGDRILFRDLTHVFRPGVLTRIAGPNGVGKTSLLRL...   \n",
       "45892  MTSRAPSREPQQDRSRATRRRLLEAAIDCLASVGWAGTTVAVVAER...   \n",
       "20933  MTRIVRQAQYLVFLVPILVVYSLFTIYPLLKAFFLSFTNFDGYSKV...   \n",
       "40971  MFAVYATEPHPDDPLAGLAVGERPEPEAPEGWVTVDVKAASLNMHD...   \n",
       "\n",
       "                                          m_protein_desc  \\\n",
       "33553                ABC transporter ATP-binding protein   \n",
       "9427                             site-specific integrase   \n",
       "199    C4-dicarboxylate TRAP transporter substrate-bi...   \n",
       "12447     amino acid ABC transporter ATP-binding protein   \n",
       "39489                               hypothetical protein   \n",
       "...                                                  ...   \n",
       "37891              carbohydrate ABC transporter permease   \n",
       "26178                ABC transporter ATP-binding protein   \n",
       "45892  TetR/AcrR family transcriptional regulator C-t...   \n",
       "20933                     sugar ABC transporter permease   \n",
       "40971  zinc-dependent alcohol dehydrogenase family pr...   \n",
       "\n",
       "                                          t_protein_desc  m_protein_len  \\\n",
       "33553                ABC transporter ATP-binding protein            348   \n",
       "9427             site-specific tyrosine recombinase XerD            256   \n",
       "199           TRAP transporter substrate-binding protein            337   \n",
       "12447                ABC transporter ATP-binding protein            250   \n",
       "39489                        TMEM165/GDT1 family protein            271   \n",
       "...                                                  ...            ...   \n",
       "37891              carbohydrate ABC transporter permease            278   \n",
       "26178  cytochrome c biogenesis heme-transporting ATPa...            257   \n",
       "45892         TetR/AcrR family transcriptional regulator            208   \n",
       "20933                     sugar ABC transporter permease            283   \n",
       "40971                             quinone oxidoreductase            328   \n",
       "\n",
       "       t_protein_len  \n",
       "33553            224  \n",
       "9427             311  \n",
       "199              334  \n",
       "12447            246  \n",
       "39489            193  \n",
       "...              ...  \n",
       "37891            282  \n",
       "26178            213  \n",
       "45892            216  \n",
       "20933            294  \n",
       "40971            321  \n",
       "\n",
       "[200 rows x 30 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a5477fd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_fasta_from_dataframe(\n",
    "        dataframe, output_file_a: str, output_file_b: str):\n",
    "    '''\n",
    "    Generates fasta file type from pandas dataframe.\n",
    "\n",
    "    Args:\n",
    "        Dataframe (pandas dataframe)\n",
    "        Names of output fasta files (str)\n",
    "\n",
    "    Returns:\n",
    "        Two fasta files with protein sequences and pair_id\n",
    "    '''\n",
    "    # meso sequence to fasta\n",
    "    with open(f'{output_file_a}', 'w') as f:\n",
    "        for _, row in dataframe.iterrows():\n",
    "            f.write(\n",
    "                '>{}\\n{}\\n'.format(\n",
    "                    (row['prot_pair_index']),\n",
    "                    row['m_protein_seq']))\n",
    "\n",
    "    # thermo sequence to fasta\n",
    "    with open(f'{output_file_b}', 'w') as f:\n",
    "        for _, row in dataframe.iterrows():\n",
    "            f.write(\n",
    "                '>{}\\n{}\\n'.format(\n",
    "                    (row['prot_pair_index']),\n",
    "                    (row['t_protein_seq'])))\n",
    "\n",
    "    # return output files\n",
    "    return [output_file_a, output_file_b]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3d0cc1c6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['seq_a.fasta', 'seq_b.fasta']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_fasta_from_dataframe(df, 'seq_a.fasta', 'seq_b.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ec4f8efb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2193416",
   "metadata": {},
   "source": [
    "Would be a good spot to figure out how to generate esm embeddings from a fasta file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7b0de76f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-13-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "#this block of code should load the files into a dataframe \n",
    "\n",
    "import os\n",
    "\n",
    "# Create an empty dataframe\n",
    "embeddings = pd.DataFrame(columns=['prot_pair_index', 'tensor'])\n",
    "\n",
    "# Set the directory path containing the PyTorch files\n",
    "directory = '/Users/loganroberts/Learn2Therm/ValidProt/data/pytorch_embeddings'\n",
    "\n",
    "# Iterate over the files in the directory\n",
    "for filename in os.listdir(directory):\n",
    "    if filename.endswith('.pt'):  # Adjust the file extension as needed\n",
    "        file_path = os.path.join(directory, filename)\n",
    "        \n",
    "        # Load the PyTorch file\n",
    "        data = torch.load(file_path)\n",
    "        \n",
    "        # Extract the tensor(s) from the loaded data\n",
    "        if isinstance(data, torch.Tensor):\n",
    "            tensors = [data]\n",
    "        elif isinstance(data, dict):\n",
    "            tensors = list(data.values())\n",
    "        else:\n",
    "            # Handle other cases based on your specific file structure\n",
    "            continue\n",
    "        \n",
    "        # Append the filename and tensor(s) to the dataframe\n",
    "        embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
    "\n",
    "#expand tensor values into individual columns\n",
    "embeddings = pd.concat([embeddings.drop('tensor', axis=1),\n",
    "                embeddings['tensor'].apply(lambda x: pd.Series(x))], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b6c050b3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prot_pair_index</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>...</th>\n",
       "      <th>1270</th>\n",
       "      <th>1271</th>\n",
       "      <th>1272</th>\n",
       "      <th>1273</th>\n",
       "      <th>1274</th>\n",
       "      <th>1275</th>\n",
       "      <th>1276</th>\n",
       "      <th>1277</th>\n",
       "      <th>1278</th>\n",
       "      <th>1279</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>111336591</td>\n",
       "      <td>0.019533</td>\n",
       "      <td>-0.024063</td>\n",
       "      <td>-0.066794</td>\n",
       "      <td>0.050212</td>\n",
       "      <td>-0.047821</td>\n",
       "      <td>-0.094027</td>\n",
       "      <td>0.005994</td>\n",
       "      <td>-0.042168</td>\n",
       "      <td>-0.076345</td>\n",
       "      <td>...</td>\n",
       "      <td>0.064117</td>\n",
       "      <td>-0.019009</td>\n",
       "      <td>-0.056704</td>\n",
       "      <td>-0.025658</td>\n",
       "      <td>-0.057087</td>\n",
       "      <td>0.068928</td>\n",
       "      <td>0.081193</td>\n",
       "      <td>-0.126052</td>\n",
       "      <td>0.009601</td>\n",
       "      <td>-0.012149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>94441498</td>\n",
       "      <td>-0.011820</td>\n",
       "      <td>-0.104690</td>\n",
       "      <td>0.038199</td>\n",
       "      <td>-0.035568</td>\n",
       "      <td>-0.113570</td>\n",
       "      <td>0.041435</td>\n",
       "      <td>-0.002200</td>\n",
       "      <td>-0.204471</td>\n",
       "      <td>-0.095335</td>\n",
       "      <td>...</td>\n",
       "      <td>0.065634</td>\n",
       "      <td>-0.101977</td>\n",
       "      <td>-0.032930</td>\n",
       "      <td>0.027585</td>\n",
       "      <td>-0.048157</td>\n",
       "      <td>-0.056852</td>\n",
       "      <td>0.132306</td>\n",
       "      <td>-0.249838</td>\n",
       "      <td>0.005303</td>\n",
       "      <td>0.080105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8507781</td>\n",
       "      <td>-0.013917</td>\n",
       "      <td>0.001229</td>\n",
       "      <td>0.015009</td>\n",
       "      <td>0.047818</td>\n",
       "      <td>-0.123596</td>\n",
       "      <td>-0.030345</td>\n",
       "      <td>0.005982</td>\n",
       "      <td>-0.083568</td>\n",
       "      <td>-0.016797</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.024941</td>\n",
       "      <td>-0.100372</td>\n",
       "      <td>0.004829</td>\n",
       "      <td>-0.036445</td>\n",
       "      <td>-0.053202</td>\n",
       "      <td>-0.054582</td>\n",
       "      <td>0.124111</td>\n",
       "      <td>-0.150853</td>\n",
       "      <td>-0.051989</td>\n",
       "      <td>0.058379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>136710423</td>\n",
       "      <td>-0.040277</td>\n",
       "      <td>-0.054379</td>\n",
       "      <td>0.014265</td>\n",
       "      <td>-0.067939</td>\n",
       "      <td>-0.015732</td>\n",
       "      <td>-0.016550</td>\n",
       "      <td>0.022490</td>\n",
       "      <td>-0.081806</td>\n",
       "      <td>-0.056216</td>\n",
       "      <td>...</td>\n",
       "      <td>0.038008</td>\n",
       "      <td>-0.067936</td>\n",
       "      <td>-0.081956</td>\n",
       "      <td>-0.026359</td>\n",
       "      <td>-0.060215</td>\n",
       "      <td>0.091697</td>\n",
       "      <td>0.028884</td>\n",
       "      <td>-0.120096</td>\n",
       "      <td>0.016685</td>\n",
       "      <td>0.030200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>49002962</td>\n",
       "      <td>-0.027329</td>\n",
       "      <td>-0.096013</td>\n",
       "      <td>-0.018252</td>\n",
       "      <td>-0.009544</td>\n",
       "      <td>-0.085416</td>\n",
       "      <td>-0.035731</td>\n",
       "      <td>0.041572</td>\n",
       "      <td>-0.023275</td>\n",
       "      <td>-0.097101</td>\n",
       "      <td>...</td>\n",
       "      <td>0.049347</td>\n",
       "      <td>-0.068024</td>\n",
       "      <td>-0.072538</td>\n",
       "      <td>-0.035347</td>\n",
       "      <td>-0.035973</td>\n",
       "      <td>0.053484</td>\n",
       "      <td>0.054946</td>\n",
       "      <td>-0.245820</td>\n",
       "      <td>0.033397</td>\n",
       "      <td>0.033206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>140227431</td>\n",
       "      <td>0.000347</td>\n",
       "      <td>-0.010710</td>\n",
       "      <td>0.057421</td>\n",
       "      <td>-0.007362</td>\n",
       "      <td>-0.086283</td>\n",
       "      <td>0.010301</td>\n",
       "      <td>0.083760</td>\n",
       "      <td>-0.098041</td>\n",
       "      <td>-0.090244</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000161</td>\n",
       "      <td>-0.054455</td>\n",
       "      <td>-0.006414</td>\n",
       "      <td>-0.062227</td>\n",
       "      <td>-0.094488</td>\n",
       "      <td>0.007952</td>\n",
       "      <td>0.105958</td>\n",
       "      <td>-0.153057</td>\n",
       "      <td>0.016046</td>\n",
       "      <td>0.086979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>95379032</td>\n",
       "      <td>-0.060202</td>\n",
       "      <td>-0.008571</td>\n",
       "      <td>0.051682</td>\n",
       "      <td>0.064621</td>\n",
       "      <td>-0.121588</td>\n",
       "      <td>-0.039067</td>\n",
       "      <td>-0.023636</td>\n",
       "      <td>-0.126189</td>\n",
       "      <td>-0.022895</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.037514</td>\n",
       "      <td>-0.084787</td>\n",
       "      <td>0.000140</td>\n",
       "      <td>-0.001242</td>\n",
       "      <td>-0.056960</td>\n",
       "      <td>-0.024581</td>\n",
       "      <td>0.117786</td>\n",
       "      <td>-0.138116</td>\n",
       "      <td>-0.025414</td>\n",
       "      <td>0.024033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>25255216</td>\n",
       "      <td>-0.032610</td>\n",
       "      <td>-0.030036</td>\n",
       "      <td>-0.021970</td>\n",
       "      <td>0.052059</td>\n",
       "      <td>-0.082433</td>\n",
       "      <td>0.042803</td>\n",
       "      <td>-0.027478</td>\n",
       "      <td>-0.097515</td>\n",
       "      <td>-0.039641</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.040955</td>\n",
       "      <td>-0.044872</td>\n",
       "      <td>-0.015927</td>\n",
       "      <td>-0.078347</td>\n",
       "      <td>-0.040029</td>\n",
       "      <td>-0.021940</td>\n",
       "      <td>0.038414</td>\n",
       "      <td>-0.149151</td>\n",
       "      <td>0.073578</td>\n",
       "      <td>0.089899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>114520866</td>\n",
       "      <td>-0.015026</td>\n",
       "      <td>-0.027358</td>\n",
       "      <td>0.058169</td>\n",
       "      <td>-0.028993</td>\n",
       "      <td>-0.017491</td>\n",
       "      <td>-0.085501</td>\n",
       "      <td>-0.015761</td>\n",
       "      <td>-0.051320</td>\n",
       "      <td>-0.056967</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.032083</td>\n",
       "      <td>-0.061828</td>\n",
       "      <td>-0.031256</td>\n",
       "      <td>-0.044584</td>\n",
       "      <td>-0.055094</td>\n",
       "      <td>0.062155</td>\n",
       "      <td>0.101345</td>\n",
       "      <td>-0.130985</td>\n",
       "      <td>0.013092</td>\n",
       "      <td>0.049958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>89307094</td>\n",
       "      <td>0.049971</td>\n",
       "      <td>-0.104723</td>\n",
       "      <td>-0.069643</td>\n",
       "      <td>0.099095</td>\n",
       "      <td>-0.090778</td>\n",
       "      <td>-0.098584</td>\n",
       "      <td>0.058866</td>\n",
       "      <td>-0.136397</td>\n",
       "      <td>-0.078372</td>\n",
       "      <td>...</td>\n",
       "      <td>0.058942</td>\n",
       "      <td>-0.003030</td>\n",
       "      <td>-0.083117</td>\n",
       "      <td>-0.072200</td>\n",
       "      <td>-0.075920</td>\n",
       "      <td>-0.073272</td>\n",
       "      <td>0.122866</td>\n",
       "      <td>-0.216860</td>\n",
       "      <td>0.012048</td>\n",
       "      <td>0.099718</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>200 rows Ã— 1281 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    prot_pair_index         0         1         2         3         4  \\\n",
       "0         111336591  0.019533 -0.024063 -0.066794  0.050212 -0.047821   \n",
       "1          94441498 -0.011820 -0.104690  0.038199 -0.035568 -0.113570   \n",
       "2           8507781 -0.013917  0.001229  0.015009  0.047818 -0.123596   \n",
       "3         136710423 -0.040277 -0.054379  0.014265 -0.067939 -0.015732   \n",
       "4          49002962 -0.027329 -0.096013 -0.018252 -0.009544 -0.085416   \n",
       "..              ...       ...       ...       ...       ...       ...   \n",
       "195       140227431  0.000347 -0.010710  0.057421 -0.007362 -0.086283   \n",
       "196        95379032 -0.060202 -0.008571  0.051682  0.064621 -0.121588   \n",
       "197        25255216 -0.032610 -0.030036 -0.021970  0.052059 -0.082433   \n",
       "198       114520866 -0.015026 -0.027358  0.058169 -0.028993 -0.017491   \n",
       "199        89307094  0.049971 -0.104723 -0.069643  0.099095 -0.090778   \n",
       "\n",
       "            5         6         7         8  ...      1270      1271  \\\n",
       "0   -0.094027  0.005994 -0.042168 -0.076345  ...  0.064117 -0.019009   \n",
       "1    0.041435 -0.002200 -0.204471 -0.095335  ...  0.065634 -0.101977   \n",
       "2   -0.030345  0.005982 -0.083568 -0.016797  ... -0.024941 -0.100372   \n",
       "3   -0.016550  0.022490 -0.081806 -0.056216  ...  0.038008 -0.067936   \n",
       "4   -0.035731  0.041572 -0.023275 -0.097101  ...  0.049347 -0.068024   \n",
       "..        ...       ...       ...       ...  ...       ...       ...   \n",
       "195  0.010301  0.083760 -0.098041 -0.090244  ...  0.000161 -0.054455   \n",
       "196 -0.039067 -0.023636 -0.126189 -0.022895  ... -0.037514 -0.084787   \n",
       "197  0.042803 -0.027478 -0.097515 -0.039641  ... -0.040955 -0.044872   \n",
       "198 -0.085501 -0.015761 -0.051320 -0.056967  ... -0.032083 -0.061828   \n",
       "199 -0.098584  0.058866 -0.136397 -0.078372  ...  0.058942 -0.003030   \n",
       "\n",
       "         1272      1273      1274      1275      1276      1277      1278  \\\n",
       "0   -0.056704 -0.025658 -0.057087  0.068928  0.081193 -0.126052  0.009601   \n",
       "1   -0.032930  0.027585 -0.048157 -0.056852  0.132306 -0.249838  0.005303   \n",
       "2    0.004829 -0.036445 -0.053202 -0.054582  0.124111 -0.150853 -0.051989   \n",
       "3   -0.081956 -0.026359 -0.060215  0.091697  0.028884 -0.120096  0.016685   \n",
       "4   -0.072538 -0.035347 -0.035973  0.053484  0.054946 -0.245820  0.033397   \n",
       "..        ...       ...       ...       ...       ...       ...       ...   \n",
       "195 -0.006414 -0.062227 -0.094488  0.007952  0.105958 -0.153057  0.016046   \n",
       "196  0.000140 -0.001242 -0.056960 -0.024581  0.117786 -0.138116 -0.025414   \n",
       "197 -0.015927 -0.078347 -0.040029 -0.021940  0.038414 -0.149151  0.073578   \n",
       "198 -0.031256 -0.044584 -0.055094  0.062155  0.101345 -0.130985  0.013092   \n",
       "199 -0.083117 -0.072200 -0.075920 -0.073272  0.122866 -0.216860  0.012048   \n",
       "\n",
       "         1279  \n",
       "0   -0.012149  \n",
       "1    0.080105  \n",
       "2    0.058379  \n",
       "3    0.030200  \n",
       "4    0.033206  \n",
       "..        ...  \n",
       "195  0.086979  \n",
       "196  0.024033  \n",
       "197  0.089899  \n",
       "198  0.049958  \n",
       "199  0.099718  \n",
       "\n",
       "[200 rows x 1281 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "embeddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b7fc1c27",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "new_df = pd.merge(embeddings, df, on='prot_pair_index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "34b67d18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prot_pair_index</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>...</th>\n",
       "      <th>bit_score_16s</th>\n",
       "      <th>m_ogt</th>\n",
       "      <th>t_ogt</th>\n",
       "      <th>ogt_difference</th>\n",
       "      <th>m_protein_seq</th>\n",
       "      <th>t_protein_seq</th>\n",
       "      <th>m_protein_desc</th>\n",
       "      <th>t_protein_desc</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>111336591</td>\n",
       "      <td>0.019533</td>\n",
       "      <td>-0.024063</td>\n",
       "      <td>-0.066794</td>\n",
       "      <td>0.050212</td>\n",
       "      <td>-0.047821</td>\n",
       "      <td>-0.094027</td>\n",
       "      <td>0.005994</td>\n",
       "      <td>-0.042168</td>\n",
       "      <td>-0.076345</td>\n",
       "      <td>...</td>\n",
       "      <td>980.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>57.5</td>\n",
       "      <td>20.5</td>\n",
       "      <td>MGLVTRQKRVAPSIYIRLKTPQKYLRKIEELSSRVIVEPWNWREPE...</td>\n",
       "      <td>MRKFSVVVTEKMLPSISDKLQERCHVREWDRTDPIPRELLLKWLAD...</td>\n",
       "      <td>D-2-hydroxyacid dehydrogenase</td>\n",
       "      <td>D-glycerate dehydrogenase</td>\n",
       "      <td>331</td>\n",
       "      <td>328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>94441498</td>\n",
       "      <td>-0.011820</td>\n",
       "      <td>-0.104690</td>\n",
       "      <td>0.038199</td>\n",
       "      <td>-0.035568</td>\n",
       "      <td>-0.113570</td>\n",
       "      <td>0.041435</td>\n",
       "      <td>-0.002200</td>\n",
       "      <td>-0.204471</td>\n",
       "      <td>-0.095335</td>\n",
       "      <td>...</td>\n",
       "      <td>1114.0</td>\n",
       "      <td>22.5</td>\n",
       "      <td>46.0</td>\n",
       "      <td>23.5</td>\n",
       "      <td>MTPSRPYIMRALYEWIVDNSCTPYILVDATMEDVSVPEQYVKDGQI...</td>\n",
       "      <td>MTSNRPYILRAFYEWITDNRCTPYILVDTHLPGVLVPQQHVNEDGQ...</td>\n",
       "      <td>ClpXP protease specificity-enhancing factor</td>\n",
       "      <td>ClpXP protease specificity-enhancing factor</td>\n",
       "      <td>133</td>\n",
       "      <td>131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8507781</td>\n",
       "      <td>-0.013917</td>\n",
       "      <td>0.001229</td>\n",
       "      <td>0.015009</td>\n",
       "      <td>0.047818</td>\n",
       "      <td>-0.123596</td>\n",
       "      <td>-0.030345</td>\n",
       "      <td>0.005982</td>\n",
       "      <td>-0.083568</td>\n",
       "      <td>-0.016797</td>\n",
       "      <td>...</td>\n",
       "      <td>1082.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>MTTTSETAAEAPRKTLETAVRMEAVNKWYGAFQALKNVDLEVRSGE...</td>\n",
       "      <td>MTLFEDQFAGRHAPSVPPKRSQGGAHAYFSAWNLHAYYGESYIVQG...</td>\n",
       "      <td>amino acid ABC transporter ATP-binding protein</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>258</td>\n",
       "      <td>259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>136710423</td>\n",
       "      <td>-0.040277</td>\n",
       "      <td>-0.054379</td>\n",
       "      <td>0.014265</td>\n",
       "      <td>-0.067939</td>\n",
       "      <td>-0.015732</td>\n",
       "      <td>-0.016550</td>\n",
       "      <td>0.022490</td>\n",
       "      <td>-0.081806</td>\n",
       "      <td>-0.056216</td>\n",
       "      <td>...</td>\n",
       "      <td>1077.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MTTTGSASPYDPPADPAASAAAAAVEQTPPRTDRRVVLVTGASSGI...</td>\n",
       "      <td>MTAEVPANDVEAPVADLSGHRALVTGAASGIGRACAEAFAAAGAEV...</td>\n",
       "      <td>SDR family NAD(P)-dependent oxidoreductase</td>\n",
       "      <td>3-hydroxybutyrate dehydrogenase</td>\n",
       "      <td>279</td>\n",
       "      <td>262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>49002962</td>\n",
       "      <td>-0.027329</td>\n",
       "      <td>-0.096013</td>\n",
       "      <td>-0.018252</td>\n",
       "      <td>-0.009544</td>\n",
       "      <td>-0.085416</td>\n",
       "      <td>-0.035731</td>\n",
       "      <td>0.041572</td>\n",
       "      <td>-0.023275</td>\n",
       "      <td>-0.097101</td>\n",
       "      <td>...</td>\n",
       "      <td>963.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>MSLTLRQVRYFVATAEIGQISQAAIHLNISQSAVTTAIKELEAMLG...</td>\n",
       "      <td>MADRRLQVFHTVARLLSFTKAAEALHMTQPAVTFQIRQLEEQFNTR...</td>\n",
       "      <td>LysR family transcriptional regulator</td>\n",
       "      <td>selenium metabolism-associated LysR family tra...</td>\n",
       "      <td>305</td>\n",
       "      <td>300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>140227431</td>\n",
       "      <td>0.000347</td>\n",
       "      <td>-0.010710</td>\n",
       "      <td>0.057421</td>\n",
       "      <td>-0.007362</td>\n",
       "      <td>-0.086283</td>\n",
       "      <td>0.010301</td>\n",
       "      <td>0.083760</td>\n",
       "      <td>-0.098041</td>\n",
       "      <td>-0.090244</td>\n",
       "      <td>...</td>\n",
       "      <td>945.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>MTAHAKITLKATLLSCAAAATLWSSMAHAQTTVRLGVGDPIGSSVG...</td>\n",
       "      <td>MLKTLTRSLMVATMLAGSAAAGWAADYTLRAVANSNENDEDYDGLV...</td>\n",
       "      <td>DctP family TRAP transporter solute-binding su...</td>\n",
       "      <td>TRAP transporter substrate-binding protein DctP</td>\n",
       "      <td>337</td>\n",
       "      <td>354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>95379032</td>\n",
       "      <td>-0.060202</td>\n",
       "      <td>-0.008571</td>\n",
       "      <td>0.051682</td>\n",
       "      <td>0.064621</td>\n",
       "      <td>-0.121588</td>\n",
       "      <td>-0.039067</td>\n",
       "      <td>-0.023636</td>\n",
       "      <td>-0.126189</td>\n",
       "      <td>-0.022895</td>\n",
       "      <td>...</td>\n",
       "      <td>958.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>52.5</td>\n",
       "      <td>22.5</td>\n",
       "      <td>MNAPIISVKDVRKSYGRGPNRFDALKGVSFDIFEGESVAIVGKSGS...</td>\n",
       "      <td>MDDRKQPSGAPEPVLSVRDLVVEYRTGQRPSDPVHRAVDGVSFDLY...</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>232</td>\n",
       "      <td>291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>25255216</td>\n",
       "      <td>-0.032610</td>\n",
       "      <td>-0.030036</td>\n",
       "      <td>-0.021970</td>\n",
       "      <td>0.052059</td>\n",
       "      <td>-0.082433</td>\n",
       "      <td>0.042803</td>\n",
       "      <td>-0.027478</td>\n",
       "      <td>-0.097515</td>\n",
       "      <td>-0.039641</td>\n",
       "      <td>...</td>\n",
       "      <td>1124.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>52.5</td>\n",
       "      <td>22.5</td>\n",
       "      <td>MTAAVEPAATDAPVGDALVSVEDLKIHFPIKRGLLVDRTVGHVYAV...</td>\n",
       "      <td>MTSTSTAPRTDRAAGTGEPVLSVRDLRVEYHTDAGAVRAVNGVSFD...</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>380</td>\n",
       "      <td>341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>114520866</td>\n",
       "      <td>-0.015026</td>\n",
       "      <td>-0.027358</td>\n",
       "      <td>0.058169</td>\n",
       "      <td>-0.028993</td>\n",
       "      <td>-0.017491</td>\n",
       "      <td>-0.085501</td>\n",
       "      <td>-0.015761</td>\n",
       "      <td>-0.051320</td>\n",
       "      <td>-0.056967</td>\n",
       "      <td>...</td>\n",
       "      <td>1154.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MSSRKSFVDVARPDSPVALVTGASGGLGQALAIELDALGCRVAVHY...</td>\n",
       "      <td>MSAPGAPLDGKVALITGAARGQGRSHALRLARDGAAIIAVDICAPI...</td>\n",
       "      <td>3-oxoacyl-ACP reductase FabG</td>\n",
       "      <td>mycofactocin-coupled SDR family oxidoreductase</td>\n",
       "      <td>253</td>\n",
       "      <td>283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>89307094</td>\n",
       "      <td>0.049971</td>\n",
       "      <td>-0.104723</td>\n",
       "      <td>-0.069643</td>\n",
       "      <td>0.099095</td>\n",
       "      <td>-0.090778</td>\n",
       "      <td>-0.098584</td>\n",
       "      <td>0.058866</td>\n",
       "      <td>-0.136397</td>\n",
       "      <td>-0.078372</td>\n",
       "      <td>...</td>\n",
       "      <td>1141.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>MKKIKVLTIFGTRPEAIKMAPVVKELNRQSDRFESLVAVTAQHREM...</td>\n",
       "      <td>MNKKLKIMTVFGTRPEAIKMAPLVKELEKYPDQIQSIVTVTAQHRE...</td>\n",
       "      <td>UDP-N-acetylglucosamine 2-epimerase (non-hydro...</td>\n",
       "      <td>UDP-N-acetylglucosamine 2-epimerase (non-hydro...</td>\n",
       "      <td>390</td>\n",
       "      <td>379</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>200 rows Ã— 1310 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    prot_pair_index         0         1         2         3         4  \\\n",
       "0         111336591  0.019533 -0.024063 -0.066794  0.050212 -0.047821   \n",
       "1          94441498 -0.011820 -0.104690  0.038199 -0.035568 -0.113570   \n",
       "2           8507781 -0.013917  0.001229  0.015009  0.047818 -0.123596   \n",
       "3         136710423 -0.040277 -0.054379  0.014265 -0.067939 -0.015732   \n",
       "4          49002962 -0.027329 -0.096013 -0.018252 -0.009544 -0.085416   \n",
       "..              ...       ...       ...       ...       ...       ...   \n",
       "195       140227431  0.000347 -0.010710  0.057421 -0.007362 -0.086283   \n",
       "196        95379032 -0.060202 -0.008571  0.051682  0.064621 -0.121588   \n",
       "197        25255216 -0.032610 -0.030036 -0.021970  0.052059 -0.082433   \n",
       "198       114520866 -0.015026 -0.027358  0.058169 -0.028993 -0.017491   \n",
       "199        89307094  0.049971 -0.104723 -0.069643  0.099095 -0.090778   \n",
       "\n",
       "            5         6         7         8  ...  bit_score_16s  m_ogt  t_ogt  \\\n",
       "0   -0.094027  0.005994 -0.042168 -0.076345  ...          980.0   37.0   57.5   \n",
       "1    0.041435 -0.002200 -0.204471 -0.095335  ...         1114.0   22.5   46.0   \n",
       "2   -0.030345  0.005982 -0.083568 -0.016797  ...         1082.0   24.0   45.0   \n",
       "3   -0.016550  0.022490 -0.081806 -0.056216  ...         1077.0   28.0   50.0   \n",
       "4   -0.035731  0.041572 -0.023275 -0.097101  ...          963.0   30.0   65.0   \n",
       "..        ...       ...       ...       ...  ...            ...    ...    ...   \n",
       "195  0.010301  0.083760 -0.098041 -0.090244  ...          945.0   29.0   54.0   \n",
       "196 -0.039067 -0.023636 -0.126189 -0.022895  ...          958.0   30.0   52.5   \n",
       "197  0.042803 -0.027478 -0.097515 -0.039641  ...         1124.0   30.0   52.5   \n",
       "198 -0.085501 -0.015761 -0.051320 -0.056967  ...         1154.0   28.0   50.0   \n",
       "199 -0.098584  0.058866 -0.136397 -0.078372  ...         1141.0   30.0   50.0   \n",
       "\n",
       "     ogt_difference                                      m_protein_seq  \\\n",
       "0              20.5  MGLVTRQKRVAPSIYIRLKTPQKYLRKIEELSSRVIVEPWNWREPE...   \n",
       "1              23.5  MTPSRPYIMRALYEWIVDNSCTPYILVDATMEDVSVPEQYVKDGQI...   \n",
       "2              21.0  MTTTSETAAEAPRKTLETAVRMEAVNKWYGAFQALKNVDLEVRSGE...   \n",
       "3              22.0  MTTTGSASPYDPPADPAASAAAAAVEQTPPRTDRRVVLVTGASSGI...   \n",
       "4              35.0  MSLTLRQVRYFVATAEIGQISQAAIHLNISQSAVTTAIKELEAMLG...   \n",
       "..              ...                                                ...   \n",
       "195            25.0  MTAHAKITLKATLLSCAAAATLWSSMAHAQTTVRLGVGDPIGSSVG...   \n",
       "196            22.5  MNAPIISVKDVRKSYGRGPNRFDALKGVSFDIFEGESVAIVGKSGS...   \n",
       "197            22.5  MTAAVEPAATDAPVGDALVSVEDLKIHFPIKRGLLVDRTVGHVYAV...   \n",
       "198            22.0  MSSRKSFVDVARPDSPVALVTGASGGLGQALAIELDALGCRVAVHY...   \n",
       "199            20.0  MKKIKVLTIFGTRPEAIKMAPVVKELNRQSDRFESLVAVTAQHREM...   \n",
       "\n",
       "                                         t_protein_seq  \\\n",
       "0    MRKFSVVVTEKMLPSISDKLQERCHVREWDRTDPIPRELLLKWLAD...   \n",
       "1    MTSNRPYILRAFYEWITDNRCTPYILVDTHLPGVLVPQQHVNEDGQ...   \n",
       "2    MTLFEDQFAGRHAPSVPPKRSQGGAHAYFSAWNLHAYYGESYIVQG...   \n",
       "3    MTAEVPANDVEAPVADLSGHRALVTGAASGIGRACAEAFAAAGAEV...   \n",
       "4    MADRRLQVFHTVARLLSFTKAAEALHMTQPAVTFQIRQLEEQFNTR...   \n",
       "..                                                 ...   \n",
       "195  MLKTLTRSLMVATMLAGSAAAGWAADYTLRAVANSNENDEDYDGLV...   \n",
       "196  MDDRKQPSGAPEPVLSVRDLVVEYRTGQRPSDPVHRAVDGVSFDLY...   \n",
       "197  MTSTSTAPRTDRAAGTGEPVLSVRDLRVEYHTDAGAVRAVNGVSFD...   \n",
       "198  MSAPGAPLDGKVALITGAARGQGRSHALRLARDGAAIIAVDICAPI...   \n",
       "199  MNKKLKIMTVFGTRPEAIKMAPLVKELEKYPDQIQSIVTVTAQHRE...   \n",
       "\n",
       "                                        m_protein_desc  \\\n",
       "0                        D-2-hydroxyacid dehydrogenase   \n",
       "1          ClpXP protease specificity-enhancing factor   \n",
       "2       amino acid ABC transporter ATP-binding protein   \n",
       "3           SDR family NAD(P)-dependent oxidoreductase   \n",
       "4                LysR family transcriptional regulator   \n",
       "..                                                 ...   \n",
       "195  DctP family TRAP transporter solute-binding su...   \n",
       "196                ABC transporter ATP-binding protein   \n",
       "197                ABC transporter ATP-binding protein   \n",
       "198                       3-oxoacyl-ACP reductase FabG   \n",
       "199  UDP-N-acetylglucosamine 2-epimerase (non-hydro...   \n",
       "\n",
       "                                        t_protein_desc  m_protein_len  \\\n",
       "0                            D-glycerate dehydrogenase            331   \n",
       "1          ClpXP protease specificity-enhancing factor            133   \n",
       "2                  ABC transporter ATP-binding protein            258   \n",
       "3                      3-hydroxybutyrate dehydrogenase            279   \n",
       "4    selenium metabolism-associated LysR family tra...            305   \n",
       "..                                                 ...            ...   \n",
       "195    TRAP transporter substrate-binding protein DctP            337   \n",
       "196                ABC transporter ATP-binding protein            232   \n",
       "197                ABC transporter ATP-binding protein            380   \n",
       "198     mycofactocin-coupled SDR family oxidoreductase            253   \n",
       "199  UDP-N-acetylglucosamine 2-epimerase (non-hydro...            390   \n",
       "\n",
       "     t_protein_len  \n",
       "0              328  \n",
       "1              131  \n",
       "2              259  \n",
       "3              262  \n",
       "4              300  \n",
       "..             ...  \n",
       "195            354  \n",
       "196            291  \n",
       "197            341  \n",
       "198            283  \n",
       "199            379  \n",
       "\n",
       "[200 rows x 1310 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6a60d1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "python pairpro/extract.py esm2_t33_650M_UR50D notebooks/seq_a.fasta data/pytorch_embeddings/ --include mean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d937efd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import subprocess\n",
    "\n",
    "# # Run the command line prompt and capture the output\n",
    "# output = subprocess.check_output(['python pairpro/extract.py esm2_t33_650M_UR50D notebooks/seq_a.fasta data/pytorch_embeddings/ --include mean'], universal_newlines=True)\n",
    "\n",
    "# # Process the captured output as needed\n",
    "# processed_data = output.strip()\n",
    "\n",
    "# # Use the processed data in your script\n",
    "# print(\"Processed data:\", processed_data)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ba75d30",
   "metadata": {},
   "source": [
    "Below code is for running on individual sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "184ced04",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# # Load ESM-2 model\n",
    "# model, alphabet = esm.pretrained.esm2_t33_650M_UR50D()\n",
    "# batch_converter = alphabet.get_batch_converter()\n",
    "# model.eval()  # disables dropout for deterministic results\n",
    "\n",
    "# # Prepare data (first 2 sequences from ESMStructuralSplitDataset superfamily / 4)\n",
    "# data = [(df['prot_pair_index'].iloc[0], df['m_protein_seq'].iloc[0])]\n",
    "\n",
    "# batch_labels, batch_strs, batch_tokens = batch_converter(data)\n",
    "# batch_lens = (batch_tokens != alphabet.padding_idx).sum(1)\n",
    "\n",
    "# # Extract per-residue representations (on CPU)\n",
    "# with torch.no_grad():\n",
    "#     results = model(batch_tokens, repr_layers=[33], return_contacts=True)\n",
    "# token_representations = results[\"representations\"][33]\n",
    "\n",
    "# # Generate per-sequence representations via averaging\n",
    "# # NOTE: token 0 is always a beginning-of-sequence token, so the first residue is token 1.\n",
    "# sequence_representations = []\n",
    "# for i, tokens_len in enumerate(batch_lens):\n",
    "#     sequence_representations.append(token_representations[i, 1 : tokens_len - 1].mean(0))\n",
    "\n",
    "# # Look at the unsupervised self-attention map contact predictions\n",
    "# # import matplotlib.pyplot as plt\n",
    "# # for (_, seq), tokens_len, attention_contacts in zip(data, batch_lens, results[\"contacts\"]):\n",
    "# #     plt.matshow(attention_contacts[: tokens_len, : tokens_len])\n",
    "# #     plt.title(seq)\n",
    "# #     plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "fa54115a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b883627d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(67, 1313)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hmmer_target = pd.read_csv('protein_match_50k')\n",
    "hmmer_target['protein_match'] = hmmer_target['protein_match'].map({'Yes': 1, 'No': 0})\n",
    "hmmer_target\n",
    "\n",
    "\n",
    "\n",
    "from sklearn.utils import resample\n",
    "\n",
    "# Assuming your data is in a pandas DataFrame called 'data'\n",
    "# Separate the majority and minority classes\n",
    "majority_class = hmmer_target[hmmer_target['protein_match'] == 1]\n",
    "minority_class = hmmer_target[hmmer_target['protein_match'] == 0]\n",
    "\n",
    "# Undersample the majority class to match the number of minority class samples\n",
    "n_samples = len(minority_class)\n",
    "undersampled_majority = resample(majority_class, n_samples=n_samples, replace=False)\n",
    "\n",
    "# Combine the undersampled majority class with the minority class\n",
    "balanced_data = pd.concat([undersampled_majority, minority_class])\n",
    "\n",
    "balanced_data\n",
    "\n",
    "df = pd.merge(df, balanced_data, on=['prot_pair_index'])\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "14015e54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1    35\n",
       "0    32\n",
       "Name: protein_match, dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['protein_match'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "06ad97a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['structure_match'] = df['protein_match']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "fa81d559",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prot_pair_index</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>...</th>\n",
       "      <th>m_protein_seq</th>\n",
       "      <th>t_protein_seq</th>\n",
       "      <th>m_protein_desc</th>\n",
       "      <th>t_protein_desc</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "      <th>Unnamed: 0_y</th>\n",
       "      <th>protein_match</th>\n",
       "      <th>Jaccard_Score</th>\n",
       "      <th>structure_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8507781</td>\n",
       "      <td>-0.013917</td>\n",
       "      <td>0.001229</td>\n",
       "      <td>0.015009</td>\n",
       "      <td>0.047818</td>\n",
       "      <td>-0.123596</td>\n",
       "      <td>-0.030345</td>\n",
       "      <td>0.005982</td>\n",
       "      <td>-0.083568</td>\n",
       "      <td>-0.016797</td>\n",
       "      <td>...</td>\n",
       "      <td>MTTTSETAAEAPRKTLETAVRMEAVNKWYGAFQALKNVDLEVRSGE...</td>\n",
       "      <td>MTLFEDQFAGRHAPSVPPKRSQGGAHAYFSAWNLHAYYGESYIVQG...</td>\n",
       "      <td>amino acid ABC transporter ATP-binding protein</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>258</td>\n",
       "      <td>259</td>\n",
       "      <td>1807</td>\n",
       "      <td>0</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9674279</td>\n",
       "      <td>-0.035063</td>\n",
       "      <td>-0.033278</td>\n",
       "      <td>-0.031553</td>\n",
       "      <td>0.050510</td>\n",
       "      <td>-0.125549</td>\n",
       "      <td>-0.023605</td>\n",
       "      <td>-0.023363</td>\n",
       "      <td>-0.054658</td>\n",
       "      <td>-0.000024</td>\n",
       "      <td>...</td>\n",
       "      <td>MVKKSIVLDDSDNYPRIIAKNLEVTYPLGSFSAVKRNRTGFKALDG...</td>\n",
       "      <td>MLDSSEVTVTADGRRIGGVLMDIRNITLRFGGVVAIRDTSFDIREG...</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>246</td>\n",
       "      <td>272</td>\n",
       "      <td>2196</td>\n",
       "      <td>1</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>47032839</td>\n",
       "      <td>-0.042536</td>\n",
       "      <td>-0.001069</td>\n",
       "      <td>0.010911</td>\n",
       "      <td>-0.078783</td>\n",
       "      <td>-0.110655</td>\n",
       "      <td>-0.029389</td>\n",
       "      <td>-0.020749</td>\n",
       "      <td>-0.072975</td>\n",
       "      <td>-0.040882</td>\n",
       "      <td>...</td>\n",
       "      <td>MTLRARLVVARPRFRLDIDLEVASGQVVALLGPNGAGKTTALRALA...</td>\n",
       "      <td>MDLEALTKRYPGQRVPAVENVTMHIPAGEIVVLVGPSGCGKTTTMK...</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>betaine/proline/choline family ABC transporter...</td>\n",
       "      <td>343</td>\n",
       "      <td>386</td>\n",
       "      <td>1982</td>\n",
       "      <td>0</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22228095</td>\n",
       "      <td>-0.048989</td>\n",
       "      <td>-0.084854</td>\n",
       "      <td>-0.017102</td>\n",
       "      <td>-0.018352</td>\n",
       "      <td>-0.079451</td>\n",
       "      <td>-0.022137</td>\n",
       "      <td>0.013828</td>\n",
       "      <td>-0.097690</td>\n",
       "      <td>-0.013558</td>\n",
       "      <td>...</td>\n",
       "      <td>MFYQLLKYIVLGPLLRLVFRPRVEGLEHVPEEGAAIIAGNHLSFAD...</td>\n",
       "      <td>MFYTFLARVVAPILRLLFRPQVEGLENVPAHGPLIIASNHLSFIDS...</td>\n",
       "      <td>1-acyl-sn-glycerol-3-phosphate acyltransferase</td>\n",
       "      <td>1-acyl-sn-glycerol-3-phosphate acyltransferase</td>\n",
       "      <td>242</td>\n",
       "      <td>226</td>\n",
       "      <td>1050</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8499524</td>\n",
       "      <td>-0.060101</td>\n",
       "      <td>-0.061766</td>\n",
       "      <td>-0.073530</td>\n",
       "      <td>0.034004</td>\n",
       "      <td>-0.124030</td>\n",
       "      <td>-0.007483</td>\n",
       "      <td>-0.046469</td>\n",
       "      <td>-0.068001</td>\n",
       "      <td>-0.033389</td>\n",
       "      <td>...</td>\n",
       "      <td>MSTQADEPRTEGDPQVPVQVGEVVDPRFRHDYLQVRDLRVVFDGFV...</td>\n",
       "      <td>MSDLTLQNVRKVYRTKGRPEVEAVRGMTLTVRSGELVGLLGPSGCG...</td>\n",
       "      <td>urea ABC transporter ATP-binding protein UrtD</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>286</td>\n",
       "      <td>356</td>\n",
       "      <td>2517</td>\n",
       "      <td>0</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>95670986</td>\n",
       "      <td>-0.026282</td>\n",
       "      <td>-0.020020</td>\n",
       "      <td>-0.034754</td>\n",
       "      <td>0.026878</td>\n",
       "      <td>-0.132900</td>\n",
       "      <td>-0.105046</td>\n",
       "      <td>0.012186</td>\n",
       "      <td>-0.045722</td>\n",
       "      <td>0.011634</td>\n",
       "      <td>...</td>\n",
       "      <td>MLKVNNVTKFFHKNNKSFCAVDSLDLHISPGEIYAFLGPNGAGKTT...</td>\n",
       "      <td>MIQFINVNKAFGDFQVLKDINLKIEAGEKVVILGPSGSGKSTLLRC...</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>amino acid ABC transporter ATP-binding protein</td>\n",
       "      <td>303</td>\n",
       "      <td>242</td>\n",
       "      <td>1033</td>\n",
       "      <td>0</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>141297786</td>\n",
       "      <td>-0.023940</td>\n",
       "      <td>-0.042270</td>\n",
       "      <td>-0.039667</td>\n",
       "      <td>-0.034018</td>\n",
       "      <td>-0.001502</td>\n",
       "      <td>-0.023285</td>\n",
       "      <td>-0.030766</td>\n",
       "      <td>-0.100879</td>\n",
       "      <td>-0.063545</td>\n",
       "      <td>...</td>\n",
       "      <td>MRAVVCPAYGPPEIIRIEERRSPAPGPGQVRVRVGAAAVNFPDVLV...</td>\n",
       "      <td>MPKGVQVVRPGGPEVMSYTDVEVRRPGEGEVLVDVSVAGVNFIDIY...</td>\n",
       "      <td>NADPH:quinone oxidoreductase family protein</td>\n",
       "      <td>quinone oxidoreductase</td>\n",
       "      <td>322</td>\n",
       "      <td>320</td>\n",
       "      <td>803</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>52135150</td>\n",
       "      <td>-0.011473</td>\n",
       "      <td>-0.009117</td>\n",
       "      <td>0.005419</td>\n",
       "      <td>0.016954</td>\n",
       "      <td>-0.113067</td>\n",
       "      <td>-0.036459</td>\n",
       "      <td>-0.026923</td>\n",
       "      <td>-0.072642</td>\n",
       "      <td>-0.044034</td>\n",
       "      <td>...</td>\n",
       "      <td>MSADNTPHGASITLRGLTKTYGSEKAVDAIDLDARPGEFLTLLGPS...</td>\n",
       "      <td>MARVSDAPEVVIDPTREPIVVVDDLHIIYRVYGSAGGGKGTAASAL...</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>362</td>\n",
       "      <td>265</td>\n",
       "      <td>1210</td>\n",
       "      <td>0</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>120473180</td>\n",
       "      <td>-0.066773</td>\n",
       "      <td>-0.072190</td>\n",
       "      <td>-0.065844</td>\n",
       "      <td>0.028404</td>\n",
       "      <td>-0.159087</td>\n",
       "      <td>-0.105056</td>\n",
       "      <td>0.067993</td>\n",
       "      <td>-0.108386</td>\n",
       "      <td>-0.046956</td>\n",
       "      <td>...</td>\n",
       "      <td>MNKKELNWRVSQSEANLSLRAFLREEKQLSKRLLSAIKFQGGEIYL...</td>\n",
       "      <td>MTPTERAYTLTLTVQEGQEGRLAAFLREQHHFSRRLLNKLKREGTV...</td>\n",
       "      <td>RluA family pseudouridine synthase</td>\n",
       "      <td>RluA family pseudouridine synthase</td>\n",
       "      <td>313</td>\n",
       "      <td>299</td>\n",
       "      <td>2597</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>95379032</td>\n",
       "      <td>-0.060202</td>\n",
       "      <td>-0.008571</td>\n",
       "      <td>0.051682</td>\n",
       "      <td>0.064621</td>\n",
       "      <td>-0.121588</td>\n",
       "      <td>-0.039067</td>\n",
       "      <td>-0.023636</td>\n",
       "      <td>-0.126189</td>\n",
       "      <td>-0.022895</td>\n",
       "      <td>...</td>\n",
       "      <td>MNAPIISVKDVRKSYGRGPNRFDALKGVSFDIFEGESVAIVGKSGS...</td>\n",
       "      <td>MDDRKQPSGAPEPVLSVRDLVVEYRTGQRPSDPVHRAVDGVSFDLY...</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>232</td>\n",
       "      <td>291</td>\n",
       "      <td>3449</td>\n",
       "      <td>1</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67 rows Ã— 1314 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   prot_pair_index         0         1         2         3         4  \\\n",
       "0          8507781 -0.013917  0.001229  0.015009  0.047818 -0.123596   \n",
       "1          9674279 -0.035063 -0.033278 -0.031553  0.050510 -0.125549   \n",
       "2         47032839 -0.042536 -0.001069  0.010911 -0.078783 -0.110655   \n",
       "3         22228095 -0.048989 -0.084854 -0.017102 -0.018352 -0.079451   \n",
       "4          8499524 -0.060101 -0.061766 -0.073530  0.034004 -0.124030   \n",
       "..             ...       ...       ...       ...       ...       ...   \n",
       "62        95670986 -0.026282 -0.020020 -0.034754  0.026878 -0.132900   \n",
       "63       141297786 -0.023940 -0.042270 -0.039667 -0.034018 -0.001502   \n",
       "64        52135150 -0.011473 -0.009117  0.005419  0.016954 -0.113067   \n",
       "65       120473180 -0.066773 -0.072190 -0.065844  0.028404 -0.159087   \n",
       "66        95379032 -0.060202 -0.008571  0.051682  0.064621 -0.121588   \n",
       "\n",
       "           5         6         7         8  ...  \\\n",
       "0  -0.030345  0.005982 -0.083568 -0.016797  ...   \n",
       "1  -0.023605 -0.023363 -0.054658 -0.000024  ...   \n",
       "2  -0.029389 -0.020749 -0.072975 -0.040882  ...   \n",
       "3  -0.022137  0.013828 -0.097690 -0.013558  ...   \n",
       "4  -0.007483 -0.046469 -0.068001 -0.033389  ...   \n",
       "..       ...       ...       ...       ...  ...   \n",
       "62 -0.105046  0.012186 -0.045722  0.011634  ...   \n",
       "63 -0.023285 -0.030766 -0.100879 -0.063545  ...   \n",
       "64 -0.036459 -0.026923 -0.072642 -0.044034  ...   \n",
       "65 -0.105056  0.067993 -0.108386 -0.046956  ...   \n",
       "66 -0.039067 -0.023636 -0.126189 -0.022895  ...   \n",
       "\n",
       "                                        m_protein_seq  \\\n",
       "0   MTTTSETAAEAPRKTLETAVRMEAVNKWYGAFQALKNVDLEVRSGE...   \n",
       "1   MVKKSIVLDDSDNYPRIIAKNLEVTYPLGSFSAVKRNRTGFKALDG...   \n",
       "2   MTLRARLVVARPRFRLDIDLEVASGQVVALLGPNGAGKTTALRALA...   \n",
       "3   MFYQLLKYIVLGPLLRLVFRPRVEGLEHVPEEGAAIIAGNHLSFAD...   \n",
       "4   MSTQADEPRTEGDPQVPVQVGEVVDPRFRHDYLQVRDLRVVFDGFV...   \n",
       "..                                                ...   \n",
       "62  MLKVNNVTKFFHKNNKSFCAVDSLDLHISPGEIYAFLGPNGAGKTT...   \n",
       "63  MRAVVCPAYGPPEIIRIEERRSPAPGPGQVRVRVGAAAVNFPDVLV...   \n",
       "64  MSADNTPHGASITLRGLTKTYGSEKAVDAIDLDARPGEFLTLLGPS...   \n",
       "65  MNKKELNWRVSQSEANLSLRAFLREEKQLSKRLLSAIKFQGGEIYL...   \n",
       "66  MNAPIISVKDVRKSYGRGPNRFDALKGVSFDIFEGESVAIVGKSGS...   \n",
       "\n",
       "                                        t_protein_seq  \\\n",
       "0   MTLFEDQFAGRHAPSVPPKRSQGGAHAYFSAWNLHAYYGESYIVQG...   \n",
       "1   MLDSSEVTVTADGRRIGGVLMDIRNITLRFGGVVAIRDTSFDIREG...   \n",
       "2   MDLEALTKRYPGQRVPAVENVTMHIPAGEIVVLVGPSGCGKTTTMK...   \n",
       "3   MFYTFLARVVAPILRLLFRPQVEGLENVPAHGPLIIASNHLSFIDS...   \n",
       "4   MSDLTLQNVRKVYRTKGRPEVEAVRGMTLTVRSGELVGLLGPSGCG...   \n",
       "..                                                ...   \n",
       "62  MIQFINVNKAFGDFQVLKDINLKIEAGEKVVILGPSGSGKSTLLRC...   \n",
       "63  MPKGVQVVRPGGPEVMSYTDVEVRRPGEGEVLVDVSVAGVNFIDIY...   \n",
       "64  MARVSDAPEVVIDPTREPIVVVDDLHIIYRVYGSAGGGKGTAASAL...   \n",
       "65  MTPTERAYTLTLTVQEGQEGRLAAFLREQHHFSRRLLNKLKREGTV...   \n",
       "66  MDDRKQPSGAPEPVLSVRDLVVEYRTGQRPSDPVHRAVDGVSFDLY...   \n",
       "\n",
       "                                    m_protein_desc  \\\n",
       "0   amino acid ABC transporter ATP-binding protein   \n",
       "1              ABC transporter ATP-binding protein   \n",
       "2              ABC transporter ATP-binding protein   \n",
       "3   1-acyl-sn-glycerol-3-phosphate acyltransferase   \n",
       "4    urea ABC transporter ATP-binding protein UrtD   \n",
       "..                                             ...   \n",
       "62             ABC transporter ATP-binding protein   \n",
       "63     NADPH:quinone oxidoreductase family protein   \n",
       "64             ABC transporter ATP-binding protein   \n",
       "65              RluA family pseudouridine synthase   \n",
       "66             ABC transporter ATP-binding protein   \n",
       "\n",
       "                                       t_protein_desc  m_protein_len  \\\n",
       "0                 ABC transporter ATP-binding protein            258   \n",
       "1                 ABC transporter ATP-binding protein            246   \n",
       "2   betaine/proline/choline family ABC transporter...            343   \n",
       "3      1-acyl-sn-glycerol-3-phosphate acyltransferase            242   \n",
       "4                 ABC transporter ATP-binding protein            286   \n",
       "..                                                ...            ...   \n",
       "62     amino acid ABC transporter ATP-binding protein            303   \n",
       "63                             quinone oxidoreductase            322   \n",
       "64                ABC transporter ATP-binding protein            362   \n",
       "65                 RluA family pseudouridine synthase            313   \n",
       "66                ABC transporter ATP-binding protein            232   \n",
       "\n",
       "    t_protein_len  Unnamed: 0_y  protein_match  Jaccard_Score  structure_match  \n",
       "0             259          1807              0       0.222222                0  \n",
       "1             272          2196              1       0.500000                1  \n",
       "2             386          1982              0       0.222222                0  \n",
       "3             226          1050              1       1.000000                1  \n",
       "4             356          2517              0       0.250000                0  \n",
       "..            ...           ...            ...            ...              ...  \n",
       "62            242          1033              0       0.200000                0  \n",
       "63            320           803              1       1.000000                1  \n",
       "64            265          1210              0       0.166667                0  \n",
       "65            299          2597              1       1.000000                1  \n",
       "66            291          3449              1       0.333333                1  \n",
       "\n",
       "[67 rows x 1314 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "337f87aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>m_protein_seq</th>\n",
       "      <th>t_protein_seq</th>\n",
       "      <th>prot_pair_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MTTTSETAAEAPRKTLETAVRMEAVNKWYGAFQALKNVDLEVRSGE...</td>\n",
       "      <td>MTLFEDQFAGRHAPSVPPKRSQGGAHAYFSAWNLHAYYGESYIVQG...</td>\n",
       "      <td>8507781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MVKKSIVLDDSDNYPRIIAKNLEVTYPLGSFSAVKRNRTGFKALDG...</td>\n",
       "      <td>MLDSSEVTVTADGRRIGGVLMDIRNITLRFGGVVAIRDTSFDIREG...</td>\n",
       "      <td>9674279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MTLRARLVVARPRFRLDIDLEVASGQVVALLGPNGAGKTTALRALA...</td>\n",
       "      <td>MDLEALTKRYPGQRVPAVENVTMHIPAGEIVVLVGPSGCGKTTTMK...</td>\n",
       "      <td>47032839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MFYQLLKYIVLGPLLRLVFRPRVEGLEHVPEEGAAIIAGNHLSFAD...</td>\n",
       "      <td>MFYTFLARVVAPILRLLFRPQVEGLENVPAHGPLIIASNHLSFIDS...</td>\n",
       "      <td>22228095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MSTQADEPRTEGDPQVPVQVGEVVDPRFRHDYLQVRDLRVVFDGFV...</td>\n",
       "      <td>MSDLTLQNVRKVYRTKGRPEVEAVRGMTLTVRSGELVGLLGPSGCG...</td>\n",
       "      <td>8499524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>MLKVNNVTKFFHKNNKSFCAVDSLDLHISPGEIYAFLGPNGAGKTT...</td>\n",
       "      <td>MIQFINVNKAFGDFQVLKDINLKIEAGEKVVILGPSGSGKSTLLRC...</td>\n",
       "      <td>95670986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>MRAVVCPAYGPPEIIRIEERRSPAPGPGQVRVRVGAAAVNFPDVLV...</td>\n",
       "      <td>MPKGVQVVRPGGPEVMSYTDVEVRRPGEGEVLVDVSVAGVNFIDIY...</td>\n",
       "      <td>141297786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>MSADNTPHGASITLRGLTKTYGSEKAVDAIDLDARPGEFLTLLGPS...</td>\n",
       "      <td>MARVSDAPEVVIDPTREPIVVVDDLHIIYRVYGSAGGGKGTAASAL...</td>\n",
       "      <td>52135150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>MNKKELNWRVSQSEANLSLRAFLREEKQLSKRLLSAIKFQGGEIYL...</td>\n",
       "      <td>MTPTERAYTLTLTVQEGQEGRLAAFLREQHHFSRRLLNKLKREGTV...</td>\n",
       "      <td>120473180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>MNAPIISVKDVRKSYGRGPNRFDALKGVSFDIFEGESVAIVGKSGS...</td>\n",
       "      <td>MDDRKQPSGAPEPVLSVRDLVVEYRTGQRPSDPVHRAVDGVSFDLY...</td>\n",
       "      <td>95379032</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        m_protein_seq  \\\n",
       "0   MTTTSETAAEAPRKTLETAVRMEAVNKWYGAFQALKNVDLEVRSGE...   \n",
       "1   MVKKSIVLDDSDNYPRIIAKNLEVTYPLGSFSAVKRNRTGFKALDG...   \n",
       "2   MTLRARLVVARPRFRLDIDLEVASGQVVALLGPNGAGKTTALRALA...   \n",
       "3   MFYQLLKYIVLGPLLRLVFRPRVEGLEHVPEEGAAIIAGNHLSFAD...   \n",
       "4   MSTQADEPRTEGDPQVPVQVGEVVDPRFRHDYLQVRDLRVVFDGFV...   \n",
       "..                                                ...   \n",
       "62  MLKVNNVTKFFHKNNKSFCAVDSLDLHISPGEIYAFLGPNGAGKTT...   \n",
       "63  MRAVVCPAYGPPEIIRIEERRSPAPGPGQVRVRVGAAAVNFPDVLV...   \n",
       "64  MSADNTPHGASITLRGLTKTYGSEKAVDAIDLDARPGEFLTLLGPS...   \n",
       "65  MNKKELNWRVSQSEANLSLRAFLREEKQLSKRLLSAIKFQGGEIYL...   \n",
       "66  MNAPIISVKDVRKSYGRGPNRFDALKGVSFDIFEGESVAIVGKSGS...   \n",
       "\n",
       "                                        t_protein_seq prot_pair_index  \n",
       "0   MTLFEDQFAGRHAPSVPPKRSQGGAHAYFSAWNLHAYYGESYIVQG...         8507781  \n",
       "1   MLDSSEVTVTADGRRIGGVLMDIRNITLRFGGVVAIRDTSFDIREG...         9674279  \n",
       "2   MDLEALTKRYPGQRVPAVENVTMHIPAGEIVVLVGPSGCGKTTTMK...        47032839  \n",
       "3   MFYTFLARVVAPILRLLFRPQVEGLENVPAHGPLIIASNHLSFIDS...        22228095  \n",
       "4   MSDLTLQNVRKVYRTKGRPEVEAVRGMTLTVRSGELVGLLGPSGCG...         8499524  \n",
       "..                                                ...             ...  \n",
       "62  MIQFINVNKAFGDFQVLKDINLKIEAGEKVVILGPSGSGKSTLLRC...        95670986  \n",
       "63  MPKGVQVVRPGGPEVMSYTDVEVRRPGEGEVLVDVSVAGVNFIDIY...       141297786  \n",
       "64  MARVSDAPEVVIDPTREPIVVVDDLHIIYRVYGSAGGGKGTAASAL...        52135150  \n",
       "65  MTPTERAYTLTLTVQEGQEGRLAAFLREQHHFSRRLLNKLKREGTV...       120473180  \n",
       "66  MDDRKQPSGAPEPVLSVRDLVVEYRTGQRPSDPVHRAVDGVSFDLY...        95379032  \n",
       "\n",
       "[67 rows x 3 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_seqs = df[['m_protein_seq', 't_protein_seq', 'prot_pair_index']]\n",
    "df_seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "084d9561",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get rid of stuff that isn't quantitative\n",
    "\n",
    "df = df.drop(columns = ['Unnamed: 0_x', 'Unnamed: 0_y','thermo_index', 'm_protein_seq', 't_protein_seq',\n",
    "                        'm_protein_desc', 't_protein_desc'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "eea27e6c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index([      'prot_pair_index',                       0,\n",
       "                             1,                       2,\n",
       "                             3,                       4,\n",
       "                             5,                       6,\n",
       "                             7,                       8,\n",
       "       ...\n",
       "       'subject_align_cov_16s',         'bit_score_16s',\n",
       "                       'm_ogt',                 't_ogt',\n",
       "              'ogt_difference',         'm_protein_len',\n",
       "               't_protein_len',         'protein_match',\n",
       "               'Jaccard_Score',       'structure_match'],\n",
       "      dtype='object', length=1307)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07cc3a80",
   "metadata": {},
   "source": [
    "From pairplot below, it looks like query_align_cov_16s and subject_align_cov_16s dont correlate with target. Let's remove them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "36d8dadb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# sns.pairplot(df.sample(500),height = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "91079edd",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.drop(columns = ['query_align_cov_16s', 'subject_align_cov_16s'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "7929625e",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>scaled_local_symmetric_percent_id_16s</th>\n",
       "      <th>bit_score_16s</th>\n",
       "      <th>m_ogt</th>\n",
       "      <th>t_ogt</th>\n",
       "      <th>ogt_difference</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "      <th>protein_match</th>\n",
       "      <th>Jaccard_Score</th>\n",
       "      <th>structure_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>67.000000</td>\n",
       "      <td>67.000000</td>\n",
       "      <td>67.000000</td>\n",
       "      <td>67.000000</td>\n",
       "      <td>67.000000</td>\n",
       "      <td>67.000000</td>\n",
       "      <td>67.000000</td>\n",
       "      <td>67.000000</td>\n",
       "      <td>67.000000</td>\n",
       "      <td>67.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>67.000000</td>\n",
       "      <td>67.000000</td>\n",
       "      <td>67.000000</td>\n",
       "      <td>67.00000</td>\n",
       "      <td>67.000000</td>\n",
       "      <td>67.000000</td>\n",
       "      <td>67.000000</td>\n",
       "      <td>67.000000</td>\n",
       "      <td>67.000000</td>\n",
       "      <td>67.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>-0.027277</td>\n",
       "      <td>-0.035986</td>\n",
       "      <td>-0.015397</td>\n",
       "      <td>0.019778</td>\n",
       "      <td>-0.075090</td>\n",
       "      <td>-0.019970</td>\n",
       "      <td>0.006330</td>\n",
       "      <td>-0.081295</td>\n",
       "      <td>-0.020644</td>\n",
       "      <td>0.023334</td>\n",
       "      <td>...</td>\n",
       "      <td>0.903224</td>\n",
       "      <td>1056.477612</td>\n",
       "      <td>27.738806</td>\n",
       "      <td>52.00000</td>\n",
       "      <td>24.261194</td>\n",
       "      <td>275.432836</td>\n",
       "      <td>262.373134</td>\n",
       "      <td>0.522388</td>\n",
       "      <td>0.483173</td>\n",
       "      <td>0.522388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.037587</td>\n",
       "      <td>0.034874</td>\n",
       "      <td>0.038562</td>\n",
       "      <td>0.045897</td>\n",
       "      <td>0.056407</td>\n",
       "      <td>0.038783</td>\n",
       "      <td>0.040957</td>\n",
       "      <td>0.042846</td>\n",
       "      <td>0.044200</td>\n",
       "      <td>0.030281</td>\n",
       "      <td>...</td>\n",
       "      <td>0.014996</td>\n",
       "      <td>73.429075</td>\n",
       "      <td>3.478678</td>\n",
       "      <td>3.97816</td>\n",
       "      <td>3.225559</td>\n",
       "      <td>53.030930</td>\n",
       "      <td>56.150187</td>\n",
       "      <td>0.503268</td>\n",
       "      <td>0.344267</td>\n",
       "      <td>0.503268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>-0.138532</td>\n",
       "      <td>-0.146711</td>\n",
       "      <td>-0.113655</td>\n",
       "      <td>-0.106057</td>\n",
       "      <td>-0.171560</td>\n",
       "      <td>-0.105056</td>\n",
       "      <td>-0.076026</td>\n",
       "      <td>-0.185853</td>\n",
       "      <td>-0.125826</td>\n",
       "      <td>-0.030235</td>\n",
       "      <td>...</td>\n",
       "      <td>0.880180</td>\n",
       "      <td>919.000000</td>\n",
       "      <td>14.000000</td>\n",
       "      <td>45.00000</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>134.000000</td>\n",
       "      <td>129.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>-0.048794</td>\n",
       "      <td>-0.061848</td>\n",
       "      <td>-0.036372</td>\n",
       "      <td>-0.009608</td>\n",
       "      <td>-0.122009</td>\n",
       "      <td>-0.040004</td>\n",
       "      <td>-0.020599</td>\n",
       "      <td>-0.107815</td>\n",
       "      <td>-0.044900</td>\n",
       "      <td>0.006553</td>\n",
       "      <td>...</td>\n",
       "      <td>0.890774</td>\n",
       "      <td>1003.500000</td>\n",
       "      <td>28.000000</td>\n",
       "      <td>50.00000</td>\n",
       "      <td>22.000000</td>\n",
       "      <td>242.000000</td>\n",
       "      <td>232.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>-0.026015</td>\n",
       "      <td>-0.032280</td>\n",
       "      <td>-0.015854</td>\n",
       "      <td>0.018101</td>\n",
       "      <td>-0.087591</td>\n",
       "      <td>-0.023647</td>\n",
       "      <td>0.002887</td>\n",
       "      <td>-0.079946</td>\n",
       "      <td>-0.021781</td>\n",
       "      <td>0.019506</td>\n",
       "      <td>...</td>\n",
       "      <td>0.902423</td>\n",
       "      <td>1054.000000</td>\n",
       "      <td>28.000000</td>\n",
       "      <td>52.50000</td>\n",
       "      <td>24.500000</td>\n",
       "      <td>269.000000</td>\n",
       "      <td>259.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>-0.009166</td>\n",
       "      <td>-0.009521</td>\n",
       "      <td>0.011781</td>\n",
       "      <td>0.049164</td>\n",
       "      <td>-0.026148</td>\n",
       "      <td>-0.001903</td>\n",
       "      <td>0.031278</td>\n",
       "      <td>-0.055150</td>\n",
       "      <td>0.006157</td>\n",
       "      <td>0.031526</td>\n",
       "      <td>...</td>\n",
       "      <td>0.911330</td>\n",
       "      <td>1097.000000</td>\n",
       "      <td>30.000000</td>\n",
       "      <td>52.50000</td>\n",
       "      <td>24.500000</td>\n",
       "      <td>321.000000</td>\n",
       "      <td>300.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.875000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>0.076832</td>\n",
       "      <td>0.049612</td>\n",
       "      <td>0.073470</td>\n",
       "      <td>0.117160</td>\n",
       "      <td>0.088323</td>\n",
       "      <td>0.068677</td>\n",
       "      <td>0.096358</td>\n",
       "      <td>0.016453</td>\n",
       "      <td>0.097132</td>\n",
       "      <td>0.125401</td>\n",
       "      <td>...</td>\n",
       "      <td>0.942958</td>\n",
       "      <td>1266.000000</td>\n",
       "      <td>37.000000</td>\n",
       "      <td>60.00000</td>\n",
       "      <td>34.000000</td>\n",
       "      <td>389.000000</td>\n",
       "      <td>396.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows Ã— 1304 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0          1          2          3          4          5  \\\n",
       "count  67.000000  67.000000  67.000000  67.000000  67.000000  67.000000   \n",
       "mean   -0.027277  -0.035986  -0.015397   0.019778  -0.075090  -0.019970   \n",
       "std     0.037587   0.034874   0.038562   0.045897   0.056407   0.038783   \n",
       "min    -0.138532  -0.146711  -0.113655  -0.106057  -0.171560  -0.105056   \n",
       "25%    -0.048794  -0.061848  -0.036372  -0.009608  -0.122009  -0.040004   \n",
       "50%    -0.026015  -0.032280  -0.015854   0.018101  -0.087591  -0.023647   \n",
       "75%    -0.009166  -0.009521   0.011781   0.049164  -0.026148  -0.001903   \n",
       "max     0.076832   0.049612   0.073470   0.117160   0.088323   0.068677   \n",
       "\n",
       "               6          7          8          9  ...  \\\n",
       "count  67.000000  67.000000  67.000000  67.000000  ...   \n",
       "mean    0.006330  -0.081295  -0.020644   0.023334  ...   \n",
       "std     0.040957   0.042846   0.044200   0.030281  ...   \n",
       "min    -0.076026  -0.185853  -0.125826  -0.030235  ...   \n",
       "25%    -0.020599  -0.107815  -0.044900   0.006553  ...   \n",
       "50%     0.002887  -0.079946  -0.021781   0.019506  ...   \n",
       "75%     0.031278  -0.055150   0.006157   0.031526  ...   \n",
       "max     0.096358   0.016453   0.097132   0.125401  ...   \n",
       "\n",
       "       scaled_local_symmetric_percent_id_16s  bit_score_16s      m_ogt  \\\n",
       "count                              67.000000      67.000000  67.000000   \n",
       "mean                                0.903224    1056.477612  27.738806   \n",
       "std                                 0.014996      73.429075   3.478678   \n",
       "min                                 0.880180     919.000000  14.000000   \n",
       "25%                                 0.890774    1003.500000  28.000000   \n",
       "50%                                 0.902423    1054.000000  28.000000   \n",
       "75%                                 0.911330    1097.000000  30.000000   \n",
       "max                                 0.942958    1266.000000  37.000000   \n",
       "\n",
       "          t_ogt  ogt_difference  m_protein_len  t_protein_len  protein_match  \\\n",
       "count  67.00000       67.000000      67.000000      67.000000      67.000000   \n",
       "mean   52.00000       24.261194     275.432836     262.373134       0.522388   \n",
       "std     3.97816        3.225559      53.030930      56.150187       0.503268   \n",
       "min    45.00000       20.000000     134.000000     129.000000       0.000000   \n",
       "25%    50.00000       22.000000     242.000000     232.000000       0.000000   \n",
       "50%    52.50000       24.500000     269.000000     259.000000       1.000000   \n",
       "75%    52.50000       24.500000     321.000000     300.000000       1.000000   \n",
       "max    60.00000       34.000000     389.000000     396.000000       1.000000   \n",
       "\n",
       "       Jaccard_Score  structure_match  \n",
       "count      67.000000        67.000000  \n",
       "mean        0.483173         0.522388  \n",
       "std         0.344267         0.503268  \n",
       "min         0.000000         0.000000  \n",
       "25%         0.200000         0.000000  \n",
       "50%         0.333333         1.000000  \n",
       "75%         0.875000         1.000000  \n",
       "max         1.000000         1.000000  \n",
       "\n",
       "[8 rows x 1304 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90871d01",
   "metadata": {},
   "source": [
    "Split data into dev and test, and then split that into train and validation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "6519dcaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#drop columns that don't exihibit signficant pearson correlation with bit_score\n",
    "\n",
    "df = df.drop(columns = ['meso_index', 'meso_protein_int_index', 'local_gap_compressed_percent_id_16s', \n",
    "                        'scaled_local_query_percent_id_16s', 'scaled_local_symmetric_percent_id_16s',\n",
    "                       'bit_score_16s', 'm_ogt', 't_ogt', 'taxa_pair_index', 'thermo_protein_int_index'\n",
    "                       , 'prot_pair_index', 'ogt_difference'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "b3b9a7c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.drop(columns = ['Jaccard_Score'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "113e6608",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index([                                  0,\n",
       "                                         1,\n",
       "                                         2,\n",
       "                                         3,\n",
       "                                         4,\n",
       "                                         5,\n",
       "                                         6,\n",
       "                                         7,\n",
       "                                         8,\n",
       "                                         9,\n",
       "       ...\n",
       "       'scaled_local_symmetric_percent_id',\n",
       "                         'query_align_len',\n",
       "                         'query_align_cov',\n",
       "                       'subject_align_len',\n",
       "                       'subject_align_cov',\n",
       "                               'bit_score',\n",
       "                           'm_protein_len',\n",
       "                           't_protein_len',\n",
       "                           'protein_match',\n",
       "                         'structure_match'],\n",
       "      dtype='object', length=1292)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "af9f23cf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(56, 1292)\n",
      "(11, 1292)\n",
      "(47, 1292)\n",
      "(9, 1292)\n"
     ]
    }
   ],
   "source": [
    "#choosing 80/20 split instead of 85/15 because of volume of data\n",
    "\n",
    "dev, test = sklearn.model_selection.train_test_split(df, test_size=0.15, random_state=1)\n",
    "\n",
    "train, val = sklearn.model_selection.train_test_split(dev, test_size=0.15, random_state=1)\n",
    "\n",
    "print(dev.shape)\n",
    "print(test.shape)\n",
    "print(train.shape)\n",
    "print(val.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "8470550d",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#ID target and features, separate into separate arrays\n",
    "\n",
    "target = [columns for columns in df[['protein_match', 'structure_match']]]\n",
    "input_features = [columns for columns in df.drop(columns=['protein_match', 'structure_match'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "73c0a63f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0,\n",
       " 1,\n",
       " 2,\n",
       " 3,\n",
       " 4,\n",
       " 5,\n",
       " 6,\n",
       " 7,\n",
       " 8,\n",
       " 9,\n",
       " 10,\n",
       " 11,\n",
       " 12,\n",
       " 13,\n",
       " 14,\n",
       " 15,\n",
       " 16,\n",
       " 17,\n",
       " 18,\n",
       " 19,\n",
       " 20,\n",
       " 21,\n",
       " 22,\n",
       " 23,\n",
       " 24,\n",
       " 25,\n",
       " 26,\n",
       " 27,\n",
       " 28,\n",
       " 29,\n",
       " 30,\n",
       " 31,\n",
       " 32,\n",
       " 33,\n",
       " 34,\n",
       " 35,\n",
       " 36,\n",
       " 37,\n",
       " 38,\n",
       " 39,\n",
       " 40,\n",
       " 41,\n",
       " 42,\n",
       " 43,\n",
       " 44,\n",
       " 45,\n",
       " 46,\n",
       " 47,\n",
       " 48,\n",
       " 49,\n",
       " 50,\n",
       " 51,\n",
       " 52,\n",
       " 53,\n",
       " 54,\n",
       " 55,\n",
       " 56,\n",
       " 57,\n",
       " 58,\n",
       " 59,\n",
       " 60,\n",
       " 61,\n",
       " 62,\n",
       " 63,\n",
       " 64,\n",
       " 65,\n",
       " 66,\n",
       " 67,\n",
       " 68,\n",
       " 69,\n",
       " 70,\n",
       " 71,\n",
       " 72,\n",
       " 73,\n",
       " 74,\n",
       " 75,\n",
       " 76,\n",
       " 77,\n",
       " 78,\n",
       " 79,\n",
       " 80,\n",
       " 81,\n",
       " 82,\n",
       " 83,\n",
       " 84,\n",
       " 85,\n",
       " 86,\n",
       " 87,\n",
       " 88,\n",
       " 89,\n",
       " 90,\n",
       " 91,\n",
       " 92,\n",
       " 93,\n",
       " 94,\n",
       " 95,\n",
       " 96,\n",
       " 97,\n",
       " 98,\n",
       " 99,\n",
       " 100,\n",
       " 101,\n",
       " 102,\n",
       " 103,\n",
       " 104,\n",
       " 105,\n",
       " 106,\n",
       " 107,\n",
       " 108,\n",
       " 109,\n",
       " 110,\n",
       " 111,\n",
       " 112,\n",
       " 113,\n",
       " 114,\n",
       " 115,\n",
       " 116,\n",
       " 117,\n",
       " 118,\n",
       " 119,\n",
       " 120,\n",
       " 121,\n",
       " 122,\n",
       " 123,\n",
       " 124,\n",
       " 125,\n",
       " 126,\n",
       " 127,\n",
       " 128,\n",
       " 129,\n",
       " 130,\n",
       " 131,\n",
       " 132,\n",
       " 133,\n",
       " 134,\n",
       " 135,\n",
       " 136,\n",
       " 137,\n",
       " 138,\n",
       " 139,\n",
       " 140,\n",
       " 141,\n",
       " 142,\n",
       " 143,\n",
       " 144,\n",
       " 145,\n",
       " 146,\n",
       " 147,\n",
       " 148,\n",
       " 149,\n",
       " 150,\n",
       " 151,\n",
       " 152,\n",
       " 153,\n",
       " 154,\n",
       " 155,\n",
       " 156,\n",
       " 157,\n",
       " 158,\n",
       " 159,\n",
       " 160,\n",
       " 161,\n",
       " 162,\n",
       " 163,\n",
       " 164,\n",
       " 165,\n",
       " 166,\n",
       " 167,\n",
       " 168,\n",
       " 169,\n",
       " 170,\n",
       " 171,\n",
       " 172,\n",
       " 173,\n",
       " 174,\n",
       " 175,\n",
       " 176,\n",
       " 177,\n",
       " 178,\n",
       " 179,\n",
       " 180,\n",
       " 181,\n",
       " 182,\n",
       " 183,\n",
       " 184,\n",
       " 185,\n",
       " 186,\n",
       " 187,\n",
       " 188,\n",
       " 189,\n",
       " 190,\n",
       " 191,\n",
       " 192,\n",
       " 193,\n",
       " 194,\n",
       " 195,\n",
       " 196,\n",
       " 197,\n",
       " 198,\n",
       " 199,\n",
       " 200,\n",
       " 201,\n",
       " 202,\n",
       " 203,\n",
       " 204,\n",
       " 205,\n",
       " 206,\n",
       " 207,\n",
       " 208,\n",
       " 209,\n",
       " 210,\n",
       " 211,\n",
       " 212,\n",
       " 213,\n",
       " 214,\n",
       " 215,\n",
       " 216,\n",
       " 217,\n",
       " 218,\n",
       " 219,\n",
       " 220,\n",
       " 221,\n",
       " 222,\n",
       " 223,\n",
       " 224,\n",
       " 225,\n",
       " 226,\n",
       " 227,\n",
       " 228,\n",
       " 229,\n",
       " 230,\n",
       " 231,\n",
       " 232,\n",
       " 233,\n",
       " 234,\n",
       " 235,\n",
       " 236,\n",
       " 237,\n",
       " 238,\n",
       " 239,\n",
       " 240,\n",
       " 241,\n",
       " 242,\n",
       " 243,\n",
       " 244,\n",
       " 245,\n",
       " 246,\n",
       " 247,\n",
       " 248,\n",
       " 249,\n",
       " 250,\n",
       " 251,\n",
       " 252,\n",
       " 253,\n",
       " 254,\n",
       " 255,\n",
       " 256,\n",
       " 257,\n",
       " 258,\n",
       " 259,\n",
       " 260,\n",
       " 261,\n",
       " 262,\n",
       " 263,\n",
       " 264,\n",
       " 265,\n",
       " 266,\n",
       " 267,\n",
       " 268,\n",
       " 269,\n",
       " 270,\n",
       " 271,\n",
       " 272,\n",
       " 273,\n",
       " 274,\n",
       " 275,\n",
       " 276,\n",
       " 277,\n",
       " 278,\n",
       " 279,\n",
       " 280,\n",
       " 281,\n",
       " 282,\n",
       " 283,\n",
       " 284,\n",
       " 285,\n",
       " 286,\n",
       " 287,\n",
       " 288,\n",
       " 289,\n",
       " 290,\n",
       " 291,\n",
       " 292,\n",
       " 293,\n",
       " 294,\n",
       " 295,\n",
       " 296,\n",
       " 297,\n",
       " 298,\n",
       " 299,\n",
       " 300,\n",
       " 301,\n",
       " 302,\n",
       " 303,\n",
       " 304,\n",
       " 305,\n",
       " 306,\n",
       " 307,\n",
       " 308,\n",
       " 309,\n",
       " 310,\n",
       " 311,\n",
       " 312,\n",
       " 313,\n",
       " 314,\n",
       " 315,\n",
       " 316,\n",
       " 317,\n",
       " 318,\n",
       " 319,\n",
       " 320,\n",
       " 321,\n",
       " 322,\n",
       " 323,\n",
       " 324,\n",
       " 325,\n",
       " 326,\n",
       " 327,\n",
       " 328,\n",
       " 329,\n",
       " 330,\n",
       " 331,\n",
       " 332,\n",
       " 333,\n",
       " 334,\n",
       " 335,\n",
       " 336,\n",
       " 337,\n",
       " 338,\n",
       " 339,\n",
       " 340,\n",
       " 341,\n",
       " 342,\n",
       " 343,\n",
       " 344,\n",
       " 345,\n",
       " 346,\n",
       " 347,\n",
       " 348,\n",
       " 349,\n",
       " 350,\n",
       " 351,\n",
       " 352,\n",
       " 353,\n",
       " 354,\n",
       " 355,\n",
       " 356,\n",
       " 357,\n",
       " 358,\n",
       " 359,\n",
       " 360,\n",
       " 361,\n",
       " 362,\n",
       " 363,\n",
       " 364,\n",
       " 365,\n",
       " 366,\n",
       " 367,\n",
       " 368,\n",
       " 369,\n",
       " 370,\n",
       " 371,\n",
       " 372,\n",
       " 373,\n",
       " 374,\n",
       " 375,\n",
       " 376,\n",
       " 377,\n",
       " 378,\n",
       " 379,\n",
       " 380,\n",
       " 381,\n",
       " 382,\n",
       " 383,\n",
       " 384,\n",
       " 385,\n",
       " 386,\n",
       " 387,\n",
       " 388,\n",
       " 389,\n",
       " 390,\n",
       " 391,\n",
       " 392,\n",
       " 393,\n",
       " 394,\n",
       " 395,\n",
       " 396,\n",
       " 397,\n",
       " 398,\n",
       " 399,\n",
       " 400,\n",
       " 401,\n",
       " 402,\n",
       " 403,\n",
       " 404,\n",
       " 405,\n",
       " 406,\n",
       " 407,\n",
       " 408,\n",
       " 409,\n",
       " 410,\n",
       " 411,\n",
       " 412,\n",
       " 413,\n",
       " 414,\n",
       " 415,\n",
       " 416,\n",
       " 417,\n",
       " 418,\n",
       " 419,\n",
       " 420,\n",
       " 421,\n",
       " 422,\n",
       " 423,\n",
       " 424,\n",
       " 425,\n",
       " 426,\n",
       " 427,\n",
       " 428,\n",
       " 429,\n",
       " 430,\n",
       " 431,\n",
       " 432,\n",
       " 433,\n",
       " 434,\n",
       " 435,\n",
       " 436,\n",
       " 437,\n",
       " 438,\n",
       " 439,\n",
       " 440,\n",
       " 441,\n",
       " 442,\n",
       " 443,\n",
       " 444,\n",
       " 445,\n",
       " 446,\n",
       " 447,\n",
       " 448,\n",
       " 449,\n",
       " 450,\n",
       " 451,\n",
       " 452,\n",
       " 453,\n",
       " 454,\n",
       " 455,\n",
       " 456,\n",
       " 457,\n",
       " 458,\n",
       " 459,\n",
       " 460,\n",
       " 461,\n",
       " 462,\n",
       " 463,\n",
       " 464,\n",
       " 465,\n",
       " 466,\n",
       " 467,\n",
       " 468,\n",
       " 469,\n",
       " 470,\n",
       " 471,\n",
       " 472,\n",
       " 473,\n",
       " 474,\n",
       " 475,\n",
       " 476,\n",
       " 477,\n",
       " 478,\n",
       " 479,\n",
       " 480,\n",
       " 481,\n",
       " 482,\n",
       " 483,\n",
       " 484,\n",
       " 485,\n",
       " 486,\n",
       " 487,\n",
       " 488,\n",
       " 489,\n",
       " 490,\n",
       " 491,\n",
       " 492,\n",
       " 493,\n",
       " 494,\n",
       " 495,\n",
       " 496,\n",
       " 497,\n",
       " 498,\n",
       " 499,\n",
       " 500,\n",
       " 501,\n",
       " 502,\n",
       " 503,\n",
       " 504,\n",
       " 505,\n",
       " 506,\n",
       " 507,\n",
       " 508,\n",
       " 509,\n",
       " 510,\n",
       " 511,\n",
       " 512,\n",
       " 513,\n",
       " 514,\n",
       " 515,\n",
       " 516,\n",
       " 517,\n",
       " 518,\n",
       " 519,\n",
       " 520,\n",
       " 521,\n",
       " 522,\n",
       " 523,\n",
       " 524,\n",
       " 525,\n",
       " 526,\n",
       " 527,\n",
       " 528,\n",
       " 529,\n",
       " 530,\n",
       " 531,\n",
       " 532,\n",
       " 533,\n",
       " 534,\n",
       " 535,\n",
       " 536,\n",
       " 537,\n",
       " 538,\n",
       " 539,\n",
       " 540,\n",
       " 541,\n",
       " 542,\n",
       " 543,\n",
       " 544,\n",
       " 545,\n",
       " 546,\n",
       " 547,\n",
       " 548,\n",
       " 549,\n",
       " 550,\n",
       " 551,\n",
       " 552,\n",
       " 553,\n",
       " 554,\n",
       " 555,\n",
       " 556,\n",
       " 557,\n",
       " 558,\n",
       " 559,\n",
       " 560,\n",
       " 561,\n",
       " 562,\n",
       " 563,\n",
       " 564,\n",
       " 565,\n",
       " 566,\n",
       " 567,\n",
       " 568,\n",
       " 569,\n",
       " 570,\n",
       " 571,\n",
       " 572,\n",
       " 573,\n",
       " 574,\n",
       " 575,\n",
       " 576,\n",
       " 577,\n",
       " 578,\n",
       " 579,\n",
       " 580,\n",
       " 581,\n",
       " 582,\n",
       " 583,\n",
       " 584,\n",
       " 585,\n",
       " 586,\n",
       " 587,\n",
       " 588,\n",
       " 589,\n",
       " 590,\n",
       " 591,\n",
       " 592,\n",
       " 593,\n",
       " 594,\n",
       " 595,\n",
       " 596,\n",
       " 597,\n",
       " 598,\n",
       " 599,\n",
       " 600,\n",
       " 601,\n",
       " 602,\n",
       " 603,\n",
       " 604,\n",
       " 605,\n",
       " 606,\n",
       " 607,\n",
       " 608,\n",
       " 609,\n",
       " 610,\n",
       " 611,\n",
       " 612,\n",
       " 613,\n",
       " 614,\n",
       " 615,\n",
       " 616,\n",
       " 617,\n",
       " 618,\n",
       " 619,\n",
       " 620,\n",
       " 621,\n",
       " 622,\n",
       " 623,\n",
       " 624,\n",
       " 625,\n",
       " 626,\n",
       " 627,\n",
       " 628,\n",
       " 629,\n",
       " 630,\n",
       " 631,\n",
       " 632,\n",
       " 633,\n",
       " 634,\n",
       " 635,\n",
       " 636,\n",
       " 637,\n",
       " 638,\n",
       " 639,\n",
       " 640,\n",
       " 641,\n",
       " 642,\n",
       " 643,\n",
       " 644,\n",
       " 645,\n",
       " 646,\n",
       " 647,\n",
       " 648,\n",
       " 649,\n",
       " 650,\n",
       " 651,\n",
       " 652,\n",
       " 653,\n",
       " 654,\n",
       " 655,\n",
       " 656,\n",
       " 657,\n",
       " 658,\n",
       " 659,\n",
       " 660,\n",
       " 661,\n",
       " 662,\n",
       " 663,\n",
       " 664,\n",
       " 665,\n",
       " 666,\n",
       " 667,\n",
       " 668,\n",
       " 669,\n",
       " 670,\n",
       " 671,\n",
       " 672,\n",
       " 673,\n",
       " 674,\n",
       " 675,\n",
       " 676,\n",
       " 677,\n",
       " 678,\n",
       " 679,\n",
       " 680,\n",
       " 681,\n",
       " 682,\n",
       " 683,\n",
       " 684,\n",
       " 685,\n",
       " 686,\n",
       " 687,\n",
       " 688,\n",
       " 689,\n",
       " 690,\n",
       " 691,\n",
       " 692,\n",
       " 693,\n",
       " 694,\n",
       " 695,\n",
       " 696,\n",
       " 697,\n",
       " 698,\n",
       " 699,\n",
       " 700,\n",
       " 701,\n",
       " 702,\n",
       " 703,\n",
       " 704,\n",
       " 705,\n",
       " 706,\n",
       " 707,\n",
       " 708,\n",
       " 709,\n",
       " 710,\n",
       " 711,\n",
       " 712,\n",
       " 713,\n",
       " 714,\n",
       " 715,\n",
       " 716,\n",
       " 717,\n",
       " 718,\n",
       " 719,\n",
       " 720,\n",
       " 721,\n",
       " 722,\n",
       " 723,\n",
       " 724,\n",
       " 725,\n",
       " 726,\n",
       " 727,\n",
       " 728,\n",
       " 729,\n",
       " 730,\n",
       " 731,\n",
       " 732,\n",
       " 733,\n",
       " 734,\n",
       " 735,\n",
       " 736,\n",
       " 737,\n",
       " 738,\n",
       " 739,\n",
       " 740,\n",
       " 741,\n",
       " 742,\n",
       " 743,\n",
       " 744,\n",
       " 745,\n",
       " 746,\n",
       " 747,\n",
       " 748,\n",
       " 749,\n",
       " 750,\n",
       " 751,\n",
       " 752,\n",
       " 753,\n",
       " 754,\n",
       " 755,\n",
       " 756,\n",
       " 757,\n",
       " 758,\n",
       " 759,\n",
       " 760,\n",
       " 761,\n",
       " 762,\n",
       " 763,\n",
       " 764,\n",
       " 765,\n",
       " 766,\n",
       " 767,\n",
       " 768,\n",
       " 769,\n",
       " 770,\n",
       " 771,\n",
       " 772,\n",
       " 773,\n",
       " 774,\n",
       " 775,\n",
       " 776,\n",
       " 777,\n",
       " 778,\n",
       " 779,\n",
       " 780,\n",
       " 781,\n",
       " 782,\n",
       " 783,\n",
       " 784,\n",
       " 785,\n",
       " 786,\n",
       " 787,\n",
       " 788,\n",
       " 789,\n",
       " 790,\n",
       " 791,\n",
       " 792,\n",
       " 793,\n",
       " 794,\n",
       " 795,\n",
       " 796,\n",
       " 797,\n",
       " 798,\n",
       " 799,\n",
       " 800,\n",
       " 801,\n",
       " 802,\n",
       " 803,\n",
       " 804,\n",
       " 805,\n",
       " 806,\n",
       " 807,\n",
       " 808,\n",
       " 809,\n",
       " 810,\n",
       " 811,\n",
       " 812,\n",
       " 813,\n",
       " 814,\n",
       " 815,\n",
       " 816,\n",
       " 817,\n",
       " 818,\n",
       " 819,\n",
       " 820,\n",
       " 821,\n",
       " 822,\n",
       " 823,\n",
       " 824,\n",
       " 825,\n",
       " 826,\n",
       " 827,\n",
       " 828,\n",
       " 829,\n",
       " 830,\n",
       " 831,\n",
       " 832,\n",
       " 833,\n",
       " 834,\n",
       " 835,\n",
       " 836,\n",
       " 837,\n",
       " 838,\n",
       " 839,\n",
       " 840,\n",
       " 841,\n",
       " 842,\n",
       " 843,\n",
       " 844,\n",
       " 845,\n",
       " 846,\n",
       " 847,\n",
       " 848,\n",
       " 849,\n",
       " 850,\n",
       " 851,\n",
       " 852,\n",
       " 853,\n",
       " 854,\n",
       " 855,\n",
       " 856,\n",
       " 857,\n",
       " 858,\n",
       " 859,\n",
       " 860,\n",
       " 861,\n",
       " 862,\n",
       " 863,\n",
       " 864,\n",
       " 865,\n",
       " 866,\n",
       " 867,\n",
       " 868,\n",
       " 869,\n",
       " 870,\n",
       " 871,\n",
       " 872,\n",
       " 873,\n",
       " 874,\n",
       " 875,\n",
       " 876,\n",
       " 877,\n",
       " 878,\n",
       " 879,\n",
       " 880,\n",
       " 881,\n",
       " 882,\n",
       " 883,\n",
       " 884,\n",
       " 885,\n",
       " 886,\n",
       " 887,\n",
       " 888,\n",
       " 889,\n",
       " 890,\n",
       " 891,\n",
       " 892,\n",
       " 893,\n",
       " 894,\n",
       " 895,\n",
       " 896,\n",
       " 897,\n",
       " 898,\n",
       " 899,\n",
       " 900,\n",
       " 901,\n",
       " 902,\n",
       " 903,\n",
       " 904,\n",
       " 905,\n",
       " 906,\n",
       " 907,\n",
       " 908,\n",
       " 909,\n",
       " 910,\n",
       " 911,\n",
       " 912,\n",
       " 913,\n",
       " 914,\n",
       " 915,\n",
       " 916,\n",
       " 917,\n",
       " 918,\n",
       " 919,\n",
       " 920,\n",
       " 921,\n",
       " 922,\n",
       " 923,\n",
       " 924,\n",
       " 925,\n",
       " 926,\n",
       " 927,\n",
       " 928,\n",
       " 929,\n",
       " 930,\n",
       " 931,\n",
       " 932,\n",
       " 933,\n",
       " 934,\n",
       " 935,\n",
       " 936,\n",
       " 937,\n",
       " 938,\n",
       " 939,\n",
       " 940,\n",
       " 941,\n",
       " 942,\n",
       " 943,\n",
       " 944,\n",
       " 945,\n",
       " 946,\n",
       " 947,\n",
       " 948,\n",
       " 949,\n",
       " 950,\n",
       " 951,\n",
       " 952,\n",
       " 953,\n",
       " 954,\n",
       " 955,\n",
       " 956,\n",
       " 957,\n",
       " 958,\n",
       " 959,\n",
       " 960,\n",
       " 961,\n",
       " 962,\n",
       " 963,\n",
       " 964,\n",
       " 965,\n",
       " 966,\n",
       " 967,\n",
       " 968,\n",
       " 969,\n",
       " 970,\n",
       " 971,\n",
       " 972,\n",
       " 973,\n",
       " 974,\n",
       " 975,\n",
       " 976,\n",
       " 977,\n",
       " 978,\n",
       " 979,\n",
       " 980,\n",
       " 981,\n",
       " 982,\n",
       " 983,\n",
       " 984,\n",
       " 985,\n",
       " 986,\n",
       " 987,\n",
       " 988,\n",
       " 989,\n",
       " 990,\n",
       " 991,\n",
       " 992,\n",
       " 993,\n",
       " 994,\n",
       " 995,\n",
       " 996,\n",
       " 997,\n",
       " 998,\n",
       " 999,\n",
       " ...]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "1ab62bff",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(56, 1290) (11, 1290) (56, 2) (11, 2)\n"
     ]
    }
   ],
   "source": [
    "#split X and y\n",
    "\n",
    "dev_X = dev[input_features].values\n",
    "test_X = test[input_features].values\n",
    "\n",
    "dev_y = dev[target].values\n",
    "test_y = test[target].values\n",
    "\n",
    "print(dev_X.shape, test_X.shape, dev_y.shape, test_y.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "9ac52ac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#same thing for training and validation data\n",
    "\n",
    "train_X = train[input_features].values\n",
    "val_X = val[input_features].values\n",
    "\n",
    "train_y = train[target].values\n",
    "val_y = val[target].values"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c374f2d7",
   "metadata": {},
   "source": [
    "Scale the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "2b48b2e5",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "scaler = sklearn.preprocessing.StandardScaler()\n",
    "dev_X = scaler.fit_transform(dev_X)\n",
    "test_X = scaler.fit_transform(test_X)\n",
    "train_X = scaler.fit_transform(train_X)\n",
    "val_X = scaler.fit_transform(val_X)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3022371",
   "metadata": {},
   "source": [
    "Train the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "be4e09b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-1 {color: black;background-color: white;}#sk-container-id-1 pre{padding: 0;}#sk-container-id-1 div.sk-toggleable {background-color: white;}#sk-container-id-1 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-1 label.sk-toggleable__label-arrow:before {content: \"â–¸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-1 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-1 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-1 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"â–¾\";}#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-1 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-1 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-1 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-1 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-1 div.sk-item {position: relative;z-index: 1;}#sk-container-id-1 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-1 div.sk-item::before, #sk-container-id-1 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-1 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-1 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-1 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-1 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-1 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-1 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-1 div.sk-label-container {text-align: center;}#sk-container-id-1 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-1 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>RandomForestClassifier()</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" checked><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">RandomForestClassifier</label><div class=\"sk-toggleable__content\"><pre>RandomForestClassifier()</pre></div></div></div></div></div>"
      ],
      "text/plain": [
       "RandomForestClassifier()"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Random Forest\n",
    "\n",
    "model = sklearn.ensemble.RandomForestClassifier()\n",
    "model.fit(dev_X, dev_y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "c96904eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Obtain predicted class probabilities\n",
    "y_pred_prob = model.predict_proba(dev_X)\n",
    "\n",
    "# Compute uncertainty estimates using out-of-bag samples\n",
    "oob_predictions = model.oob_decision_function_\n",
    "\n",
    "uncertainty = np.mean(np.var(oob_predictions, axis=1), axis=0)\n",
    "\n",
    "# 'y_pred_prob' contains predicted class probabilities, and 'uncertainty' represents the uncertainty estimates\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "b64e62db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0.09550174, 0.09550174])"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uncertainty"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "b87faf14",
   "metadata": {},
   "outputs": [],
   "source": [
    "import joblib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "5ad183bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['trained_model.pkl']"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "joblib.dump(model, 'trained_model.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "3e12d31d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-2 {color: black;background-color: white;}#sk-container-id-2 pre{padding: 0;}#sk-container-id-2 div.sk-toggleable {background-color: white;}#sk-container-id-2 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-2 label.sk-toggleable__label-arrow:before {content: \"â–¸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-2 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-2 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-2 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-2 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-2 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-2 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"â–¾\";}#sk-container-id-2 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-2 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-2 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-2 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-2 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-2 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-2 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-2 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-2 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-2 div.sk-item {position: relative;z-index: 1;}#sk-container-id-2 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-2 div.sk-item::before, #sk-container-id-2 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-2 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-2 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-2 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-2 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-2 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-2 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-2 div.sk-label-container {text-align: center;}#sk-container-id-2 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-2 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-2\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>RandomForestClassifier(oob_score=True)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-2\" type=\"checkbox\" checked><label for=\"sk-estimator-id-2\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">RandomForestClassifier</label><div class=\"sk-toggleable__content\"><pre>RandomForestClassifier(oob_score=True)</pre></div></div></div></div></div>"
      ],
      "text/plain": [
       "RandomForestClassifier(oob_score=True)"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "joblib.load('trained_model.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69ebf269",
   "metadata": {},
   "source": [
    "Test the model, report relevant statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "c73327e9",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model score is: 0.5454545454545454\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "11"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "score = model.score(test_X, test_y)\n",
    "print('Model score is: {}'.format(score))\n",
    "\n",
    "preds = model.predict(test_X)\n",
    "len(preds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "669f962e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1, 1],\n",
       "       [1, 1],\n",
       "       [1, 1],\n",
       "       [1, 1],\n",
       "       [0, 0],\n",
       "       [1, 1],\n",
       "       [0, 0],\n",
       "       [1, 1],\n",
       "       [0, 0],\n",
       "       [0, 0],\n",
       "       [1, 1]])"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "7f62dbd9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy: 0.5454545454545454\n",
      "Precision: [0.42857143 0.42857143]\n",
      "Recall: [0.75 0.75]\n",
      "F1 Score: [0.54545455 0.54545455]\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score\n",
    "\n",
    "true_labels = test_y\n",
    "\n",
    "predicted_labels = preds\n",
    "\n",
    "# Calculate accuracy\n",
    "accuracy = accuracy_score(true_labels, predicted_labels)\n",
    "\n",
    "# Calculate precision\n",
    "precision = precision_score(true_labels, predicted_labels, average=None)\n",
    "\n",
    "# Calculate recall\n",
    "recall = recall_score(true_labels, predicted_labels, average=None)\n",
    "\n",
    "# Calculate F1 score\n",
    "f1 = f1_score(true_labels, predicted_labels, average=None)\n",
    "\n",
    "# Print the scores\n",
    "print(\"Accuracy:\", accuracy)\n",
    "print(\"Precision:\", precision)\n",
    "print(\"Recall:\", recall)\n",
    "print(\"F1 Score:\", f1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0ab9c8b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
