{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "006c3f36",
   "metadata": {},
   "source": [
    "Progress:\n",
    "\n",
    "DONE:\n",
    "-get package to work from command line\n",
    "-generate output from fasta files\n",
    "-programatically get package to work on one sequence\n",
    "-figure out indexing/confirm that package runs through dataframe in order\n",
    "\n",
    "NOT DONE:\n",
    "-send outfiles files into dataframe\n",
    "-get command line argument incorporated into script\n",
    "-scale beyond 5 sequences\n",
    "-train model with embeddings and compare results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "61c53eee",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats\n",
    "import sklearn.preprocessing\n",
    "import sklearn.model_selection\n",
    "import sklearn.ensemble\n",
    "import sklearn.feature_selection\n",
    "import unittest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "aae5ffbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "import esm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "ba87d47f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/loganroberts/Learn2Therm/ValidProt/notebooks\n"
     ]
    }
   ],
   "source": [
    "cd /Users/loganroberts/Learn2Therm/ValidProt/notebooks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "f90c1cdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#convert to pandas df\n",
    "df = pd.read_csv('learn2therm_sample_50k_exploration.csv').sample(5,random_state=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "51d90c54",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>thermo_index</th>\n",
       "      <th>...</th>\n",
       "      <th>bit_score_16s</th>\n",
       "      <th>m_ogt</th>\n",
       "      <th>t_ogt</th>\n",
       "      <th>ogt_difference</th>\n",
       "      <th>m_protein_seq</th>\n",
       "      <th>t_protein_seq</th>\n",
       "      <th>m_protein_desc</th>\n",
       "      <th>t_protein_desc</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>33553</th>\n",
       "      <td>33553</td>\n",
       "      <td>0.303922</td>\n",
       "      <td>0.276786</td>\n",
       "      <td>0.216783</td>\n",
       "      <td>199</td>\n",
       "      <td>0.888393</td>\n",
       "      <td>211</td>\n",
       "      <td>0.606322</td>\n",
       "      <td>221</td>\n",
       "      <td>9827</td>\n",
       "      <td>...</td>\n",
       "      <td>1004.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>MTNCKVDNKGWINVSDGSQLLDIQHLHTGFRIKDEYYDAVDDVSFE...</td>\n",
       "      <td>MDIKLHHVNKSYETGVETINVLKDIDLTIESGSWLTITGPSGSGKT...</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>348</td>\n",
       "      <td>224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9427</th>\n",
       "      <td>9427</td>\n",
       "      <td>0.303371</td>\n",
       "      <td>0.173633</td>\n",
       "      <td>0.190476</td>\n",
       "      <td>178</td>\n",
       "      <td>0.572347</td>\n",
       "      <td>196</td>\n",
       "      <td>0.765625</td>\n",
       "      <td>70</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>1035.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>MRNVAALSMPPRVPVEEGRTFTVDEGKRLLDAVADHRMGVLVLLAL...</td>\n",
       "      <td>MSGTRPDRVIRAFVDHLVVERGMARNTVDSYARDLRRYHAHLQARG...</td>\n",
       "      <td>site-specific integrase</td>\n",
       "      <td>site-specific tyrosine recombinase XerD</td>\n",
       "      <td>256</td>\n",
       "      <td>311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>199</td>\n",
       "      <td>0.292453</td>\n",
       "      <td>0.185629</td>\n",
       "      <td>0.184799</td>\n",
       "      <td>213</td>\n",
       "      <td>0.637725</td>\n",
       "      <td>215</td>\n",
       "      <td>0.637982</td>\n",
       "      <td>140</td>\n",
       "      <td>11324</td>\n",
       "      <td>...</td>\n",
       "      <td>989.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MFTLKHATAASAVVLGLVFATETFAADVTWNVSLWGKRRAFTEHVE...</td>\n",
       "      <td>MKIAREFAGALAGTLAGFAMAAVTALPVFAEELKFANFTPPFHTIN...</td>\n",
       "      <td>C4-dicarboxylate TRAP transporter substrate-bi...</td>\n",
       "      <td>TRAP transporter substrate-binding protein</td>\n",
       "      <td>337</td>\n",
       "      <td>334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12447</th>\n",
       "      <td>12447</td>\n",
       "      <td>0.406250</td>\n",
       "      <td>0.369919</td>\n",
       "      <td>0.366936</td>\n",
       "      <td>221</td>\n",
       "      <td>0.898374</td>\n",
       "      <td>230</td>\n",
       "      <td>0.920000</td>\n",
       "      <td>315</td>\n",
       "      <td>14963</td>\n",
       "      <td>...</td>\n",
       "      <td>1081.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>52.5</td>\n",
       "      <td>24.5</td>\n",
       "      <td>MVKADRVHKNFGALKVLKGISLEVGRGEVLCLVGPSGSGKSTFLRC...</td>\n",
       "      <td>MPPVLEARDIAKSFGPTPALRGLSLTIDAGEVVAITGPSGSGKSTL...</td>\n",
       "      <td>amino acid ABC transporter ATP-binding protein</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>250</td>\n",
       "      <td>246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39489</th>\n",
       "      <td>39489</td>\n",
       "      <td>0.412371</td>\n",
       "      <td>0.207254</td>\n",
       "      <td>0.172414</td>\n",
       "      <td>111</td>\n",
       "      <td>0.575129</td>\n",
       "      <td>102</td>\n",
       "      <td>0.376384</td>\n",
       "      <td>47</td>\n",
       "      <td>12872</td>\n",
       "      <td>...</td>\n",
       "      <td>1044.0</td>\n",
       "      <td>21.5</td>\n",
       "      <td>45.0</td>\n",
       "      <td>23.5</td>\n",
       "      <td>MIISLPAALGAAVLFGVASALQAYAARTEPPGSGLRGLLRVVLRLP...</td>\n",
       "      <td>MISLTVTAVVFGVVFLAELPDKTALAGLVLGTRYRASYVFAGVAAA...</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>TMEM165/GDT1 family protein</td>\n",
       "      <td>271</td>\n",
       "      <td>193</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0  local_gap_compressed_percent_id  \\\n",
       "33553       33553                         0.303922   \n",
       "9427         9427                         0.303371   \n",
       "199           199                         0.292453   \n",
       "12447       12447                         0.406250   \n",
       "39489       39489                         0.412371   \n",
       "\n",
       "       scaled_local_query_percent_id  scaled_local_symmetric_percent_id  \\\n",
       "33553                       0.276786                           0.216783   \n",
       "9427                        0.173633                           0.190476   \n",
       "199                         0.185629                           0.184799   \n",
       "12447                       0.369919                           0.366936   \n",
       "39489                       0.207254                           0.172414   \n",
       "\n",
       "       query_align_len  query_align_cov  subject_align_len  subject_align_cov  \\\n",
       "33553              199         0.888393                211           0.606322   \n",
       "9427               178         0.572347                196           0.765625   \n",
       "199                213         0.637725                215           0.637982   \n",
       "12447              221         0.898374                230           0.920000   \n",
       "39489              111         0.575129                102           0.376384   \n",
       "\n",
       "       bit_score  thermo_index  ...  bit_score_16s  m_ogt  t_ogt  \\\n",
       "33553        221          9827  ...         1004.0   26.0   50.0   \n",
       "9427          70           875  ...         1035.0   30.0   50.0   \n",
       "199          140         11324  ...          989.0   32.0   54.0   \n",
       "12447        315         14963  ...         1081.0   28.0   52.5   \n",
       "39489         47         12872  ...         1044.0   21.5   45.0   \n",
       "\n",
       "       ogt_difference                                      m_protein_seq  \\\n",
       "33553            24.0  MTNCKVDNKGWINVSDGSQLLDIQHLHTGFRIKDEYYDAVDDVSFE...   \n",
       "9427             20.0  MRNVAALSMPPRVPVEEGRTFTVDEGKRLLDAVADHRMGVLVLLAL...   \n",
       "199              22.0  MFTLKHATAASAVVLGLVFATETFAADVTWNVSLWGKRRAFTEHVE...   \n",
       "12447            24.5  MVKADRVHKNFGALKVLKGISLEVGRGEVLCLVGPSGSGKSTFLRC...   \n",
       "39489            23.5  MIISLPAALGAAVLFGVASALQAYAARTEPPGSGLRGLLRVVLRLP...   \n",
       "\n",
       "                                           t_protein_seq  \\\n",
       "33553  MDIKLHHVNKSYETGVETINVLKDIDLTIESGSWLTITGPSGSGKT...   \n",
       "9427   MSGTRPDRVIRAFVDHLVVERGMARNTVDSYARDLRRYHAHLQARG...   \n",
       "199    MKIAREFAGALAGTLAGFAMAAVTALPVFAEELKFANFTPPFHTIN...   \n",
       "12447  MPPVLEARDIAKSFGPTPALRGLSLTIDAGEVVAITGPSGSGKSTL...   \n",
       "39489  MISLTVTAVVFGVVFLAELPDKTALAGLVLGTRYRASYVFAGVAAA...   \n",
       "\n",
       "                                          m_protein_desc  \\\n",
       "33553                ABC transporter ATP-binding protein   \n",
       "9427                             site-specific integrase   \n",
       "199    C4-dicarboxylate TRAP transporter substrate-bi...   \n",
       "12447     amino acid ABC transporter ATP-binding protein   \n",
       "39489                               hypothetical protein   \n",
       "\n",
       "                                   t_protein_desc  m_protein_len  \\\n",
       "33553         ABC transporter ATP-binding protein            348   \n",
       "9427      site-specific tyrosine recombinase XerD            256   \n",
       "199    TRAP transporter substrate-binding protein            337   \n",
       "12447         ABC transporter ATP-binding protein            250   \n",
       "39489                 TMEM165/GDT1 family protein            271   \n",
       "\n",
       "       t_protein_len  \n",
       "33553            224  \n",
       "9427             311  \n",
       "199              334  \n",
       "12447            246  \n",
       "39489            193  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "a5477fd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_fasta_from_dataframe(\n",
    "        dataframe, output_file_a: str, output_file_b: str):\n",
    "    '''\n",
    "    Generates fasta file type from pandas dataframe.\n",
    "\n",
    "    Args:\n",
    "        Dataframe (pandas dataframe)\n",
    "        Names of output fasta files (str)\n",
    "\n",
    "    Returns:\n",
    "        Two fasta files with protein sequences and pair_id\n",
    "    '''\n",
    "    # meso sequence to fasta\n",
    "    with open(f'{output_file_a}', 'w') as f:\n",
    "        for _, row in dataframe.iterrows():\n",
    "            f.write(\n",
    "                '>{}\\n{}\\n'.format(\n",
    "                    (row['prot_pair_index']),\n",
    "                    row['m_protein_seq']))\n",
    "\n",
    "    # thermo sequence to fasta\n",
    "    with open(f'{output_file_b}', 'w') as f:\n",
    "        for _, row in dataframe.iterrows():\n",
    "            f.write(\n",
    "                '>{}\\n{}\\n'.format(\n",
    "                    (row['prot_pair_index']),\n",
    "                    (row['t_protein_seq'])))\n",
    "\n",
    "    # return output files\n",
    "    return [output_file_a, output_file_b]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "3d0cc1c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['seq_a.fasta', 'seq_b.fasta']"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_fasta_from_dataframe(df, 'seq_a.fasta', 'seq_b.fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2193416",
   "metadata": {},
   "source": [
    "Would be a good spot to figure out how to generate esm embeddings from a fasta file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "f9caa733",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-132-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-132-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-132-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-132-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
      "<ipython-input-132-2b3af67eb1b2>:29: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "#this block of code should load the files into a dataframe \n",
    "\n",
    "import os\n",
    "\n",
    "# Create an empty dataframe\n",
    "embeddings = pd.DataFrame(columns=['prot_pair_index', 'tensor'])\n",
    "\n",
    "# Set the directory path containing the PyTorch files\n",
    "directory = '/Users/loganroberts/Learn2Therm/ValidProt/data/pytorch_embeddings'\n",
    "\n",
    "# Iterate over the files in the directory\n",
    "for filename in os.listdir(directory):\n",
    "    if filename.endswith('.pt'):  # Adjust the file extension as needed\n",
    "        file_path = os.path.join(directory, filename)\n",
    "        \n",
    "        # Load the PyTorch file\n",
    "        data = torch.load(file_path)\n",
    "        \n",
    "        # Extract the tensor(s) from the loaded data\n",
    "        if isinstance(data, torch.Tensor):\n",
    "            tensors = [data]\n",
    "        elif isinstance(data, dict):\n",
    "            tensors = list(data.values())\n",
    "        else:\n",
    "            # Handle other cases based on your specific file structure\n",
    "            continue\n",
    "        \n",
    "        # Append the filename and tensor(s) to the dataframe\n",
    "        embeddings = embeddings.append({'prot_pair_index': int(re.sub('.pt', '', filename)), 'tensor': tensors[1][33].numpy()}, ignore_index=True)\n",
    "\n",
    "#expand tensor values into individual columns\n",
    "embeddings = pd.concat([embeddings.drop('tensor', axis=1),\n",
    "                embeddings['tensor'].apply(lambda x: pd.Series(x))], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "bb01b4e7",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prot_pair_index</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>...</th>\n",
       "      <th>1270</th>\n",
       "      <th>1271</th>\n",
       "      <th>1272</th>\n",
       "      <th>1273</th>\n",
       "      <th>1274</th>\n",
       "      <th>1275</th>\n",
       "      <th>1276</th>\n",
       "      <th>1277</th>\n",
       "      <th>1278</th>\n",
       "      <th>1279</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>170424926</td>\n",
       "      <td>-0.089057</td>\n",
       "      <td>-0.048351</td>\n",
       "      <td>-0.072258</td>\n",
       "      <td>-0.022964</td>\n",
       "      <td>0.011155</td>\n",
       "      <td>0.049804</td>\n",
       "      <td>-0.011335</td>\n",
       "      <td>-0.145744</td>\n",
       "      <td>-0.019996</td>\n",
       "      <td>...</td>\n",
       "      <td>0.141867</td>\n",
       "      <td>-0.116855</td>\n",
       "      <td>-0.058509</td>\n",
       "      <td>-0.069415</td>\n",
       "      <td>-0.059710</td>\n",
       "      <td>0.154020</td>\n",
       "      <td>-0.028967</td>\n",
       "      <td>-0.131237</td>\n",
       "      <td>-0.002997</td>\n",
       "      <td>0.196756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>172778855</td>\n",
       "      <td>-0.004024</td>\n",
       "      <td>-0.011217</td>\n",
       "      <td>-0.035143</td>\n",
       "      <td>0.051118</td>\n",
       "      <td>-0.118710</td>\n",
       "      <td>-0.000500</td>\n",
       "      <td>-0.000316</td>\n",
       "      <td>-0.089665</td>\n",
       "      <td>-0.068695</td>\n",
       "      <td>...</td>\n",
       "      <td>0.025760</td>\n",
       "      <td>-0.009402</td>\n",
       "      <td>-0.013473</td>\n",
       "      <td>-0.062470</td>\n",
       "      <td>-0.019069</td>\n",
       "      <td>0.039852</td>\n",
       "      <td>0.118363</td>\n",
       "      <td>-0.105894</td>\n",
       "      <td>-0.026337</td>\n",
       "      <td>0.059242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>102140025</td>\n",
       "      <td>0.049990</td>\n",
       "      <td>-0.100387</td>\n",
       "      <td>-0.035571</td>\n",
       "      <td>-0.016107</td>\n",
       "      <td>-0.132159</td>\n",
       "      <td>0.012995</td>\n",
       "      <td>0.004166</td>\n",
       "      <td>0.067679</td>\n",
       "      <td>-0.088118</td>\n",
       "      <td>...</td>\n",
       "      <td>0.035702</td>\n",
       "      <td>-0.111666</td>\n",
       "      <td>-0.110127</td>\n",
       "      <td>0.046714</td>\n",
       "      <td>-0.061923</td>\n",
       "      <td>0.065274</td>\n",
       "      <td>0.012571</td>\n",
       "      <td>-0.138126</td>\n",
       "      <td>0.054840</td>\n",
       "      <td>0.069625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>76961413</td>\n",
       "      <td>-0.074474</td>\n",
       "      <td>-0.027899</td>\n",
       "      <td>0.005806</td>\n",
       "      <td>0.005537</td>\n",
       "      <td>-0.070316</td>\n",
       "      <td>0.009532</td>\n",
       "      <td>0.079727</td>\n",
       "      <td>-0.107065</td>\n",
       "      <td>-0.062455</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.026527</td>\n",
       "      <td>-0.105266</td>\n",
       "      <td>0.017469</td>\n",
       "      <td>-0.045476</td>\n",
       "      <td>-0.106896</td>\n",
       "      <td>0.073323</td>\n",
       "      <td>0.118733</td>\n",
       "      <td>-0.177319</td>\n",
       "      <td>-0.010493</td>\n",
       "      <td>0.087448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9602481</td>\n",
       "      <td>0.001106</td>\n",
       "      <td>-0.054233</td>\n",
       "      <td>0.053005</td>\n",
       "      <td>0.033410</td>\n",
       "      <td>-0.124973</td>\n",
       "      <td>0.018081</td>\n",
       "      <td>0.025389</td>\n",
       "      <td>-0.076978</td>\n",
       "      <td>0.003235</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.064183</td>\n",
       "      <td>-0.088116</td>\n",
       "      <td>-0.041833</td>\n",
       "      <td>-0.008492</td>\n",
       "      <td>-0.011153</td>\n",
       "      <td>-0.081377</td>\n",
       "      <td>0.110826</td>\n",
       "      <td>-0.107516</td>\n",
       "      <td>-0.019247</td>\n",
       "      <td>0.061446</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1281 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  prot_pair_index         0         1         2         3         4         5  \\\n",
       "0       170424926 -0.089057 -0.048351 -0.072258 -0.022964  0.011155  0.049804   \n",
       "1       172778855 -0.004024 -0.011217 -0.035143  0.051118 -0.118710 -0.000500   \n",
       "2       102140025  0.049990 -0.100387 -0.035571 -0.016107 -0.132159  0.012995   \n",
       "3        76961413 -0.074474 -0.027899  0.005806  0.005537 -0.070316  0.009532   \n",
       "4         9602481  0.001106 -0.054233  0.053005  0.033410 -0.124973  0.018081   \n",
       "\n",
       "          6         7         8  ...      1270      1271      1272      1273  \\\n",
       "0 -0.011335 -0.145744 -0.019996  ...  0.141867 -0.116855 -0.058509 -0.069415   \n",
       "1 -0.000316 -0.089665 -0.068695  ...  0.025760 -0.009402 -0.013473 -0.062470   \n",
       "2  0.004166  0.067679 -0.088118  ...  0.035702 -0.111666 -0.110127  0.046714   \n",
       "3  0.079727 -0.107065 -0.062455  ... -0.026527 -0.105266  0.017469 -0.045476   \n",
       "4  0.025389 -0.076978  0.003235  ... -0.064183 -0.088116 -0.041833 -0.008492   \n",
       "\n",
       "       1274      1275      1276      1277      1278      1279  \n",
       "0 -0.059710  0.154020 -0.028967 -0.131237 -0.002997  0.196756  \n",
       "1 -0.019069  0.039852  0.118363 -0.105894 -0.026337  0.059242  \n",
       "2 -0.061923  0.065274  0.012571 -0.138126  0.054840  0.069625  \n",
       "3 -0.106896  0.073323  0.118733 -0.177319 -0.010493  0.087448  \n",
       "4 -0.011153 -0.081377  0.110826 -0.107516 -0.019247  0.061446  \n",
       "\n",
       "[5 rows x 1281 columns]"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "embeddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "ad8e4fea",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "new_df = pd.merge(embeddings, df, on='prot_pair_index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "1689ff2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prot_pair_index</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>...</th>\n",
       "      <th>bit_score_16s</th>\n",
       "      <th>m_ogt</th>\n",
       "      <th>t_ogt</th>\n",
       "      <th>ogt_difference</th>\n",
       "      <th>m_protein_seq</th>\n",
       "      <th>t_protein_seq</th>\n",
       "      <th>m_protein_desc</th>\n",
       "      <th>t_protein_desc</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>170424926</td>\n",
       "      <td>-0.089057</td>\n",
       "      <td>-0.048351</td>\n",
       "      <td>-0.072258</td>\n",
       "      <td>-0.022964</td>\n",
       "      <td>0.011155</td>\n",
       "      <td>0.049804</td>\n",
       "      <td>-0.011335</td>\n",
       "      <td>-0.145744</td>\n",
       "      <td>-0.019996</td>\n",
       "      <td>...</td>\n",
       "      <td>1044.0</td>\n",
       "      <td>21.5</td>\n",
       "      <td>45.0</td>\n",
       "      <td>23.5</td>\n",
       "      <td>MIISLPAALGAAVLFGVASALQAYAARTEPPGSGLRGLLRVVLRLP...</td>\n",
       "      <td>MISLTVTAVVFGVVFLAELPDKTALAGLVLGTRYRASYVFAGVAAA...</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>TMEM165/GDT1 family protein</td>\n",
       "      <td>271</td>\n",
       "      <td>193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>172778855</td>\n",
       "      <td>-0.004024</td>\n",
       "      <td>-0.011217</td>\n",
       "      <td>-0.035143</td>\n",
       "      <td>0.051118</td>\n",
       "      <td>-0.118710</td>\n",
       "      <td>-0.000500</td>\n",
       "      <td>-0.000316</td>\n",
       "      <td>-0.089665</td>\n",
       "      <td>-0.068695</td>\n",
       "      <td>...</td>\n",
       "      <td>1004.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>MTNCKVDNKGWINVSDGSQLLDIQHLHTGFRIKDEYYDAVDDVSFE...</td>\n",
       "      <td>MDIKLHHVNKSYETGVETINVLKDIDLTIESGSWLTITGPSGSGKT...</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>348</td>\n",
       "      <td>224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>102140025</td>\n",
       "      <td>0.049990</td>\n",
       "      <td>-0.100387</td>\n",
       "      <td>-0.035571</td>\n",
       "      <td>-0.016107</td>\n",
       "      <td>-0.132159</td>\n",
       "      <td>0.012995</td>\n",
       "      <td>0.004166</td>\n",
       "      <td>0.067679</td>\n",
       "      <td>-0.088118</td>\n",
       "      <td>...</td>\n",
       "      <td>1035.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>MRNVAALSMPPRVPVEEGRTFTVDEGKRLLDAVADHRMGVLVLLAL...</td>\n",
       "      <td>MSGTRPDRVIRAFVDHLVVERGMARNTVDSYARDLRRYHAHLQARG...</td>\n",
       "      <td>site-specific integrase</td>\n",
       "      <td>site-specific tyrosine recombinase XerD</td>\n",
       "      <td>256</td>\n",
       "      <td>311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>76961413</td>\n",
       "      <td>-0.074474</td>\n",
       "      <td>-0.027899</td>\n",
       "      <td>0.005806</td>\n",
       "      <td>0.005537</td>\n",
       "      <td>-0.070316</td>\n",
       "      <td>0.009532</td>\n",
       "      <td>0.079727</td>\n",
       "      <td>-0.107065</td>\n",
       "      <td>-0.062455</td>\n",
       "      <td>...</td>\n",
       "      <td>989.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MFTLKHATAASAVVLGLVFATETFAADVTWNVSLWGKRRAFTEHVE...</td>\n",
       "      <td>MKIAREFAGALAGTLAGFAMAAVTALPVFAEELKFANFTPPFHTIN...</td>\n",
       "      <td>C4-dicarboxylate TRAP transporter substrate-bi...</td>\n",
       "      <td>TRAP transporter substrate-binding protein</td>\n",
       "      <td>337</td>\n",
       "      <td>334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9602481</td>\n",
       "      <td>0.001106</td>\n",
       "      <td>-0.054233</td>\n",
       "      <td>0.053005</td>\n",
       "      <td>0.033410</td>\n",
       "      <td>-0.124973</td>\n",
       "      <td>0.018081</td>\n",
       "      <td>0.025389</td>\n",
       "      <td>-0.076978</td>\n",
       "      <td>0.003235</td>\n",
       "      <td>...</td>\n",
       "      <td>1081.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>52.5</td>\n",
       "      <td>24.5</td>\n",
       "      <td>MVKADRVHKNFGALKVLKGISLEVGRGEVLCLVGPSGSGKSTFLRC...</td>\n",
       "      <td>MPPVLEARDIAKSFGPTPALRGLSLTIDAGEVVAITGPSGSGKSTL...</td>\n",
       "      <td>amino acid ABC transporter ATP-binding protein</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>250</td>\n",
       "      <td>246</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1310 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  prot_pair_index         0         1         2         3         4         5  \\\n",
       "0       170424926 -0.089057 -0.048351 -0.072258 -0.022964  0.011155  0.049804   \n",
       "1       172778855 -0.004024 -0.011217 -0.035143  0.051118 -0.118710 -0.000500   \n",
       "2       102140025  0.049990 -0.100387 -0.035571 -0.016107 -0.132159  0.012995   \n",
       "3        76961413 -0.074474 -0.027899  0.005806  0.005537 -0.070316  0.009532   \n",
       "4         9602481  0.001106 -0.054233  0.053005  0.033410 -0.124973  0.018081   \n",
       "\n",
       "          6         7         8  ...  bit_score_16s  m_ogt  t_ogt  \\\n",
       "0 -0.011335 -0.145744 -0.019996  ...         1044.0   21.5   45.0   \n",
       "1 -0.000316 -0.089665 -0.068695  ...         1004.0   26.0   50.0   \n",
       "2  0.004166  0.067679 -0.088118  ...         1035.0   30.0   50.0   \n",
       "3  0.079727 -0.107065 -0.062455  ...          989.0   32.0   54.0   \n",
       "4  0.025389 -0.076978  0.003235  ...         1081.0   28.0   52.5   \n",
       "\n",
       "   ogt_difference                                      m_protein_seq  \\\n",
       "0            23.5  MIISLPAALGAAVLFGVASALQAYAARTEPPGSGLRGLLRVVLRLP...   \n",
       "1            24.0  MTNCKVDNKGWINVSDGSQLLDIQHLHTGFRIKDEYYDAVDDVSFE...   \n",
       "2            20.0  MRNVAALSMPPRVPVEEGRTFTVDEGKRLLDAVADHRMGVLVLLAL...   \n",
       "3            22.0  MFTLKHATAASAVVLGLVFATETFAADVTWNVSLWGKRRAFTEHVE...   \n",
       "4            24.5  MVKADRVHKNFGALKVLKGISLEVGRGEVLCLVGPSGSGKSTFLRC...   \n",
       "\n",
       "                                       t_protein_seq  \\\n",
       "0  MISLTVTAVVFGVVFLAELPDKTALAGLVLGTRYRASYVFAGVAAA...   \n",
       "1  MDIKLHHVNKSYETGVETINVLKDIDLTIESGSWLTITGPSGSGKT...   \n",
       "2  MSGTRPDRVIRAFVDHLVVERGMARNTVDSYARDLRRYHAHLQARG...   \n",
       "3  MKIAREFAGALAGTLAGFAMAAVTALPVFAEELKFANFTPPFHTIN...   \n",
       "4  MPPVLEARDIAKSFGPTPALRGLSLTIDAGEVVAITGPSGSGKSTL...   \n",
       "\n",
       "                                      m_protein_desc  \\\n",
       "0                               hypothetical protein   \n",
       "1                ABC transporter ATP-binding protein   \n",
       "2                            site-specific integrase   \n",
       "3  C4-dicarboxylate TRAP transporter substrate-bi...   \n",
       "4     amino acid ABC transporter ATP-binding protein   \n",
       "\n",
       "                               t_protein_desc  m_protein_len  t_protein_len  \n",
       "0                 TMEM165/GDT1 family protein            271            193  \n",
       "1         ABC transporter ATP-binding protein            348            224  \n",
       "2     site-specific tyrosine recombinase XerD            256            311  \n",
       "3  TRAP transporter substrate-binding protein            337            334  \n",
       "4         ABC transporter ATP-binding protein            250            246  \n",
       "\n",
       "[5 rows x 1310 columns]"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd261ce4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "d937efd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import subprocess\n",
    "\n",
    "# # Run the command line prompt and capture the output\n",
    "# output = subprocess.check_output(['esm-extract esm2_t36_3B_UR50D seq_a.fasta data/ --include mean'], universal_newlines=True)\n",
    "\n",
    "# # Process the captured output as needed\n",
    "# processed_data = output.strip()\n",
    "\n",
    "# # Use the processed data in your script\n",
    "# print(\"Processed data:\", processed_data)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44e374b6",
   "metadata": {},
   "source": [
    "Below code is for running on individual sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "184ced04",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# # Load ESM-2 model\n",
    "# model, alphabet = esm.pretrained.esm2_t33_650M_UR50D()\n",
    "# batch_converter = alphabet.get_batch_converter()\n",
    "# model.eval()  # disables dropout for deterministic results\n",
    "\n",
    "# # Prepare data (first 2 sequences from ESMStructuralSplitDataset superfamily / 4)\n",
    "# data = [(df['prot_pair_index'].iloc[0], df['m_protein_seq'].iloc[0])]\n",
    "\n",
    "# batch_labels, batch_strs, batch_tokens = batch_converter(data)\n",
    "# batch_lens = (batch_tokens != alphabet.padding_idx).sum(1)\n",
    "\n",
    "# # Extract per-residue representations (on CPU)\n",
    "# with torch.no_grad():\n",
    "#     results = model(batch_tokens, repr_layers=[33], return_contacts=True)\n",
    "# token_representations = results[\"representations\"][33]\n",
    "\n",
    "# # Generate per-sequence representations via averaging\n",
    "# # NOTE: token 0 is always a beginning-of-sequence token, so the first residue is token 1.\n",
    "# sequence_representations = []\n",
    "# for i, tokens_len in enumerate(batch_lens):\n",
    "#     sequence_representations.append(token_representations[i, 1 : tokens_len - 1].mean(0))\n",
    "\n",
    "# # Look at the unsupervised self-attention map contact predictions\n",
    "# # import matplotlib.pyplot as plt\n",
    "# # for (_, seq), tokens_len, attention_contacts in zip(data, batch_lens, results[\"contacts\"]):\n",
    "# #     plt.matshow(attention_contacts[: tokens_len, : tokens_len])\n",
    "# #     plt.title(seq)\n",
    "# #     plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "418ca509",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "b883627d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(17108, 33)"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hmmer_target = pd.read_csv('protein_match_50k')\n",
    "hmmer_target['protein_match'] = hmmer_target['protein_match'].map({'Yes': 1, 'No': 0})\n",
    "hmmer_target\n",
    "\n",
    "\n",
    "\n",
    "from sklearn.utils import resample\n",
    "\n",
    "# Assuming your data is in a pandas DataFrame called 'data'\n",
    "# Separate the majority and minority classes\n",
    "majority_class = hmmer_target[hmmer_target['protein_match'] == 1]\n",
    "minority_class = hmmer_target[hmmer_target['protein_match'] == 0]\n",
    "\n",
    "# Undersample the majority class to match the number of minority class samples\n",
    "n_samples = len(minority_class)\n",
    "undersampled_majority = resample(majority_class, n_samples=n_samples, replace=False)\n",
    "\n",
    "# Combine the undersampled majority class with the minority class\n",
    "balanced_data = pd.concat([undersampled_majority, minority_class])\n",
    "\n",
    "balanced_data\n",
    "\n",
    "df = pd.merge(df, balanced_data, on=['prot_pair_index'])\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "14015e54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1    8554\n",
       "0    8554\n",
       "Name: protein_match, dtype: int64"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['protein_match'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "06ad97a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['structure_match'] = df['protein_match']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "fa81d559",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0_x</th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>thermo_index</th>\n",
       "      <th>...</th>\n",
       "      <th>m_protein_seq</th>\n",
       "      <th>t_protein_seq</th>\n",
       "      <th>m_protein_desc</th>\n",
       "      <th>t_protein_desc</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "      <th>Unnamed: 0_y</th>\n",
       "      <th>protein_match</th>\n",
       "      <th>Jaccard_Score</th>\n",
       "      <th>structure_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.267380</td>\n",
       "      <td>0.254453</td>\n",
       "      <td>148</td>\n",
       "      <td>0.791444</td>\n",
       "      <td>163</td>\n",
       "      <td>0.791262</td>\n",
       "      <td>127</td>\n",
       "      <td>8919</td>\n",
       "      <td>...</td>\n",
       "      <td>MVTTGERQRNARGEGARLRLEIVAATQALLADGETATLRSIARRAG...</td>\n",
       "      <td>MNRPTYHHGDLRAAILTEAARLVAERGAERVSLRELAREAGVSHAA...</td>\n",
       "      <td>TetR/AcrR family transcriptional regulator</td>\n",
       "      <td>TetR/AcrR family transcriptional regulator</td>\n",
       "      <td>206</td>\n",
       "      <td>187</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.295082</td>\n",
       "      <td>0.291498</td>\n",
       "      <td>218</td>\n",
       "      <td>0.893443</td>\n",
       "      <td>221</td>\n",
       "      <td>0.884000</td>\n",
       "      <td>158</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>MQIKDSVAVVTGGASGLGLATTKRLLDAGGSVVVIDLKGEDVVAEL...</td>\n",
       "      <td>MTGTVVITGGSRGIGAATAVLAAERGWQVAVSFRERRDAAEQVVRR...</td>\n",
       "      <td>3-hydroxyacyl-CoA dehydrogenase</td>\n",
       "      <td>SDR family oxidoreductase</td>\n",
       "      <td>250</td>\n",
       "      <td>244</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>0.75</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11</td>\n",
       "      <td>0.275281</td>\n",
       "      <td>0.182156</td>\n",
       "      <td>0.152174</td>\n",
       "      <td>185</td>\n",
       "      <td>0.687732</td>\n",
       "      <td>197</td>\n",
       "      <td>0.525333</td>\n",
       "      <td>48</td>\n",
       "      <td>14963</td>\n",
       "      <td>...</td>\n",
       "      <td>MRAIGELWPTFDDVHEIAVLRGGGLGDLMFALPAIDALAAAYPEAR...</td>\n",
       "      <td>MATTTAEIGVIGGSGFYSFLDDPHEVTVQTPYGPPSDPIAVGTVAG...</td>\n",
       "      <td>glycosyltransferase family 9 protein</td>\n",
       "      <td>S-methyl-5'-thioadenosine phosphorylase</td>\n",
       "      <td>375</td>\n",
       "      <td>269</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>18</td>\n",
       "      <td>0.394495</td>\n",
       "      <td>0.346774</td>\n",
       "      <td>0.369099</td>\n",
       "      <td>108</td>\n",
       "      <td>0.870968</td>\n",
       "      <td>106</td>\n",
       "      <td>0.972477</td>\n",
       "      <td>130</td>\n",
       "      <td>14963</td>\n",
       "      <td>...</td>\n",
       "      <td>MHELSIAQSVVEAVCERADGRRVRSVRMVVGALCAVVPDSMLFCFE...</td>\n",
       "      <td>MGLCEAIVDAVLQRAEGRRVRSVRVRVAGHPVVREVVDQGFAMAAA...</td>\n",
       "      <td>hydrogenase maturation nickel metallochaperone...</td>\n",
       "      <td>hydrogenase maturation nickel metallochaperone...</td>\n",
       "      <td>109</td>\n",
       "      <td>124</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>20</td>\n",
       "      <td>0.268908</td>\n",
       "      <td>0.232727</td>\n",
       "      <td>0.247582</td>\n",
       "      <td>239</td>\n",
       "      <td>0.869091</td>\n",
       "      <td>236</td>\n",
       "      <td>0.975207</td>\n",
       "      <td>191</td>\n",
       "      <td>1674</td>\n",
       "      <td>...</td>\n",
       "      <td>MKKAVLEVRDVTKRVKGRTLLSHMSFSIYKNEICGFVGPNGSGKTT...</td>\n",
       "      <td>MNETHFGARDLSLAYGKKRIIEALNIDIPIGYITALIGRNGSGKST...</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>242</td>\n",
       "      <td>275</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "      <td>0.40</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17103</th>\n",
       "      <td>49995</td>\n",
       "      <td>0.344828</td>\n",
       "      <td>0.169972</td>\n",
       "      <td>0.192308</td>\n",
       "      <td>180</td>\n",
       "      <td>0.509915</td>\n",
       "      <td>200</td>\n",
       "      <td>0.738007</td>\n",
       "      <td>53</td>\n",
       "      <td>14963</td>\n",
       "      <td>...</td>\n",
       "      <td>MDNATFRLGDDLSVRLPGHSRWIGQVEREQRWLPWLAPRLPLTVST...</td>\n",
       "      <td>MPPQPRPLRPNDPREIGGFALLGRLGEGGQGTVYLGGAPDGRRVAV...</td>\n",
       "      <td>aminoglycoside phosphotransferase family protein</td>\n",
       "      <td>serine/threonine protein kinase</td>\n",
       "      <td>271</td>\n",
       "      <td>353</td>\n",
       "      <td>4875</td>\n",
       "      <td>0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17104</th>\n",
       "      <td>49996</td>\n",
       "      <td>0.417989</td>\n",
       "      <td>0.389163</td>\n",
       "      <td>0.392060</td>\n",
       "      <td>190</td>\n",
       "      <td>0.935961</td>\n",
       "      <td>187</td>\n",
       "      <td>0.935000</td>\n",
       "      <td>314</td>\n",
       "      <td>7134</td>\n",
       "      <td>...</td>\n",
       "      <td>MFRTGVKAEIGRSLAVVGEAEDVERAVRVVLEQRPDVVLLDVHLPG...</td>\n",
       "      <td>MILEAEPDIVVVGEAGDGEKAVEEARALQPDVVLMDIRMPRKDGVE...</td>\n",
       "      <td>response regulator transcription factor</td>\n",
       "      <td>response regulator transcription factor</td>\n",
       "      <td>200</td>\n",
       "      <td>203</td>\n",
       "      <td>4876</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17105</th>\n",
       "      <td>49997</td>\n",
       "      <td>0.352174</td>\n",
       "      <td>0.239645</td>\n",
       "      <td>0.242152</td>\n",
       "      <td>230</td>\n",
       "      <td>0.680473</td>\n",
       "      <td>227</td>\n",
       "      <td>0.685801</td>\n",
       "      <td>279</td>\n",
       "      <td>7134</td>\n",
       "      <td>...</td>\n",
       "      <td>MIRLAELTKTYPGQQHPAVDGISMEVAEGEIVVLVGPSGCGKTTTL...</td>\n",
       "      <td>MTEQPILSARGLTVDFRLRGGRRARAVDGVDLDLAPGEVLALAGES...</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>331</td>\n",
       "      <td>338</td>\n",
       "      <td>4877</td>\n",
       "      <td>0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17106</th>\n",
       "      <td>49998</td>\n",
       "      <td>0.334764</td>\n",
       "      <td>0.331915</td>\n",
       "      <td>0.329810</td>\n",
       "      <td>232</td>\n",
       "      <td>0.987234</td>\n",
       "      <td>230</td>\n",
       "      <td>0.966387</td>\n",
       "      <td>281</td>\n",
       "      <td>11324</td>\n",
       "      <td>...</td>\n",
       "      <td>MSESHAGALLSVRGLTAGYGGATALDGVSLTVAAGETVALLGANGA...</td>\n",
       "      <td>MSLLTTSGLTRHFSGIHAVEGVDFTLEAGEIRALIGSNGAGKTTLV...</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>238</td>\n",
       "      <td>235</td>\n",
       "      <td>4878</td>\n",
       "      <td>1</td>\n",
       "      <td>0.50</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17107</th>\n",
       "      <td>49999</td>\n",
       "      <td>0.336207</td>\n",
       "      <td>0.184834</td>\n",
       "      <td>0.174497</td>\n",
       "      <td>113</td>\n",
       "      <td>0.535545</td>\n",
       "      <td>117</td>\n",
       "      <td>0.495763</td>\n",
       "      <td>165</td>\n",
       "      <td>7134</td>\n",
       "      <td>...</td>\n",
       "      <td>MTGQGASGSETGVSDDSPITVVVVDDQELLRTGLRDLAEHDGDIAV...</td>\n",
       "      <td>MTTEAQPRRVVLAEDEALIRLDLAEMLAEEGYEVVGQAADGETALA...</td>\n",
       "      <td>response regulator transcription factor</td>\n",
       "      <td>response regulator</td>\n",
       "      <td>236</td>\n",
       "      <td>211</td>\n",
       "      <td>4879</td>\n",
       "      <td>0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17108 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0_x  local_gap_compressed_percent_id  \\\n",
       "0                 5                         0.333333   \n",
       "1                 9                         0.333333   \n",
       "2                11                         0.275281   \n",
       "3                18                         0.394495   \n",
       "4                20                         0.268908   \n",
       "...             ...                              ...   \n",
       "17103         49995                         0.344828   \n",
       "17104         49996                         0.417989   \n",
       "17105         49997                         0.352174   \n",
       "17106         49998                         0.334764   \n",
       "17107         49999                         0.336207   \n",
       "\n",
       "       scaled_local_query_percent_id  scaled_local_symmetric_percent_id  \\\n",
       "0                           0.267380                           0.254453   \n",
       "1                           0.295082                           0.291498   \n",
       "2                           0.182156                           0.152174   \n",
       "3                           0.346774                           0.369099   \n",
       "4                           0.232727                           0.247582   \n",
       "...                              ...                                ...   \n",
       "17103                       0.169972                           0.192308   \n",
       "17104                       0.389163                           0.392060   \n",
       "17105                       0.239645                           0.242152   \n",
       "17106                       0.331915                           0.329810   \n",
       "17107                       0.184834                           0.174497   \n",
       "\n",
       "       query_align_len  query_align_cov  subject_align_len  subject_align_cov  \\\n",
       "0                  148         0.791444                163           0.791262   \n",
       "1                  218         0.893443                221           0.884000   \n",
       "2                  185         0.687732                197           0.525333   \n",
       "3                  108         0.870968                106           0.972477   \n",
       "4                  239         0.869091                236           0.975207   \n",
       "...                ...              ...                ...                ...   \n",
       "17103              180         0.509915                200           0.738007   \n",
       "17104              190         0.935961                187           0.935000   \n",
       "17105              230         0.680473                227           0.685801   \n",
       "17106              232         0.987234                230           0.966387   \n",
       "17107              113         0.535545                117           0.495763   \n",
       "\n",
       "       bit_score  thermo_index  ...  \\\n",
       "0            127          8919  ...   \n",
       "1            158           875  ...   \n",
       "2             48         14963  ...   \n",
       "3            130         14963  ...   \n",
       "4            191          1674  ...   \n",
       "...          ...           ...  ...   \n",
       "17103         53         14963  ...   \n",
       "17104        314          7134  ...   \n",
       "17105        279          7134  ...   \n",
       "17106        281         11324  ...   \n",
       "17107        165          7134  ...   \n",
       "\n",
       "                                           m_protein_seq  \\\n",
       "0      MVTTGERQRNARGEGARLRLEIVAATQALLADGETATLRSIARRAG...   \n",
       "1      MQIKDSVAVVTGGASGLGLATTKRLLDAGGSVVVIDLKGEDVVAEL...   \n",
       "2      MRAIGELWPTFDDVHEIAVLRGGGLGDLMFALPAIDALAAAYPEAR...   \n",
       "3      MHELSIAQSVVEAVCERADGRRVRSVRMVVGALCAVVPDSMLFCFE...   \n",
       "4      MKKAVLEVRDVTKRVKGRTLLSHMSFSIYKNEICGFVGPNGSGKTT...   \n",
       "...                                                  ...   \n",
       "17103  MDNATFRLGDDLSVRLPGHSRWIGQVEREQRWLPWLAPRLPLTVST...   \n",
       "17104  MFRTGVKAEIGRSLAVVGEAEDVERAVRVVLEQRPDVVLLDVHLPG...   \n",
       "17105  MIRLAELTKTYPGQQHPAVDGISMEVAEGEIVVLVGPSGCGKTTTL...   \n",
       "17106  MSESHAGALLSVRGLTAGYGGATALDGVSLTVAAGETVALLGANGA...   \n",
       "17107  MTGQGASGSETGVSDDSPITVVVVDDQELLRTGLRDLAEHDGDIAV...   \n",
       "\n",
       "                                           t_protein_seq  \\\n",
       "0      MNRPTYHHGDLRAAILTEAARLVAERGAERVSLRELAREAGVSHAA...   \n",
       "1      MTGTVVITGGSRGIGAATAVLAAERGWQVAVSFRERRDAAEQVVRR...   \n",
       "2      MATTTAEIGVIGGSGFYSFLDDPHEVTVQTPYGPPSDPIAVGTVAG...   \n",
       "3      MGLCEAIVDAVLQRAEGRRVRSVRVRVAGHPVVREVVDQGFAMAAA...   \n",
       "4      MNETHFGARDLSLAYGKKRIIEALNIDIPIGYITALIGRNGSGKST...   \n",
       "...                                                  ...   \n",
       "17103  MPPQPRPLRPNDPREIGGFALLGRLGEGGQGTVYLGGAPDGRRVAV...   \n",
       "17104  MILEAEPDIVVVGEAGDGEKAVEEARALQPDVVLMDIRMPRKDGVE...   \n",
       "17105  MTEQPILSARGLTVDFRLRGGRRARAVDGVDLDLAPGEVLALAGES...   \n",
       "17106  MSLLTTSGLTRHFSGIHAVEGVDFTLEAGEIRALIGSNGAGKTTLV...   \n",
       "17107  MTTEAQPRRVVLAEDEALIRLDLAEMLAEEGYEVVGQAADGETALA...   \n",
       "\n",
       "                                          m_protein_desc  \\\n",
       "0             TetR/AcrR family transcriptional regulator   \n",
       "1                        3-hydroxyacyl-CoA dehydrogenase   \n",
       "2                   glycosyltransferase family 9 protein   \n",
       "3      hydrogenase maturation nickel metallochaperone...   \n",
       "4                    ABC transporter ATP-binding protein   \n",
       "...                                                  ...   \n",
       "17103   aminoglycoside phosphotransferase family protein   \n",
       "17104            response regulator transcription factor   \n",
       "17105                ABC transporter ATP-binding protein   \n",
       "17106                ABC transporter ATP-binding protein   \n",
       "17107            response regulator transcription factor   \n",
       "\n",
       "                                          t_protein_desc  m_protein_len  \\\n",
       "0             TetR/AcrR family transcriptional regulator            206   \n",
       "1                              SDR family oxidoreductase            250   \n",
       "2                S-methyl-5'-thioadenosine phosphorylase            375   \n",
       "3      hydrogenase maturation nickel metallochaperone...            109   \n",
       "4                    ABC transporter ATP-binding protein            242   \n",
       "...                                                  ...            ...   \n",
       "17103                    serine/threonine protein kinase            271   \n",
       "17104            response regulator transcription factor            200   \n",
       "17105                ABC transporter ATP-binding protein            331   \n",
       "17106                ABC transporter ATP-binding protein            238   \n",
       "17107                                 response regulator            236   \n",
       "\n",
       "       t_protein_len  Unnamed: 0_y  protein_match  Jaccard_Score  \\\n",
       "0                187             5              1           1.00   \n",
       "1                244             9              1           0.75   \n",
       "2                269            11              0           0.00   \n",
       "3                124            18              1           1.00   \n",
       "4                275            20              1           0.40   \n",
       "...              ...           ...            ...            ...   \n",
       "17103            353          4875              0           0.25   \n",
       "17104            203          4876              1           1.00   \n",
       "17105            338          4877              0           0.25   \n",
       "17106            235          4878              1           0.50   \n",
       "17107            211          4879              0           0.25   \n",
       "\n",
       "       structure_match  \n",
       "0                    1  \n",
       "1                    1  \n",
       "2                    0  \n",
       "3                    1  \n",
       "4                    1  \n",
       "...                ...  \n",
       "17103                0  \n",
       "17104                1  \n",
       "17105                0  \n",
       "17106                1  \n",
       "17107                0  \n",
       "\n",
       "[17108 rows x 34 columns]"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "337f87aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>m_protein_seq</th>\n",
       "      <th>t_protein_seq</th>\n",
       "      <th>prot_pair_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MVTTGERQRNARGEGARLRLEIVAATQALLADGETATLRSIARRAG...</td>\n",
       "      <td>MNRPTYHHGDLRAAILTEAARLVAERGAERVSLRELAREAGVSHAA...</td>\n",
       "      <td>20752996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MQIKDSVAVVTGGASGLGLATTKRLLDAGGSVVVIDLKGEDVVAEL...</td>\n",
       "      <td>MTGTVVITGGSRGIGAATAVLAAERGWQVAVSFRERRDAAEQVVRR...</td>\n",
       "      <td>85102366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MRAIGELWPTFDDVHEIAVLRGGGLGDLMFALPAIDALAAAYPEAR...</td>\n",
       "      <td>MATTTAEIGVIGGSGFYSFLDDPHEVTVQTPYGPPSDPIAVGTVAG...</td>\n",
       "      <td>37128170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MHELSIAQSVVEAVCERADGRRVRSVRMVVGALCAVVPDSMLFCFE...</td>\n",
       "      <td>MGLCEAIVDAVLQRAEGRRVRSVRVRVAGHPVVREVVDQGFAMAAA...</td>\n",
       "      <td>166997534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MKKAVLEVRDVTKRVKGRTLLSHMSFSIYKNEICGFVGPNGSGKTT...</td>\n",
       "      <td>MNETHFGARDLSLAYGKKRIIEALNIDIPIGYITALIGRNGSGKST...</td>\n",
       "      <td>157332201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17103</th>\n",
       "      <td>MDNATFRLGDDLSVRLPGHSRWIGQVEREQRWLPWLAPRLPLTVST...</td>\n",
       "      <td>MPPQPRPLRPNDPREIGGFALLGRLGEGGQGTVYLGGAPDGRRVAV...</td>\n",
       "      <td>78849058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17104</th>\n",
       "      <td>MFRTGVKAEIGRSLAVVGEAEDVERAVRVVLEQRPDVVLLDVHLPG...</td>\n",
       "      <td>MILEAEPDIVVVGEAGDGEKAVEEARALQPDVVLMDIRMPRKDGVE...</td>\n",
       "      <td>108797464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17105</th>\n",
       "      <td>MIRLAELTKTYPGQQHPAVDGISMEVAEGEIVVLVGPSGCGKTTTL...</td>\n",
       "      <td>MTEQPILSARGLTVDFRLRGGRRARAVDGVDLDLAPGEVLALAGES...</td>\n",
       "      <td>161110219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17106</th>\n",
       "      <td>MSESHAGALLSVRGLTAGYGGATALDGVSLTVAAGETVALLGANGA...</td>\n",
       "      <td>MSLLTTSGLTRHFSGIHAVEGVDFTLEAGEIRALIGSNGAGKTTLV...</td>\n",
       "      <td>74177185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17107</th>\n",
       "      <td>MTGQGASGSETGVSDDSPITVVVVDDQELLRTGLRDLAEHDGDIAV...</td>\n",
       "      <td>MTTEAQPRRVVLAEDEALIRLDLAEMLAEEGYEVVGQAADGETALA...</td>\n",
       "      <td>143737378</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17108 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           m_protein_seq  \\\n",
       "0      MVTTGERQRNARGEGARLRLEIVAATQALLADGETATLRSIARRAG...   \n",
       "1      MQIKDSVAVVTGGASGLGLATTKRLLDAGGSVVVIDLKGEDVVAEL...   \n",
       "2      MRAIGELWPTFDDVHEIAVLRGGGLGDLMFALPAIDALAAAYPEAR...   \n",
       "3      MHELSIAQSVVEAVCERADGRRVRSVRMVVGALCAVVPDSMLFCFE...   \n",
       "4      MKKAVLEVRDVTKRVKGRTLLSHMSFSIYKNEICGFVGPNGSGKTT...   \n",
       "...                                                  ...   \n",
       "17103  MDNATFRLGDDLSVRLPGHSRWIGQVEREQRWLPWLAPRLPLTVST...   \n",
       "17104  MFRTGVKAEIGRSLAVVGEAEDVERAVRVVLEQRPDVVLLDVHLPG...   \n",
       "17105  MIRLAELTKTYPGQQHPAVDGISMEVAEGEIVVLVGPSGCGKTTTL...   \n",
       "17106  MSESHAGALLSVRGLTAGYGGATALDGVSLTVAAGETVALLGANGA...   \n",
       "17107  MTGQGASGSETGVSDDSPITVVVVDDQELLRTGLRDLAEHDGDIAV...   \n",
       "\n",
       "                                           t_protein_seq  prot_pair_index  \n",
       "0      MNRPTYHHGDLRAAILTEAARLVAERGAERVSLRELAREAGVSHAA...         20752996  \n",
       "1      MTGTVVITGGSRGIGAATAVLAAERGWQVAVSFRERRDAAEQVVRR...         85102366  \n",
       "2      MATTTAEIGVIGGSGFYSFLDDPHEVTVQTPYGPPSDPIAVGTVAG...         37128170  \n",
       "3      MGLCEAIVDAVLQRAEGRRVRSVRVRVAGHPVVREVVDQGFAMAAA...        166997534  \n",
       "4      MNETHFGARDLSLAYGKKRIIEALNIDIPIGYITALIGRNGSGKST...        157332201  \n",
       "...                                                  ...              ...  \n",
       "17103  MPPQPRPLRPNDPREIGGFALLGRLGEGGQGTVYLGGAPDGRRVAV...         78849058  \n",
       "17104  MILEAEPDIVVVGEAGDGEKAVEEARALQPDVVLMDIRMPRKDGVE...        108797464  \n",
       "17105  MTEQPILSARGLTVDFRLRGGRRARAVDGVDLDLAPGEVLALAGES...        161110219  \n",
       "17106  MSLLTTSGLTRHFSGIHAVEGVDFTLEAGEIRALIGSNGAGKTTLV...         74177185  \n",
       "17107  MTTEAQPRRVVLAEDEALIRLDLAEMLAEEGYEVVGQAADGETALA...        143737378  \n",
       "\n",
       "[17108 rows x 3 columns]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_seqs = df[['m_protein_seq', 't_protein_seq', 'prot_pair_index']]\n",
    "df_seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "084d9561",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get rid of stuff that isn't quantitative\n",
    "\n",
    "df = df.drop(columns = ['Unnamed: 0_x', 'Unnamed: 0_y','thermo_index', 'm_protein_seq', 't_protein_seq',\n",
    "                        'm_protein_desc', 't_protein_desc'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "eea27e6c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['local_gap_compressed_percent_id', 'scaled_local_query_percent_id',\n",
       "       'scaled_local_symmetric_percent_id', 'query_align_len',\n",
       "       'query_align_cov', 'subject_align_len', 'subject_align_cov',\n",
       "       'bit_score', 'meso_index', 'prot_pair_index', 'meso_protein_int_index',\n",
       "       'thermo_protein_int_index', 'taxa_pair_index',\n",
       "       'local_gap_compressed_percent_id_16s',\n",
       "       'scaled_local_query_percent_id_16s',\n",
       "       'scaled_local_symmetric_percent_id_16s', 'query_align_cov_16s',\n",
       "       'subject_align_cov_16s', 'bit_score_16s', 'm_ogt', 't_ogt',\n",
       "       'ogt_difference', 'm_protein_len', 't_protein_len', 'protein_match',\n",
       "       'Jaccard_Score', 'structure_match'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07cc3a80",
   "metadata": {},
   "source": [
    "From pairplot below, it looks like query_align_cov_16s and subject_align_cov_16s dont correlate with target. Let's remove them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "36d8dadb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# sns.pairplot(df.sample(500),height = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "91079edd",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.drop(columns = ['query_align_cov_16s', 'subject_align_cov_16s'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "7929625e",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>meso_index</th>\n",
       "      <th>prot_pair_index</th>\n",
       "      <th>...</th>\n",
       "      <th>scaled_local_symmetric_percent_id_16s</th>\n",
       "      <th>bit_score_16s</th>\n",
       "      <th>m_ogt</th>\n",
       "      <th>t_ogt</th>\n",
       "      <th>ogt_difference</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "      <th>protein_match</th>\n",
       "      <th>Jaccard_Score</th>\n",
       "      <th>structure_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>17108.000000</td>\n",
       "      <td>17108.000000</td>\n",
       "      <td>17108.000000</td>\n",
       "      <td>17108.000000</td>\n",
       "      <td>17108.000000</td>\n",
       "      <td>17108.000000</td>\n",
       "      <td>17108.000000</td>\n",
       "      <td>17108.000000</td>\n",
       "      <td>17108.000000</td>\n",
       "      <td>1.710800e+04</td>\n",
       "      <td>...</td>\n",
       "      <td>17108.000000</td>\n",
       "      <td>17108.000000</td>\n",
       "      <td>17108.000000</td>\n",
       "      <td>17108.000000</td>\n",
       "      <td>17108.000000</td>\n",
       "      <td>17108.000000</td>\n",
       "      <td>17108.000000</td>\n",
       "      <td>17108.000000</td>\n",
       "      <td>17108.000000</td>\n",
       "      <td>17108.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>0.335028</td>\n",
       "      <td>0.253711</td>\n",
       "      <td>0.246046</td>\n",
       "      <td>199.578677</td>\n",
       "      <td>0.761698</td>\n",
       "      <td>198.636369</td>\n",
       "      <td>0.731170</td>\n",
       "      <td>204.140402</td>\n",
       "      <td>8156.811667</td>\n",
       "      <td>9.144065e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>0.904827</td>\n",
       "      <td>1060.437807</td>\n",
       "      <td>27.391454</td>\n",
       "      <td>51.547288</td>\n",
       "      <td>24.155834</td>\n",
       "      <td>277.208090</td>\n",
       "      <td>263.709843</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.454811</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.081731</td>\n",
       "      <td>0.099072</td>\n",
       "      <td>0.101347</td>\n",
       "      <td>59.651100</td>\n",
       "      <td>0.145737</td>\n",
       "      <td>59.862260</td>\n",
       "      <td>0.190915</td>\n",
       "      <td>156.853579</td>\n",
       "      <td>4594.459364</td>\n",
       "      <td>5.221655e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>0.015290</td>\n",
       "      <td>75.675183</td>\n",
       "      <td>3.687852</td>\n",
       "      <td>3.783911</td>\n",
       "      <td>3.593249</td>\n",
       "      <td>62.927629</td>\n",
       "      <td>68.167674</td>\n",
       "      <td>0.500015</td>\n",
       "      <td>0.340119</td>\n",
       "      <td>0.500015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.174274</td>\n",
       "      <td>0.081301</td>\n",
       "      <td>0.045977</td>\n",
       "      <td>21.000000</td>\n",
       "      <td>0.284553</td>\n",
       "      <td>21.000000</td>\n",
       "      <td>0.104558</td>\n",
       "      <td>35.000000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>8.858000e+03</td>\n",
       "      <td>...</td>\n",
       "      <td>0.880026</td>\n",
       "      <td>891.000000</td>\n",
       "      <td>7.000000</td>\n",
       "      <td>41.000000</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>30.000000</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>0.289855</td>\n",
       "      <td>0.193427</td>\n",
       "      <td>0.189189</td>\n",
       "      <td>166.000000</td>\n",
       "      <td>0.638779</td>\n",
       "      <td>165.000000</td>\n",
       "      <td>0.612091</td>\n",
       "      <td>108.000000</td>\n",
       "      <td>4255.000000</td>\n",
       "      <td>4.693841e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>0.893394</td>\n",
       "      <td>1008.000000</td>\n",
       "      <td>27.500000</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>22.000000</td>\n",
       "      <td>239.000000</td>\n",
       "      <td>227.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>0.318182</td>\n",
       "      <td>0.233645</td>\n",
       "      <td>0.228994</td>\n",
       "      <td>207.000000</td>\n",
       "      <td>0.763974</td>\n",
       "      <td>207.000000</td>\n",
       "      <td>0.756434</td>\n",
       "      <td>177.000000</td>\n",
       "      <td>8160.000000</td>\n",
       "      <td>9.196549e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>0.903417</td>\n",
       "      <td>1055.000000</td>\n",
       "      <td>28.000000</td>\n",
       "      <td>52.500000</td>\n",
       "      <td>24.500000</td>\n",
       "      <td>270.000000</td>\n",
       "      <td>260.000000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.320513</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>0.352941</td>\n",
       "      <td>0.282258</td>\n",
       "      <td>0.273333</td>\n",
       "      <td>231.000000</td>\n",
       "      <td>0.890869</td>\n",
       "      <td>229.250000</td>\n",
       "      <td>0.891032</td>\n",
       "      <td>244.000000</td>\n",
       "      <td>12103.000000</td>\n",
       "      <td>1.361882e+08</td>\n",
       "      <td>...</td>\n",
       "      <td>0.914642</td>\n",
       "      <td>1107.000000</td>\n",
       "      <td>30.000000</td>\n",
       "      <td>52.500000</td>\n",
       "      <td>24.500000</td>\n",
       "      <td>325.000000</td>\n",
       "      <td>314.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>0.985075</td>\n",
       "      <td>0.985075</td>\n",
       "      <td>0.985075</td>\n",
       "      <td>398.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>398.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1750.000000</td>\n",
       "      <td>16345.000000</td>\n",
       "      <td>1.814866e+08</td>\n",
       "      <td>...</td>\n",
       "      <td>0.996439</td>\n",
       "      <td>1679.000000</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>78.000000</td>\n",
       "      <td>56.000000</td>\n",
       "      <td>400.000000</td>\n",
       "      <td>400.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       local_gap_compressed_percent_id  scaled_local_query_percent_id  \\\n",
       "count                     17108.000000                   17108.000000   \n",
       "mean                          0.335028                       0.253711   \n",
       "std                           0.081731                       0.099072   \n",
       "min                           0.174274                       0.081301   \n",
       "25%                           0.289855                       0.193427   \n",
       "50%                           0.318182                       0.233645   \n",
       "75%                           0.352941                       0.282258   \n",
       "max                           0.985075                       0.985075   \n",
       "\n",
       "       scaled_local_symmetric_percent_id  query_align_len  query_align_cov  \\\n",
       "count                       17108.000000     17108.000000     17108.000000   \n",
       "mean                            0.246046       199.578677         0.761698   \n",
       "std                             0.101347        59.651100         0.145737   \n",
       "min                             0.045977        21.000000         0.284553   \n",
       "25%                             0.189189       166.000000         0.638779   \n",
       "50%                             0.228994       207.000000         0.763974   \n",
       "75%                             0.273333       231.000000         0.890869   \n",
       "max                             0.985075       398.000000         1.000000   \n",
       "\n",
       "       subject_align_len  subject_align_cov     bit_score    meso_index  \\\n",
       "count       17108.000000       17108.000000  17108.000000  17108.000000   \n",
       "mean          198.636369           0.731170    204.140402   8156.811667   \n",
       "std            59.862260           0.190915    156.853579   4594.459364   \n",
       "min            21.000000           0.104558     35.000000      5.000000   \n",
       "25%           165.000000           0.612091    108.000000   4255.000000   \n",
       "50%           207.000000           0.756434    177.000000   8160.000000   \n",
       "75%           229.250000           0.891032    244.000000  12103.000000   \n",
       "max           398.000000           1.000000   1750.000000  16345.000000   \n",
       "\n",
       "       prot_pair_index  ...  scaled_local_symmetric_percent_id_16s  \\\n",
       "count     1.710800e+04  ...                           17108.000000   \n",
       "mean      9.144065e+07  ...                               0.904827   \n",
       "std       5.221655e+07  ...                               0.015290   \n",
       "min       8.858000e+03  ...                               0.880026   \n",
       "25%       4.693841e+07  ...                               0.893394   \n",
       "50%       9.196549e+07  ...                               0.903417   \n",
       "75%       1.361882e+08  ...                               0.914642   \n",
       "max       1.814866e+08  ...                               0.996439   \n",
       "\n",
       "       bit_score_16s         m_ogt         t_ogt  ogt_difference  \\\n",
       "count   17108.000000  17108.000000  17108.000000    17108.000000   \n",
       "mean     1060.437807     27.391454     51.547288       24.155834   \n",
       "std        75.675183      3.687852      3.783911        3.593249   \n",
       "min       891.000000      7.000000     41.000000       20.000000   \n",
       "25%      1008.000000     27.500000     50.000000       22.000000   \n",
       "50%      1055.000000     28.000000     52.500000       24.500000   \n",
       "75%      1107.000000     30.000000     52.500000       24.500000   \n",
       "max      1679.000000     40.000000     78.000000       56.000000   \n",
       "\n",
       "       m_protein_len  t_protein_len  protein_match  Jaccard_Score  \\\n",
       "count   17108.000000   17108.000000   17108.000000   17108.000000   \n",
       "mean      277.208090     263.709843       0.500000       0.454811   \n",
       "std        62.927629      68.167674       0.500015       0.340119   \n",
       "min        30.000000      29.000000       0.000000       0.000000   \n",
       "25%       239.000000     227.000000       0.000000       0.200000   \n",
       "50%       270.000000     260.000000       0.500000       0.320513   \n",
       "75%       325.000000     314.000000       1.000000       0.750000   \n",
       "max       400.000000     400.000000       1.000000       1.000000   \n",
       "\n",
       "       structure_match  \n",
       "count     17108.000000  \n",
       "mean          0.500000  \n",
       "std           0.500015  \n",
       "min           0.000000  \n",
       "25%           0.000000  \n",
       "50%           0.500000  \n",
       "75%           1.000000  \n",
       "max           1.000000  \n",
       "\n",
       "[8 rows x 25 columns]"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90871d01",
   "metadata": {},
   "source": [
    "Split data into dev and test, and then split that into train and validation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "6519dcaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#drop columns that don't exihibit signficant pearson correlation with bit_score\n",
    "\n",
    "df = df.drop(columns = ['meso_index', 'meso_protein_int_index', 'local_gap_compressed_percent_id_16s', \n",
    "                        'scaled_local_query_percent_id_16s', 'scaled_local_symmetric_percent_id_16s',\n",
    "                       'bit_score_16s', 'm_ogt', 't_ogt', 'taxa_pair_index', 'thermo_protein_int_index'\n",
    "                       , 'prot_pair_index', 'ogt_difference'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "b3b9a7c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.drop(columns = ['Jaccard_Score'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "113e6608",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['local_gap_compressed_percent_id', 'scaled_local_query_percent_id',\n",
       "       'scaled_local_symmetric_percent_id', 'query_align_len',\n",
       "       'query_align_cov', 'subject_align_len', 'subject_align_cov',\n",
       "       'bit_score', 'm_protein_len', 't_protein_len', 'protein_match',\n",
       "       'structure_match'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "af9f23cf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(14541, 12)\n",
      "(2567, 12)\n",
      "(12359, 12)\n",
      "(2182, 12)\n"
     ]
    }
   ],
   "source": [
    "#choosing 80/20 split instead of 85/15 because of volume of data\n",
    "\n",
    "dev, test = sklearn.model_selection.train_test_split(df, test_size=0.15, random_state=1)\n",
    "\n",
    "train, val = sklearn.model_selection.train_test_split(dev, test_size=0.15, random_state=1)\n",
    "\n",
    "print(dev.shape)\n",
    "print(test.shape)\n",
    "print(train.shape)\n",
    "print(val.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "8470550d",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#ID target and features, separate into separate arrays\n",
    "\n",
    "target = [columns for columns in df[['protein_match', 'structure_match']]]\n",
    "input_features = [columns for columns in df.drop(columns=['protein_match', 'structure_match'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "73c0a63f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['local_gap_compressed_percent_id',\n",
       " 'scaled_local_query_percent_id',\n",
       " 'scaled_local_symmetric_percent_id',\n",
       " 'query_align_len',\n",
       " 'query_align_cov',\n",
       " 'subject_align_len',\n",
       " 'subject_align_cov',\n",
       " 'bit_score',\n",
       " 'm_protein_len',\n",
       " 't_protein_len']"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "1ab62bff",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(14541, 10) (2567, 10) (14541, 2) (2567, 2)\n"
     ]
    }
   ],
   "source": [
    "#split X and y\n",
    "\n",
    "dev_X = dev[input_features].values\n",
    "test_X = test[input_features].values\n",
    "\n",
    "dev_y = dev[target].values\n",
    "test_y = test[target].values\n",
    "\n",
    "print(dev_X.shape, test_X.shape, dev_y.shape, test_y.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "9ac52ac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#same thing for training and validation data\n",
    "\n",
    "train_X = train[input_features].values\n",
    "val_X = val[input_features].values\n",
    "\n",
    "train_y = train[target].values\n",
    "val_y = val[target].values"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c374f2d7",
   "metadata": {},
   "source": [
    "Scale the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "2b48b2e5",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "scaler = sklearn.preprocessing.StandardScaler()\n",
    "dev_X = scaler.fit_transform(dev_X)\n",
    "test_X = scaler.fit_transform(test_X)\n",
    "train_X = scaler.fit_transform(train_X)\n",
    "val_X = scaler.fit_transform(val_X)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3022371",
   "metadata": {},
   "source": [
    "Train the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "be4e09b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-1 {color: black;background-color: white;}#sk-container-id-1 pre{padding: 0;}#sk-container-id-1 div.sk-toggleable {background-color: white;}#sk-container-id-1 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-1 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-1 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-1 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-1 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-1 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-1 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-1 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-1 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-1 div.sk-item {position: relative;z-index: 1;}#sk-container-id-1 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-1 div.sk-item::before, #sk-container-id-1 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-1 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-1 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-1 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-1 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-1 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-1 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-1 div.sk-label-container {text-align: center;}#sk-container-id-1 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-1 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>RandomForestClassifier(oob_score=True)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" checked><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">RandomForestClassifier</label><div class=\"sk-toggleable__content\"><pre>RandomForestClassifier(oob_score=True)</pre></div></div></div></div></div>"
      ],
      "text/plain": [
       "RandomForestClassifier(oob_score=True)"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Random Forest\n",
    "\n",
    "model = sklearn.ensemble.RandomForestClassifier(oob_score=True)\n",
    "model.fit(dev_X, dev_y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "c96904eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Obtain predicted class probabilities\n",
    "y_pred_prob = model.predict_proba(dev_X)\n",
    "\n",
    "# Compute uncertainty estimates using out-of-bag samples\n",
    "oob_predictions = model.oob_decision_function_\n",
    "\n",
    "uncertainty = np.mean(np.var(oob_predictions, axis=1), axis=0)\n",
    "\n",
    "# 'y_pred_prob' contains predicted class probabilities, and 'uncertainty' represents the uncertainty estimates\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "b64e62db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0.09550174, 0.09550174])"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uncertainty"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "b87faf14",
   "metadata": {},
   "outputs": [],
   "source": [
    "import joblib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "5ad183bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['trained_model.pkl']"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "joblib.dump(model, 'trained_model.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "3e12d31d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-2 {color: black;background-color: white;}#sk-container-id-2 pre{padding: 0;}#sk-container-id-2 div.sk-toggleable {background-color: white;}#sk-container-id-2 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-2 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-2 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-2 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-2 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-2 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-2 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-2 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-container-id-2 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-2 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-2 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-2 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-2 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-2 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-2 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-2 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-2 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-2 div.sk-item {position: relative;z-index: 1;}#sk-container-id-2 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-2 div.sk-item::before, #sk-container-id-2 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-2 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-2 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-2 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-2 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-2 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-2 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-2 div.sk-label-container {text-align: center;}#sk-container-id-2 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-2 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-2\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>RandomForestClassifier(oob_score=True)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-2\" type=\"checkbox\" checked><label for=\"sk-estimator-id-2\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">RandomForestClassifier</label><div class=\"sk-toggleable__content\"><pre>RandomForestClassifier(oob_score=True)</pre></div></div></div></div></div>"
      ],
      "text/plain": [
       "RandomForestClassifier(oob_score=True)"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "joblib.load('trained_model.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69ebf269",
   "metadata": {},
   "source": [
    "Test the model, report relevant statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "c73327e9",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model score is: 0.7635372029606544\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "2567"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "score = model.score(test_X, test_y)\n",
    "print('Model score is: {}'.format(score))\n",
    "\n",
    "preds = model.predict(test_X)\n",
    "len(preds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "669f962e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1, 1],\n",
       "       [0, 0],\n",
       "       [0, 0],\n",
       "       ...,\n",
       "       [0, 0],\n",
       "       [0, 0],\n",
       "       [0, 0]])"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "7f62dbd9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy: 0.7635372029606544\n",
      "Precision: [0.80278019 0.80278019]\n",
      "Recall: [0.70858896 0.70858896]\n",
      "F1 Score: [0.75274949 0.75274949]\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score\n",
    "\n",
    "true_labels = test_y\n",
    "\n",
    "predicted_labels = preds\n",
    "\n",
    "# Calculate accuracy\n",
    "accuracy = accuracy_score(true_labels, predicted_labels)\n",
    "\n",
    "# Calculate precision\n",
    "precision = precision_score(true_labels, predicted_labels, average=None)\n",
    "\n",
    "# Calculate recall\n",
    "recall = recall_score(true_labels, predicted_labels, average=None)\n",
    "\n",
    "# Calculate F1 score\n",
    "f1 = f1_score(true_labels, predicted_labels, average=None)\n",
    "\n",
    "# Print the scores\n",
    "print(\"Accuracy:\", accuracy)\n",
    "print(\"Precision:\", precision)\n",
    "print(\"Recall:\", recall)\n",
    "print(\"F1 Score:\", f1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0ab9c8b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
