{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c693f524",
   "metadata": {},
   "source": [
    "### Overview\n",
    "\n",
    "This notebook contains cells for intake of new l2t database 5/2023 as well as edits to the original preprocessing component for building FAFSA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b5a9ff3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "\n",
    "import duckdb\n",
    "import pandas as pd\n",
    "\n",
    "import os\n",
    "\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fcece608",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Paths for parquet files from extracted database\n",
    "folder = '/mnt/s/FAFSA'\n",
    "\n",
    "prot_pair_path = folder + '/protein_pairs'\n",
    "prot_path = folder + '/proteins'\n",
    "taxa_pair_al_path = folder + '/taxa_pairs/alignment'\n",
    "taxa_pair_lab_path = folder + '/taxa_pairs/pair_labels'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "068fc729",
   "metadata": {},
   "source": [
    "### Example of protein pairs parquet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "8f152740",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>thermo_pid</th>\n",
       "      <th>meso_pid</th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>local_E_value</th>\n",
       "      <th>query_align_start</th>\n",
       "      <th>query_align_end</th>\n",
       "      <th>subject_align_end</th>\n",
       "      <th>subject_align_start</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>thermo_taxid</th>\n",
       "      <th>meso_taxid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E8R444</td>\n",
       "      <td>A0A5B9W4X6</td>\n",
       "      <td>0.540000</td>\n",
       "      <td>0.457627</td>\n",
       "      <td>0.482143</td>\n",
       "      <td>2.290000e-34</td>\n",
       "      <td>15</td>\n",
       "      <td>114</td>\n",
       "      <td>101</td>\n",
       "      <td>2</td>\n",
       "      <td>100</td>\n",
       "      <td>0.847458</td>\n",
       "      <td>100</td>\n",
       "      <td>0.943396</td>\n",
       "      <td>276.0</td>\n",
       "      <td>128</td>\n",
       "      <td>406548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E8R3P9</td>\n",
       "      <td>A0A5B9W2H7</td>\n",
       "      <td>0.569444</td>\n",
       "      <td>0.445652</td>\n",
       "      <td>0.465909</td>\n",
       "      <td>5.510000e-29</td>\n",
       "      <td>18</td>\n",
       "      <td>92</td>\n",
       "      <td>84</td>\n",
       "      <td>14</td>\n",
       "      <td>75</td>\n",
       "      <td>0.815217</td>\n",
       "      <td>71</td>\n",
       "      <td>0.845238</td>\n",
       "      <td>236.0</td>\n",
       "      <td>128</td>\n",
       "      <td>406548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E8R0N8</td>\n",
       "      <td>A0A5B9W0K1</td>\n",
       "      <td>0.469231</td>\n",
       "      <td>0.308081</td>\n",
       "      <td>0.346591</td>\n",
       "      <td>3.140000e-24</td>\n",
       "      <td>11</td>\n",
       "      <td>184</td>\n",
       "      <td>141</td>\n",
       "      <td>15</td>\n",
       "      <td>174</td>\n",
       "      <td>0.878788</td>\n",
       "      <td>127</td>\n",
       "      <td>0.824675</td>\n",
       "      <td>220.0</td>\n",
       "      <td>128</td>\n",
       "      <td>406548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E8R1F9</td>\n",
       "      <td>A0A5B9W1N8</td>\n",
       "      <td>0.715190</td>\n",
       "      <td>0.624309</td>\n",
       "      <td>0.631285</td>\n",
       "      <td>1.290000e-81</td>\n",
       "      <td>15</td>\n",
       "      <td>172</td>\n",
       "      <td>166</td>\n",
       "      <td>9</td>\n",
       "      <td>158</td>\n",
       "      <td>0.872928</td>\n",
       "      <td>158</td>\n",
       "      <td>0.892655</td>\n",
       "      <td>600.0</td>\n",
       "      <td>128</td>\n",
       "      <td>406548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E8QWA5</td>\n",
       "      <td>A0A5B9W5Q2</td>\n",
       "      <td>0.580000</td>\n",
       "      <td>0.471545</td>\n",
       "      <td>0.505447</td>\n",
       "      <td>3.960000e-69</td>\n",
       "      <td>44</td>\n",
       "      <td>245</td>\n",
       "      <td>207</td>\n",
       "      <td>7</td>\n",
       "      <td>202</td>\n",
       "      <td>0.821138</td>\n",
       "      <td>201</td>\n",
       "      <td>0.943662</td>\n",
       "      <td>528.0</td>\n",
       "      <td>128</td>\n",
       "      <td>406548</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  thermo_pid    meso_pid  local_gap_compressed_percent_id  \\\n",
       "0     E8R444  A0A5B9W4X6                         0.540000   \n",
       "1     E8R3P9  A0A5B9W2H7                         0.569444   \n",
       "2     E8R0N8  A0A5B9W0K1                         0.469231   \n",
       "3     E8R1F9  A0A5B9W1N8                         0.715190   \n",
       "4     E8QWA5  A0A5B9W5Q2                         0.580000   \n",
       "\n",
       "   scaled_local_query_percent_id  scaled_local_symmetric_percent_id  \\\n",
       "0                       0.457627                           0.482143   \n",
       "1                       0.445652                           0.465909   \n",
       "2                       0.308081                           0.346591   \n",
       "3                       0.624309                           0.631285   \n",
       "4                       0.471545                           0.505447   \n",
       "\n",
       "   local_E_value  query_align_start  query_align_end  subject_align_end  \\\n",
       "0   2.290000e-34                 15              114                101   \n",
       "1   5.510000e-29                 18               92                 84   \n",
       "2   3.140000e-24                 11              184                141   \n",
       "3   1.290000e-81                 15              172                166   \n",
       "4   3.960000e-69                 44              245                207   \n",
       "\n",
       "   subject_align_start  query_align_len  query_align_cov  subject_align_len  \\\n",
       "0                    2              100         0.847458                100   \n",
       "1                   14               75         0.815217                 71   \n",
       "2                   15              174         0.878788                127   \n",
       "3                    9              158         0.872928                158   \n",
       "4                    7              202         0.821138                201   \n",
       "\n",
       "   subject_align_cov  bit_score  thermo_taxid  meso_taxid  \n",
       "0           0.943396      276.0           128      406548  \n",
       "1           0.845238      236.0           128      406548  \n",
       "2           0.824675      220.0           128      406548  \n",
       "3           0.892655      600.0           128      406548  \n",
       "4           0.943662      528.0           128      406548  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "duckdb.read_parquet(prot_pair_path + '/align_taxa_128-406548.parquet').df().head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50135c6b",
   "metadata": {},
   "source": [
    "### Example of proteins parquet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "9d46701d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pid</th>\n",
       "      <th>taxid</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>alphafold_id</th>\n",
       "      <th>proteome</th>\n",
       "      <th>protein_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A2A5R0G1</td>\n",
       "      <td>373386</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A0A2A5R0G1</td>\n",
       "      <td>UP000219689</td>\n",
       "      <td>MLADLLSESYATDLEESWENERTATPVRAFAVRLHQTGCSLRETTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A2A5R046</td>\n",
       "      <td>373386</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A0A2A5R046</td>\n",
       "      <td>UP000219689</td>\n",
       "      <td>MPENDRLNGCLDEINLEFVEREATPRLLMKLSIQLHLAGLSLSNTV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A1H7MLA0</td>\n",
       "      <td>302484</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A0A1H7MLA0</td>\n",
       "      <td>UP000183894</td>\n",
       "      <td>MGIVSSKTQALQEVASVDDFLNVAATETVPLFEHLEFEFLLEYDVF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E1REH3</td>\n",
       "      <td>54120</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UP000006565</td>\n",
       "      <td>MVKGFDHSYAFFLGCIAPNRYPGCEASAIRTSAKLGIELLPLKGAS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E1REH4</td>\n",
       "      <td>54120</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E1REH4</td>\n",
       "      <td>UP000006565</td>\n",
       "      <td>MAVEKNYGNPDLEKKLADRNYYTSDSHKDFSKRVEKISGTMSHMCF...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          pid   taxid pdb_id alphafold_id     proteome  \\\n",
       "0  A0A2A5R0G1  373386    NaN   A0A2A5R0G1  UP000219689   \n",
       "1  A0A2A5R046  373386    NaN   A0A2A5R046  UP000219689   \n",
       "2  A0A1H7MLA0  302484    NaN   A0A1H7MLA0  UP000183894   \n",
       "3      E1REH3   54120    NaN          NaN  UP000006565   \n",
       "4      E1REH4   54120    NaN       E1REH4  UP000006565   \n",
       "\n",
       "                                         protein_seq  \n",
       "0  MLADLLSESYATDLEESWENERTATPVRAFAVRLHQTGCSLRETTT...  \n",
       "1  MPENDRLNGCLDEINLEFVEREATPRLLMKLSIQLHLAGLSLSNTV...  \n",
       "2  MGIVSSKTQALQEVASVDDFLNVAATETVPLFEHLEFEFLLEYDVF...  \n",
       "3  MVKGFDHSYAFFLGCIAPNRYPGCEASAIRTSAKLGIELLPLKGAS...  \n",
       "4  MAVEKNYGNPDLEKKLADRNYYTSDSHKDFSKRVEKISGTMSHMCF...  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "duckdb.read_parquet(prot_path + '/uniprot_chunk_0.parquet').df().head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11454f0e",
   "metadata": {},
   "source": [
    "### Example of taxa pairs alignment parquet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "98918332",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>local_E_value</th>\n",
       "      <th>query_align_start</th>\n",
       "      <th>query_align_end</th>\n",
       "      <th>subject_align_end</th>\n",
       "      <th>subject_align_start</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>__index_level_0__</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1281767</td>\n",
       "      <td>1413211</td>\n",
       "      <td>0.971045</td>\n",
       "      <td>0.865324</td>\n",
       "      <td>0.892278</td>\n",
       "      <td>0.0</td>\n",
       "      <td>178</td>\n",
       "      <td>1589</td>\n",
       "      <td>1486</td>\n",
       "      <td>73</td>\n",
       "      <td>1412</td>\n",
       "      <td>0.888609</td>\n",
       "      <td>1414</td>\n",
       "      <td>0.947086</td>\n",
       "      <td>1283.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1281767</td>\n",
       "      <td>545502</td>\n",
       "      <td>0.930840</td>\n",
       "      <td>0.830082</td>\n",
       "      <td>0.867763</td>\n",
       "      <td>0.0</td>\n",
       "      <td>178</td>\n",
       "      <td>1589</td>\n",
       "      <td>1437</td>\n",
       "      <td>29</td>\n",
       "      <td>1412</td>\n",
       "      <td>0.888609</td>\n",
       "      <td>1409</td>\n",
       "      <td>0.971054</td>\n",
       "      <td>1104.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1281767</td>\n",
       "      <td>157733</td>\n",
       "      <td>0.927311</td>\n",
       "      <td>0.826935</td>\n",
       "      <td>0.847469</td>\n",
       "      <td>0.0</td>\n",
       "      <td>178</td>\n",
       "      <td>1589</td>\n",
       "      <td>1493</td>\n",
       "      <td>80</td>\n",
       "      <td>1412</td>\n",
       "      <td>0.888609</td>\n",
       "      <td>1414</td>\n",
       "      <td>0.935185</td>\n",
       "      <td>1096.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1281767</td>\n",
       "      <td>86665</td>\n",
       "      <td>0.926606</td>\n",
       "      <td>0.826306</td>\n",
       "      <td>0.847644</td>\n",
       "      <td>0.0</td>\n",
       "      <td>178</td>\n",
       "      <td>1589</td>\n",
       "      <td>1491</td>\n",
       "      <td>80</td>\n",
       "      <td>1412</td>\n",
       "      <td>0.888609</td>\n",
       "      <td>1412</td>\n",
       "      <td>0.935719</td>\n",
       "      <td>1095.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1281767</td>\n",
       "      <td>79885</td>\n",
       "      <td>0.926606</td>\n",
       "      <td>0.826306</td>\n",
       "      <td>0.848465</td>\n",
       "      <td>0.0</td>\n",
       "      <td>178</td>\n",
       "      <td>1589</td>\n",
       "      <td>1486</td>\n",
       "      <td>73</td>\n",
       "      <td>1412</td>\n",
       "      <td>0.888609</td>\n",
       "      <td>1414</td>\n",
       "      <td>0.938911</td>\n",
       "      <td>1091.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  query_id subject_id  local_gap_compressed_percent_id  \\\n",
       "0  1281767    1413211                         0.971045   \n",
       "1  1281767     545502                         0.930840   \n",
       "2  1281767     157733                         0.927311   \n",
       "3  1281767      86665                         0.926606   \n",
       "4  1281767      79885                         0.926606   \n",
       "\n",
       "   scaled_local_query_percent_id  scaled_local_symmetric_percent_id  \\\n",
       "0                       0.865324                           0.892278   \n",
       "1                       0.830082                           0.867763   \n",
       "2                       0.826935                           0.847469   \n",
       "3                       0.826306                           0.847644   \n",
       "4                       0.826306                           0.848465   \n",
       "\n",
       "   local_E_value  query_align_start  query_align_end  subject_align_end  \\\n",
       "0            0.0                178             1589               1486   \n",
       "1            0.0                178             1589               1437   \n",
       "2            0.0                178             1589               1493   \n",
       "3            0.0                178             1589               1491   \n",
       "4            0.0                178             1589               1486   \n",
       "\n",
       "   subject_align_start  query_align_len  query_align_cov  subject_align_len  \\\n",
       "0                   73             1412         0.888609               1414   \n",
       "1                   29             1412         0.888609               1409   \n",
       "2                   80             1412         0.888609               1414   \n",
       "3                   80             1412         0.888609               1412   \n",
       "4                   73             1412         0.888609               1414   \n",
       "\n",
       "   subject_align_cov  bit_score  __index_level_0__  \n",
       "0           0.947086     1283.0                  0  \n",
       "1           0.971054     1104.0                  1  \n",
       "2           0.935185     1096.0                  2  \n",
       "3           0.935719     1095.0                  3  \n",
       "4           0.938911     1091.0                  4  "
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "duckdb.read_parquet(taxa_pair_al_path + '/taxa_pair_blast_chunk_0.parquet').df().head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "326decf2",
   "metadata": {},
   "source": [
    "### Example of taxa pairs label parquet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "725c6ea2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>is_pair</th>\n",
       "      <th>__index_level_0__</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   is_pair  __index_level_0__\n",
       "0    False                  0\n",
       "1    False                  1\n",
       "2    False                  2\n",
       "3    False                  3\n",
       "4    False                  4"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "duckdb.read_parquet(taxa_pair_lab_path + '/taxa_pair_blast_chunk_0.parquet').df().head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94f18045",
   "metadata": {},
   "source": [
    "### Uniprot IDs file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b72d99b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "uniprot = pd.read_csv(folder + '/uniprot/proteome_metadata.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e17e2eea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>pid</th>\n",
       "      <th>species_taxid</th>\n",
       "      <th>strain_taxid</th>\n",
       "      <th>qualifier</th>\n",
       "      <th>completeness</th>\n",
       "      <th>num_proteins</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>UP000464621</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>Other proteome</td>\n",
       "      <td>Outlier (low value)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>UP000325048</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>Redundant proteome</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>1778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>UP000321374</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>Other proteome</td>\n",
       "      <td>Close to standard (low value)</td>\n",
       "      <td>1125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>UP000095230</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>Other proteome</td>\n",
       "      <td>Standard</td>\n",
       "      <td>3900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>UP000252468</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>Other proteome</td>\n",
       "      <td>Close to standard (high value)</td>\n",
       "      <td>4310</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0          pid  species_taxid  strain_taxid           qualifier  \\\n",
       "0           0  UP000464621              9             9      Other proteome   \n",
       "1           1  UP000325048             14            14  Redundant proteome   \n",
       "2           2  UP000321374             17            17      Other proteome   \n",
       "3           3  UP000095230             23            23      Other proteome   \n",
       "4           4  UP000252468             24            24      Other proteome   \n",
       "\n",
       "                     completeness  num_proteins  \n",
       "0             Outlier (low value)             2  \n",
       "1                         Unknown          1778  \n",
       "2   Close to standard (low value)          1125  \n",
       "3                        Standard          3900  \n",
       "4  Close to standard (high value)          4310  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniprot.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be6a8ebd",
   "metadata": {},
   "source": [
    "### Example of taxa parquet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "60bff6ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxid</th>\n",
       "      <th>16s_seq</th>\n",
       "      <th>16s_len</th>\n",
       "      <th>temperature</th>\n",
       "      <th>superkingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>__index_level_0__</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>617123</td>\n",
       "      <td>GATGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGAAGCT...</td>\n",
       "      <td>1594</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1239.0</td>\n",
       "      <td>186801.0</td>\n",
       "      <td>186802.0</td>\n",
       "      <td>186803.0</td>\n",
       "      <td>1164882.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1281767</td>\n",
       "      <td>GCTCAGGACGAACGCTGGCGGCGTGCCTAATACATGCAAGTCGAAT...</td>\n",
       "      <td>1589</td>\n",
       "      <td>50.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1239.0</td>\n",
       "      <td>91061.0</td>\n",
       "      <td>1385.0</td>\n",
       "      <td>186817.0</td>\n",
       "      <td>1494427.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1329262</td>\n",
       "      <td>CAGAGTTTGATCCTGGCTCAGGATGAACGCTGGCGGCGTGCTTAAC...</td>\n",
       "      <td>1589</td>\n",
       "      <td>30.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1239.0</td>\n",
       "      <td>186801.0</td>\n",
       "      <td>186802.0</td>\n",
       "      <td>186803.0</td>\n",
       "      <td>1780122.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>402877</td>\n",
       "      <td>GACGAACGCTGGCGGCGTGCCTAACACATGCAAGTCGAACGGACTA...</td>\n",
       "      <td>1589</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1239.0</td>\n",
       "      <td>186801.0</td>\n",
       "      <td>186802.0</td>\n",
       "      <td>68298.0</td>\n",
       "      <td>862.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1485586</td>\n",
       "      <td>GCTGCCCTTCAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGC...</td>\n",
       "      <td>1588</td>\n",
       "      <td>28.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>201174.0</td>\n",
       "      <td>1760.0</td>\n",
       "      <td>85011.0</td>\n",
       "      <td>2062.0</td>\n",
       "      <td>1883.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     taxid                                            16s_seq  16s_len  \\\n",
       "0   617123  GATGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGAAGCT...     1594   \n",
       "1  1281767  GCTCAGGACGAACGCTGGCGGCGTGCCTAATACATGCAAGTCGAAT...     1589   \n",
       "2  1329262  CAGAGTTTGATCCTGGCTCAGGATGAACGCTGGCGGCGTGCTTAAC...     1589   \n",
       "3   402877  GACGAACGCTGGCGGCGTGCCTAACACATGCAAGTCGAACGGACTA...     1589   \n",
       "4  1485586  GCTGCCCTTCAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGC...     1588   \n",
       "\n",
       "   temperature  superkingdom    phylum     class     order    family  \\\n",
       "0         37.0           2.0    1239.0  186801.0  186802.0  186803.0   \n",
       "1         50.0           2.0    1239.0   91061.0    1385.0  186817.0   \n",
       "2         30.0           2.0    1239.0  186801.0  186802.0  186803.0   \n",
       "3         37.0           2.0    1239.0  186801.0  186802.0   68298.0   \n",
       "4         28.0           2.0  201174.0    1760.0   85011.0    2062.0   \n",
       "\n",
       "       genus  __index_level_0__  \n",
       "0  1164882.0                  0  \n",
       "1  1494427.0                  1  \n",
       "2  1780122.0                  2  \n",
       "3      862.0                  3  \n",
       "4     1883.0                  4  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "duckdb.read_parquet(folder + '/taxa.parquet').df().head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08da72f4",
   "metadata": {},
   "source": [
    "### Build learn2therm database\n",
    "\n",
    "This code took about 1.5 hr to run on my home desktop with 8c/16t and 32 Gb of memory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2ddaf784",
   "metadata": {},
   "outputs": [],
   "source": [
    "con = duckdb.connect(folder+'/FAFSA.db')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "044cfd1f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "dd470934223a422c99f513ba867dcb9c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "FloatProgress(value=0.0, layout=Layout(width='100%'), style=ProgressStyle(bar_color='black'))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6043.670356133 seconds\n"
     ]
    }
   ],
   "source": [
    "# Protein pairs\n",
    "all_pairs = prot_pair_path + '/*.parquet'\n",
    "\n",
    "t1 = time.monotonic()\n",
    "\n",
    "con.execute(f\"\"\"CREATE OR REPLACE TABLE protein_pairs AS SELECT * FROM '{all_pairs}'\"\"\")\n",
    "\n",
    "elapsed = time.monotonic() - t1\n",
    "\n",
    "print(f'{elapsed} seconds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7672daa7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5ee4570603f54f3fa8ef274e638d7934",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "FloatProgress(value=0.0, layout=Layout(width='100%'), style=ProgressStyle(bar_color='black'))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "64.46746825300033 seconds\n"
     ]
    }
   ],
   "source": [
    "# Proteins\n",
    "all_prot = prot_path + '/*.parquet'\n",
    "\n",
    "t1 = time.monotonic()\n",
    "\n",
    "con.execute(f\"\"\"CREATE OR REPLACE TABLE proteins AS SELECT * FROM '{all_prot}'\"\"\")\n",
    "\n",
    "elapsed = time.monotonic() - t1\n",
    "\n",
    "print(f'{elapsed} seconds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6f138ae7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8b24f947ad7f400cb2f7ff2119926892",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "FloatProgress(value=0.0, layout=Layout(width='100%'), style=ProgressStyle(bar_color='black'))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5.828336216000025 seconds\n"
     ]
    }
   ],
   "source": [
    "# Taxa pair al\n",
    "\n",
    "all_taxa_pair_al = taxa_pair_al_path + '/*.parquet'\n",
    "\n",
    "t1 = time.monotonic()\n",
    "\n",
    "con.execute(f\"\"\"CREATE OR REPLACE TABLE taxa_pairs_al AS SELECT * FROM '{all_taxa_pair_al}'\"\"\")\n",
    "\n",
    "elapsed = time.monotonic() - t1\n",
    "\n",
    "print(f'{elapsed} seconds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1b360433",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.7024612730001536 seconds\n"
     ]
    }
   ],
   "source": [
    "# Taxa pair lab\n",
    "\n",
    "all_taxa_pair_lab = taxa_pair_lab_path + '/*.parquet'\n",
    "\n",
    "t1 = time.monotonic()\n",
    "\n",
    "con.execute(f\"\"\"CREATE OR REPLACE TABLE taxa_pairs_lab AS SELECT * FROM '{all_taxa_pair_lab}'\"\"\")\n",
    "\n",
    "elapsed = time.monotonic() - t1\n",
    "\n",
    "print(f'{elapsed} seconds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8d729c73",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.6066281989988056 seconds\n"
     ]
    }
   ],
   "source": [
    "# Taxa \n",
    "\n",
    "all_taxa = folder + '/taxa.parquet'\n",
    "\n",
    "t1 = time.monotonic()\n",
    "\n",
    "con.execute(f\"\"\"CREATE OR REPLACE TABLE taxa AS SELECT * FROM '{all_taxa}'\"\"\")\n",
    "\n",
    "elapsed = time.monotonic() - t1\n",
    "\n",
    "print(f'{elapsed} seconds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "411dfebb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count(thermo_pid)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>82174541</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   count(thermo_pid)\n",
       "0           82174541"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "con.execute(\"\"\"SELECT COUNT(thermo_pid) FROM protein_pairs\"\"\").df()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3bbf667c",
   "metadata": {},
   "source": [
    "*Total pairs in l2t is now 82 million* "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ffaa16a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<duckdb.DuckDBPyConnection at 0x7fdba99fc870>"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Build uniprot table (not necessary)\n",
    "uniprot_path = folder + '/uniprot/proteome_metadata.csv'\n",
    "con.execute(f\"\"\"CREATE OR REPLACE TABLE uniprot AS SELECT * FROM '{uniprot_path}'\"\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fa738a64",
   "metadata": {},
   "outputs": [],
   "source": [
    "con.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b213f4a",
   "metadata": {},
   "source": [
    "### Build FAFSA database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f27280bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Updated version of original function in preprocessing\n",
    "def connect_db(path: str):\n",
    "    '''\n",
    "    Runs duckdb.connect() function on database path. Returns a\n",
    "    duckdb.DuckDBPyConnection object and prints execution time.\n",
    "\n",
    "    Args:\n",
    "        path (str): Path to DuckDB database file containing learn2therm.\n",
    "\n",
    "    Returns:\n",
    "        con (duckdb.DuckDBPyConnection): A DuckDB connection object linking script to\n",
    "        learn2therm database.\n",
    "\n",
    "    Raises:\n",
    "        VersionError: DuckDB installation is not one of 0.7.0 or 0.7.1.\n",
    "        AttributeError: Input database contains no tables.\n",
    "    '''\n",
    "    s_time = time.time()\n",
    "\n",
    "    version = duckdb.__version__\n",
    "    print(f'DuckDB version {version} detected.')\n",
    "\n",
    "    # Checks for compatible installation of duckdb.\n",
    "    if not version in ['0.7.0', '0.7.1']:\n",
    "        raise VersionError(\"\"\"learn2therm was generated using DuckDB storage version 39. It is only\n",
    "                           compatible with duckdb versions 0.6.0 and 0.6.1. Please check your\n",
    "                           installation. Refer to https://duckdb.org/internals/storage.html for more\n",
    "                           details.\"\"\")\n",
    "\n",
    "        e_time = time.time()\n",
    "        elapsed_time = e_time - s_time\n",
    "        print(f'Finished with VersionError. Execution time: {elapsed_time} seconds')\n",
    "\n",
    "    print('Connecting to database...')\n",
    "    con = duckdb.connect(path)\n",
    "\n",
    "    tables = con.execute(\"\"\"SELECT TABLE_NAME\n",
    "                            FROM INFORMATION_SCHEMA.TABLES\n",
    "                            WHERE TABLE_TYPE='BASE TABLE'\"\"\").df()\n",
    "\n",
    "    if tables.shape[0] < 1:\n",
    "        raise AttributeError('Input database is empty.')\n",
    "\n",
    "    e_time = time.time()\n",
    "    elapsed_time = e_time - s_time\n",
    "    print(f'Connection established! Execution time: {elapsed_time} seconds')\n",
    "    return con, tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8bfec85d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DuckDB version 0.7.1 detected.\n",
      "Connecting to database...\n",
      "Connection established! Execution time: 0.44511961936950684 seconds\n"
     ]
    }
   ],
   "source": [
    "con, tables = connect_db('/mnt/s/FAFSA/FAFSA.db')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6331e960",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>table_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>fafsa_final</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>fafsa_proteins</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>fafsa_protein_pairs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>fafsa_ogt_taxa_pairs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>fafsa_taxa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>fafsa_taxa_pairs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>taxa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>taxa_pairs_lab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>taxa_pairs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>proteins</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>protein_pairs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              table_name\n",
       "0            fafsa_final\n",
       "1         fafsa_proteins\n",
       "2    fafsa_protein_pairs\n",
       "3   fafsa_ogt_taxa_pairs\n",
       "4             fafsa_taxa\n",
       "5       fafsa_taxa_pairs\n",
       "6                   taxa\n",
       "7         taxa_pairs_lab\n",
       "8             taxa_pairs\n",
       "9               proteins\n",
       "10         protein_pairs"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Should include fafsa tables now that construction has been done once\n",
    "tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "723a88c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<duckdb.DuckDBPyConnection at 0x7fbaeccd8ab0>"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Fix table name for consistency with original database\n",
    "con.execute(\"\"\"ALTER TABLE taxa_pairs_al RENAME TO taxa_pairs\"\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "271b59a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Updated from original preprocessing function\n",
    "def build_fafsa(con, min_ogt_diff: int = 20, min_16s: int = 1300,\n",
    "                    plots: bool = False):\n",
    "    '''\n",
    "    Converts learn2therm DuckDB database into a DuckDB database for FAFSA by adding filtered and\n",
    "    constructed tables.\n",
    "\n",
    "    Args:\n",
    "        con (duckdb.DuckDBPyConnection): DuckDB connection object. Links script to DuckDB SQL \n",
    "        database.\n",
    "        min_ogt_diff (int): Cutoff for minimum difference in optimal growth temperature between \n",
    "        thermophile and mesophile pairs. Default 20 deg C.\n",
    "        min_16s (int): Cutoff for minimum 16S read length for taxa. Default 1300 bp. Filters out \n",
    "        organisms with poor or incomplete 16S sequencing.\n",
    "        plots (bool): Boolean to determine whether the user wants Sankey plots diagramming the fate\n",
    "        of learn2therm samples during FAFSA construction to be saved in ./plots.\n",
    "\n",
    "    Returns:\n",
    "        None. Database object is modified in place.\n",
    "\n",
    "    Raises:\n",
    "        ValueError: Optimal growth temperature difference must be positive.\n",
    "        ValueError: Minimum 16S sequence read is 1 bp.\n",
    "        AttributeError: Database must be in the learn2therm format.\n",
    "    '''\n",
    "\n",
    "    if min_ogt_diff < 0:\n",
    "        raise ValueError('Optimal growth temperature difference must be positive.')\n",
    "\n",
    "    if min_16s < 1:\n",
    "        raise ValueError('16S must have at least 1 bp read.')\n",
    "\n",
    "    tables = con.execute(\"\"\"SELECT TABLE_NAME\n",
    "                            FROM INFORMATION_SCHEMA.TABLES\n",
    "                            WHERE TABLE_TYPE='BASE TABLE'\"\"\").df()\n",
    "\n",
    "    # Check if proper tables exist in database. If they do not, raise an error.\n",
    "    if (item in tables for item in ['proteins', 'protein_pairs', 'taxa', 'taxa_pairs']):\n",
    "        pass\n",
    "\n",
    "    else:\n",
    "        raise AttributeError('Database is not formatted for learn2therm.')\n",
    "\n",
    "    s_time = time.time()\n",
    "    print('Constructing fafsa_taxa_pairs...')\n",
    "\n",
    "    # Builds FAFSA taxa pair table using only paired taxa from learn2therm\n",
    "    taxa_pairs_cmd = \"\"\"CREATE OR REPLACE TABLE fafsa_taxa_pairs AS\n",
    "                        SELECT *\n",
    "                        FROM taxa_pairs\n",
    "                        INNER JOIN taxa_pairs_lab ON (taxa_pairs.__index_level_0__ = taxa_pairs_lab.__index_level_0__)\n",
    "                        WHERE taxa_pairs_lab.is_pair = True\"\"\"\n",
    "    con.execute(taxa_pairs_cmd)\n",
    "\n",
    "    e_time = time.time()\n",
    "    elapsed_time = e_time - s_time\n",
    "    print(f'Finished constructing fafsa_taxa_pairs. Execution time: {elapsed_time} seconds')\n",
    "    print('Constructing fafsa_taxa...')\n",
    "\n",
    "    # Builds FAFSA taxa table using only paired taxa from learn2therm.\n",
    "    taxa_cmd = f\"\"\"CREATE OR REPLACE TABLE fafsa_taxa AS\n",
    "                   SELECT *\n",
    "                   FROM taxa\n",
    "                   WHERE taxid IN \n",
    "                   (SELECT DISTINCT query_id FROM fafsa_taxa_pairs)\n",
    "                   OR taxid IN\n",
    "                   (SELECT DISTINCT subject_id FROM fafsa_taxa_pairs)\"\"\"\n",
    "    con.execute(taxa_cmd)\n",
    "\n",
    "    e_time2 = time.time()\n",
    "    elapsed_time = e_time2 - e_time\n",
    "    print(f'Finished constructing fafsa_taxa. Execution time: {elapsed_time} seconds')\n",
    "    print('Filtering on ogt and 16S sequence parameters...')\n",
    "\n",
    "    # Builds FAFSA table containing taxa pairs and their associated optimal growth temperatures\n",
    "    # (ogt). Excludes 16S sequences and ogt difference below cutoff values from function input.\n",
    "    ogt_pairs_cmd = f\"\"\"CREATE OR REPLACE TABLE fafsa_ogt_taxa_pairs AS SELECT fafsa_taxa_pairs.*,\n",
    "                        taxa_m.temperature AS meso_ogt,\n",
    "                        taxa_t.temperature AS thermo_ogt,\n",
    "                        taxa_t.temperature - taxa_m.temperature AS ogt_diff,\n",
    "                        taxa_m.\"16s_len\" AS meso_16s_len,\n",
    "                        taxa_t.\"16s_len\" AS thermo_16s_len\n",
    "                        FROM fafsa_taxa_pairs\n",
    "                        JOIN fafsa_taxa AS taxa_m ON (fafsa_taxa_pairs.subject_id = taxa_m.taxid)\n",
    "                        JOIN fafsa_taxa AS taxa_t ON (fafsa_taxa_pairs.query_id = taxa_t.taxid)\n",
    "                        WHERE ogt_diff >= {min_ogt_diff}\n",
    "                        AND meso_16s_len >= {min_16s}\n",
    "                        AND thermo_16s_len >= {min_16s}\"\"\"\n",
    "    con.execute(ogt_pairs_cmd)\n",
    "\n",
    "    e_time3 = time.time()\n",
    "    elapsed_time = e_time3 - e_time2\n",
    "    print(f'Finished filtering. Execution time: {elapsed_time} seconds')\n",
    "    print('Constructing fafsa_protein_pairs...')\n",
    "\n",
    "    # Builds FAFSA table containing protein pairs\n",
    "    protein_pair_cmd = \"\"\"CREATE OR REPLACE TABLE fafsa_protein_pairs AS\n",
    "                          SELECT \n",
    "                          protein_pairs.meso_pid,\n",
    "                          protein_pairs.thermo_pid,\n",
    "                          protein_pairs.bit_score,\n",
    "                          protein_pairs.local_gap_compressed_percent_id,\n",
    "                          protein_pairs.scaled_local_query_percent_id,\n",
    "                          protein_pairs.scaled_local_symmetric_percent_id,\n",
    "                          protein_pairs.query_align_len,\n",
    "                          protein_pairs.query_align_cov,\n",
    "                          protein_pairs.subject_align_len,\n",
    "                          protein_pairs.subject_align_cov,\n",
    "                          otp.meso_ogt AS m_ogt,\n",
    "                          otp.thermo_ogt AS t_ogt,\n",
    "                          otp.ogt_diff AS ogt_difference\n",
    "                          FROM protein_pairs\n",
    "                          INNER JOIN fafsa_ogt_taxa_pairs AS otp\n",
    "                          ON (protein_pairs.thermo_taxid = otp.query_id)\n",
    "                          AND (protein_pairs.meso_taxid = otp.subject_id)\"\"\"\n",
    "    con.execute(protein_pair_cmd)\n",
    "\n",
    "    e_time4 = time.time()\n",
    "    elapsed_time = e_time4 - e_time3\n",
    "    print(f'Finished constructing fafsa_protein_pairs. Execution time: {elapsed_time} seconds')\n",
    "    print('Constructing fafsa_proteins...')\n",
    "\n",
    "    # Builds FAFSA table containing proteins that belong to taxa from fafsa_taxa_pairs.\n",
    "    prot_filt_cmd = \"\"\"CREATE OR REPLACE TABLE fafsa_proteins AS SELECT *\n",
    "                       FROM proteins\n",
    "                       WHERE pid IN (SELECT DISTINCT meso_pid FROM fafsa_protein_pairs) OR\n",
    "                       pid IN (SELECT DISTINCT thermo_pid FROM fafsa_protein_pairs)\n",
    "                    \"\"\"\n",
    "    con.execute(prot_filt_cmd)\n",
    "\n",
    "    e_time5 = time.time()\n",
    "    elapsed_time = e_time5 - e_time4\n",
    "    print(f'Finished constructing fafsa_proteins. Execution time: {elapsed_time} seconds')\n",
    "    print('Constructing final dataset...')\n",
    "\n",
    "    # Builds final FAFSA data table for downstream sampling.\n",
    "    big_table_cmd = \"\"\"CREATE OR REPLACE TABLE fafsa_final AS\n",
    "                       SELECT \n",
    "                       proteins_m.protein_seq AS m_protein_seq,\n",
    "                       proteins_t.protein_seq AS t_protein_seq,\n",
    "                       proteins_m.alphafold_id AS meso_alphafold_id,\n",
    "                       proteins_t.alphafold_id AS thermo_alphafold_id,\n",
    "                       fafsa_protein_pairs.*,\n",
    "                       LENGTH(proteins_m.protein_seq) AS m_protein_len,\n",
    "                       LENGTH(proteins_t.protein_seq) AS t_protein_len\n",
    "                       FROM fafsa_protein_pairs\n",
    "                       JOIN fafsa_proteins AS proteins_m\n",
    "                       ON (fafsa_protein_pairs.meso_pid = proteins_m.pid)\n",
    "                       JOIN fafsa_proteins AS proteins_t\n",
    "                       ON (fafsa_protein_pairs.thermo_pid =\n",
    "                           proteins_t.pid)\"\"\"\n",
    "    con.execute(big_table_cmd)\n",
    "    print('Finishing up...')\n",
    "    con.commit()\n",
    "    con.close()\n",
    "\n",
    "    et_final = time.time()\n",
    "    elapsed_time = et_final - e_time5\n",
    "    print(f'Finished. Total execution time: {elapsed_time} seconds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "60f05f0d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructing fafsa_taxa_pairs...\n",
      "Finished constructing fafsa_taxa_pairs. Execution time: 0.4976813793182373 seconds\n",
      "Constructing fafsa_taxa...\n",
      "Finished constructing fafsa_taxa. Execution time: 0.3858222961425781 seconds\n",
      "Filtering on ogt and 16S sequence parameters...\n",
      "Finished filtering. Execution time: 1.0931081771850586 seconds\n",
      "Constructing fafsa_protein_pairs...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "182e354875264ecc875df63e368d7699",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "FloatProgress(value=0.0, layout=Layout(width='100%'), style=ProgressStyle(bar_color='black'))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished constructing fafsa_protein_pairs. Execution time: 10.692678689956665 seconds\n",
      "Constructing fafsa_proteins...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0bd218167958446799a624a5e21e761b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "FloatProgress(value=0.0, layout=Layout(width='100%'), style=ProgressStyle(bar_color='black'))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished constructing fafsa_proteins. Execution time: 12.06689977645874 seconds\n",
      "Constructing final dataset...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "994cd8014e3e4ea6a5a40355025132b7",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "FloatProgress(value=0.0, layout=Layout(width='100%'), style=ProgressStyle(bar_color='black'))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finishing up...\n",
      "Finished. Total execution time: 65.12362670898438 seconds\n"
     ]
    }
   ],
   "source": [
    "build_fafsa(con)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "18d4b5b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DuckDB version 0.7.1 detected.\n",
      "Connecting to database...\n",
      "Connection established! Execution time: 0.27312803268432617 seconds\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>m_protein_seq</th>\n",
       "      <th>t_protein_seq</th>\n",
       "      <th>meso_alphafold_id</th>\n",
       "      <th>thermo_alphafold_id</th>\n",
       "      <th>meso_pid</th>\n",
       "      <th>thermo_pid</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>m_ogt</th>\n",
       "      <th>t_ogt</th>\n",
       "      <th>ogt_difference</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MQKRILIVDDEPAIRDMVAFALRKGEYEPVHAGDALEAQTAIADRV...</td>\n",
       "      <td>MRILLVEDEAPLRETLAARLKREGFAVDAAQDGEEGLYMGREVPFD...</td>\n",
       "      <td>A0A0R0CEE8</td>\n",
       "      <td>A0A562DKC7</td>\n",
       "      <td>A0A0R0CEE8</td>\n",
       "      <td>A0A562DKC7</td>\n",
       "      <td>320.0</td>\n",
       "      <td>0.349776</td>\n",
       "      <td>0.343612</td>\n",
       "      <td>0.342105</td>\n",
       "      <td>220</td>\n",
       "      <td>0.969163</td>\n",
       "      <td>223</td>\n",
       "      <td>0.973799</td>\n",
       "      <td>28.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>229</td>\n",
       "      <td>227</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       m_protein_seq  \\\n",
       "0  MQKRILIVDDEPAIRDMVAFALRKGEYEPVHAGDALEAQTAIADRV...   \n",
       "\n",
       "                                       t_protein_seq meso_alphafold_id  \\\n",
       "0  MRILLVEDEAPLRETLAARLKREGFAVDAAQDGEEGLYMGREVPFD...        A0A0R0CEE8   \n",
       "\n",
       "  thermo_alphafold_id    meso_pid  thermo_pid  bit_score  \\\n",
       "0          A0A562DKC7  A0A0R0CEE8  A0A562DKC7      320.0   \n",
       "\n",
       "   local_gap_compressed_percent_id  scaled_local_query_percent_id  \\\n",
       "0                         0.349776                       0.343612   \n",
       "\n",
       "   scaled_local_symmetric_percent_id  query_align_len  query_align_cov  \\\n",
       "0                           0.342105              220         0.969163   \n",
       "\n",
       "   subject_align_len  subject_align_cov  m_ogt  t_ogt  ogt_difference  \\\n",
       "0                223           0.973799   28.0   49.0            21.0   \n",
       "\n",
       "   m_protein_len  t_protein_len  \n",
       "0            229            227  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "con, tables = connect_db('/mnt/s/FAFSA/FAFSA.db')\n",
    "\n",
    "con.execute(\"\"\"SELECT * FROM fafsa_final LIMIT 1\"\"\").df()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0dd8a214",
   "metadata": {},
   "source": [
    "**Looks good**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76797d7a",
   "metadata": {},
   "source": [
    "### Build FAFSA_lite, a version of FAFSA with only 1 million pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "cfb22059",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "35d1332a984745f7a1e47ea2aed8bebb",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "FloatProgress(value=0.0, layout=Layout(width='100%'), style=ProgressStyle(bar_color='black'))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fafsa_lite = con.execute(\"\"\"SELECT * FROM fafsa_final USING SAMPLE 1000000\"\"\").df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "111cadec",
   "metadata": {},
   "outputs": [],
   "source": [
    "fafsa_lite.to_parquet('/mnt/s/FAFSA/fafsa_lite.parquet')\n",
    "con.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "e5ce9f55",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Connect to new database and create fafsa_final table using FAFSA_lite data\n",
    "con = duckdb.connect('/mnt/s/FAFSA/FAFSA_lite.db')\n",
    "\n",
    "con.execute(\"\"\"CREATE OR REPLACE TABLE fafsa_final AS SELECT * FROM '/mnt/s/FAFSA/fafsa_lite.parquet'\"\"\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
