{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6c1fdfb1",
   "metadata": {},
   "source": [
    "The purpose of this notebook is to document data exploration processes with an n=50,000 sample of our Learn2Therm protein database. Doing so provides a workspace to select and optimize a model for predicting protein pair functionality between two sequences. Several classifier architectures were tested prior to the release of the final version of this component."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84922e82",
   "metadata": {},
   "source": [
    "This notebook includes exploration of a Random Forest Classifier."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59b867a6",
   "metadata": {},
   "source": [
    "Problems:\n",
    "\n",
    "1) get_protein_descriptors returns a dictionary of dataframes. If one descriptor is passed, it just returns a dictionary. Need to ensure that output is a dataframe regardless of how many descriptors are passed\n",
    "2) Need to append dataframe as new descriptors are added, making function more robust\n",
    "3) Need to add code to deal with fasta format in get protein descriptors\n",
    "4) need to write unit tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "61c53eee",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats\n",
    "import sklearn.preprocessing\n",
    "import sklearn.model_selection\n",
    "import sklearn.ensemble\n",
    "import sklearn.feature_selection\n",
    "import unittest\n",
    "import iFeatureOmegaCLI\n",
    "import Bio.SeqIO\n",
    "import io\n",
    "from io import StringIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "ba87d47f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/loganroberts/Learn2Therm/ValidProt/FAFSA\n"
     ]
    }
   ],
   "source": [
    "cd /Users/loganroberts/Learn2Therm/ValidProt/FAFSA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "f90c1cdd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#convert to pandas df\n",
    "df = pd.read_csv('learn2therm_sample_50k.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "5d55b25f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/loganroberts/Learn2Therm/ValidProt/notebooks\n"
     ]
    }
   ],
   "source": [
    "cd /Users/loganroberts/Learn2Therm/ValidProt/notebooks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "49d634f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_fasta_from_dataframe(dataframe, output_file:str):\n",
    "    #adjust this to write function with BioPython\n",
    "\n",
    "    with open(output_file, 'w') as f:\n",
    "        for _, row in df.iterrows():\n",
    "            f.write('>{}\\n{}\\n'.format((row['meso_index']), row['m_protein_seq']))\n",
    "    return output_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "653eb96d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_protein_descriptors(fasta:str, descriptors=[]):\n",
    "    \n",
    "    \"\"\"\n",
    "    Generates features from a protein sequence\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    Fasta file with protein sequences.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    Vector of descriptors\n",
    "    \"\"\"\n",
    "    \n",
    "    #create iProtein object\n",
    "    protein = iFeatureOmegaCLI.iProtein(fasta)\n",
    "    \n",
    "    #not sure why we need this yet. Right now it is stored in local directory.\n",
    "    params = protein.import_parameters('protein_parameters.json')\n",
    "    \n",
    "    protein_descriptors = {}\n",
    "    \n",
    "    for descriptor in descriptors:\n",
    "        protein.get_descriptor(descriptor)\n",
    "        protein_descriptors.update({f'{descriptor}':protein.encodings})\n",
    "        \n",
    "    return protein_descriptors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "49cf5076",
   "metadata": {},
   "outputs": [],
   "source": [
    "#this function is only necessary if we get a fasta file with a descriptor generated from component 3\n",
    "\n",
    "def remove_fasta_description(filename:str):\n",
    "    \n",
    "    \"\"\"\n",
    "    Removes description from fasta file so that iProtein can read the input.\n",
    "    Might not be necessary in our current format.\n",
    "    \"\"\"\n",
    "    \n",
    "    #assign unwanted string to object\n",
    "    string_to_remove = \"<unknown description>\"\n",
    "\n",
    "    #open file\n",
    "    with open(filename, \"r\") as file:\n",
    "        content = file.read()\n",
    "    \n",
    "    # Remove the string\n",
    "    new_content = content.replace(string_to_remove, \"\")\n",
    "\n",
    "    #overwrite file without string\n",
    "    with open(filename, \"w\") as file:\n",
    "        seq = file.write(new_content)\n",
    "        \n",
    "    return seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "e93e310e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#combine both of the above functions\n",
    "\n",
    "def fasta_to_descriptors(fasta:str, descriptors=[]):\n",
    "\n",
    "    #remove description from fasta file\n",
    "    remove_fasta_description(fasta)\n",
    "    \n",
    "    #return protein descriptors\n",
    "    return get_protein_descriptors(fasta, descriptors=descriptors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "114c3d10",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_new_dataframe(dataframe, output_file, descriptors=[]):\n",
    "    \"\"\"\n",
    "    Creates new dataframe with descriptors added.\n",
    "    \"\"\"\n",
    "    \n",
    "    fasta = get_fasta_from_dataframe(dataframe, output_file)\n",
    "    \n",
    "    feature_dict = get_protein_descriptors(fasta, descriptors)\n",
    "    \n",
    "    df = dataframe.reset_index()\n",
    "    \n",
    "    for desc in descriptors:\n",
    "        \n",
    "        feature_dict[desc].index = feature_dict[desc].index.astype(int)\n",
    "        features = feature_dict[desc].reset_index()\n",
    "        \n",
    "        df = pd.merge(df, features, how='outer', left_index=True, right_index=True)\n",
    "        \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "29bdd03f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File imported successfully.\n"
     ]
    }
   ],
   "source": [
    "df = create_new_dataframe(df, 'meso_50k.fasta', descriptors=['AAC', 'GAAC'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a3ef17af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write the dataframe to a fasta file\n",
    "#adjust this to write function with BioPython\n",
    "\n",
    "with open('meso_50k.fasta', 'w') as f:\n",
    "    for i, row in df.iterrows():\n",
    "        f.write('>{}\\n{}\\n'.format((row['meso_index']), row['m_protein_seq']))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f0a9171",
   "metadata": {},
   "source": [
    "Let's do the same thing with the whole database:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "dda9aa47",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_l2t = get_protein_descriptors('meso_50k.fasta', descriptors=['AAC','GAAC'])['AAC']\n",
    "df_l2t.index = df_l2t.index.astype(int)\n",
    "df_l2t = df_l2t.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "fd4977c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "indexed_df = df.set_index(('meso_index'))\n",
    "indexed_df = indexed_df.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a10fc649",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>meso_index</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>...</th>\n",
       "      <th>AAC_M</th>\n",
       "      <th>AAC_N</th>\n",
       "      <th>AAC_P</th>\n",
       "      <th>AAC_Q</th>\n",
       "      <th>AAC_R</th>\n",
       "      <th>AAC_S</th>\n",
       "      <th>AAC_T</th>\n",
       "      <th>AAC_V</th>\n",
       "      <th>AAC_W</th>\n",
       "      <th>AAC_Y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12897</td>\n",
       "      <td>0</td>\n",
       "      <td>0.287582</td>\n",
       "      <td>0.217822</td>\n",
       "      <td>0.215686</td>\n",
       "      <td>160</td>\n",
       "      <td>0.792079</td>\n",
       "      <td>152</td>\n",
       "      <td>0.737864</td>\n",
       "      <td>131</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.004854</td>\n",
       "      <td>0.053398</td>\n",
       "      <td>0.038835</td>\n",
       "      <td>0.131068</td>\n",
       "      <td>0.033981</td>\n",
       "      <td>0.048544</td>\n",
       "      <td>0.106796</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.029126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13026</td>\n",
       "      <td>1</td>\n",
       "      <td>0.319635</td>\n",
       "      <td>0.295359</td>\n",
       "      <td>0.297872</td>\n",
       "      <td>218</td>\n",
       "      <td>0.919831</td>\n",
       "      <td>226</td>\n",
       "      <td>0.969957</td>\n",
       "      <td>282</td>\n",
       "      <td>...</td>\n",
       "      <td>0.055794</td>\n",
       "      <td>0.025751</td>\n",
       "      <td>0.025751</td>\n",
       "      <td>0.042918</td>\n",
       "      <td>0.090129</td>\n",
       "      <td>0.042918</td>\n",
       "      <td>0.047210</td>\n",
       "      <td>0.081545</td>\n",
       "      <td>0.004292</td>\n",
       "      <td>0.030043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8203</td>\n",
       "      <td>2</td>\n",
       "      <td>0.279621</td>\n",
       "      <td>0.234127</td>\n",
       "      <td>0.218924</td>\n",
       "      <td>211</td>\n",
       "      <td>0.837302</td>\n",
       "      <td>210</td>\n",
       "      <td>0.731707</td>\n",
       "      <td>96</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010453</td>\n",
       "      <td>0.020906</td>\n",
       "      <td>0.034843</td>\n",
       "      <td>0.013937</td>\n",
       "      <td>0.104530</td>\n",
       "      <td>0.031359</td>\n",
       "      <td>0.055749</td>\n",
       "      <td>0.114983</td>\n",
       "      <td>0.010453</td>\n",
       "      <td>0.017422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3340</td>\n",
       "      <td>3</td>\n",
       "      <td>0.327273</td>\n",
       "      <td>0.200743</td>\n",
       "      <td>0.214712</td>\n",
       "      <td>166</td>\n",
       "      <td>0.617100</td>\n",
       "      <td>163</td>\n",
       "      <td>0.696581</td>\n",
       "      <td>175</td>\n",
       "      <td>...</td>\n",
       "      <td>0.029915</td>\n",
       "      <td>0.008547</td>\n",
       "      <td>0.047009</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>0.051282</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.051282</td>\n",
       "      <td>0.094017</td>\n",
       "      <td>0.029915</td>\n",
       "      <td>0.025641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>14020</td>\n",
       "      <td>4</td>\n",
       "      <td>0.338710</td>\n",
       "      <td>0.318182</td>\n",
       "      <td>0.287671</td>\n",
       "      <td>60</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>71</td>\n",
       "      <td>0.887500</td>\n",
       "      <td>61</td>\n",
       "      <td>...</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.087500</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.025000</td>\n",
       "      <td>0.087500</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.062500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49995</th>\n",
       "      <td>8772</td>\n",
       "      <td>49995</td>\n",
       "      <td>0.344828</td>\n",
       "      <td>0.169972</td>\n",
       "      <td>0.192308</td>\n",
       "      <td>180</td>\n",
       "      <td>0.509915</td>\n",
       "      <td>200</td>\n",
       "      <td>0.738007</td>\n",
       "      <td>53</td>\n",
       "      <td>...</td>\n",
       "      <td>0.011070</td>\n",
       "      <td>0.014760</td>\n",
       "      <td>0.081181</td>\n",
       "      <td>0.022140</td>\n",
       "      <td>0.073801</td>\n",
       "      <td>0.036900</td>\n",
       "      <td>0.055351</td>\n",
       "      <td>0.081181</td>\n",
       "      <td>0.036900</td>\n",
       "      <td>0.007380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49996</th>\n",
       "      <td>1395</td>\n",
       "      <td>49996</td>\n",
       "      <td>0.417989</td>\n",
       "      <td>0.389163</td>\n",
       "      <td>0.392060</td>\n",
       "      <td>190</td>\n",
       "      <td>0.935961</td>\n",
       "      <td>187</td>\n",
       "      <td>0.935000</td>\n",
       "      <td>314</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010000</td>\n",
       "      <td>0.005000</td>\n",
       "      <td>0.025000</td>\n",
       "      <td>0.020000</td>\n",
       "      <td>0.120000</td>\n",
       "      <td>0.060000</td>\n",
       "      <td>0.030000</td>\n",
       "      <td>0.150000</td>\n",
       "      <td>0.005000</td>\n",
       "      <td>0.015000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49997</th>\n",
       "      <td>8513</td>\n",
       "      <td>49997</td>\n",
       "      <td>0.352174</td>\n",
       "      <td>0.239645</td>\n",
       "      <td>0.242152</td>\n",
       "      <td>230</td>\n",
       "      <td>0.680473</td>\n",
       "      <td>227</td>\n",
       "      <td>0.685801</td>\n",
       "      <td>279</td>\n",
       "      <td>...</td>\n",
       "      <td>0.015106</td>\n",
       "      <td>0.012085</td>\n",
       "      <td>0.060423</td>\n",
       "      <td>0.057402</td>\n",
       "      <td>0.063444</td>\n",
       "      <td>0.030211</td>\n",
       "      <td>0.057402</td>\n",
       "      <td>0.078550</td>\n",
       "      <td>0.009063</td>\n",
       "      <td>0.012085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49998</th>\n",
       "      <td>4295</td>\n",
       "      <td>49998</td>\n",
       "      <td>0.334764</td>\n",
       "      <td>0.331915</td>\n",
       "      <td>0.329810</td>\n",
       "      <td>232</td>\n",
       "      <td>0.987234</td>\n",
       "      <td>230</td>\n",
       "      <td>0.966387</td>\n",
       "      <td>281</td>\n",
       "      <td>...</td>\n",
       "      <td>0.016807</td>\n",
       "      <td>0.008403</td>\n",
       "      <td>0.046218</td>\n",
       "      <td>0.016807</td>\n",
       "      <td>0.096639</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.029412</td>\n",
       "      <td>0.088235</td>\n",
       "      <td>0.004202</td>\n",
       "      <td>0.008403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49999</th>\n",
       "      <td>106</td>\n",
       "      <td>49999</td>\n",
       "      <td>0.336207</td>\n",
       "      <td>0.184834</td>\n",
       "      <td>0.174497</td>\n",
       "      <td>113</td>\n",
       "      <td>0.535545</td>\n",
       "      <td>117</td>\n",
       "      <td>0.495763</td>\n",
       "      <td>165</td>\n",
       "      <td>...</td>\n",
       "      <td>0.025424</td>\n",
       "      <td>0.016949</td>\n",
       "      <td>0.042373</td>\n",
       "      <td>0.029661</td>\n",
       "      <td>0.101695</td>\n",
       "      <td>0.076271</td>\n",
       "      <td>0.055085</td>\n",
       "      <td>0.088983</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012712</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>50000 rows × 51 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       meso_index  Unnamed: 0  local_gap_compressed_percent_id  \\\n",
       "0           12897           0                         0.287582   \n",
       "1           13026           1                         0.319635   \n",
       "2            8203           2                         0.279621   \n",
       "3            3340           3                         0.327273   \n",
       "4           14020           4                         0.338710   \n",
       "...           ...         ...                              ...   \n",
       "49995        8772       49995                         0.344828   \n",
       "49996        1395       49996                         0.417989   \n",
       "49997        8513       49997                         0.352174   \n",
       "49998        4295       49998                         0.334764   \n",
       "49999         106       49999                         0.336207   \n",
       "\n",
       "       scaled_local_query_percent_id  scaled_local_symmetric_percent_id  \\\n",
       "0                           0.217822                           0.215686   \n",
       "1                           0.295359                           0.297872   \n",
       "2                           0.234127                           0.218924   \n",
       "3                           0.200743                           0.214712   \n",
       "4                           0.318182                           0.287671   \n",
       "...                              ...                                ...   \n",
       "49995                       0.169972                           0.192308   \n",
       "49996                       0.389163                           0.392060   \n",
       "49997                       0.239645                           0.242152   \n",
       "49998                       0.331915                           0.329810   \n",
       "49999                       0.184834                           0.174497   \n",
       "\n",
       "       query_align_len  query_align_cov  subject_align_len  subject_align_cov  \\\n",
       "0                  160         0.792079                152           0.737864   \n",
       "1                  218         0.919831                226           0.969957   \n",
       "2                  211         0.837302                210           0.731707   \n",
       "3                  166         0.617100                163           0.696581   \n",
       "4                   60         0.909091                 71           0.887500   \n",
       "...                ...              ...                ...                ...   \n",
       "49995              180         0.509915                200           0.738007   \n",
       "49996              190         0.935961                187           0.935000   \n",
       "49997              230         0.680473                227           0.685801   \n",
       "49998              232         0.987234                230           0.966387   \n",
       "49999              113         0.535545                117           0.495763   \n",
       "\n",
       "       bit_score  ...     AAC_M     AAC_N     AAC_P     AAC_Q     AAC_R  \\\n",
       "0            131  ...  0.009709  0.004854  0.053398  0.038835  0.131068   \n",
       "1            282  ...  0.055794  0.025751  0.025751  0.042918  0.090129   \n",
       "2             96  ...  0.010453  0.020906  0.034843  0.013937  0.104530   \n",
       "3            175  ...  0.029915  0.008547  0.047009  0.025641  0.051282   \n",
       "4             61  ...  0.050000  0.037500  0.050000  0.037500  0.087500   \n",
       "...          ...  ...       ...       ...       ...       ...       ...   \n",
       "49995         53  ...  0.011070  0.014760  0.081181  0.022140  0.073801   \n",
       "49996        314  ...  0.010000  0.005000  0.025000  0.020000  0.120000   \n",
       "49997        279  ...  0.015106  0.012085  0.060423  0.057402  0.063444   \n",
       "49998        281  ...  0.016807  0.008403  0.046218  0.016807  0.096639   \n",
       "49999        165  ...  0.025424  0.016949  0.042373  0.029661  0.101695   \n",
       "\n",
       "          AAC_S     AAC_T     AAC_V     AAC_W     AAC_Y  \n",
       "0      0.033981  0.048544  0.106796  0.009709  0.029126  \n",
       "1      0.042918  0.047210  0.081545  0.004292  0.030043  \n",
       "2      0.031359  0.055749  0.114983  0.010453  0.017422  \n",
       "3      0.055556  0.051282  0.094017  0.029915  0.025641  \n",
       "4      0.037500  0.025000  0.087500  0.037500  0.062500  \n",
       "...         ...       ...       ...       ...       ...  \n",
       "49995  0.036900  0.055351  0.081181  0.036900  0.007380  \n",
       "49996  0.060000  0.030000  0.150000  0.005000  0.015000  \n",
       "49997  0.030211  0.057402  0.078550  0.009063  0.012085  \n",
       "49998  0.071429  0.029412  0.088235  0.004202  0.008403  \n",
       "49999  0.076271  0.055085  0.088983  0.000000  0.012712  \n",
       "\n",
       "[50000 rows x 51 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df= pd.merge(indexed_df, df_l2t, how='outer', left_index=True, right_index=True)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95f6ec84",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b7be5ead",
   "metadata": {},
   "source": [
    "Scratch work for above function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d863dd63",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'meso_input_copy.fasta.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-16-531eaf9cc340>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mstring_to_remove\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"<unknown description>\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msequence\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"r\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mfile\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m     \u001b[0mcontent\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfile\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'meso_input_copy.fasta.txt'"
     ]
    }
   ],
   "source": [
    "sequence = \"meso_input_copy.fasta.txt\"\n",
    "string_to_remove = \"<unknown description>\"\n",
    "\n",
    "with open(sequence, \"r\") as file:\n",
    "    content = file.read()\n",
    "    \n",
    "# Remove the string\n",
    "new_content = content.replace(string_to_remove, \"\")\n",
    "\n",
    "with open(sequence, \"w\") as file:\n",
    "    file.write(new_content)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9a78f6e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#just figured out that iFeature cannot read the <unknown descriptor> part of the fasta sequence\n",
    "\n",
    "\n",
    "protein = iFeatureOmegaCLI.iProtein('meso_input_copy.fasta.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4981b0a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "protein.display_feature_types()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96d2c565",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df['t_protein_len'].describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04026501",
   "metadata": {},
   "source": [
    "The bit-score provides a better rule-of-thumb for inferring homology. For average length proteins, a bit score of 50 is almost always significant. A bit score of 40 is only significant (E() < 0.001) in searches of protein databases with fewer than 7000 entries. Increasing the score by 10 bits increases the significance 210=1000-fold, so 50 bits would be significant in a database with less than 7 million entries (10 times SwissProt, and within a factor of 3 of the largest protein databases). Thus, the NCBI Blast web site uses a color code of blue for alignment with scores between 40–50 bits; and green for scores between 50–80 bits. In the yeast vs human example, the alignments with less than 20% identity had scores ranging from 55 – 170 bits. Except for very long proteins and very large databases, 50 bits of similarity score will always be statistically significant and is a much better rule-of-thumb for inferring homology in protein alignments.\n",
    "\n",
    "Pearson et al., 2013: An Introduction to Sequence Similarity (“Homology”) Searching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "273884e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['protein_match'] = ((df['bit_score'] > 50 ) & (df['query_align_cov'] > 0.8))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "14015e54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True     25887\n",
       "False    24113\n",
       "Name: protein_match, dtype: int64"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['protein_match'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "b27836e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_x</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>...</th>\n",
       "      <th>AAC_V</th>\n",
       "      <th>AAC_W</th>\n",
       "      <th>AAC_Y</th>\n",
       "      <th>index</th>\n",
       "      <th>GAAC_alphatic</th>\n",
       "      <th>GAAC_aromatic</th>\n",
       "      <th>GAAC_postivecharge</th>\n",
       "      <th>GAAC_negativecharge</th>\n",
       "      <th>GAAC_uncharge</th>\n",
       "      <th>protein_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.287582</td>\n",
       "      <td>0.217822</td>\n",
       "      <td>0.215686</td>\n",
       "      <td>160</td>\n",
       "      <td>0.792079</td>\n",
       "      <td>152</td>\n",
       "      <td>0.737864</td>\n",
       "      <td>131</td>\n",
       "      <td>...</td>\n",
       "      <td>0.106796</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.029126</td>\n",
       "      <td>12897</td>\n",
       "      <td>0.451456</td>\n",
       "      <td>0.053398</td>\n",
       "      <td>0.165049</td>\n",
       "      <td>0.140777</td>\n",
       "      <td>0.189320</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.319635</td>\n",
       "      <td>0.295359</td>\n",
       "      <td>0.297872</td>\n",
       "      <td>218</td>\n",
       "      <td>0.919831</td>\n",
       "      <td>226</td>\n",
       "      <td>0.969957</td>\n",
       "      <td>282</td>\n",
       "      <td>...</td>\n",
       "      <td>0.081545</td>\n",
       "      <td>0.004292</td>\n",
       "      <td>0.030043</td>\n",
       "      <td>13026</td>\n",
       "      <td>0.412017</td>\n",
       "      <td>0.064378</td>\n",
       "      <td>0.154506</td>\n",
       "      <td>0.184549</td>\n",
       "      <td>0.184549</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0.279621</td>\n",
       "      <td>0.234127</td>\n",
       "      <td>0.218924</td>\n",
       "      <td>211</td>\n",
       "      <td>0.837302</td>\n",
       "      <td>210</td>\n",
       "      <td>0.731707</td>\n",
       "      <td>96</td>\n",
       "      <td>...</td>\n",
       "      <td>0.114983</td>\n",
       "      <td>0.010453</td>\n",
       "      <td>0.017422</td>\n",
       "      <td>8203</td>\n",
       "      <td>0.543554</td>\n",
       "      <td>0.059233</td>\n",
       "      <td>0.135889</td>\n",
       "      <td>0.097561</td>\n",
       "      <td>0.163763</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0.327273</td>\n",
       "      <td>0.200743</td>\n",
       "      <td>0.214712</td>\n",
       "      <td>166</td>\n",
       "      <td>0.617100</td>\n",
       "      <td>163</td>\n",
       "      <td>0.696581</td>\n",
       "      <td>175</td>\n",
       "      <td>...</td>\n",
       "      <td>0.094017</td>\n",
       "      <td>0.029915</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>3340</td>\n",
       "      <td>0.551282</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.094017</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.188034</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>0.338710</td>\n",
       "      <td>0.318182</td>\n",
       "      <td>0.287671</td>\n",
       "      <td>60</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>71</td>\n",
       "      <td>0.887500</td>\n",
       "      <td>61</td>\n",
       "      <td>...</td>\n",
       "      <td>0.087500</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>14020</td>\n",
       "      <td>0.425000</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>0.112500</td>\n",
       "      <td>0.075000</td>\n",
       "      <td>0.187500</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49995</th>\n",
       "      <td>49995</td>\n",
       "      <td>49995</td>\n",
       "      <td>0.344828</td>\n",
       "      <td>0.169972</td>\n",
       "      <td>0.192308</td>\n",
       "      <td>180</td>\n",
       "      <td>0.509915</td>\n",
       "      <td>200</td>\n",
       "      <td>0.738007</td>\n",
       "      <td>53</td>\n",
       "      <td>...</td>\n",
       "      <td>0.081181</td>\n",
       "      <td>0.036900</td>\n",
       "      <td>0.007380</td>\n",
       "      <td>8772</td>\n",
       "      <td>0.468635</td>\n",
       "      <td>0.081181</td>\n",
       "      <td>0.099631</td>\n",
       "      <td>0.132841</td>\n",
       "      <td>0.217712</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49996</th>\n",
       "      <td>49996</td>\n",
       "      <td>49996</td>\n",
       "      <td>0.417989</td>\n",
       "      <td>0.389163</td>\n",
       "      <td>0.392060</td>\n",
       "      <td>190</td>\n",
       "      <td>0.935961</td>\n",
       "      <td>187</td>\n",
       "      <td>0.935000</td>\n",
       "      <td>314</td>\n",
       "      <td>...</td>\n",
       "      <td>0.150000</td>\n",
       "      <td>0.005000</td>\n",
       "      <td>0.015000</td>\n",
       "      <td>1395</td>\n",
       "      <td>0.485000</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.170000</td>\n",
       "      <td>0.155000</td>\n",
       "      <td>0.140000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49997</th>\n",
       "      <td>49997</td>\n",
       "      <td>49997</td>\n",
       "      <td>0.352174</td>\n",
       "      <td>0.239645</td>\n",
       "      <td>0.242152</td>\n",
       "      <td>230</td>\n",
       "      <td>0.680473</td>\n",
       "      <td>227</td>\n",
       "      <td>0.685801</td>\n",
       "      <td>279</td>\n",
       "      <td>...</td>\n",
       "      <td>0.078550</td>\n",
       "      <td>0.009063</td>\n",
       "      <td>0.012085</td>\n",
       "      <td>8513</td>\n",
       "      <td>0.510574</td>\n",
       "      <td>0.039275</td>\n",
       "      <td>0.108761</td>\n",
       "      <td>0.120846</td>\n",
       "      <td>0.220544</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49998</th>\n",
       "      <td>49998</td>\n",
       "      <td>49998</td>\n",
       "      <td>0.334764</td>\n",
       "      <td>0.331915</td>\n",
       "      <td>0.329810</td>\n",
       "      <td>232</td>\n",
       "      <td>0.987234</td>\n",
       "      <td>230</td>\n",
       "      <td>0.966387</td>\n",
       "      <td>281</td>\n",
       "      <td>...</td>\n",
       "      <td>0.088235</td>\n",
       "      <td>0.004202</td>\n",
       "      <td>0.008403</td>\n",
       "      <td>4295</td>\n",
       "      <td>0.567227</td>\n",
       "      <td>0.042017</td>\n",
       "      <td>0.117647</td>\n",
       "      <td>0.100840</td>\n",
       "      <td>0.172269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49999</th>\n",
       "      <td>49999</td>\n",
       "      <td>49999</td>\n",
       "      <td>0.336207</td>\n",
       "      <td>0.184834</td>\n",
       "      <td>0.174497</td>\n",
       "      <td>113</td>\n",
       "      <td>0.535545</td>\n",
       "      <td>117</td>\n",
       "      <td>0.495763</td>\n",
       "      <td>165</td>\n",
       "      <td>...</td>\n",
       "      <td>0.088983</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012712</td>\n",
       "      <td>106</td>\n",
       "      <td>0.483051</td>\n",
       "      <td>0.021186</td>\n",
       "      <td>0.122881</td>\n",
       "      <td>0.152542</td>\n",
       "      <td>0.220339</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>50000 rows × 59 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       index_x  Unnamed: 0  local_gap_compressed_percent_id  \\\n",
       "0            0           0                         0.287582   \n",
       "1            1           1                         0.319635   \n",
       "2            2           2                         0.279621   \n",
       "3            3           3                         0.327273   \n",
       "4            4           4                         0.338710   \n",
       "...        ...         ...                              ...   \n",
       "49995    49995       49995                         0.344828   \n",
       "49996    49996       49996                         0.417989   \n",
       "49997    49997       49997                         0.352174   \n",
       "49998    49998       49998                         0.334764   \n",
       "49999    49999       49999                         0.336207   \n",
       "\n",
       "       scaled_local_query_percent_id  scaled_local_symmetric_percent_id  \\\n",
       "0                           0.217822                           0.215686   \n",
       "1                           0.295359                           0.297872   \n",
       "2                           0.234127                           0.218924   \n",
       "3                           0.200743                           0.214712   \n",
       "4                           0.318182                           0.287671   \n",
       "...                              ...                                ...   \n",
       "49995                       0.169972                           0.192308   \n",
       "49996                       0.389163                           0.392060   \n",
       "49997                       0.239645                           0.242152   \n",
       "49998                       0.331915                           0.329810   \n",
       "49999                       0.184834                           0.174497   \n",
       "\n",
       "       query_align_len  query_align_cov  subject_align_len  subject_align_cov  \\\n",
       "0                  160         0.792079                152           0.737864   \n",
       "1                  218         0.919831                226           0.969957   \n",
       "2                  211         0.837302                210           0.731707   \n",
       "3                  166         0.617100                163           0.696581   \n",
       "4                   60         0.909091                 71           0.887500   \n",
       "...                ...              ...                ...                ...   \n",
       "49995              180         0.509915                200           0.738007   \n",
       "49996              190         0.935961                187           0.935000   \n",
       "49997              230         0.680473                227           0.685801   \n",
       "49998              232         0.987234                230           0.966387   \n",
       "49999              113         0.535545                117           0.495763   \n",
       "\n",
       "       bit_score  ...     AAC_V     AAC_W     AAC_Y  index  GAAC_alphatic  \\\n",
       "0            131  ...  0.106796  0.009709  0.029126  12897       0.451456   \n",
       "1            282  ...  0.081545  0.004292  0.030043  13026       0.412017   \n",
       "2             96  ...  0.114983  0.010453  0.017422   8203       0.543554   \n",
       "3            175  ...  0.094017  0.029915  0.025641   3340       0.551282   \n",
       "4             61  ...  0.087500  0.037500  0.062500  14020       0.425000   \n",
       "...          ...  ...       ...       ...       ...    ...            ...   \n",
       "49995         53  ...  0.081181  0.036900  0.007380   8772       0.468635   \n",
       "49996        314  ...  0.150000  0.005000  0.015000   1395       0.485000   \n",
       "49997        279  ...  0.078550  0.009063  0.012085   8513       0.510574   \n",
       "49998        281  ...  0.088235  0.004202  0.008403   4295       0.567227   \n",
       "49999        165  ...  0.088983  0.000000  0.012712    106       0.483051   \n",
       "\n",
       "       GAAC_aromatic  GAAC_postivecharge  GAAC_negativecharge  GAAC_uncharge  \\\n",
       "0           0.053398            0.165049             0.140777       0.189320   \n",
       "1           0.064378            0.154506             0.184549       0.184549   \n",
       "2           0.059233            0.135889             0.097561       0.163763   \n",
       "3           0.111111            0.094017             0.055556       0.188034   \n",
       "4           0.200000            0.112500             0.075000       0.187500   \n",
       "...              ...                 ...                  ...            ...   \n",
       "49995       0.081181            0.099631             0.132841       0.217712   \n",
       "49996       0.050000            0.170000             0.155000       0.140000   \n",
       "49997       0.039275            0.108761             0.120846       0.220544   \n",
       "49998       0.042017            0.117647             0.100840       0.172269   \n",
       "49999       0.021186            0.122881             0.152542       0.220339   \n",
       "\n",
       "       protein_match  \n",
       "0              False  \n",
       "1               True  \n",
       "2               True  \n",
       "3              False  \n",
       "4               True  \n",
       "...              ...  \n",
       "49995          False  \n",
       "49996           True  \n",
       "49997          False  \n",
       "49998           True  \n",
       "49999          False  \n",
       "\n",
       "[50000 rows x 59 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "084d9561",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get rid of stuff that isn't quantitative\n",
    "\n",
    "df = df.drop(columns = ['Unnamed: 0','thermo_index', 'm_protein_seq', 't_protein_seq',\n",
    "                        'm_protein_desc', 't_protein_desc'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "eea27e6c",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_x</th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>meso_index</th>\n",
       "      <th>...</th>\n",
       "      <th>AAC_V</th>\n",
       "      <th>AAC_W</th>\n",
       "      <th>AAC_Y</th>\n",
       "      <th>index</th>\n",
       "      <th>GAAC_alphatic</th>\n",
       "      <th>GAAC_aromatic</th>\n",
       "      <th>GAAC_postivecharge</th>\n",
       "      <th>GAAC_negativecharge</th>\n",
       "      <th>GAAC_uncharge</th>\n",
       "      <th>protein_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0.287582</td>\n",
       "      <td>0.217822</td>\n",
       "      <td>0.215686</td>\n",
       "      <td>160</td>\n",
       "      <td>0.792079</td>\n",
       "      <td>152</td>\n",
       "      <td>0.737864</td>\n",
       "      <td>131</td>\n",
       "      <td>12897</td>\n",
       "      <td>...</td>\n",
       "      <td>0.106796</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.029126</td>\n",
       "      <td>12897</td>\n",
       "      <td>0.451456</td>\n",
       "      <td>0.053398</td>\n",
       "      <td>0.165049</td>\n",
       "      <td>0.140777</td>\n",
       "      <td>0.189320</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0.319635</td>\n",
       "      <td>0.295359</td>\n",
       "      <td>0.297872</td>\n",
       "      <td>218</td>\n",
       "      <td>0.919831</td>\n",
       "      <td>226</td>\n",
       "      <td>0.969957</td>\n",
       "      <td>282</td>\n",
       "      <td>13026</td>\n",
       "      <td>...</td>\n",
       "      <td>0.081545</td>\n",
       "      <td>0.004292</td>\n",
       "      <td>0.030043</td>\n",
       "      <td>13026</td>\n",
       "      <td>0.412017</td>\n",
       "      <td>0.064378</td>\n",
       "      <td>0.154506</td>\n",
       "      <td>0.184549</td>\n",
       "      <td>0.184549</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0.279621</td>\n",
       "      <td>0.234127</td>\n",
       "      <td>0.218924</td>\n",
       "      <td>211</td>\n",
       "      <td>0.837302</td>\n",
       "      <td>210</td>\n",
       "      <td>0.731707</td>\n",
       "      <td>96</td>\n",
       "      <td>8203</td>\n",
       "      <td>...</td>\n",
       "      <td>0.114983</td>\n",
       "      <td>0.010453</td>\n",
       "      <td>0.017422</td>\n",
       "      <td>8203</td>\n",
       "      <td>0.543554</td>\n",
       "      <td>0.059233</td>\n",
       "      <td>0.135889</td>\n",
       "      <td>0.097561</td>\n",
       "      <td>0.163763</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0.327273</td>\n",
       "      <td>0.200743</td>\n",
       "      <td>0.214712</td>\n",
       "      <td>166</td>\n",
       "      <td>0.617100</td>\n",
       "      <td>163</td>\n",
       "      <td>0.696581</td>\n",
       "      <td>175</td>\n",
       "      <td>3340</td>\n",
       "      <td>...</td>\n",
       "      <td>0.094017</td>\n",
       "      <td>0.029915</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>3340</td>\n",
       "      <td>0.551282</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.094017</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.188034</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0.338710</td>\n",
       "      <td>0.318182</td>\n",
       "      <td>0.287671</td>\n",
       "      <td>60</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>71</td>\n",
       "      <td>0.887500</td>\n",
       "      <td>61</td>\n",
       "      <td>14020</td>\n",
       "      <td>...</td>\n",
       "      <td>0.087500</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>14020</td>\n",
       "      <td>0.425000</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>0.112500</td>\n",
       "      <td>0.075000</td>\n",
       "      <td>0.187500</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   index_x  local_gap_compressed_percent_id  scaled_local_query_percent_id  \\\n",
       "0        0                         0.287582                       0.217822   \n",
       "1        1                         0.319635                       0.295359   \n",
       "2        2                         0.279621                       0.234127   \n",
       "3        3                         0.327273                       0.200743   \n",
       "4        4                         0.338710                       0.318182   \n",
       "\n",
       "   scaled_local_symmetric_percent_id  query_align_len  query_align_cov  \\\n",
       "0                           0.215686              160         0.792079   \n",
       "1                           0.297872              218         0.919831   \n",
       "2                           0.218924              211         0.837302   \n",
       "3                           0.214712              166         0.617100   \n",
       "4                           0.287671               60         0.909091   \n",
       "\n",
       "   subject_align_len  subject_align_cov  bit_score  meso_index  ...     AAC_V  \\\n",
       "0                152           0.737864        131       12897  ...  0.106796   \n",
       "1                226           0.969957        282       13026  ...  0.081545   \n",
       "2                210           0.731707         96        8203  ...  0.114983   \n",
       "3                163           0.696581        175        3340  ...  0.094017   \n",
       "4                 71           0.887500         61       14020  ...  0.087500   \n",
       "\n",
       "      AAC_W     AAC_Y  index  GAAC_alphatic  GAAC_aromatic  \\\n",
       "0  0.009709  0.029126  12897       0.451456       0.053398   \n",
       "1  0.004292  0.030043  13026       0.412017       0.064378   \n",
       "2  0.010453  0.017422   8203       0.543554       0.059233   \n",
       "3  0.029915  0.025641   3340       0.551282       0.111111   \n",
       "4  0.037500  0.062500  14020       0.425000       0.200000   \n",
       "\n",
       "   GAAC_postivecharge  GAAC_negativecharge  GAAC_uncharge  protein_match  \n",
       "0            0.165049             0.140777       0.189320          False  \n",
       "1            0.154506             0.184549       0.184549           True  \n",
       "2            0.135889             0.097561       0.163763           True  \n",
       "3            0.094017             0.055556       0.188034          False  \n",
       "4            0.112500             0.075000       0.187500           True  \n",
       "\n",
       "[5 rows x 53 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07cc3a80",
   "metadata": {},
   "source": [
    "From pairplot below, it looks like query_align_cov_16s and subject_align_cov_16s dont correlate with target. Let's remove them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "36d8dadb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# sns.pairplot(df.sample(500),height = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "91079edd",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.drop(columns = ['query_align_cov_16s', 'subject_align_cov_16s'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "7929625e",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_x</th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>meso_index</th>\n",
       "      <th>...</th>\n",
       "      <th>AAC_T</th>\n",
       "      <th>AAC_V</th>\n",
       "      <th>AAC_W</th>\n",
       "      <th>AAC_Y</th>\n",
       "      <th>index</th>\n",
       "      <th>GAAC_alphatic</th>\n",
       "      <th>GAAC_aromatic</th>\n",
       "      <th>GAAC_postivecharge</th>\n",
       "      <th>GAAC_negativecharge</th>\n",
       "      <th>GAAC_uncharge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.00000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>24999.500000</td>\n",
       "      <td>0.349026</td>\n",
       "      <td>0.276478</td>\n",
       "      <td>0.270792</td>\n",
       "      <td>206.011660</td>\n",
       "      <td>0.791507</td>\n",
       "      <td>205.414920</td>\n",
       "      <td>0.770875</td>\n",
       "      <td>230.63206</td>\n",
       "      <td>8098.858120</td>\n",
       "      <td>...</td>\n",
       "      <td>0.057258</td>\n",
       "      <td>0.088333</td>\n",
       "      <td>0.009929</td>\n",
       "      <td>0.019106</td>\n",
       "      <td>8098.858120</td>\n",
       "      <td>0.477431</td>\n",
       "      <td>0.059405</td>\n",
       "      <td>0.126355</td>\n",
       "      <td>0.118392</td>\n",
       "      <td>0.218418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>14433.901067</td>\n",
       "      <td>0.096608</td>\n",
       "      <td>0.114404</td>\n",
       "      <td>0.116039</td>\n",
       "      <td>64.260901</td>\n",
       "      <td>0.144893</td>\n",
       "      <td>64.430548</td>\n",
       "      <td>0.179955</td>\n",
       "      <td>186.44859</td>\n",
       "      <td>4608.982941</td>\n",
       "      <td>...</td>\n",
       "      <td>0.016214</td>\n",
       "      <td>0.022656</td>\n",
       "      <td>0.008839</td>\n",
       "      <td>0.011123</td>\n",
       "      <td>4608.982941</td>\n",
       "      <td>0.052699</td>\n",
       "      <td>0.026374</td>\n",
       "      <td>0.028722</td>\n",
       "      <td>0.030594</td>\n",
       "      <td>0.033043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.164103</td>\n",
       "      <td>0.081301</td>\n",
       "      <td>0.045977</td>\n",
       "      <td>19.000000</td>\n",
       "      <td>0.284553</td>\n",
       "      <td>19.000000</td>\n",
       "      <td>0.082019</td>\n",
       "      <td>35.00000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>0.183673</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.024735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>12499.750000</td>\n",
       "      <td>0.293750</td>\n",
       "      <td>0.204545</td>\n",
       "      <td>0.202020</td>\n",
       "      <td>172.000000</td>\n",
       "      <td>0.673288</td>\n",
       "      <td>172.000000</td>\n",
       "      <td>0.656854</td>\n",
       "      <td>118.00000</td>\n",
       "      <td>4121.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.046296</td>\n",
       "      <td>0.072874</td>\n",
       "      <td>0.003690</td>\n",
       "      <td>0.011494</td>\n",
       "      <td>4121.000000</td>\n",
       "      <td>0.443983</td>\n",
       "      <td>0.041667</td>\n",
       "      <td>0.109966</td>\n",
       "      <td>0.106122</td>\n",
       "      <td>0.196787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>24999.500000</td>\n",
       "      <td>0.325000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.244648</td>\n",
       "      <td>210.000000</td>\n",
       "      <td>0.811321</td>\n",
       "      <td>209.000000</td>\n",
       "      <td>0.806584</td>\n",
       "      <td>186.00000</td>\n",
       "      <td>8046.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.056225</td>\n",
       "      <td>0.087500</td>\n",
       "      <td>0.007905</td>\n",
       "      <td>0.017391</td>\n",
       "      <td>8046.000000</td>\n",
       "      <td>0.477273</td>\n",
       "      <td>0.054264</td>\n",
       "      <td>0.128492</td>\n",
       "      <td>0.121569</td>\n",
       "      <td>0.217791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>37499.250000</td>\n",
       "      <td>0.367257</td>\n",
       "      <td>0.308300</td>\n",
       "      <td>0.301639</td>\n",
       "      <td>240.000000</td>\n",
       "      <td>0.923077</td>\n",
       "      <td>240.000000</td>\n",
       "      <td>0.923077</td>\n",
       "      <td>267.00000</td>\n",
       "      <td>12103.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.067114</td>\n",
       "      <td>0.103226</td>\n",
       "      <td>0.014409</td>\n",
       "      <td>0.025000</td>\n",
       "      <td>12103.000000</td>\n",
       "      <td>0.510417</td>\n",
       "      <td>0.070755</td>\n",
       "      <td>0.143918</td>\n",
       "      <td>0.136519</td>\n",
       "      <td>0.239057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>49999.000000</td>\n",
       "      <td>0.985075</td>\n",
       "      <td>0.985075</td>\n",
       "      <td>0.985075</td>\n",
       "      <td>400.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>399.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1788.00000</td>\n",
       "      <td>16564.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.208696</td>\n",
       "      <td>0.277778</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>0.156863</td>\n",
       "      <td>16564.000000</td>\n",
       "      <td>0.876325</td>\n",
       "      <td>0.252336</td>\n",
       "      <td>0.625000</td>\n",
       "      <td>0.411111</td>\n",
       "      <td>0.533333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 50 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            index_x  local_gap_compressed_percent_id  \\\n",
       "count  50000.000000                     50000.000000   \n",
       "mean   24999.500000                         0.349026   \n",
       "std    14433.901067                         0.096608   \n",
       "min        0.000000                         0.164103   \n",
       "25%    12499.750000                         0.293750   \n",
       "50%    24999.500000                         0.325000   \n",
       "75%    37499.250000                         0.367257   \n",
       "max    49999.000000                         0.985075   \n",
       "\n",
       "       scaled_local_query_percent_id  scaled_local_symmetric_percent_id  \\\n",
       "count                   50000.000000                       50000.000000   \n",
       "mean                        0.276478                           0.270792   \n",
       "std                         0.114404                           0.116039   \n",
       "min                         0.081301                           0.045977   \n",
       "25%                         0.204545                           0.202020   \n",
       "50%                         0.250000                           0.244648   \n",
       "75%                         0.308300                           0.301639   \n",
       "max                         0.985075                           0.985075   \n",
       "\n",
       "       query_align_len  query_align_cov  subject_align_len  subject_align_cov  \\\n",
       "count     50000.000000     50000.000000       50000.000000       50000.000000   \n",
       "mean        206.011660         0.791507         205.414920           0.770875   \n",
       "std          64.260901         0.144893          64.430548           0.179955   \n",
       "min          19.000000         0.284553          19.000000           0.082019   \n",
       "25%         172.000000         0.673288         172.000000           0.656854   \n",
       "50%         210.000000         0.811321         209.000000           0.806584   \n",
       "75%         240.000000         0.923077         240.000000           0.923077   \n",
       "max         400.000000         1.000000         399.000000           1.000000   \n",
       "\n",
       "         bit_score    meso_index  ...         AAC_T         AAC_V  \\\n",
       "count  50000.00000  50000.000000  ...  50000.000000  50000.000000   \n",
       "mean     230.63206   8098.858120  ...      0.057258      0.088333   \n",
       "std      186.44859   4608.982941  ...      0.016214      0.022656   \n",
       "min       35.00000      5.000000  ...      0.000000      0.000000   \n",
       "25%      118.00000   4121.000000  ...      0.046296      0.072874   \n",
       "50%      186.00000   8046.000000  ...      0.056225      0.087500   \n",
       "75%      267.00000  12103.000000  ...      0.067114      0.103226   \n",
       "max     1788.00000  16564.000000  ...      0.208696      0.277778   \n",
       "\n",
       "              AAC_W         AAC_Y         index  GAAC_alphatic  GAAC_aromatic  \\\n",
       "count  50000.000000  50000.000000  50000.000000   50000.000000   50000.000000   \n",
       "mean       0.009929      0.019106   8098.858120       0.477431       0.059405   \n",
       "std        0.008839      0.011123   4608.982941       0.052699       0.026374   \n",
       "min        0.000000      0.000000      5.000000       0.183673       0.000000   \n",
       "25%        0.003690      0.011494   4121.000000       0.443983       0.041667   \n",
       "50%        0.007905      0.017391   8046.000000       0.477273       0.054264   \n",
       "75%        0.014409      0.025000  12103.000000       0.510417       0.070755   \n",
       "max        0.107143      0.156863  16564.000000       0.876325       0.252336   \n",
       "\n",
       "       GAAC_postivecharge  GAAC_negativecharge  GAAC_uncharge  \n",
       "count        50000.000000         50000.000000   50000.000000  \n",
       "mean             0.126355             0.118392       0.218418  \n",
       "std              0.028722             0.030594       0.033043  \n",
       "min              0.000000             0.000000       0.024735  \n",
       "25%              0.109966             0.106122       0.196787  \n",
       "50%              0.128492             0.121569       0.217791  \n",
       "75%              0.143918             0.136519       0.239057  \n",
       "max              0.625000             0.411111       0.533333  \n",
       "\n",
       "[8 rows x 50 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90871d01",
   "metadata": {},
   "source": [
    "Split data into dev and test, and then split that into train and validation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "6519dcaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#drop columns that don't exihibit signficant pearson correlation with bit_score\n",
    "\n",
    "df = df.drop(columns = ['meso_index', 'meso_protein_int_index', 'local_gap_compressed_percent_id_16s', \n",
    "                        'scaled_local_query_percent_id_16s', 'scaled_local_symmetric_percent_id_16s',\n",
    "                       'bit_score_16s', 'm_ogt', 't_ogt', 'taxa_pair_index', 'thermo_protein_int_index'\n",
    "                       , 'prot_pair_index', 'ogt_difference'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "af9f23cf",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(42500, 39)\n",
      "(7500, 39)\n",
      "(36125, 39)\n",
      "(6375, 39)\n"
     ]
    }
   ],
   "source": [
    "#choosing 80/20 split instead of 85/15 because of volume of data\n",
    "\n",
    "dev, test = sklearn.model_selection.train_test_split(df, test_size=0.15, random_state=1)\n",
    "\n",
    "train, val = sklearn.model_selection.train_test_split(dev, test_size=0.15, random_state=1)\n",
    "\n",
    "print(dev.shape)\n",
    "print(test.shape)\n",
    "print(train.shape)\n",
    "print(val.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "8470550d",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#ID target and features, separate into separate arrays\n",
    "\n",
    "target = 'protein_match'\n",
    "input_features = [columns for columns in df]\n",
    "input_features.remove(target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "c62edd28",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['index_x', 'local_gap_compressed_percent_id', 'scaled_local_query_percent_id', 'scaled_local_symmetric_percent_id', 'query_align_len', 'query_align_cov', 'subject_align_len', 'subject_align_cov', 'bit_score', 'm_protein_len', 't_protein_len', 'index_y', 'AAC_A', 'AAC_C', 'AAC_D', 'AAC_E', 'AAC_F', 'AAC_G', 'AAC_H', 'AAC_I', 'AAC_K', 'AAC_L', 'AAC_M', 'AAC_N', 'AAC_P', 'AAC_Q', 'AAC_R', 'AAC_S', 'AAC_T', 'AAC_V', 'AAC_W', 'AAC_Y', 'index', 'GAAC_alphatic', 'GAAC_aromatic', 'GAAC_postivecharge', 'GAAC_negativecharge', 'GAAC_uncharge']\n",
      "protein_match\n"
     ]
    }
   ],
   "source": [
    "print(input_features)\n",
    "print(target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "1ab62bff",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(42500, 38) (7500, 38) (42500, 1) (7500, 1)\n"
     ]
    }
   ],
   "source": [
    "#split X and y\n",
    "\n",
    "dev_X = dev[input_features].values\n",
    "test_X = test[input_features].values\n",
    "\n",
    "dev_y = dev[target].values.reshape(-1,1)\n",
    "test_y = test[target].values.reshape(-1,1)  \n",
    "\n",
    "print(dev_X.shape, test_X.shape, dev_y.shape, test_y.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "9ac52ac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#same thing for training and validation data\n",
    "\n",
    "train_X = train[input_features].values\n",
    "val_X = val[input_features].values\n",
    "\n",
    "train_y = train[target].values.reshape(-1,1)\n",
    "val_y = val[target].values.reshape(-1,1) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c374f2d7",
   "metadata": {},
   "source": [
    "Scale the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "2b48b2e5",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "scaler = sklearn.preprocessing.StandardScaler()\n",
    "dev_X = scaler.fit_transform(dev_X)\n",
    "test_X = scaler.fit_transform(test_X)\n",
    "train_X = scaler.fit_transform(train_X)\n",
    "val_X = scaler.fit_transform(val_X)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3022371",
   "metadata": {},
   "source": [
    "Train the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "be4e09b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-2 {color: black;background-color: white;}#sk-container-id-2 pre{padding: 0;}#sk-container-id-2 div.sk-toggleable {background-color: white;}#sk-container-id-2 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-2 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-2 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-2 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-2 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-2 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-2 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-2 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-container-id-2 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-2 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-2 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-2 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-2 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-2 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-2 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-2 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-2 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-2 div.sk-item {position: relative;z-index: 1;}#sk-container-id-2 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-2 div.sk-item::before, #sk-container-id-2 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-2 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-2 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-2 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-2 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-2 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-2 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-2 div.sk-label-container {text-align: center;}#sk-container-id-2 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-2 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-2\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>RandomForestClassifier(max_features=0.5, max_samples=0.5, min_samples_split=10,\n",
       "                       min_weight_fraction_leaf=0.000215, n_estimators=150)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-2\" type=\"checkbox\" checked><label for=\"sk-estimator-id-2\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">RandomForestClassifier</label><div class=\"sk-toggleable__content\"><pre>RandomForestClassifier(max_features=0.5, max_samples=0.5, min_samples_split=10,\n",
       "                       min_weight_fraction_leaf=0.000215, n_estimators=150)</pre></div></div></div></div></div>"
      ],
      "text/plain": [
       "RandomForestClassifier(max_features=0.5, max_samples=0.5, min_samples_split=10,\n",
       "                       min_weight_fraction_leaf=0.000215, n_estimators=150)"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Random Forest\n",
    "\n",
    "model = sklearn.ensemble.RandomForestClassifier(n_estimators=150, max_depth=None, max_samples=0.5,\n",
    "                                                max_features=0.5, min_weight_fraction_leaf=0.000215,\n",
    "                                               min_samples_split=10)\n",
    "# model = sklearn.ensemble.RandomForestClassifier()\n",
    "model.fit(train_X, train_y.ravel())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69ebf269",
   "metadata": {},
   "source": [
    "Test the model, report relevant statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "c73327e9",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model score is: 0.9927843137254901\n",
      "[False False False ... False  True False]\n"
     ]
    }
   ],
   "source": [
    "score = model.score(val_X, val_y)\n",
    "print('Model score is: {}'.format(score))\n",
    "\n",
    "preds = model.predict(test_X)\n",
    "print(preds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "4e9dce62",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<sklearn.metrics._plot.confusion_matrix.ConfusionMatrixDisplay at 0x7fea041546d0>"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#confusion matrix\n",
    "\n",
    "confusion_matrix = sklearn.metrics.confusion_matrix(preds, test_y)\n",
    "sklearn.metrics.ConfusionMatrixDisplay(confusion_matrix).plot()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21feaafb",
   "metadata": {},
   "source": [
    "Convert the above code into functions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "bf1d563c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def train_model(dataframe, columns = [],  target = []):\n",
    "    \"\"\"\n",
    "    Takes dataframe and splits it into a training and testing set. \n",
    "    Note: Data is called train and test, but this test set is currently\n",
    "    closer to a validation set. Keeping nomenclature to keep model robust.\n",
    "    Trains a KNN classifier model with selected data.\n",
    "    \n",
    "    Params\n",
    "    ----------\n",
    "    dataframe: Pandas dataframe \n",
    "    columns: list of strings, representing input features\n",
    "    target: list of strings, representing target feature(s)\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    -Sk-learn model object\n",
    "    -train data (features)\n",
    "    -train data (target)\n",
    "    -validation data (features)\n",
    "    -validation data (target)\n",
    "    \"\"\"\n",
    "    #split data\n",
    "    dev, test = sklearn.model_selection.train_test_split(dataframe, test_size=0.15, random_state=1)\n",
    "    \n",
    "    #test input arguments\n",
    "    assert \"pandas.core.frame.DataFrame\" in str(type(dev))\n",
    "    assert \"pandas.core.frame.DataFrame\" in str(type(test))\n",
    "    assert \"str\" in str(type(columns[0]))\n",
    "    assert \"str\" in str(type(target[0]))\n",
    "   \n",
    "    #split into input and output feature(s)\n",
    "    dev_X = dev[columns].values\n",
    "    test_X = test[columns].values\n",
    "\n",
    "    dev_y = dev[target].values.reshape(-1,1)\n",
    "    test_y = test[target].values.reshape(-1,1)\n",
    "    \n",
    "    #scale data\n",
    "    scaler = sklearn.preprocessing.StandardScaler()\n",
    "    dev_X = scaler.fit_transform(dev_X)\n",
    "    test_X = scaler.fit_transform(test_X)\n",
    "    \n",
    "    #train model\n",
    "    model = sklearn.ensemble.RandomForestClassifier()\n",
    "    model = model.fit(dev_X, dev_y.ravel())\n",
    "    \n",
    "    return model, dev_X, dev_y, test_X, test_y\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "e40acd6d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "model, dev_X, dev_y, test_X, test_y = train_model(df, columns = input_features, \n",
    "                                          target='protein_match')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "46729a81",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "..\n",
      "----------------------------------------------------------------------\n",
      "Ran 2 tests in 8.896s\n",
      "\n",
      "OK\n"
     ]
    }
   ],
   "source": [
    "class TestModelTraining(unittest.TestCase):\n",
    "    \n",
    "    def test_invalid_inputs(self):\n",
    "    \n",
    "        #test that input data type is correct\n",
    "        \n",
    "        try:\n",
    "            train_model([1,2,3], columns = 'string', target = 'string')\n",
    "            self.assertTrue(False)\n",
    "        except AssertionError:\n",
    "            self.assertTrue(True)\n",
    "    \n",
    "#     def test_input_distro(self):\n",
    "        \n",
    "#         #test that dev and test features have similar Jensen Shannon Distribution\n",
    "        \n",
    "#         JSD = (\n",
    "#             scipy.stats.bootstrap((train_reg(split_data(df)[0], split_data(df)[1],\n",
    "#                                 columns = input_features, target=target)[1], train_reg(split_data(df)[0], \n",
    "#                                 split_data(df)[1], columns = input_features, target=target)[3]), \n",
    "#                                   JSD_dev_and_test, n_resamples=1000, \n",
    "#                                   batch=5, method='percentile')\n",
    "#         )\n",
    "\n",
    "#         div = JSD.confidence_interval[1]\n",
    "        \n",
    "#         #asserts that the divergence between data sets is sufficiently low\n",
    "#         assert abs(div) < 0.3, \"Warning! High JSD between dev and test set!\"\n",
    "        \n",
    "    def test_output_format(self):\n",
    "        \n",
    "        #asserts that function returns 4 objects to be assigned to pearson_corr, model, test_X, test_y\n",
    "        assert len(train_model(df, columns = input_features, \n",
    "                                          target = target)) == 5\n",
    "        \n",
    "\n",
    "suite = unittest.TestLoader().loadTestsFromTestCase(TestModelTraining)\n",
    "_ = unittest.TextTestRunner().run(suite)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "47f766ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "#need to fix this\n",
    "\n",
    "def evaluate_model(model, test_X, test_y):\n",
    "    \n",
    "    \"\"\"\n",
    "    Takes a trained model and test data and tests the model.\n",
    "    \n",
    "    Params\n",
    "    ----------\n",
    "    model: sklearn.neighbors.KNeighborsClassifier\n",
    "    test_X: numpy array\n",
    "    test_y: numpy array\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    Vector of predictions based on the model (numpy array)\n",
    "    \"\"\"\n",
    "    \n",
    "    #test input arguments\n",
    "    assert \"sklearn\" in str(type(model))\n",
    "    assert \"numpy.ndarray\" in str(type(test_X))\n",
    "    assert \"numpy.ndarray\" in str(type(test_y))\n",
    "    \n",
    "    preds = model.predict(test_X)\n",
    "    \n",
    "    return preds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "7bbb0242",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([False, False, False, ..., False,  True, False])"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preds = evaluate_model(model, test_X, test_y)\n",
    "preds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "8b2fdac6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "...\n",
      "----------------------------------------------------------------------\n",
      "Ran 3 tests in 24.048s\n",
      "\n",
      "OK\n"
     ]
    }
   ],
   "source": [
    "class TestModelPerformance(unittest.TestCase):\n",
    "\n",
    "    def test_asserts(self):\n",
    "        model, _, _, test_X, test_y = train_model(\n",
    "            df, columns=input_features, target='protein_match'\n",
    "        )\n",
    "        # assert that input types are correct\n",
    "        with self.assertRaises(AssertionError):\n",
    "            evaluate_model(model, [1, 2, 3], test_y)\n",
    "            \n",
    "    def test_model_output(self):\n",
    "        model, _, _, test_X, test_y = train_model(\n",
    "            df, columns=input_features, target='protein_match'\n",
    "        )\n",
    "        # assert output type is correct\n",
    "        output = evaluate_model(model, test_X, test_y)\n",
    "        self.assertIsInstance(output, np.ndarray)\n",
    "        \n",
    "    def test_pred_dimension(self):\n",
    "        model, _, _, test_X, test_y = train_model(df, \n",
    "            columns=input_features, target='protein_match'\n",
    "        )\n",
    "        # want to check that the number of predictions is equal to the number of test examples\n",
    "        preds = evaluate_model(model, test_X, test_y)\n",
    "        self.assertEqual(len(test_y), len(preds))\n",
    "\n",
    "suite = unittest.TestLoader().loadTestsFromTestCase(TestModelPerformance)\n",
    "_ = unittest.TextTestRunner().run(suite)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9b79929",
   "metadata": {},
   "source": [
    "A good test here can be adding outliers and seeing how much the score and predictions change (probably for the above function)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "dae1e64f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_model(model, test_X, test_y):\n",
    "    \"\"\"\n",
    "    Takes a test KNN Classifier model and plots the confusion matrix.\n",
    "    \n",
    "    Params\n",
    "    ----------\n",
    "    model: sklearn.neighbors.KNeighborsClassifier\n",
    "    test_X: numpy array\n",
    "    test_y: numpy array\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    -Confusion predictions vs. observations\n",
    "    -Model score\n",
    "    \"\"\"\n",
    "    \n",
    "    #test input arguments\n",
    "    assert \"sklearn\" in str(type(model))\n",
    "    assert \"numpy.ndarray\" in str(type(test_X))\n",
    "    assert \"numpy.ndarray\" in str(type(test_y))\n",
    "    \n",
    "    score = model.score(test_X, test_y)\n",
    "    preds = evaluate_model(model, test_X, test_y)\n",
    "   \n",
    "    # plot confusion matrix\n",
    "    confusion_matrix = sklearn.metrics.confusion_matrix(preds, test_y)\n",
    "    cm_plot = sklearn.metrics.ConfusionMatrixDisplay(confusion_matrix)\n",
    "    \n",
    "    cm_plot.plot(cmap=plt.cm.Blues)\n",
    "    cm_plot.ax_.set_title('Confusion Matrix')\n",
    "    \n",
    "    return score\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "c923e39b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9957333333333334"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#make some appropriate display labels here\n",
    "\n",
    "plot_model(model, test_X, test_y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15e9e191",
   "metadata": {},
   "outputs": [],
   "source": [
    "#wrapper function\n",
    "\n",
    "def RF_wrapper(dataframe):\n",
    "    \"\"\"\n",
    "    Takes dataframe and runs it through kNN model.\n",
    "    \n",
    "    Params\n",
    "    ----------\n",
    "    dataframe: Pandas dataframe \n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    -Target feature predictions\n",
    "    -Parity plot\n",
    "    \"\"\"\n",
    "    \n",
    "    assert 'pandas.core.frame.DataFrame' in str(type(dataframe))\n",
    "\n",
    "    #user inputs target feature\n",
    "    target = 'protein_match'\n",
    "    \n",
    "    #define input features\n",
    "    input_features = [columns for columns in dataframe]\n",
    "    \n",
    "    input_features.remove(target)\n",
    "\n",
    "    #train the model based off data split\n",
    "    model, dev_X, dev_y, test_X, test_y = train_model(\n",
    "                                        dataframe, columns=input_features, \n",
    "                                        target=target\n",
    "                                                   )\n",
    "    \n",
    "    #test the model and return predictions\n",
    "    preds = evaluate_model(model, test_X, test_y)\n",
    "\n",
    "    #plot the results of the model\n",
    "    plot_model(model, test_X, test_y)\n",
    "    \n",
    "    return preds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbbd8fa8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "RF_wrapper(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37c0e2b2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "class TestWrapper(unittest.TestCase):\n",
    "    \n",
    "    def test_wrapper_input(self):\n",
    "        #test that input data type is correct\n",
    "        try:\n",
    "            RF_wrapper([1,2,3])\n",
    "            self.assertTrue(False)\n",
    "        except AssertionError:\n",
    "            self.assertTrue(True)\n",
    "\n",
    "    def test_wrapper_output(self):\n",
    "        model, _, _, test_X, test_y = train_model(\n",
    "            df, \n",
    "            columns=input_features, target='protein_match'\n",
    "        )\n",
    "        # assert output type is correct\n",
    "        output = evaluate_model(model, test_X, test_y)\n",
    "        self.assertIsInstance(output, np.ndarray)\n",
    "        \n",
    "    def test_output_dimension(self):\n",
    "        model, _, _, test_X, test_y = train_model(\n",
    "            df, \n",
    "            columns=input_features, target='protein_match'\n",
    "        )\n",
    "        # want to check that the number of predictions is equal to the number of test examples\n",
    "        preds = evaluate_model(model, test_X, test_y)\n",
    "        self.assertEqual(len(test_y), len(preds))\n",
    "        \n",
    "\n",
    "suite = unittest.TestLoader().loadTestsFromTestCase(TestWrapper)\n",
    "_ = unittest.TextTestRunner().run(suite)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02c53a58",
   "metadata": {},
   "source": [
    "**IGNORE BELOW**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13d52919",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #plot empirical distribution of scaled_local_query_percent_id\n",
    "\n",
    "# target = df['protein_match']\n",
    "# #create histplot\n",
    "# fig, ax = plt.subplots()\n",
    "\n",
    "# sns.histplot(data = df, x = target)\n",
    "\n",
    "# ax.set_title('Sample Histogram', fontsize=16)\n",
    "# ax.set_xlabel('Class', fontsize=14)\n",
    "# ax.set_ylabel('Count', fontsize=14)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e3d1950",
   "metadata": {},
   "source": [
    "Things I've tried to improve model: \n",
    "\n",
    "1. drop bit scores over 1000\n",
    "2. Switch to 85/15 train/test split\n",
    "3. Lasso regression - not great\n",
    "4. Ridge regression - R2 = 0.87, about the same as normal Linear regression\n",
    "5. KNN regression (n_neighbors optimized at 8) gives best result, R2 = ~0.93 --> \n",
    "*got it up to .942 by removing a few features\n",
    "6. Decision tree regressor was slightly worse than KNN\n",
    "7. Input DT classifier and RF classifier. RF classifier has best performance\n",
    "8. Changed target to binary protein function match"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5220b26b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
