{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "265e8129",
   "metadata": {},
   "source": [
    "This notebook uses iFeatureOmega, a feature generation software, to add to our feature space for a RandomForestClassifier that predicts protein pair functionality."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b88f847",
   "metadata": {},
   "source": [
    "To do:\n",
    "\n",
    "1) Write unit tests for iFeatureOmega\n",
    "2) time trial for different descriptors - DONE\n",
    "3) Figure out how to append meso and thermo descriptors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "61c53eee",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats\n",
    "import sklearn.preprocessing\n",
    "import sklearn.model_selection\n",
    "import sklearn.ensemble\n",
    "import sklearn.feature_selection\n",
    "import unittest\n",
    "import iFeatureOmegaCLI\n",
    "import Bio.SeqIO\n",
    "import Bio.SeqRecord\n",
    "import io\n",
    "from io import StringIO\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "ba87d47f",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/loganroberts/Learn2Therm/ValidProt/FAFSA\n"
     ]
    }
   ],
   "source": [
    "cd /Users/loganroberts/Learn2Therm/ValidProt/FAFSA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "f90c1cdd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Unnamed: 0', 'local_gap_compressed_percent_id',\n",
       "       'scaled_local_query_percent_id', 'scaled_local_symmetric_percent_id',\n",
       "       'query_align_len', 'query_align_cov', 'subject_align_len',\n",
       "       'subject_align_cov', 'bit_score', 'thermo_index', 'meso_index',\n",
       "       'prot_pair_index', 'meso_protein_int_index', 'thermo_protein_int_index',\n",
       "       'taxa_pair_index', 'local_gap_compressed_percent_id_16s',\n",
       "       'scaled_local_query_percent_id_16s',\n",
       "       'scaled_local_symmetric_percent_id_16s', 'query_align_cov_16s',\n",
       "       'subject_align_cov_16s', 'bit_score_16s', 'm_ogt', 't_ogt',\n",
       "       'ogt_difference', 'm_protein_seq', 't_protein_seq', 'm_protein_desc',\n",
       "       't_protein_desc', 'm_protein_len', 't_protein_len'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#convert to pandas df\n",
    "df = pd.read_csv('learn2therm_sample_50k.csv')\n",
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "d7f0a563",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/loganroberts/Learn2Therm/ValidProt/notebooks\n"
     ]
    }
   ],
   "source": [
    "cd /Users/loganroberts/Learn2Therm/ValidProt/notebooks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "043bf4dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "target = pd.read_csv('protein_match_6k.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "dbe2a5c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prot_pair_index</th>\n",
       "      <th>Jaccard_Score</th>\n",
       "      <th>protein_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>48641291</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>92992745</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>157628663</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>136708305</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>133672542</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6343</th>\n",
       "      <td>55489429</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6344</th>\n",
       "      <td>172293605</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6345</th>\n",
       "      <td>47082975</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6346</th>\n",
       "      <td>154513027</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6347</th>\n",
       "      <td>90211763</td>\n",
       "      <td>0.2</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6348 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      prot_pair_index  Jaccard_Score protein_match\n",
       "0            48641291            1.0           Yes\n",
       "1            92992745            1.0           Yes\n",
       "2           157628663            1.0           Yes\n",
       "3           136708305            1.0           Yes\n",
       "4           133672542            1.0           Yes\n",
       "...               ...            ...           ...\n",
       "6343         55489429            0.5           Yes\n",
       "6344        172293605            1.0           Yes\n",
       "6345         47082975            1.0           Yes\n",
       "6346        154513027            1.0           Yes\n",
       "6347         90211763            0.2            No\n",
       "\n",
       "[6348 rows x 3 columns]"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "id": "792508bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6348, 32)"
      ]
     },
     "execution_count": 178,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.merge(df, target, on=['prot_pair_index'])\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "9fd94531",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>thermo_index</th>\n",
       "      <th>...</th>\n",
       "      <th>t_ogt</th>\n",
       "      <th>ogt_difference</th>\n",
       "      <th>m_protein_seq</th>\n",
       "      <th>t_protein_seq</th>\n",
       "      <th>m_protein_desc</th>\n",
       "      <th>t_protein_desc</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "      <th>Jaccard_Score</th>\n",
       "      <th>protein_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0.287582</td>\n",
       "      <td>0.217822</td>\n",
       "      <td>0.215686</td>\n",
       "      <td>160</td>\n",
       "      <td>0.792079</td>\n",
       "      <td>152</td>\n",
       "      <td>0.737864</td>\n",
       "      <td>131</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>50.0</td>\n",
       "      <td>22.5</td>\n",
       "      <td>MAESGTSRRADHLVPVPGPDAEPPAVADELLRAVGRGDEQAFGRLY...</td>\n",
       "      <td>MPSQITESERIELAERFERDALPLLDQLYSAALRMTRNPADAEDLV...</td>\n",
       "      <td>ECF RNA polymerase sigma factor SigK</td>\n",
       "      <td>sigma-70 family RNA polymerase sigma factor</td>\n",
       "      <td>206</td>\n",
       "      <td>202</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0.319635</td>\n",
       "      <td>0.295359</td>\n",
       "      <td>0.297872</td>\n",
       "      <td>218</td>\n",
       "      <td>0.919831</td>\n",
       "      <td>226</td>\n",
       "      <td>0.969957</td>\n",
       "      <td>282</td>\n",
       "      <td>11324</td>\n",
       "      <td>...</td>\n",
       "      <td>54.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>MARIALVDDDRNILTSVSMTLEAEGFEVETYNDGQSALDAFNKRMP...</td>\n",
       "      <td>MRVLLVEDDPNTSRSIEMMLTHANLNVYATDMGEEGIDLAKLYDYD...</td>\n",
       "      <td>response regulator transcription factor</td>\n",
       "      <td>response regulator transcription factor</td>\n",
       "      <td>233</td>\n",
       "      <td>237</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0.279621</td>\n",
       "      <td>0.234127</td>\n",
       "      <td>0.218924</td>\n",
       "      <td>211</td>\n",
       "      <td>0.837302</td>\n",
       "      <td>210</td>\n",
       "      <td>0.731707</td>\n",
       "      <td>96</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>50.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MKDTVVFVTGAARGIGAHTARLAVARGARVALVGLEPHLLADLAAE...</td>\n",
       "      <td>MTPEQIFSGQTAIVTGGASGIGAATVEHIARRGGRVFSVDLSYDSP...</td>\n",
       "      <td>SDR family oxidoreductase</td>\n",
       "      <td>SDR family oxidoreductase</td>\n",
       "      <td>287</td>\n",
       "      <td>252</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0.327273</td>\n",
       "      <td>0.200743</td>\n",
       "      <td>0.214712</td>\n",
       "      <td>166</td>\n",
       "      <td>0.617100</td>\n",
       "      <td>163</td>\n",
       "      <td>0.696581</td>\n",
       "      <td>175</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>50.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MTSGLWERVLDGVWVTIQLLVLSALLATAVSFVVGIARTHRLWIVR...</td>\n",
       "      <td>MAMSRRKRGQLARGIQYAILVIVVVVLALLADWGKIGKAFFDWEAA...</td>\n",
       "      <td>ectoine/hydroxyectoine ABC transporter permeas...</td>\n",
       "      <td>amino acid ABC transporter permease</td>\n",
       "      <td>234</td>\n",
       "      <td>269</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0.338710</td>\n",
       "      <td>0.318182</td>\n",
       "      <td>0.287671</td>\n",
       "      <td>60</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>71</td>\n",
       "      <td>0.887500</td>\n",
       "      <td>61</td>\n",
       "      <td>9827</td>\n",
       "      <td>...</td>\n",
       "      <td>50.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>MIISLRRGLRFIRFIVFFAALVYLFYHVLDLFNGWISPVDQYQMPT...</td>\n",
       "      <td>MKRMVWRTLKVFIIFIACTLLFYFGLRFMHLEYEQFHRYEPPEGPA...</td>\n",
       "      <td>YqzK family protein</td>\n",
       "      <td>YqzK family protein</td>\n",
       "      <td>80</td>\n",
       "      <td>66</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6343</th>\n",
       "      <td>6495</td>\n",
       "      <td>0.387574</td>\n",
       "      <td>0.367978</td>\n",
       "      <td>0.360882</td>\n",
       "      <td>336</td>\n",
       "      <td>0.943820</td>\n",
       "      <td>337</td>\n",
       "      <td>0.910811</td>\n",
       "      <td>588</td>\n",
       "      <td>11707</td>\n",
       "      <td>...</td>\n",
       "      <td>50.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>MDNEKPIVDFSALRDELGKNYEPVQVMDGQAKVIKPEIVSKYSNDQ...</td>\n",
       "      <td>MKLEFPILQIVDENGNLVNENYRKAITEELVKKFYYHMIRIRTFDR...</td>\n",
       "      <td>pyruvate dehydrogenase (acetyl-transferring) E...</td>\n",
       "      <td>pyruvate dehydrogenase (acetyl-transferring) E...</td>\n",
       "      <td>370</td>\n",
       "      <td>356</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6344</th>\n",
       "      <td>6496</td>\n",
       "      <td>0.488462</td>\n",
       "      <td>0.488462</td>\n",
       "      <td>0.455197</td>\n",
       "      <td>258</td>\n",
       "      <td>0.992308</td>\n",
       "      <td>263</td>\n",
       "      <td>0.882550</td>\n",
       "      <td>619</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>50.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>MANPYLDDLRLAHVLADSADNLSMDRFGALDLEVSTKPDMTYVTES...</td>\n",
       "      <td>MTRPYADDIHLAKRLADTADAITVPRYSVRDLTVRTKPDRSPVTDA...</td>\n",
       "      <td>histidinol phosphatase</td>\n",
       "      <td>histidinol-phosphatase</td>\n",
       "      <td>298</td>\n",
       "      <td>260</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6345</th>\n",
       "      <td>6497</td>\n",
       "      <td>0.383838</td>\n",
       "      <td>0.376238</td>\n",
       "      <td>0.374384</td>\n",
       "      <td>296</td>\n",
       "      <td>0.976898</td>\n",
       "      <td>299</td>\n",
       "      <td>0.977124</td>\n",
       "      <td>436</td>\n",
       "      <td>5381</td>\n",
       "      <td>...</td>\n",
       "      <td>45.0</td>\n",
       "      <td>21.5</td>\n",
       "      <td>MGHDHDHSHGTATTNRTKLAWAFGITFTILIAEVIGAILTNSLALL...</td>\n",
       "      <td>MAHQHGPARPEHASGRYLKRLMAAFGIGLVFMVLEVVVGVLTGSLA...</td>\n",
       "      <td>cation diffusion facilitator family transporter</td>\n",
       "      <td>cation diffusion facilitator family transporter</td>\n",
       "      <td>306</td>\n",
       "      <td>303</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6346</th>\n",
       "      <td>6498</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>78</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>78</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>258</td>\n",
       "      <td>14963</td>\n",
       "      <td>...</td>\n",
       "      <td>52.5</td>\n",
       "      <td>22.5</td>\n",
       "      <td>MAKPALRKPKKKSNPLKAAKISYVDYKDTALLRKFISDRGKIRARR...</td>\n",
       "      <td>MAKQPPRKPKKKVCVFCQEKISYVDYKDTALLRKFISDRGKIRARR...</td>\n",
       "      <td>30S ribosomal protein S18</td>\n",
       "      <td>30S ribosomal protein S18</td>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6347</th>\n",
       "      <td>6499</td>\n",
       "      <td>0.343220</td>\n",
       "      <td>0.223757</td>\n",
       "      <td>0.264274</td>\n",
       "      <td>234</td>\n",
       "      <td>0.646409</td>\n",
       "      <td>238</td>\n",
       "      <td>0.948207</td>\n",
       "      <td>347</td>\n",
       "      <td>7134</td>\n",
       "      <td>...</td>\n",
       "      <td>52.5</td>\n",
       "      <td>24.5</td>\n",
       "      <td>MTTHVPGDALVSLNAVNKHYGQLHVLKDINLQVRKGEVVVVIGPSG...</td>\n",
       "      <td>MSDGLTLRGITKVLGGRTIVDGFDLEVRRGELVCLLGPSGCGKTTT...</td>\n",
       "      <td>amino acid ABC transporter ATP-binding protein</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>251</td>\n",
       "      <td>362</td>\n",
       "      <td>0.2</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6348 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0  local_gap_compressed_percent_id  \\\n",
       "0              0                         0.287582   \n",
       "1              1                         0.319635   \n",
       "2              2                         0.279621   \n",
       "3              3                         0.327273   \n",
       "4              4                         0.338710   \n",
       "...          ...                              ...   \n",
       "6343        6495                         0.387574   \n",
       "6344        6496                         0.488462   \n",
       "6345        6497                         0.383838   \n",
       "6346        6498                         0.769231   \n",
       "6347        6499                         0.343220   \n",
       "\n",
       "      scaled_local_query_percent_id  scaled_local_symmetric_percent_id  \\\n",
       "0                          0.217822                           0.215686   \n",
       "1                          0.295359                           0.297872   \n",
       "2                          0.234127                           0.218924   \n",
       "3                          0.200743                           0.214712   \n",
       "4                          0.318182                           0.287671   \n",
       "...                             ...                                ...   \n",
       "6343                       0.367978                           0.360882   \n",
       "6344                       0.488462                           0.455197   \n",
       "6345                       0.376238                           0.374384   \n",
       "6346                       0.769231                           0.769231   \n",
       "6347                       0.223757                           0.264274   \n",
       "\n",
       "      query_align_len  query_align_cov  subject_align_len  subject_align_cov  \\\n",
       "0                 160         0.792079                152           0.737864   \n",
       "1                 218         0.919831                226           0.969957   \n",
       "2                 211         0.837302                210           0.731707   \n",
       "3                 166         0.617100                163           0.696581   \n",
       "4                  60         0.909091                 71           0.887500   \n",
       "...               ...              ...                ...                ...   \n",
       "6343              336         0.943820                337           0.910811   \n",
       "6344              258         0.992308                263           0.882550   \n",
       "6345              296         0.976898                299           0.977124   \n",
       "6346               78         1.000000                 78           1.000000   \n",
       "6347              234         0.646409                238           0.948207   \n",
       "\n",
       "      bit_score  thermo_index  ...  t_ogt  ogt_difference  \\\n",
       "0           131           875  ...   50.0            22.5   \n",
       "1           282         11324  ...   54.0            29.0   \n",
       "2            96           875  ...   50.0            22.0   \n",
       "3           175           875  ...   50.0            22.0   \n",
       "4            61          9827  ...   50.0            20.0   \n",
       "...         ...           ...  ...    ...             ...   \n",
       "6343        588         11707  ...   50.0            20.0   \n",
       "6344        619           875  ...   50.0            20.0   \n",
       "6345        436          5381  ...   45.0            21.5   \n",
       "6346        258         14963  ...   52.5            22.5   \n",
       "6347        347          7134  ...   52.5            24.5   \n",
       "\n",
       "                                          m_protein_seq  \\\n",
       "0     MAESGTSRRADHLVPVPGPDAEPPAVADELLRAVGRGDEQAFGRLY...   \n",
       "1     MARIALVDDDRNILTSVSMTLEAEGFEVETYNDGQSALDAFNKRMP...   \n",
       "2     MKDTVVFVTGAARGIGAHTARLAVARGARVALVGLEPHLLADLAAE...   \n",
       "3     MTSGLWERVLDGVWVTIQLLVLSALLATAVSFVVGIARTHRLWIVR...   \n",
       "4     MIISLRRGLRFIRFIVFFAALVYLFYHVLDLFNGWISPVDQYQMPT...   \n",
       "...                                                 ...   \n",
       "6343  MDNEKPIVDFSALRDELGKNYEPVQVMDGQAKVIKPEIVSKYSNDQ...   \n",
       "6344  MANPYLDDLRLAHVLADSADNLSMDRFGALDLEVSTKPDMTYVTES...   \n",
       "6345  MGHDHDHSHGTATTNRTKLAWAFGITFTILIAEVIGAILTNSLALL...   \n",
       "6346  MAKPALRKPKKKSNPLKAAKISYVDYKDTALLRKFISDRGKIRARR...   \n",
       "6347  MTTHVPGDALVSLNAVNKHYGQLHVLKDINLQVRKGEVVVVIGPSG...   \n",
       "\n",
       "                                          t_protein_seq  \\\n",
       "0     MPSQITESERIELAERFERDALPLLDQLYSAALRMTRNPADAEDLV...   \n",
       "1     MRVLLVEDDPNTSRSIEMMLTHANLNVYATDMGEEGIDLAKLYDYD...   \n",
       "2     MTPEQIFSGQTAIVTGGASGIGAATVEHIARRGGRVFSVDLSYDSP...   \n",
       "3     MAMSRRKRGQLARGIQYAILVIVVVVLALLADWGKIGKAFFDWEAA...   \n",
       "4     MKRMVWRTLKVFIIFIACTLLFYFGLRFMHLEYEQFHRYEPPEGPA...   \n",
       "...                                                 ...   \n",
       "6343  MKLEFPILQIVDENGNLVNENYRKAITEELVKKFYYHMIRIRTFDR...   \n",
       "6344  MTRPYADDIHLAKRLADTADAITVPRYSVRDLTVRTKPDRSPVTDA...   \n",
       "6345  MAHQHGPARPEHASGRYLKRLMAAFGIGLVFMVLEVVVGVLTGSLA...   \n",
       "6346  MAKQPPRKPKKKVCVFCQEKISYVDYKDTALLRKFISDRGKIRARR...   \n",
       "6347  MSDGLTLRGITKVLGGRTIVDGFDLEVRRGELVCLLGPSGCGKTTT...   \n",
       "\n",
       "                                         m_protein_desc  \\\n",
       "0                  ECF RNA polymerase sigma factor SigK   \n",
       "1               response regulator transcription factor   \n",
       "2                             SDR family oxidoreductase   \n",
       "3     ectoine/hydroxyectoine ABC transporter permeas...   \n",
       "4                                   YqzK family protein   \n",
       "...                                                 ...   \n",
       "6343  pyruvate dehydrogenase (acetyl-transferring) E...   \n",
       "6344                             histidinol phosphatase   \n",
       "6345    cation diffusion facilitator family transporter   \n",
       "6346                          30S ribosomal protein S18   \n",
       "6347     amino acid ABC transporter ATP-binding protein   \n",
       "\n",
       "                                         t_protein_desc  m_protein_len  \\\n",
       "0           sigma-70 family RNA polymerase sigma factor            206   \n",
       "1               response regulator transcription factor            233   \n",
       "2                             SDR family oxidoreductase            287   \n",
       "3                   amino acid ABC transporter permease            234   \n",
       "4                                   YqzK family protein             80   \n",
       "...                                                 ...            ...   \n",
       "6343  pyruvate dehydrogenase (acetyl-transferring) E...            370   \n",
       "6344                             histidinol-phosphatase            298   \n",
       "6345    cation diffusion facilitator family transporter            306   \n",
       "6346                          30S ribosomal protein S18             78   \n",
       "6347                ABC transporter ATP-binding protein            251   \n",
       "\n",
       "      t_protein_len  Jaccard_Score  protein_match  \n",
       "0               202            1.0            Yes  \n",
       "1               237            1.0            Yes  \n",
       "2               252            1.0            Yes  \n",
       "3               269            1.0            Yes  \n",
       "4                66            1.0            Yes  \n",
       "...             ...            ...            ...  \n",
       "6343            356            0.5            Yes  \n",
       "6344            260            1.0            Yes  \n",
       "6345            303            1.0            Yes  \n",
       "6346             78            1.0            Yes  \n",
       "6347            362            0.2             No  \n",
       "\n",
       "[6348 rows x 32 columns]"
      ]
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "99990cd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_fasta_from_dataframe(dataframe, output_file_a, output_file_b):\n",
    "    #adjust this to write function with BioPython\n",
    "    #separate functions for each of the input sequences\n",
    "    #in training, seq_a = meso and seq_b = thermo\n",
    "    \n",
    "    \n",
    "    #meso sequence to fasta\n",
    "    with open(output_file_a, 'w') as f:\n",
    "        for _, row in df.iterrows():\n",
    "            f.write('>{}\\n{}\\n'.format((row['prot_pair_index']), row['m_protein_seq']))\n",
    "    \n",
    "    #thermo sequence to fasta\n",
    "    with open(output_file_b, 'w') as f:\n",
    "        for _, row in df.iterrows():\n",
    "            f.write('>{}\\n{}\\n'.format((row['prot_pair_index']), (row['t_protein_seq'])))\n",
    "   \n",
    "    #return output files\n",
    "    return [output_file_a, output_file_b]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "9bd56759",
   "metadata": {},
   "outputs": [],
   "source": [
    "#getting an error in the try/except block\n",
    "\n",
    "\n",
    "from Bio import Seq, SeqIO\n",
    "\n",
    "def get_fasta_from_dataframe_biopython(dataframe, output_file:str):\n",
    "    # Create a list of SeqRecord objects\n",
    "    records = []\n",
    "    for _, row in dataframe.iterrows():\n",
    "        try:\n",
    "            seq = Seq(row['m_protein_seq'])\n",
    "        except Exception as e:\n",
    "            print(f\"Skipping row {row.name}: {str(e)}\")\n",
    "            continue\n",
    "        record = SeqIO.SeqRecord(\n",
    "            seq=seq,\n",
    "            id=row['meso_index'],\n",
    "            description=\"\"\n",
    "        )\n",
    "        records.append(record)\n",
    "    \n",
    "    # Write the records to a FASTA file\n",
    "    with open(output_file, 'w') as f:\n",
    "        SeqIO.write(records, f, 'fasta')\n",
    "    \n",
    "    return output_file\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "653eb96d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_protein_descriptors(fasta_file, descriptors=[]):\n",
    "    \n",
    "    \"\"\"\n",
    "    Generates features from a protein sequence\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    Fasta file with amino acid sequences.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    Vector of descriptors\n",
    "    \"\"\"\n",
    "    \n",
    "    #create iProtein object\n",
    "    protein = iFeatureOmegaCLI.iProtein(fasta_file)\n",
    "    \n",
    "    #not sure why we need this yet. Right now it is stored in local directory.\n",
    "    params = protein.import_parameters('protein_parameters.json')\n",
    "    \n",
    "    protein_descriptors = {}\n",
    "    \n",
    "    for descriptor in descriptors:\n",
    "        protein.get_descriptor(descriptor)\n",
    "        protein_descriptors.update({f'{descriptor}':protein.encodings})\n",
    "        \n",
    "    return protein_descriptors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "57eb54f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def compute_descriptor_difference(fasta_files, descriptors=[]):\n",
    "        \n",
    "#     desc_a = get_protein_descriptors(fasta_files[0], descriptors)\n",
    "#     desc_b = get_protein_descriptors(fasta_files[1], descriptors)\n",
    "    \n",
    "#     desc_dict = {}\n",
    "\n",
    "#     for key in desc_a:\n",
    "#         desc_dict[key] = desc_a[key] - desc_b[key]\n",
    "    \n",
    "#     return desc_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "id": "fe48f193",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_new_dataframe(dataframe, output_files, descriptors=[]):\n",
    "    \"\"\"\n",
    "    Creates new dataframe with descriptors added.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    Pandas dataframe, list of descriptors as strings, output file name.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    Dataframe including vector(s) of descriptors\n",
    "    \"\"\"\n",
    "\n",
    "    fasta_files = get_fasta_from_dataframe(dataframe, output_files[0], output_files[1])\n",
    "    \n",
    "    def compute_descriptor_difference(fasta_files, descriptors=[]):\n",
    "        \"\"\"\n",
    "        Generates dictionary of descriptors for each of the two input sequences.\n",
    "        Computes the difference between each instance of a descriptor.\n",
    "        \n",
    "        Parameters\n",
    "        ----------\n",
    "        List of two fasta files (str) and list of descriptors (str).\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        Dictionary with difference between descriptors for each of the \n",
    "        input sequences.\n",
    "        \"\"\"\n",
    "        desc_a = get_protein_descriptors(fasta_files[0], descriptors)\n",
    "        desc_b = get_protein_descriptors(fasta_files[1], descriptors)\n",
    "\n",
    "        feature_dict = {}\n",
    "\n",
    "        for key in desc_a:\n",
    "            feature_dict[key] = desc_a[key] - desc_b[key]\n",
    "\n",
    "        return feature_dict\n",
    "    \n",
    "    feature_dict = compute_descriptor_difference(fasta_files, descriptors)\n",
    "\n",
    "\n",
    "    df = dataframe.reset_index()\n",
    "\n",
    "    for desc in descriptors:\n",
    "\n",
    "        feature_dict[desc].index = feature_dict[desc].index.astype(int)\n",
    "        features = feature_dict[desc].reset_index()\n",
    "\n",
    "        df = pd.merge(\n",
    "            df,\n",
    "            features,\n",
    "            how='outer',\n",
    "            left_index=True,\n",
    "            right_index=True)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "id": "b90003a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File imported successfully.\n",
      "File imported successfully.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_x</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>...</th>\n",
       "      <th>AAC_T</th>\n",
       "      <th>AAC_V</th>\n",
       "      <th>AAC_W</th>\n",
       "      <th>AAC_Y</th>\n",
       "      <th>index</th>\n",
       "      <th>GAAC_alphatic</th>\n",
       "      <th>GAAC_aromatic</th>\n",
       "      <th>GAAC_postivecharge</th>\n",
       "      <th>GAAC_negativecharge</th>\n",
       "      <th>GAAC_uncharge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.287582</td>\n",
       "      <td>0.217822</td>\n",
       "      <td>0.215686</td>\n",
       "      <td>160</td>\n",
       "      <td>0.792079</td>\n",
       "      <td>152</td>\n",
       "      <td>0.737864</td>\n",
       "      <td>131</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000961</td>\n",
       "      <td>0.062242</td>\n",
       "      <td>-0.000192</td>\n",
       "      <td>-0.010478</td>\n",
       "      <td>48641291</td>\n",
       "      <td>0.065318</td>\n",
       "      <td>-0.025810</td>\n",
       "      <td>0.011583</td>\n",
       "      <td>-0.017639</td>\n",
       "      <td>-0.033452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.319635</td>\n",
       "      <td>0.295359</td>\n",
       "      <td>0.297872</td>\n",
       "      <td>218</td>\n",
       "      <td>0.919831</td>\n",
       "      <td>226</td>\n",
       "      <td>0.969957</td>\n",
       "      <td>282</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.016081</td>\n",
       "      <td>0.018254</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000507</td>\n",
       "      <td>92992745</td>\n",
       "      <td>-0.014143</td>\n",
       "      <td>0.009525</td>\n",
       "      <td>-0.010051</td>\n",
       "      <td>0.032651</td>\n",
       "      <td>-0.017982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0.279621</td>\n",
       "      <td>0.234127</td>\n",
       "      <td>0.218924</td>\n",
       "      <td>211</td>\n",
       "      <td>0.837302</td>\n",
       "      <td>210</td>\n",
       "      <td>0.731707</td>\n",
       "      <td>96</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.007743</td>\n",
       "      <td>0.011808</td>\n",
       "      <td>0.002516</td>\n",
       "      <td>-0.018293</td>\n",
       "      <td>157628663</td>\n",
       "      <td>0.087205</td>\n",
       "      <td>-0.020132</td>\n",
       "      <td>0.040650</td>\n",
       "      <td>0.006291</td>\n",
       "      <td>-0.114015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0.327273</td>\n",
       "      <td>0.200743</td>\n",
       "      <td>0.214712</td>\n",
       "      <td>166</td>\n",
       "      <td>0.617100</td>\n",
       "      <td>163</td>\n",
       "      <td>0.696581</td>\n",
       "      <td>175</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.011915</td>\n",
       "      <td>0.019668</td>\n",
       "      <td>0.018762</td>\n",
       "      <td>-0.015251</td>\n",
       "      <td>136708305</td>\n",
       "      <td>0.015966</td>\n",
       "      <td>0.014457</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>-0.011359</td>\n",
       "      <td>-0.020144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>0.338710</td>\n",
       "      <td>0.318182</td>\n",
       "      <td>0.287671</td>\n",
       "      <td>60</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>71</td>\n",
       "      <td>0.887500</td>\n",
       "      <td>61</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.005303</td>\n",
       "      <td>0.042045</td>\n",
       "      <td>0.022348</td>\n",
       "      <td>0.017045</td>\n",
       "      <td>133672542</td>\n",
       "      <td>0.091667</td>\n",
       "      <td>-0.027273</td>\n",
       "      <td>-0.084470</td>\n",
       "      <td>-0.031061</td>\n",
       "      <td>0.051136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6343</th>\n",
       "      <td>6343</td>\n",
       "      <td>6495</td>\n",
       "      <td>0.387574</td>\n",
       "      <td>0.367978</td>\n",
       "      <td>0.360882</td>\n",
       "      <td>336</td>\n",
       "      <td>0.943820</td>\n",
       "      <td>337</td>\n",
       "      <td>0.910811</td>\n",
       "      <td>588</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.004616</td>\n",
       "      <td>0.011388</td>\n",
       "      <td>-0.006043</td>\n",
       "      <td>0.003811</td>\n",
       "      <td>55489429</td>\n",
       "      <td>0.004783</td>\n",
       "      <td>-0.017871</td>\n",
       "      <td>-0.019040</td>\n",
       "      <td>0.008624</td>\n",
       "      <td>0.023504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6344</th>\n",
       "      <td>6344</td>\n",
       "      <td>6496</td>\n",
       "      <td>0.488462</td>\n",
       "      <td>0.488462</td>\n",
       "      <td>0.455197</td>\n",
       "      <td>258</td>\n",
       "      <td>0.992308</td>\n",
       "      <td>263</td>\n",
       "      <td>0.882550</td>\n",
       "      <td>619</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.011693</td>\n",
       "      <td>-0.033867</td>\n",
       "      <td>-0.001058</td>\n",
       "      <td>0.000903</td>\n",
       "      <td>172293605</td>\n",
       "      <td>-0.039210</td>\n",
       "      <td>0.011306</td>\n",
       "      <td>-0.000800</td>\n",
       "      <td>0.021631</td>\n",
       "      <td>0.007073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6345</th>\n",
       "      <td>6345</td>\n",
       "      <td>6497</td>\n",
       "      <td>0.383838</td>\n",
       "      <td>0.376238</td>\n",
       "      <td>0.374384</td>\n",
       "      <td>296</td>\n",
       "      <td>0.976898</td>\n",
       "      <td>299</td>\n",
       "      <td>0.977124</td>\n",
       "      <td>436</td>\n",
       "      <td>...</td>\n",
       "      <td>0.029023</td>\n",
       "      <td>-0.056753</td>\n",
       "      <td>-0.003397</td>\n",
       "      <td>-0.013266</td>\n",
       "      <td>47082975</td>\n",
       "      <td>-0.035009</td>\n",
       "      <td>-0.017019</td>\n",
       "      <td>0.008801</td>\n",
       "      <td>0.002362</td>\n",
       "      <td>0.040866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6346</th>\n",
       "      <td>6346</td>\n",
       "      <td>6498</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>78</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>78</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>258</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.051282</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>154513027</td>\n",
       "      <td>0.064103</td>\n",
       "      <td>-0.012821</td>\n",
       "      <td>0.012821</td>\n",
       "      <td>-0.025641</td>\n",
       "      <td>-0.038462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6347</th>\n",
       "      <td>6347</td>\n",
       "      <td>6499</td>\n",
       "      <td>0.343220</td>\n",
       "      <td>0.223757</td>\n",
       "      <td>0.264274</td>\n",
       "      <td>234</td>\n",
       "      <td>0.646409</td>\n",
       "      <td>238</td>\n",
       "      <td>0.948207</td>\n",
       "      <td>347</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.017268</td>\n",
       "      <td>-0.016740</td>\n",
       "      <td>-0.002762</td>\n",
       "      <td>-0.018115</td>\n",
       "      <td>90211763</td>\n",
       "      <td>0.012778</td>\n",
       "      <td>-0.012646</td>\n",
       "      <td>-0.005745</td>\n",
       "      <td>0.001959</td>\n",
       "      <td>0.003654</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6348 rows × 60 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      index_x  Unnamed: 0  local_gap_compressed_percent_id  \\\n",
       "0           0           0                         0.287582   \n",
       "1           1           1                         0.319635   \n",
       "2           2           2                         0.279621   \n",
       "3           3           3                         0.327273   \n",
       "4           4           4                         0.338710   \n",
       "...       ...         ...                              ...   \n",
       "6343     6343        6495                         0.387574   \n",
       "6344     6344        6496                         0.488462   \n",
       "6345     6345        6497                         0.383838   \n",
       "6346     6346        6498                         0.769231   \n",
       "6347     6347        6499                         0.343220   \n",
       "\n",
       "      scaled_local_query_percent_id  scaled_local_symmetric_percent_id  \\\n",
       "0                          0.217822                           0.215686   \n",
       "1                          0.295359                           0.297872   \n",
       "2                          0.234127                           0.218924   \n",
       "3                          0.200743                           0.214712   \n",
       "4                          0.318182                           0.287671   \n",
       "...                             ...                                ...   \n",
       "6343                       0.367978                           0.360882   \n",
       "6344                       0.488462                           0.455197   \n",
       "6345                       0.376238                           0.374384   \n",
       "6346                       0.769231                           0.769231   \n",
       "6347                       0.223757                           0.264274   \n",
       "\n",
       "      query_align_len  query_align_cov  subject_align_len  subject_align_cov  \\\n",
       "0                 160         0.792079                152           0.737864   \n",
       "1                 218         0.919831                226           0.969957   \n",
       "2                 211         0.837302                210           0.731707   \n",
       "3                 166         0.617100                163           0.696581   \n",
       "4                  60         0.909091                 71           0.887500   \n",
       "...               ...              ...                ...                ...   \n",
       "6343              336         0.943820                337           0.910811   \n",
       "6344              258         0.992308                263           0.882550   \n",
       "6345              296         0.976898                299           0.977124   \n",
       "6346               78         1.000000                 78           1.000000   \n",
       "6347              234         0.646409                238           0.948207   \n",
       "\n",
       "      bit_score  ...     AAC_T     AAC_V     AAC_W     AAC_Y      index  \\\n",
       "0           131  ... -0.000961  0.062242 -0.000192 -0.010478   48641291   \n",
       "1           282  ... -0.016081  0.018254  0.000072  0.000507   92992745   \n",
       "2            96  ... -0.007743  0.011808  0.002516 -0.018293  157628663   \n",
       "3           175  ... -0.011915  0.019668  0.018762 -0.015251  136708305   \n",
       "4            61  ... -0.005303  0.042045  0.022348  0.017045  133672542   \n",
       "...         ...  ...       ...       ...       ...       ...        ...   \n",
       "6343        588  ... -0.004616  0.011388 -0.006043  0.003811   55489429   \n",
       "6344        619  ... -0.011693 -0.033867 -0.001058  0.000903  172293605   \n",
       "6345        436  ...  0.029023 -0.056753 -0.003397 -0.013266   47082975   \n",
       "6346        258  ... -0.051282  0.000000  0.000000  0.000000  154513027   \n",
       "6347        347  ... -0.017268 -0.016740 -0.002762 -0.018115   90211763   \n",
       "\n",
       "      GAAC_alphatic  GAAC_aromatic  GAAC_postivecharge  GAAC_negativecharge  \\\n",
       "0          0.065318      -0.025810            0.011583            -0.017639   \n",
       "1         -0.014143       0.009525           -0.010051             0.032651   \n",
       "2          0.087205      -0.020132            0.040650             0.006291   \n",
       "3          0.015966       0.014457            0.001080            -0.011359   \n",
       "4          0.091667      -0.027273           -0.084470            -0.031061   \n",
       "...             ...            ...                 ...                  ...   \n",
       "6343       0.004783      -0.017871           -0.019040             0.008624   \n",
       "6344      -0.039210       0.011306           -0.000800             0.021631   \n",
       "6345      -0.035009      -0.017019            0.008801             0.002362   \n",
       "6346       0.064103      -0.012821            0.012821            -0.025641   \n",
       "6347       0.012778      -0.012646           -0.005745             0.001959   \n",
       "\n",
       "      GAAC_uncharge  \n",
       "0         -0.033452  \n",
       "1         -0.017982  \n",
       "2         -0.114015  \n",
       "3         -0.020144  \n",
       "4          0.051136  \n",
       "...             ...  \n",
       "6343       0.023504  \n",
       "6344       0.007073  \n",
       "6345       0.040866  \n",
       "6346      -0.038462  \n",
       "6347       0.003654  \n",
       "\n",
       "[6348 rows x 60 columns]"
      ]
     },
     "execution_count": 185,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = create_new_dataframe(df, ['seq_50k_a.fasta', 'seq_50k_b.fasta'], descriptors=['AAC', 'GAAC'])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "id": "49cf5076",
   "metadata": {},
   "outputs": [],
   "source": [
    "#this function is only necessary if we get a fasta file with a descriptor generated from component 3\n",
    "\n",
    "def remove_fasta_description(filename:str):\n",
    "    \n",
    "    \"\"\"\n",
    "    Removes description from fasta file so that iProtein can read the input.\n",
    "    Might not be necessary in our current format.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    Fasta file with protein sequences.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    Fasta file with protein sequences, sans description.\n",
    "    \"\"\"\n",
    "    \n",
    "    #assign unwanted string to object\n",
    "    string_to_remove = \"<unknown description>\"\n",
    "\n",
    "    #open file\n",
    "    with open(filename, \"r\") as file:\n",
    "        content = file.read()\n",
    "    \n",
    "    # Remove the string\n",
    "    new_content = content.replace(string_to_remove, \"\")\n",
    "\n",
    "    #overwrite file without string\n",
    "    with open(filename, \"w\") as file:\n",
    "        seq = file.write(new_content)\n",
    "        \n",
    "    return seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "e93e310e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#combine both of the above functions\n",
    "\n",
    "def fasta_to_descriptors(fasta:str, descriptors=[]):\n",
    "\n",
    "    #remove description from fasta file\n",
    "    remove_fasta_description(fasta)\n",
    "    \n",
    "    #return protein descriptors\n",
    "    return get_protein_descriptors(fasta, descriptors=descriptors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "114c3d10",
   "metadata": {},
   "outputs": [],
   "source": [
    "#need to figure out how to pass fasta for both descriptors and append to one dataframe\n",
    "#set this back to what it was originally. Gonna calculate difference for each descriptors\n",
    "\n",
    "# def create_new_dataframe(dataframe, fasta_files:list, descriptors=[]):\n",
    "#     \"\"\"\n",
    "#     Creates new dataframe with descriptors added.\n",
    "    \n",
    "#     Parameters\n",
    "#     ----------\n",
    "#     Pandas dataframe, list of descriptors as strings, output file name.\n",
    "\n",
    "#     Returns\n",
    "#     -------\n",
    "#     Dataframe including vector(s) of descriptors\n",
    "#     \"\"\"\n",
    "    \n",
    "#     #descriptors for meso sequence\n",
    "#     fasta = get_fasta_from_dataframe(dataframe, fasta_files[0], fasta_files[1])\n",
    "    \n",
    "#     #create feature dictionary for sequence a\n",
    "#     feature_dict_a = get_protein_descriptors(fasta[0], 'Meso', descriptors)\n",
    "#     print(feature_dict_a)\n",
    "    \n",
    "#     #create feature dictionary for sequence b\n",
    "#     feature_dict_b = get_protein_descriptors(fasta[1], 'Thermo', descriptors)\n",
    "    \n",
    "#     df = dataframe.reset_index()\n",
    "    \n",
    "#     #merge meso sequences\n",
    "#     for desc in descriptors:\n",
    "        \n",
    "#         feature_dict_a[desc+'_Meso'].index = feature_dict_a[desc+'_Meso'].index.astype(int)\n",
    "#         features_a = feature_dict_a[desc+'_Meso'].reset_index()\n",
    "        \n",
    "#     df = pd.merge(df, features_a, how='outer', left_index=True, right_index=True)\n",
    "    \n",
    "#     #merge thermo sequences\n",
    "# #     for desc in descriptors:\n",
    "        \n",
    "# #         feature_dict_b[desc+'_Thermo'].index = feature_dict_b[desc+'_Thermo'].index.astype(int)\n",
    "# #         features_b = feature_dict_b[desc+'_Thermo'].reset_index()\n",
    "        \n",
    "# #     df = pd.merge(df, features_b, how='outer', left_index=True, right_index=True)\n",
    "        \n",
    "#     return df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49c582c7",
   "metadata": {},
   "source": [
    "Let's see how long generating some of the descriptors takes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "id": "f073f547",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_x</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>...</th>\n",
       "      <th>AAC_T</th>\n",
       "      <th>AAC_V</th>\n",
       "      <th>AAC_W</th>\n",
       "      <th>AAC_Y</th>\n",
       "      <th>index</th>\n",
       "      <th>GAAC_alphatic</th>\n",
       "      <th>GAAC_aromatic</th>\n",
       "      <th>GAAC_postivecharge</th>\n",
       "      <th>GAAC_negativecharge</th>\n",
       "      <th>GAAC_uncharge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.287582</td>\n",
       "      <td>0.217822</td>\n",
       "      <td>0.215686</td>\n",
       "      <td>160</td>\n",
       "      <td>0.792079</td>\n",
       "      <td>152</td>\n",
       "      <td>0.737864</td>\n",
       "      <td>131</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000961</td>\n",
       "      <td>0.062242</td>\n",
       "      <td>-0.000192</td>\n",
       "      <td>-0.010478</td>\n",
       "      <td>48641291</td>\n",
       "      <td>0.065318</td>\n",
       "      <td>-0.025810</td>\n",
       "      <td>0.011583</td>\n",
       "      <td>-0.017639</td>\n",
       "      <td>-0.033452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.319635</td>\n",
       "      <td>0.295359</td>\n",
       "      <td>0.297872</td>\n",
       "      <td>218</td>\n",
       "      <td>0.919831</td>\n",
       "      <td>226</td>\n",
       "      <td>0.969957</td>\n",
       "      <td>282</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.016081</td>\n",
       "      <td>0.018254</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000507</td>\n",
       "      <td>92992745</td>\n",
       "      <td>-0.014143</td>\n",
       "      <td>0.009525</td>\n",
       "      <td>-0.010051</td>\n",
       "      <td>0.032651</td>\n",
       "      <td>-0.017982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0.279621</td>\n",
       "      <td>0.234127</td>\n",
       "      <td>0.218924</td>\n",
       "      <td>211</td>\n",
       "      <td>0.837302</td>\n",
       "      <td>210</td>\n",
       "      <td>0.731707</td>\n",
       "      <td>96</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.007743</td>\n",
       "      <td>0.011808</td>\n",
       "      <td>0.002516</td>\n",
       "      <td>-0.018293</td>\n",
       "      <td>157628663</td>\n",
       "      <td>0.087205</td>\n",
       "      <td>-0.020132</td>\n",
       "      <td>0.040650</td>\n",
       "      <td>0.006291</td>\n",
       "      <td>-0.114015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0.327273</td>\n",
       "      <td>0.200743</td>\n",
       "      <td>0.214712</td>\n",
       "      <td>166</td>\n",
       "      <td>0.617100</td>\n",
       "      <td>163</td>\n",
       "      <td>0.696581</td>\n",
       "      <td>175</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.011915</td>\n",
       "      <td>0.019668</td>\n",
       "      <td>0.018762</td>\n",
       "      <td>-0.015251</td>\n",
       "      <td>136708305</td>\n",
       "      <td>0.015966</td>\n",
       "      <td>0.014457</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>-0.011359</td>\n",
       "      <td>-0.020144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>0.338710</td>\n",
       "      <td>0.318182</td>\n",
       "      <td>0.287671</td>\n",
       "      <td>60</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>71</td>\n",
       "      <td>0.887500</td>\n",
       "      <td>61</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.005303</td>\n",
       "      <td>0.042045</td>\n",
       "      <td>0.022348</td>\n",
       "      <td>0.017045</td>\n",
       "      <td>133672542</td>\n",
       "      <td>0.091667</td>\n",
       "      <td>-0.027273</td>\n",
       "      <td>-0.084470</td>\n",
       "      <td>-0.031061</td>\n",
       "      <td>0.051136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6343</th>\n",
       "      <td>6343</td>\n",
       "      <td>6495</td>\n",
       "      <td>0.387574</td>\n",
       "      <td>0.367978</td>\n",
       "      <td>0.360882</td>\n",
       "      <td>336</td>\n",
       "      <td>0.943820</td>\n",
       "      <td>337</td>\n",
       "      <td>0.910811</td>\n",
       "      <td>588</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.004616</td>\n",
       "      <td>0.011388</td>\n",
       "      <td>-0.006043</td>\n",
       "      <td>0.003811</td>\n",
       "      <td>55489429</td>\n",
       "      <td>0.004783</td>\n",
       "      <td>-0.017871</td>\n",
       "      <td>-0.019040</td>\n",
       "      <td>0.008624</td>\n",
       "      <td>0.023504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6344</th>\n",
       "      <td>6344</td>\n",
       "      <td>6496</td>\n",
       "      <td>0.488462</td>\n",
       "      <td>0.488462</td>\n",
       "      <td>0.455197</td>\n",
       "      <td>258</td>\n",
       "      <td>0.992308</td>\n",
       "      <td>263</td>\n",
       "      <td>0.882550</td>\n",
       "      <td>619</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.011693</td>\n",
       "      <td>-0.033867</td>\n",
       "      <td>-0.001058</td>\n",
       "      <td>0.000903</td>\n",
       "      <td>172293605</td>\n",
       "      <td>-0.039210</td>\n",
       "      <td>0.011306</td>\n",
       "      <td>-0.000800</td>\n",
       "      <td>0.021631</td>\n",
       "      <td>0.007073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6345</th>\n",
       "      <td>6345</td>\n",
       "      <td>6497</td>\n",
       "      <td>0.383838</td>\n",
       "      <td>0.376238</td>\n",
       "      <td>0.374384</td>\n",
       "      <td>296</td>\n",
       "      <td>0.976898</td>\n",
       "      <td>299</td>\n",
       "      <td>0.977124</td>\n",
       "      <td>436</td>\n",
       "      <td>...</td>\n",
       "      <td>0.029023</td>\n",
       "      <td>-0.056753</td>\n",
       "      <td>-0.003397</td>\n",
       "      <td>-0.013266</td>\n",
       "      <td>47082975</td>\n",
       "      <td>-0.035009</td>\n",
       "      <td>-0.017019</td>\n",
       "      <td>0.008801</td>\n",
       "      <td>0.002362</td>\n",
       "      <td>0.040866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6346</th>\n",
       "      <td>6346</td>\n",
       "      <td>6498</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>78</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>78</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>258</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.051282</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>154513027</td>\n",
       "      <td>0.064103</td>\n",
       "      <td>-0.012821</td>\n",
       "      <td>0.012821</td>\n",
       "      <td>-0.025641</td>\n",
       "      <td>-0.038462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6347</th>\n",
       "      <td>6347</td>\n",
       "      <td>6499</td>\n",
       "      <td>0.343220</td>\n",
       "      <td>0.223757</td>\n",
       "      <td>0.264274</td>\n",
       "      <td>234</td>\n",
       "      <td>0.646409</td>\n",
       "      <td>238</td>\n",
       "      <td>0.948207</td>\n",
       "      <td>347</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.017268</td>\n",
       "      <td>-0.016740</td>\n",
       "      <td>-0.002762</td>\n",
       "      <td>-0.018115</td>\n",
       "      <td>90211763</td>\n",
       "      <td>0.012778</td>\n",
       "      <td>-0.012646</td>\n",
       "      <td>-0.005745</td>\n",
       "      <td>0.001959</td>\n",
       "      <td>0.003654</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6348 rows × 60 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      index_x  Unnamed: 0  local_gap_compressed_percent_id  \\\n",
       "0           0           0                         0.287582   \n",
       "1           1           1                         0.319635   \n",
       "2           2           2                         0.279621   \n",
       "3           3           3                         0.327273   \n",
       "4           4           4                         0.338710   \n",
       "...       ...         ...                              ...   \n",
       "6343     6343        6495                         0.387574   \n",
       "6344     6344        6496                         0.488462   \n",
       "6345     6345        6497                         0.383838   \n",
       "6346     6346        6498                         0.769231   \n",
       "6347     6347        6499                         0.343220   \n",
       "\n",
       "      scaled_local_query_percent_id  scaled_local_symmetric_percent_id  \\\n",
       "0                          0.217822                           0.215686   \n",
       "1                          0.295359                           0.297872   \n",
       "2                          0.234127                           0.218924   \n",
       "3                          0.200743                           0.214712   \n",
       "4                          0.318182                           0.287671   \n",
       "...                             ...                                ...   \n",
       "6343                       0.367978                           0.360882   \n",
       "6344                       0.488462                           0.455197   \n",
       "6345                       0.376238                           0.374384   \n",
       "6346                       0.769231                           0.769231   \n",
       "6347                       0.223757                           0.264274   \n",
       "\n",
       "      query_align_len  query_align_cov  subject_align_len  subject_align_cov  \\\n",
       "0                 160         0.792079                152           0.737864   \n",
       "1                 218         0.919831                226           0.969957   \n",
       "2                 211         0.837302                210           0.731707   \n",
       "3                 166         0.617100                163           0.696581   \n",
       "4                  60         0.909091                 71           0.887500   \n",
       "...               ...              ...                ...                ...   \n",
       "6343              336         0.943820                337           0.910811   \n",
       "6344              258         0.992308                263           0.882550   \n",
       "6345              296         0.976898                299           0.977124   \n",
       "6346               78         1.000000                 78           1.000000   \n",
       "6347              234         0.646409                238           0.948207   \n",
       "\n",
       "      bit_score  ...     AAC_T     AAC_V     AAC_W     AAC_Y      index  \\\n",
       "0           131  ... -0.000961  0.062242 -0.000192 -0.010478   48641291   \n",
       "1           282  ... -0.016081  0.018254  0.000072  0.000507   92992745   \n",
       "2            96  ... -0.007743  0.011808  0.002516 -0.018293  157628663   \n",
       "3           175  ... -0.011915  0.019668  0.018762 -0.015251  136708305   \n",
       "4            61  ... -0.005303  0.042045  0.022348  0.017045  133672542   \n",
       "...         ...  ...       ...       ...       ...       ...        ...   \n",
       "6343        588  ... -0.004616  0.011388 -0.006043  0.003811   55489429   \n",
       "6344        619  ... -0.011693 -0.033867 -0.001058  0.000903  172293605   \n",
       "6345        436  ...  0.029023 -0.056753 -0.003397 -0.013266   47082975   \n",
       "6346        258  ... -0.051282  0.000000  0.000000  0.000000  154513027   \n",
       "6347        347  ... -0.017268 -0.016740 -0.002762 -0.018115   90211763   \n",
       "\n",
       "      GAAC_alphatic  GAAC_aromatic  GAAC_postivecharge  GAAC_negativecharge  \\\n",
       "0          0.065318      -0.025810            0.011583            -0.017639   \n",
       "1         -0.014143       0.009525           -0.010051             0.032651   \n",
       "2          0.087205      -0.020132            0.040650             0.006291   \n",
       "3          0.015966       0.014457            0.001080            -0.011359   \n",
       "4          0.091667      -0.027273           -0.084470            -0.031061   \n",
       "...             ...            ...                 ...                  ...   \n",
       "6343       0.004783      -0.017871           -0.019040             0.008624   \n",
       "6344      -0.039210       0.011306           -0.000800             0.021631   \n",
       "6345      -0.035009      -0.017019            0.008801             0.002362   \n",
       "6346       0.064103      -0.012821            0.012821            -0.025641   \n",
       "6347       0.012778      -0.012646           -0.005745             0.001959   \n",
       "\n",
       "      GAAC_uncharge  \n",
       "0         -0.033452  \n",
       "1         -0.017982  \n",
       "2         -0.114015  \n",
       "3         -0.020144  \n",
       "4          0.051136  \n",
       "...             ...  \n",
       "6343       0.023504  \n",
       "6344       0.007073  \n",
       "6345       0.040866  \n",
       "6346      -0.038462  \n",
       "6347       0.003654  \n",
       "\n",
       "[6348 rows x 60 columns]"
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "id": "dd299dfd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Yes    5243\n",
       "No     1105\n",
       "Name: protein_match, dtype: int64"
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df\n",
    "df['protein_match'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "id": "38c1120f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "        ----- Available feature types ------        \n",
      "        \n",
      "        AAC                                                Amino acid composition\n",
      "        EAAC                                               Enhanced amino acid composition\n",
      "        CKSAAP type 1                                      Composition of k-spaced amino acid pairs type 1 - normalized\n",
      "        CKSAAP type 2                                      Composition of k-spaced amino acid pairs type 2 - raw count\n",
      "        DPC type 1                                         Dipeptide composition type 1 - normalized\n",
      "        DPC type 2                                         Dipeptide composition type 2 - raw count\n",
      "        TPC type 1                                         Tripeptide composition type 1 - normalized\n",
      "        TPC type 2                                         Tripeptide composition type 1 - raw count\n",
      "        CTDC                                               Composition\n",
      "        CTDT                                               Transition\n",
      "        CTDD                                               Distribution\n",
      "        CTriad                                             Conjoint triad\n",
      "        KSCTriad                                           Conjoint k-spaced triad\n",
      "        ASDC                                               Adaptive skip dipeptide composition\n",
      "        DistancePair                                       PseAAC of distance-pairs and reduced alphabe\n",
      "        GAAC                                               Grouped amino acid composition\n",
      "        EGAAC                                              Enhanced grouped amino acid composition\n",
      "        CKSAAGP type 1                                     Composition of k-spaced amino acid group pairs type 1- normalized\n",
      "        CKSAAGP type 2                                     Composition of k-spaced amino acid group pairs type 2- raw count\n",
      "        GDPC type 1                                        Grouped dipeptide composition type 1 - normalized\n",
      "        GDPC type 2                                        Grouped dipeptide composition type 2 - raw count\n",
      "        GTPC type 1                                        Grouped tripeptide composition type 1 - normalized\n",
      "        GTPC type 2                                        Grouped tripeptide composition type 1 - raw count\n",
      "        Moran                                              Moran\n",
      "        Geary                                              Geary\n",
      "        NMBroto                                            Normalized Moreau-Broto\n",
      "        AC                                                 Auto covariance\n",
      "        CC                                                 Cross covariance\n",
      "        ACC                                                Auto-cross covariance\n",
      "        SOCNumber                                          Sequence-order-coupling number\n",
      "        QSOrder                                            Quasi-sequence-order descriptors\n",
      "        PAAC                                               Pseudo-amino acid composition\n",
      "        APAAC                                              Amphiphilic PAAC\n",
      "        PseKRAAC type 1                                    Pseudo K-tuple reduced amino acids composition type 1\n",
      "        PseKRAAC type 2                                    Pseudo K-tuple reduced amino acids composition type 2\n",
      "        PseKRAAC type 3A                                   Pseudo K-tuple reduced amino acids composition type 3A\n",
      "        PseKRAAC type 3B                                   Pseudo K-tuple reduced amino acids composition type 3B\n",
      "        PseKRAAC type 4                                    Pseudo K-tuple reduced amino acids composition type 4\n",
      "        PseKRAAC type 5                                    Pseudo K-tuple reduced amino acids composition type 5\n",
      "        PseKRAAC type 6A                                   Pseudo K-tuple reduced amino acids composition type 6A\n",
      "        PseKRAAC type 6B                                   Pseudo K-tuple reduced amino acids composition type 6B\n",
      "        PseKRAAC type 6C                                   Pseudo K-tuple reduced amino acids composition type 6C\n",
      "        PseKRAAC type 7                                    Pseudo K-tuple reduced amino acids composition type 7\n",
      "        PseKRAAC type 8                                    Pseudo K-tuple reduced amino acids composition type 8\n",
      "        PseKRAAC type 9                                    Pseudo K-tuple reduced amino acids composition type 9\n",
      "        PseKRAAC type 10                                   Pseudo K-tuple reduced amino acids composition type 10\n",
      "        PseKRAAC type 11                                   Pseudo K-tuple reduced amino acids composition type 11\n",
      "        PseKRAAC type 12                                   Pseudo K-tuple reduced amino acids composition type 12\n",
      "        PseKRAAC type 13                                   Pseudo K-tuple reduced amino acids composition type 13\n",
      "        PseKRAAC type 14                                   Pseudo K-tuple reduced amino acids composition type 14\n",
      "        PseKRAAC type 15                                   Pseudo K-tuple reduced amino acids composition type 15\n",
      "        PseKRAAC type 16                                   Pseudo K-tuple reduced amino acids composition type 16\n",
      "        binary                                             Binary\n",
      "        binary_6bit                                        Binary\n",
      "        binary_5bit type 1                                 Binary\n",
      "        binary_5bit type 2                                 Binary\n",
      "        binary_3bit type 1                                 Binary\n",
      "        binary_3bit type 2                                 Binary\n",
      "        binary_3bit type 3                                 Binary\n",
      "        binary_3bit type 4                                 Binary\n",
      "        binary_3bit type 5                                 Binary\n",
      "        binary_3bit type 6                                 Binary\n",
      "        binary_3bit type 7                                 Binary\n",
      "        AESNN3                                             Learn from alignments\n",
      "        OPF_10bit                                          Overlapping property features - 10 bit\n",
      "        OPF_7bit type 1                                    Overlapping property features - 7 bit type 1\n",
      "        OPF_7bit type 2                                    Overlapping property features - 7 bit type 2\n",
      "        OPF_7bit type 3                                    Overlapping property features - 7 bit type 3\n",
      "        AAIndex                                            AAIndex\n",
      "        BLOSUM62                                           BLOSUM62\n",
      "        ZScale                                             Z-Scales index\n",
      "        KNN                                                K-nearest neighbor\n",
      "\n",
      "        Note: the first column is the names of availables feature types while the second column is description.  \n",
      "        \n",
      "        \n"
     ]
    }
   ],
   "source": [
    "#iFeature properties\n",
    "\n",
    "protein = iFeatureOmegaCLI.iProtein('meso_50k.fasta')\n",
    "protein.display_feature_types()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "aabaa4a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "this list comes from a combination of reading through the features and determining which might be useful\n",
    "and timing some of the feature generations. those that took more than 30ish seconds were eliminated\n",
    "Also removed those that have really high dimensionality (>4000)\n",
    "\"\"\"\n",
    "\n",
    "feature_list = ['AAC', 'GAAC', 'DistancePair',\n",
    "               'DPC type 1', 'CTDC', 'CTDT', 'CTDD', 'CTriad',\n",
    "                'CKSAAGP type 1', 'PseKRAAC type 1', 'PseKRAAC type 2', 'PseKRAAC type 3A',\n",
    "                'PseKRAAC type 3B', 'PseKRAAC type 4', 'PseKRAAC type 5',\n",
    "                'APAAC', 'QSOrder']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "id": "1aa26a7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "id": "7dda7ea4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def time_feature_generation(dataframe, file_name:str, descriptor:str):\n",
    "    \n",
    "    \"\"\"\n",
    "    Times how long it takes to generate a specific descriptor.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    Pandas dataframe, fasta file name, descriptor name (string).\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    Time to generate descriptor.\n",
    "    \"\"\"\n",
    "\n",
    "    # Record the current time\n",
    "    start_time = time.time()\n",
    "\n",
    "    # Code to be timed goes here\n",
    "    create_new_dataframe(dataframe, file_name, descriptors=[descriptor])\n",
    "\n",
    "    # Record the time again and calculate the elapsed time\n",
    "    end_time = time.time()\n",
    "    elapsed_time = end_time - start_time\n",
    "\n",
    "    # Print the elapsed time\n",
    "    return elapsed_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "id": "4e17abbc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File imported successfully.\n",
      "File imported successfully.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-184-97c201eec18b>:50: FutureWarning: Passing 'suffixes' which cause duplicate columns {'index_x'} in the result is deprecated and will raise a MergeError in a future version.\n",
      "  df = pd.merge(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "11.930752992630005"
      ]
     },
     "execution_count": 195,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "time_feature_generation(df, 'meso_50k.fasta', 'QSOrder')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "id": "3faadf61",
   "metadata": {},
   "outputs": [],
   "source": [
    "def time_feature_dict(dataframe, file_name:str, feature_list):\n",
    "    \n",
    "    \"\"\"\n",
    "    Iterates through dictionary of protein descriptors and returns the time it takes to run the code.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    Pandas dataframe, fasta file name, descriptors (list of strings).\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    Dictionary of descriptors and time to generate them.\n",
    "    \"\"\"\n",
    "    \n",
    "    protein = iFeatureOmegaCLI.iProtein(file_name)\n",
    "    \n",
    "    time_dict = {}\n",
    "    \n",
    "    for feature in feature_list:\n",
    "        time_dict[feature] = time_feature_generation(dataframe, file_name, feature)\n",
    "    \n",
    "    return time_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "id": "2c9fdb81",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# time_feature_dict(df, 'meso_50k.fasta', feature_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "id": "beee8301",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       1.0\n",
       "1       1.0\n",
       "2       1.0\n",
       "3       1.0\n",
       "4       1.0\n",
       "       ... \n",
       "6343    0.5\n",
       "6344    1.0\n",
       "6345    1.0\n",
       "6346    1.0\n",
       "6347    0.2\n",
       "Name: Jaccard_Score, Length: 6348, dtype: float64"
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Jaccard_Score']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "id": "7fbe24b5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_x</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>...</th>\n",
       "      <th>AAC_T</th>\n",
       "      <th>AAC_V</th>\n",
       "      <th>AAC_W</th>\n",
       "      <th>AAC_Y</th>\n",
       "      <th>index</th>\n",
       "      <th>GAAC_alphatic</th>\n",
       "      <th>GAAC_aromatic</th>\n",
       "      <th>GAAC_postivecharge</th>\n",
       "      <th>GAAC_negativecharge</th>\n",
       "      <th>GAAC_uncharge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.287582</td>\n",
       "      <td>0.217822</td>\n",
       "      <td>0.215686</td>\n",
       "      <td>160</td>\n",
       "      <td>0.792079</td>\n",
       "      <td>152</td>\n",
       "      <td>0.737864</td>\n",
       "      <td>131</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000961</td>\n",
       "      <td>0.062242</td>\n",
       "      <td>-0.000192</td>\n",
       "      <td>-0.010478</td>\n",
       "      <td>48641291</td>\n",
       "      <td>0.065318</td>\n",
       "      <td>-0.025810</td>\n",
       "      <td>0.011583</td>\n",
       "      <td>-0.017639</td>\n",
       "      <td>-0.033452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.319635</td>\n",
       "      <td>0.295359</td>\n",
       "      <td>0.297872</td>\n",
       "      <td>218</td>\n",
       "      <td>0.919831</td>\n",
       "      <td>226</td>\n",
       "      <td>0.969957</td>\n",
       "      <td>282</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.016081</td>\n",
       "      <td>0.018254</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000507</td>\n",
       "      <td>92992745</td>\n",
       "      <td>-0.014143</td>\n",
       "      <td>0.009525</td>\n",
       "      <td>-0.010051</td>\n",
       "      <td>0.032651</td>\n",
       "      <td>-0.017982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0.279621</td>\n",
       "      <td>0.234127</td>\n",
       "      <td>0.218924</td>\n",
       "      <td>211</td>\n",
       "      <td>0.837302</td>\n",
       "      <td>210</td>\n",
       "      <td>0.731707</td>\n",
       "      <td>96</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.007743</td>\n",
       "      <td>0.011808</td>\n",
       "      <td>0.002516</td>\n",
       "      <td>-0.018293</td>\n",
       "      <td>157628663</td>\n",
       "      <td>0.087205</td>\n",
       "      <td>-0.020132</td>\n",
       "      <td>0.040650</td>\n",
       "      <td>0.006291</td>\n",
       "      <td>-0.114015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0.327273</td>\n",
       "      <td>0.200743</td>\n",
       "      <td>0.214712</td>\n",
       "      <td>166</td>\n",
       "      <td>0.617100</td>\n",
       "      <td>163</td>\n",
       "      <td>0.696581</td>\n",
       "      <td>175</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.011915</td>\n",
       "      <td>0.019668</td>\n",
       "      <td>0.018762</td>\n",
       "      <td>-0.015251</td>\n",
       "      <td>136708305</td>\n",
       "      <td>0.015966</td>\n",
       "      <td>0.014457</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>-0.011359</td>\n",
       "      <td>-0.020144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>0.338710</td>\n",
       "      <td>0.318182</td>\n",
       "      <td>0.287671</td>\n",
       "      <td>60</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>71</td>\n",
       "      <td>0.887500</td>\n",
       "      <td>61</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.005303</td>\n",
       "      <td>0.042045</td>\n",
       "      <td>0.022348</td>\n",
       "      <td>0.017045</td>\n",
       "      <td>133672542</td>\n",
       "      <td>0.091667</td>\n",
       "      <td>-0.027273</td>\n",
       "      <td>-0.084470</td>\n",
       "      <td>-0.031061</td>\n",
       "      <td>0.051136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6343</th>\n",
       "      <td>6343</td>\n",
       "      <td>6495</td>\n",
       "      <td>0.387574</td>\n",
       "      <td>0.367978</td>\n",
       "      <td>0.360882</td>\n",
       "      <td>336</td>\n",
       "      <td>0.943820</td>\n",
       "      <td>337</td>\n",
       "      <td>0.910811</td>\n",
       "      <td>588</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.004616</td>\n",
       "      <td>0.011388</td>\n",
       "      <td>-0.006043</td>\n",
       "      <td>0.003811</td>\n",
       "      <td>55489429</td>\n",
       "      <td>0.004783</td>\n",
       "      <td>-0.017871</td>\n",
       "      <td>-0.019040</td>\n",
       "      <td>0.008624</td>\n",
       "      <td>0.023504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6344</th>\n",
       "      <td>6344</td>\n",
       "      <td>6496</td>\n",
       "      <td>0.488462</td>\n",
       "      <td>0.488462</td>\n",
       "      <td>0.455197</td>\n",
       "      <td>258</td>\n",
       "      <td>0.992308</td>\n",
       "      <td>263</td>\n",
       "      <td>0.882550</td>\n",
       "      <td>619</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.011693</td>\n",
       "      <td>-0.033867</td>\n",
       "      <td>-0.001058</td>\n",
       "      <td>0.000903</td>\n",
       "      <td>172293605</td>\n",
       "      <td>-0.039210</td>\n",
       "      <td>0.011306</td>\n",
       "      <td>-0.000800</td>\n",
       "      <td>0.021631</td>\n",
       "      <td>0.007073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6345</th>\n",
       "      <td>6345</td>\n",
       "      <td>6497</td>\n",
       "      <td>0.383838</td>\n",
       "      <td>0.376238</td>\n",
       "      <td>0.374384</td>\n",
       "      <td>296</td>\n",
       "      <td>0.976898</td>\n",
       "      <td>299</td>\n",
       "      <td>0.977124</td>\n",
       "      <td>436</td>\n",
       "      <td>...</td>\n",
       "      <td>0.029023</td>\n",
       "      <td>-0.056753</td>\n",
       "      <td>-0.003397</td>\n",
       "      <td>-0.013266</td>\n",
       "      <td>47082975</td>\n",
       "      <td>-0.035009</td>\n",
       "      <td>-0.017019</td>\n",
       "      <td>0.008801</td>\n",
       "      <td>0.002362</td>\n",
       "      <td>0.040866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6346</th>\n",
       "      <td>6346</td>\n",
       "      <td>6498</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>78</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>78</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>258</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.051282</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>154513027</td>\n",
       "      <td>0.064103</td>\n",
       "      <td>-0.012821</td>\n",
       "      <td>0.012821</td>\n",
       "      <td>-0.025641</td>\n",
       "      <td>-0.038462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6347</th>\n",
       "      <td>6347</td>\n",
       "      <td>6499</td>\n",
       "      <td>0.343220</td>\n",
       "      <td>0.223757</td>\n",
       "      <td>0.264274</td>\n",
       "      <td>234</td>\n",
       "      <td>0.646409</td>\n",
       "      <td>238</td>\n",
       "      <td>0.948207</td>\n",
       "      <td>347</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.017268</td>\n",
       "      <td>-0.016740</td>\n",
       "      <td>-0.002762</td>\n",
       "      <td>-0.018115</td>\n",
       "      <td>90211763</td>\n",
       "      <td>0.012778</td>\n",
       "      <td>-0.012646</td>\n",
       "      <td>-0.005745</td>\n",
       "      <td>0.001959</td>\n",
       "      <td>0.003654</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6348 rows × 60 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      index_x  Unnamed: 0  local_gap_compressed_percent_id  \\\n",
       "0           0           0                         0.287582   \n",
       "1           1           1                         0.319635   \n",
       "2           2           2                         0.279621   \n",
       "3           3           3                         0.327273   \n",
       "4           4           4                         0.338710   \n",
       "...       ...         ...                              ...   \n",
       "6343     6343        6495                         0.387574   \n",
       "6344     6344        6496                         0.488462   \n",
       "6345     6345        6497                         0.383838   \n",
       "6346     6346        6498                         0.769231   \n",
       "6347     6347        6499                         0.343220   \n",
       "\n",
       "      scaled_local_query_percent_id  scaled_local_symmetric_percent_id  \\\n",
       "0                          0.217822                           0.215686   \n",
       "1                          0.295359                           0.297872   \n",
       "2                          0.234127                           0.218924   \n",
       "3                          0.200743                           0.214712   \n",
       "4                          0.318182                           0.287671   \n",
       "...                             ...                                ...   \n",
       "6343                       0.367978                           0.360882   \n",
       "6344                       0.488462                           0.455197   \n",
       "6345                       0.376238                           0.374384   \n",
       "6346                       0.769231                           0.769231   \n",
       "6347                       0.223757                           0.264274   \n",
       "\n",
       "      query_align_len  query_align_cov  subject_align_len  subject_align_cov  \\\n",
       "0                 160         0.792079                152           0.737864   \n",
       "1                 218         0.919831                226           0.969957   \n",
       "2                 211         0.837302                210           0.731707   \n",
       "3                 166         0.617100                163           0.696581   \n",
       "4                  60         0.909091                 71           0.887500   \n",
       "...               ...              ...                ...                ...   \n",
       "6343              336         0.943820                337           0.910811   \n",
       "6344              258         0.992308                263           0.882550   \n",
       "6345              296         0.976898                299           0.977124   \n",
       "6346               78         1.000000                 78           1.000000   \n",
       "6347              234         0.646409                238           0.948207   \n",
       "\n",
       "      bit_score  ...     AAC_T     AAC_V     AAC_W     AAC_Y      index  \\\n",
       "0           131  ... -0.000961  0.062242 -0.000192 -0.010478   48641291   \n",
       "1           282  ... -0.016081  0.018254  0.000072  0.000507   92992745   \n",
       "2            96  ... -0.007743  0.011808  0.002516 -0.018293  157628663   \n",
       "3           175  ... -0.011915  0.019668  0.018762 -0.015251  136708305   \n",
       "4            61  ... -0.005303  0.042045  0.022348  0.017045  133672542   \n",
       "...         ...  ...       ...       ...       ...       ...        ...   \n",
       "6343        588  ... -0.004616  0.011388 -0.006043  0.003811   55489429   \n",
       "6344        619  ... -0.011693 -0.033867 -0.001058  0.000903  172293605   \n",
       "6345        436  ...  0.029023 -0.056753 -0.003397 -0.013266   47082975   \n",
       "6346        258  ... -0.051282  0.000000  0.000000  0.000000  154513027   \n",
       "6347        347  ... -0.017268 -0.016740 -0.002762 -0.018115   90211763   \n",
       "\n",
       "      GAAC_alphatic  GAAC_aromatic  GAAC_postivecharge  GAAC_negativecharge  \\\n",
       "0          0.065318      -0.025810            0.011583            -0.017639   \n",
       "1         -0.014143       0.009525           -0.010051             0.032651   \n",
       "2          0.087205      -0.020132            0.040650             0.006291   \n",
       "3          0.015966       0.014457            0.001080            -0.011359   \n",
       "4          0.091667      -0.027273           -0.084470            -0.031061   \n",
       "...             ...            ...                 ...                  ...   \n",
       "6343       0.004783      -0.017871           -0.019040             0.008624   \n",
       "6344      -0.039210       0.011306           -0.000800             0.021631   \n",
       "6345      -0.035009      -0.017019            0.008801             0.002362   \n",
       "6346       0.064103      -0.012821            0.012821            -0.025641   \n",
       "6347       0.012778      -0.012646           -0.005745             0.001959   \n",
       "\n",
       "      GAAC_uncharge  \n",
       "0         -0.033452  \n",
       "1         -0.017982  \n",
       "2         -0.114015  \n",
       "3         -0.020144  \n",
       "4          0.051136  \n",
       "...             ...  \n",
       "6343       0.023504  \n",
       "6344       0.007073  \n",
       "6345       0.040866  \n",
       "6346      -0.038462  \n",
       "6347       0.003654  \n",
       "\n",
       "[6348 rows x 60 columns]"
      ]
     },
     "execution_count": 199,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "id": "0261f575",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['index_x', 'Unnamed: 0', 'local_gap_compressed_percent_id',\n",
       "       'scaled_local_query_percent_id', 'scaled_local_symmetric_percent_id',\n",
       "       'query_align_len', 'query_align_cov', 'subject_align_len',\n",
       "       'subject_align_cov', 'bit_score', 'thermo_index', 'meso_index',\n",
       "       'prot_pair_index', 'meso_protein_int_index', 'thermo_protein_int_index',\n",
       "       'taxa_pair_index', 'local_gap_compressed_percent_id_16s',\n",
       "       'scaled_local_query_percent_id_16s',\n",
       "       'scaled_local_symmetric_percent_id_16s', 'query_align_cov_16s',\n",
       "       'subject_align_cov_16s', 'bit_score_16s', 'm_ogt', 't_ogt',\n",
       "       'ogt_difference', 'm_protein_seq', 't_protein_seq', 'm_protein_desc',\n",
       "       't_protein_desc', 'm_protein_len', 't_protein_len', 'Jaccard_Score',\n",
       "       'protein_match', 'index_y', 'AAC_A', 'AAC_C', 'AAC_D', 'AAC_E', 'AAC_F',\n",
       "       'AAC_G', 'AAC_H', 'AAC_I', 'AAC_K', 'AAC_L', 'AAC_M', 'AAC_N', 'AAC_P',\n",
       "       'AAC_Q', 'AAC_R', 'AAC_S', 'AAC_T', 'AAC_V', 'AAC_W', 'AAC_Y', 'index',\n",
       "       'GAAC_alphatic', 'GAAC_aromatic', 'GAAC_postivecharge',\n",
       "       'GAAC_negativecharge', 'GAAC_uncharge'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "220a1d24",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b7be5ead",
   "metadata": {},
   "source": [
    "# Scratch work for above function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "id": "d863dd63",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# sequence = \"meso_input_copy.fasta.txt\"\n",
    "# string_to_remove = \"<unknown description>\"\n",
    "\n",
    "# with open(sequence, \"r\") as file:\n",
    "#     content = file.read()\n",
    "    \n",
    "# # Remove the string\n",
    "# new_content = content.replace(string_to_remove, \"\")\n",
    "\n",
    "# with open(sequence, \"w\") as file:\n",
    "#     file.write(new_content)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "id": "96d2c565",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    6348.000000\n",
       "mean      264.121141\n",
       "std        67.865076\n",
       "min        35.000000\n",
       "25%       229.000000\n",
       "50%       262.000000\n",
       "75%       312.000000\n",
       "max       400.000000\n",
       "Name: t_protein_len, dtype: float64"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['t_protein_len'].describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04026501",
   "metadata": {},
   "source": [
    "The bit-score provides a better rule-of-thumb for inferring homology. For average length proteins, a bit score of 50 is almost always significant. A bit score of 40 is only significant (E() < 0.001) in searches of protein databases with fewer than 7000 entries. Increasing the score by 10 bits increases the significance 210=1000-fold, so 50 bits would be significant in a database with less than 7 million entries (10 times SwissProt, and within a factor of 3 of the largest protein databases). Thus, the NCBI Blast web site uses a color code of blue for alignment with scores between 40–50 bits; and green for scores between 50–80 bits. In the yeast vs human example, the alignments with less than 20% identity had scores ranging from 55 – 170 bits. Except for very long proteins and very large databases, 50 bits of similarity score will always be statistically significant and is a much better rule-of-thumb for inferring homology in protein alignments.\n",
    "\n",
    "Pearson et al., 2013: An Introduction to Sequence Similarity (“Homology”) Searching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "id": "14015e54",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Yes    5243\n",
       "No     1105\n",
       "Name: protein_match, dtype: int64"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['protein_match'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "id": "b27836e4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_x</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>...</th>\n",
       "      <th>AAC_T</th>\n",
       "      <th>AAC_V</th>\n",
       "      <th>AAC_W</th>\n",
       "      <th>AAC_Y</th>\n",
       "      <th>index</th>\n",
       "      <th>GAAC_alphatic</th>\n",
       "      <th>GAAC_aromatic</th>\n",
       "      <th>GAAC_postivecharge</th>\n",
       "      <th>GAAC_negativecharge</th>\n",
       "      <th>GAAC_uncharge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.287582</td>\n",
       "      <td>0.217822</td>\n",
       "      <td>0.215686</td>\n",
       "      <td>160</td>\n",
       "      <td>0.792079</td>\n",
       "      <td>152</td>\n",
       "      <td>0.737864</td>\n",
       "      <td>131</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000961</td>\n",
       "      <td>0.062242</td>\n",
       "      <td>-0.000192</td>\n",
       "      <td>-0.010478</td>\n",
       "      <td>48641291</td>\n",
       "      <td>0.065318</td>\n",
       "      <td>-0.025810</td>\n",
       "      <td>0.011583</td>\n",
       "      <td>-0.017639</td>\n",
       "      <td>-0.033452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.319635</td>\n",
       "      <td>0.295359</td>\n",
       "      <td>0.297872</td>\n",
       "      <td>218</td>\n",
       "      <td>0.919831</td>\n",
       "      <td>226</td>\n",
       "      <td>0.969957</td>\n",
       "      <td>282</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.016081</td>\n",
       "      <td>0.018254</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000507</td>\n",
       "      <td>92992745</td>\n",
       "      <td>-0.014143</td>\n",
       "      <td>0.009525</td>\n",
       "      <td>-0.010051</td>\n",
       "      <td>0.032651</td>\n",
       "      <td>-0.017982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0.279621</td>\n",
       "      <td>0.234127</td>\n",
       "      <td>0.218924</td>\n",
       "      <td>211</td>\n",
       "      <td>0.837302</td>\n",
       "      <td>210</td>\n",
       "      <td>0.731707</td>\n",
       "      <td>96</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.007743</td>\n",
       "      <td>0.011808</td>\n",
       "      <td>0.002516</td>\n",
       "      <td>-0.018293</td>\n",
       "      <td>157628663</td>\n",
       "      <td>0.087205</td>\n",
       "      <td>-0.020132</td>\n",
       "      <td>0.040650</td>\n",
       "      <td>0.006291</td>\n",
       "      <td>-0.114015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0.327273</td>\n",
       "      <td>0.200743</td>\n",
       "      <td>0.214712</td>\n",
       "      <td>166</td>\n",
       "      <td>0.617100</td>\n",
       "      <td>163</td>\n",
       "      <td>0.696581</td>\n",
       "      <td>175</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.011915</td>\n",
       "      <td>0.019668</td>\n",
       "      <td>0.018762</td>\n",
       "      <td>-0.015251</td>\n",
       "      <td>136708305</td>\n",
       "      <td>0.015966</td>\n",
       "      <td>0.014457</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>-0.011359</td>\n",
       "      <td>-0.020144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>0.338710</td>\n",
       "      <td>0.318182</td>\n",
       "      <td>0.287671</td>\n",
       "      <td>60</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>71</td>\n",
       "      <td>0.887500</td>\n",
       "      <td>61</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.005303</td>\n",
       "      <td>0.042045</td>\n",
       "      <td>0.022348</td>\n",
       "      <td>0.017045</td>\n",
       "      <td>133672542</td>\n",
       "      <td>0.091667</td>\n",
       "      <td>-0.027273</td>\n",
       "      <td>-0.084470</td>\n",
       "      <td>-0.031061</td>\n",
       "      <td>0.051136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6343</th>\n",
       "      <td>6343</td>\n",
       "      <td>6495</td>\n",
       "      <td>0.387574</td>\n",
       "      <td>0.367978</td>\n",
       "      <td>0.360882</td>\n",
       "      <td>336</td>\n",
       "      <td>0.943820</td>\n",
       "      <td>337</td>\n",
       "      <td>0.910811</td>\n",
       "      <td>588</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.004616</td>\n",
       "      <td>0.011388</td>\n",
       "      <td>-0.006043</td>\n",
       "      <td>0.003811</td>\n",
       "      <td>55489429</td>\n",
       "      <td>0.004783</td>\n",
       "      <td>-0.017871</td>\n",
       "      <td>-0.019040</td>\n",
       "      <td>0.008624</td>\n",
       "      <td>0.023504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6344</th>\n",
       "      <td>6344</td>\n",
       "      <td>6496</td>\n",
       "      <td>0.488462</td>\n",
       "      <td>0.488462</td>\n",
       "      <td>0.455197</td>\n",
       "      <td>258</td>\n",
       "      <td>0.992308</td>\n",
       "      <td>263</td>\n",
       "      <td>0.882550</td>\n",
       "      <td>619</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.011693</td>\n",
       "      <td>-0.033867</td>\n",
       "      <td>-0.001058</td>\n",
       "      <td>0.000903</td>\n",
       "      <td>172293605</td>\n",
       "      <td>-0.039210</td>\n",
       "      <td>0.011306</td>\n",
       "      <td>-0.000800</td>\n",
       "      <td>0.021631</td>\n",
       "      <td>0.007073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6345</th>\n",
       "      <td>6345</td>\n",
       "      <td>6497</td>\n",
       "      <td>0.383838</td>\n",
       "      <td>0.376238</td>\n",
       "      <td>0.374384</td>\n",
       "      <td>296</td>\n",
       "      <td>0.976898</td>\n",
       "      <td>299</td>\n",
       "      <td>0.977124</td>\n",
       "      <td>436</td>\n",
       "      <td>...</td>\n",
       "      <td>0.029023</td>\n",
       "      <td>-0.056753</td>\n",
       "      <td>-0.003397</td>\n",
       "      <td>-0.013266</td>\n",
       "      <td>47082975</td>\n",
       "      <td>-0.035009</td>\n",
       "      <td>-0.017019</td>\n",
       "      <td>0.008801</td>\n",
       "      <td>0.002362</td>\n",
       "      <td>0.040866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6346</th>\n",
       "      <td>6346</td>\n",
       "      <td>6498</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>78</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>78</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>258</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.051282</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>154513027</td>\n",
       "      <td>0.064103</td>\n",
       "      <td>-0.012821</td>\n",
       "      <td>0.012821</td>\n",
       "      <td>-0.025641</td>\n",
       "      <td>-0.038462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6347</th>\n",
       "      <td>6347</td>\n",
       "      <td>6499</td>\n",
       "      <td>0.343220</td>\n",
       "      <td>0.223757</td>\n",
       "      <td>0.264274</td>\n",
       "      <td>234</td>\n",
       "      <td>0.646409</td>\n",
       "      <td>238</td>\n",
       "      <td>0.948207</td>\n",
       "      <td>347</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.017268</td>\n",
       "      <td>-0.016740</td>\n",
       "      <td>-0.002762</td>\n",
       "      <td>-0.018115</td>\n",
       "      <td>90211763</td>\n",
       "      <td>0.012778</td>\n",
       "      <td>-0.012646</td>\n",
       "      <td>-0.005745</td>\n",
       "      <td>0.001959</td>\n",
       "      <td>0.003654</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6348 rows × 60 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      index_x  Unnamed: 0  local_gap_compressed_percent_id  \\\n",
       "0           0           0                         0.287582   \n",
       "1           1           1                         0.319635   \n",
       "2           2           2                         0.279621   \n",
       "3           3           3                         0.327273   \n",
       "4           4           4                         0.338710   \n",
       "...       ...         ...                              ...   \n",
       "6343     6343        6495                         0.387574   \n",
       "6344     6344        6496                         0.488462   \n",
       "6345     6345        6497                         0.383838   \n",
       "6346     6346        6498                         0.769231   \n",
       "6347     6347        6499                         0.343220   \n",
       "\n",
       "      scaled_local_query_percent_id  scaled_local_symmetric_percent_id  \\\n",
       "0                          0.217822                           0.215686   \n",
       "1                          0.295359                           0.297872   \n",
       "2                          0.234127                           0.218924   \n",
       "3                          0.200743                           0.214712   \n",
       "4                          0.318182                           0.287671   \n",
       "...                             ...                                ...   \n",
       "6343                       0.367978                           0.360882   \n",
       "6344                       0.488462                           0.455197   \n",
       "6345                       0.376238                           0.374384   \n",
       "6346                       0.769231                           0.769231   \n",
       "6347                       0.223757                           0.264274   \n",
       "\n",
       "      query_align_len  query_align_cov  subject_align_len  subject_align_cov  \\\n",
       "0                 160         0.792079                152           0.737864   \n",
       "1                 218         0.919831                226           0.969957   \n",
       "2                 211         0.837302                210           0.731707   \n",
       "3                 166         0.617100                163           0.696581   \n",
       "4                  60         0.909091                 71           0.887500   \n",
       "...               ...              ...                ...                ...   \n",
       "6343              336         0.943820                337           0.910811   \n",
       "6344              258         0.992308                263           0.882550   \n",
       "6345              296         0.976898                299           0.977124   \n",
       "6346               78         1.000000                 78           1.000000   \n",
       "6347              234         0.646409                238           0.948207   \n",
       "\n",
       "      bit_score  ...     AAC_T     AAC_V     AAC_W     AAC_Y      index  \\\n",
       "0           131  ... -0.000961  0.062242 -0.000192 -0.010478   48641291   \n",
       "1           282  ... -0.016081  0.018254  0.000072  0.000507   92992745   \n",
       "2            96  ... -0.007743  0.011808  0.002516 -0.018293  157628663   \n",
       "3           175  ... -0.011915  0.019668  0.018762 -0.015251  136708305   \n",
       "4            61  ... -0.005303  0.042045  0.022348  0.017045  133672542   \n",
       "...         ...  ...       ...       ...       ...       ...        ...   \n",
       "6343        588  ... -0.004616  0.011388 -0.006043  0.003811   55489429   \n",
       "6344        619  ... -0.011693 -0.033867 -0.001058  0.000903  172293605   \n",
       "6345        436  ...  0.029023 -0.056753 -0.003397 -0.013266   47082975   \n",
       "6346        258  ... -0.051282  0.000000  0.000000  0.000000  154513027   \n",
       "6347        347  ... -0.017268 -0.016740 -0.002762 -0.018115   90211763   \n",
       "\n",
       "      GAAC_alphatic  GAAC_aromatic  GAAC_postivecharge  GAAC_negativecharge  \\\n",
       "0          0.065318      -0.025810            0.011583            -0.017639   \n",
       "1         -0.014143       0.009525           -0.010051             0.032651   \n",
       "2          0.087205      -0.020132            0.040650             0.006291   \n",
       "3          0.015966       0.014457            0.001080            -0.011359   \n",
       "4          0.091667      -0.027273           -0.084470            -0.031061   \n",
       "...             ...            ...                 ...                  ...   \n",
       "6343       0.004783      -0.017871           -0.019040             0.008624   \n",
       "6344      -0.039210       0.011306           -0.000800             0.021631   \n",
       "6345      -0.035009      -0.017019            0.008801             0.002362   \n",
       "6346       0.064103      -0.012821            0.012821            -0.025641   \n",
       "6347       0.012778      -0.012646           -0.005745             0.001959   \n",
       "\n",
       "      GAAC_uncharge  \n",
       "0         -0.033452  \n",
       "1         -0.017982  \n",
       "2         -0.114015  \n",
       "3         -0.020144  \n",
       "4          0.051136  \n",
       "...             ...  \n",
       "6343       0.023504  \n",
       "6344       0.007073  \n",
       "6345       0.040866  \n",
       "6346      -0.038462  \n",
       "6347       0.003654  \n",
       "\n",
       "[6348 rows x 60 columns]"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "id": "7c305f9a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0          0\n",
       "1          1\n",
       "2          2\n",
       "3          3\n",
       "4          4\n",
       "        ... \n",
       "6343    6343\n",
       "6344    6344\n",
       "6345    6345\n",
       "6346    6346\n",
       "6347    6347\n",
       "Name: index_x, Length: 6348, dtype: int64"
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['index_x']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "id": "084d9561",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get rid of stuff that isn't quantitative\n",
    "\n",
    "df = df.drop(columns = (['Unnamed: 0', 'index_x', 'thermo_index',\n",
    "                         'm_protein_seq', 't_protein_seq', 'm_protein_desc', 't_protein_desc']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "id": "eea27e6c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>meso_index</th>\n",
       "      <th>prot_pair_index</th>\n",
       "      <th>...</th>\n",
       "      <th>AAC_T</th>\n",
       "      <th>AAC_V</th>\n",
       "      <th>AAC_W</th>\n",
       "      <th>AAC_Y</th>\n",
       "      <th>index</th>\n",
       "      <th>GAAC_alphatic</th>\n",
       "      <th>GAAC_aromatic</th>\n",
       "      <th>GAAC_postivecharge</th>\n",
       "      <th>GAAC_negativecharge</th>\n",
       "      <th>GAAC_uncharge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.287582</td>\n",
       "      <td>0.217822</td>\n",
       "      <td>0.215686</td>\n",
       "      <td>160</td>\n",
       "      <td>0.792079</td>\n",
       "      <td>152</td>\n",
       "      <td>0.737864</td>\n",
       "      <td>131</td>\n",
       "      <td>12897</td>\n",
       "      <td>48641291</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000961</td>\n",
       "      <td>0.062242</td>\n",
       "      <td>-0.000192</td>\n",
       "      <td>-0.010478</td>\n",
       "      <td>48641291</td>\n",
       "      <td>0.065318</td>\n",
       "      <td>-0.025810</td>\n",
       "      <td>0.011583</td>\n",
       "      <td>-0.017639</td>\n",
       "      <td>-0.033452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.319635</td>\n",
       "      <td>0.295359</td>\n",
       "      <td>0.297872</td>\n",
       "      <td>218</td>\n",
       "      <td>0.919831</td>\n",
       "      <td>226</td>\n",
       "      <td>0.969957</td>\n",
       "      <td>282</td>\n",
       "      <td>13026</td>\n",
       "      <td>92992745</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.016081</td>\n",
       "      <td>0.018254</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000507</td>\n",
       "      <td>92992745</td>\n",
       "      <td>-0.014143</td>\n",
       "      <td>0.009525</td>\n",
       "      <td>-0.010051</td>\n",
       "      <td>0.032651</td>\n",
       "      <td>-0.017982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.279621</td>\n",
       "      <td>0.234127</td>\n",
       "      <td>0.218924</td>\n",
       "      <td>211</td>\n",
       "      <td>0.837302</td>\n",
       "      <td>210</td>\n",
       "      <td>0.731707</td>\n",
       "      <td>96</td>\n",
       "      <td>8203</td>\n",
       "      <td>157628663</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.007743</td>\n",
       "      <td>0.011808</td>\n",
       "      <td>0.002516</td>\n",
       "      <td>-0.018293</td>\n",
       "      <td>157628663</td>\n",
       "      <td>0.087205</td>\n",
       "      <td>-0.020132</td>\n",
       "      <td>0.040650</td>\n",
       "      <td>0.006291</td>\n",
       "      <td>-0.114015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.327273</td>\n",
       "      <td>0.200743</td>\n",
       "      <td>0.214712</td>\n",
       "      <td>166</td>\n",
       "      <td>0.617100</td>\n",
       "      <td>163</td>\n",
       "      <td>0.696581</td>\n",
       "      <td>175</td>\n",
       "      <td>3340</td>\n",
       "      <td>136708305</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.011915</td>\n",
       "      <td>0.019668</td>\n",
       "      <td>0.018762</td>\n",
       "      <td>-0.015251</td>\n",
       "      <td>136708305</td>\n",
       "      <td>0.015966</td>\n",
       "      <td>0.014457</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>-0.011359</td>\n",
       "      <td>-0.020144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.338710</td>\n",
       "      <td>0.318182</td>\n",
       "      <td>0.287671</td>\n",
       "      <td>60</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>71</td>\n",
       "      <td>0.887500</td>\n",
       "      <td>61</td>\n",
       "      <td>14020</td>\n",
       "      <td>133672542</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.005303</td>\n",
       "      <td>0.042045</td>\n",
       "      <td>0.022348</td>\n",
       "      <td>0.017045</td>\n",
       "      <td>133672542</td>\n",
       "      <td>0.091667</td>\n",
       "      <td>-0.027273</td>\n",
       "      <td>-0.084470</td>\n",
       "      <td>-0.031061</td>\n",
       "      <td>0.051136</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   local_gap_compressed_percent_id  scaled_local_query_percent_id  \\\n",
       "0                         0.287582                       0.217822   \n",
       "1                         0.319635                       0.295359   \n",
       "2                         0.279621                       0.234127   \n",
       "3                         0.327273                       0.200743   \n",
       "4                         0.338710                       0.318182   \n",
       "\n",
       "   scaled_local_symmetric_percent_id  query_align_len  query_align_cov  \\\n",
       "0                           0.215686              160         0.792079   \n",
       "1                           0.297872              218         0.919831   \n",
       "2                           0.218924              211         0.837302   \n",
       "3                           0.214712              166         0.617100   \n",
       "4                           0.287671               60         0.909091   \n",
       "\n",
       "   subject_align_len  subject_align_cov  bit_score  meso_index  \\\n",
       "0                152           0.737864        131       12897   \n",
       "1                226           0.969957        282       13026   \n",
       "2                210           0.731707         96        8203   \n",
       "3                163           0.696581        175        3340   \n",
       "4                 71           0.887500         61       14020   \n",
       "\n",
       "   prot_pair_index  ...     AAC_T     AAC_V     AAC_W     AAC_Y      index  \\\n",
       "0         48641291  ... -0.000961  0.062242 -0.000192 -0.010478   48641291   \n",
       "1         92992745  ... -0.016081  0.018254  0.000072  0.000507   92992745   \n",
       "2        157628663  ... -0.007743  0.011808  0.002516 -0.018293  157628663   \n",
       "3        136708305  ... -0.011915  0.019668  0.018762 -0.015251  136708305   \n",
       "4        133672542  ... -0.005303  0.042045  0.022348  0.017045  133672542   \n",
       "\n",
       "   GAAC_alphatic  GAAC_aromatic  GAAC_postivecharge  GAAC_negativecharge  \\\n",
       "0       0.065318      -0.025810            0.011583            -0.017639   \n",
       "1      -0.014143       0.009525           -0.010051             0.032651   \n",
       "2       0.087205      -0.020132            0.040650             0.006291   \n",
       "3       0.015966       0.014457            0.001080            -0.011359   \n",
       "4       0.091667      -0.027273           -0.084470            -0.031061   \n",
       "\n",
       "   GAAC_uncharge  \n",
       "0      -0.033452  \n",
       "1      -0.017982  \n",
       "2      -0.114015  \n",
       "3      -0.020144  \n",
       "4       0.051136  \n",
       "\n",
       "[5 rows x 53 columns]"
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07cc3a80",
   "metadata": {},
   "source": [
    "From pairplot below, it looks like query_align_cov_16s and subject_align_cov_16s dont correlate with target. Let's remove them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "91079edd",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.drop(columns = ['query_align_cov_16s', 'subject_align_cov_16s'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "id": "36d8dadb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# sns.pairplot(df.sample(500),height = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "id": "7929625e",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>meso_index</th>\n",
       "      <th>prot_pair_index</th>\n",
       "      <th>...</th>\n",
       "      <th>AAC_T</th>\n",
       "      <th>AAC_V</th>\n",
       "      <th>AAC_W</th>\n",
       "      <th>AAC_Y</th>\n",
       "      <th>index</th>\n",
       "      <th>GAAC_alphatic</th>\n",
       "      <th>GAAC_aromatic</th>\n",
       "      <th>GAAC_postivecharge</th>\n",
       "      <th>GAAC_negativecharge</th>\n",
       "      <th>GAAC_uncharge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>6348.000000</td>\n",
       "      <td>6348.000000</td>\n",
       "      <td>6348.000000</td>\n",
       "      <td>6348.000000</td>\n",
       "      <td>6348.000000</td>\n",
       "      <td>6348.000000</td>\n",
       "      <td>6348.000000</td>\n",
       "      <td>6348.000000</td>\n",
       "      <td>6348.000000</td>\n",
       "      <td>6.348000e+03</td>\n",
       "      <td>...</td>\n",
       "      <td>6348.000000</td>\n",
       "      <td>6348.000000</td>\n",
       "      <td>6348.000000</td>\n",
       "      <td>6348.000000</td>\n",
       "      <td>6.348000e+03</td>\n",
       "      <td>6348.000000</td>\n",
       "      <td>6348.000000</td>\n",
       "      <td>6348.000000</td>\n",
       "      <td>6348.000000</td>\n",
       "      <td>6348.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>0.348174</td>\n",
       "      <td>0.276047</td>\n",
       "      <td>0.270881</td>\n",
       "      <td>208.339162</td>\n",
       "      <td>0.792409</td>\n",
       "      <td>207.604600</td>\n",
       "      <td>0.773547</td>\n",
       "      <td>233.339162</td>\n",
       "      <td>8098.207152</td>\n",
       "      <td>9.059579e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002973</td>\n",
       "      <td>-0.000936</td>\n",
       "      <td>-0.000822</td>\n",
       "      <td>-0.000874</td>\n",
       "      <td>9.059579e+07</td>\n",
       "      <td>0.004287</td>\n",
       "      <td>-0.001298</td>\n",
       "      <td>-0.008658</td>\n",
       "      <td>-0.002680</td>\n",
       "      <td>0.008348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.096390</td>\n",
       "      <td>0.113913</td>\n",
       "      <td>0.115180</td>\n",
       "      <td>63.261682</td>\n",
       "      <td>0.145697</td>\n",
       "      <td>63.264993</td>\n",
       "      <td>0.175379</td>\n",
       "      <td>186.940699</td>\n",
       "      <td>4565.102004</td>\n",
       "      <td>5.250268e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>0.020149</td>\n",
       "      <td>0.024541</td>\n",
       "      <td>0.008551</td>\n",
       "      <td>0.011881</td>\n",
       "      <td>5.250268e+07</td>\n",
       "      <td>0.045827</td>\n",
       "      <td>0.019876</td>\n",
       "      <td>0.023366</td>\n",
       "      <td>0.022492</td>\n",
       "      <td>0.038984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.193431</td>\n",
       "      <td>0.106383</td>\n",
       "      <td>0.066929</td>\n",
       "      <td>23.000000</td>\n",
       "      <td>0.327935</td>\n",
       "      <td>23.000000</td>\n",
       "      <td>0.083916</td>\n",
       "      <td>37.000000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>8.858000e+03</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.081561</td>\n",
       "      <td>-0.109956</td>\n",
       "      <td>-0.057143</td>\n",
       "      <td>-0.136364</td>\n",
       "      <td>8.858000e+03</td>\n",
       "      <td>-0.177149</td>\n",
       "      <td>-0.090909</td>\n",
       "      <td>-0.152017</td>\n",
       "      <td>-0.211716</td>\n",
       "      <td>-0.174242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>0.292571</td>\n",
       "      <td>0.205948</td>\n",
       "      <td>0.203093</td>\n",
       "      <td>176.000000</td>\n",
       "      <td>0.672716</td>\n",
       "      <td>176.000000</td>\n",
       "      <td>0.657086</td>\n",
       "      <td>121.000000</td>\n",
       "      <td>4243.250000</td>\n",
       "      <td>4.530101e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.009939</td>\n",
       "      <td>-0.016708</td>\n",
       "      <td>-0.005470</td>\n",
       "      <td>-0.008303</td>\n",
       "      <td>4.530101e+07</td>\n",
       "      <td>-0.026531</td>\n",
       "      <td>-0.013625</td>\n",
       "      <td>-0.022834</td>\n",
       "      <td>-0.016520</td>\n",
       "      <td>-0.016369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>0.323720</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.244565</td>\n",
       "      <td>211.000000</td>\n",
       "      <td>0.811727</td>\n",
       "      <td>210.000000</td>\n",
       "      <td>0.805808</td>\n",
       "      <td>188.000000</td>\n",
       "      <td>8046.000000</td>\n",
       "      <td>9.085625e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002729</td>\n",
       "      <td>-0.000799</td>\n",
       "      <td>-0.000080</td>\n",
       "      <td>-0.000484</td>\n",
       "      <td>9.085625e+07</td>\n",
       "      <td>0.003705</td>\n",
       "      <td>-0.001190</td>\n",
       "      <td>-0.007898</td>\n",
       "      <td>-0.002856</td>\n",
       "      <td>0.007810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>0.366709</td>\n",
       "      <td>0.306723</td>\n",
       "      <td>0.300314</td>\n",
       "      <td>241.000000</td>\n",
       "      <td>0.925000</td>\n",
       "      <td>240.000000</td>\n",
       "      <td>0.924258</td>\n",
       "      <td>270.000000</td>\n",
       "      <td>12069.000000</td>\n",
       "      <td>1.360285e+08</td>\n",
       "      <td>...</td>\n",
       "      <td>0.015633</td>\n",
       "      <td>0.014542</td>\n",
       "      <td>0.004072</td>\n",
       "      <td>0.006596</td>\n",
       "      <td>1.360285e+08</td>\n",
       "      <td>0.034317</td>\n",
       "      <td>0.011286</td>\n",
       "      <td>0.006209</td>\n",
       "      <td>0.011004</td>\n",
       "      <td>0.033613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>0.959350</td>\n",
       "      <td>0.951613</td>\n",
       "      <td>0.955466</td>\n",
       "      <td>398.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>398.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1565.000000</td>\n",
       "      <td>16345.000000</td>\n",
       "      <td>1.814286e+08</td>\n",
       "      <td>...</td>\n",
       "      <td>0.100446</td>\n",
       "      <td>0.113636</td>\n",
       "      <td>0.043723</td>\n",
       "      <td>0.061224</td>\n",
       "      <td>1.814286e+08</td>\n",
       "      <td>0.295455</td>\n",
       "      <td>0.086914</td>\n",
       "      <td>0.098889</td>\n",
       "      <td>0.106208</td>\n",
       "      <td>0.193513</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 50 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       local_gap_compressed_percent_id  scaled_local_query_percent_id  \\\n",
       "count                      6348.000000                    6348.000000   \n",
       "mean                          0.348174                       0.276047   \n",
       "std                           0.096390                       0.113913   \n",
       "min                           0.193431                       0.106383   \n",
       "25%                           0.292571                       0.205948   \n",
       "50%                           0.323720                       0.250000   \n",
       "75%                           0.366709                       0.306723   \n",
       "max                           0.959350                       0.951613   \n",
       "\n",
       "       scaled_local_symmetric_percent_id  query_align_len  query_align_cov  \\\n",
       "count                        6348.000000      6348.000000      6348.000000   \n",
       "mean                            0.270881       208.339162         0.792409   \n",
       "std                             0.115180        63.261682         0.145697   \n",
       "min                             0.066929        23.000000         0.327935   \n",
       "25%                             0.203093       176.000000         0.672716   \n",
       "50%                             0.244565       211.000000         0.811727   \n",
       "75%                             0.300314       241.000000         0.925000   \n",
       "max                             0.955466       398.000000         1.000000   \n",
       "\n",
       "       subject_align_len  subject_align_cov    bit_score    meso_index  \\\n",
       "count        6348.000000        6348.000000  6348.000000   6348.000000   \n",
       "mean          207.604600           0.773547   233.339162   8098.207152   \n",
       "std            63.264993           0.175379   186.940699   4565.102004   \n",
       "min            23.000000           0.083916    37.000000      5.000000   \n",
       "25%           176.000000           0.657086   121.000000   4243.250000   \n",
       "50%           210.000000           0.805808   188.000000   8046.000000   \n",
       "75%           240.000000           0.924258   270.000000  12069.000000   \n",
       "max           398.000000           1.000000  1565.000000  16345.000000   \n",
       "\n",
       "       prot_pair_index  ...        AAC_T        AAC_V        AAC_W  \\\n",
       "count     6.348000e+03  ...  6348.000000  6348.000000  6348.000000   \n",
       "mean      9.059579e+07  ...     0.002973    -0.000936    -0.000822   \n",
       "std       5.250268e+07  ...     0.020149     0.024541     0.008551   \n",
       "min       8.858000e+03  ...    -0.081561    -0.109956    -0.057143   \n",
       "25%       4.530101e+07  ...    -0.009939    -0.016708    -0.005470   \n",
       "50%       9.085625e+07  ...     0.002729    -0.000799    -0.000080   \n",
       "75%       1.360285e+08  ...     0.015633     0.014542     0.004072   \n",
       "max       1.814286e+08  ...     0.100446     0.113636     0.043723   \n",
       "\n",
       "             AAC_Y         index  GAAC_alphatic  GAAC_aromatic  \\\n",
       "count  6348.000000  6.348000e+03    6348.000000    6348.000000   \n",
       "mean     -0.000874  9.059579e+07       0.004287      -0.001298   \n",
       "std       0.011881  5.250268e+07       0.045827       0.019876   \n",
       "min      -0.136364  8.858000e+03      -0.177149      -0.090909   \n",
       "25%      -0.008303  4.530101e+07      -0.026531      -0.013625   \n",
       "50%      -0.000484  9.085625e+07       0.003705      -0.001190   \n",
       "75%       0.006596  1.360285e+08       0.034317       0.011286   \n",
       "max       0.061224  1.814286e+08       0.295455       0.086914   \n",
       "\n",
       "       GAAC_postivecharge  GAAC_negativecharge  GAAC_uncharge  \n",
       "count         6348.000000          6348.000000    6348.000000  \n",
       "mean            -0.008658            -0.002680       0.008348  \n",
       "std              0.023366             0.022492       0.038984  \n",
       "min             -0.152017            -0.211716      -0.174242  \n",
       "25%             -0.022834            -0.016520      -0.016369  \n",
       "50%             -0.007898            -0.002856       0.007810  \n",
       "75%              0.006209             0.011004       0.033613  \n",
       "max              0.098889             0.106208       0.193513  \n",
       "\n",
       "[8 rows x 50 columns]"
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90871d01",
   "metadata": {},
   "source": [
    "Split data into dev and test, and then split that into train and validation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "id": "6519dcaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#drop columns that don't exihibit signficant pearson correlation with bit_score\n",
    "\n",
    "df = df.drop(columns = ['meso_index', 'meso_protein_int_index', 'local_gap_compressed_percent_id_16s', \n",
    "                        'scaled_local_query_percent_id_16s', 'scaled_local_symmetric_percent_id_16s',\n",
    "                       'bit_score_16s', 'm_ogt', 't_ogt', 'taxa_pair_index', 'thermo_protein_int_index'\n",
    "                       , 'prot_pair_index', 'ogt_difference'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "da410bc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.drop(columns=['protein_match'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "id": "af9f23cf",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5395, 38)\n",
      "(953, 38)\n",
      "(4585, 38)\n",
      "(810, 38)\n"
     ]
    }
   ],
   "source": [
    "#choosing 80/20 split instead of 85/15 because of volume of data\n",
    "\n",
    "dev, test = sklearn.model_selection.train_test_split(df, test_size=0.15, random_state=1)\n",
    "\n",
    "train, val = sklearn.model_selection.train_test_split(dev, test_size=0.15, random_state=1)\n",
    "\n",
    "print(dev.shape)\n",
    "print(test.shape)\n",
    "print(train.shape)\n",
    "print(val.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "8470550d",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#ID target and features, separate into separate arrays\n",
    "\n",
    "target = 'Jaccard_Score'\n",
    "input_features = [columns for columns in df]\n",
    "input_features.remove(target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "id": "c62edd28",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['local_gap_compressed_percent_id', 'scaled_local_query_percent_id', 'scaled_local_symmetric_percent_id', 'query_align_len', 'query_align_cov', 'subject_align_len', 'subject_align_cov', 'bit_score', 'm_protein_len', 't_protein_len', 'index_y', 'AAC_A', 'AAC_C', 'AAC_D', 'AAC_E', 'AAC_F', 'AAC_G', 'AAC_H', 'AAC_I', 'AAC_K', 'AAC_L', 'AAC_M', 'AAC_N', 'AAC_P', 'AAC_Q', 'AAC_R', 'AAC_S', 'AAC_T', 'AAC_V', 'AAC_W', 'AAC_Y', 'index', 'GAAC_alphatic', 'GAAC_aromatic', 'GAAC_postivecharge', 'GAAC_negativecharge', 'GAAC_uncharge']\n",
      "Jaccard_Score\n"
     ]
    }
   ],
   "source": [
    "print(input_features)\n",
    "print(target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "id": "1ab62bff",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5395, 37) (953, 37) (5395, 1) (953, 1)\n"
     ]
    }
   ],
   "source": [
    "#split X and y\n",
    "\n",
    "dev_X = dev[input_features].values\n",
    "test_X = test[input_features].values\n",
    "\n",
    "dev_y = dev[target].values.reshape(-1,1)\n",
    "test_y = test[target].values.reshape(-1,1)  \n",
    "\n",
    "print(dev_X.shape, test_X.shape, dev_y.shape, test_y.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "id": "9ac52ac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#same thing for training and validation data\n",
    "\n",
    "train_X = train[input_features].values\n",
    "val_X = val[input_features].values\n",
    "\n",
    "train_y = train[target].values.reshape(-1,1)\n",
    "val_y = val[target].values.reshape(-1,1) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c374f2d7",
   "metadata": {},
   "source": [
    "Scale the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "id": "2b48b2e5",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "scaler = sklearn.preprocessing.StandardScaler()\n",
    "dev_X = scaler.fit_transform(dev_X)\n",
    "test_X = scaler.fit_transform(test_X)\n",
    "train_X = scaler.fit_transform(train_X)\n",
    "val_X = scaler.fit_transform(val_X)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3022371",
   "metadata": {},
   "source": [
    "Train the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "id": "be4e09b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-3 {color: black;background-color: white;}#sk-container-id-3 pre{padding: 0;}#sk-container-id-3 div.sk-toggleable {background-color: white;}#sk-container-id-3 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-3 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-3 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-3 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-3 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-3 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-3 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-3 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-container-id-3 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-3 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-3 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-3 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-3 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-3 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-3 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-3 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-3 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-3 div.sk-item {position: relative;z-index: 1;}#sk-container-id-3 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-3 div.sk-item::before, #sk-container-id-3 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-3 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-3 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-3 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-3 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-3 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-3 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-3 div.sk-label-container {text-align: center;}#sk-container-id-3 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-3 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-3\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>RandomForestRegressor()</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-3\" type=\"checkbox\" checked><label for=\"sk-estimator-id-3\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">RandomForestRegressor</label><div class=\"sk-toggleable__content\"><pre>RandomForestRegressor()</pre></div></div></div></div></div>"
      ],
      "text/plain": [
       "RandomForestRegressor()"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Random Forest\n",
    "\n",
    "model = sklearn.ensemble.RandomForestRegressor()\n",
    "\n",
    "model.fit(train_X, train_y.ravel())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69ebf269",
   "metadata": {},
   "source": [
    "Test the model, report relevant statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "id": "c73327e9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model score is: 0.3669055969884438\n",
      "[0.60390795 0.96       0.6739881  0.43842857 0.52303571 0.61281602\n",
      " 0.66175938 0.80086111 0.42849964 0.6289127  0.5912619  0.69910714\n",
      " 0.38770635 0.66678968 0.46654365 0.48885442 0.26077778 0.44075577\n",
      " 0.91191667 0.48374603 0.76127778 0.05313492 0.76395238 0.40683442\n",
      " 0.96416667 0.82580159 0.28683333 0.20842857 0.67293651 0.65707143\n",
      " 0.81506349 0.81621429 0.69366667 0.74704762 0.91341667 0.82391667\n",
      " 0.87766667 0.7987381  0.64969322 0.59069872 0.77635317 0.97029487\n",
      " 0.41793162 0.77604762 0.74879762 0.76415995 0.75471429 0.45893362\n",
      " 0.56573918 0.48990476 0.94983333 0.60159127 0.37009524 0.68502778\n",
      " 0.42476465 0.85961111 0.9072381  0.87058333 0.71476623 0.73829437\n",
      " 0.50654906 0.77716667 0.74466667 0.85607937 0.63891667 0.95821429\n",
      " 0.74572619 0.87158333 0.7576746  0.953      0.5851562  0.80288889\n",
      " 0.62020238 0.65015751 0.4565174  0.4435689  0.81195635 0.66874603\n",
      " 0.64269048 0.55076984 0.53645665 0.42031566 0.85414286 0.91166667\n",
      " 0.73866667 0.86316667 0.54149878 0.84422619 0.47167011 0.66699675\n",
      " 0.54203066 0.8287619  0.69436419 0.62390476 0.0735     0.19066667\n",
      " 0.61735276 0.80896429 0.87825    0.66393434 0.34888095 0.92452381\n",
      " 0.62202201 0.92630952 0.89130952 0.8374881  0.48874747 0.52305125\n",
      " 0.6142619  0.68762302 0.51584799 0.8337381  0.45695491 0.89687302\n",
      " 0.64877381 0.50782973 0.32986255 0.62325    0.91117582 0.71940476\n",
      " 0.37959921 0.73528968 0.38786935 0.4303943  0.35185714 0.50772758\n",
      " 0.66883333 0.6514881  0.62746917 0.62685714 0.69146429 0.38393559\n",
      " 0.41339735 0.94288462 0.01666667 0.36905556 0.935      0.68135714\n",
      " 0.62376984 0.49356926 0.97283333 0.57056025 0.88382143 0.90964286\n",
      " 0.61596032 0.56547619 0.56517929 0.02141941 0.41174134 0.78660714\n",
      " 0.66577381 0.62183207 0.84983333 0.45466252 0.69991667 0.50301659\n",
      " 0.56087879 0.97       0.57415476 0.93958333 0.81635317 0.68257143\n",
      " 0.75213889 0.67777597 0.79728175 0.88582143 0.49261362 0.39560462\n",
      " 0.61219444 0.72016667 0.03666667 0.60905117 0.77483333 0.5936075\n",
      " 0.49965945 0.69776587 0.70771429 0.66298413 0.78933333 0.6747406\n",
      " 0.06755556 0.4658083  0.58962843 0.36155952 0.98       0.76588492\n",
      " 0.59819573 0.48025086 0.57290859 0.82913095 0.60267711 0.9905\n",
      " 0.42158694 0.13478846 0.72262302 0.58742136 0.71625    0.59442063\n",
      " 0.59101984 0.46311905 0.73963203 0.80680556 0.90661111 0.7685\n",
      " 0.60961697 0.88532143 0.70882143 0.79211111 0.96297619 0.51783242\n",
      " 0.78122619 0.37957071 0.71172294 0.81235317 0.43544372 0.39661419\n",
      " 0.58438095 0.65115079 0.76669048 0.52025072 0.80123016 0.51924603\n",
      " 0.62705556 0.41397024 0.77769841 0.62814286 0.8615119  0.54800794\n",
      " 0.81094048 0.82921176 0.58235965 0.71353247 0.81005952 0.84169048\n",
      " 0.76429762 0.42715512 0.14485714 0.0197619  0.44448629 0.53236905\n",
      " 0.54224026 0.54200794 0.77513095 0.43198088 0.59981349 0.62689322\n",
      " 0.77349206 0.5954062  0.4959787  0.81268254 0.48892388 0.46653752\n",
      " 0.57252128 0.60801587 0.59346429 0.90613095 0.64833333 0.92152381\n",
      " 0.69997222 0.73083333 0.41229076 0.54759632 0.56438961 0.67934799\n",
      " 0.57128571 0.47284993 0.66600794 0.83335714 0.7867406  0.62499711\n",
      " 0.05       0.37072872 0.62583117 0.48499892 0.14850794 0.2722381\n",
      " 0.448364   0.7562619  0.35714177 0.38887085 0.63211652 0.5672886\n",
      " 0.66170238 0.87128571 0.72760859 0.6170749  0.49239683 0.70392063\n",
      " 0.53767605 0.79583333 0.43436905 0.59466667 0.61721429 0.64827381\n",
      " 0.76163492 0.69877214 0.10785714 0.48183838 0.84719048 0.64633333\n",
      " 0.79382143 0.97333333 0.61821501 0.75409524 0.77572619 0.83489286\n",
      " 0.68296978 0.57119048 0.51992587 0.40197222 0.72842696 0.8000119\n",
      " 0.86105556 0.62353896 0.90994444 0.53674206 0.54163203 0.73578571\n",
      " 0.31872641 0.8772619  0.81153175 0.64848016 0.74752778 0.88495599\n",
      " 0.50675794 0.53769048 0.55950419 0.50251659 0.06166667 0.85538095\n",
      " 0.78994877 0.44916342 0.55119913 0.87313889 0.81599206 0.5453378\n",
      " 0.54514683 0.0444127  0.39117893 0.35422619 0.67204076 0.66635714\n",
      " 0.93478571 0.32166613 0.88666667 0.72313095 0.59968651 0.87066667\n",
      " 0.48898485 0.83666667 0.68432684 0.60542063 0.87102381 0.56450433\n",
      " 0.31345346 0.72746429 0.50685823 0.88977381 0.65093867 0.48593218\n",
      " 0.5880873  0.43259524 0.71049603 0.51714633 0.7762619  0.81018254\n",
      " 0.1938956  0.80164286 0.41950938 0.69036111 0.60720635 0.59272908\n",
      " 0.64843651 0.45194408 0.32917216 0.91632143 0.87803571 0.86478571\n",
      " 0.66105303 0.59328571 0.88496429 0.42774759 0.76572619 0.65613095\n",
      " 0.66717857 0.48752778 0.55355952 0.69676623 0.88742424 0.71191667\n",
      " 0.67436416 0.45705411 0.52780556 0.94154762 0.38717929 0.75444048\n",
      " 0.79513095 0.67617857 0.0193956  0.76977778 0.64071825 0.46007937\n",
      " 0.87636905 0.69059524 0.61062698 0.6135754  0.81778175 0.91616667\n",
      " 0.82346429 0.58850397 0.59294048 0.8213373  0.59170635 0.42043254\n",
      " 0.99       0.79914194 0.54277074 0.40077128 0.379      0.50102381\n",
      " 0.60883333 0.64926546 0.892      0.5858956  0.62955952 0.46349351\n",
      " 0.90744048 0.44451476 0.97666667 0.60670058 0.1157619  0.50450108\n",
      " 0.57549512 0.63109127 0.87264286 0.77054113 0.65810714 0.40208442\n",
      " 0.21841941 0.88474603 0.42739502 0.57019841 0.80973377 0.82878175\n",
      " 0.85377778 0.70072619 0.86816667 0.67492857 0.40101984 0.72809127\n",
      " 0.41024156 0.76035859 0.06355952 0.64671429 0.27266667 0.48404762\n",
      " 0.6455     0.45394661 0.14859524 0.6116562  0.7273254  0.7639062\n",
      " 0.48611111 0.56638997 0.38666847 0.59792208 0.32971517 0.76197619\n",
      " 0.89909524 0.68320238 0.89464286 0.09087179 0.75965476 0.67578571\n",
      " 0.5585119  0.53593651 0.95291667 0.95233333 0.45993001 0.98532468\n",
      " 0.16653846 0.93375    0.83535714 0.70997619 0.47512585 0.91917857\n",
      " 0.43424603 0.80219048 0.55673557 0.5223254  0.10746429 0.63802381\n",
      " 0.50076441 0.64653968 0.76328571 0.53734921 0.79578571 0.6487312\n",
      " 0.43820527 0.66292202 0.67446429 0.75732814 0.79484524 0.54774731\n",
      " 0.72428571 0.49862338 0.63835456 0.74325794 0.48398052 0.50309596\n",
      " 0.4143804  0.93083333 0.54153571 0.60584524 0.51284704 0.65102778\n",
      " 0.31244715 0.47137662 0.85835714 0.64771429 0.54006746 0.55857143\n",
      " 0.32060498 0.75414286 0.68685553 0.65803716 0.85469048 0.52251587\n",
      " 0.7327381  0.64419048 0.59061905 0.53255339 0.5631746  0.49603175\n",
      " 0.4549614  0.84658333 0.90358333 0.40605339 0.40081602 0.8147619\n",
      " 0.94866667 0.84616667 0.57564502 0.5490754  0.3347215  0.62742532\n",
      " 0.91783333 0.55980159 0.88354762 0.68608333 0.90642857 0.35429672\n",
      " 0.79301365 0.67319444 0.37586183 0.63845774 0.91989286 0.58582937\n",
      " 0.867      0.58934596 0.80646429 0.42984416 0.69621825 0.45950104\n",
      " 0.64250794 0.73399206 0.58465152 0.56659127 0.55189033 0.48313276\n",
      " 0.63840476 0.4569228  0.76135714 0.57907143 0.77242857 0.63508586\n",
      " 0.36170238 0.623509   0.53528175 0.76090476 0.7152482  0.44692677\n",
      " 0.80834524 0.53635859 0.37730014 0.40934921 0.88909524 0.65820879\n",
      " 0.68878571 0.57478571 0.4026912  0.61169048 0.76317063 0.41206043\n",
      " 0.63256746 0.09333333 0.52307251 0.82280159 0.79002381 0.64021898\n",
      " 0.55052201 0.89874603 0.52315096 0.60175    0.55852525 0.8184246\n",
      " 0.44782053 0.67301587 0.88641667 0.75998443 0.73602381 0.12522894\n",
      " 0.76984921 0.75261111 0.5174455  0.46250036 0.66565476 0.68044913\n",
      " 0.56322375 0.46783584 0.55000902 0.55457937 0.50052381 0.78199206\n",
      " 0.5910754  0.73380556 0.55131313 0.81137302 0.34649278 0.90083333\n",
      " 0.8107767  0.47707937 0.59786255 0.43908933 0.64815476 0.54838889\n",
      " 0.53634632 0.11488889 0.89071429 0.8808254  0.38820796 0.48089064\n",
      " 0.77505159 0.52122222 0.71736905 0.05280037 0.63877381 0.80753175\n",
      " 0.79020238 0.47693651 0.66215476 0.79840079 0.48550647 0.40820504\n",
      " 0.65398413 0.47509921 0.90541392 0.57138595 0.66510714 0.80741667\n",
      " 0.85611905 0.02       0.73402381 0.77064286 0.54016017 0.62596248\n",
      " 0.9555     1.         0.47909722 0.39487843 0.638      0.83034524\n",
      " 0.99       0.75130952 0.4954152  0.86846429 0.56149603 0.86004762\n",
      " 0.7008254  0.70718254 0.94716667 0.9475     0.57103571 0.92016667\n",
      " 0.34497294 0.78792857 0.62747619 0.40930556 0.67183333 0.37532359\n",
      " 0.48954834 0.41977381 0.53071429 0.55050216 0.72569048 0.8770671\n",
      " 0.95715801 0.51727381 0.84855952 0.91442857 0.73307143 0.80851587\n",
      " 0.84082143 0.85008059 0.27271429 0.03666667 0.45583405 0.12222222\n",
      " 0.864      0.7907381  0.88994444 0.88589286 0.67883333 0.60234163\n",
      " 0.40073088 0.953      0.64835317 0.893      0.89191667 0.52757251\n",
      " 0.61663004 0.51506025 0.57743182 0.52419048 0.35208039 0.73360714\n",
      " 0.87625    0.57855844 0.48676335 0.84839683 0.59146481 0.44279804\n",
      " 0.41785786 0.47821825 0.69441667 0.60378175 0.64753463 0.4335653\n",
      " 0.57283333 0.85958333 0.62083333 0.55592857 0.92366667 0.1837619\n",
      " 0.63881868 0.59256982 0.44547763 0.47148954 0.50650794 0.28555952\n",
      " 0.40494841 0.75363492 0.99218182 0.60377235 0.41056746 0.75452381\n",
      " 0.46617979 0.57315152 0.62810714 0.56555556 0.79219048 0.6303658\n",
      " 0.51604509 0.91320238 0.45944625 0.39959524 0.48565693 0.81244048\n",
      " 0.71001984 0.53468723 0.56155356 0.49086255 0.45391739 0.51196703\n",
      " 0.91530556 0.68430952 0.8117381  0.74989683 0.11416667 0.57970238\n",
      " 0.68159524 0.39001118 0.4410873  0.55963095 0.60065873 0.60137266\n",
      " 0.59859993 0.8114246  0.70086508 0.63968651 0.92919048 0.58033333\n",
      " 0.97       0.994      0.90378571 0.43201587 0.73819048 0.4311526\n",
      " 0.73146429 0.49311869 0.80015873 0.87298016 0.36502958 0.70365959\n",
      " 0.79413492 0.61955375 0.46687879 0.45076299 0.83280159 0.50518254\n",
      " 0.81271429 0.65075    0.63609554 0.81780556 0.51755952 0.76806277\n",
      " 0.79378175 0.53771337 0.65279437 0.60695238 0.64912013 0.75101587\n",
      " 0.87096429 0.45159088 0.37485714 0.83824026 0.42978746 0.53219444\n",
      " 0.51896573 0.61516017 0.52378053 0.38310534 0.56958333 0.82489408\n",
      " 0.79477778 0.7293254  0.64619048 0.71241667 0.37680717 0.62392857\n",
      " 0.78559524 0.53924625 0.86575    0.58667929 0.64461905 0.71244322\n",
      " 0.77890476 0.5560994  0.73654762 0.44888131 0.89442063 0.46535317\n",
      " 0.72094697 0.86425    0.79166667 0.60312698 0.76428175 0.38023882\n",
      " 0.77413492 0.46038961 0.92738095 0.55128968 0.77114286 0.42064791\n",
      " 0.79465873 0.63675672 0.82445238 0.60500397 0.81936905 0.3971627\n",
      " 0.46451371 0.61687049 0.98833333 0.2487381  0.84641667 0.65917041\n",
      " 0.68634127 0.61608438 0.58933333 0.47199567 0.8574881  0.64319048\n",
      " 0.8482619  0.78878571 0.72435137 0.56425794 0.76165079 0.43168254\n",
      " 0.93609091 0.40271032 0.89088095 0.60843651 0.5085184  0.79228571\n",
      " 0.48447253 0.53628968 0.63523413 0.45596585 0.47757648 0.68048016\n",
      " 0.59004762 0.52799906 0.55313095 0.91155556 0.78621429 0.83542857\n",
      " 0.83972619 0.81415476 0.5680754  0.64117063 0.44316739 0.80585714\n",
      " 0.46719048 0.68350397 0.41334127 0.63283333 0.0497619  0.53311905\n",
      " 0.81499206 0.7272381  0.6302381  0.82175794 0.98333333 0.72098543\n",
      " 0.81971429 0.7810873  0.47458514 0.57339358 0.63341667 0.08883333\n",
      " 0.55533492 0.71550072 0.43749603 0.45669695 0.58042063 0.61989683\n",
      " 0.6355754  0.48979831 0.76455375 0.95125    0.54860714 0.72511905\n",
      " 0.93166667 0.65053571 0.81945238 0.04809524 0.8515     0.81605952\n",
      " 0.8898373  0.54764286 0.83142857 0.81834127 0.882     ]\n"
     ]
    }
   ],
   "source": [
    "score = model.score(val_X, val_y)\n",
    "print('Model score is: {}'.format(score))\n",
    "\n",
    "preds = model.predict(test_X)\n",
    "print(preds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "4e9dce62",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<sklearn.metrics._plot.confusion_matrix.ConfusionMatrixDisplay at 0x7fd1d679ef10>"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#confusion matrix\n",
    "\n",
    "confusion_matrix = sklearn.metrics.confusion_matrix(preds, test_y)\n",
    "sklearn.metrics.ConfusionMatrixDisplay(confusion_matrix).plot()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21feaafb",
   "metadata": {},
   "source": [
    "Convert the above code into functions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "bf1d563c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def train_model(dataframe, columns = [],  target = []):\n",
    "    \"\"\"\n",
    "    Takes dataframe and splits it into a training and testing set. \n",
    "    Note: Data is called train and test, but this test set is currently\n",
    "    closer to a validation set. Keeping nomenclature to keep model robust.\n",
    "    Trains a KNN classifier model with selected data.\n",
    "    \n",
    "    Params\n",
    "    ----------\n",
    "    dataframe: Pandas dataframe \n",
    "    columns: list of strings, representing input features\n",
    "    target: list of strings, representing target feature(s)\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    -Sk-learn model object\n",
    "    -train data (features)\n",
    "    -train data (target)\n",
    "    -validation data (features)\n",
    "    -validation data (target)\n",
    "    \"\"\"\n",
    "    #split data\n",
    "    dev, test = sklearn.model_selection.train_test_split(dataframe, test_size=0.15, random_state=1)\n",
    "    \n",
    "    #test input arguments\n",
    "    assert \"pandas.core.frame.DataFrame\" in str(type(dev))\n",
    "    assert \"pandas.core.frame.DataFrame\" in str(type(test))\n",
    "    assert \"str\" in str(type(columns[0]))\n",
    "    assert \"str\" in str(type(target[0]))\n",
    "   \n",
    "    #split into input and output feature(s)\n",
    "    dev_X = dev[columns].values\n",
    "    test_X = test[columns].values\n",
    "\n",
    "    dev_y = dev[target].values.reshape(-1,1)\n",
    "    test_y = test[target].values.reshape(-1,1)\n",
    "    \n",
    "    #scale data\n",
    "    scaler = sklearn.preprocessing.StandardScaler()\n",
    "    dev_X = scaler.fit_transform(dev_X)\n",
    "    test_X = scaler.fit_transform(test_X)\n",
    "    \n",
    "    #train model\n",
    "    model = sklearn.ensemble.RandomForestClassifier()\n",
    "    model = model.fit(dev_X, dev_y.ravel())\n",
    "    \n",
    "    return model, dev_X, dev_y, test_X, test_y\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "e40acd6d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "model, dev_X, dev_y, test_X, test_y = train_model(df, columns = input_features, \n",
    "                                          target='protein_match')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "46729a81",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "..\n",
      "----------------------------------------------------------------------\n",
      "Ran 2 tests in 14.866s\n",
      "\n",
      "OK\n"
     ]
    }
   ],
   "source": [
    "class TestModelTraining(unittest.TestCase):\n",
    "    \n",
    "    def test_invalid_inputs(self):\n",
    "    \n",
    "        #test that input data type is correct\n",
    "        \n",
    "        try:\n",
    "            train_model([1,2,3], columns = 'string', target = 'string')\n",
    "            self.assertTrue(False)\n",
    "        except AssertionError:\n",
    "            self.assertTrue(True)\n",
    "    \n",
    "#     def test_input_distro(self):\n",
    "        \n",
    "#         #test that dev and test features have similar Jensen Shannon Distribution\n",
    "        \n",
    "#         JSD = (\n",
    "#             scipy.stats.bootstrap((train_reg(split_data(df)[0], split_data(df)[1],\n",
    "#                                 columns = input_features, target=target)[1], train_reg(split_data(df)[0], \n",
    "#                                 split_data(df)[1], columns = input_features, target=target)[3]), \n",
    "#                                   JSD_dev_and_test, n_resamples=1000, \n",
    "#                                   batch=5, method='percentile')\n",
    "#         )\n",
    "\n",
    "#         div = JSD.confidence_interval[1]\n",
    "        \n",
    "#         #asserts that the divergence between data sets is sufficiently low\n",
    "#         assert abs(div) < 0.3, \"Warning! High JSD between dev and test set!\"\n",
    "        \n",
    "    def test_output_format(self):\n",
    "        \n",
    "        #asserts that function returns 4 objects to be assigned to pearson_corr, model, test_X, test_y\n",
    "        assert len(train_model(df, columns = input_features, \n",
    "                                          target = target)) == 5\n",
    "        \n",
    "\n",
    "suite = unittest.TestLoader().loadTestsFromTestCase(TestModelTraining)\n",
    "_ = unittest.TextTestRunner().run(suite)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "47f766ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "#need to fix this\n",
    "\n",
    "def evaluate_model(model, test_X, test_y):\n",
    "    \n",
    "    \"\"\"\n",
    "    Takes a trained model and test data and tests the model.\n",
    "    \n",
    "    Params\n",
    "    ----------\n",
    "    model: sklearn.neighbors.KNeighborsClassifier\n",
    "    test_X: numpy array\n",
    "    test_y: numpy array\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    Vector of predictions based on the model (numpy array)\n",
    "    \"\"\"\n",
    "    \n",
    "    #test input arguments\n",
    "    assert \"sklearn\" in str(type(model))\n",
    "    assert \"numpy.ndarray\" in str(type(test_X))\n",
    "    assert \"numpy.ndarray\" in str(type(test_y))\n",
    "    \n",
    "    preds = model.predict(test_X)\n",
    "    \n",
    "    return preds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "7bbb0242",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([False, False, False, ..., False,  True, False])"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preds = evaluate_model(model, test_X, test_y)\n",
    "preds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "8b2fdac6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "...\n",
      "----------------------------------------------------------------------\n",
      "Ran 3 tests in 236.519s\n",
      "\n",
      "OK\n"
     ]
    }
   ],
   "source": [
    "class TestModelPerformance(unittest.TestCase):\n",
    "\n",
    "    def test_asserts(self):\n",
    "        model, _, _, test_X, test_y = train_model(\n",
    "            df, columns=input_features, target='protein_match'\n",
    "        )\n",
    "        # assert that input types are correct\n",
    "        with self.assertRaises(AssertionError):\n",
    "            evaluate_model(model, [1, 2, 3], test_y)\n",
    "            \n",
    "    def test_model_output(self):\n",
    "        model, _, _, test_X, test_y = train_model(\n",
    "            df, columns=input_features, target='protein_match'\n",
    "        )\n",
    "        # assert output type is correct\n",
    "        output = evaluate_model(model, test_X, test_y)\n",
    "        self.assertIsInstance(output, np.ndarray)\n",
    "        \n",
    "    def test_pred_dimension(self):\n",
    "        model, _, _, test_X, test_y = train_model(df, \n",
    "            columns=input_features, target='protein_match'\n",
    "        )\n",
    "        # want to check that the number of predictions is equal to the number of test examples\n",
    "        preds = evaluate_model(model, test_X, test_y)\n",
    "        self.assertEqual(len(test_y), len(preds))\n",
    "\n",
    "suite = unittest.TestLoader().loadTestsFromTestCase(TestModelPerformance)\n",
    "_ = unittest.TextTestRunner().run(suite)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9b79929",
   "metadata": {},
   "source": [
    "A good test here can be adding outliers and seeing how much the score and predictions change (probably for the above function)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "dae1e64f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_model(model, test_X, test_y):\n",
    "    \"\"\"\n",
    "    Takes a test KNN Classifier model and plots the confusion matrix.\n",
    "    \n",
    "    Params\n",
    "    ----------\n",
    "    model: sklearn.neighbors.KNeighborsClassifier\n",
    "    test_X: numpy array\n",
    "    test_y: numpy array\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    -Confusion predictions vs. observations\n",
    "    -Model score\n",
    "    \"\"\"\n",
    "    \n",
    "    #test input arguments\n",
    "    assert \"sklearn\" in str(type(model))\n",
    "    assert \"numpy.ndarray\" in str(type(test_X))\n",
    "    assert \"numpy.ndarray\" in str(type(test_y))\n",
    "    \n",
    "    score = model.score(test_X, test_y)\n",
    "    preds = evaluate_model(model, test_X, test_y)\n",
    "   \n",
    "    # plot confusion matrix\n",
    "    confusion_matrix = sklearn.metrics.confusion_matrix(preds, test_y)\n",
    "    cm_plot = sklearn.metrics.ConfusionMatrixDisplay(confusion_matrix)\n",
    "    \n",
    "    cm_plot.plot(cmap=plt.cm.Blues)\n",
    "    cm_plot.ax_.set_title('Confusion Matrix')\n",
    "    \n",
    "    return score\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "c923e39b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9904"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#make some appropriate display labels here\n",
    "\n",
    "plot_model(model, test_X, test_y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "15e9e191",
   "metadata": {},
   "outputs": [],
   "source": [
    "#wrapper function\n",
    "\n",
    "def RF_wrapper(dataframe):\n",
    "    \"\"\"\n",
    "    Takes dataframe and runs it through kNN model.\n",
    "    \n",
    "    Params\n",
    "    ----------\n",
    "    dataframe: Pandas dataframe \n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    -Target feature predictions\n",
    "    -Parity plot\n",
    "    \"\"\"\n",
    "    \n",
    "    assert 'pandas.core.frame.DataFrame' in str(type(dataframe))\n",
    "\n",
    "    #user inputs target feature\n",
    "    target = 'protein_match'\n",
    "    \n",
    "    #define input features\n",
    "    input_features = [columns for columns in dataframe]\n",
    "    \n",
    "    input_features.remove(target)\n",
    "\n",
    "    #train the model based off data split\n",
    "    model, dev_X, dev_y, test_X, test_y = train_model(\n",
    "                                        dataframe, columns=input_features, \n",
    "                                        target=target\n",
    "                                                   )\n",
    "    \n",
    "    #test the model and return predictions\n",
    "    preds = evaluate_model(model, test_X, test_y)\n",
    "\n",
    "    #plot the results of the model\n",
    "    plot_model(model, test_X, test_y)\n",
    "    \n",
    "    return preds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "cbbd8fa8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([False, False, False, ..., False,  True, False])"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RF_wrapper(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "37c0e2b2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "...\n",
      "----------------------------------------------------------------------\n",
      "Ran 3 tests in 158.947s\n",
      "\n",
      "OK\n"
     ]
    }
   ],
   "source": [
    "class TestWrapper(unittest.TestCase):\n",
    "    \n",
    "    def test_wrapper_input(self):\n",
    "        #test that input data type is correct\n",
    "        try:\n",
    "            RF_wrapper([1,2,3])\n",
    "            self.assertTrue(False)\n",
    "        except AssertionError:\n",
    "            self.assertTrue(True)\n",
    "\n",
    "    def test_wrapper_output(self):\n",
    "        model, _, _, test_X, test_y = train_model(\n",
    "            df, \n",
    "            columns=input_features, target='protein_match'\n",
    "        )\n",
    "        # assert output type is correct\n",
    "        output = evaluate_model(model, test_X, test_y)\n",
    "        self.assertIsInstance(output, np.ndarray)\n",
    "        \n",
    "    def test_output_dimension(self):\n",
    "        model, _, _, test_X, test_y = train_model(\n",
    "            df, \n",
    "            columns=input_features, target='protein_match'\n",
    "        )\n",
    "        # want to check that the number of predictions is equal to the number of test examples\n",
    "        preds = evaluate_model(model, test_X, test_y)\n",
    "        self.assertEqual(len(test_y), len(preds))\n",
    "        \n",
    "\n",
    "suite = unittest.TestLoader().loadTestsFromTestCase(TestWrapper)\n",
    "_ = unittest.TextTestRunner().run(suite)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02c53a58",
   "metadata": {},
   "source": [
    "**IGNORE BELOW**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "13d52919",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #plot empirical distribution of scaled_local_query_percent_id\n",
    "\n",
    "# target = df['protein_match']\n",
    "# #create histplot\n",
    "# fig, ax = plt.subplots()\n",
    "\n",
    "# sns.histplot(data = df, x = target)\n",
    "\n",
    "# ax.set_title('Sample Histogram', fontsize=16)\n",
    "# ax.set_xlabel('Class', fontsize=14)\n",
    "# ax.set_ylabel('Count', fontsize=14)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e3d1950",
   "metadata": {},
   "source": [
    "Things I've tried to improve model: \n",
    "\n",
    "1. drop bit scores over 1000\n",
    "2. Switch to 85/15 train/test split\n",
    "3. Lasso regression - not great\n",
    "4. Ridge regression - R2 = 0.87, about the same as normal Linear regression\n",
    "5. KNN regression (n_neighbors optimized at 8) gives best result, R2 = ~0.93 --> \n",
    "*got it up to .942 by removing a few features\n",
    "6. Decision tree regressor was slightly worse than KNN\n",
    "7. Input DT classifier and RF classifier. RF classifier has best performance\n",
    "8. Changed target to binary protein function match"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5220b26b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
