{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6c1fdfb1",
   "metadata": {},
   "source": [
    "The purpose of this notebook is to document data exploration processes with an n=50,000 sample of our Learn2Therm protein database. Doing so provides a workspace to select and optimize a model for predicting protein pair functionality between two sequences. Several classifier architectures were tested prior to the release of the final version of this component."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84922e82",
   "metadata": {},
   "source": [
    "This notebook includes exploration of a Random Forest Classifier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "61c53eee",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats\n",
    "import sklearn.preprocessing\n",
    "import sklearn.model_selection\n",
    "import sklearn.ensemble\n",
    "import sklearn.feature_selection\n",
    "import unittest\n",
    "import iFeatureOmegaCLI\n",
    "import Bio.SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "ba87d47f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/loganroberts/Learn2Therm/ValidProt/FAFSA\n"
     ]
    }
   ],
   "source": [
    "cd /Users/loganroberts/Learn2Therm/ValidProt/FAFSA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "f90c1cdd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#convert to pandas df\n",
    "df = pd.read_csv('learn2therm_sample_50k.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "5d55b25f",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/loganroberts/Learn2Therm/ValidProt/notebooks\n"
     ]
    }
   ],
   "source": [
    "cd /Users/loganroberts/Learn2Therm/ValidProt/notebooks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "49cf5076",
   "metadata": {},
   "outputs": [],
   "source": [
    "#this function is only necessary if we get a fasta file with a descriptor generated from component 3\n",
    "\n",
    "def remove_fasta_description(filename:str):\n",
    "    \n",
    "    \"\"\"\n",
    "    Removes description from fasta file so that iProtein can read the input.\n",
    "    \"\"\"\n",
    "    \n",
    "    #assign unwanted string to object\n",
    "    string_to_remove = \"<unknown description>\"\n",
    "\n",
    "    #open file\n",
    "    with open(filename, \"r\") as file:\n",
    "        content = file.read()\n",
    "    \n",
    "    # Remove the string\n",
    "    new_content = content.replace(string_to_remove, \"\")\n",
    "\n",
    "    #overwrite file without string\n",
    "    with open(filename, \"w\") as file:\n",
    "        seq = file.write(new_content)\n",
    "        \n",
    "    return seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "6d9e9e84",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_protein_descriptors(sequences:str, descriptors=[]):\n",
    "    \n",
    "    \"\"\"\n",
    "    Input: Fasta file with protein sequences. \n",
    "    Must be a .txt file.\n",
    "    \n",
    "    Output: Vector of descriptors\n",
    "    \"\"\"\n",
    "    \n",
    "    #create iProtein object\n",
    "    protein = iFeatureOmegaCLI.iProtein(sequences)\n",
    "    \n",
    "    #not sure why we need this yet. Right now it is stored in local directory.\n",
    "    params = protein.import_parameters('protein_parameters.json')\n",
    "    \n",
    "    protein_descriptors = {}\n",
    "    \n",
    "    for descriptor in descriptors:\n",
    "        protein.get_descriptor(descriptor)\n",
    "        protein_descriptors.update({f'{descriptor}':protein.encodings})\n",
    "        \n",
    "    return protein_descriptors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "6a2ee685",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein = iFeatureOmegaCLI.iProtein('meso_input_copy.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "dfc02c9c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "        ----- Available feature types ------        \n",
      "        \n",
      "        AAC                                                Amino acid composition\n",
      "        EAAC                                               Enhanced amino acid composition\n",
      "        CKSAAP type 1                                      Composition of k-spaced amino acid pairs type 1 - normalized\n",
      "        CKSAAP type 2                                      Composition of k-spaced amino acid pairs type 2 - raw count\n",
      "        DPC type 1                                         Dipeptide composition type 1 - normalized\n",
      "        DPC type 2                                         Dipeptide composition type 2 - raw count\n",
      "        TPC type 1                                         Tripeptide composition type 1 - normalized\n",
      "        TPC type 2                                         Tripeptide composition type 1 - raw count\n",
      "        CTDC                                               Composition\n",
      "        CTDT                                               Transition\n",
      "        CTDD                                               Distribution\n",
      "        CTriad                                             Conjoint triad\n",
      "        KSCTriad                                           Conjoint k-spaced triad\n",
      "        ASDC                                               Adaptive skip dipeptide composition\n",
      "        DistancePair                                       PseAAC of distance-pairs and reduced alphabe\n",
      "        GAAC                                               Grouped amino acid composition\n",
      "        EGAAC                                              Enhanced grouped amino acid composition\n",
      "        CKSAAGP type 1                                     Composition of k-spaced amino acid group pairs type 1- normalized\n",
      "        CKSAAGP type 2                                     Composition of k-spaced amino acid group pairs type 2- raw count\n",
      "        GDPC type 1                                        Grouped dipeptide composition type 1 - normalized\n",
      "        GDPC type 2                                        Grouped dipeptide composition type 2 - raw count\n",
      "        GTPC type 1                                        Grouped tripeptide composition type 1 - normalized\n",
      "        GTPC type 2                                        Grouped tripeptide composition type 1 - raw count\n",
      "        Moran                                              Moran\n",
      "        Geary                                              Geary\n",
      "        NMBroto                                            Normalized Moreau-Broto\n",
      "        AC                                                 Auto covariance\n",
      "        CC                                                 Cross covariance\n",
      "        ACC                                                Auto-cross covariance\n",
      "        SOCNumber                                          Sequence-order-coupling number\n",
      "        QSOrder                                            Quasi-sequence-order descriptors\n",
      "        PAAC                                               Pseudo-amino acid composition\n",
      "        APAAC                                              Amphiphilic PAAC\n",
      "        PseKRAAC type 1                                    Pseudo K-tuple reduced amino acids composition type 1\n",
      "        PseKRAAC type 2                                    Pseudo K-tuple reduced amino acids composition type 2\n",
      "        PseKRAAC type 3A                                   Pseudo K-tuple reduced amino acids composition type 3A\n",
      "        PseKRAAC type 3B                                   Pseudo K-tuple reduced amino acids composition type 3B\n",
      "        PseKRAAC type 4                                    Pseudo K-tuple reduced amino acids composition type 4\n",
      "        PseKRAAC type 5                                    Pseudo K-tuple reduced amino acids composition type 5\n",
      "        PseKRAAC type 6A                                   Pseudo K-tuple reduced amino acids composition type 6A\n",
      "        PseKRAAC type 6B                                   Pseudo K-tuple reduced amino acids composition type 6B\n",
      "        PseKRAAC type 6C                                   Pseudo K-tuple reduced amino acids composition type 6C\n",
      "        PseKRAAC type 7                                    Pseudo K-tuple reduced amino acids composition type 7\n",
      "        PseKRAAC type 8                                    Pseudo K-tuple reduced amino acids composition type 8\n",
      "        PseKRAAC type 9                                    Pseudo K-tuple reduced amino acids composition type 9\n",
      "        PseKRAAC type 10                                   Pseudo K-tuple reduced amino acids composition type 10\n",
      "        PseKRAAC type 11                                   Pseudo K-tuple reduced amino acids composition type 11\n",
      "        PseKRAAC type 12                                   Pseudo K-tuple reduced amino acids composition type 12\n",
      "        PseKRAAC type 13                                   Pseudo K-tuple reduced amino acids composition type 13\n",
      "        PseKRAAC type 14                                   Pseudo K-tuple reduced amino acids composition type 14\n",
      "        PseKRAAC type 15                                   Pseudo K-tuple reduced amino acids composition type 15\n",
      "        PseKRAAC type 16                                   Pseudo K-tuple reduced amino acids composition type 16\n",
      "        binary                                             Binary\n",
      "        binary_6bit                                        Binary\n",
      "        binary_5bit type 1                                 Binary\n",
      "        binary_5bit type 2                                 Binary\n",
      "        binary_3bit type 1                                 Binary\n",
      "        binary_3bit type 2                                 Binary\n",
      "        binary_3bit type 3                                 Binary\n",
      "        binary_3bit type 4                                 Binary\n",
      "        binary_3bit type 5                                 Binary\n",
      "        binary_3bit type 6                                 Binary\n",
      "        binary_3bit type 7                                 Binary\n",
      "        AESNN3                                             Learn from alignments\n",
      "        OPF_10bit                                          Overlapping property features - 10 bit\n",
      "        OPF_7bit type 1                                    Overlapping property features - 7 bit type 1\n",
      "        OPF_7bit type 2                                    Overlapping property features - 7 bit type 2\n",
      "        OPF_7bit type 3                                    Overlapping property features - 7 bit type 3\n",
      "        AAIndex                                            AAIndex\n",
      "        BLOSUM62                                           BLOSUM62\n",
      "        ZScale                                             Z-Scales index\n",
      "        KNN                                                K-nearest neighbor\n",
      "\n",
      "        Note: the first column is the names of availables feature types while the second column is description.  \n",
      "        \n",
      "        \n"
     ]
    }
   ],
   "source": [
    "protein.display_feature_types()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "35e7c460",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File imported successfully.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'AAC':           AAC_A     AAC_C     AAC_D     AAC_E     AAC_F     AAC_G     AAC_H  \\\n",
       " 12897  0.145631  0.009709  0.072816  0.067961  0.014563  0.067961  0.019417   \n",
       " 13026  0.064378  0.000000  0.115880  0.068670  0.030043  0.051502  0.008584   \n",
       " 8203   0.205575  0.006969  0.062718  0.034843  0.031359  0.094077  0.020906   \n",
       " 3340   0.119658  0.000000  0.012821  0.042735  0.055556  0.094017  0.004274   \n",
       " 14020  0.050000  0.000000  0.025000  0.050000  0.100000  0.075000  0.012500   \n",
       " ...         ...       ...       ...       ...       ...       ...       ...   \n",
       " 3888   0.184906  0.011321  0.075472  0.049057  0.026415  0.109434  0.011321   \n",
       " 2837   0.136364  0.007576  0.041667  0.075758  0.034091  0.090909  0.026515   \n",
       " 15339  0.114007  0.003257  0.039088  0.048860  0.042345  0.074919  0.026059   \n",
       " 13840  0.145631  0.009709  0.067961  0.043689  0.029126  0.053398  0.058252   \n",
       " 11636  0.141379  0.003448  0.051724  0.065517  0.003448  0.093103  0.027586   \n",
       " \n",
       "           AAC_I     AAC_K     AAC_L     AAC_M     AAC_N     AAC_P     AAC_Q  \\\n",
       " 12897  0.014563  0.014563  0.106796  0.009709  0.004854  0.053398  0.038835   \n",
       " 13026  0.051502  0.055794  0.107296  0.055794  0.025751  0.025751  0.042918   \n",
       " 8203   0.024390  0.010453  0.094077  0.010453  0.020906  0.034843  0.013937   \n",
       " 3340   0.059829  0.038462  0.153846  0.029915  0.008547  0.047009  0.025641   \n",
       " 14020  0.062500  0.012500  0.100000  0.050000  0.037500  0.050000  0.037500   \n",
       " ...         ...       ...       ...       ...       ...       ...       ...   \n",
       " 3888   0.011321  0.015094  0.143396  0.015094  0.003774  0.052830  0.030189   \n",
       " 2837   0.049242  0.022727  0.121212  0.030303  0.041667  0.026515  0.041667   \n",
       " 15339  0.045603  0.052117  0.127036  0.022801  0.022801  0.055375  0.039088   \n",
       " 13840  0.063107  0.004854  0.097087  0.004854  0.009709  0.053398  0.024272   \n",
       " 11636  0.017241  0.006897  0.134483  0.017241  0.006897  0.082759  0.027586   \n",
       " \n",
       "           AAC_R     AAC_S     AAC_T     AAC_V     AAC_W     AAC_Y  \n",
       " 12897  0.131068  0.033981  0.048544  0.106796  0.009709  0.029126  \n",
       " 13026  0.090129  0.042918  0.047210  0.081545  0.004292  0.030043  \n",
       " 8203   0.104530  0.031359  0.055749  0.114983  0.010453  0.017422  \n",
       " 3340   0.051282  0.055556  0.051282  0.094017  0.029915  0.025641  \n",
       " 14020  0.087500  0.037500  0.025000  0.087500  0.037500  0.062500  \n",
       " ...         ...       ...       ...       ...       ...       ...  \n",
       " 3888   0.071698  0.033962  0.045283  0.094340  0.007547  0.007547  \n",
       " 2837   0.087121  0.056818  0.034091  0.049242  0.003788  0.022727  \n",
       " 15339  0.052117  0.065147  0.055375  0.078176  0.013029  0.022801  \n",
       " 13840  0.101942  0.053398  0.038835  0.092233  0.029126  0.019417  \n",
       " 11636  0.110345  0.031034  0.055172  0.100000  0.013793  0.010345  \n",
       " \n",
       " [500 rows x 20 columns],\n",
       " 'GAAC':        GAAC_alphatic  GAAC_aromatic  GAAC_postivecharge  GAAC_negativecharge  \\\n",
       " 12897       0.451456       0.053398            0.165049             0.140777   \n",
       " 13026       0.412017       0.064378            0.154506             0.184549   \n",
       " 8203        0.543554       0.059233            0.135889             0.097561   \n",
       " 3340        0.551282       0.111111            0.094017             0.055556   \n",
       " 14020       0.425000       0.200000            0.112500             0.075000   \n",
       " ...              ...            ...                 ...                  ...   \n",
       " 3888        0.558491       0.041509            0.098113             0.124528   \n",
       " 2837        0.477273       0.060606            0.136364             0.117424   \n",
       " 15339       0.462541       0.078176            0.130293             0.087948   \n",
       " 13840       0.456311       0.077670            0.165049             0.111650   \n",
       " 11636       0.503448       0.027586            0.144828             0.117241   \n",
       " \n",
       "        GAAC_uncharge  \n",
       " 12897       0.189320  \n",
       " 13026       0.184549  \n",
       " 8203        0.163763  \n",
       " 3340        0.188034  \n",
       " 14020       0.187500  \n",
       " ...              ...  \n",
       " 3888        0.177358  \n",
       " 2837        0.208333  \n",
       " 15339       0.241042  \n",
       " 13840       0.189320  \n",
       " 11636       0.206897  \n",
       " \n",
       " [500 rows x 5 columns]}"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_protein_descriptors(\"meso_input_copy.fasta.txt\", descriptors=['AAC', 'GAAC'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "177b9443",
   "metadata": {},
   "outputs": [],
   "source": [
    "#combine both of the above functions\n",
    "\n",
    "def fasta_to_descriptors(fasta:str):\n",
    "\n",
    "    #remove description from fasta file\n",
    "    remove_fasta_description(fasta)\n",
    "    \n",
    "    #return protein descriptors\n",
    "    return get_protein_descriptors(fasta, descriptors=['AAC', 'GAAC'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "22c12a19",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File imported successfully.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAC_A</th>\n",
       "      <th>AAC_C</th>\n",
       "      <th>AAC_D</th>\n",
       "      <th>AAC_E</th>\n",
       "      <th>AAC_F</th>\n",
       "      <th>AAC_G</th>\n",
       "      <th>AAC_H</th>\n",
       "      <th>AAC_I</th>\n",
       "      <th>AAC_K</th>\n",
       "      <th>AAC_L</th>\n",
       "      <th>AAC_M</th>\n",
       "      <th>AAC_N</th>\n",
       "      <th>AAC_P</th>\n",
       "      <th>AAC_Q</th>\n",
       "      <th>AAC_R</th>\n",
       "      <th>AAC_S</th>\n",
       "      <th>AAC_T</th>\n",
       "      <th>AAC_V</th>\n",
       "      <th>AAC_W</th>\n",
       "      <th>AAC_Y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12897</th>\n",
       "      <td>0.145631</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.072816</td>\n",
       "      <td>0.067961</td>\n",
       "      <td>0.014563</td>\n",
       "      <td>0.067961</td>\n",
       "      <td>0.019417</td>\n",
       "      <td>0.014563</td>\n",
       "      <td>0.014563</td>\n",
       "      <td>0.106796</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.004854</td>\n",
       "      <td>0.053398</td>\n",
       "      <td>0.038835</td>\n",
       "      <td>0.131068</td>\n",
       "      <td>0.033981</td>\n",
       "      <td>0.048544</td>\n",
       "      <td>0.106796</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.029126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13026</th>\n",
       "      <td>0.064378</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.115880</td>\n",
       "      <td>0.068670</td>\n",
       "      <td>0.030043</td>\n",
       "      <td>0.051502</td>\n",
       "      <td>0.008584</td>\n",
       "      <td>0.051502</td>\n",
       "      <td>0.055794</td>\n",
       "      <td>0.107296</td>\n",
       "      <td>0.055794</td>\n",
       "      <td>0.025751</td>\n",
       "      <td>0.025751</td>\n",
       "      <td>0.042918</td>\n",
       "      <td>0.090129</td>\n",
       "      <td>0.042918</td>\n",
       "      <td>0.047210</td>\n",
       "      <td>0.081545</td>\n",
       "      <td>0.004292</td>\n",
       "      <td>0.030043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8203</th>\n",
       "      <td>0.205575</td>\n",
       "      <td>0.006969</td>\n",
       "      <td>0.062718</td>\n",
       "      <td>0.034843</td>\n",
       "      <td>0.031359</td>\n",
       "      <td>0.094077</td>\n",
       "      <td>0.020906</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.010453</td>\n",
       "      <td>0.094077</td>\n",
       "      <td>0.010453</td>\n",
       "      <td>0.020906</td>\n",
       "      <td>0.034843</td>\n",
       "      <td>0.013937</td>\n",
       "      <td>0.104530</td>\n",
       "      <td>0.031359</td>\n",
       "      <td>0.055749</td>\n",
       "      <td>0.114983</td>\n",
       "      <td>0.010453</td>\n",
       "      <td>0.017422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3340</th>\n",
       "      <td>0.119658</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012821</td>\n",
       "      <td>0.042735</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.094017</td>\n",
       "      <td>0.004274</td>\n",
       "      <td>0.059829</td>\n",
       "      <td>0.038462</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.029915</td>\n",
       "      <td>0.008547</td>\n",
       "      <td>0.047009</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>0.051282</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.051282</td>\n",
       "      <td>0.094017</td>\n",
       "      <td>0.029915</td>\n",
       "      <td>0.025641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14020</th>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.025000</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.075000</td>\n",
       "      <td>0.012500</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>0.012500</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.087500</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.025000</td>\n",
       "      <td>0.087500</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.062500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3888</th>\n",
       "      <td>0.184906</td>\n",
       "      <td>0.011321</td>\n",
       "      <td>0.075472</td>\n",
       "      <td>0.049057</td>\n",
       "      <td>0.026415</td>\n",
       "      <td>0.109434</td>\n",
       "      <td>0.011321</td>\n",
       "      <td>0.011321</td>\n",
       "      <td>0.015094</td>\n",
       "      <td>0.143396</td>\n",
       "      <td>0.015094</td>\n",
       "      <td>0.003774</td>\n",
       "      <td>0.052830</td>\n",
       "      <td>0.030189</td>\n",
       "      <td>0.071698</td>\n",
       "      <td>0.033962</td>\n",
       "      <td>0.045283</td>\n",
       "      <td>0.094340</td>\n",
       "      <td>0.007547</td>\n",
       "      <td>0.007547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2837</th>\n",
       "      <td>0.136364</td>\n",
       "      <td>0.007576</td>\n",
       "      <td>0.041667</td>\n",
       "      <td>0.075758</td>\n",
       "      <td>0.034091</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>0.026515</td>\n",
       "      <td>0.049242</td>\n",
       "      <td>0.022727</td>\n",
       "      <td>0.121212</td>\n",
       "      <td>0.030303</td>\n",
       "      <td>0.041667</td>\n",
       "      <td>0.026515</td>\n",
       "      <td>0.041667</td>\n",
       "      <td>0.087121</td>\n",
       "      <td>0.056818</td>\n",
       "      <td>0.034091</td>\n",
       "      <td>0.049242</td>\n",
       "      <td>0.003788</td>\n",
       "      <td>0.022727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15339</th>\n",
       "      <td>0.114007</td>\n",
       "      <td>0.003257</td>\n",
       "      <td>0.039088</td>\n",
       "      <td>0.048860</td>\n",
       "      <td>0.042345</td>\n",
       "      <td>0.074919</td>\n",
       "      <td>0.026059</td>\n",
       "      <td>0.045603</td>\n",
       "      <td>0.052117</td>\n",
       "      <td>0.127036</td>\n",
       "      <td>0.022801</td>\n",
       "      <td>0.022801</td>\n",
       "      <td>0.055375</td>\n",
       "      <td>0.039088</td>\n",
       "      <td>0.052117</td>\n",
       "      <td>0.065147</td>\n",
       "      <td>0.055375</td>\n",
       "      <td>0.078176</td>\n",
       "      <td>0.013029</td>\n",
       "      <td>0.022801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13840</th>\n",
       "      <td>0.145631</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.067961</td>\n",
       "      <td>0.043689</td>\n",
       "      <td>0.029126</td>\n",
       "      <td>0.053398</td>\n",
       "      <td>0.058252</td>\n",
       "      <td>0.063107</td>\n",
       "      <td>0.004854</td>\n",
       "      <td>0.097087</td>\n",
       "      <td>0.004854</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.053398</td>\n",
       "      <td>0.024272</td>\n",
       "      <td>0.101942</td>\n",
       "      <td>0.053398</td>\n",
       "      <td>0.038835</td>\n",
       "      <td>0.092233</td>\n",
       "      <td>0.029126</td>\n",
       "      <td>0.019417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11636</th>\n",
       "      <td>0.141379</td>\n",
       "      <td>0.003448</td>\n",
       "      <td>0.051724</td>\n",
       "      <td>0.065517</td>\n",
       "      <td>0.003448</td>\n",
       "      <td>0.093103</td>\n",
       "      <td>0.027586</td>\n",
       "      <td>0.017241</td>\n",
       "      <td>0.006897</td>\n",
       "      <td>0.134483</td>\n",
       "      <td>0.017241</td>\n",
       "      <td>0.006897</td>\n",
       "      <td>0.082759</td>\n",
       "      <td>0.027586</td>\n",
       "      <td>0.110345</td>\n",
       "      <td>0.031034</td>\n",
       "      <td>0.055172</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.013793</td>\n",
       "      <td>0.010345</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>500 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          AAC_A     AAC_C     AAC_D     AAC_E     AAC_F     AAC_G     AAC_H  \\\n",
       "12897  0.145631  0.009709  0.072816  0.067961  0.014563  0.067961  0.019417   \n",
       "13026  0.064378  0.000000  0.115880  0.068670  0.030043  0.051502  0.008584   \n",
       "8203   0.205575  0.006969  0.062718  0.034843  0.031359  0.094077  0.020906   \n",
       "3340   0.119658  0.000000  0.012821  0.042735  0.055556  0.094017  0.004274   \n",
       "14020  0.050000  0.000000  0.025000  0.050000  0.100000  0.075000  0.012500   \n",
       "...         ...       ...       ...       ...       ...       ...       ...   \n",
       "3888   0.184906  0.011321  0.075472  0.049057  0.026415  0.109434  0.011321   \n",
       "2837   0.136364  0.007576  0.041667  0.075758  0.034091  0.090909  0.026515   \n",
       "15339  0.114007  0.003257  0.039088  0.048860  0.042345  0.074919  0.026059   \n",
       "13840  0.145631  0.009709  0.067961  0.043689  0.029126  0.053398  0.058252   \n",
       "11636  0.141379  0.003448  0.051724  0.065517  0.003448  0.093103  0.027586   \n",
       "\n",
       "          AAC_I     AAC_K     AAC_L     AAC_M     AAC_N     AAC_P     AAC_Q  \\\n",
       "12897  0.014563  0.014563  0.106796  0.009709  0.004854  0.053398  0.038835   \n",
       "13026  0.051502  0.055794  0.107296  0.055794  0.025751  0.025751  0.042918   \n",
       "8203   0.024390  0.010453  0.094077  0.010453  0.020906  0.034843  0.013937   \n",
       "3340   0.059829  0.038462  0.153846  0.029915  0.008547  0.047009  0.025641   \n",
       "14020  0.062500  0.012500  0.100000  0.050000  0.037500  0.050000  0.037500   \n",
       "...         ...       ...       ...       ...       ...       ...       ...   \n",
       "3888   0.011321  0.015094  0.143396  0.015094  0.003774  0.052830  0.030189   \n",
       "2837   0.049242  0.022727  0.121212  0.030303  0.041667  0.026515  0.041667   \n",
       "15339  0.045603  0.052117  0.127036  0.022801  0.022801  0.055375  0.039088   \n",
       "13840  0.063107  0.004854  0.097087  0.004854  0.009709  0.053398  0.024272   \n",
       "11636  0.017241  0.006897  0.134483  0.017241  0.006897  0.082759  0.027586   \n",
       "\n",
       "          AAC_R     AAC_S     AAC_T     AAC_V     AAC_W     AAC_Y  \n",
       "12897  0.131068  0.033981  0.048544  0.106796  0.009709  0.029126  \n",
       "13026  0.090129  0.042918  0.047210  0.081545  0.004292  0.030043  \n",
       "8203   0.104530  0.031359  0.055749  0.114983  0.010453  0.017422  \n",
       "3340   0.051282  0.055556  0.051282  0.094017  0.029915  0.025641  \n",
       "14020  0.087500  0.037500  0.025000  0.087500  0.037500  0.062500  \n",
       "...         ...       ...       ...       ...       ...       ...  \n",
       "3888   0.071698  0.033962  0.045283  0.094340  0.007547  0.007547  \n",
       "2837   0.087121  0.056818  0.034091  0.049242  0.003788  0.022727  \n",
       "15339  0.052117  0.065147  0.055375  0.078176  0.013029  0.022801  \n",
       "13840  0.101942  0.053398  0.038835  0.092233  0.029126  0.019417  \n",
       "11636  0.110345  0.031034  0.055172  0.100000  0.013793  0.010345  \n",
       "\n",
       "[500 rows x 20 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = fasta_to_descriptors(\"meso_input_copy.fasta\")['AAC']\n",
    "df2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ca9096d",
   "metadata": {},
   "source": [
    "Let's try and align the dataframe with generated features by index.\n",
    "\n",
    "Note: This requires input of the same AA's, which we dont have rn."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "48123577",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>thermo_index</th>\n",
       "      <th>...</th>\n",
       "      <th>bit_score_16s</th>\n",
       "      <th>m_ogt</th>\n",
       "      <th>t_ogt</th>\n",
       "      <th>ogt_difference</th>\n",
       "      <th>m_protein_seq</th>\n",
       "      <th>t_protein_seq</th>\n",
       "      <th>m_protein_desc</th>\n",
       "      <th>t_protein_desc</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>meso_index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>13251</td>\n",
       "      <td>0.285106</td>\n",
       "      <td>0.259690</td>\n",
       "      <td>0.223706</td>\n",
       "      <td>239</td>\n",
       "      <td>0.926357</td>\n",
       "      <td>238</td>\n",
       "      <td>0.697947</td>\n",
       "      <td>177</td>\n",
       "      <td>14963</td>\n",
       "      <td>...</td>\n",
       "      <td>1155.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>52.5</td>\n",
       "      <td>24.5</td>\n",
       "      <td>MSQPTPASATPPGGHLLDLRDLHVEFRTQEGVAKVINGVSYHLDAG...</td>\n",
       "      <td>MAKRIEVSGLNAYYGRVRAIEDISMTIEPRSVTAFIGPSGCGKSTF...</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>phosphate ABC transporter ATP-binding protein ...</td>\n",
       "      <td>341</td>\n",
       "      <td>258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>22239</td>\n",
       "      <td>0.309783</td>\n",
       "      <td>0.235537</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>184</td>\n",
       "      <td>0.760331</td>\n",
       "      <td>188</td>\n",
       "      <td>0.746032</td>\n",
       "      <td>92</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>1153.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MHIDNTAAIVTGGASGLGAATSQALAERGAKVFAFDLPTAIEGAQG...</td>\n",
       "      <td>MTHRTSTKAPVAVVTGASAGLGLAVTTALTRHGWHVVADARNAARL...</td>\n",
       "      <td>SDR family NAD(P)-dependent oxidoreductase</td>\n",
       "      <td>SDR family oxidoreductase</td>\n",
       "      <td>252</td>\n",
       "      <td>242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>43963</td>\n",
       "      <td>0.257778</td>\n",
       "      <td>0.207885</td>\n",
       "      <td>0.211679</td>\n",
       "      <td>226</td>\n",
       "      <td>0.810036</td>\n",
       "      <td>223</td>\n",
       "      <td>0.828996</td>\n",
       "      <td>148</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>1153.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MTPDPPFGERWALITGATAGIGAAFAERLASDGWHLVLVARDAGRL...</td>\n",
       "      <td>MTRPFAGATAIVTGAASGIGAEIARQMATGGAFVLATDVAEEPLHA...</td>\n",
       "      <td>SDR family oxidoreductase</td>\n",
       "      <td>SDR family oxidoreductase</td>\n",
       "      <td>269</td>\n",
       "      <td>279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>32945</td>\n",
       "      <td>0.324873</td>\n",
       "      <td>0.178771</td>\n",
       "      <td>0.188235</td>\n",
       "      <td>196</td>\n",
       "      <td>0.547486</td>\n",
       "      <td>208</td>\n",
       "      <td>0.645963</td>\n",
       "      <td>94</td>\n",
       "      <td>7134</td>\n",
       "      <td>...</td>\n",
       "      <td>1139.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>52.5</td>\n",
       "      <td>24.5</td>\n",
       "      <td>MDLVSEMTYRRLGDSGLVVSVVGIGCNNFGRKLDLDGTRAVVDAAL...</td>\n",
       "      <td>MQEWNPTRRQLMKAGGLGAAAAVVPSMLAGEAAASSAPSEPKDLIT...</td>\n",
       "      <td>aldo/keto reductase</td>\n",
       "      <td>aldo/keto reductase</td>\n",
       "      <td>322</td>\n",
       "      <td>358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>49249</td>\n",
       "      <td>0.347826</td>\n",
       "      <td>0.226415</td>\n",
       "      <td>0.261343</td>\n",
       "      <td>203</td>\n",
       "      <td>0.638365</td>\n",
       "      <td>211</td>\n",
       "      <td>0.905579</td>\n",
       "      <td>170</td>\n",
       "      <td>14963</td>\n",
       "      <td>...</td>\n",
       "      <td>1155.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>52.5</td>\n",
       "      <td>24.5</td>\n",
       "      <td>MTQLQARGLVKSYGPTPALHGITLDVAEGEIVAVTGPSGCGKSTLL...</td>\n",
       "      <td>MIEVRDVRKTFTVRSGRLRRRRTMVRAVDGVSFTVREGEFVGYLGS...</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>ATP-binding cassette domain-containing protein</td>\n",
       "      <td>233</td>\n",
       "      <td>318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16345</th>\n",
       "      <td>24189</td>\n",
       "      <td>0.350962</td>\n",
       "      <td>0.323009</td>\n",
       "      <td>0.268382</td>\n",
       "      <td>207</td>\n",
       "      <td>0.915929</td>\n",
       "      <td>230</td>\n",
       "      <td>0.723270</td>\n",
       "      <td>118</td>\n",
       "      <td>16507</td>\n",
       "      <td>...</td>\n",
       "      <td>1150.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MSERVRIDVGEGSPEGANSAYLLPDRGVLIDPGPPTERAWSNLLAG...</td>\n",
       "      <td>MSETERANVHALPISIEYGGRPLTITPVVLETDRGLVLVDVGPEGA...</td>\n",
       "      <td>MBL fold metallo-hydrolase</td>\n",
       "      <td>MBL fold metallo-hydrolase</td>\n",
       "      <td>318</td>\n",
       "      <td>226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16345</th>\n",
       "      <td>46461</td>\n",
       "      <td>0.344444</td>\n",
       "      <td>0.187879</td>\n",
       "      <td>0.195276</td>\n",
       "      <td>199</td>\n",
       "      <td>0.603030</td>\n",
       "      <td>183</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>131</td>\n",
       "      <td>16507</td>\n",
       "      <td>...</td>\n",
       "      <td>1150.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>METGIAQEVTAGDCTDLYYIDTGMYETDEYGAVYVLESERPAVVET...</td>\n",
       "      <td>MERISLSNGAFEGNNNAYLFAGGEDVVLVDTGDWMPATREQLEAEF...</td>\n",
       "      <td>MBL fold metallo-hydrolase</td>\n",
       "      <td>MBL fold metallo-hydrolase</td>\n",
       "      <td>305</td>\n",
       "      <td>330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16345</th>\n",
       "      <td>17892</td>\n",
       "      <td>0.286996</td>\n",
       "      <td>0.224561</td>\n",
       "      <td>0.247104</td>\n",
       "      <td>238</td>\n",
       "      <td>0.835088</td>\n",
       "      <td>220</td>\n",
       "      <td>0.944206</td>\n",
       "      <td>220</td>\n",
       "      <td>16507</td>\n",
       "      <td>...</td>\n",
       "      <td>1150.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MIEFQSVTYGFEGPPVLEDCSLEIDDGEFVLLAGANGSGKTTLLRH...</td>\n",
       "      <td>MSDADATTSTESAASNEVTTTGADSRRRLDDPILELEGVVKRFGGI...</td>\n",
       "      <td>energy-coupling factor ABC transporter ATP-bin...</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>233</td>\n",
       "      <td>285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16345</th>\n",
       "      <td>9073</td>\n",
       "      <td>0.395833</td>\n",
       "      <td>0.295950</td>\n",
       "      <td>0.301587</td>\n",
       "      <td>240</td>\n",
       "      <td>0.747664</td>\n",
       "      <td>239</td>\n",
       "      <td>0.773463</td>\n",
       "      <td>391</td>\n",
       "      <td>16507</td>\n",
       "      <td>...</td>\n",
       "      <td>1150.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MSESDTAGDRARLVASRLVDAVRERSATRAGTRLRSNGQVRLGSAI...</td>\n",
       "      <td>MSLETPPTDTDSLDRRKLRWQRRLEAVRNWLNVLKADRMGQIGLTI...</td>\n",
       "      <td>ABC transporter permease</td>\n",
       "      <td>ABC transporter permease</td>\n",
       "      <td>309</td>\n",
       "      <td>321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16564</th>\n",
       "      <td>36348</td>\n",
       "      <td>0.619403</td>\n",
       "      <td>0.617100</td>\n",
       "      <td>0.615955</td>\n",
       "      <td>267</td>\n",
       "      <td>0.992565</td>\n",
       "      <td>269</td>\n",
       "      <td>0.996296</td>\n",
       "      <td>782</td>\n",
       "      <td>16424</td>\n",
       "      <td>...</td>\n",
       "      <td>1161.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>MSVAAGGGGEPAIDPKKTAALGIIGGFVGIYLTPYFGVIGPLLAAL...</td>\n",
       "      <td>MTVAAGGPTAAAIPESRLLALGIIGGLVGIYLFGVPVIGSLLACLG...</td>\n",
       "      <td>tetrahydromethanopterin S-methyltransferase su...</td>\n",
       "      <td>tetrahydromethanopterin S-methyltransferase su...</td>\n",
       "      <td>270</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>50000 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Unnamed: 0  local_gap_compressed_percent_id  \\\n",
       "meso_index                                                \n",
       "5                13251                         0.285106   \n",
       "5                22239                         0.309783   \n",
       "5                43963                         0.257778   \n",
       "5                32945                         0.324873   \n",
       "5                49249                         0.347826   \n",
       "...                ...                              ...   \n",
       "16345            24189                         0.350962   \n",
       "16345            46461                         0.344444   \n",
       "16345            17892                         0.286996   \n",
       "16345             9073                         0.395833   \n",
       "16564            36348                         0.619403   \n",
       "\n",
       "            scaled_local_query_percent_id  scaled_local_symmetric_percent_id  \\\n",
       "meso_index                                                                     \n",
       "5                                0.259690                           0.223706   \n",
       "5                                0.235537                           0.230769   \n",
       "5                                0.207885                           0.211679   \n",
       "5                                0.178771                           0.188235   \n",
       "5                                0.226415                           0.261343   \n",
       "...                                   ...                                ...   \n",
       "16345                            0.323009                           0.268382   \n",
       "16345                            0.187879                           0.195276   \n",
       "16345                            0.224561                           0.247104   \n",
       "16345                            0.295950                           0.301587   \n",
       "16564                            0.617100                           0.615955   \n",
       "\n",
       "            query_align_len  query_align_cov  subject_align_len  \\\n",
       "meso_index                                                        \n",
       "5                       239         0.926357                238   \n",
       "5                       184         0.760331                188   \n",
       "5                       226         0.810036                223   \n",
       "5                       196         0.547486                208   \n",
       "5                       203         0.638365                211   \n",
       "...                     ...              ...                ...   \n",
       "16345                   207         0.915929                230   \n",
       "16345                   199         0.603030                183   \n",
       "16345                   238         0.835088                220   \n",
       "16345                   240         0.747664                239   \n",
       "16564                   267         0.992565                269   \n",
       "\n",
       "            subject_align_cov  bit_score  thermo_index  ...  bit_score_16s  \\\n",
       "meso_index                                              ...                  \n",
       "5                    0.697947        177         14963  ...         1155.0   \n",
       "5                    0.746032         92           875  ...         1153.0   \n",
       "5                    0.828996        148           875  ...         1153.0   \n",
       "5                    0.645963         94          7134  ...         1139.0   \n",
       "5                    0.905579        170         14963  ...         1155.0   \n",
       "...                       ...        ...           ...  ...            ...   \n",
       "16345                0.723270        118         16507  ...         1150.0   \n",
       "16345                0.600000        131         16507  ...         1150.0   \n",
       "16345                0.944206        220         16507  ...         1150.0   \n",
       "16345                0.773463        391         16507  ...         1150.0   \n",
       "16564                0.996296        782         16424  ...         1161.0   \n",
       "\n",
       "            m_ogt  t_ogt  ogt_difference  \\\n",
       "meso_index                                 \n",
       "5            28.0   52.5            24.5   \n",
       "5            28.0   50.0            22.0   \n",
       "5            28.0   50.0            22.0   \n",
       "5            28.0   52.5            24.5   \n",
       "5            28.0   52.5            24.5   \n",
       "...           ...    ...             ...   \n",
       "16345        30.0   52.0            22.0   \n",
       "16345        30.0   52.0            22.0   \n",
       "16345        30.0   52.0            22.0   \n",
       "16345        30.0   52.0            22.0   \n",
       "16564        35.0   70.0            35.0   \n",
       "\n",
       "                                                m_protein_seq  \\\n",
       "meso_index                                                      \n",
       "5           MSQPTPASATPPGGHLLDLRDLHVEFRTQEGVAKVINGVSYHLDAG...   \n",
       "5           MHIDNTAAIVTGGASGLGAATSQALAERGAKVFAFDLPTAIEGAQG...   \n",
       "5           MTPDPPFGERWALITGATAGIGAAFAERLASDGWHLVLVARDAGRL...   \n",
       "5           MDLVSEMTYRRLGDSGLVVSVVGIGCNNFGRKLDLDGTRAVVDAAL...   \n",
       "5           MTQLQARGLVKSYGPTPALHGITLDVAEGEIVAVTGPSGCGKSTLL...   \n",
       "...                                                       ...   \n",
       "16345       MSERVRIDVGEGSPEGANSAYLLPDRGVLIDPGPPTERAWSNLLAG...   \n",
       "16345       METGIAQEVTAGDCTDLYYIDTGMYETDEYGAVYVLESERPAVVET...   \n",
       "16345       MIEFQSVTYGFEGPPVLEDCSLEIDDGEFVLLAGANGSGKTTLLRH...   \n",
       "16345       MSESDTAGDRARLVASRLVDAVRERSATRAGTRLRSNGQVRLGSAI...   \n",
       "16564       MSVAAGGGGEPAIDPKKTAALGIIGGFVGIYLTPYFGVIGPLLAAL...   \n",
       "\n",
       "                                                t_protein_seq  \\\n",
       "meso_index                                                      \n",
       "5           MAKRIEVSGLNAYYGRVRAIEDISMTIEPRSVTAFIGPSGCGKSTF...   \n",
       "5           MTHRTSTKAPVAVVTGASAGLGLAVTTALTRHGWHVVADARNAARL...   \n",
       "5           MTRPFAGATAIVTGAASGIGAEIARQMATGGAFVLATDVAEEPLHA...   \n",
       "5           MQEWNPTRRQLMKAGGLGAAAAVVPSMLAGEAAASSAPSEPKDLIT...   \n",
       "5           MIEVRDVRKTFTVRSGRLRRRRTMVRAVDGVSFTVREGEFVGYLGS...   \n",
       "...                                                       ...   \n",
       "16345       MSETERANVHALPISIEYGGRPLTITPVVLETDRGLVLVDVGPEGA...   \n",
       "16345       MERISLSNGAFEGNNNAYLFAGGEDVVLVDTGDWMPATREQLEAEF...   \n",
       "16345       MSDADATTSTESAASNEVTTTGADSRRRLDDPILELEGVVKRFGGI...   \n",
       "16345       MSLETPPTDTDSLDRRKLRWQRRLEAVRNWLNVLKADRMGQIGLTI...   \n",
       "16564       MTVAAGGPTAAAIPESRLLALGIIGGLVGIYLFGVPVIGSLLACLG...   \n",
       "\n",
       "                                               m_protein_desc  \\\n",
       "meso_index                                                      \n",
       "5                         ABC transporter ATP-binding protein   \n",
       "5                  SDR family NAD(P)-dependent oxidoreductase   \n",
       "5                                   SDR family oxidoreductase   \n",
       "5                                         aldo/keto reductase   \n",
       "5                         ABC transporter ATP-binding protein   \n",
       "...                                                       ...   \n",
       "16345                              MBL fold metallo-hydrolase   \n",
       "16345                              MBL fold metallo-hydrolase   \n",
       "16345       energy-coupling factor ABC transporter ATP-bin...   \n",
       "16345                                ABC transporter permease   \n",
       "16564       tetrahydromethanopterin S-methyltransferase su...   \n",
       "\n",
       "                                               t_protein_desc  m_protein_len  \\\n",
       "meso_index                                                                     \n",
       "5           phosphate ABC transporter ATP-binding protein ...            341   \n",
       "5                                   SDR family oxidoreductase            252   \n",
       "5                                   SDR family oxidoreductase            269   \n",
       "5                                         aldo/keto reductase            322   \n",
       "5              ATP-binding cassette domain-containing protein            233   \n",
       "...                                                       ...            ...   \n",
       "16345                              MBL fold metallo-hydrolase            318   \n",
       "16345                              MBL fold metallo-hydrolase            305   \n",
       "16345                     ABC transporter ATP-binding protein            233   \n",
       "16345                                ABC transporter permease            309   \n",
       "16564       tetrahydromethanopterin S-methyltransferase su...            270   \n",
       "\n",
       "            t_protein_len  \n",
       "meso_index                 \n",
       "5                     258  \n",
       "5                     242  \n",
       "5                     279  \n",
       "5                     358  \n",
       "5                     318  \n",
       "...                   ...  \n",
       "16345                 226  \n",
       "16345                 330  \n",
       "16345                 285  \n",
       "16345                 321  \n",
       "16564                 269  \n",
       "\n",
       "[50000 rows x 29 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.set_index('meso_index').sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "36b12d9a",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAC_A</th>\n",
       "      <th>AAC_C</th>\n",
       "      <th>AAC_D</th>\n",
       "      <th>AAC_E</th>\n",
       "      <th>AAC_F</th>\n",
       "      <th>AAC_G</th>\n",
       "      <th>AAC_H</th>\n",
       "      <th>AAC_I</th>\n",
       "      <th>AAC_K</th>\n",
       "      <th>AAC_L</th>\n",
       "      <th>AAC_M</th>\n",
       "      <th>AAC_N</th>\n",
       "      <th>AAC_P</th>\n",
       "      <th>AAC_Q</th>\n",
       "      <th>AAC_R</th>\n",
       "      <th>AAC_S</th>\n",
       "      <th>AAC_T</th>\n",
       "      <th>AAC_V</th>\n",
       "      <th>AAC_W</th>\n",
       "      <th>AAC_Y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>0.164000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.076000</td>\n",
       "      <td>0.044000</td>\n",
       "      <td>0.032000</td>\n",
       "      <td>0.108000</td>\n",
       "      <td>0.020000</td>\n",
       "      <td>0.032000</td>\n",
       "      <td>0.008000</td>\n",
       "      <td>0.112000</td>\n",
       "      <td>0.020000</td>\n",
       "      <td>0.032000</td>\n",
       "      <td>0.060000</td>\n",
       "      <td>0.012000</td>\n",
       "      <td>0.072000</td>\n",
       "      <td>0.028000</td>\n",
       "      <td>0.064000</td>\n",
       "      <td>0.104000</td>\n",
       "      <td>0.004000</td>\n",
       "      <td>0.008000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10112</th>\n",
       "      <td>0.096386</td>\n",
       "      <td>0.004016</td>\n",
       "      <td>0.064257</td>\n",
       "      <td>0.068273</td>\n",
       "      <td>0.024096</td>\n",
       "      <td>0.084337</td>\n",
       "      <td>0.024096</td>\n",
       "      <td>0.028112</td>\n",
       "      <td>0.024096</td>\n",
       "      <td>0.156627</td>\n",
       "      <td>0.016064</td>\n",
       "      <td>0.028112</td>\n",
       "      <td>0.048193</td>\n",
       "      <td>0.048193</td>\n",
       "      <td>0.084337</td>\n",
       "      <td>0.040161</td>\n",
       "      <td>0.028112</td>\n",
       "      <td>0.120482</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10144</th>\n",
       "      <td>0.080702</td>\n",
       "      <td>0.003509</td>\n",
       "      <td>0.031579</td>\n",
       "      <td>0.024561</td>\n",
       "      <td>0.098246</td>\n",
       "      <td>0.063158</td>\n",
       "      <td>0.007018</td>\n",
       "      <td>0.126316</td>\n",
       "      <td>0.031579</td>\n",
       "      <td>0.122807</td>\n",
       "      <td>0.028070</td>\n",
       "      <td>0.049123</td>\n",
       "      <td>0.035088</td>\n",
       "      <td>0.035088</td>\n",
       "      <td>0.031579</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.045614</td>\n",
       "      <td>0.063158</td>\n",
       "      <td>0.024561</td>\n",
       "      <td>0.045614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10144</th>\n",
       "      <td>0.142259</td>\n",
       "      <td>0.004184</td>\n",
       "      <td>0.041841</td>\n",
       "      <td>0.079498</td>\n",
       "      <td>0.016736</td>\n",
       "      <td>0.083682</td>\n",
       "      <td>0.008368</td>\n",
       "      <td>0.075314</td>\n",
       "      <td>0.050209</td>\n",
       "      <td>0.096234</td>\n",
       "      <td>0.033473</td>\n",
       "      <td>0.037657</td>\n",
       "      <td>0.016736</td>\n",
       "      <td>0.025105</td>\n",
       "      <td>0.041841</td>\n",
       "      <td>0.075314</td>\n",
       "      <td>0.075314</td>\n",
       "      <td>0.071130</td>\n",
       "      <td>0.008368</td>\n",
       "      <td>0.016736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10144</th>\n",
       "      <td>0.086420</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021605</td>\n",
       "      <td>0.024691</td>\n",
       "      <td>0.058642</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.006173</td>\n",
       "      <td>0.095679</td>\n",
       "      <td>0.049383</td>\n",
       "      <td>0.145062</td>\n",
       "      <td>0.030864</td>\n",
       "      <td>0.058642</td>\n",
       "      <td>0.043210</td>\n",
       "      <td>0.018519</td>\n",
       "      <td>0.033951</td>\n",
       "      <td>0.077160</td>\n",
       "      <td>0.043210</td>\n",
       "      <td>0.086420</td>\n",
       "      <td>0.015432</td>\n",
       "      <td>0.049383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9754</th>\n",
       "      <td>0.143312</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.066879</td>\n",
       "      <td>0.054140</td>\n",
       "      <td>0.022293</td>\n",
       "      <td>0.105096</td>\n",
       "      <td>0.025478</td>\n",
       "      <td>0.025478</td>\n",
       "      <td>0.015924</td>\n",
       "      <td>0.092357</td>\n",
       "      <td>0.012739</td>\n",
       "      <td>0.015924</td>\n",
       "      <td>0.076433</td>\n",
       "      <td>0.022293</td>\n",
       "      <td>0.063694</td>\n",
       "      <td>0.031847</td>\n",
       "      <td>0.057325</td>\n",
       "      <td>0.143312</td>\n",
       "      <td>0.012739</td>\n",
       "      <td>0.012739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9757</th>\n",
       "      <td>0.205993</td>\n",
       "      <td>0.003745</td>\n",
       "      <td>0.052434</td>\n",
       "      <td>0.052434</td>\n",
       "      <td>0.003745</td>\n",
       "      <td>0.089888</td>\n",
       "      <td>0.014981</td>\n",
       "      <td>0.026217</td>\n",
       "      <td>0.007491</td>\n",
       "      <td>0.086142</td>\n",
       "      <td>0.018727</td>\n",
       "      <td>0.026217</td>\n",
       "      <td>0.044944</td>\n",
       "      <td>0.044944</td>\n",
       "      <td>0.071161</td>\n",
       "      <td>0.059925</td>\n",
       "      <td>0.063670</td>\n",
       "      <td>0.104869</td>\n",
       "      <td>0.014981</td>\n",
       "      <td>0.007491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9778</th>\n",
       "      <td>0.116883</td>\n",
       "      <td>0.006494</td>\n",
       "      <td>0.081169</td>\n",
       "      <td>0.064935</td>\n",
       "      <td>0.022727</td>\n",
       "      <td>0.087662</td>\n",
       "      <td>0.022727</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.025974</td>\n",
       "      <td>0.136364</td>\n",
       "      <td>0.019481</td>\n",
       "      <td>0.016234</td>\n",
       "      <td>0.045455</td>\n",
       "      <td>0.032468</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>0.022727</td>\n",
       "      <td>0.064935</td>\n",
       "      <td>0.077922</td>\n",
       "      <td>0.003247</td>\n",
       "      <td>0.025974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9931</th>\n",
       "      <td>0.134503</td>\n",
       "      <td>0.005848</td>\n",
       "      <td>0.049708</td>\n",
       "      <td>0.032164</td>\n",
       "      <td>0.017544</td>\n",
       "      <td>0.093567</td>\n",
       "      <td>0.017544</td>\n",
       "      <td>0.020468</td>\n",
       "      <td>0.008772</td>\n",
       "      <td>0.152047</td>\n",
       "      <td>0.035088</td>\n",
       "      <td>0.008772</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.093567</td>\n",
       "      <td>0.073099</td>\n",
       "      <td>0.040936</td>\n",
       "      <td>0.067251</td>\n",
       "      <td>0.008772</td>\n",
       "      <td>0.032164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>0.081897</td>\n",
       "      <td>0.012931</td>\n",
       "      <td>0.051724</td>\n",
       "      <td>0.047414</td>\n",
       "      <td>0.047414</td>\n",
       "      <td>0.064655</td>\n",
       "      <td>0.025862</td>\n",
       "      <td>0.038793</td>\n",
       "      <td>0.038793</td>\n",
       "      <td>0.094828</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.047414</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.051724</td>\n",
       "      <td>0.056034</td>\n",
       "      <td>0.060345</td>\n",
       "      <td>0.077586</td>\n",
       "      <td>0.090517</td>\n",
       "      <td>0.012931</td>\n",
       "      <td>0.030172</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>500 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          AAC_A     AAC_C     AAC_D     AAC_E     AAC_F     AAC_G     AAC_H  \\\n",
       "100    0.164000  0.000000  0.076000  0.044000  0.032000  0.108000  0.020000   \n",
       "10112  0.096386  0.004016  0.064257  0.068273  0.024096  0.084337  0.024096   \n",
       "10144  0.080702  0.003509  0.031579  0.024561  0.098246  0.063158  0.007018   \n",
       "10144  0.142259  0.004184  0.041841  0.079498  0.016736  0.083682  0.008368   \n",
       "10144  0.086420  0.000000  0.021605  0.024691  0.058642  0.055556  0.006173   \n",
       "...         ...       ...       ...       ...       ...       ...       ...   \n",
       "9754   0.143312  0.000000  0.066879  0.054140  0.022293  0.105096  0.025478   \n",
       "9757   0.205993  0.003745  0.052434  0.052434  0.003745  0.089888  0.014981   \n",
       "9778   0.116883  0.006494  0.081169  0.064935  0.022727  0.087662  0.022727   \n",
       "9931   0.134503  0.005848  0.049708  0.032164  0.017544  0.093567  0.017544   \n",
       "998    0.081897  0.012931  0.051724  0.047414  0.047414  0.064655  0.025862   \n",
       "\n",
       "          AAC_I     AAC_K     AAC_L     AAC_M     AAC_N     AAC_P     AAC_Q  \\\n",
       "100    0.032000  0.008000  0.112000  0.020000  0.032000  0.060000  0.012000   \n",
       "10112  0.028112  0.024096  0.156627  0.016064  0.028112  0.048193  0.048193   \n",
       "10144  0.126316  0.031579  0.122807  0.028070  0.049123  0.035088  0.035088   \n",
       "10144  0.075314  0.050209  0.096234  0.033473  0.037657  0.016736  0.025105   \n",
       "10144  0.095679  0.049383  0.145062  0.030864  0.058642  0.043210  0.018519   \n",
       "...         ...       ...       ...       ...       ...       ...       ...   \n",
       "9754   0.025478  0.015924  0.092357  0.012739  0.015924  0.076433  0.022293   \n",
       "9757   0.026217  0.007491  0.086142  0.018727  0.026217  0.044944  0.044944   \n",
       "9778   0.035714  0.025974  0.136364  0.019481  0.016234  0.045455  0.032468   \n",
       "9931   0.020468  0.008772  0.152047  0.035088  0.008772  0.052632  0.055556   \n",
       "998    0.038793  0.038793  0.094828  0.034483  0.047414  0.034483  0.051724   \n",
       "\n",
       "          AAC_R     AAC_S     AAC_T     AAC_V     AAC_W     AAC_Y  \n",
       "100    0.072000  0.028000  0.064000  0.104000  0.004000  0.008000  \n",
       "10112  0.084337  0.040161  0.028112  0.120482  0.000000  0.012048  \n",
       "10144  0.031579  0.052632  0.045614  0.063158  0.024561  0.045614  \n",
       "10144  0.041841  0.075314  0.075314  0.071130  0.008368  0.016736  \n",
       "10144  0.033951  0.077160  0.043210  0.086420  0.015432  0.049383  \n",
       "...         ...       ...       ...       ...       ...       ...  \n",
       "9754   0.063694  0.031847  0.057325  0.143312  0.012739  0.012739  \n",
       "9757   0.071161  0.059925  0.063670  0.104869  0.014981  0.007491  \n",
       "9778   0.090909  0.022727  0.064935  0.077922  0.003247  0.025974  \n",
       "9931   0.093567  0.073099  0.040936  0.067251  0.008772  0.032164  \n",
       "998    0.056034  0.060345  0.077586  0.090517  0.012931  0.030172  \n",
       "\n",
       "[500 rows x 20 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2.sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "4d5577e0",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>thermo_index</th>\n",
       "      <th>...</th>\n",
       "      <th>bit_score_16s</th>\n",
       "      <th>m_ogt</th>\n",
       "      <th>t_ogt</th>\n",
       "      <th>ogt_difference</th>\n",
       "      <th>m_protein_seq</th>\n",
       "      <th>t_protein_seq</th>\n",
       "      <th>m_protein_desc</th>\n",
       "      <th>t_protein_desc</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0.287582</td>\n",
       "      <td>0.217822</td>\n",
       "      <td>0.215686</td>\n",
       "      <td>160</td>\n",
       "      <td>0.792079</td>\n",
       "      <td>152</td>\n",
       "      <td>0.737864</td>\n",
       "      <td>131</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>1153.0</td>\n",
       "      <td>27.5</td>\n",
       "      <td>50.0</td>\n",
       "      <td>22.5</td>\n",
       "      <td>MAESGTSRRADHLVPVPGPDAEPPAVADELLRAVGRGDEQAFGRLY...</td>\n",
       "      <td>MPSQITESERIELAERFERDALPLLDQLYSAALRMTRNPADAEDLV...</td>\n",
       "      <td>ECF RNA polymerase sigma factor SigK</td>\n",
       "      <td>sigma-70 family RNA polymerase sigma factor</td>\n",
       "      <td>206</td>\n",
       "      <td>202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0.319635</td>\n",
       "      <td>0.295359</td>\n",
       "      <td>0.297872</td>\n",
       "      <td>218</td>\n",
       "      <td>0.919831</td>\n",
       "      <td>226</td>\n",
       "      <td>0.969957</td>\n",
       "      <td>282</td>\n",
       "      <td>11324</td>\n",
       "      <td>...</td>\n",
       "      <td>1014.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>MARIALVDDDRNILTSVSMTLEAEGFEVETYNDGQSALDAFNKRMP...</td>\n",
       "      <td>MRVLLVEDDPNTSRSIEMMLTHANLNVYATDMGEEGIDLAKLYDYD...</td>\n",
       "      <td>response regulator transcription factor</td>\n",
       "      <td>response regulator transcription factor</td>\n",
       "      <td>233</td>\n",
       "      <td>237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0.279621</td>\n",
       "      <td>0.234127</td>\n",
       "      <td>0.218924</td>\n",
       "      <td>211</td>\n",
       "      <td>0.837302</td>\n",
       "      <td>210</td>\n",
       "      <td>0.731707</td>\n",
       "      <td>96</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>1138.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MKDTVVFVTGAARGIGAHTARLAVARGARVALVGLEPHLLADLAAE...</td>\n",
       "      <td>MTPEQIFSGQTAIVTGGASGIGAATVEHIARRGGRVFSVDLSYDSP...</td>\n",
       "      <td>SDR family oxidoreductase</td>\n",
       "      <td>SDR family oxidoreductase</td>\n",
       "      <td>287</td>\n",
       "      <td>252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0.327273</td>\n",
       "      <td>0.200743</td>\n",
       "      <td>0.214712</td>\n",
       "      <td>166</td>\n",
       "      <td>0.617100</td>\n",
       "      <td>163</td>\n",
       "      <td>0.696581</td>\n",
       "      <td>175</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>1077.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MTSGLWERVLDGVWVTIQLLVLSALLATAVSFVVGIARTHRLWIVR...</td>\n",
       "      <td>MAMSRRKRGQLARGIQYAILVIVVVVLALLADWGKIGKAFFDWEAA...</td>\n",
       "      <td>ectoine/hydroxyectoine ABC transporter permeas...</td>\n",
       "      <td>amino acid ABC transporter permease</td>\n",
       "      <td>234</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0.338710</td>\n",
       "      <td>0.318182</td>\n",
       "      <td>0.287671</td>\n",
       "      <td>60</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>71</td>\n",
       "      <td>0.887500</td>\n",
       "      <td>61</td>\n",
       "      <td>9827</td>\n",
       "      <td>...</td>\n",
       "      <td>991.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>MIISLRRGLRFIRFIVFFAALVYLFYHVLDLFNGWISPVDQYQMPT...</td>\n",
       "      <td>MKRMVWRTLKVFIIFIACTLLFYFGLRFMHLEYEQFHRYEPPEGPA...</td>\n",
       "      <td>YqzK family protein</td>\n",
       "      <td>YqzK family protein</td>\n",
       "      <td>80</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  local_gap_compressed_percent_id  scaled_local_query_percent_id  \\\n",
       "0           0                         0.287582                       0.217822   \n",
       "1           1                         0.319635                       0.295359   \n",
       "2           2                         0.279621                       0.234127   \n",
       "3           3                         0.327273                       0.200743   \n",
       "4           4                         0.338710                       0.318182   \n",
       "\n",
       "   scaled_local_symmetric_percent_id  query_align_len  query_align_cov  \\\n",
       "0                           0.215686              160         0.792079   \n",
       "1                           0.297872              218         0.919831   \n",
       "2                           0.218924              211         0.837302   \n",
       "3                           0.214712              166         0.617100   \n",
       "4                           0.287671               60         0.909091   \n",
       "\n",
       "   subject_align_len  subject_align_cov  bit_score  thermo_index  ...  \\\n",
       "0                152           0.737864        131           875  ...   \n",
       "1                226           0.969957        282         11324  ...   \n",
       "2                210           0.731707         96           875  ...   \n",
       "3                163           0.696581        175           875  ...   \n",
       "4                 71           0.887500         61          9827  ...   \n",
       "\n",
       "   bit_score_16s  m_ogt  t_ogt  ogt_difference  \\\n",
       "0         1153.0   27.5   50.0            22.5   \n",
       "1         1014.0   25.0   54.0            29.0   \n",
       "2         1138.0   28.0   50.0            22.0   \n",
       "3         1077.0   28.0   50.0            22.0   \n",
       "4          991.0   30.0   50.0            20.0   \n",
       "\n",
       "                                       m_protein_seq  \\\n",
       "0  MAESGTSRRADHLVPVPGPDAEPPAVADELLRAVGRGDEQAFGRLY...   \n",
       "1  MARIALVDDDRNILTSVSMTLEAEGFEVETYNDGQSALDAFNKRMP...   \n",
       "2  MKDTVVFVTGAARGIGAHTARLAVARGARVALVGLEPHLLADLAAE...   \n",
       "3  MTSGLWERVLDGVWVTIQLLVLSALLATAVSFVVGIARTHRLWIVR...   \n",
       "4  MIISLRRGLRFIRFIVFFAALVYLFYHVLDLFNGWISPVDQYQMPT...   \n",
       "\n",
       "                                       t_protein_seq  \\\n",
       "0  MPSQITESERIELAERFERDALPLLDQLYSAALRMTRNPADAEDLV...   \n",
       "1  MRVLLVEDDPNTSRSIEMMLTHANLNVYATDMGEEGIDLAKLYDYD...   \n",
       "2  MTPEQIFSGQTAIVTGGASGIGAATVEHIARRGGRVFSVDLSYDSP...   \n",
       "3  MAMSRRKRGQLARGIQYAILVIVVVVLALLADWGKIGKAFFDWEAA...   \n",
       "4  MKRMVWRTLKVFIIFIACTLLFYFGLRFMHLEYEQFHRYEPPEGPA...   \n",
       "\n",
       "                                      m_protein_desc  \\\n",
       "0               ECF RNA polymerase sigma factor SigK   \n",
       "1            response regulator transcription factor   \n",
       "2                          SDR family oxidoreductase   \n",
       "3  ectoine/hydroxyectoine ABC transporter permeas...   \n",
       "4                                YqzK family protein   \n",
       "\n",
       "                                t_protein_desc  m_protein_len  t_protein_len  \n",
       "0  sigma-70 family RNA polymerase sigma factor            206            202  \n",
       "1      response regulator transcription factor            233            237  \n",
       "2                    SDR family oxidoreductase            287            252  \n",
       "3          amino acid ABC transporter permease            234            269  \n",
       "4                          YqzK family protein             80             66  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "e14bb1d3",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>meso_index</th>\n",
       "      <th>m_protein_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12897</td>\n",
       "      <td>MAESGTSRRADHLVPVPGPDAEPPAVADELLRAVGRGDEQAFGRLY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13026</td>\n",
       "      <td>MARIALVDDDRNILTSVSMTLEAEGFEVETYNDGQSALDAFNKRMP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8203</td>\n",
       "      <td>MKDTVVFVTGAARGIGAHTARLAVARGARVALVGLEPHLLADLAAE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3340</td>\n",
       "      <td>MTSGLWERVLDGVWVTIQLLVLSALLATAVSFVVGIARTHRLWIVR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>14020</td>\n",
       "      <td>MIISLRRGLRFIRFIVFFAALVYLFYHVLDLFNGWISPVDQYQMPT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   meso_index                                      m_protein_seq\n",
       "0       12897  MAESGTSRRADHLVPVPGPDAEPPAVADELLRAVGRGDEQAFGRLY...\n",
       "1       13026  MARIALVDDDRNILTSVSMTLEAEGFEVETYNDGQSALDAFNKRMP...\n",
       "2        8203  MKDTVVFVTGAARGIGAHTARLAVARGARVALVGLEPHLLADLAAE...\n",
       "3        3340  MTSGLWERVLDGVWVTIQLLVLSALLATAVSFVVGIARTHRLWIVR...\n",
       "4       14020  MIISLRRGLRFIRFIVFFAALVYLFYHVLDLFNGWISPVDQYQMPT..."
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_seqs = df[['meso_index', 'm_protein_seq']]\n",
    "df_seqs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "20c76d56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>thermo_index</th>\n",
       "      <th>...</th>\n",
       "      <th>bit_score_16s</th>\n",
       "      <th>m_ogt</th>\n",
       "      <th>t_ogt</th>\n",
       "      <th>ogt_difference</th>\n",
       "      <th>m_protein_seq</th>\n",
       "      <th>t_protein_seq</th>\n",
       "      <th>m_protein_desc</th>\n",
       "      <th>t_protein_desc</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>meso_index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12897</th>\n",
       "      <td>0</td>\n",
       "      <td>0.287582</td>\n",
       "      <td>0.217822</td>\n",
       "      <td>0.215686</td>\n",
       "      <td>160</td>\n",
       "      <td>0.792079</td>\n",
       "      <td>152</td>\n",
       "      <td>0.737864</td>\n",
       "      <td>131</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>1153.0</td>\n",
       "      <td>27.5</td>\n",
       "      <td>50.0</td>\n",
       "      <td>22.5</td>\n",
       "      <td>MAESGTSRRADHLVPVPGPDAEPPAVADELLRAVGRGDEQAFGRLY...</td>\n",
       "      <td>MPSQITESERIELAERFERDALPLLDQLYSAALRMTRNPADAEDLV...</td>\n",
       "      <td>ECF RNA polymerase sigma factor SigK</td>\n",
       "      <td>sigma-70 family RNA polymerase sigma factor</td>\n",
       "      <td>206</td>\n",
       "      <td>202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13026</th>\n",
       "      <td>1</td>\n",
       "      <td>0.319635</td>\n",
       "      <td>0.295359</td>\n",
       "      <td>0.297872</td>\n",
       "      <td>218</td>\n",
       "      <td>0.919831</td>\n",
       "      <td>226</td>\n",
       "      <td>0.969957</td>\n",
       "      <td>282</td>\n",
       "      <td>11324</td>\n",
       "      <td>...</td>\n",
       "      <td>1014.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>MARIALVDDDRNILTSVSMTLEAEGFEVETYNDGQSALDAFNKRMP...</td>\n",
       "      <td>MRVLLVEDDPNTSRSIEMMLTHANLNVYATDMGEEGIDLAKLYDYD...</td>\n",
       "      <td>response regulator transcription factor</td>\n",
       "      <td>response regulator transcription factor</td>\n",
       "      <td>233</td>\n",
       "      <td>237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8203</th>\n",
       "      <td>2</td>\n",
       "      <td>0.279621</td>\n",
       "      <td>0.234127</td>\n",
       "      <td>0.218924</td>\n",
       "      <td>211</td>\n",
       "      <td>0.837302</td>\n",
       "      <td>210</td>\n",
       "      <td>0.731707</td>\n",
       "      <td>96</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>1138.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MKDTVVFVTGAARGIGAHTARLAVARGARVALVGLEPHLLADLAAE...</td>\n",
       "      <td>MTPEQIFSGQTAIVTGGASGIGAATVEHIARRGGRVFSVDLSYDSP...</td>\n",
       "      <td>SDR family oxidoreductase</td>\n",
       "      <td>SDR family oxidoreductase</td>\n",
       "      <td>287</td>\n",
       "      <td>252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3340</th>\n",
       "      <td>3</td>\n",
       "      <td>0.327273</td>\n",
       "      <td>0.200743</td>\n",
       "      <td>0.214712</td>\n",
       "      <td>166</td>\n",
       "      <td>0.617100</td>\n",
       "      <td>163</td>\n",
       "      <td>0.696581</td>\n",
       "      <td>175</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>1077.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MTSGLWERVLDGVWVTIQLLVLSALLATAVSFVVGIARTHRLWIVR...</td>\n",
       "      <td>MAMSRRKRGQLARGIQYAILVIVVVVLALLADWGKIGKAFFDWEAA...</td>\n",
       "      <td>ectoine/hydroxyectoine ABC transporter permeas...</td>\n",
       "      <td>amino acid ABC transporter permease</td>\n",
       "      <td>234</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14020</th>\n",
       "      <td>4</td>\n",
       "      <td>0.338710</td>\n",
       "      <td>0.318182</td>\n",
       "      <td>0.287671</td>\n",
       "      <td>60</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>71</td>\n",
       "      <td>0.887500</td>\n",
       "      <td>61</td>\n",
       "      <td>9827</td>\n",
       "      <td>...</td>\n",
       "      <td>991.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>MIISLRRGLRFIRFIVFFAALVYLFYHVLDLFNGWISPVDQYQMPT...</td>\n",
       "      <td>MKRMVWRTLKVFIIFIACTLLFYFGLRFMHLEYEQFHRYEPPEGPA...</td>\n",
       "      <td>YqzK family protein</td>\n",
       "      <td>YqzK family protein</td>\n",
       "      <td>80</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3582</th>\n",
       "      <td>5</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.267380</td>\n",
       "      <td>0.254453</td>\n",
       "      <td>148</td>\n",
       "      <td>0.791444</td>\n",
       "      <td>163</td>\n",
       "      <td>0.791262</td>\n",
       "      <td>127</td>\n",
       "      <td>8919</td>\n",
       "      <td>...</td>\n",
       "      <td>942.0</td>\n",
       "      <td>17.5</td>\n",
       "      <td>42.0</td>\n",
       "      <td>24.5</td>\n",
       "      <td>MVTTGERQRNARGEGARLRLEIVAATQALLADGETATLRSIARRAG...</td>\n",
       "      <td>MNRPTYHHGDLRAAILTEAARLVAERGAERVSLRELAREAGVSHAA...</td>\n",
       "      <td>TetR/AcrR family transcriptional regulator</td>\n",
       "      <td>TetR/AcrR family transcriptional regulator</td>\n",
       "      <td>206</td>\n",
       "      <td>187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13920</th>\n",
       "      <td>6</td>\n",
       "      <td>0.283582</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.240506</td>\n",
       "      <td>133</td>\n",
       "      <td>0.810976</td>\n",
       "      <td>134</td>\n",
       "      <td>0.881579</td>\n",
       "      <td>99</td>\n",
       "      <td>7134</td>\n",
       "      <td>...</td>\n",
       "      <td>1011.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>52.5</td>\n",
       "      <td>24.5</td>\n",
       "      <td>MGSTDRPDLAAMLAPLTRTLIAMERPVLETYGLTMWAYSVLVALSR...</td>\n",
       "      <td>MGHKDELIARIQRAEREFRRSVVSQAATDFFSVDLTMPQLRVVFFL...</td>\n",
       "      <td>MarR family transcriptional regulator</td>\n",
       "      <td>MarR family transcriptional regulator</td>\n",
       "      <td>152</td>\n",
       "      <td>164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6370</th>\n",
       "      <td>7</td>\n",
       "      <td>0.268482</td>\n",
       "      <td>0.190083</td>\n",
       "      <td>0.189300</td>\n",
       "      <td>248</td>\n",
       "      <td>0.683196</td>\n",
       "      <td>271</td>\n",
       "      <td>0.740437</td>\n",
       "      <td>132</td>\n",
       "      <td>14963</td>\n",
       "      <td>...</td>\n",
       "      <td>1054.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>52.5</td>\n",
       "      <td>25.5</td>\n",
       "      <td>MPRWVKIAVIVAVAAIVGVASVPLSQLGSGTITVTAQFADAAGLYN...</td>\n",
       "      <td>MKVPFRERNPVPIALASFTALAVLMLGAYFIEDLPVIGGGTTYSAR...</td>\n",
       "      <td>MCE family protein</td>\n",
       "      <td>MCE family protein</td>\n",
       "      <td>366</td>\n",
       "      <td>363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9289</th>\n",
       "      <td>8</td>\n",
       "      <td>0.265306</td>\n",
       "      <td>0.201550</td>\n",
       "      <td>0.210243</td>\n",
       "      <td>320</td>\n",
       "      <td>0.826873</td>\n",
       "      <td>294</td>\n",
       "      <td>0.828169</td>\n",
       "      <td>140</td>\n",
       "      <td>1674</td>\n",
       "      <td>...</td>\n",
       "      <td>1011.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>57.5</td>\n",
       "      <td>27.5</td>\n",
       "      <td>MSITTKEKRFGIIERYREFLPVTENTPKLTLHEGDTPLIHAASLSA...</td>\n",
       "      <td>MSGVEIEKGYFGQFGGSFVFPELQEVLDYLAEQFERYKDDPEFKQE...</td>\n",
       "      <td>threonine synthase</td>\n",
       "      <td>tryptophan synthase subunit beta</td>\n",
       "      <td>355</td>\n",
       "      <td>387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>742</th>\n",
       "      <td>9</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.295082</td>\n",
       "      <td>0.291498</td>\n",
       "      <td>218</td>\n",
       "      <td>0.893443</td>\n",
       "      <td>221</td>\n",
       "      <td>0.884000</td>\n",
       "      <td>158</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>1085.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>MQIKDSVAVVTGGASGLGLATTKRLLDAGGSVVVIDLKGEDVVAEL...</td>\n",
       "      <td>MTGTVVITGGSRGIGAATAVLAAERGWQVAVSFRERRDAAEQVVRR...</td>\n",
       "      <td>3-hydroxyacyl-CoA dehydrogenase</td>\n",
       "      <td>SDR family oxidoreductase</td>\n",
       "      <td>250</td>\n",
       "      <td>244</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Unnamed: 0  local_gap_compressed_percent_id  \\\n",
       "meso_index                                                \n",
       "12897                0                         0.287582   \n",
       "13026                1                         0.319635   \n",
       "8203                 2                         0.279621   \n",
       "3340                 3                         0.327273   \n",
       "14020                4                         0.338710   \n",
       "3582                 5                         0.333333   \n",
       "13920                6                         0.283582   \n",
       "6370                 7                         0.268482   \n",
       "9289                 8                         0.265306   \n",
       "742                  9                         0.333333   \n",
       "\n",
       "            scaled_local_query_percent_id  scaled_local_symmetric_percent_id  \\\n",
       "meso_index                                                                     \n",
       "12897                            0.217822                           0.215686   \n",
       "13026                            0.295359                           0.297872   \n",
       "8203                             0.234127                           0.218924   \n",
       "3340                             0.200743                           0.214712   \n",
       "14020                            0.318182                           0.287671   \n",
       "3582                             0.267380                           0.254453   \n",
       "13920                            0.231707                           0.240506   \n",
       "6370                             0.190083                           0.189300   \n",
       "9289                             0.201550                           0.210243   \n",
       "742                              0.295082                           0.291498   \n",
       "\n",
       "            query_align_len  query_align_cov  subject_align_len  \\\n",
       "meso_index                                                        \n",
       "12897                   160         0.792079                152   \n",
       "13026                   218         0.919831                226   \n",
       "8203                    211         0.837302                210   \n",
       "3340                    166         0.617100                163   \n",
       "14020                    60         0.909091                 71   \n",
       "3582                    148         0.791444                163   \n",
       "13920                   133         0.810976                134   \n",
       "6370                    248         0.683196                271   \n",
       "9289                    320         0.826873                294   \n",
       "742                     218         0.893443                221   \n",
       "\n",
       "            subject_align_cov  bit_score  thermo_index  ...  bit_score_16s  \\\n",
       "meso_index                                              ...                  \n",
       "12897                0.737864        131           875  ...         1153.0   \n",
       "13026                0.969957        282         11324  ...         1014.0   \n",
       "8203                 0.731707         96           875  ...         1138.0   \n",
       "3340                 0.696581        175           875  ...         1077.0   \n",
       "14020                0.887500         61          9827  ...          991.0   \n",
       "3582                 0.791262        127          8919  ...          942.0   \n",
       "13920                0.881579         99          7134  ...         1011.0   \n",
       "6370                 0.740437        132         14963  ...         1054.0   \n",
       "9289                 0.828169        140          1674  ...         1011.0   \n",
       "742                  0.884000        158           875  ...         1085.0   \n",
       "\n",
       "            m_ogt  t_ogt  ogt_difference  \\\n",
       "meso_index                                 \n",
       "12897        27.5   50.0            22.5   \n",
       "13026        25.0   54.0            29.0   \n",
       "8203         28.0   50.0            22.0   \n",
       "3340         28.0   50.0            22.0   \n",
       "14020        30.0   50.0            20.0   \n",
       "3582         17.5   42.0            24.5   \n",
       "13920        28.0   52.5            24.5   \n",
       "6370         27.0   52.5            25.5   \n",
       "9289         30.0   57.5            27.5   \n",
       "742          26.0   50.0            24.0   \n",
       "\n",
       "                                                m_protein_seq  \\\n",
       "meso_index                                                      \n",
       "12897       MAESGTSRRADHLVPVPGPDAEPPAVADELLRAVGRGDEQAFGRLY...   \n",
       "13026       MARIALVDDDRNILTSVSMTLEAEGFEVETYNDGQSALDAFNKRMP...   \n",
       "8203        MKDTVVFVTGAARGIGAHTARLAVARGARVALVGLEPHLLADLAAE...   \n",
       "3340        MTSGLWERVLDGVWVTIQLLVLSALLATAVSFVVGIARTHRLWIVR...   \n",
       "14020       MIISLRRGLRFIRFIVFFAALVYLFYHVLDLFNGWISPVDQYQMPT...   \n",
       "3582        MVTTGERQRNARGEGARLRLEIVAATQALLADGETATLRSIARRAG...   \n",
       "13920       MGSTDRPDLAAMLAPLTRTLIAMERPVLETYGLTMWAYSVLVALSR...   \n",
       "6370        MPRWVKIAVIVAVAAIVGVASVPLSQLGSGTITVTAQFADAAGLYN...   \n",
       "9289        MSITTKEKRFGIIERYREFLPVTENTPKLTLHEGDTPLIHAASLSA...   \n",
       "742         MQIKDSVAVVTGGASGLGLATTKRLLDAGGSVVVIDLKGEDVVAEL...   \n",
       "\n",
       "                                                t_protein_seq  \\\n",
       "meso_index                                                      \n",
       "12897       MPSQITESERIELAERFERDALPLLDQLYSAALRMTRNPADAEDLV...   \n",
       "13026       MRVLLVEDDPNTSRSIEMMLTHANLNVYATDMGEEGIDLAKLYDYD...   \n",
       "8203        MTPEQIFSGQTAIVTGGASGIGAATVEHIARRGGRVFSVDLSYDSP...   \n",
       "3340        MAMSRRKRGQLARGIQYAILVIVVVVLALLADWGKIGKAFFDWEAA...   \n",
       "14020       MKRMVWRTLKVFIIFIACTLLFYFGLRFMHLEYEQFHRYEPPEGPA...   \n",
       "3582        MNRPTYHHGDLRAAILTEAARLVAERGAERVSLRELAREAGVSHAA...   \n",
       "13920       MGHKDELIARIQRAEREFRRSVVSQAATDFFSVDLTMPQLRVVFFL...   \n",
       "6370        MKVPFRERNPVPIALASFTALAVLMLGAYFIEDLPVIGGGTTYSAR...   \n",
       "9289        MSGVEIEKGYFGQFGGSFVFPELQEVLDYLAEQFERYKDDPEFKQE...   \n",
       "742         MTGTVVITGGSRGIGAATAVLAAERGWQVAVSFRERRDAAEQVVRR...   \n",
       "\n",
       "                                               m_protein_desc  \\\n",
       "meso_index                                                      \n",
       "12897                    ECF RNA polymerase sigma factor SigK   \n",
       "13026                 response regulator transcription factor   \n",
       "8203                                SDR family oxidoreductase   \n",
       "3340        ectoine/hydroxyectoine ABC transporter permeas...   \n",
       "14020                                     YqzK family protein   \n",
       "3582               TetR/AcrR family transcriptional regulator   \n",
       "13920                   MarR family transcriptional regulator   \n",
       "6370                                       MCE family protein   \n",
       "9289                                       threonine synthase   \n",
       "742                           3-hydroxyacyl-CoA dehydrogenase   \n",
       "\n",
       "                                         t_protein_desc  m_protein_len  \\\n",
       "meso_index                                                               \n",
       "12897       sigma-70 family RNA polymerase sigma factor            206   \n",
       "13026           response regulator transcription factor            233   \n",
       "8203                          SDR family oxidoreductase            287   \n",
       "3340                amino acid ABC transporter permease            234   \n",
       "14020                               YqzK family protein             80   \n",
       "3582         TetR/AcrR family transcriptional regulator            206   \n",
       "13920             MarR family transcriptional regulator            152   \n",
       "6370                                 MCE family protein            366   \n",
       "9289                   tryptophan synthase subunit beta            355   \n",
       "742                           SDR family oxidoreductase            250   \n",
       "\n",
       "            t_protein_len  \n",
       "meso_index                 \n",
       "12897                 202  \n",
       "13026                 237  \n",
       "8203                  252  \n",
       "3340                  269  \n",
       "14020                  66  \n",
       "3582                  187  \n",
       "13920                 164  \n",
       "6370                  363  \n",
       "9289                  387  \n",
       "742                   244  \n",
       "\n",
       "[10 rows x 29 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "small_df = df.set_index('meso_index')[:10]\n",
    "small_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "0f182aa7",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAC_A</th>\n",
       "      <th>AAC_C</th>\n",
       "      <th>AAC_D</th>\n",
       "      <th>AAC_E</th>\n",
       "      <th>AAC_F</th>\n",
       "      <th>AAC_G</th>\n",
       "      <th>AAC_H</th>\n",
       "      <th>AAC_I</th>\n",
       "      <th>AAC_K</th>\n",
       "      <th>AAC_L</th>\n",
       "      <th>AAC_M</th>\n",
       "      <th>AAC_N</th>\n",
       "      <th>AAC_P</th>\n",
       "      <th>AAC_Q</th>\n",
       "      <th>AAC_R</th>\n",
       "      <th>AAC_S</th>\n",
       "      <th>AAC_T</th>\n",
       "      <th>AAC_V</th>\n",
       "      <th>AAC_W</th>\n",
       "      <th>AAC_Y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12897</th>\n",
       "      <td>0.145631</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.072816</td>\n",
       "      <td>0.067961</td>\n",
       "      <td>0.014563</td>\n",
       "      <td>0.067961</td>\n",
       "      <td>0.019417</td>\n",
       "      <td>0.014563</td>\n",
       "      <td>0.014563</td>\n",
       "      <td>0.106796</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.004854</td>\n",
       "      <td>0.053398</td>\n",
       "      <td>0.038835</td>\n",
       "      <td>0.131068</td>\n",
       "      <td>0.033981</td>\n",
       "      <td>0.048544</td>\n",
       "      <td>0.106796</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.029126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13026</th>\n",
       "      <td>0.064378</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.115880</td>\n",
       "      <td>0.068670</td>\n",
       "      <td>0.030043</td>\n",
       "      <td>0.051502</td>\n",
       "      <td>0.008584</td>\n",
       "      <td>0.051502</td>\n",
       "      <td>0.055794</td>\n",
       "      <td>0.107296</td>\n",
       "      <td>0.055794</td>\n",
       "      <td>0.025751</td>\n",
       "      <td>0.025751</td>\n",
       "      <td>0.042918</td>\n",
       "      <td>0.090129</td>\n",
       "      <td>0.042918</td>\n",
       "      <td>0.047210</td>\n",
       "      <td>0.081545</td>\n",
       "      <td>0.004292</td>\n",
       "      <td>0.030043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8203</th>\n",
       "      <td>0.205575</td>\n",
       "      <td>0.006969</td>\n",
       "      <td>0.062718</td>\n",
       "      <td>0.034843</td>\n",
       "      <td>0.031359</td>\n",
       "      <td>0.094077</td>\n",
       "      <td>0.020906</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.010453</td>\n",
       "      <td>0.094077</td>\n",
       "      <td>0.010453</td>\n",
       "      <td>0.020906</td>\n",
       "      <td>0.034843</td>\n",
       "      <td>0.013937</td>\n",
       "      <td>0.104530</td>\n",
       "      <td>0.031359</td>\n",
       "      <td>0.055749</td>\n",
       "      <td>0.114983</td>\n",
       "      <td>0.010453</td>\n",
       "      <td>0.017422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3340</th>\n",
       "      <td>0.119658</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012821</td>\n",
       "      <td>0.042735</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.094017</td>\n",
       "      <td>0.004274</td>\n",
       "      <td>0.059829</td>\n",
       "      <td>0.038462</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.029915</td>\n",
       "      <td>0.008547</td>\n",
       "      <td>0.047009</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>0.051282</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.051282</td>\n",
       "      <td>0.094017</td>\n",
       "      <td>0.029915</td>\n",
       "      <td>0.025641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14020</th>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.025000</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.075000</td>\n",
       "      <td>0.012500</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>0.012500</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.087500</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.025000</td>\n",
       "      <td>0.087500</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.062500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3582</th>\n",
       "      <td>0.179612</td>\n",
       "      <td>0.004854</td>\n",
       "      <td>0.092233</td>\n",
       "      <td>0.043689</td>\n",
       "      <td>0.033981</td>\n",
       "      <td>0.063107</td>\n",
       "      <td>0.029126</td>\n",
       "      <td>0.033981</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.101942</td>\n",
       "      <td>0.024272</td>\n",
       "      <td>0.014563</td>\n",
       "      <td>0.038835</td>\n",
       "      <td>0.024272</td>\n",
       "      <td>0.106796</td>\n",
       "      <td>0.043689</td>\n",
       "      <td>0.048544</td>\n",
       "      <td>0.072816</td>\n",
       "      <td>0.019417</td>\n",
       "      <td>0.014563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13920</th>\n",
       "      <td>0.131579</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.085526</td>\n",
       "      <td>0.065789</td>\n",
       "      <td>0.006579</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.039474</td>\n",
       "      <td>0.006579</td>\n",
       "      <td>0.157895</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.006579</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.105263</td>\n",
       "      <td>0.059211</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.006579</td>\n",
       "      <td>0.013158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6370</th>\n",
       "      <td>0.139344</td>\n",
       "      <td>0.005464</td>\n",
       "      <td>0.081967</td>\n",
       "      <td>0.019126</td>\n",
       "      <td>0.021858</td>\n",
       "      <td>0.103825</td>\n",
       "      <td>0.008197</td>\n",
       "      <td>0.054645</td>\n",
       "      <td>0.032787</td>\n",
       "      <td>0.103825</td>\n",
       "      <td>0.024590</td>\n",
       "      <td>0.043716</td>\n",
       "      <td>0.027322</td>\n",
       "      <td>0.046448</td>\n",
       "      <td>0.060109</td>\n",
       "      <td>0.054645</td>\n",
       "      <td>0.071038</td>\n",
       "      <td>0.084699</td>\n",
       "      <td>0.002732</td>\n",
       "      <td>0.013661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9289</th>\n",
       "      <td>0.132394</td>\n",
       "      <td>0.014085</td>\n",
       "      <td>0.042254</td>\n",
       "      <td>0.070423</td>\n",
       "      <td>0.025352</td>\n",
       "      <td>0.087324</td>\n",
       "      <td>0.005634</td>\n",
       "      <td>0.084507</td>\n",
       "      <td>0.059155</td>\n",
       "      <td>0.078873</td>\n",
       "      <td>0.019718</td>\n",
       "      <td>0.045070</td>\n",
       "      <td>0.053521</td>\n",
       "      <td>0.030986</td>\n",
       "      <td>0.033803</td>\n",
       "      <td>0.053521</td>\n",
       "      <td>0.064789</td>\n",
       "      <td>0.064789</td>\n",
       "      <td>0.002817</td>\n",
       "      <td>0.030986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>742</th>\n",
       "      <td>0.136000</td>\n",
       "      <td>0.008000</td>\n",
       "      <td>0.056000</td>\n",
       "      <td>0.056000</td>\n",
       "      <td>0.024000</td>\n",
       "      <td>0.116000</td>\n",
       "      <td>0.008000</td>\n",
       "      <td>0.072000</td>\n",
       "      <td>0.044000</td>\n",
       "      <td>0.092000</td>\n",
       "      <td>0.020000</td>\n",
       "      <td>0.032000</td>\n",
       "      <td>0.048000</td>\n",
       "      <td>0.028000</td>\n",
       "      <td>0.052000</td>\n",
       "      <td>0.048000</td>\n",
       "      <td>0.048000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          AAC_A     AAC_C     AAC_D     AAC_E     AAC_F     AAC_G     AAC_H  \\\n",
       "12897  0.145631  0.009709  0.072816  0.067961  0.014563  0.067961  0.019417   \n",
       "13026  0.064378  0.000000  0.115880  0.068670  0.030043  0.051502  0.008584   \n",
       "8203   0.205575  0.006969  0.062718  0.034843  0.031359  0.094077  0.020906   \n",
       "3340   0.119658  0.000000  0.012821  0.042735  0.055556  0.094017  0.004274   \n",
       "14020  0.050000  0.000000  0.025000  0.050000  0.100000  0.075000  0.012500   \n",
       "3582   0.179612  0.004854  0.092233  0.043689  0.033981  0.063107  0.029126   \n",
       "13920  0.131579  0.000000  0.085526  0.065789  0.006579  0.052632  0.026316   \n",
       "6370   0.139344  0.005464  0.081967  0.019126  0.021858  0.103825  0.008197   \n",
       "9289   0.132394  0.014085  0.042254  0.070423  0.025352  0.087324  0.005634   \n",
       "742    0.136000  0.008000  0.056000  0.056000  0.024000  0.116000  0.008000   \n",
       "\n",
       "          AAC_I     AAC_K     AAC_L     AAC_M     AAC_N     AAC_P     AAC_Q  \\\n",
       "12897  0.014563  0.014563  0.106796  0.009709  0.004854  0.053398  0.038835   \n",
       "13026  0.051502  0.055794  0.107296  0.055794  0.025751  0.025751  0.042918   \n",
       "8203   0.024390  0.010453  0.094077  0.010453  0.020906  0.034843  0.013937   \n",
       "3340   0.059829  0.038462  0.153846  0.029915  0.008547  0.047009  0.025641   \n",
       "14020  0.062500  0.012500  0.100000  0.050000  0.037500  0.050000  0.037500   \n",
       "3582   0.033981  0.009709  0.101942  0.024272  0.014563  0.038835  0.024272   \n",
       "13920  0.039474  0.006579  0.157895  0.026316  0.006579  0.052632  0.052632   \n",
       "6370   0.054645  0.032787  0.103825  0.024590  0.043716  0.027322  0.046448   \n",
       "9289   0.084507  0.059155  0.078873  0.019718  0.045070  0.053521  0.030986   \n",
       "742    0.072000  0.044000  0.092000  0.020000  0.032000  0.048000  0.028000   \n",
       "\n",
       "          AAC_R     AAC_S     AAC_T     AAC_V     AAC_W     AAC_Y  \n",
       "12897  0.131068  0.033981  0.048544  0.106796  0.009709  0.029126  \n",
       "13026  0.090129  0.042918  0.047210  0.081545  0.004292  0.030043  \n",
       "8203   0.104530  0.031359  0.055749  0.114983  0.010453  0.017422  \n",
       "3340   0.051282  0.055556  0.051282  0.094017  0.029915  0.025641  \n",
       "14020  0.087500  0.037500  0.025000  0.087500  0.037500  0.062500  \n",
       "3582   0.106796  0.043689  0.048544  0.072816  0.019417  0.014563  \n",
       "13920  0.105263  0.059211  0.052632  0.052632  0.006579  0.013158  \n",
       "6370   0.060109  0.054645  0.071038  0.084699  0.002732  0.013661  \n",
       "9289   0.033803  0.053521  0.064789  0.064789  0.002817  0.030986  \n",
       "742    0.052000  0.048000  0.048000  0.100000  0.000000  0.012000  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "small_50k = learn2therm_50k[:10]\n",
    "small_50k"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "191fd583",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>thermo_index</th>\n",
       "      <th>...</th>\n",
       "      <th>AAC_M</th>\n",
       "      <th>AAC_N</th>\n",
       "      <th>AAC_P</th>\n",
       "      <th>AAC_Q</th>\n",
       "      <th>AAC_R</th>\n",
       "      <th>AAC_S</th>\n",
       "      <th>AAC_T</th>\n",
       "      <th>AAC_V</th>\n",
       "      <th>AAC_W</th>\n",
       "      <th>AAC_Y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>742</th>\n",
       "      <td>9.0</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.295082</td>\n",
       "      <td>0.291498</td>\n",
       "      <td>218.0</td>\n",
       "      <td>0.893443</td>\n",
       "      <td>221.0</td>\n",
       "      <td>0.884000</td>\n",
       "      <td>158.0</td>\n",
       "      <td>875.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3340</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.327273</td>\n",
       "      <td>0.200743</td>\n",
       "      <td>0.214712</td>\n",
       "      <td>166.0</td>\n",
       "      <td>0.617100</td>\n",
       "      <td>163.0</td>\n",
       "      <td>0.696581</td>\n",
       "      <td>175.0</td>\n",
       "      <td>875.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3582</th>\n",
       "      <td>5.0</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.267380</td>\n",
       "      <td>0.254453</td>\n",
       "      <td>148.0</td>\n",
       "      <td>0.791444</td>\n",
       "      <td>163.0</td>\n",
       "      <td>0.791262</td>\n",
       "      <td>127.0</td>\n",
       "      <td>8919.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6370</th>\n",
       "      <td>7.0</td>\n",
       "      <td>0.268482</td>\n",
       "      <td>0.190083</td>\n",
       "      <td>0.189300</td>\n",
       "      <td>248.0</td>\n",
       "      <td>0.683196</td>\n",
       "      <td>271.0</td>\n",
       "      <td>0.740437</td>\n",
       "      <td>132.0</td>\n",
       "      <td>14963.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8203</th>\n",
       "      <td>2.0</td>\n",
       "      <td>0.279621</td>\n",
       "      <td>0.234127</td>\n",
       "      <td>0.218924</td>\n",
       "      <td>211.0</td>\n",
       "      <td>0.837302</td>\n",
       "      <td>210.0</td>\n",
       "      <td>0.731707</td>\n",
       "      <td>96.0</td>\n",
       "      <td>875.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9289</th>\n",
       "      <td>8.0</td>\n",
       "      <td>0.265306</td>\n",
       "      <td>0.201550</td>\n",
       "      <td>0.210243</td>\n",
       "      <td>320.0</td>\n",
       "      <td>0.826873</td>\n",
       "      <td>294.0</td>\n",
       "      <td>0.828169</td>\n",
       "      <td>140.0</td>\n",
       "      <td>1674.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12897</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.287582</td>\n",
       "      <td>0.217822</td>\n",
       "      <td>0.215686</td>\n",
       "      <td>160.0</td>\n",
       "      <td>0.792079</td>\n",
       "      <td>152.0</td>\n",
       "      <td>0.737864</td>\n",
       "      <td>131.0</td>\n",
       "      <td>875.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13026</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.319635</td>\n",
       "      <td>0.295359</td>\n",
       "      <td>0.297872</td>\n",
       "      <td>218.0</td>\n",
       "      <td>0.919831</td>\n",
       "      <td>226.0</td>\n",
       "      <td>0.969957</td>\n",
       "      <td>282.0</td>\n",
       "      <td>11324.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13920</th>\n",
       "      <td>6.0</td>\n",
       "      <td>0.283582</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.240506</td>\n",
       "      <td>133.0</td>\n",
       "      <td>0.810976</td>\n",
       "      <td>134.0</td>\n",
       "      <td>0.881579</td>\n",
       "      <td>99.0</td>\n",
       "      <td>7134.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14020</th>\n",
       "      <td>4.0</td>\n",
       "      <td>0.338710</td>\n",
       "      <td>0.318182</td>\n",
       "      <td>0.287671</td>\n",
       "      <td>60.0</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>71.0</td>\n",
       "      <td>0.887500</td>\n",
       "      <td>61.0</td>\n",
       "      <td>9827.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12897</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.004854</td>\n",
       "      <td>0.053398</td>\n",
       "      <td>0.038835</td>\n",
       "      <td>0.131068</td>\n",
       "      <td>0.033981</td>\n",
       "      <td>0.048544</td>\n",
       "      <td>0.106796</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.029126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13026</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.055794</td>\n",
       "      <td>0.025751</td>\n",
       "      <td>0.025751</td>\n",
       "      <td>0.042918</td>\n",
       "      <td>0.090129</td>\n",
       "      <td>0.042918</td>\n",
       "      <td>0.047210</td>\n",
       "      <td>0.081545</td>\n",
       "      <td>0.004292</td>\n",
       "      <td>0.030043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13920</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.006579</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.105263</td>\n",
       "      <td>0.059211</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.006579</td>\n",
       "      <td>0.013158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14020</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.087500</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.025000</td>\n",
       "      <td>0.087500</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.062500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3340</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.029915</td>\n",
       "      <td>0.008547</td>\n",
       "      <td>0.047009</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>0.051282</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.051282</td>\n",
       "      <td>0.094017</td>\n",
       "      <td>0.029915</td>\n",
       "      <td>0.025641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3582</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.024272</td>\n",
       "      <td>0.014563</td>\n",
       "      <td>0.038835</td>\n",
       "      <td>0.024272</td>\n",
       "      <td>0.106796</td>\n",
       "      <td>0.043689</td>\n",
       "      <td>0.048544</td>\n",
       "      <td>0.072816</td>\n",
       "      <td>0.019417</td>\n",
       "      <td>0.014563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6370</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.024590</td>\n",
       "      <td>0.043716</td>\n",
       "      <td>0.027322</td>\n",
       "      <td>0.046448</td>\n",
       "      <td>0.060109</td>\n",
       "      <td>0.054645</td>\n",
       "      <td>0.071038</td>\n",
       "      <td>0.084699</td>\n",
       "      <td>0.002732</td>\n",
       "      <td>0.013661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>742</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.020000</td>\n",
       "      <td>0.032000</td>\n",
       "      <td>0.048000</td>\n",
       "      <td>0.028000</td>\n",
       "      <td>0.052000</td>\n",
       "      <td>0.048000</td>\n",
       "      <td>0.048000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8203</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010453</td>\n",
       "      <td>0.020906</td>\n",
       "      <td>0.034843</td>\n",
       "      <td>0.013937</td>\n",
       "      <td>0.104530</td>\n",
       "      <td>0.031359</td>\n",
       "      <td>0.055749</td>\n",
       "      <td>0.114983</td>\n",
       "      <td>0.010453</td>\n",
       "      <td>0.017422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9289</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.019718</td>\n",
       "      <td>0.045070</td>\n",
       "      <td>0.053521</td>\n",
       "      <td>0.030986</td>\n",
       "      <td>0.033803</td>\n",
       "      <td>0.053521</td>\n",
       "      <td>0.064789</td>\n",
       "      <td>0.064789</td>\n",
       "      <td>0.002817</td>\n",
       "      <td>0.030986</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows × 49 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0  local_gap_compressed_percent_id  \\\n",
       "742           9.0                         0.333333   \n",
       "3340          3.0                         0.327273   \n",
       "3582          5.0                         0.333333   \n",
       "6370          7.0                         0.268482   \n",
       "8203          2.0                         0.279621   \n",
       "9289          8.0                         0.265306   \n",
       "12897         0.0                         0.287582   \n",
       "13026         1.0                         0.319635   \n",
       "13920         6.0                         0.283582   \n",
       "14020         4.0                         0.338710   \n",
       "12897         NaN                              NaN   \n",
       "13026         NaN                              NaN   \n",
       "13920         NaN                              NaN   \n",
       "14020         NaN                              NaN   \n",
       "3340          NaN                              NaN   \n",
       "3582          NaN                              NaN   \n",
       "6370          NaN                              NaN   \n",
       "742           NaN                              NaN   \n",
       "8203          NaN                              NaN   \n",
       "9289          NaN                              NaN   \n",
       "\n",
       "       scaled_local_query_percent_id  scaled_local_symmetric_percent_id  \\\n",
       "742                         0.295082                           0.291498   \n",
       "3340                        0.200743                           0.214712   \n",
       "3582                        0.267380                           0.254453   \n",
       "6370                        0.190083                           0.189300   \n",
       "8203                        0.234127                           0.218924   \n",
       "9289                        0.201550                           0.210243   \n",
       "12897                       0.217822                           0.215686   \n",
       "13026                       0.295359                           0.297872   \n",
       "13920                       0.231707                           0.240506   \n",
       "14020                       0.318182                           0.287671   \n",
       "12897                            NaN                                NaN   \n",
       "13026                            NaN                                NaN   \n",
       "13920                            NaN                                NaN   \n",
       "14020                            NaN                                NaN   \n",
       "3340                             NaN                                NaN   \n",
       "3582                             NaN                                NaN   \n",
       "6370                             NaN                                NaN   \n",
       "742                              NaN                                NaN   \n",
       "8203                             NaN                                NaN   \n",
       "9289                             NaN                                NaN   \n",
       "\n",
       "       query_align_len  query_align_cov  subject_align_len  subject_align_cov  \\\n",
       "742              218.0         0.893443              221.0           0.884000   \n",
       "3340             166.0         0.617100              163.0           0.696581   \n",
       "3582             148.0         0.791444              163.0           0.791262   \n",
       "6370             248.0         0.683196              271.0           0.740437   \n",
       "8203             211.0         0.837302              210.0           0.731707   \n",
       "9289             320.0         0.826873              294.0           0.828169   \n",
       "12897            160.0         0.792079              152.0           0.737864   \n",
       "13026            218.0         0.919831              226.0           0.969957   \n",
       "13920            133.0         0.810976              134.0           0.881579   \n",
       "14020             60.0         0.909091               71.0           0.887500   \n",
       "12897              NaN              NaN                NaN                NaN   \n",
       "13026              NaN              NaN                NaN                NaN   \n",
       "13920              NaN              NaN                NaN                NaN   \n",
       "14020              NaN              NaN                NaN                NaN   \n",
       "3340               NaN              NaN                NaN                NaN   \n",
       "3582               NaN              NaN                NaN                NaN   \n",
       "6370               NaN              NaN                NaN                NaN   \n",
       "742                NaN              NaN                NaN                NaN   \n",
       "8203               NaN              NaN                NaN                NaN   \n",
       "9289               NaN              NaN                NaN                NaN   \n",
       "\n",
       "       bit_score  thermo_index  ...     AAC_M     AAC_N     AAC_P     AAC_Q  \\\n",
       "742        158.0         875.0  ...       NaN       NaN       NaN       NaN   \n",
       "3340       175.0         875.0  ...       NaN       NaN       NaN       NaN   \n",
       "3582       127.0        8919.0  ...       NaN       NaN       NaN       NaN   \n",
       "6370       132.0       14963.0  ...       NaN       NaN       NaN       NaN   \n",
       "8203        96.0         875.0  ...       NaN       NaN       NaN       NaN   \n",
       "9289       140.0        1674.0  ...       NaN       NaN       NaN       NaN   \n",
       "12897      131.0         875.0  ...       NaN       NaN       NaN       NaN   \n",
       "13026      282.0       11324.0  ...       NaN       NaN       NaN       NaN   \n",
       "13920       99.0        7134.0  ...       NaN       NaN       NaN       NaN   \n",
       "14020       61.0        9827.0  ...       NaN       NaN       NaN       NaN   \n",
       "12897        NaN           NaN  ...  0.009709  0.004854  0.053398  0.038835   \n",
       "13026        NaN           NaN  ...  0.055794  0.025751  0.025751  0.042918   \n",
       "13920        NaN           NaN  ...  0.026316  0.006579  0.052632  0.052632   \n",
       "14020        NaN           NaN  ...  0.050000  0.037500  0.050000  0.037500   \n",
       "3340         NaN           NaN  ...  0.029915  0.008547  0.047009  0.025641   \n",
       "3582         NaN           NaN  ...  0.024272  0.014563  0.038835  0.024272   \n",
       "6370         NaN           NaN  ...  0.024590  0.043716  0.027322  0.046448   \n",
       "742          NaN           NaN  ...  0.020000  0.032000  0.048000  0.028000   \n",
       "8203         NaN           NaN  ...  0.010453  0.020906  0.034843  0.013937   \n",
       "9289         NaN           NaN  ...  0.019718  0.045070  0.053521  0.030986   \n",
       "\n",
       "          AAC_R     AAC_S     AAC_T     AAC_V     AAC_W     AAC_Y  \n",
       "742         NaN       NaN       NaN       NaN       NaN       NaN  \n",
       "3340        NaN       NaN       NaN       NaN       NaN       NaN  \n",
       "3582        NaN       NaN       NaN       NaN       NaN       NaN  \n",
       "6370        NaN       NaN       NaN       NaN       NaN       NaN  \n",
       "8203        NaN       NaN       NaN       NaN       NaN       NaN  \n",
       "9289        NaN       NaN       NaN       NaN       NaN       NaN  \n",
       "12897       NaN       NaN       NaN       NaN       NaN       NaN  \n",
       "13026       NaN       NaN       NaN       NaN       NaN       NaN  \n",
       "13920       NaN       NaN       NaN       NaN       NaN       NaN  \n",
       "14020       NaN       NaN       NaN       NaN       NaN       NaN  \n",
       "12897  0.131068  0.033981  0.048544  0.106796  0.009709  0.029126  \n",
       "13026  0.090129  0.042918  0.047210  0.081545  0.004292  0.030043  \n",
       "13920  0.105263  0.059211  0.052632  0.052632  0.006579  0.013158  \n",
       "14020  0.087500  0.037500  0.025000  0.087500  0.037500  0.062500  \n",
       "3340   0.051282  0.055556  0.051282  0.094017  0.029915  0.025641  \n",
       "3582   0.106796  0.043689  0.048544  0.072816  0.019417  0.014563  \n",
       "6370   0.060109  0.054645  0.071038  0.084699  0.002732  0.013661  \n",
       "742    0.052000  0.048000  0.048000  0.100000  0.000000  0.012000  \n",
       "8203   0.104530  0.031359  0.055749  0.114983  0.010453  0.017422  \n",
       "9289   0.033803  0.053521  0.064789  0.064789  0.002817  0.030986  \n",
       "\n",
       "[20 rows x 49 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df3= pd.merge(small_df, small_50k.sort_index(), how='outer', left_index=True, right_index=True)\n",
    "df3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "0f86b3a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index([    742,    3340,    3582,    6370,    8203,    9289,   12897,   13026,\n",
      "         13920,   14020, '12897', '13026', '13920', '14020',  '3340',  '3582',\n",
      "        '6370',   '742',  '8203',  '9289'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(df3.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "a3ef17af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write the dataframe to a fasta file\n",
    "#adjust this to write function with BioPython\n",
    "\n",
    "with open('meso_50k.fasta', 'w') as f:\n",
    "    for i, row in df.iterrows():\n",
    "        f.write('>{}\\n{}\\n'.format(row['meso_index'], row['m_protein_seq']))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "1889b505",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File imported successfully.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAC_A</th>\n",
       "      <th>AAC_C</th>\n",
       "      <th>AAC_D</th>\n",
       "      <th>AAC_E</th>\n",
       "      <th>AAC_F</th>\n",
       "      <th>AAC_G</th>\n",
       "      <th>AAC_H</th>\n",
       "      <th>AAC_I</th>\n",
       "      <th>AAC_K</th>\n",
       "      <th>AAC_L</th>\n",
       "      <th>AAC_M</th>\n",
       "      <th>AAC_N</th>\n",
       "      <th>AAC_P</th>\n",
       "      <th>AAC_Q</th>\n",
       "      <th>AAC_R</th>\n",
       "      <th>AAC_S</th>\n",
       "      <th>AAC_T</th>\n",
       "      <th>AAC_V</th>\n",
       "      <th>AAC_W</th>\n",
       "      <th>AAC_Y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12897</th>\n",
       "      <td>0.145631</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.072816</td>\n",
       "      <td>0.067961</td>\n",
       "      <td>0.014563</td>\n",
       "      <td>0.067961</td>\n",
       "      <td>0.019417</td>\n",
       "      <td>0.014563</td>\n",
       "      <td>0.014563</td>\n",
       "      <td>0.106796</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.004854</td>\n",
       "      <td>0.053398</td>\n",
       "      <td>0.038835</td>\n",
       "      <td>0.131068</td>\n",
       "      <td>0.033981</td>\n",
       "      <td>0.048544</td>\n",
       "      <td>0.106796</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.029126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13026</th>\n",
       "      <td>0.064378</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.115880</td>\n",
       "      <td>0.068670</td>\n",
       "      <td>0.030043</td>\n",
       "      <td>0.051502</td>\n",
       "      <td>0.008584</td>\n",
       "      <td>0.051502</td>\n",
       "      <td>0.055794</td>\n",
       "      <td>0.107296</td>\n",
       "      <td>0.055794</td>\n",
       "      <td>0.025751</td>\n",
       "      <td>0.025751</td>\n",
       "      <td>0.042918</td>\n",
       "      <td>0.090129</td>\n",
       "      <td>0.042918</td>\n",
       "      <td>0.047210</td>\n",
       "      <td>0.081545</td>\n",
       "      <td>0.004292</td>\n",
       "      <td>0.030043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8203</th>\n",
       "      <td>0.205575</td>\n",
       "      <td>0.006969</td>\n",
       "      <td>0.062718</td>\n",
       "      <td>0.034843</td>\n",
       "      <td>0.031359</td>\n",
       "      <td>0.094077</td>\n",
       "      <td>0.020906</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.010453</td>\n",
       "      <td>0.094077</td>\n",
       "      <td>0.010453</td>\n",
       "      <td>0.020906</td>\n",
       "      <td>0.034843</td>\n",
       "      <td>0.013937</td>\n",
       "      <td>0.104530</td>\n",
       "      <td>0.031359</td>\n",
       "      <td>0.055749</td>\n",
       "      <td>0.114983</td>\n",
       "      <td>0.010453</td>\n",
       "      <td>0.017422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3340</th>\n",
       "      <td>0.119658</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012821</td>\n",
       "      <td>0.042735</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.094017</td>\n",
       "      <td>0.004274</td>\n",
       "      <td>0.059829</td>\n",
       "      <td>0.038462</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.029915</td>\n",
       "      <td>0.008547</td>\n",
       "      <td>0.047009</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>0.051282</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.051282</td>\n",
       "      <td>0.094017</td>\n",
       "      <td>0.029915</td>\n",
       "      <td>0.025641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14020</th>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.025000</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.075000</td>\n",
       "      <td>0.012500</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>0.012500</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.087500</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.025000</td>\n",
       "      <td>0.087500</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.062500</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          AAC_A     AAC_C     AAC_D     AAC_E     AAC_F     AAC_G     AAC_H  \\\n",
       "12897  0.145631  0.009709  0.072816  0.067961  0.014563  0.067961  0.019417   \n",
       "13026  0.064378  0.000000  0.115880  0.068670  0.030043  0.051502  0.008584   \n",
       "8203   0.205575  0.006969  0.062718  0.034843  0.031359  0.094077  0.020906   \n",
       "3340   0.119658  0.000000  0.012821  0.042735  0.055556  0.094017  0.004274   \n",
       "14020  0.050000  0.000000  0.025000  0.050000  0.100000  0.075000  0.012500   \n",
       "\n",
       "          AAC_I     AAC_K     AAC_L     AAC_M     AAC_N     AAC_P     AAC_Q  \\\n",
       "12897  0.014563  0.014563  0.106796  0.009709  0.004854  0.053398  0.038835   \n",
       "13026  0.051502  0.055794  0.107296  0.055794  0.025751  0.025751  0.042918   \n",
       "8203   0.024390  0.010453  0.094077  0.010453  0.020906  0.034843  0.013937   \n",
       "3340   0.059829  0.038462  0.153846  0.029915  0.008547  0.047009  0.025641   \n",
       "14020  0.062500  0.012500  0.100000  0.050000  0.037500  0.050000  0.037500   \n",
       "\n",
       "          AAC_R     AAC_S     AAC_T     AAC_V     AAC_W     AAC_Y  \n",
       "12897  0.131068  0.033981  0.048544  0.106796  0.009709  0.029126  \n",
       "13026  0.090129  0.042918  0.047210  0.081545  0.004292  0.030043  \n",
       "8203   0.104530  0.031359  0.055749  0.114983  0.010453  0.017422  \n",
       "3340   0.051282  0.055556  0.051282  0.094017  0.029915  0.025641  \n",
       "14020  0.087500  0.037500  0.025000  0.087500  0.037500  0.062500  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "learn2therm_50k = get_protein_descriptors('meso_50k.fasta', descriptors=['AAC', 'GAAC'])['AAC']\n",
    "learn2therm_50k.head()\n",
    "#why does this not require a .txt file? I think it accepts fasta, just needs descriptor to be gone"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "5c8f4a5a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>thermo_index</th>\n",
       "      <th>...</th>\n",
       "      <th>bit_score_16s</th>\n",
       "      <th>m_ogt</th>\n",
       "      <th>t_ogt</th>\n",
       "      <th>ogt_difference</th>\n",
       "      <th>m_protein_seq</th>\n",
       "      <th>t_protein_seq</th>\n",
       "      <th>m_protein_desc</th>\n",
       "      <th>t_protein_desc</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>meso_index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12897</th>\n",
       "      <td>0</td>\n",
       "      <td>0.287582</td>\n",
       "      <td>0.217822</td>\n",
       "      <td>0.215686</td>\n",
       "      <td>160</td>\n",
       "      <td>0.792079</td>\n",
       "      <td>152</td>\n",
       "      <td>0.737864</td>\n",
       "      <td>131</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>1153.0</td>\n",
       "      <td>27.5</td>\n",
       "      <td>50.0</td>\n",
       "      <td>22.5</td>\n",
       "      <td>MAESGTSRRADHLVPVPGPDAEPPAVADELLRAVGRGDEQAFGRLY...</td>\n",
       "      <td>MPSQITESERIELAERFERDALPLLDQLYSAALRMTRNPADAEDLV...</td>\n",
       "      <td>ECF RNA polymerase sigma factor SigK</td>\n",
       "      <td>sigma-70 family RNA polymerase sigma factor</td>\n",
       "      <td>206</td>\n",
       "      <td>202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13026</th>\n",
       "      <td>1</td>\n",
       "      <td>0.319635</td>\n",
       "      <td>0.295359</td>\n",
       "      <td>0.297872</td>\n",
       "      <td>218</td>\n",
       "      <td>0.919831</td>\n",
       "      <td>226</td>\n",
       "      <td>0.969957</td>\n",
       "      <td>282</td>\n",
       "      <td>11324</td>\n",
       "      <td>...</td>\n",
       "      <td>1014.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>MARIALVDDDRNILTSVSMTLEAEGFEVETYNDGQSALDAFNKRMP...</td>\n",
       "      <td>MRVLLVEDDPNTSRSIEMMLTHANLNVYATDMGEEGIDLAKLYDYD...</td>\n",
       "      <td>response regulator transcription factor</td>\n",
       "      <td>response regulator transcription factor</td>\n",
       "      <td>233</td>\n",
       "      <td>237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8203</th>\n",
       "      <td>2</td>\n",
       "      <td>0.279621</td>\n",
       "      <td>0.234127</td>\n",
       "      <td>0.218924</td>\n",
       "      <td>211</td>\n",
       "      <td>0.837302</td>\n",
       "      <td>210</td>\n",
       "      <td>0.731707</td>\n",
       "      <td>96</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>1138.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MKDTVVFVTGAARGIGAHTARLAVARGARVALVGLEPHLLADLAAE...</td>\n",
       "      <td>MTPEQIFSGQTAIVTGGASGIGAATVEHIARRGGRVFSVDLSYDSP...</td>\n",
       "      <td>SDR family oxidoreductase</td>\n",
       "      <td>SDR family oxidoreductase</td>\n",
       "      <td>287</td>\n",
       "      <td>252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3340</th>\n",
       "      <td>3</td>\n",
       "      <td>0.327273</td>\n",
       "      <td>0.200743</td>\n",
       "      <td>0.214712</td>\n",
       "      <td>166</td>\n",
       "      <td>0.617100</td>\n",
       "      <td>163</td>\n",
       "      <td>0.696581</td>\n",
       "      <td>175</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>1077.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MTSGLWERVLDGVWVTIQLLVLSALLATAVSFVVGIARTHRLWIVR...</td>\n",
       "      <td>MAMSRRKRGQLARGIQYAILVIVVVVLALLADWGKIGKAFFDWEAA...</td>\n",
       "      <td>ectoine/hydroxyectoine ABC transporter permeas...</td>\n",
       "      <td>amino acid ABC transporter permease</td>\n",
       "      <td>234</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14020</th>\n",
       "      <td>4</td>\n",
       "      <td>0.338710</td>\n",
       "      <td>0.318182</td>\n",
       "      <td>0.287671</td>\n",
       "      <td>60</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>71</td>\n",
       "      <td>0.887500</td>\n",
       "      <td>61</td>\n",
       "      <td>9827</td>\n",
       "      <td>...</td>\n",
       "      <td>991.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>MIISLRRGLRFIRFIVFFAALVYLFYHVLDLFNGWISPVDQYQMPT...</td>\n",
       "      <td>MKRMVWRTLKVFIIFIACTLLFYFGLRFMHLEYEQFHRYEPPEGPA...</td>\n",
       "      <td>YqzK family protein</td>\n",
       "      <td>YqzK family protein</td>\n",
       "      <td>80</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Unnamed: 0  local_gap_compressed_percent_id  \\\n",
       "meso_index                                                \n",
       "12897                0                         0.287582   \n",
       "13026                1                         0.319635   \n",
       "8203                 2                         0.279621   \n",
       "3340                 3                         0.327273   \n",
       "14020                4                         0.338710   \n",
       "\n",
       "            scaled_local_query_percent_id  scaled_local_symmetric_percent_id  \\\n",
       "meso_index                                                                     \n",
       "12897                            0.217822                           0.215686   \n",
       "13026                            0.295359                           0.297872   \n",
       "8203                             0.234127                           0.218924   \n",
       "3340                             0.200743                           0.214712   \n",
       "14020                            0.318182                           0.287671   \n",
       "\n",
       "            query_align_len  query_align_cov  subject_align_len  \\\n",
       "meso_index                                                        \n",
       "12897                   160         0.792079                152   \n",
       "13026                   218         0.919831                226   \n",
       "8203                    211         0.837302                210   \n",
       "3340                    166         0.617100                163   \n",
       "14020                    60         0.909091                 71   \n",
       "\n",
       "            subject_align_cov  bit_score  thermo_index  ...  bit_score_16s  \\\n",
       "meso_index                                              ...                  \n",
       "12897                0.737864        131           875  ...         1153.0   \n",
       "13026                0.969957        282         11324  ...         1014.0   \n",
       "8203                 0.731707         96           875  ...         1138.0   \n",
       "3340                 0.696581        175           875  ...         1077.0   \n",
       "14020                0.887500         61          9827  ...          991.0   \n",
       "\n",
       "            m_ogt  t_ogt  ogt_difference  \\\n",
       "meso_index                                 \n",
       "12897        27.5   50.0            22.5   \n",
       "13026        25.0   54.0            29.0   \n",
       "8203         28.0   50.0            22.0   \n",
       "3340         28.0   50.0            22.0   \n",
       "14020        30.0   50.0            20.0   \n",
       "\n",
       "                                                m_protein_seq  \\\n",
       "meso_index                                                      \n",
       "12897       MAESGTSRRADHLVPVPGPDAEPPAVADELLRAVGRGDEQAFGRLY...   \n",
       "13026       MARIALVDDDRNILTSVSMTLEAEGFEVETYNDGQSALDAFNKRMP...   \n",
       "8203        MKDTVVFVTGAARGIGAHTARLAVARGARVALVGLEPHLLADLAAE...   \n",
       "3340        MTSGLWERVLDGVWVTIQLLVLSALLATAVSFVVGIARTHRLWIVR...   \n",
       "14020       MIISLRRGLRFIRFIVFFAALVYLFYHVLDLFNGWISPVDQYQMPT...   \n",
       "\n",
       "                                                t_protein_seq  \\\n",
       "meso_index                                                      \n",
       "12897       MPSQITESERIELAERFERDALPLLDQLYSAALRMTRNPADAEDLV...   \n",
       "13026       MRVLLVEDDPNTSRSIEMMLTHANLNVYATDMGEEGIDLAKLYDYD...   \n",
       "8203        MTPEQIFSGQTAIVTGGASGIGAATVEHIARRGGRVFSVDLSYDSP...   \n",
       "3340        MAMSRRKRGQLARGIQYAILVIVVVVLALLADWGKIGKAFFDWEAA...   \n",
       "14020       MKRMVWRTLKVFIIFIACTLLFYFGLRFMHLEYEQFHRYEPPEGPA...   \n",
       "\n",
       "                                               m_protein_desc  \\\n",
       "meso_index                                                      \n",
       "12897                    ECF RNA polymerase sigma factor SigK   \n",
       "13026                 response regulator transcription factor   \n",
       "8203                                SDR family oxidoreductase   \n",
       "3340        ectoine/hydroxyectoine ABC transporter permeas...   \n",
       "14020                                     YqzK family protein   \n",
       "\n",
       "                                         t_protein_desc  m_protein_len  \\\n",
       "meso_index                                                               \n",
       "12897       sigma-70 family RNA polymerase sigma factor            206   \n",
       "13026           response regulator transcription factor            233   \n",
       "8203                          SDR family oxidoreductase            287   \n",
       "3340                amino acid ABC transporter permease            234   \n",
       "14020                               YqzK family protein             80   \n",
       "\n",
       "            t_protein_len  \n",
       "meso_index                 \n",
       "12897                 202  \n",
       "13026                 237  \n",
       "8203                  252  \n",
       "3340                  269  \n",
       "14020                  66  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.set_index('meso_index').head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "a8ef1d59",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>thermo_index</th>\n",
       "      <th>...</th>\n",
       "      <th>AAC_M</th>\n",
       "      <th>AAC_N</th>\n",
       "      <th>AAC_P</th>\n",
       "      <th>AAC_Q</th>\n",
       "      <th>AAC_R</th>\n",
       "      <th>AAC_S</th>\n",
       "      <th>AAC_T</th>\n",
       "      <th>AAC_V</th>\n",
       "      <th>AAC_W</th>\n",
       "      <th>AAC_Y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12897</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.287582</td>\n",
       "      <td>0.217822</td>\n",
       "      <td>0.215686</td>\n",
       "      <td>160.0</td>\n",
       "      <td>0.792079</td>\n",
       "      <td>152.0</td>\n",
       "      <td>0.737864</td>\n",
       "      <td>131.0</td>\n",
       "      <td>875.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13026</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.319635</td>\n",
       "      <td>0.295359</td>\n",
       "      <td>0.297872</td>\n",
       "      <td>218.0</td>\n",
       "      <td>0.919831</td>\n",
       "      <td>226.0</td>\n",
       "      <td>0.969957</td>\n",
       "      <td>282.0</td>\n",
       "      <td>11324.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8203</th>\n",
       "      <td>2.0</td>\n",
       "      <td>0.279621</td>\n",
       "      <td>0.234127</td>\n",
       "      <td>0.218924</td>\n",
       "      <td>211.0</td>\n",
       "      <td>0.837302</td>\n",
       "      <td>210.0</td>\n",
       "      <td>0.731707</td>\n",
       "      <td>96.0</td>\n",
       "      <td>875.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3340</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.327273</td>\n",
       "      <td>0.200743</td>\n",
       "      <td>0.214712</td>\n",
       "      <td>166.0</td>\n",
       "      <td>0.617100</td>\n",
       "      <td>163.0</td>\n",
       "      <td>0.696581</td>\n",
       "      <td>175.0</td>\n",
       "      <td>875.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14020</th>\n",
       "      <td>4.0</td>\n",
       "      <td>0.338710</td>\n",
       "      <td>0.318182</td>\n",
       "      <td>0.287671</td>\n",
       "      <td>60.0</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>71.0</td>\n",
       "      <td>0.887500</td>\n",
       "      <td>61.0</td>\n",
       "      <td>9827.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8772</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.011070</td>\n",
       "      <td>0.014760</td>\n",
       "      <td>0.081181</td>\n",
       "      <td>0.022140</td>\n",
       "      <td>0.073801</td>\n",
       "      <td>0.036900</td>\n",
       "      <td>0.055351</td>\n",
       "      <td>0.081181</td>\n",
       "      <td>0.036900</td>\n",
       "      <td>0.007380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1395</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010000</td>\n",
       "      <td>0.005000</td>\n",
       "      <td>0.025000</td>\n",
       "      <td>0.020000</td>\n",
       "      <td>0.120000</td>\n",
       "      <td>0.060000</td>\n",
       "      <td>0.030000</td>\n",
       "      <td>0.150000</td>\n",
       "      <td>0.005000</td>\n",
       "      <td>0.015000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8513</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.015106</td>\n",
       "      <td>0.012085</td>\n",
       "      <td>0.060423</td>\n",
       "      <td>0.057402</td>\n",
       "      <td>0.063444</td>\n",
       "      <td>0.030211</td>\n",
       "      <td>0.057402</td>\n",
       "      <td>0.078550</td>\n",
       "      <td>0.009063</td>\n",
       "      <td>0.012085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4295</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.016807</td>\n",
       "      <td>0.008403</td>\n",
       "      <td>0.046218</td>\n",
       "      <td>0.016807</td>\n",
       "      <td>0.096639</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.029412</td>\n",
       "      <td>0.088235</td>\n",
       "      <td>0.004202</td>\n",
       "      <td>0.008403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.025424</td>\n",
       "      <td>0.016949</td>\n",
       "      <td>0.042373</td>\n",
       "      <td>0.029661</td>\n",
       "      <td>0.101695</td>\n",
       "      <td>0.076271</td>\n",
       "      <td>0.055085</td>\n",
       "      <td>0.088983</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012712</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100000 rows × 49 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0  local_gap_compressed_percent_id  \\\n",
       "12897         0.0                         0.287582   \n",
       "13026         1.0                         0.319635   \n",
       "8203          2.0                         0.279621   \n",
       "3340          3.0                         0.327273   \n",
       "14020         4.0                         0.338710   \n",
       "...           ...                              ...   \n",
       "8772          NaN                              NaN   \n",
       "1395          NaN                              NaN   \n",
       "8513          NaN                              NaN   \n",
       "4295          NaN                              NaN   \n",
       "106           NaN                              NaN   \n",
       "\n",
       "       scaled_local_query_percent_id  scaled_local_symmetric_percent_id  \\\n",
       "12897                       0.217822                           0.215686   \n",
       "13026                       0.295359                           0.297872   \n",
       "8203                        0.234127                           0.218924   \n",
       "3340                        0.200743                           0.214712   \n",
       "14020                       0.318182                           0.287671   \n",
       "...                              ...                                ...   \n",
       "8772                             NaN                                NaN   \n",
       "1395                             NaN                                NaN   \n",
       "8513                             NaN                                NaN   \n",
       "4295                             NaN                                NaN   \n",
       "106                              NaN                                NaN   \n",
       "\n",
       "       query_align_len  query_align_cov  subject_align_len  subject_align_cov  \\\n",
       "12897            160.0         0.792079              152.0           0.737864   \n",
       "13026            218.0         0.919831              226.0           0.969957   \n",
       "8203             211.0         0.837302              210.0           0.731707   \n",
       "3340             166.0         0.617100              163.0           0.696581   \n",
       "14020             60.0         0.909091               71.0           0.887500   \n",
       "...                ...              ...                ...                ...   \n",
       "8772               NaN              NaN                NaN                NaN   \n",
       "1395               NaN              NaN                NaN                NaN   \n",
       "8513               NaN              NaN                NaN                NaN   \n",
       "4295               NaN              NaN                NaN                NaN   \n",
       "106                NaN              NaN                NaN                NaN   \n",
       "\n",
       "       bit_score  thermo_index  ...     AAC_M     AAC_N     AAC_P     AAC_Q  \\\n",
       "12897      131.0         875.0  ...       NaN       NaN       NaN       NaN   \n",
       "13026      282.0       11324.0  ...       NaN       NaN       NaN       NaN   \n",
       "8203        96.0         875.0  ...       NaN       NaN       NaN       NaN   \n",
       "3340       175.0         875.0  ...       NaN       NaN       NaN       NaN   \n",
       "14020       61.0        9827.0  ...       NaN       NaN       NaN       NaN   \n",
       "...          ...           ...  ...       ...       ...       ...       ...   \n",
       "8772         NaN           NaN  ...  0.011070  0.014760  0.081181  0.022140   \n",
       "1395         NaN           NaN  ...  0.010000  0.005000  0.025000  0.020000   \n",
       "8513         NaN           NaN  ...  0.015106  0.012085  0.060423  0.057402   \n",
       "4295         NaN           NaN  ...  0.016807  0.008403  0.046218  0.016807   \n",
       "106          NaN           NaN  ...  0.025424  0.016949  0.042373  0.029661   \n",
       "\n",
       "          AAC_R     AAC_S     AAC_T     AAC_V     AAC_W     AAC_Y  \n",
       "12897       NaN       NaN       NaN       NaN       NaN       NaN  \n",
       "13026       NaN       NaN       NaN       NaN       NaN       NaN  \n",
       "8203        NaN       NaN       NaN       NaN       NaN       NaN  \n",
       "3340        NaN       NaN       NaN       NaN       NaN       NaN  \n",
       "14020       NaN       NaN       NaN       NaN       NaN       NaN  \n",
       "...         ...       ...       ...       ...       ...       ...  \n",
       "8772   0.073801  0.036900  0.055351  0.081181  0.036900  0.007380  \n",
       "1395   0.120000  0.060000  0.030000  0.150000  0.005000  0.015000  \n",
       "8513   0.063444  0.030211  0.057402  0.078550  0.009063  0.012085  \n",
       "4295   0.096639  0.071429  0.029412  0.088235  0.004202  0.008403  \n",
       "106    0.101695  0.076271  0.055085  0.088983  0.000000  0.012712  \n",
       "\n",
       "[100000 rows x 49 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_joined = pd.concat([df.set_index('meso_index'), learn2therm_50k], axis=0)\n",
    "df_joined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "21da85a7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAC_A</th>\n",
       "      <th>AAC_C</th>\n",
       "      <th>AAC_D</th>\n",
       "      <th>AAC_E</th>\n",
       "      <th>AAC_F</th>\n",
       "      <th>AAC_G</th>\n",
       "      <th>AAC_H</th>\n",
       "      <th>AAC_I</th>\n",
       "      <th>AAC_K</th>\n",
       "      <th>AAC_L</th>\n",
       "      <th>AAC_M</th>\n",
       "      <th>AAC_N</th>\n",
       "      <th>AAC_P</th>\n",
       "      <th>AAC_Q</th>\n",
       "      <th>AAC_R</th>\n",
       "      <th>AAC_S</th>\n",
       "      <th>AAC_T</th>\n",
       "      <th>AAC_V</th>\n",
       "      <th>AAC_W</th>\n",
       "      <th>AAC_Y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12897</th>\n",
       "      <td>0.145631</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.072816</td>\n",
       "      <td>0.067961</td>\n",
       "      <td>0.014563</td>\n",
       "      <td>0.067961</td>\n",
       "      <td>0.019417</td>\n",
       "      <td>0.014563</td>\n",
       "      <td>0.014563</td>\n",
       "      <td>0.106796</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.004854</td>\n",
       "      <td>0.053398</td>\n",
       "      <td>0.038835</td>\n",
       "      <td>0.131068</td>\n",
       "      <td>0.033981</td>\n",
       "      <td>0.048544</td>\n",
       "      <td>0.106796</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.029126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13026</th>\n",
       "      <td>0.064378</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.115880</td>\n",
       "      <td>0.068670</td>\n",
       "      <td>0.030043</td>\n",
       "      <td>0.051502</td>\n",
       "      <td>0.008584</td>\n",
       "      <td>0.051502</td>\n",
       "      <td>0.055794</td>\n",
       "      <td>0.107296</td>\n",
       "      <td>0.055794</td>\n",
       "      <td>0.025751</td>\n",
       "      <td>0.025751</td>\n",
       "      <td>0.042918</td>\n",
       "      <td>0.090129</td>\n",
       "      <td>0.042918</td>\n",
       "      <td>0.047210</td>\n",
       "      <td>0.081545</td>\n",
       "      <td>0.004292</td>\n",
       "      <td>0.030043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8203</th>\n",
       "      <td>0.205575</td>\n",
       "      <td>0.006969</td>\n",
       "      <td>0.062718</td>\n",
       "      <td>0.034843</td>\n",
       "      <td>0.031359</td>\n",
       "      <td>0.094077</td>\n",
       "      <td>0.020906</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.010453</td>\n",
       "      <td>0.094077</td>\n",
       "      <td>0.010453</td>\n",
       "      <td>0.020906</td>\n",
       "      <td>0.034843</td>\n",
       "      <td>0.013937</td>\n",
       "      <td>0.104530</td>\n",
       "      <td>0.031359</td>\n",
       "      <td>0.055749</td>\n",
       "      <td>0.114983</td>\n",
       "      <td>0.010453</td>\n",
       "      <td>0.017422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3340</th>\n",
       "      <td>0.119658</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012821</td>\n",
       "      <td>0.042735</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.094017</td>\n",
       "      <td>0.004274</td>\n",
       "      <td>0.059829</td>\n",
       "      <td>0.038462</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.029915</td>\n",
       "      <td>0.008547</td>\n",
       "      <td>0.047009</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>0.051282</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.051282</td>\n",
       "      <td>0.094017</td>\n",
       "      <td>0.029915</td>\n",
       "      <td>0.025641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14020</th>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.025000</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.075000</td>\n",
       "      <td>0.012500</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>0.012500</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.087500</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.025000</td>\n",
       "      <td>0.087500</td>\n",
       "      <td>0.037500</td>\n",
       "      <td>0.062500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8772</th>\n",
       "      <td>0.158672</td>\n",
       "      <td>0.007380</td>\n",
       "      <td>0.092251</td>\n",
       "      <td>0.040590</td>\n",
       "      <td>0.036900</td>\n",
       "      <td>0.088561</td>\n",
       "      <td>0.014760</td>\n",
       "      <td>0.018450</td>\n",
       "      <td>0.011070</td>\n",
       "      <td>0.110701</td>\n",
       "      <td>0.011070</td>\n",
       "      <td>0.014760</td>\n",
       "      <td>0.081181</td>\n",
       "      <td>0.022140</td>\n",
       "      <td>0.073801</td>\n",
       "      <td>0.036900</td>\n",
       "      <td>0.055351</td>\n",
       "      <td>0.081181</td>\n",
       "      <td>0.036900</td>\n",
       "      <td>0.007380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1395</th>\n",
       "      <td>0.105000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.065000</td>\n",
       "      <td>0.090000</td>\n",
       "      <td>0.030000</td>\n",
       "      <td>0.080000</td>\n",
       "      <td>0.020000</td>\n",
       "      <td>0.040000</td>\n",
       "      <td>0.030000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.010000</td>\n",
       "      <td>0.005000</td>\n",
       "      <td>0.025000</td>\n",
       "      <td>0.020000</td>\n",
       "      <td>0.120000</td>\n",
       "      <td>0.060000</td>\n",
       "      <td>0.030000</td>\n",
       "      <td>0.150000</td>\n",
       "      <td>0.005000</td>\n",
       "      <td>0.015000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8513</th>\n",
       "      <td>0.126888</td>\n",
       "      <td>0.003021</td>\n",
       "      <td>0.063444</td>\n",
       "      <td>0.057402</td>\n",
       "      <td>0.018127</td>\n",
       "      <td>0.105740</td>\n",
       "      <td>0.021148</td>\n",
       "      <td>0.069486</td>\n",
       "      <td>0.024169</td>\n",
       "      <td>0.114804</td>\n",
       "      <td>0.015106</td>\n",
       "      <td>0.012085</td>\n",
       "      <td>0.060423</td>\n",
       "      <td>0.057402</td>\n",
       "      <td>0.063444</td>\n",
       "      <td>0.030211</td>\n",
       "      <td>0.057402</td>\n",
       "      <td>0.078550</td>\n",
       "      <td>0.009063</td>\n",
       "      <td>0.012085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4295</th>\n",
       "      <td>0.168067</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.046218</td>\n",
       "      <td>0.054622</td>\n",
       "      <td>0.029412</td>\n",
       "      <td>0.126050</td>\n",
       "      <td>0.004202</td>\n",
       "      <td>0.008403</td>\n",
       "      <td>0.016807</td>\n",
       "      <td>0.159664</td>\n",
       "      <td>0.016807</td>\n",
       "      <td>0.008403</td>\n",
       "      <td>0.046218</td>\n",
       "      <td>0.016807</td>\n",
       "      <td>0.096639</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.029412</td>\n",
       "      <td>0.088235</td>\n",
       "      <td>0.004202</td>\n",
       "      <td>0.008403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>0.101695</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.084746</td>\n",
       "      <td>0.067797</td>\n",
       "      <td>0.008475</td>\n",
       "      <td>0.088983</td>\n",
       "      <td>0.012712</td>\n",
       "      <td>0.059322</td>\n",
       "      <td>0.008475</td>\n",
       "      <td>0.118644</td>\n",
       "      <td>0.025424</td>\n",
       "      <td>0.016949</td>\n",
       "      <td>0.042373</td>\n",
       "      <td>0.029661</td>\n",
       "      <td>0.101695</td>\n",
       "      <td>0.076271</td>\n",
       "      <td>0.055085</td>\n",
       "      <td>0.088983</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012712</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>50000 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          AAC_A     AAC_C     AAC_D     AAC_E     AAC_F     AAC_G     AAC_H  \\\n",
       "12897  0.145631  0.009709  0.072816  0.067961  0.014563  0.067961  0.019417   \n",
       "13026  0.064378  0.000000  0.115880  0.068670  0.030043  0.051502  0.008584   \n",
       "8203   0.205575  0.006969  0.062718  0.034843  0.031359  0.094077  0.020906   \n",
       "3340   0.119658  0.000000  0.012821  0.042735  0.055556  0.094017  0.004274   \n",
       "14020  0.050000  0.000000  0.025000  0.050000  0.100000  0.075000  0.012500   \n",
       "...         ...       ...       ...       ...       ...       ...       ...   \n",
       "8772   0.158672  0.007380  0.092251  0.040590  0.036900  0.088561  0.014760   \n",
       "1395   0.105000  0.000000  0.065000  0.090000  0.030000  0.080000  0.020000   \n",
       "8513   0.126888  0.003021  0.063444  0.057402  0.018127  0.105740  0.021148   \n",
       "4295   0.168067  0.000000  0.046218  0.054622  0.029412  0.126050  0.004202   \n",
       "106    0.101695  0.000000  0.084746  0.067797  0.008475  0.088983  0.012712   \n",
       "\n",
       "          AAC_I     AAC_K     AAC_L     AAC_M     AAC_N     AAC_P     AAC_Q  \\\n",
       "12897  0.014563  0.014563  0.106796  0.009709  0.004854  0.053398  0.038835   \n",
       "13026  0.051502  0.055794  0.107296  0.055794  0.025751  0.025751  0.042918   \n",
       "8203   0.024390  0.010453  0.094077  0.010453  0.020906  0.034843  0.013937   \n",
       "3340   0.059829  0.038462  0.153846  0.029915  0.008547  0.047009  0.025641   \n",
       "14020  0.062500  0.012500  0.100000  0.050000  0.037500  0.050000  0.037500   \n",
       "...         ...       ...       ...       ...       ...       ...       ...   \n",
       "8772   0.018450  0.011070  0.110701  0.011070  0.014760  0.081181  0.022140   \n",
       "1395   0.040000  0.030000  0.100000  0.010000  0.005000  0.025000  0.020000   \n",
       "8513   0.069486  0.024169  0.114804  0.015106  0.012085  0.060423  0.057402   \n",
       "4295   0.008403  0.016807  0.159664  0.016807  0.008403  0.046218  0.016807   \n",
       "106    0.059322  0.008475  0.118644  0.025424  0.016949  0.042373  0.029661   \n",
       "\n",
       "          AAC_R     AAC_S     AAC_T     AAC_V     AAC_W     AAC_Y  \n",
       "12897  0.131068  0.033981  0.048544  0.106796  0.009709  0.029126  \n",
       "13026  0.090129  0.042918  0.047210  0.081545  0.004292  0.030043  \n",
       "8203   0.104530  0.031359  0.055749  0.114983  0.010453  0.017422  \n",
       "3340   0.051282  0.055556  0.051282  0.094017  0.029915  0.025641  \n",
       "14020  0.087500  0.037500  0.025000  0.087500  0.037500  0.062500  \n",
       "...         ...       ...       ...       ...       ...       ...  \n",
       "8772   0.073801  0.036900  0.055351  0.081181  0.036900  0.007380  \n",
       "1395   0.120000  0.060000  0.030000  0.150000  0.005000  0.015000  \n",
       "8513   0.063444  0.030211  0.057402  0.078550  0.009063  0.012085  \n",
       "4295   0.096639  0.071429  0.029412  0.088235  0.004202  0.008403  \n",
       "106    0.101695  0.076271  0.055085  0.088983  0.000000  0.012712  \n",
       "\n",
       "[50000 rows x 20 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "learn2therm_50k"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "29197a17",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n"
     ]
    }
   ],
   "source": [
    "# check if the index is properly aligned\n",
    "print(df.set_index('meso_index').reset_index().index.equals(learn2therm_50k.sort_index().index))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7be5ead",
   "metadata": {},
   "source": [
    "Scratch work for above function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "d863dd63",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequence = \"meso_input_copy.fasta.txt\"\n",
    "string_to_remove = \"<unknown description>\"\n",
    "\n",
    "with open(sequence, \"r\") as file:\n",
    "    content = file.read()\n",
    "    \n",
    "# Remove the string\n",
    "new_content = content.replace(string_to_remove, \"\")\n",
    "\n",
    "with open(sequence, \"w\") as file:\n",
    "    file.write(new_content)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e9a78f6e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#just figured out that iFeature cannot read the <unknown descriptor> part of the fasta sequence\n",
    "\n",
    "\n",
    "protein = iFeatureOmegaCLI.iProtein('meso_input_copy.fasta.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "b4981b0a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "        ----- Available feature types ------        \n",
      "        \n",
      "        AAC                                                Amino acid composition\n",
      "        EAAC                                               Enhanced amino acid composition\n",
      "        CKSAAP type 1                                      Composition of k-spaced amino acid pairs type 1 - normalized\n",
      "        CKSAAP type 2                                      Composition of k-spaced amino acid pairs type 2 - raw count\n",
      "        DPC type 1                                         Dipeptide composition type 1 - normalized\n",
      "        DPC type 2                                         Dipeptide composition type 2 - raw count\n",
      "        TPC type 1                                         Tripeptide composition type 1 - normalized\n",
      "        TPC type 2                                         Tripeptide composition type 1 - raw count\n",
      "        CTDC                                               Composition\n",
      "        CTDT                                               Transition\n",
      "        CTDD                                               Distribution\n",
      "        CTriad                                             Conjoint triad\n",
      "        KSCTriad                                           Conjoint k-spaced triad\n",
      "        ASDC                                               Adaptive skip dipeptide composition\n",
      "        DistancePair                                       PseAAC of distance-pairs and reduced alphabe\n",
      "        GAAC                                               Grouped amino acid composition\n",
      "        EGAAC                                              Enhanced grouped amino acid composition\n",
      "        CKSAAGP type 1                                     Composition of k-spaced amino acid group pairs type 1- normalized\n",
      "        CKSAAGP type 2                                     Composition of k-spaced amino acid group pairs type 2- raw count\n",
      "        GDPC type 1                                        Grouped dipeptide composition type 1 - normalized\n",
      "        GDPC type 2                                        Grouped dipeptide composition type 2 - raw count\n",
      "        GTPC type 1                                        Grouped tripeptide composition type 1 - normalized\n",
      "        GTPC type 2                                        Grouped tripeptide composition type 1 - raw count\n",
      "        Moran                                              Moran\n",
      "        Geary                                              Geary\n",
      "        NMBroto                                            Normalized Moreau-Broto\n",
      "        AC                                                 Auto covariance\n",
      "        CC                                                 Cross covariance\n",
      "        ACC                                                Auto-cross covariance\n",
      "        SOCNumber                                          Sequence-order-coupling number\n",
      "        QSOrder                                            Quasi-sequence-order descriptors\n",
      "        PAAC                                               Pseudo-amino acid composition\n",
      "        APAAC                                              Amphiphilic PAAC\n",
      "        PseKRAAC type 1                                    Pseudo K-tuple reduced amino acids composition type 1\n",
      "        PseKRAAC type 2                                    Pseudo K-tuple reduced amino acids composition type 2\n",
      "        PseKRAAC type 3A                                   Pseudo K-tuple reduced amino acids composition type 3A\n",
      "        PseKRAAC type 3B                                   Pseudo K-tuple reduced amino acids composition type 3B\n",
      "        PseKRAAC type 4                                    Pseudo K-tuple reduced amino acids composition type 4\n",
      "        PseKRAAC type 5                                    Pseudo K-tuple reduced amino acids composition type 5\n",
      "        PseKRAAC type 6A                                   Pseudo K-tuple reduced amino acids composition type 6A\n",
      "        PseKRAAC type 6B                                   Pseudo K-tuple reduced amino acids composition type 6B\n",
      "        PseKRAAC type 6C                                   Pseudo K-tuple reduced amino acids composition type 6C\n",
      "        PseKRAAC type 7                                    Pseudo K-tuple reduced amino acids composition type 7\n",
      "        PseKRAAC type 8                                    Pseudo K-tuple reduced amino acids composition type 8\n",
      "        PseKRAAC type 9                                    Pseudo K-tuple reduced amino acids composition type 9\n",
      "        PseKRAAC type 10                                   Pseudo K-tuple reduced amino acids composition type 10\n",
      "        PseKRAAC type 11                                   Pseudo K-tuple reduced amino acids composition type 11\n",
      "        PseKRAAC type 12                                   Pseudo K-tuple reduced amino acids composition type 12\n",
      "        PseKRAAC type 13                                   Pseudo K-tuple reduced amino acids composition type 13\n",
      "        PseKRAAC type 14                                   Pseudo K-tuple reduced amino acids composition type 14\n",
      "        PseKRAAC type 15                                   Pseudo K-tuple reduced amino acids composition type 15\n",
      "        PseKRAAC type 16                                   Pseudo K-tuple reduced amino acids composition type 16\n",
      "        binary                                             Binary\n",
      "        binary_6bit                                        Binary\n",
      "        binary_5bit type 1                                 Binary\n",
      "        binary_5bit type 2                                 Binary\n",
      "        binary_3bit type 1                                 Binary\n",
      "        binary_3bit type 2                                 Binary\n",
      "        binary_3bit type 3                                 Binary\n",
      "        binary_3bit type 4                                 Binary\n",
      "        binary_3bit type 5                                 Binary\n",
      "        binary_3bit type 6                                 Binary\n",
      "        binary_3bit type 7                                 Binary\n",
      "        AESNN3                                             Learn from alignments\n",
      "        OPF_10bit                                          Overlapping property features - 10 bit\n",
      "        OPF_7bit type 1                                    Overlapping property features - 7 bit type 1\n",
      "        OPF_7bit type 2                                    Overlapping property features - 7 bit type 2\n",
      "        OPF_7bit type 3                                    Overlapping property features - 7 bit type 3\n",
      "        AAIndex                                            AAIndex\n",
      "        BLOSUM62                                           BLOSUM62\n",
      "        ZScale                                             Z-Scales index\n",
      "        KNN                                                K-nearest neighbor\n",
      "\n",
      "        Note: the first column is the names of availables feature types while the second column is description.  \n",
      "        \n",
      "        \n"
     ]
    }
   ],
   "source": [
    "protein.display_feature_types()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "96d2c565",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    50000.000000\n",
       "mean       261.367700\n",
       "std         69.675548\n",
       "min         27.000000\n",
       "25%        226.000000\n",
       "50%        260.000000\n",
       "75%        311.000000\n",
       "max        400.000000\n",
       "Name: t_protein_len, dtype: float64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['t_protein_len'].describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04026501",
   "metadata": {},
   "source": [
    "The bit-score provides a better rule-of-thumb for inferring homology. For average length proteins, a bit score of 50 is almost always significant. A bit score of 40 is only significant (E() < 0.001) in searches of protein databases with fewer than 7000 entries. Increasing the score by 10 bits increases the significance 210=1000-fold, so 50 bits would be significant in a database with less than 7 million entries (10 times SwissProt, and within a factor of 3 of the largest protein databases). Thus, the NCBI Blast web site uses a color code of blue for alignment with scores between 40–50 bits; and green for scores between 50–80 bits. In the yeast vs human example, the alignments with less than 20% identity had scores ranging from 55 – 170 bits. Except for very long proteins and very large databases, 50 bits of similarity score will always be statistically significant and is a much better rule-of-thumb for inferring homology in protein alignments.\n",
    "\n",
    "Pearson et al., 2013: An Introduction to Sequence Similarity (“Homology”) Searching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "273884e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['protein_match'] = ((df['bit_score'] > 50 ) & (df['query_align_cov'] > 0.8))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "14015e54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True     25887\n",
       "False    24113\n",
       "Name: protein_match, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['protein_match'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "084d9561",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get rid of stuff that isn't quantitative\n",
    "\n",
    "df = df.drop(columns = ['Unnamed: 0','thermo_index', 'm_protein_seq', 't_protein_seq',\n",
    "                        'm_protein_desc', 't_protein_desc'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "eea27e6c",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>meso_index</th>\n",
       "      <th>prot_pair_index</th>\n",
       "      <th>...</th>\n",
       "      <th>scaled_local_symmetric_percent_id_16s</th>\n",
       "      <th>query_align_cov_16s</th>\n",
       "      <th>subject_align_cov_16s</th>\n",
       "      <th>bit_score_16s</th>\n",
       "      <th>m_ogt</th>\n",
       "      <th>t_ogt</th>\n",
       "      <th>ogt_difference</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "      <th>protein_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.287582</td>\n",
       "      <td>0.217822</td>\n",
       "      <td>0.215686</td>\n",
       "      <td>160</td>\n",
       "      <td>0.792079</td>\n",
       "      <td>152</td>\n",
       "      <td>0.737864</td>\n",
       "      <td>131</td>\n",
       "      <td>12897</td>\n",
       "      <td>48641291</td>\n",
       "      <td>...</td>\n",
       "      <td>0.926461</td>\n",
       "      <td>0.998040</td>\n",
       "      <td>0.998020</td>\n",
       "      <td>1153.0</td>\n",
       "      <td>27.5</td>\n",
       "      <td>50.0</td>\n",
       "      <td>22.5</td>\n",
       "      <td>206</td>\n",
       "      <td>202</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.319635</td>\n",
       "      <td>0.295359</td>\n",
       "      <td>0.297872</td>\n",
       "      <td>218</td>\n",
       "      <td>0.919831</td>\n",
       "      <td>226</td>\n",
       "      <td>0.969957</td>\n",
       "      <td>282</td>\n",
       "      <td>13026</td>\n",
       "      <td>92992745</td>\n",
       "      <td>...</td>\n",
       "      <td>0.902522</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1014.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>233</td>\n",
       "      <td>237</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.279621</td>\n",
       "      <td>0.234127</td>\n",
       "      <td>0.218924</td>\n",
       "      <td>211</td>\n",
       "      <td>0.837302</td>\n",
       "      <td>210</td>\n",
       "      <td>0.731707</td>\n",
       "      <td>96</td>\n",
       "      <td>8203</td>\n",
       "      <td>157628663</td>\n",
       "      <td>...</td>\n",
       "      <td>0.923531</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1138.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>287</td>\n",
       "      <td>252</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.327273</td>\n",
       "      <td>0.200743</td>\n",
       "      <td>0.214712</td>\n",
       "      <td>166</td>\n",
       "      <td>0.617100</td>\n",
       "      <td>163</td>\n",
       "      <td>0.696581</td>\n",
       "      <td>175</td>\n",
       "      <td>3340</td>\n",
       "      <td>136708305</td>\n",
       "      <td>...</td>\n",
       "      <td>0.909745</td>\n",
       "      <td>0.996734</td>\n",
       "      <td>0.996726</td>\n",
       "      <td>1077.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>234</td>\n",
       "      <td>269</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.338710</td>\n",
       "      <td>0.318182</td>\n",
       "      <td>0.287671</td>\n",
       "      <td>60</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>71</td>\n",
       "      <td>0.887500</td>\n",
       "      <td>61</td>\n",
       "      <td>14020</td>\n",
       "      <td>133672542</td>\n",
       "      <td>...</td>\n",
       "      <td>0.887107</td>\n",
       "      <td>0.998084</td>\n",
       "      <td>0.998067</td>\n",
       "      <td>991.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>80</td>\n",
       "      <td>66</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   local_gap_compressed_percent_id  scaled_local_query_percent_id  \\\n",
       "0                         0.287582                       0.217822   \n",
       "1                         0.319635                       0.295359   \n",
       "2                         0.279621                       0.234127   \n",
       "3                         0.327273                       0.200743   \n",
       "4                         0.338710                       0.318182   \n",
       "\n",
       "   scaled_local_symmetric_percent_id  query_align_len  query_align_cov  \\\n",
       "0                           0.215686              160         0.792079   \n",
       "1                           0.297872              218         0.919831   \n",
       "2                           0.218924              211         0.837302   \n",
       "3                           0.214712              166         0.617100   \n",
       "4                           0.287671               60         0.909091   \n",
       "\n",
       "   subject_align_len  subject_align_cov  bit_score  meso_index  \\\n",
       "0                152           0.737864        131       12897   \n",
       "1                226           0.969957        282       13026   \n",
       "2                210           0.731707         96        8203   \n",
       "3                163           0.696581        175        3340   \n",
       "4                 71           0.887500         61       14020   \n",
       "\n",
       "   prot_pair_index  ...  scaled_local_symmetric_percent_id_16s  \\\n",
       "0         48641291  ...                               0.926461   \n",
       "1         92992745  ...                               0.902522   \n",
       "2        157628663  ...                               0.923531   \n",
       "3        136708305  ...                               0.909745   \n",
       "4        133672542  ...                               0.887107   \n",
       "\n",
       "   query_align_cov_16s  subject_align_cov_16s  bit_score_16s  m_ogt  t_ogt  \\\n",
       "0             0.998040               0.998020         1153.0   27.5   50.0   \n",
       "1             1.000000               1.000000         1014.0   25.0   54.0   \n",
       "2             1.000000               1.000000         1138.0   28.0   50.0   \n",
       "3             0.996734               0.996726         1077.0   28.0   50.0   \n",
       "4             0.998084               0.998067          991.0   30.0   50.0   \n",
       "\n",
       "   ogt_difference  m_protein_len  t_protein_len  protein_match  \n",
       "0            22.5            206            202          False  \n",
       "1            29.0            233            237           True  \n",
       "2            22.0            287            252           True  \n",
       "3            22.0            234            269          False  \n",
       "4            20.0             80             66           True  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07cc3a80",
   "metadata": {},
   "source": [
    "From pairplot below, it looks like query_align_cov_16s and subject_align_cov_16s dont correlate with target. Let's remove them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "36d8dadb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# sns.pairplot(df.sample(500),height = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "91079edd",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.drop(columns = ['query_align_cov_16s', 'subject_align_cov_16s'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7929625e",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>meso_index</th>\n",
       "      <th>prot_pair_index</th>\n",
       "      <th>...</th>\n",
       "      <th>taxa_pair_index</th>\n",
       "      <th>local_gap_compressed_percent_id_16s</th>\n",
       "      <th>scaled_local_query_percent_id_16s</th>\n",
       "      <th>scaled_local_symmetric_percent_id_16s</th>\n",
       "      <th>bit_score_16s</th>\n",
       "      <th>m_ogt</th>\n",
       "      <th>t_ogt</th>\n",
       "      <th>ogt_difference</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.00000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>5.000000e+04</td>\n",
       "      <td>...</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "      <td>50000.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>0.349026</td>\n",
       "      <td>0.276478</td>\n",
       "      <td>0.270792</td>\n",
       "      <td>206.011660</td>\n",
       "      <td>0.791507</td>\n",
       "      <td>205.414920</td>\n",
       "      <td>0.770875</td>\n",
       "      <td>230.63206</td>\n",
       "      <td>8098.858120</td>\n",
       "      <td>9.105934e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>362827.905140</td>\n",
       "      <td>0.906828</td>\n",
       "      <td>0.903266</td>\n",
       "      <td>0.904959</td>\n",
       "      <td>1061.404480</td>\n",
       "      <td>27.348365</td>\n",
       "      <td>51.505410</td>\n",
       "      <td>24.157045</td>\n",
       "      <td>269.985200</td>\n",
       "      <td>261.367700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.096608</td>\n",
       "      <td>0.114404</td>\n",
       "      <td>0.116039</td>\n",
       "      <td>64.260901</td>\n",
       "      <td>0.144893</td>\n",
       "      <td>64.430548</td>\n",
       "      <td>0.179955</td>\n",
       "      <td>186.44859</td>\n",
       "      <td>4608.982941</td>\n",
       "      <td>5.233883e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>229264.926253</td>\n",
       "      <td>0.015240</td>\n",
       "      <td>0.015626</td>\n",
       "      <td>0.015300</td>\n",
       "      <td>75.659661</td>\n",
       "      <td>3.737201</td>\n",
       "      <td>3.816972</td>\n",
       "      <td>3.586797</td>\n",
       "      <td>65.935753</td>\n",
       "      <td>69.675548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.164103</td>\n",
       "      <td>0.081301</td>\n",
       "      <td>0.045977</td>\n",
       "      <td>19.000000</td>\n",
       "      <td>0.284553</td>\n",
       "      <td>19.000000</td>\n",
       "      <td>0.082019</td>\n",
       "      <td>35.00000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>1.219000e+03</td>\n",
       "      <td>...</td>\n",
       "      <td>4885.000000</td>\n",
       "      <td>0.876590</td>\n",
       "      <td>0.836055</td>\n",
       "      <td>0.880026</td>\n",
       "      <td>891.000000</td>\n",
       "      <td>7.000000</td>\n",
       "      <td>41.000000</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>30.000000</td>\n",
       "      <td>27.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>0.293750</td>\n",
       "      <td>0.204545</td>\n",
       "      <td>0.202020</td>\n",
       "      <td>172.000000</td>\n",
       "      <td>0.673288</td>\n",
       "      <td>172.000000</td>\n",
       "      <td>0.656854</td>\n",
       "      <td>118.00000</td>\n",
       "      <td>4121.000000</td>\n",
       "      <td>4.645832e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>173976.750000</td>\n",
       "      <td>0.895995</td>\n",
       "      <td>0.893020</td>\n",
       "      <td>0.893531</td>\n",
       "      <td>1009.000000</td>\n",
       "      <td>27.000000</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>22.000000</td>\n",
       "      <td>233.000000</td>\n",
       "      <td>226.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>0.325000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.244648</td>\n",
       "      <td>210.000000</td>\n",
       "      <td>0.811321</td>\n",
       "      <td>209.000000</td>\n",
       "      <td>0.806584</td>\n",
       "      <td>186.00000</td>\n",
       "      <td>8046.000000</td>\n",
       "      <td>9.160212e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>312362.000000</td>\n",
       "      <td>0.905983</td>\n",
       "      <td>0.902326</td>\n",
       "      <td>0.903543</td>\n",
       "      <td>1056.000000</td>\n",
       "      <td>28.000000</td>\n",
       "      <td>52.500000</td>\n",
       "      <td>24.500000</td>\n",
       "      <td>267.000000</td>\n",
       "      <td>260.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>0.367257</td>\n",
       "      <td>0.308300</td>\n",
       "      <td>0.301639</td>\n",
       "      <td>240.000000</td>\n",
       "      <td>0.923077</td>\n",
       "      <td>240.000000</td>\n",
       "      <td>0.923077</td>\n",
       "      <td>267.00000</td>\n",
       "      <td>12103.000000</td>\n",
       "      <td>1.360713e+08</td>\n",
       "      <td>...</td>\n",
       "      <td>617814.000000</td>\n",
       "      <td>0.916172</td>\n",
       "      <td>0.912576</td>\n",
       "      <td>0.914698</td>\n",
       "      <td>1108.000000</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>52.500000</td>\n",
       "      <td>24.500000</td>\n",
       "      <td>317.000000</td>\n",
       "      <td>311.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>0.985075</td>\n",
       "      <td>0.985075</td>\n",
       "      <td>0.985075</td>\n",
       "      <td>400.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>399.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1788.00000</td>\n",
       "      <td>16564.000000</td>\n",
       "      <td>1.814866e+08</td>\n",
       "      <td>...</td>\n",
       "      <td>744518.000000</td>\n",
       "      <td>0.999351</td>\n",
       "      <td>0.993544</td>\n",
       "      <td>0.996439</td>\n",
       "      <td>1679.000000</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>78.000000</td>\n",
       "      <td>56.000000</td>\n",
       "      <td>400.000000</td>\n",
       "      <td>400.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       local_gap_compressed_percent_id  scaled_local_query_percent_id  \\\n",
       "count                     50000.000000                   50000.000000   \n",
       "mean                          0.349026                       0.276478   \n",
       "std                           0.096608                       0.114404   \n",
       "min                           0.164103                       0.081301   \n",
       "25%                           0.293750                       0.204545   \n",
       "50%                           0.325000                       0.250000   \n",
       "75%                           0.367257                       0.308300   \n",
       "max                           0.985075                       0.985075   \n",
       "\n",
       "       scaled_local_symmetric_percent_id  query_align_len  query_align_cov  \\\n",
       "count                       50000.000000     50000.000000     50000.000000   \n",
       "mean                            0.270792       206.011660         0.791507   \n",
       "std                             0.116039        64.260901         0.144893   \n",
       "min                             0.045977        19.000000         0.284553   \n",
       "25%                             0.202020       172.000000         0.673288   \n",
       "50%                             0.244648       210.000000         0.811321   \n",
       "75%                             0.301639       240.000000         0.923077   \n",
       "max                             0.985075       400.000000         1.000000   \n",
       "\n",
       "       subject_align_len  subject_align_cov    bit_score    meso_index  \\\n",
       "count       50000.000000       50000.000000  50000.00000  50000.000000   \n",
       "mean          205.414920           0.770875    230.63206   8098.858120   \n",
       "std            64.430548           0.179955    186.44859   4608.982941   \n",
       "min            19.000000           0.082019     35.00000      5.000000   \n",
       "25%           172.000000           0.656854    118.00000   4121.000000   \n",
       "50%           209.000000           0.806584    186.00000   8046.000000   \n",
       "75%           240.000000           0.923077    267.00000  12103.000000   \n",
       "max           399.000000           1.000000   1788.00000  16564.000000   \n",
       "\n",
       "       prot_pair_index  ...  taxa_pair_index  \\\n",
       "count     5.000000e+04  ...     50000.000000   \n",
       "mean      9.105934e+07  ...    362827.905140   \n",
       "std       5.233883e+07  ...    229264.926253   \n",
       "min       1.219000e+03  ...      4885.000000   \n",
       "25%       4.645832e+07  ...    173976.750000   \n",
       "50%       9.160212e+07  ...    312362.000000   \n",
       "75%       1.360713e+08  ...    617814.000000   \n",
       "max       1.814866e+08  ...    744518.000000   \n",
       "\n",
       "       local_gap_compressed_percent_id_16s  scaled_local_query_percent_id_16s  \\\n",
       "count                         50000.000000                       50000.000000   \n",
       "mean                              0.906828                           0.903266   \n",
       "std                               0.015240                           0.015626   \n",
       "min                               0.876590                           0.836055   \n",
       "25%                               0.895995                           0.893020   \n",
       "50%                               0.905983                           0.902326   \n",
       "75%                               0.916172                           0.912576   \n",
       "max                               0.999351                           0.993544   \n",
       "\n",
       "       scaled_local_symmetric_percent_id_16s  bit_score_16s         m_ogt  \\\n",
       "count                           50000.000000   50000.000000  50000.000000   \n",
       "mean                                0.904959    1061.404480     27.348365   \n",
       "std                                 0.015300      75.659661      3.737201   \n",
       "min                                 0.880026     891.000000      7.000000   \n",
       "25%                                 0.893531    1009.000000     27.000000   \n",
       "50%                                 0.903543    1056.000000     28.000000   \n",
       "75%                                 0.914698    1108.000000     29.000000   \n",
       "max                                 0.996439    1679.000000     40.000000   \n",
       "\n",
       "              t_ogt  ogt_difference  m_protein_len  t_protein_len  \n",
       "count  50000.000000    50000.000000   50000.000000   50000.000000  \n",
       "mean      51.505410       24.157045     269.985200     261.367700  \n",
       "std        3.816972        3.586797      65.935753      69.675548  \n",
       "min       41.000000       20.000000      30.000000      27.000000  \n",
       "25%       50.000000       22.000000     233.000000     226.000000  \n",
       "50%       52.500000       24.500000     267.000000     260.000000  \n",
       "75%       52.500000       24.500000     317.000000     311.000000  \n",
       "max       78.000000       56.000000     400.000000     400.000000  \n",
       "\n",
       "[8 rows x 22 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90871d01",
   "metadata": {},
   "source": [
    "Split data into dev and test, and then split that into train and validation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6519dcaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#drop columns that don't exihibit signficant pearson correlation with bit_score\n",
    "\n",
    "df = df.drop(columns = ['meso_index', 'meso_protein_int_index', 'local_gap_compressed_percent_id_16s', \n",
    "                        'scaled_local_query_percent_id_16s', 'scaled_local_symmetric_percent_id_16s',\n",
    "                       'bit_score_16s', 'm_ogt', 't_ogt', 'taxa_pair_index', 'thermo_protein_int_index'\n",
    "                       , 'prot_pair_index', 'ogt_difference'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "af9f23cf",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(42500, 11)\n",
      "(7500, 11)\n",
      "(36125, 11)\n",
      "(6375, 11)\n"
     ]
    }
   ],
   "source": [
    "#choosing 80/20 split instead of 85/15 because of volume of data\n",
    "\n",
    "dev, test = sklearn.model_selection.train_test_split(df, test_size=0.15, random_state=1)\n",
    "\n",
    "train, val = sklearn.model_selection.train_test_split(dev, test_size=0.15, random_state=1)\n",
    "\n",
    "print(dev.shape)\n",
    "print(test.shape)\n",
    "print(train.shape)\n",
    "print(val.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8470550d",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#ID target and features, separate into separate arrays\n",
    "\n",
    "target = 'protein_match'\n",
    "input_features = [columns for columns in df]\n",
    "input_features.remove(target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c62edd28",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['local_gap_compressed_percent_id', 'scaled_local_query_percent_id', 'scaled_local_symmetric_percent_id', 'query_align_len', 'query_align_cov', 'subject_align_len', 'subject_align_cov', 'bit_score', 'm_protein_len', 't_protein_len']\n",
      "protein_match\n"
     ]
    }
   ],
   "source": [
    "print(input_features)\n",
    "print(target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1ab62bff",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(42500, 10) (7500, 10) (42500, 1) (7500, 1)\n"
     ]
    }
   ],
   "source": [
    "#split X and y\n",
    "\n",
    "dev_X = dev[input_features].values\n",
    "test_X = test[input_features].values\n",
    "\n",
    "dev_y = dev[target].values.reshape(-1,1)\n",
    "test_y = test[target].values.reshape(-1,1)  \n",
    "\n",
    "print(dev_X.shape, test_X.shape, dev_y.shape, test_y.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "9ac52ac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#same thing for training and validation data\n",
    "\n",
    "train_X = train[input_features].values\n",
    "val_X = val[input_features].values\n",
    "\n",
    "train_y = train[target].values.reshape(-1,1)\n",
    "val_y = val[target].values.reshape(-1,1) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c374f2d7",
   "metadata": {},
   "source": [
    "Scale the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "2b48b2e5",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "scaler = sklearn.preprocessing.StandardScaler()\n",
    "dev_X = scaler.fit_transform(dev_X)\n",
    "test_X = scaler.fit_transform(test_X)\n",
    "train_X = scaler.fit_transform(train_X)\n",
    "val_X = scaler.fit_transform(val_X)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3022371",
   "metadata": {},
   "source": [
    "Train the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "be4e09b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-1 {color: black;background-color: white;}#sk-container-id-1 pre{padding: 0;}#sk-container-id-1 div.sk-toggleable {background-color: white;}#sk-container-id-1 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-1 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-1 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-1 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-1 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-1 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-1 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-1 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-1 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-1 div.sk-item {position: relative;z-index: 1;}#sk-container-id-1 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-1 div.sk-item::before, #sk-container-id-1 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-1 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-1 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-1 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-1 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-1 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-1 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-1 div.sk-label-container {text-align: center;}#sk-container-id-1 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-1 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>RandomForestClassifier(max_features=0.5, max_samples=0.5, min_samples_split=10,\n",
       "                       min_weight_fraction_leaf=0.000215, n_estimators=150)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" checked><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">RandomForestClassifier</label><div class=\"sk-toggleable__content\"><pre>RandomForestClassifier(max_features=0.5, max_samples=0.5, min_samples_split=10,\n",
       "                       min_weight_fraction_leaf=0.000215, n_estimators=150)</pre></div></div></div></div></div>"
      ],
      "text/plain": [
       "RandomForestClassifier(max_features=0.5, max_samples=0.5, min_samples_split=10,\n",
       "                       min_weight_fraction_leaf=0.000215, n_estimators=150)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Random Forest\n",
    "\n",
    "model = sklearn.ensemble.RandomForestClassifier(n_estimators=150, max_depth=None, max_samples=0.5,\n",
    "                                                max_features=0.5, min_weight_fraction_leaf=0.000215,\n",
    "                                               min_samples_split=10)\n",
    "# model = sklearn.ensemble.RandomForestClassifier()\n",
    "model.fit(train_X, train_y.ravel())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69ebf269",
   "metadata": {},
   "source": [
    "Test the model, report relevant statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "c73327e9",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model score is: 0.9927843137254901\n",
      "[False False False ... False  True False]\n"
     ]
    }
   ],
   "source": [
    "score = model.score(val_X, val_y)\n",
    "print('Model score is: {}'.format(score))\n",
    "\n",
    "preds = model.predict(test_X)\n",
    "print(preds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4e9dce62",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<sklearn.metrics._plot.confusion_matrix.ConfusionMatrixDisplay at 0x7ff020090ac0>"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#confusion matrix\n",
    "\n",
    "confusion_matrix = sklearn.metrics.confusion_matrix(preds, test_y)\n",
    "sklearn.metrics.ConfusionMatrixDisplay(confusion_matrix).plot()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21feaafb",
   "metadata": {},
   "source": [
    "Convert the above code into functions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "bf1d563c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def train_model(dataframe, columns = [],  target = []):\n",
    "    \"\"\"\n",
    "    Takes dataframe and splits it into a training and testing set. \n",
    "    Note: Data is called train and test, but this test set is currently\n",
    "    closer to a validation set. Keeping nomenclature to keep model robust.\n",
    "    Trains a KNN classifier model with selected data.\n",
    "    \n",
    "    Params\n",
    "    ----------\n",
    "    dataframe: Pandas dataframe \n",
    "    columns: list of strings, representing input features\n",
    "    target: list of strings, representing target feature(s)\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    -Sk-learn model object\n",
    "    -train data (features)\n",
    "    -train data (target)\n",
    "    -validation data (features)\n",
    "    -validation data (target)\n",
    "    \"\"\"\n",
    "    #split data\n",
    "    dev, test = sklearn.model_selection.train_test_split(dataframe, test_size=0.15, random_state=1)\n",
    "    \n",
    "    #test input arguments\n",
    "    assert \"pandas.core.frame.DataFrame\" in str(type(dev))\n",
    "    assert \"pandas.core.frame.DataFrame\" in str(type(test))\n",
    "    assert \"str\" in str(type(columns[0]))\n",
    "    assert \"str\" in str(type(target[0]))\n",
    "   \n",
    "    #split into input and output feature(s)\n",
    "    dev_X = dev[columns].values\n",
    "    test_X = test[columns].values\n",
    "\n",
    "    dev_y = dev[target].values.reshape(-1,1)\n",
    "    test_y = test[target].values.reshape(-1,1)\n",
    "    \n",
    "    #scale data\n",
    "    scaler = sklearn.preprocessing.StandardScaler()\n",
    "    dev_X = scaler.fit_transform(dev_X)\n",
    "    test_X = scaler.fit_transform(test_X)\n",
    "    \n",
    "    #train model\n",
    "    model = sklearn.ensemble.RandomForestClassifier()\n",
    "    model = model.fit(dev_X, dev_y.ravel())\n",
    "    \n",
    "    return model, dev_X, dev_y, test_X, test_y\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "e40acd6d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "model, dev_X, dev_y, test_X, test_y = train_model(df, columns = input_features, \n",
    "                                          target='protein_match')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "46729a81",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "..\n",
      "----------------------------------------------------------------------\n",
      "Ran 2 tests in 2.559s\n",
      "\n",
      "OK\n"
     ]
    }
   ],
   "source": [
    "class TestModelTraining(unittest.TestCase):\n",
    "    \n",
    "    def test_invalid_inputs(self):\n",
    "    \n",
    "        #test that input data type is correct\n",
    "        \n",
    "        try:\n",
    "            train_model([1,2,3], columns = 'string', target = 'string')\n",
    "            self.assertTrue(False)\n",
    "        except AssertionError:\n",
    "            self.assertTrue(True)\n",
    "    \n",
    "#     def test_input_distro(self):\n",
    "        \n",
    "#         #test that dev and test features have similar Jensen Shannon Distribution\n",
    "        \n",
    "#         JSD = (\n",
    "#             scipy.stats.bootstrap((train_reg(split_data(df)[0], split_data(df)[1],\n",
    "#                                 columns = input_features, target=target)[1], train_reg(split_data(df)[0], \n",
    "#                                 split_data(df)[1], columns = input_features, target=target)[3]), \n",
    "#                                   JSD_dev_and_test, n_resamples=1000, \n",
    "#                                   batch=5, method='percentile')\n",
    "#         )\n",
    "\n",
    "#         div = JSD.confidence_interval[1]\n",
    "        \n",
    "#         #asserts that the divergence between data sets is sufficiently low\n",
    "#         assert abs(div) < 0.3, \"Warning! High JSD between dev and test set!\"\n",
    "        \n",
    "    def test_output_format(self):\n",
    "        \n",
    "        #asserts that function returns 4 objects to be assigned to pearson_corr, model, test_X, test_y\n",
    "        assert len(train_model(df, columns = input_features, \n",
    "                                          target = target)) == 5\n",
    "        \n",
    "\n",
    "suite = unittest.TestLoader().loadTestsFromTestCase(TestModelTraining)\n",
    "_ = unittest.TextTestRunner().run(suite)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "47f766ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "#need to fix this\n",
    "\n",
    "def evaluate_model(model, test_X, test_y):\n",
    "    \n",
    "    \"\"\"\n",
    "    Takes a trained model and test data and tests the model.\n",
    "    \n",
    "    Params\n",
    "    ----------\n",
    "    model: sklearn.neighbors.KNeighborsClassifier\n",
    "    test_X: numpy array\n",
    "    test_y: numpy array\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    Vector of predictions based on the model (numpy array)\n",
    "    \"\"\"\n",
    "    \n",
    "    #test input arguments\n",
    "    assert \"sklearn\" in str(type(model))\n",
    "    assert \"numpy.ndarray\" in str(type(test_X))\n",
    "    assert \"numpy.ndarray\" in str(type(test_y))\n",
    "    \n",
    "    preds = model.predict(test_X)\n",
    "    \n",
    "    return preds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "7bbb0242",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([False, False, False, ..., False,  True, False])"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preds = evaluate_model(model, test_X, test_y)\n",
    "preds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "8b2fdac6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "...\n",
      "----------------------------------------------------------------------\n",
      "Ran 3 tests in 8.492s\n",
      "\n",
      "OK\n"
     ]
    }
   ],
   "source": [
    "class TestModelPerformance(unittest.TestCase):\n",
    "\n",
    "    def test_asserts(self):\n",
    "        model, _, _, test_X, test_y = train_model(\n",
    "            df, columns=input_features, target='protein_match'\n",
    "        )\n",
    "        # assert that input types are correct\n",
    "        with self.assertRaises(AssertionError):\n",
    "            evaluate_model(model, [1, 2, 3], test_y)\n",
    "            \n",
    "    def test_model_output(self):\n",
    "        model, _, _, test_X, test_y = train_model(\n",
    "            df, columns=input_features, target='protein_match'\n",
    "        )\n",
    "        # assert output type is correct\n",
    "        output = evaluate_model(model, test_X, test_y)\n",
    "        self.assertIsInstance(output, np.ndarray)\n",
    "        \n",
    "    def test_pred_dimension(self):\n",
    "        model, _, _, test_X, test_y = train_model(df, \n",
    "            columns=input_features, target='protein_match'\n",
    "        )\n",
    "        # want to check that the number of predictions is equal to the number of test examples\n",
    "        preds = evaluate_model(model, test_X, test_y)\n",
    "        self.assertEqual(len(test_y), len(preds))\n",
    "\n",
    "suite = unittest.TestLoader().loadTestsFromTestCase(TestModelPerformance)\n",
    "_ = unittest.TextTestRunner().run(suite)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9b79929",
   "metadata": {},
   "source": [
    "A good test here can be adding outliers and seeing how much the score and predictions change (probably for the above function)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "dae1e64f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_model(model, test_X, test_y):\n",
    "    \"\"\"\n",
    "    Takes a test KNN Classifier model and plots the confusion matrix.\n",
    "    \n",
    "    Params\n",
    "    ----------\n",
    "    model: sklearn.neighbors.KNeighborsClassifier\n",
    "    test_X: numpy array\n",
    "    test_y: numpy array\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    -Confusion predictions vs. observations\n",
    "    -Model score\n",
    "    \"\"\"\n",
    "    \n",
    "    #test input arguments\n",
    "    assert \"sklearn\" in str(type(model))\n",
    "    assert \"numpy.ndarray\" in str(type(test_X))\n",
    "    assert \"numpy.ndarray\" in str(type(test_y))\n",
    "    \n",
    "    score = model.score(test_X, test_y)\n",
    "    preds = evaluate_model(model, test_X, test_y)\n",
    "   \n",
    "    # plot confusion matrix\n",
    "    confusion_matrix = sklearn.metrics.confusion_matrix(preds, test_y)\n",
    "    cm_plot = sklearn.metrics.ConfusionMatrixDisplay(confusion_matrix)\n",
    "    \n",
    "    cm_plot.plot(cmap=plt.cm.Blues)\n",
    "    cm_plot.ax_.set_title('Confusion Matrix')\n",
    "    \n",
    "    return score\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "c923e39b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9989333333333333"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAT4AAAEWCAYAAAD/x/trAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Il7ecAAAACXBIWXMAAAsTAAALEwEAmpwYAAAgVElEQVR4nO3deZwV1Z338c+3G1kMCLIKimAENbgT4hodow5CdMRkEhM1CfqQIU40zhPjZHSeeaIx8XklmRk1iRqj0XEhrqNGYoxKUMclLoBLIqIBxQVEWW1BEenm9/xRp/FKerkX+va9t+v79lWvvnWq6tTvNvLjnDpVpxQRmJnlSV2lAzAz62xOfGaWO058ZpY7TnxmljtOfGaWO058ZpY7TnxdjKRekn4rqUHSrVtQz0mS7uvI2CpB0u8lTa50HFZdnPgqRNKJkmZLWiNpSfoL+ukOqPoLwBBgQER8cXMriYhfR8T4DojnIyQdJikk3bFJ+d6p/MEi6zlP0rT29ouIiRFx7WaGa12UE18FSDoTuBj4f2RJakfgMmBSB1Q/AvhLRDR2QF3lsgw4UNKAgrLJwF866gTK+P9va1lEeOnEBegLrAG+2MY+PcgS4xtpuRjokbYdBiwCvgMsBZYAp6Rt3wc+ANanc0wBzgOmFdQ9EgigW1o/GXgZWA0sBE4qKH+k4LiDgFlAQ/p5UMG2B4EfAI+meu4DBrby3Zrjvxw4LZXVA4uB7wEPFuz7U+B14B1gDnBIKp+wyfd8tiCOC1Ica4FRqezrafsvgNsK6v8xMBNQpf+/8NK5i/9F7HwHAj2BO9rY5/8ABwD7AHsD+wH/VrB9O7IEuj1ZcrtU0rYRcS5ZK/LmiOgdEVe1FYikjwE/AyZGRB+y5PZMC/v1B36X9h0AXAj8bpMW24nAKcBgoDtwVlvnBq4DvpY+HwU8R5bkC80i+x30B24AbpXUMyLu2eR77l1wzFeBqUAf4NVN6vsOsKekkyUdQva7mxwRfm4zZ5z4Ot8AYHm03RU9CTg/IpZGxDKyltxXC7avT9vXR8TdZK2eXTczng3AHpJ6RcSSiJjbwj5HA/Mj4vqIaIyIG4EXgL8r2Oe/IuIvEbEWuIUsYbUqIv4I9Je0K1kCvK6FfaZFxIp0zv8kawm39z2viYi56Zj1m9T3Htnv8UJgGvCtiFjUTn3WBTnxdb4VwEBJ3drYZxgfba28mso21rFJ4nwP6F1qIBHxLvAl4FRgiaTfSdqtiHiaY9q+YP3NzYjneuB04DO00AKWdJakeWmE+m2yVu7Adup8va2NEfEEWddeZAnacsiJr/M9BqwDjmtjnzfIBima7chfdwOL9S6wdcH6doUbI+LeiPhbYChZK+7KIuJpjmnxZsbU7Hrgm8DdqTW2UeqKfhc4Htg2IvqRXV9Uc+it1Nlmt1XSaWQtxzdS/ZZDTnydLCIayC7iXyrpOElbS9pK0kRJP0m73Qj8m6RBkgam/du9daMVzwCHStpRUl/gnOYNkoZImpSu9a0j6zJvaKGOu4Fd0i043SR9CRgD3LWZMQEQEQuBvyG7prmpPkAj2QhwN0nfA7Yp2P4WMLKUkVtJuwA/BL5C1uX9rqR9Ni96q2VOfBWQrledSTZgsYyse3Y68Ju0yw+B2cCfgD8DT6WyzTnXDODmVNccPpqs6lIcbwAryZLQP7ZQxwrgGLLBgRVkLaVjImL55sS0Sd2PRERLrdl7gXvIbnF5FXifj3Zjm2/OXiHpqfbOky4tTAN+HBHPRsR84F+B6yX12JLvYLVHHtAys7xxi8/McseJz8xyx4nPzHLHic/Mcqetm2g7nbr1CnXvU+kwrAT7fmLHSodgJXj11VdYvny52t+zdfXbjIhoXFvUvrF22b0RMWFLzlcO1ZX4uvehx67HVzoMK8GjT1xS6RCsBAfvP26L64jGtUX/PX3/mUvbe9KmIqoq8ZlZLRDU+IxfTnxmVhoBdfWVjmKLOPGZWem0RZcJK86Jz8xK5K6umeWRW3xmliui5lt8tR29mVWAshZfMUt7NUmvSPqzpGckzU5l/SXNkDQ//dw2lUvSzyQtkPQnSWML6pmc9p9fzOtEnfjMrHR19cUtxflMROwTEc03GZ4NzIyI0WQvgzo7lU8ERqdlKtnLo5rfCXMusD/Z+2nObU6WrYZfwlc1M2Pj4EYxy+aZBDS/C/laPpytfBJwXWQeB/pJGkr2sqoZEbEyIlYBM8jexNcqJz4zK40opas7UNLsgmXqJrUFcJ+kOQXbhkTEkvT5TbJ3T0P2jpfCyWgXpbLWylvlwQ0zK13xrbnlBV3Ylnw6IhZLGgzMkPRC4caICEkdPluyW3xmVqKO6+pGxOL0cynZm/b2A95KXVjSz6Vp98XA8ILDd0hlrZW3yonPzEojoL6+uKWtaqSPSerT/BkYT/Zi+elA88jsZODO9Hk68LU0unsA0JC6xPcC4yVtmwY1xqeyVrmra2al65gbmIcAdyirqxtwQ0TcI2kWcIukKWQvmmqeCuZu4LPAArJ3N58CEBErJf0AmJX2Oz8iVrZ1Yic+MytRxzyyFhEvA3u3UL4COKKF8gBOa6Wuq4Griz23E5+Zlc6PrJlZ7tT4I2tOfGZWmiIfR6tmTnxmVjpPRGpm+eL5+Mwsj9zVNbNc6QLz8TnxmVmJ3NU1szzy4IaZ5Y6v8ZlZrshdXTPLI7f4zCxv5MRnZnmSzTzvxGdmeSKhOic+M8sZt/jMLHec+Mwsd5z4zCxflJYa5sRnZiURcovPzPKnrs5PbphZzrjFZ2b54mt8ZpZHbvGZWa54cMPMcsmPrJlZvshdXTPLISc+M8sdJz4zyxUPbphZPtV23nPiM7MSqfYfWavt6M2sIiQVtRRZV72kpyXdldZ3kvSEpAWSbpbUPZX3SOsL0vaRBXWck8pflHRUe+d04jOz0qnIpTj/BMwrWP8xcFFEjAJWAVNS+RRgVSq/KO2HpDHAl4HdgQnAZZLafOO5u7qb6dk7v8+a99bRtGEDjY0bOHzyTzj/jOM46pA9WL++iYWLlnPa+dN4Z81aAHYfNYwLzzmBPr17EhuCwyf/hDqJa340hZE7DKRpQ3Dvw3/m+5dMr/A3y7e9jv0evbfuQX1dHd261fHAdf9S6ZCqUkcNbkjaATgauAA4U1nFhwMnpl2uBc4DfgFMSp8B/hu4JO0/CbgpItYBCyUtAPYDHmvtvGVNfJImAD8F6oFfRcSPynm+zvZ3p/6UlQ3vblx/4IkX+P6l02lq2sB5p0/izJPHc94ld1JfX8cvz5/Mqedex3PzF7Nt34+xvrGJHlt14+fTZvLInPls1a2eOy/7FkceNIY//PH5Cn4r++3l/8SAfr0rHUbVKqUbW4SLge8CfdL6AODtiGhM64uA7dPn7YHXASKiUVJD2n974PGCOguPaVHZurqpqXkpMBEYA5yQmqRd1gNPvEBT0wYAZj23kGFD+gFw+P67MXfBYp6bvxiAVQ3vsmFDsHbdeh6ZMx+A9Y1NPPvi6wwb3K8SoZuVpIRrfAMlzS5YphbUcQywNCLmdHb85Wzx7QcsiIiXASTdRNYk7RLNmYjg9ktOJyK45o5HufaORz+y/SvHHsgdM54CYOcRg4mA//7ZaQzctje33zeHn13/h4/sv03vXkw4ZE8uv+nBzvoK1gJJfP70S5DEyZ87mJM//+lKh1SVSnhWd3lEjGtl28HAsZI+C/QEtiHrIfaT1C21+nYAFqf9FwPDgUWSugF9gRUF5c0Kj2lRORPfxmZpsgjYf9Od0r8A2b8CW9VO92LiP1zEkmUNDNy2N3dccjrzX3mTPz79EgDfOeUoGhs3cMvvZwHQrb6eA/b+OIdP/nfWvv8Bv7nsDJ554TUemvUXAOrr67jqgpP55c0P8uriFRX7Tga/v/LbDBvcj2UrV/O50y9h9MjtOHjsqEqHVXU6oqsbEecA56T6DgPOioiTJN0KfAG4CZgM3JkOmZ7WH0vb74+IkDQduEHShcAwYDTwZFvnrviobkRcERHjImKcuvWqdDhFW7KsAYDlq9Zw14N/YuzuIwE44Zj9Gf/pPZj6f6/ZuO8bb73NH59+iZUN77J23Xpm/HEue+/64T9QF//rCbz02jIuv/HBTvwG1pLmSw2D+vfhmMP24qm5r1Q0nqqkjr2dpQX/QjbQsYDsGt5VqfwqYEAqPxM4GyAi5gK3kPUm7wFOi4imtk5QzsRXcvOzVmzdszu9t+6x8fPhB+zGvJfe4IgDP8EZXz2SE7/zS9auW79x/5mPP8+YUcPo1WMr6uvrOHjsKF5c+CYA/+fUY9imdy/OufC2inwX+9C7a9ex+t33N36+//EX+MTOwyocVfURIBW3FCsiHoyIY9LnlyNiv4gYFRFfTKO1RMT7aX1U2v5ywfEXRMTOEbFrRPy+vfOVs6s7CxgtaSeyhPdlPhyirmmDBvRh2k/+AYD6bvXcds9sZj42jzm3n0uP7t2449LTAZj951c480c30bB6LZfdcD8zr/suRDDj0bnc9+hchg3ux1lTJvDiwjf5n2nZbRNX3vI/XH9nq6PwVkbLVqzmK9+9EoCmxib+fsI4jjyoS4/Hbabaf1ZXEVG+yrOLlheT3c5ydURc0Nb+dVsPjh67Hl+2eKzjrZp1SaVDsBIcvP845syZvUVZq+d2u8SIyT8vat+//GTCnDYGNyqmrPfxRcTdwN3lPIeZdbISu7HVyE9umFlJBNR56nkzyxu3+Mwsd2p9cMOJz8xK42t8ZpY3QjU/EakTn5mVzC0+M8sdX+Mzs3zxNT4zy5vsWd3aznxOfGZWshrPe058ZlY6P7lhZvkid3XNLGea5+OrZU58Zlai2p+Pz4nPzEpW43nPic/MSiQPbphZzvg+PjPLJSc+M8udGs97TnxmVjq3+MwsXzxJgZnlTTYRaW1nPic+MytZXY03+Zz4zKxkNZ73nPjMrDTyJAVmlkc1fomv9cQn6edAtLY9Is4oS0RmVvW68uDG7E6LwsxqhshGdmtZq4kvIq4tXJe0dUS8V/6QzKza1XiDj3bfCizpQEnPAy+k9b0lXVb2yMysOimbj6+YpVoV8zr0i4GjgBUAEfEscGgZYzKzKicVt7Rdh3pKelLSs5LmSvp+Kt9J0hOSFki6WVL3VN4jrS9I20cW1HVOKn9R0lHtxV9M4iMiXt+kqKmY48ys6xHZDczFLO1YBxweEXsD+wATJB0A/Bi4KCJGAauAKWn/KcCqVH5R2g9JY4AvA7sDE4DLJNW3deJiEt/rkg4CQtJWks4C5hVxnJl1UXV1KmppS2TWpNWt0hLA4cB/p/JrgePS50lpnbT9CGX96UnATRGxLiIWAguA/dqMv4jveCpwGrA98AZZZj6tiOPMrAsqtpubGnwDJc0uWKZ+tC7VS3oGWArMAF4C3o6IxrTLIrLcQ/r5OkDa3gAMKCxv4ZgWtXsDc0QsB05qbz8zy48SntVdHhHjWtsYEU3APpL6AXcAu215dO0rZlT345J+K2mZpKWS7pT08c4Izsyqk4pcihURbwMPAAcC/SQ1N8p2ABanz4uB4QBpe1+yQdeN5S0c06Jiuro3ALcAQ4FhwK3AjUUcZ2ZdVEfcziJpUGrpIakX8Ldk4wcPAF9Iu00G7kyfp6d10vb7IyJS+ZfTqO9OwGjgybbOXcyzultHxPUF69Mk/XMRx5lZF5SN6nZIVUOBa9MIbB1wS0Tcle4bvknSD4GngavS/lcB10taAKwkG8klIuZKugV4HmgETktd6Fa19axu//Tx95LOBm4iG3H5EnD35n1PM6t56piJSCPiT8C+LZS/TAujshHxPvDFVuq6ALig2HO31eKbQ5bomr/hNwrPA5xT7EnMrGup5qcyitHWs7o7dWYgZlYbOrCrWzFFzccnaQ9gDNCzuSwiritXUGZW3bpsi6+ZpHOBw8gS393AROARwInPLKdqO+0VdzvLF4AjgDcj4hRgb7L7Z8wshySor1NRS7Uqpqu7NiI2SGqUtA3ZoyXD2zvIzLquLt/VBWanmwyvJBvpXQM8Vs6gzKy61XjeK+pZ3W+mj5dLugfYJt1/Y2Y5JIqacqqqtXUD89i2tkXEU+UJycyqWhGTjFa7tlp8/9nGtuY5szrUvp/YkUefuKSjq7Uy+vkjL1U6BCvBW2vWdUg9XfYaX0R8pjMDMbPaIKC+qyY+M7PWVPGdKkVx4jOzkjnxmVmuZNPK13bmK2YGZkn6iqTvpfUdJbX5Ig8z69rqVNxSrYp5ZO0ysumgT0jrq4FLyxaRmVW9jnivbiUV09XdPyLGSnoaICJWNb/g18zyR0C3as5qRSgm8a1PU0MHZPPkAxvKGpWZVbUaz3tFJb6fkb32bbCkC8hma/m3skZlZlVL6sKPrDWLiF9LmkM2NZWA4yJiXtkjM7OqVeN5r6iJSHcE3gN+W1gWEa+VMzAzq17VPGJbjGK6ur/jw5cO9QR2Al4Edi9jXGZWpQRVPcloMYrp6u5ZuJ5mbflmK7ubWVdX5ffoFaPkJzci4ilJ+5cjGDOrDarxt24Uc43vzILVOmAs8EbZIjKzqpaX10v2KfjcSHbN77byhGNmtaBLJ75043KfiDirk+IxsxpQ65MUtDX1fLeIaJR0cGcGZGbVLXu9ZKWj2DJttfieJLue94yk6cCtwLvNGyPi9jLHZmZVqss/uUF2794KsndsNN/PF4ATn1kOdfXBjcFpRPc5Pkx4zaKsUZlZVavxBl+bia8e6A0t3rDjxGeWW6KuC9/HtyQizu+0SMysJoiOafFJGg5cBwwha0xdERE/ldQfuBkYCbwCHJ/mARXwU+CzZPMHnNz8fm9Jk/lw1qgfRsS1bZ27rbGZ2k7pZlYegm51KmppRyPwnYgYAxwAnCZpDHA2MDMiRgMz0zrARGB0WqYCvwBIifJcYH9gP+BcSdu2deK2Et8R7UVtZvnT3OLb0qnnI2JJc4stIlYD84DtgUlAc4vtWuC49HkScF1kHgf6SRoKHAXMiIiVEbEKmAFMaOvcbb1QfGXbYZtZXpVwO8tASbML1q+IiCs23UnSSGBf4AlgSEQsSZveJOsKQ5YUXy84bFEqa628VX69pJmVrIRrfMsjYlzbdak32WOw/zsi3il8KiQiQlKHD6bW+P3XZtbZRJY4ilnarUvaiizp/brgoYi3UheW9HNpKl8MDC84fIdU1lp5q5z4zKw0yrq6xSxtVpM17a4C5kXEhQWbpgOT0+fJwJ0F5V9L7/o+AGhIXeJ7gfGStk2DGuNTWavc1TWzkmRPbnTITR8HA18F/izpmVT2r8CPgFskTQFeBY5P2+4mu5VlAdntLKdANh4h6QfArLTf+e2NUTjxmVnJOiLtRcQjbVT1V3eVREQAp7VS19XA1cWe24nPzErWlR9ZMzNrgbrufHxmZi1pHtWtZU58ZlayPMzHZ2b2IXXhqefNzFrirq6Z5ZJbfGaWO7Wd9pz4zKxEAurd4jOzvKnxvOfEZ2alEqrxzq4Tn5mVzC0+M8uV7HaW2s58TnxmVpoi3qdR7Zz4zKxkfmTNzHIlm4i00lFsGSc+MyuZR3XNLHdqvKfrxFdO769bz9FTL2bd+kaaGps49oh9OecbR1c6rFxqWLWa2264h3dXvwfAuAP35MC/Gcu90x/ixbkvU19fT/+BfTnuhPH06tWTRa++yfRb/gBAEHzmqAMZs9coAObPe4W773iQiA2M3X8PDj1yv4p9r0pxi68Vkq4GjgGWRsQe5TpPNevRvRt3/uIMem/dg/WNTUz8+oUcedAYPrXnTpUOLXfq6sSEYw9l2PAhrHv/Ay6/8NfsvOsIdt5lBEce/Wnq6+u477cP8/AfZjH+7w5h8NABfOPME6mvr2N1wxou+49p7Lr7x5HgrtvuZ/Kpn2ebfn345UU3sNseOzN4uwGV/oqdpitc4yvn7DLXABPKWH/Vk0TvrXsAsL6xifWNTTU/q0Wt6tO3N8OGDwGgR8/uDBrSn3ca1jBqtxHU12d/DXYYMZR33l4DQPfuW20sb2xsovmx/EWvvUn/gf3oP7Af3brVs+e+u/LCcy91/heqpCJfLVnNI79la/FFxEOSRpar/lrR1LSBw776YxYuWsaULx7KuD1GVjqk3Fu1soEli5axw4jtPlL+1BPPsee+u25cf/3VJfzmxvtoWLWaz580IWv9vb2Gvv36bNxnm769WfTam50We7Wo3pRWnIpf45M0FZgKMHzHHSscTcerr6/j4RvOoWH1e3zln6/k+QVvMGbUsEqHlVvr1n3ATf91FxM/9zf07NljY/n/zHiC+vo69vrkbhvLho8YyrfOnsyyt1Zw+w33MvoTIysQcfXpwPfqVkzFJ1KNiCsiYlxEjBs0cFClwymbvn225pBP7sLMx56vdCi51dTUxE3/dRd7fXI3xuw1emP500/O5cW5C/n7r0xs8VLEoCED6N69O0uXLKdPv940vL1647Z3GtawTd/enRJ/NVGRS7WqeOLrypavWk1DGkVc+/4HPPDkC4weOaTCUeVTRPCbm2YwaEh/Dj7skxvL5897hUfun81JXz+W7t232li+akUDTU0bAHh75TssX7qSfv37sv3w7Vi5bBWrVjTQ2NjEn59+kd12/3inf5+Kq/HMV/Gublf25vJ3+OZ519O0YQMbNgSfO3IsEw7Zs9Jh5dJrC9/g2dnzGDJ0IJf9+zQAjjz6YO6+4wEaG5u49he3A7DDiO049vgjefXlxTw8cxb19fVI4pgvHM7HevcC4Oi/P5zrfnk7GzYEY/ffncFDB1bse1VKrXd1y3k7y43AYcBASYuAcyPiqnKdrxrtMXp7Hvr12ZUOw4ARH9+e8y/69l+V7zKm5VuL9vnUGPb51JgWt+0yZqdWj8uL2k575R3VPaFcdZtZhdV45nNX18xKkl2+q+3M58RnZqXxfHxmlkc1nvec+MysVKr5Ry99H5+ZlUwqbmm/Hl0taamk5wrK+kuaIWl++rltKpekn0laIOlPksYWHDM57T9f0uT2zuvEZ2YlKfbe5SLbhNfw15OZnA3MjIjRwMy0DjARGJ2WqcAvIEuUwLnA/sB+wLnNybI1TnxmVroOynwR8RCwcpPiScC16fO1wHEF5ddF5nGgn6ShwFHAjIhYGRGrgBm0MzOUr/GZWcnKfDvLkIhYkj6/CTQ/57k98HrBfotSWWvlrXLiM7OSlTC2MVDS7IL1KyLiimIPjoiQFKXEVgwnPjMrTWn38S2PiHElnuEtSUMjYknqyi5N5YuB4QX77ZDKFpM9HltY/mBbJ/A1PjMrmYr8bzNNB5pHZicDdxaUfy2N7h4ANKQu8b3AeEnbpkGN8amsVW7xmVlJRMc9udHSZCbAj4BbJE0BXgWOT7vfDXwWWAC8B5wCEBErJf0AmJX2Oz8iNh0w+QgnPjMrWUcNbbQxmckRLewbwGmt1HM1cHWx53XiM7PS1faDG058ZlY6T0RqZrlT22nPic/MNkeNZz4nPjMriSciNbP88USkZpZHNZ73nPjMrFS1PxGpE5+ZlazG854Tn5mVpoRJRquWE5+Zla7GM58Tn5mVzLezmFnu+BqfmeWLoM6Jz8zyp7YznxOfmZWkIycirRQnPjMrWY3nPSc+MyudW3xmljt+ZM3Mcqe2054Tn5mVSJ6WyszyyE9umFn+1Hbec+Izs9LVeN5z4jOzUsmvlzSzfOkKT27UVToAM7PO5hafmZWs1lt8TnxmVjLfzmJm+eIbmM0sb7rC4IYTn5mVzF1dM8sdt/jMLHdqPO858ZnZZqjxzOfEZ2YlEdT8I2uKiErHsJGkZcCrlY6jDAYCyysdhJWkq/6ZjYiIQVtSgaR7yH4/xVgeERO25HzlUFWJr6uSNDsixlU6Diue/8y6Nj+ra2a548RnZrnjxNc5rqh0AFYy/5l1Yb7GZ2a54xafmeWOE5+Z5Y4TXxlJmiDpRUkLJJ1d6XisfZKulrRU0nOVjsXKx4mvTCTVA5cCE4ExwAmSxlQ2KivCNUDV3XBrHcuJr3z2AxZExMsR8QFwEzCpwjFZOyLiIWBlpeOw8nLiK5/tgdcL1helMjOrMCc+M8sdJ77yWQwML1jfIZWZWYU58ZXPLGC0pJ0kdQe+DEyvcExmhhNf2UREI3A6cC8wD7glIuZWNiprj6QbgceAXSUtkjSl0jFZx/Mja2aWO27xmVnuOPGZWe448ZlZ7jjxmVnuOPGZWe448dUQSU2SnpH0nKRbJW29BXVdI+kL6fOv2ppAQdJhkg7ajHO8Iumv3sbVWvkm+6wp8VznSTqr1Bgtn5z4asvaiNgnIvYAPgBOLdwoabPekxwRX4+I59vY5TCg5MRnVq2c+GrXw8Co1Bp7WNJ04HlJ9ZL+XdIsSX+S9A0AZS5J8wP+ARjcXJGkByWNS58nSHpK0rOSZkoaSZZgv51am4dIGiTptnSOWZIOTscOkHSfpLmSfkX27uk2SfqNpDnpmKmbbLsolc+UNCiV7SzpnnTMw5J265DfpuXKZrUQrLJSy24icE8qGgvsERELU/JoiIhPSeoBPCrpPmBfYFeyuQGHAM8DV29S7yDgSuDQVFf/iFgp6XJgTUT8R9rvBuCiiHhE0o5kT6d8AjgXeCQizpd0NFDMUw//K52jFzBL0m0RsQL4GDA7Ir4t6Xup7tPJXgJ0akTMl7Q/cBlw+Gb8Gi3HnPhqSy9Jz6TPDwNXkXVBn4yIhal8PLBX8/U7oC8wGjgUuDEimoA3JN3fQv0HAA811xURrc1LdyQwRtrYoNtGUu90js+nY38naVUR3+kMSZ9Ln4enWFcAG4CbU/k04PZ0joOAWwvO3aOIc5h9hBNfbVkbEfsUFqQE8G5hEfCtiLh3k/0+24Fx1AEHRMT7LcRSNEmHkSXRAyPiPUkPAj1b2T3Sed/e9HdgVipf4+t67gX+UdJWAJJ2kfQx4CHgS+ka4FDgMy0c+zhwqKSd0rH9U/lqoE/BfvcB32pekbRP+vgQcGIqmwhs206sfYFVKentRtbibFYHNLdaTyTrQr8DLJT0xXQOSdq7nXOY/RUnvq7nV2TX755KL8z5JVnL/g5gftp2HdkMJB8REcuAqWTdymf5sKv5W+BzzYMbwBnAuDR48jwfji5/nyxxziXr8r7WTqz3AN0kzQN+RJZ4m70L7Je+w+HA+an8JGBKim8uns7fNoNnZzGz3HGLz8xyx4nPzHLHic/McseJz8xyx4nPzHLHic/McseJz8xy5/8DKCrIBrOC2ZMAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#make some appropriate display labels here\n",
    "\n",
    "plot_model(model, test_X, test_y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "15e9e191",
   "metadata": {},
   "outputs": [],
   "source": [
    "#wrapper function\n",
    "\n",
    "def RF_wrapper(dataframe):\n",
    "    \"\"\"\n",
    "    Takes dataframe and runs it through kNN model.\n",
    "    \n",
    "    Params\n",
    "    ----------\n",
    "    dataframe: Pandas dataframe \n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    -Target feature predictions\n",
    "    -Parity plot\n",
    "    \"\"\"\n",
    "    \n",
    "    assert 'pandas.core.frame.DataFrame' in str(type(dataframe))\n",
    "\n",
    "    #user inputs target feature\n",
    "    target = 'protein_match'\n",
    "    \n",
    "    #define input features\n",
    "    input_features = [columns for columns in dataframe]\n",
    "    \n",
    "    input_features.remove(target)\n",
    "\n",
    "    #train the model based off data split\n",
    "    model, dev_X, dev_y, test_X, test_y = train_model(\n",
    "                                        dataframe, columns=input_features, \n",
    "                                        target=target\n",
    "                                                   )\n",
    "    \n",
    "    #test the model and return predictions\n",
    "    preds = evaluate_model(model, test_X, test_y)\n",
    "\n",
    "    #plot the results of the model\n",
    "    plot_model(model, test_X, test_y)\n",
    "    \n",
    "    return preds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "cbbd8fa8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([False, False, False, ..., False,  True, False])"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAT4AAAEWCAYAAAD/x/trAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Il7ecAAAACXBIWXMAAAsTAAALEwEAmpwYAAAgI0lEQVR4nO3debgV1Znv8e/vHARUUGQUFMEIDmgcaFpUEmOMjRBtUZOYOCTotZu2W5PcGG9a+/YNiYn3ydCJSVpNotFWNGo0akRDVOLQDokKjhGHQBwiiDIjIKIH3v6j1sEtnmEXnH323qd+H596zq5VVavefYDXtWpVrVJEYGZWJA3VDsDMrLM58ZlZ4TjxmVnhOPGZWeE48ZlZ4TjxmVnhOPF1MZK2lnSbpJWSbtyCek6WdFdHxlYNkn4naXK147Da4sRXJZJOkjRb0mpJC9M/0I90QNWfBgYB/SLiM5tbSUT8MiLGd0A87yPpMEkh6ZZNyvdL5feVWc83JF3T3n4RMTEirtrMcK2LcuKrAklnAz8C/j9ZktoFuASY1AHVDwP+HBFNHVBXpSwGDpbUr6RsMvDnjjqBMv77bS2LCC+duADbA6uBz7SxTw+yxPhaWn4E9EjbDgPmA18FFgELgdPStm8C7wDvpnOcDnwDuKak7uFAAN3S+qnAi8Aq4CXg5JLyB0uOOwSYBaxMPw8p2XYf8C3goVTPXUD/Vr5bc/w/A85MZY3AAuDrwH0l+/4YeBV4E3gM+Ggqn7DJ93yqJI4LUhxrgRGp7B/S9p8CN5XU/13gbkDV/nvhpXMX/x+x8x0M9ARuaWOf/wscBOwP7AccCPx7yfYdyRLoTmTJ7WJJO0TEVLJW5K8ioldEXN5WIJK2BX4CTIyI3mTJ7ckW9usL/Dbt2w/4IfDbTVpsJwGnAQOB7sA5bZ0bmAZ8IX0+EniGLMmXmkX2O+gLXAvcKKlnRNyxyffcr+SYzwNTgN7AK5vU91Xgw5JOlfRRst/d5Ijwc5sF48TX+foBS6LtrujJwPkRsSgiFpO15D5fsv3dtP3diJhB1urZYzPj2QDsI2nriFgYEXNa2OcoYG5EXB0RTRFxHfA88Pcl+/xXRPw5ItYCN5AlrFZFxB+AvpL2IEuA01rY55qIWJrO+QOylnB73/PKiJiTjnl3k/reIvs9/hC4BvhiRMxvpz7rgpz4Ot9SoL+kbm3sM4T3t1ZeSWUb69gkcb4F9MobSESsAT4LnAEslPRbSXuWEU9zTDuVrL++GfFcDZwFfJwWWsCSzpH0XBqhXkHWyu3fTp2vtrUxIh4h69qLLEFbATnxdb4/AuuAY9vY5zWyQYpmu/DBbmC51gDblKzvWLoxIu6MiL8DBpO14i4rI57mmBZsZkzNrgb+BZiRWmMbpa7o14ATgB0iog/Z9UU1h95KnW12WyWdSdZyfC3VbwXkxNfJImIl2UX8iyUdK2kbSVtJmijpe2m364B/lzRAUv+0f7u3brTiSeBQSbtI2h44r3mDpEGSJqVrfevIuswbWqhjBrB7ugWnm6TPAqOA2zczJgAi4iXgY2TXNDfVG2giGwHuJunrwHYl298AhucZuZW0O/Bt4BSyLu/XJO2/edFbPXPiq4J0vepssgGLxWTds7OA36Rdvg3MBp4G/gQ8nso251wzgV+luh7j/cmqIcXxGrCMLAn9cwt1LAWOJhscWErWUjo6IpZsTkyb1P1gRLTUmr0TuIPsFpdXgLd5fze2+ebspZIeb+886dLCNcB3I+KpiJgL/BtwtaQeW/IdrP7IA1pmVjRu8ZlZ4TjxmVnhOPGZWeE48ZlZ4bR1E22nU7etQ917VzsMy+GAvXapdgiWwyuvvMySJUvU/p6ta9xuWETT2rL2jbWL74yICVtyvkqorcTXvTc99jih2mFYDg89clG1Q7Acxo0ds8V1RNPasv+dvv3kxe09aVMVNZX4zKweCOp8xi8nPjPLR0BDY7Wj2CJOfGaWn7boMmHVOfGZWU7u6ppZEbnFZ2aFIuq+xVff0ZtZFShr8ZWztFeT9LKkP0l6UtLsVNZX0kxJc9PPHVK5JP1E0jxJT0saXVLP5LT/3HJeJ+rEZ2b5NTSWt5Tn4xGxf0Q032R4LnB3RIwkexnUual8IjAyLVPIXh7V/E6YqcBYsvfTTG1Olq2Gn+OrmpmxcXCjnGXzTAKa34V8Fe/NVj4JmBaZh4E+kgaTvaxqZkQsi4jlwEyyN/G1yonPzPIRebq6/SXNLlmmbFJbAHdJeqxk26CIWJg+v0727mnI3vFSOhnt/FTWWnmrPLhhZvmV35pbUtKFbclHImKBpIHATEnPl26MiJDU4bMlu8VnZjl1XFc3Ihakn4vI3rR3IPBG6sKSfi5Kuy8AhpYcvnMqa628VU58ZpaPgMbG8pa2qpG2ldS7+TMwnuzF8tOB5pHZycCt6fN04AtpdPcgYGXqEt8JjJe0QxrUGJ/KWuWurpnl1zE3MA8CblFWVzfg2oi4Q9Is4AZJp5O9aKp5KpgZwCeBeWTvbj4NICKWSfoWMCvtd35ELGvrxE58ZpZTxzyyFhEvAvu1UL4U+EQL5QGc2UpdVwBXlHtuJz4zy8+PrJlZ4dT5I2tOfGaWT5mPo9UyJz4zy88TkZpZsXg+PjMrInd1zaxQusB8fE58ZpaTu7pmVkQe3DCzwvE1PjMrFLmra2ZF5BafmRWNnPjMrEiymeed+MysSCTU4MRnZgXjFp+ZFY4Tn5kVjhOfmRWL0lLHnPjMLBcht/jMrHgaGvzkhpkVjFt8ZlYsvsZnZkXkFp+ZFYoHN8yskPzImpkVi9zVNbMCcuIzs8Jx4jOzQvHghpkVU33nPSc+M8tJ9f/IWn1Hb2ZVIamspcy6GiU9Ien2tL6rpEckzZP0K0ndU3mPtD4vbR9eUsd5qfwFSUe2d04nPjPLT2Uu5fky8FzJ+neBCyNiBLAcOD2Vnw4sT+UXpv2QNAr4HLA3MAG4RFKbbzx3V3czPXXrN1n91jrWb9hAU9MGDp/8Pc7/0rEc+dF9ePfd9bw0fwlnnn8Nb65eC8DeI4bww/NOpHevnsSG4PDJ32PdO03st+dQLpn6eXr22IqZD83h3B/8usrfrNj2Pebr9NqmB40NDXTr1sC90/612iHVpI4a3JC0M3AUcAFwtrKKDwdOSrtcBXwD+CkwKX0G+DVwUdp/EnB9RKwDXpI0DzgQ+GNr561o4pM0Afgx0Aj8IiK+U8nzdba/P+PHLFu5ZuP6vY88zzcvns769Rv4xlmTOPvU8XzjoltpbGzg5+dP5oyp03hm7gJ22H5b3m1aD8APzv0sX77gWmY/8zI3/vifOeKQUfz+D89W6ysZcNvPvky/Pr2qHUbNytONBfpLml2yfmlEXFqy/iPga0DvtN4PWBERTWl9PrBT+rwT8CpARDRJWpn23wl4uKTO0mNaVLGubmpqXgxMBEYBJ6YmaZd17yPPs379BgBmPfMSQwb1AeDwsXsyZ94Cnpm7AIDlK9ewYUMwqN929N62J7OfeRmA63/7KEd9bN9qhG6WS45rfEsiYkzJcmlJHUcDiyLisc6Ov5ItvgOBeRHxIoCk68mapF2iORMR3HzRWUQEV97yEFfd8tD7tp9yzMHcMvNxAHYbNpAI+PVPzqT/Dr24+a7H+MnVv2fwwD68tmjFxmNeW7SCwQP6dOK3sE1J4vizLkISpx43jlOP/0i1Q6pJHfSs7jjgGEmfBHoC25H1EPtI6pZafTsDC9L+C4ChwHxJ3YDtgaUl5c1Kj2lRJRPfxmZpMh8Yu+lOkqYAUwDYqn66FxP/8UIWLl5J/x16cctFZzH35df5wxN/AeCrpx1JU9MGbvjdLAC6NTZy0H4f4vDJ32ft2+/wm0u+xJPP/5U3V79dza9gLfjdZV9hyMA+LF62iuPOuoiRw3dk3OgR1Q6r5nTENb6IOA84L9V3GHBORJws6Ubg08D1wGTg1nTI9LT+x7T9nogISdOBayX9EBgCjAQebevcVR/VjYhLm5vB6rZ1tcMp28LFKwFYsnw1t9/3NKP3Hg7AiUePZfxH9mHK/7ty476vvbGCPzzxF5atXMPade8y8w9z2G+PoSxctIIhA/ts3G/IwD4sXLyi876EfUDzn8eAvr05+rB9eXzOy1WNpyapY29nacG/kg10zCO7hnd5Kr8c6JfKzwbOBYiIOcANZL3JO4AzI2J9WyeoZOLL3fysF9v07E6vbXps/Hz4QXvy3F9e4xMH78WXPn8EJ33156xd9+7G/e9++FlGjRjC1j22orGxgXGjR/DCS6/zxtI3WbXmbcbsMxyAzx11IDP+++lqfCUD1qxdx6o1b2/8fM/Dz7PXbkOqHFXtESCVt5QrIu6LiKPT5xcj4sCIGBERn0mjtUTE22l9RNr+YsnxF0TEbhGxR0T8rr3zVbKrOwsYKWlXsoT3Od4boq5rA/r15prv/SMAjd0auemO2dz9x+d47Oap9OjejVsuPguA2X96mbO/cz0rV63lkmvv4e5pX4MIZj40h7semgPAOd+9gUumnkLPHlvx+z88y0yP6FbN4qWrOOVrlwGwvmk9n5owhiMO6dLjcZup/p/VVURUrvLsouWPyG5nuSIiLmhr/4ZtBkaPPU6oWDzW8ZbPuqjaIVgO48aO4bHHZm9R1uq54+4xbPJ/lrXvn7834bGIGLMl56uEit7HFxEzgBmVPIeZdbKc3dha5Cc3zCwXAQ2eet7MisYtPjMrnHof3HDiM7N8fI3PzIpGqO4nInXiM7Pc3OIzs8LxNT4zKxZf4zOzosme1a3vzOfEZ2a51Xnec+Izs/z85IaZFYvc1TWzgmmej6+eOfGZWU71Px+fE5+Z5Vbnec+Jz8xykgc3zKxgfB+fmRWSE5+ZFU6d5z0nPjPLzy0+MysWT1JgZkWTTURa35nPic/Mcmuo8yafE5+Z5Vbnec+Jz8zykScpMLMiqvNLfK0nPkn/CURr2yPiSxWJyMxqXlce3JjdaVGYWd0Q2chuPWs18UXEVaXrkraJiLcqH5KZ1bo6b/DR7luBJR0s6Vng+bS+n6RLKh6ZmdUmZfPxlbPUqnJeh/4j4EhgKUBEPAUcWsGYzKzGSeUtbdehnpIelfSUpDmSvpnKd5X0iKR5kn4lqXsq75HW56Xtw0vqOi+VvyDpyPbiLyfxERGvblK0vpzjzKzrEdkNzOUs7VgHHB4R+wH7AxMkHQR8F7gwIkYAy4HT0/6nA8tT+YVpPySNAj4H7A1MAC6R1NjWictJfK9KOgQISVtJOgd4rozjzKyLamhQWUtbIrM6rW6VlgAOB36dyq8Cjk2fJ6V10vZPKOtPTwKuj4h1EfESMA84sM34y/iOZwBnAjsBr5Fl5jPLOM7MuqByu7mpwddf0uySZcr761KjpCeBRcBM4C/AiohoSrvMJ8s9pJ+vAqTtK4F+peUtHNOidm9gjoglwMnt7WdmxZHjWd0lETGmtY0RsR7YX1If4BZgzy2Prn3ljOp+SNJtkhZLWiTpVkkf6ozgzKw2qcylXBGxArgXOBjoI6m5UbYzsCB9XgAMBUjbtycbdN1Y3sIxLSqnq3stcAMwGBgC3AhcV8ZxZtZFdcTtLJIGpJYekrYG/o5s/OBe4NNpt8nArenz9LRO2n5PREQq/1wa9d0VGAk82ta5y3lWd5uIuLpk/RpJ/6eM48ysC8pGdTukqsHAVWkEtgG4ISJuT/cNXy/p28ATwOVp/8uBqyXNA5aRjeQSEXMk3QA8CzQBZ6YudKvaela3b/r4O0nnAteTjbh8Fpixed/TzOqeOmYi0oh4GjighfIXaWFUNiLeBj7TSl0XABeUe+62WnyPkSW65m/4T6XnAc4r9yRm1rXU8lMZ5WjrWd1dOzMQM6sPHdjVrZqy5uOTtA8wCujZXBYR0yoVlJnVti7b4msmaSpwGFnimwFMBB4EnPjMCqq+0155t7N8GvgE8HpEnAbsR3b/jJkVkASNDSprqVXldHXXRsQGSU2StiN7tGRoeweZWdfV5bu6wOx0k+FlZCO9q4E/VjIoM6ttdZ73ynpW91/Sx59JugPYLt1/Y2YFJMqacqqmtXUD8+i2tkXE45UJycxqWhmTjNa6tlp8P2hjW/OcWR3qgL124aFHLuroaq2C/vPBv1Q7BMvhjdXrOqSeLnuNLyI+3pmBmFl9ENDYVROfmVlravhOlbI48ZlZbk58ZlYo2bTy9Z35ypmBWZJOkfT1tL6LpDZf5GFmXVuDyltqVTmPrF1CNh30iWl9FXBxxSIys5rXEe/VraZyurpjI2K0pCcAImJ58wt+zax4BHSr5axWhnIS37tpauiAbJ58YENFozKzmlbnea+sxPcTste+DZR0AdlsLf9e0ajMrGZJXfiRtWYR8UtJj5FNTSXg2Ih4ruKRmVnNqvO8V9ZEpLsAbwG3lZZFxF8rGZiZ1a5aHrEtRzld3d/y3kuHegK7Ai8Ae1cwLjOrUYKanmS0HOV0dT9cup5mbfmXVnY3s66uxu/RK0fuJzci4nFJYysRjJnVB9X5WzfKucZ3dslqAzAaeK1iEZlZTSvK6yV7l3xuIrvmd1NlwjGzetClE1+6cbl3RJzTSfGYWR2o90kK2pp6vltENEka15kBmVlty14vWe0otkxbLb5Hya7nPSlpOnAjsKZ5Y0TcXOHYzKxGdfknN8ju3VtK9o6N5vv5AnDiMyugrj64MTCN6D7DewmvWVQ0KjOraXXe4Gsz8TUCvaDFG3ac+MwKSzR04fv4FkbE+Z0WiZnVBdExLT5JQ4FpwCCyxtSlEfFjSX2BXwHDgZeBE9I8oAJ+DHySbP6AU5vf7y1pMu/NGvXtiLiqrXO3NTZT3yndzCpD0K1BZS3taAK+GhGjgIOAMyWNAs4F7o6IkcDdaR1gIjAyLVOAnwKkRDkVGAscCEyVtENbJ24r8X2ivajNrHiaW3xbOvV8RCxsbrFFxCrgOWAnYBLQ3GK7Cjg2fZ4ETIvMw0AfSYOBI4GZEbEsIpYDM4EJbZ27rReKL2s7bDMrqhy3s/SXNLtk/dKIuHTTnSQNBw4AHgEGRcTCtOl1sq4wZEnx1ZLD5qey1spb5ddLmlluOa7xLYmIMW3XpV5kj8H+74h4s/SpkIgISR0+mFrn91+bWWcTWeIoZ2m3LmkrsqT3y5KHIt5IXVjSz0WpfAEwtOTwnVNZa+WtcuIzs3yUdXXLWdqsJmvaXQ48FxE/LNk0HZicPk8Gbi0p/0J61/dBwMrUJb4TGC9phzSoMT6VtcpdXTPLJXtyo0Nu+hgHfB74k6QnU9m/Ad8BbpB0OvAKcELaNoPsVpZ5ZLeznAbZeISkbwGz0n7ntzdG4cRnZrl1RNqLiAfbqOoDd5VERABntlLXFcAV5Z7bic/McuvKj6yZmbVAXXc+PjOzljSP6tYzJz4zy60I8/GZmb1HXXjqeTOzlrira2aF5BafmRVOfac9Jz4zy0lAo1t8ZlY0dZ73nPjMLC+hOu/sOvGZWW5u8ZlZoWS3s9R35nPiM7N8ynifRq1z4jOz3PzImpkVSjYRabWj2DJOfGaWm0d1zaxw6ryn68RXafse83V6bdODxoYGunVr4N5p/1rtkApp5fJV3HTtHaxZ9RYAYw7+MAd/bDR3Tr+fF+a8SGNjI337b8+xJ45n6617Mv+V15l+w+8BCIKPH3kwo/YdAcDc515mxi33EbGB0WP34dAjDqza96oWt/haIekK4GhgUUTsU6nz1IPbfvZl+vXpVe0wCq2hQUw45lCGDB3Eurff4Wc//CW77TGM3XYfxhFHfYTGxgbuuu0BHvj9LMb//UcZOLgf/3T2STQ2NrBq5Wou+Y9r2GPvDyHB7Tfdw+Qzjme7Pr35+YXXsuc+uzFwx37V/oqdpitc46vk7DJXAhMqWL9Z2Xpv34shQwcB0KNndwYM6subK1czYs9hNDZm/wx2HjaYN1esBqB79602ljc1raf5sfz5f32dvv370Ld/H7p1a+TDB+zB88/8pfO/UDWV+WrJWh75rViLLyLulzS8UvXXC0kcf9ZFSOLU48Zx6vEfqXZIhbd82UoWzl/MzsN2fF/54488w4cP2GPj+quvLOQ3193FyuWrOP7kCVnrb8Vqtu/Te+M+223fi/l/fb3TYq8VtZvSylP1a3ySpgBTAIbuskuVo+l4v7vsKwwZ2IfFy1Zx3FkXMXL4jowbPaLaYRXWunXvcP1/3c7E4z5Gz549Npb/98xHaGxsYN+/2XNj2dBhg/niuZNZ/MZSbr72TkbuNbwKEdeeDnyvbtVUfSLViLg0IsZExJgB/QdUO5wON2RgHwAG9O3N0Yfty+NzXq5qPEW2fv16rv+v29n3b/Zk1L4jN5Y/8egcXpjzEp86ZWKLE2wOGNSP7t27s2jhEnr36cXKFas2bntz5Wq22754129V5lKrqp74urI1a9exas3bGz/f8/Dz7LXbkCpHVUwRwW+un8mAQX0Zd9jfbCyf+9zLPHjPbE7+h2Po3n2rjeXLl65k/foNAKxY9iZLFi2jT9/t2WnojixbvJzlS1fS1LSePz3xAnvu/aFO/z5VV+eZr+pd3a5s8dJVnPK1ywBY37SeT00YwxGHjKpyVMX015de46nZzzFocH8u+f41ABxx1Dhm3HIvTU3rueqnNwOw87AdOeaEI3jlxQU8cPcsGhsbkcTRnz6cbXttDcBRnzqcaT+/mQ0bgtFj92bg4P5V+17VUu9d3UreznIdcBjQX9J8YGpEXF6p89Wi4Tv358Frz6t2GAYM+9BOnH/hVz5QvvuoXVvcf/+/HcX+f9vy/6R2H7Vrq8cVRX2nvcqO6p5YqbrNrMrqPPO5q2tmuWSX7+o78znxmVk+no/PzIqozvOeE5+Z5aW6f6G47+Mzs9yk8pb269EVkhZJeqakrK+kmZLmpp87pHJJ+omkeZKeljS65JjJaf+5kia3d14nPjPLpdx7l8tsE17JByczORe4OyJGAnendYCJwMi0TAF+ClmiBKYCY4EDganNybI1Tnxmll8HZb6IuB9YtknxJOCq9Pkq4NiS8mmReRjoI2kwcCQwMyKWRcRyYCbtzAzla3xmlluFb2cZFBEL0+fXgUHp807AqyX7zU9lrZW3yonPzHLLMbbRX9LskvVLI+LScg+OiJAUeWIrhxOfmeWT7z6+JRExJucZ3pA0OCIWpq7solS+ABhast/OqWwB2eOxpeX3tXUCX+Mzs9xU5n+baTrQPDI7Gbi1pPwLaXT3IGBl6hLfCYyXtEMa1BifylrlFp+Z5SI67smNliYzAb4D3CDpdOAV4IS0+wzgk8A84C3gNICIWCbpW8CstN/5EbHpgMn7OPGZWW4dNbTRxmQmn2hh3wDObKWeK4Aryj2vE5+Z5VffD2448ZlZfp6I1MwKp77TnhOfmW2OOs98TnxmlosnIjWz4vFEpGZWRHWe95z4zCyv+p+I1InPzHKr87znxGdm+eSYZLRmOfGZWX51nvmc+MwsN9/OYmaF42t8ZlYsggYnPjMrnvrOfE58ZpZLR05EWi1OfGaWW53nPSc+M8vPLT4zKxw/smZmhVPfac+Jz8xykqelMrMi8pMbZlY89Z33nPjMLL86z3tOfGaWl/x6STMrlq7w5EZDtQMwM+tsbvGZWW713uJz4jOz3Hw7i5kVi29gNrOi6QqDG058Zpabu7pmVjhu8ZlZ4dR53nPiM7PNUOeZz4nPzHIR1P0ja4qIasewkaTFwCvVjqMC+gNLqh2E5dJV/8yGRcSALalA0h1kv59yLImICVtyvkqoqcTXVUmaHRFjqh2Hlc9/Zl2bn9U1s8Jx4jOzwnHi6xyXVjsAy81/Zl2Yr/GZWeG4xWdmhePEZ2aF48RXQZImSHpB0jxJ51Y7HmufpCskLZL0TLVjscpx4qsQSY3AxcBEYBRwoqRR1Y3KynAlUHM33FrHcuKrnAOBeRHxYkS8A1wPTKpyTNaOiLgfWFbtOKyynPgqZyfg1ZL1+anMzKrMic/MCseJr3IWAENL1ndOZWZWZU58lTMLGClpV0ndgc8B06sck5nhxFcxEdEEnAXcCTwH3BARc6oblbVH0nXAH4E9JM2XdHq1Y7KO50fWzKxw3OIzs8Jx4jOzwnHiM7PCceIzs8Jx4jOzwnHiqyOS1kt6UtIzkm6UtM0W1HWlpE+nz79oawIFSYdJOmQzzvGypA+8jau18k32WZ3zXN+QdE7eGK2YnPjqy9qI2D8i9gHeAc4o3Shps96THBH/EBHPtrHLYUDuxGdWq5z46tcDwIjUGntA0nTgWUmNkr4vaZakpyX9E4AyF6X5AX8PDGyuSNJ9ksakzxMkPS7pKUl3SxpOlmC/klqbH5U0QNJN6RyzJI1Lx/aTdJekOZJ+Qfbu6TZJ+o2kx9IxUzbZdmEqv1vSgFS2m6Q70jEPSNqzQ36bViib1UKw6kotu4nAHaloNLBPRLyUksfKiPhbST2AhyTdBRwA7EE2N+Ag4Fngik3qHQBcBhya6uobEcsk/QxYHRH/kfa7FrgwIh6UtAvZ0yl7AVOBByPifElHAeU89fC/0jm2BmZJuikilgLbArMj4iuSvp7qPovsJUBnRMRcSWOBS4DDN+PXaAXmxFdftpb0ZPr8AHA5WRf00Yh4KZWPB/Ztvn4HbA+MBA4FrouI9cBrku5pof6DgPub64qI1ualOwIYJW1s0G0nqVc6x/Hp2N9KWl7Gd/qSpOPS56Ep1qXABuBXqfwa4OZ0jkOAG0vO3aOMc5i9jxNffVkbEfuXFqQEsKa0CPhiRNy5yX6f7MA4GoCDIuLtFmIpm6TDyJLowRHxlqT7gJ6t7B7pvCs2/R2Y5eVrfF3PncA/S9oKQNLukrYF7gc+m64BDgY+3sKxDwOHSto1Hds3la8CepfsdxfwxeYVSfunj/cDJ6WyicAO7cS6PbA8Jb09yVqczRqA5lbrSWRd6DeBlyR9Jp1DkvZr5xxmH+DE1/X8guz63ePphTk/J2vZ3wLMTdumkc1A8j4RsRiYQtatfIr3upq3Acc1D24AXwLGpMGTZ3lvdPmbZIlzDlmX96/txHoH0E3Sc8B3yBJvszXAgek7HA6cn8pPBk5P8c3B0/nbZvDsLGZWOG7xmVnhOPGZWeE48ZlZ4TjxmVnhOPGZWeE48ZlZ4TjxmVnh/A9z0LMjUrlUpQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "RF_wrapper(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "37c0e2b2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "...\n",
      "----------------------------------------------------------------------\n",
      "Ran 3 tests in 6.192s\n",
      "\n",
      "OK\n"
     ]
    }
   ],
   "source": [
    "class TestWrapper(unittest.TestCase):\n",
    "    \n",
    "    def test_wrapper_input(self):\n",
    "        #test that input data type is correct\n",
    "        try:\n",
    "            RF_wrapper([1,2,3])\n",
    "            self.assertTrue(False)\n",
    "        except AssertionError:\n",
    "            self.assertTrue(True)\n",
    "\n",
    "    def test_wrapper_output(self):\n",
    "        model, _, _, test_X, test_y = train_model(\n",
    "            df, \n",
    "            columns=input_features, target='protein_match'\n",
    "        )\n",
    "        # assert output type is correct\n",
    "        output = evaluate_model(model, test_X, test_y)\n",
    "        self.assertIsInstance(output, np.ndarray)\n",
    "        \n",
    "    def test_output_dimension(self):\n",
    "        model, _, _, test_X, test_y = train_model(\n",
    "            df, \n",
    "            columns=input_features, target='protein_match'\n",
    "        )\n",
    "        # want to check that the number of predictions is equal to the number of test examples\n",
    "        preds = evaluate_model(model, test_X, test_y)\n",
    "        self.assertEqual(len(test_y), len(preds))\n",
    "        \n",
    "\n",
    "suite = unittest.TestLoader().loadTestsFromTestCase(TestWrapper)\n",
    "_ = unittest.TextTestRunner().run(suite)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02c53a58",
   "metadata": {},
   "source": [
    "**IGNORE BELOW**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "13d52919",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #plot empirical distribution of scaled_local_query_percent_id\n",
    "\n",
    "# target = df['protein_match']\n",
    "# #create histplot\n",
    "# fig, ax = plt.subplots()\n",
    "\n",
    "# sns.histplot(data = df, x = target)\n",
    "\n",
    "# ax.set_title('Sample Histogram', fontsize=16)\n",
    "# ax.set_xlabel('Class', fontsize=14)\n",
    "# ax.set_ylabel('Count', fontsize=14)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e3d1950",
   "metadata": {},
   "source": [
    "Things I've tried to improve model: \n",
    "\n",
    "1. drop bit scores over 1000\n",
    "2. Switch to 85/15 train/test split\n",
    "3. Lasso regression - not great\n",
    "4. Ridge regression - R2 = 0.87, about the same as normal Linear regression\n",
    "5. KNN regression (n_neighbors optimized at 8) gives best result, R2 = ~0.93 --> \n",
    "*got it up to .942 by removing a few features\n",
    "6. Decision tree regressor was slightly worse than KNN\n",
    "7. Input DT classifier and RF classifier. RF classifier has best performance\n",
    "8. Changed target to binary protein function match"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5220b26b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
